// @kbn/security-solution-plugin/public/investigations/routes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.routes = void 0;
var _react = _interopRequireDefault(require("react"));
var _securitySolutionNavigation = require("@kbn/security-solution-navigation");
var _constants = require("../../common/constants");
var _investigations = require("./investigations");
var _plugin_template_wrapper = require("../common/components/plugin_template_wrapper");
var _security_route_page_wrapper = require("../common/components/security_route_page_wrapper");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const InvestigationsPage = /*#__PURE__*/_react.default.memo(function InvestigationsPage() {
  return /*#__PURE__*/_react.default.createElement(_plugin_template_wrapper.PluginTemplateWrapper, null, /*#__PURE__*/_react.default.createElement(_security_route_page_wrapper.SecurityRoutePageWrapper, {
    pageName: _securitySolutionNavigation.SecurityPageName.investigations,
    redirectOnMissing: true
  }, /*#__PURE__*/_react.default.createElement(_investigations.Investigations, null)));
});
const routes = exports.routes = [{
  path: _constants.INVESTIGATIONS_PATH,
  component: InvestigationsPage
}];
