// @kbn/fleet-plugin/server/routes/fleet_server_hosts/handler.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.putFleetServerHostHandler = exports.postFleetServerHost = exports.getFleetServerHostHandler = exports.getAllFleetServerHostsHandler = exports.deleteFleetServerHostHandler = void 0;
var _server = require("@kbn/core/server");
var _lodash = require("lodash");
var _constants = require("../../constants");
var _errors = require("../../errors");
var _services = require("../../services");
var _fleet_server_host = require("../../services/fleet_server_host");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

async function checkFleetServerHostsWriteAPIsAllowed(soClient, hostUrls) {
  const cloudSetup = _services.appContextService.getCloud();
  if (!(cloudSetup !== null && cloudSetup !== void 0 && cloudSetup.isServerlessEnabled)) {
    return;
  }

  // Fleet Server hosts must have the default host URL in serverless.
  const serverlessDefaultFleetServerHost = await (0, _fleet_server_host.getFleetServerHost)(soClient, _constants.SERVERLESS_DEFAULT_FLEET_SERVER_HOST_ID);
  if (!(0, _lodash.isEqual)(hostUrls, serverlessDefaultFleetServerHost.host_urls)) {
    throw new _errors.FleetServerHostUnauthorizedError(`Fleet server host must have default URL in serverless: ${serverlessDefaultFleetServerHost.host_urls}`);
  }
}
const postFleetServerHost = async (context, request, response) => {
  const coreContext = await context.core;
  const soClient = coreContext.savedObjects.client;
  const esClient = coreContext.elasticsearch.client.asInternalUser;
  try {
    // In serverless, allow create fleet server host if host url is same as default.
    await checkFleetServerHostsWriteAPIsAllowed(soClient, request.body.host_urls);
    const {
      id,
      ...data
    } = request.body;
    const FleetServerHost = await (0, _fleet_server_host.createFleetServerHost)(soClient, {
      ...data,
      is_preconfigured: false
    }, {
      id
    });
    if (FleetServerHost.is_default) {
      await _services.agentPolicyService.bumpAllAgentPolicies(esClient);
    }
    const body = {
      item: FleetServerHost
    };
    return response.ok({
      body
    });
  } catch (error) {
    return (0, _errors.defaultFleetErrorHandler)({
      error,
      response
    });
  }
};
exports.postFleetServerHost = postFleetServerHost;
const getFleetServerHostHandler = async (context, request, response) => {
  const soClient = (await context.core).savedObjects.client;
  try {
    const item = await (0, _fleet_server_host.getFleetServerHost)(soClient, request.params.itemId);
    const body = {
      item
    };
    return response.ok({
      body
    });
  } catch (error) {
    if (_server.SavedObjectsErrorHelpers.isNotFoundError(error)) {
      return response.notFound({
        body: {
          message: `Fleet server ${request.params.itemId} not found`
        }
      });
    }
    return (0, _errors.defaultFleetErrorHandler)({
      error,
      response
    });
  }
};
exports.getFleetServerHostHandler = getFleetServerHostHandler;
const deleteFleetServerHostHandler = async (context, request, response) => {
  try {
    const coreContext = await context.core;
    const soClient = coreContext.savedObjects.client;
    const esClient = coreContext.elasticsearch.client.asInternalUser;
    await (0, _fleet_server_host.deleteFleetServerHost)(soClient, esClient, request.params.itemId);
    const body = {
      id: request.params.itemId
    };
    return response.ok({
      body
    });
  } catch (error) {
    if (_server.SavedObjectsErrorHelpers.isNotFoundError(error)) {
      return response.notFound({
        body: {
          message: `Fleet server ${request.params.itemId} not found`
        }
      });
    }
    return (0, _errors.defaultFleetErrorHandler)({
      error,
      response
    });
  }
};
exports.deleteFleetServerHostHandler = deleteFleetServerHostHandler;
const putFleetServerHostHandler = async (context, request, response) => {
  try {
    const coreContext = await await context.core;
    const esClient = coreContext.elasticsearch.client.asInternalUser;
    const soClient = coreContext.savedObjects.client;

    // In serverless, allow update fleet server host if host url is same as default.
    if (request.body.host_urls) {
      await checkFleetServerHostsWriteAPIsAllowed(soClient, request.body.host_urls);
    }
    const item = await (0, _fleet_server_host.updateFleetServerHost)(soClient, request.params.itemId, request.body);
    const body = {
      item
    };
    if (item.is_default) {
      await _services.agentPolicyService.bumpAllAgentPolicies(esClient);
    } else {
      await _services.agentPolicyService.bumpAllAgentPoliciesForFleetServerHosts(esClient, item.id);
    }
    return response.ok({
      body
    });
  } catch (error) {
    if (_server.SavedObjectsErrorHelpers.isNotFoundError(error)) {
      return response.notFound({
        body: {
          message: `Fleet server ${request.params.itemId} not found`
        }
      });
    }
    return (0, _errors.defaultFleetErrorHandler)({
      error,
      response
    });
  }
};
exports.putFleetServerHostHandler = putFleetServerHostHandler;
const getAllFleetServerHostsHandler = async (context, request, response) => {
  const soClient = (await context.core).savedObjects.client;
  try {
    const res = await (0, _fleet_server_host.listFleetServerHosts)(soClient);
    const body = {
      items: res.items,
      page: res.page,
      perPage: res.perPage,
      total: res.total
    };
    return response.ok({
      body
    });
  } catch (error) {
    return (0, _errors.defaultFleetErrorHandler)({
      error,
      response
    });
  }
};
exports.getAllFleetServerHostsHandler = getAllFleetServerHostsHandler;
