// @kbn/osquery-plugin/server/routes/live_query/get_live_query_results_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLiveQueryResultsRoute = void 0;
var _lodash = require("lodash");
var _rxjs = require("rxjs");
var _route_validation = require("../../utils/build_validation/route_validation");
var _constants = require("../../../common/constants");
var _common = require("../../../common");
var _search_strategy = require("../../../common/search_strategy");
var _build_query = require("../../../common/utils/build_query");
var _utils = require("./utils");
var _api = require("../../../common/api");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getLiveQueryResultsRoute = router => {
  router.versioned.get({
    access: 'public',
    path: '/api/osquery/live_queries/{id}/results/{actionId}',
    options: {
      tags: [`access:${_common.PLUGIN_ID}-read`]
    }
  }).addVersion({
    version: _constants.API_VERSIONS.public.v1,
    validate: {
      request: {
        query: (0, _route_validation.buildRouteValidation)(_api.getLiveQueryResultsRequestQuerySchema),
        params: (0, _route_validation.buildRouteValidation)(_api.getLiveQueryResultsRequestParamsSchema)
      }
    }
  }, async (context, request, response) => {
    const abortSignal = getRequestAbortedSignal(request.events.aborted$);
    try {
      var _actionDetails$_sourc, _request$query$page, _request$query$pageSi, _request$query$sortOr, _request$query$sort;
      const search = await context.search;
      const {
        actionDetails
      } = await (0, _rxjs.lastValueFrom)(search.search({
        actionId: request.params.id,
        kuery: request.query.kuery,
        factoryQueryType: _search_strategy.OsqueryQueries.actionDetails
      }, {
        abortSignal,
        strategy: 'osquerySearchStrategy'
      }));
      const queries = actionDetails === null || actionDetails === void 0 ? void 0 : (_actionDetails$_sourc = actionDetails._source) === null || _actionDetails$_sourc === void 0 ? void 0 : _actionDetails$_sourc.queries;
      await (0, _rxjs.lastValueFrom)((0, _rxjs.zip)(...(0, _lodash.map)(queries, query => {
        var _query$agents$length, _query$agents;
        return (0, _utils.getActionResponses)(search, query.action_id, (_query$agents$length = (_query$agents = query.agents) === null || _query$agents === void 0 ? void 0 : _query$agents.length) !== null && _query$agents$length !== void 0 ? _query$agents$length : 0);
      })));
      const res = await (0, _rxjs.lastValueFrom)(search.search({
        actionId: request.params.actionId,
        factoryQueryType: _search_strategy.OsqueryQueries.results,
        kuery: request.query.kuery,
        pagination: (0, _build_query.generateTablePaginationOptions)((_request$query$page = request.query.page) !== null && _request$query$page !== void 0 ? _request$query$page : 0, (_request$query$pageSi = request.query.pageSize) !== null && _request$query$pageSi !== void 0 ? _request$query$pageSi : 100),
        sort: [{
          direction: (_request$query$sortOr = request.query.sortOrder) !== null && _request$query$sortOr !== void 0 ? _request$query$sortOr : _search_strategy.Direction.desc,
          field: (_request$query$sort = request.query.sort) !== null && _request$query$sort !== void 0 ? _request$query$sort : '@timestamp'
        }]
      }, {
        abortSignal,
        strategy: 'osquerySearchStrategy'
      }));
      return response.ok({
        body: {
          data: res
        }
      });
    } catch (e) {
      var _e$statusCode;
      return response.customError({
        statusCode: (_e$statusCode = e.statusCode) !== null && _e$statusCode !== void 0 ? _e$statusCode : 500,
        body: {
          message: e.message
        }
      });
    }
  });
};
exports.getLiveQueryResultsRoute = getLiveQueryResultsRoute;
function getRequestAbortedSignal(aborted$) {
  const controller = new AbortController();
  aborted$.subscribe(() => controller.abort());
  return controller.signal;
}
