// @kbn/security-solution-plugin/public/attack_discovery/routes.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.routes = exports.AttackDiscoveryRoutes = void 0;
var _react = _interopRequireDefault(require("react"));
var _public = require("@kbn/usage-collection-plugin/public");
var _pages = require("./pages");
var _types = require("../app/types");
var _constants = require("../../common/constants");
var _plugin_template_wrapper = require("../common/components/plugin_template_wrapper");
var _security_route_page_wrapper = require("../common/components/security_route_page_wrapper");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AttackDiscoveryRoutes = () => /*#__PURE__*/_react.default.createElement(_plugin_template_wrapper.PluginTemplateWrapper, null, /*#__PURE__*/_react.default.createElement(_public.TrackApplicationView, {
  viewId: _types.SecurityPageName.attackDiscovery
}, /*#__PURE__*/_react.default.createElement(_security_route_page_wrapper.SecurityRoutePageWrapper, {
  pageName: _types.SecurityPageName.attackDiscovery
}, /*#__PURE__*/_react.default.createElement(_pages.AttackDiscoveryPage, null))));
exports.AttackDiscoveryRoutes = AttackDiscoveryRoutes;
const routes = exports.routes = [{
  path: _constants.ATTACK_DISCOVERY_PATH,
  component: AttackDiscoveryRoutes
}];
