// @kbn/ux-plugin/public/components/app/rum_dashboard/panels/visitor_breakdowns.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VisitorBreakdownsPanel = VisitorBreakdownsPanel;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _visitor_breakdown = require("../visitor_breakdown");
var _visitor_breakdown_map = require("../visitor_breakdown_map");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function VisitorBreakdownsPanel() {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    wrap: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    style: {
      flexBasis: 650
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_visitor_breakdown_map.VisitorBreakdownMap, null))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    style: {
      flexBasis: 650
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_visitor_breakdown.VisitorBreakdown, null))));
}
