// @kbn/saved-objects-management-plugin/public/management_section/saved_objects_edition_page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactRouterDom = require("react-router-dom");
var _queryString = require("query-string");
var _i18n = require("@kbn/i18n");
var _sharedUxLinkRedirectApp = require("@kbn/shared-ux-link-redirect-app");
var _object_view = require("./object_view");
require("./saved_objects_edition_page.scss");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const SavedObjectsEditionPage = ({
  coreStart,
  setBreadcrumbs,
  history
}) => {
  const {
    type,
    id
  } = (0, _reactRouterDom.useParams)();
  const capabilities = coreStart.application.capabilities;
  const docLinks = coreStart.docLinks.links;
  const {
    search
  } = (0, _reactRouterDom.useLocation)();
  const query = (0, _queryString.parse)(search);
  (0, _react.useEffect)(() => {
    setBreadcrumbs([{
      text: _i18n.i18n.translate('savedObjectsManagement.breadcrumb.index', {
        defaultMessage: 'Saved objects'
      }),
      href: '/'
    }, {
      text: _i18n.i18n.translate('savedObjectsManagement.breadcrumb.inspect', {
        defaultMessage: 'Inspect {savedObjectType}',
        values: {
          savedObjectType: type
        }
      })
    }]);
  }, [setBreadcrumbs, type]);
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "savedObjectsManagementEditionPage"
  }, /*#__PURE__*/_react.default.createElement(_sharedUxLinkRedirectApp.RedirectAppLinks, {
    coreStart: {
      application: coreStart.application
    }
  }, /*#__PURE__*/_react.default.createElement(_object_view.SavedObjectEdition, {
    id: id,
    savedObjectType: type,
    http: coreStart.http,
    overlays: coreStart.overlays,
    notifications: coreStart.notifications,
    capabilities: capabilities,
    notFoundType: query.notFound,
    uiSettings: coreStart.uiSettings,
    history: history,
    docLinks: docLinks,
    settings: coreStart.settings,
    theme: coreStart.theme
  })));
};

// eslint-disable-next-line import/no-default-export
exports.default = SavedObjectsEditionPage;
module.exports = exports.default;
