// @kbn/uptime-plugin/public/legacy_uptime/contexts/uptime_refresh_context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useUptimeRefreshContext = exports.UptimeRefreshContextProvider = exports.UptimeRefreshContext = void 0;
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const defaultContext = {
  lastRefresh: 0,
  refreshApp: () => {
    throw new Error('App refresh was not initialized, set it when you invoke the context');
  }
};
const UptimeRefreshContext = exports.UptimeRefreshContext = /*#__PURE__*/(0, _react.createContext)(defaultContext);
const UptimeRefreshContextProvider = ({
  children
}) => {
  const [lastRefresh, setLastRefresh] = (0, _react.useState)(Date.now());
  const refreshApp = () => {
    const refreshTime = Date.now();
    setLastRefresh(refreshTime);
  };
  const value = (0, _react.useMemo)(() => {
    return {
      lastRefresh,
      refreshApp
    };
  }, [lastRefresh]);
  return /*#__PURE__*/_react.default.createElement(UptimeRefreshContext.Provider, {
    value: value,
    children: children
  });
};
exports.UptimeRefreshContextProvider = UptimeRefreshContextProvider;
const useUptimeRefreshContext = () => (0, _react.useContext)(UptimeRefreshContext);
exports.useUptimeRefreshContext = useUptimeRefreshContext;
