// @kbn/triggers-actions-ui-plugin/public/common/truncated_text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TruncatedText = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LINE_CLAMP = 2;
const styles = {
  truncatedText: (0, _react2.css)`
    text-overflow: ellipsis;
    display: -webkit-box;
    -webkit-line-clamp: ${LINE_CLAMP};
    -webkit-box-orient: vertical;
    overflow: hidden;
    word-break: break-word;
  `
};
const TruncatedTextComponent = ({
  text
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "xs",
  color: "subdued",
  css: styles.truncatedText
}, text);
TruncatedTextComponent.displayName = 'TruncatedText';
const TruncatedText = exports.TruncatedText = /*#__PURE__*/_react.default.memo(TruncatedTextComponent);
