// @kbn/profiling-plugin/common/storage_explorer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.indexLifecyclePhaseRt = exports.indexLifeCyclePhaseToDataTier = exports.IndexLifecyclePhaseSelectOption = void 0;
var t = _interopRequireWildcard(require("io-ts"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let IndexLifecyclePhaseSelectOption = exports.IndexLifecyclePhaseSelectOption = /*#__PURE__*/function (IndexLifecyclePhaseSelectOption) {
  IndexLifecyclePhaseSelectOption["All"] = "all";
  IndexLifecyclePhaseSelectOption["Hot"] = "hot";
  IndexLifecyclePhaseSelectOption["Warm"] = "warm";
  IndexLifecyclePhaseSelectOption["Cold"] = "cold";
  IndexLifecyclePhaseSelectOption["Frozen"] = "frozen";
  return IndexLifecyclePhaseSelectOption;
}({});
const indexLifecyclePhaseRt = exports.indexLifecyclePhaseRt = t.type({
  indexLifecyclePhase: t.union([t.literal(IndexLifecyclePhaseSelectOption.All), t.literal(IndexLifecyclePhaseSelectOption.Hot), t.literal(IndexLifecyclePhaseSelectOption.Warm), t.literal(IndexLifecyclePhaseSelectOption.Cold), t.literal(IndexLifecyclePhaseSelectOption.Frozen)])
});
const indexLifeCyclePhaseToDataTier = exports.indexLifeCyclePhaseToDataTier = {
  [IndexLifecyclePhaseSelectOption.Hot]: 'data_hot',
  [IndexLifecyclePhaseSelectOption.Warm]: 'data_warm',
  [IndexLifecyclePhaseSelectOption.Cold]: 'data_cold',
  [IndexLifecyclePhaseSelectOption.Frozen]: 'data_frozen'
};
