// @kbn/vis-type-xy-plugin/public/types/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThresholdLineStyle = exports.ScaleType = exports.InterpolationMode = exports.ChartMode = exports.AxisType = exports.AxisMode = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let ChartMode = exports.ChartMode = /*#__PURE__*/function (ChartMode) {
  ChartMode["Normal"] = "normal";
  ChartMode["Stacked"] = "stacked";
  return ChartMode;
}({});
let InterpolationMode = exports.InterpolationMode = /*#__PURE__*/function (InterpolationMode) {
  InterpolationMode["Linear"] = "linear";
  InterpolationMode["Cardinal"] = "cardinal";
  InterpolationMode["StepAfter"] = "step-after";
  return InterpolationMode;
}({});
let AxisType = exports.AxisType = /*#__PURE__*/function (AxisType) {
  AxisType["Category"] = "category";
  AxisType["Value"] = "value";
  return AxisType;
}({});
let ScaleType = exports.ScaleType = /*#__PURE__*/function (ScaleType) {
  ScaleType["Linear"] = "linear";
  ScaleType["Log"] = "log";
  ScaleType["SquareRoot"] = "square root";
  return ScaleType;
}({});
let AxisMode = exports.AxisMode = /*#__PURE__*/function (AxisMode) {
  AxisMode["Normal"] = "normal";
  AxisMode["Percentage"] = "percentage";
  AxisMode["Wiggle"] = "wiggle";
  AxisMode["Silhouette"] = "silhouette";
  return AxisMode;
}({});
let ThresholdLineStyle = exports.ThresholdLineStyle = /*#__PURE__*/function (ThresholdLineStyle) {
  ThresholdLineStyle["Full"] = "full";
  ThresholdLineStyle["Dashed"] = "dashed";
  ThresholdLineStyle["DotDashed"] = "dot-dashed";
  return ThresholdLineStyle;
}({});
