// @kbn/integration-assistant-plugin/public/components/create_integration/create_integration_landing/integration_assistant_card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IntegrationAssistantCard = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _elasticAssistant = require("@kbn/elastic-assistant");
var _react2 = require("@emotion/react");
var _use_authorization = require("../../../common/hooks/use_authorization");
var _authorization = require("../../../common/components/authorization");
var _use_navigate = require("../../../common/hooks/use_navigate");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useAssistantCardCss = () => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return (0, _react2.css)`
    /* compensate for EuiCard children margin-block-start */
    margin-block-start: calc(${euiTheme.size.s} * -2);
  `;
};
const IntegrationAssistantCard = exports.IntegrationAssistantCard = /*#__PURE__*/_react.default.memo(() => {
  const {
    canExecuteConnectors
  } = (0, _use_authorization.useAuthorization)();
  const navigate = (0, _use_navigate.useNavigate)();
  const assistantCardCss = useAssistantCardCss();
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    display: "plain",
    hasBorder: true,
    paddingSize: "l",
    title: '' // title shown inside the child component
    ,
    betaBadgeProps: {
      label: i18n.TECH_PREVIEW,
      tooltipContent: i18n.TECH_PREVIEW_TOOLTIP
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "row",
    gutterSize: "l",
    alignItems: "center",
    justifyContent: "center",
    css: assistantCardCss
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_elasticAssistant.AssistantAvatar, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "xs",
    alignItems: "flexStart",
    justifyContent: "flexStart"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("h3", null, i18n.ASSISTANT_TITLE))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s",
    color: "subdued",
    textAlign: "left"
  }, i18n.ASSISTANT_DESCRIPTION)))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, canExecuteConnectors ? /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    onClick: () => navigate(_use_navigate.Page.assistant)
  }, i18n.ASSISTANT_BUTTON) : /*#__PURE__*/_react.default.createElement(_authorization.MissingPrivilegesTooltip, {
    canExecuteConnectors: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    disabled: true
  }, i18n.ASSISTANT_BUTTON)))));
});
IntegrationAssistantCard.displayName = 'IntegrationAssistantCard';
