// @kbn/security-solution-plugin/public/attack_discovery/attack_discovery_panel/view_in_ai_assistant/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewInAiAssistant = void 0;
var _elasticAssistant = require("@kbn/elastic-assistant");
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var i18n = _interopRequireWildcard(require("./translations"));
var _use_view_in_ai_assistant = require("./use_view_in_ai_assistant");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ViewInAiAssistantComponent = ({
  attackDiscovery,
  compact = false,
  replacements
}) => {
  const {
    showAssistantOverlay,
    disabled
  } = (0, _use_view_in_ai_assistant.useViewInAiAssistant)({
    attackDiscovery,
    replacements
  });
  return compact ? /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    "data-test-subj": "viewInAiAssistantCompact",
    disabled: disabled,
    iconType: "expand",
    onClick: showAssistantOverlay,
    size: "xs"
  }, i18n.VIEW_IN_AI_ASSISTANT) : /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "viewInAiAssistant",
    disabled: disabled,
    onClick: showAssistantOverlay,
    size: "s"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "xs"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "assistantAvatar",
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_elasticAssistant.AssistantAvatar, {
    size: "xs"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "viewInAiAssistantLabel",
    grow: false
  }, i18n.VIEW_IN_AI_ASSISTANT)));
};
ViewInAiAssistantComponent.displayName = 'ViewInAiAssistant';
const ViewInAiAssistant = exports.ViewInAiAssistant = /*#__PURE__*/_react.default.memo(ViewInAiAssistantComponent);
