// @kbn/security-solution-plugin/public/common/components/loader/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Loader = void 0;
var _eui = require("@elastic/eui");
var _polished = require("polished");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Aside = _styledComponents.default.aside`
  padding: ${({
  theme
}) => theme.eui.euiSizeM};

  ${({
  overlay,
  overlayBackground,
  theme
}) => overlay && (0, _styledComponents.css)`
      background: ${overlayBackground ? (0, _polished.rgba)(overlayBackground, 0.9) : (0, _polished.rgba)(theme.eui.euiColorEmptyShade, 0.9)};
      bottom: 0;
      left: 0;
      position: absolute;
      right: 0;
      top: 0;
      z-index: ${theme.eui.euiZLevel1};
    `}
`;
Aside.displayName = 'Aside';
const FlexGroup = (0, _styledComponents.default)(_eui.EuiFlexGroup).attrs(() => ({
  alignItems: 'center',
  direction: 'column',
  gutterSize: 's',
  justifyContent: 'center'
}))`
  ${({
  overlay
}) => overlay && (0, _styledComponents.css)`
      height: 100%;
    `}
`;
FlexGroup.displayName = 'FlexGroup';
const Loader = exports.Loader = /*#__PURE__*/_react.default.memo(({
  children,
  overlay,
  overlayBackground,
  size
}) => /*#__PURE__*/_react.default.createElement(Aside, {
  overlay: overlay,
  overlayBackground: overlayBackground
}, /*#__PURE__*/_react.default.createElement(FlexGroup, {
  overlay: {
    overlay
  }
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
  "data-test-subj": "loading-spinner",
  size: size
})), children && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  color: "subdued",
  size: "s"
}, /*#__PURE__*/_react.default.createElement("p", null, children))))));
Loader.displayName = 'Loader';
