// @kbn/security-solution-plugin/public/sourcerer/experimental/redux/selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sourcererAdapterSelector = void 0;
var _toolkit = require("@reduxjs/toolkit");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Compatibility layer / adapter for legacy selector consumers.
 * It is used in useSecuritySolutionDataView hook as alternative data source (behind a flag)
 */
const sourcererAdapterSelector = exports.sourcererAdapterSelector = (0, _toolkit.createSelector)([state => state.dataViewPicker], dataViewPicker => {
  return {
    loading: dataViewPicker.state === 'loading',
    dataViewId: dataViewPicker.dataView.id || '',
    patternList: dataViewPicker.patternList,
    indicesExist: true,
    browserFields: {},
    activePatterns: dataViewPicker.patternList,
    runtimeMappings: {},
    selectedPatterns: dataViewPicker.patternList,
    indexPattern: {
      fields: [],
      title: dataViewPicker.dataView.title || ''
    },
    sourcererDataView: {}
  };
});
