// @kbn/observability-ai-assistant-plugin/public/service/create_mock_chat_service.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createMockChatService = void 0;
var _common = require("../../common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createMockChatService = () => {
  const mockChatService = {
    chat: jest.fn(),
    complete: jest.fn(),
    sendAnalyticsEvent: jest.fn(),
    getFunctions: jest.fn().mockReturnValue([]),
    hasFunction: jest.fn().mockReturnValue(false),
    hasRenderFunction: jest.fn().mockReturnValue(true),
    renderFunction: jest.fn(),
    getSystemMessage: jest.fn().mockReturnValue({
      '@timestamp': new Date().toISOString(),
      message: {
        role: _common.MessageRole.System,
        content: 'system'
      }
    })
  };
  return mockChatService;
};
exports.createMockChatService = createMockChatService;
