// @kbn/data-view-management-plugin/public/components/empty_index_list_prompt/empty_index_list_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyIndexListPrompt = void 0;
require("./empty_index_list_prompt.scss");
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const EmptyIndexListPrompt = ({
  onRefresh,
  createAnyway,
  canSaveIndexPattern,
  addDataUrl,
  navigateToApp
}) => {
  const createAnywayLink = /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    color: "subdued",
    textAlign: "center",
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "indexPatternManagement.createDataView.emptyState.createAnywayTxt",
    defaultMessage: "You can also {link}",
    values: {
      link: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        onClick: () => createAnyway(),
        "data-test-subj": "createAnyway"
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternManagement.createDataView.emptyState.createAnywayLink",
        defaultMessage: "create a data view against hidden, system or default indices."
      }))
    }
  }));
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    className: "inpEmptyState",
    "data-test-subj": "indexPatternEmptyState",
    color: "subdued",
    hasShadow: false,
    paddingSize: "xl",
    css: (0, _react2.css)`
        margin: auto;
      `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPageHeader, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, null, /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "indexPatternManagement.createDataView.emptyState.noDataTitle",
    defaultMessage: "Ready to try Kibana? First, you need data."
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xl"
  }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGrid, {
    className: "inpEmptyState__cardGrid",
    columns: 3,
    responsive: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    className: "inpEmptyState__card",
    onClick: () => {
      navigateToApp('integrations', {
        path: '/browse'
      });
    },
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "database",
      color: "subdued"
    }),
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.integrationCardTitle",
      defaultMessage: "Add integration"
    }),
    description: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.integrationCardDescription",
      defaultMessage: "Add data from a variety of sources."
    })
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    onClick: () => navigateToApp('home', {
      path: '#/tutorial_directory/fileDataViz'
    }),
    className: "inpEmptyState__card",
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "document",
      color: "subdued"
    }),
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.uploadCardTitle",
      defaultMessage: "Upload a file"
    }),
    description: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.uploadCardDescription",
      defaultMessage: "Import a CSV, NDJSON, or log file."
    })
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    className: "inpEmptyState__card",
    onClick: () => {
      navigateToApp('home', {
        path: '#/tutorial_directory/sampleData'
      });
    },
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "heatmap",
      color: "subdued"
    }),
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.sampleDataCardTitle",
      defaultMessage: "Add sample data"
    }),
    description: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "indexPatternManagement.createDataView.emptyState.sampleDataCardDescription",
      defaultMessage: "Load a data set and a Kibana dashboard."
    })
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xxl"
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "inpEmptyState__footer"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    className: "inpEmptyState__footerFlexItem"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    listItems: [{
      title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternManagement.createDataView.emptyState.learnMore",
        defaultMessage: "Want to learn more?"
      }),
      description: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        href: addDataUrl,
        target: "_blank",
        external: true
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternManagement.createDataView.emptyState.readDocs",
        defaultMessage: "Read documentation"
      }))
    }]
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    className: "inpEmptyState__footerFlexItem"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    listItems: [{
      title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternManagement.createDataView.emptyState.haveData",
        defaultMessage: "Think you already have data?"
      }),
      description: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        onClick: onRefresh,
        "data-test-subj": "refreshIndicesButton"
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternManagement.createDataView.emptyState.checkDataButton",
        defaultMessage: "Check for new data"
      }), ' ', /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
        type: "refresh",
        size: "s"
      }))
    }]
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), canSaveIndexPattern && createAnywayLink)));
};
exports.EmptyIndexListPrompt = EmptyIndexListPrompt;
