// @kbn/security-solution-plugin/public/overview/components/overview_cti_links/cti_disabled_module.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CtiDisabledModuleComponent = exports.CtiDisabledModule = void 0;
var _react = _interopRequireDefault(require("react"));
var i18n = _interopRequireWildcard(require("./translations"));
var _disabled_link_panel = require("../link_panel/disabled_link_panel");
var _threat_intel_panel_view = require("./threat_intel_panel_view");
var _use_integrations_page_link = require("./use_integrations_page_link");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CtiDisabledModuleComponent = () => {
  const integrationsLink = (0, _use_integrations_page_link.useIntegrationsPageLink)();
  return /*#__PURE__*/_react.default.createElement(_disabled_link_panel.DisabledLinkPanel, {
    bodyCopy: i18n.DANGER_BODY,
    buttonCopy: i18n.DANGER_BUTTON,
    dataTestSubjPrefix: "cti",
    docLink: integrationsLink,
    listItems: [],
    titleCopy: i18n.DANGER_TITLE,
    LinkPanelViewComponent: _threat_intel_panel_view.ThreatIntelPanelView
  });
};
exports.CtiDisabledModuleComponent = CtiDisabledModuleComponent;
const CtiDisabledModule = exports.CtiDisabledModule = /*#__PURE__*/_react.default.memo(CtiDisabledModuleComponent);
CtiDisabledModule.displayName = 'CtiDisabledModule';
