// @kbn/infra-plugin/public/alerting/log_threshold/components/expression_editor/log_view_switcher.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogViewSwitcher = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const description = _i18n.i18n.translate('xpack.infra.logs.alertFlyout.logViewDescription', {
  defaultMessage: 'Log View'
});
/**
 * TODO: this component is called LogViewSwitcher because it will allow,
 * in a following implementation, to switch between the available logViews
 * with the support of multi-logView concept.
 * It currently renders a read-only expression to tell the user to which logView
 * is the new alert associated with.
 */
const LogViewSwitcher = ({
  logView
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: logView.indices
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiExpression, {
    description: description,
    value: logView.name
  }))));
};
exports.LogViewSwitcher = LogViewSwitcher;
