// @kbn/core-saved-objects-migration-server-internal/src/zdt/utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "buildIndexMappings", {
  enumerable: true,
  get: function () {
    return _build_index_mappings.buildIndexMappings;
  }
});
Object.defineProperty(exports, "buildIndexMeta", {
  enumerable: true,
  get: function () {
    return _build_index_mappings.buildIndexMeta;
  }
});
Object.defineProperty(exports, "buildMigratorConfigs", {
  enumerable: true,
  get: function () {
    return _get_migrator_configs.buildMigratorConfigs;
  }
});
Object.defineProperty(exports, "checkIndexCurrentAlgorithm", {
  enumerable: true,
  get: function () {
    return _check_index_algorithm.checkIndexCurrentAlgorithm;
  }
});
Object.defineProperty(exports, "checkVersionCompatibility", {
  enumerable: true,
  get: function () {
    return _check_version_compatibility.checkVersionCompatibility;
  }
});
Object.defineProperty(exports, "createDocumentTransformFn", {
  enumerable: true,
  get: function () {
    return _transform_raw_docs.createDocumentTransformFn;
  }
});
Object.defineProperty(exports, "generateAdditiveMappingDiff", {
  enumerable: true,
  get: function () {
    return _generate_additive_mapping_diff.generateAdditiveMappingDiff;
  }
});
Object.defineProperty(exports, "getAliasActions", {
  enumerable: true,
  get: function () {
    return _get_alias_actions.getAliasActions;
  }
});
Object.defineProperty(exports, "getCreationAliases", {
  enumerable: true,
  get: function () {
    return _get_creation_aliases.getCreationAliases;
  }
});
Object.defineProperty(exports, "getCurrentIndex", {
  enumerable: true,
  get: function () {
    return _get_current_index.getCurrentIndex;
  }
});
Object.defineProperty(exports, "getOutdatedDocumentsQuery", {
  enumerable: true,
  get: function () {
    return _outdated_documents_query.getOutdatedDocumentsQuery;
  }
});
Object.defineProperty(exports, "removePropertiesFromV2", {
  enumerable: true,
  get: function () {
    return _update_index_meta.removePropertiesFromV2;
  }
});
Object.defineProperty(exports, "setMetaDocMigrationComplete", {
  enumerable: true,
  get: function () {
    return _update_index_meta.setMetaDocMigrationComplete;
  }
});
Object.defineProperty(exports, "setMetaDocMigrationStarted", {
  enumerable: true,
  get: function () {
    return _update_index_meta.setMetaDocMigrationStarted;
  }
});
Object.defineProperty(exports, "setMetaMappingMigrationComplete", {
  enumerable: true,
  get: function () {
    return _update_index_meta.setMetaMappingMigrationComplete;
  }
});
var _get_migrator_configs = require("./get_migrator_configs");
var _get_current_index = require("./get_current_index");
var _check_version_compatibility = require("./check_version_compatibility");
var _build_index_mappings = require("./build_index_mappings");
var _get_alias_actions = require("./get_alias_actions");
var _generate_additive_mapping_diff = require("./generate_additive_mapping_diff");
var _outdated_documents_query = require("./outdated_documents_query");
var _transform_raw_docs = require("./transform_raw_docs");
var _update_index_meta = require("./update_index_meta");
var _check_index_algorithm = require("./check_index_algorithm");
var _get_creation_aliases = require("./get_creation_aliases");
