// @kbn/visualizations-plugin/public/utils/saved_visualization_references/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "extractControlsReferences", {
  enumerable: true,
  get: function () {
    return _controls_references.extractControlsReferences;
  }
});
Object.defineProperty(exports, "extractReferences", {
  enumerable: true,
  get: function () {
    return _saved_visualization_references.extractReferences;
  }
});
Object.defineProperty(exports, "extractTimeSeriesReferences", {
  enumerable: true,
  get: function () {
    return _timeseries_references.extractTimeSeriesReferences;
  }
});
Object.defineProperty(exports, "injectControlsReferences", {
  enumerable: true,
  get: function () {
    return _controls_references.injectControlsReferences;
  }
});
Object.defineProperty(exports, "injectReferences", {
  enumerable: true,
  get: function () {
    return _saved_visualization_references.injectReferences;
  }
});
Object.defineProperty(exports, "injectTimeSeriesReferences", {
  enumerable: true,
  get: function () {
    return _timeseries_references.injectTimeSeriesReferences;
  }
});
var _controls_references = require("./controls_references");
var _timeseries_references = require("./timeseries_references");
var _saved_visualization_references = require("./saved_visualization_references");
