// @kbn/vis-type-heatmap-plugin/public/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScaleType = exports.AxisType = exports.AxisMode = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
// ToDo: move them to constants
let ScaleType = exports.ScaleType = /*#__PURE__*/function (ScaleType) {
  ScaleType["Linear"] = "linear";
  ScaleType["Log"] = "log";
  ScaleType["SquareRoot"] = "square root";
  return ScaleType;
}({});
let AxisType = exports.AxisType = /*#__PURE__*/function (AxisType) {
  AxisType["Category"] = "category";
  AxisType["Value"] = "value";
  return AxisType;
}({});
let AxisMode = exports.AxisMode = /*#__PURE__*/function (AxisMode) {
  AxisMode["Normal"] = "normal";
  AxisMode["Percentage"] = "percentage";
  AxisMode["Wiggle"] = "wiggle";
  AxisMode["Silhouette"] = "silhouette";
  return AxisMode;
}({});
