// @kbn/infra-plugin/public/pages/404.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotFoundPage = void 0;
var _react = _interopRequireDefault(require("react"));
var _sharedUxPromptNotFound = require("@kbn/shared-ux-prompt-not-found");
var _use_kibana = require("../hooks/use_kibana");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const NotFoundPage = ({
  title
}) => {
  const {
    services: {
      observabilityShared: {
        navigation: {
          PageTemplate
        }
      }
    }
  } = (0, _use_kibana.useKibanaContextForPlugin)();
  return /*#__PURE__*/_react.default.createElement(PageTemplate, {
    pageHeader: {
      pageTitle: title
    },
    "data-test-subj": "infraNotFoundPage"
  }, /*#__PURE__*/_react.default.createElement(_sharedUxPromptNotFound.NotFoundPrompt, null));
};
exports.NotFoundPage = NotFoundPage;
