// @kbn/ml-plugin/public/application/data_frame_analytics/pages/analytics_creation/components/continue_button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ContinueButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const continueButtonText = _i18n.i18n.translate('xpack.ml.dataframe.analytics.creation.continueButtonText', {
  defaultMessage: 'Continue'
});
const ContinueButton = ({
  isDisabled,
  onClick
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
  "data-test-subj": "mlAnalyticsCreateJobWizardContinueButton",
  isDisabled: isDisabled,
  size: "s",
  onClick: onClick
}, continueButtonText)));
exports.ContinueButton = ContinueButton;
