// @kbn/share-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CSV_QUOTE_VALUES_SETTING", {
  enumerable: true,
  get: function () {
    return _constants.CSV_QUOTE_VALUES_SETTING;
  }
});
Object.defineProperty(exports, "CSV_SEPARATOR_SETTING", {
  enumerable: true,
  get: function () {
    return _constants.CSV_SEPARATOR_SETTING;
  }
});
Object.defineProperty(exports, "downloadFileAs", {
  enumerable: true,
  get: function () {
    return _download_as.downloadFileAs;
  }
});
Object.defineProperty(exports, "downloadMultipleAs", {
  enumerable: true,
  get: function () {
    return _download_as.downloadMultipleAs;
  }
});
exports.plugin = plugin;
Object.defineProperty(exports, "useLocatorUrl", {
  enumerable: true,
  get: function () {
    return _use_locator_url.useLocatorUrl;
  }
});
var _constants = require("../common/constants");
var _use_locator_url = require("../common/url_service/locators/use_locator_url");
var _plugin = require("./plugin");
var _download_as = require("./lib/download_as");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function plugin(ctx) {
  return new _plugin.SharePlugin(ctx);
}
