// @kbn/remote-clusters-plugin/public/application/store/action_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.REMOVE_CLUSTERS_START = exports.REMOVE_CLUSTERS_FINISH = exports.REFRESH_CLUSTERS_SUCCESS = exports.OPEN_DETAIL_PANEL = exports.LOAD_CLUSTERS_SUCCESS = exports.LOAD_CLUSTERS_START = exports.LOAD_CLUSTERS_FAILURE = exports.EDIT_CLUSTER_SUCCESS = exports.EDIT_CLUSTER_STOP = exports.EDIT_CLUSTER_START = exports.EDIT_CLUSTER_SAVE = exports.EDIT_CLUSTER_FAILURE = exports.CLOSE_DETAIL_PANEL = exports.CLEAR_EDIT_CLUSTER_ERRORS = exports.CLEAR_ADD_CLUSTER_ERRORS = exports.ADD_CLUSTER_SUCCESS = exports.ADD_CLUSTER_START = exports.ADD_CLUSTER_FAILURE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// Detail panel
const OPEN_DETAIL_PANEL = exports.OPEN_DETAIL_PANEL = 'OPEN_DETAIL_PANEL';
const CLOSE_DETAIL_PANEL = exports.CLOSE_DETAIL_PANEL = 'CLOSE_DETAIL_PANEL';

// Load clusters
const LOAD_CLUSTERS_START = exports.LOAD_CLUSTERS_START = 'LOAD_CLUSTERS_START';
const LOAD_CLUSTERS_SUCCESS = exports.LOAD_CLUSTERS_SUCCESS = 'LOAD_CLUSTERS_SUCCESS';
const LOAD_CLUSTERS_FAILURE = exports.LOAD_CLUSTERS_FAILURE = 'LOAD_CLUSTERS_FAILURE';

// Refresh clusters
const REFRESH_CLUSTERS_SUCCESS = exports.REFRESH_CLUSTERS_SUCCESS = 'REFRESH_CLUSTERS_SUCCESS';

// Add cluster
const ADD_CLUSTER_START = exports.ADD_CLUSTER_START = 'ADD_CLUSTER_START';
const ADD_CLUSTER_SUCCESS = exports.ADD_CLUSTER_SUCCESS = 'ADD_CLUSTER_SUCCESS';
const ADD_CLUSTER_FAILURE = exports.ADD_CLUSTER_FAILURE = 'ADD_CLUSTER_FAILURE';
const CLEAR_ADD_CLUSTER_ERRORS = exports.CLEAR_ADD_CLUSTER_ERRORS = 'CLEAR_ADD_CLUSTER_ERRORS';

// Edit cluster
const EDIT_CLUSTER_START = exports.EDIT_CLUSTER_START = 'EDIT_CLUSTER_START';
const EDIT_CLUSTER_STOP = exports.EDIT_CLUSTER_STOP = 'EDIT_CLUSTER_STOP';
const EDIT_CLUSTER_SAVE = exports.EDIT_CLUSTER_SAVE = 'EDIT_CLUSTER_SAVE';
const EDIT_CLUSTER_SUCCESS = exports.EDIT_CLUSTER_SUCCESS = 'EDIT_CLUSTER_SUCCESS';
const EDIT_CLUSTER_FAILURE = exports.EDIT_CLUSTER_FAILURE = 'EDIT_CLUSTER_FAILURE';
const CLEAR_EDIT_CLUSTER_ERRORS = exports.CLEAR_EDIT_CLUSTER_ERRORS = 'CLEAR_EDIT_CLUSTER_ERRORS';

// Remove clusters
const REMOVE_CLUSTERS_START = exports.REMOVE_CLUSTERS_START = 'REMOVE_CLUSTERS_START';
const REMOVE_CLUSTERS_FINISH = exports.REMOVE_CLUSTERS_FINISH = 'REMOVE_CLUSTERS_FINISH';
