// @kbn/actions-plugin/common/routes/connector/response/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "allConnectorsResponseSchema", {
  enumerable: true,
  get: function () {
    return _latest.allConnectorsResponseSchema;
  }
});
Object.defineProperty(exports, "connectorResponseSchema", {
  enumerable: true,
  get: function () {
    return _latest.connectorResponseSchema;
  }
});
Object.defineProperty(exports, "connectorResponseSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.connectorResponseSchema;
  }
});
Object.defineProperty(exports, "connectorTypesResponseSchema", {
  enumerable: true,
  get: function () {
    return _latest.connectorTypesResponseSchema;
  }
});
Object.defineProperty(exports, "connectorTypesResponseSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.connectorTypesResponseSchema;
  }
});
Object.defineProperty(exports, "connectorWithExtraFindDataSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.allConnectorsResponseSchema;
  }
});
var _latest = require("./schemas/latest");
var _v = require("./schemas/v1");
