// @kbn/spaces-plugin/server/saved_objects/schemas.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpacesSavedObjectSchemas = void 0;
var _configSchema = require("@kbn/config-schema");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// These fields are not searched or aggregated on
const SpacesSavedObjectSchemas = exports.SpacesSavedObjectSchemas = {
  '8.8.0': _configSchema.schema.object({
    name: _configSchema.schema.string({
      minLength: 1
    }),
    description: _configSchema.schema.maybe(_configSchema.schema.string()),
    initials: _configSchema.schema.maybe(_configSchema.schema.string()),
    color: _configSchema.schema.maybe(_configSchema.schema.string()),
    disabledFeatures: _configSchema.schema.maybe(_configSchema.schema.arrayOf(_configSchema.schema.string())),
    imageUrl: _configSchema.schema.maybe(_configSchema.schema.string()),
    _reserved: _configSchema.schema.maybe(_configSchema.schema.boolean())
  })
};
