// @kbn/alerting-plugin/server/routes/lib/rewrite_maintenance_window.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.rewritePartialMaintenanceBodyRes = exports.rewriteMaintenanceWindowRes = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const rewriteMaintenanceWindowRes = ({
  expirationDate,
  rRule,
  createdBy,
  updatedBy,
  createdAt,
  updatedAt,
  eventStartTime,
  eventEndTime,
  ...rest
}) => ({
  ...rest,
  expiration_date: expirationDate,
  r_rule: rRule,
  created_by: createdBy,
  updated_by: updatedBy,
  created_at: createdAt,
  updated_at: updatedAt,
  event_start_time: eventStartTime,
  event_end_time: eventEndTime
});
exports.rewriteMaintenanceWindowRes = rewriteMaintenanceWindowRes;
const rewritePartialMaintenanceBodyRes = ({
  expirationDate,
  rRule,
  createdBy,
  updatedBy,
  createdAt,
  updatedAt,
  eventStartTime,
  eventEndTime,
  ...rest
}) => ({
  ...rest,
  expiration_date: expirationDate,
  r_rule: rRule,
  created_by: createdBy,
  updated_by: updatedBy,
  created_at: createdAt,
  updated_at: updatedAt,
  event_start_time: eventStartTime,
  event_end_time: eventEndTime
});
exports.rewritePartialMaintenanceBodyRes = rewritePartialMaintenanceBodyRes;
