// @kbn/security-solution-plugin/public/timelines/components/open_timeline/export_timeline/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditTimelineActionsComponent = exports.EditTimelineActions = void 0;
var _react = _interopRequireDefault(require("react"));
var _export_timeline = require("./export_timeline");
var _delete_timeline_modal = require("../delete_timeline_modal");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EditTimelineActionsComponent = ({
  deleteTimelines,
  ids,
  savedSearchIds,
  isEnableDownloader,
  isDeleteTimelineModalOpen,
  onComplete,
  title
}) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_export_timeline.TimelineDownloader, {
  "data-test-subj": "TimelineDownloader",
  exportedIds: ids,
  isEnableDownloader: isEnableDownloader,
  onComplete: onComplete
}), deleteTimelines != null && /*#__PURE__*/_react.default.createElement(_delete_timeline_modal.DeleteTimelineModalOverlay, {
  "data-test-subj": "DeleteTimelineModalOverlay",
  deleteTimelines: deleteTimelines,
  isModalOpen: isDeleteTimelineModalOpen,
  onComplete: onComplete,
  savedObjectIds: ids,
  savedSearchIds: savedSearchIds,
  title: title
}));
exports.EditTimelineActionsComponent = EditTimelineActionsComponent;
const EditTimelineActions = exports.EditTimelineActions = /*#__PURE__*/_react.default.memo(EditTimelineActionsComponent);
