// @kbn/index-lifecycle-management-plugin/public/shared_imports.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "APP_WRAPPER_CLASS", {
  enumerable: true,
  get: function () {
    return _public3.APP_WRAPPER_CLASS;
  }
});
Object.defineProperty(exports, "CheckBoxField", {
  enumerable: true,
  get: function () {
    return _components.CheckBoxField;
  }
});
Object.defineProperty(exports, "ComboBoxField", {
  enumerable: true,
  get: function () {
    return _components.ComboBoxField;
  }
});
Object.defineProperty(exports, "Form", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.Form;
  }
});
Object.defineProperty(exports, "KibanaContextProvider", {
  enumerable: true,
  get: function () {
    return _public.KibanaContextProvider;
  }
});
Object.defineProperty(exports, "KibanaRenderContextProvider", {
  enumerable: true,
  get: function () {
    return _reactKibanaContextRender.KibanaRenderContextProvider;
  }
});
Object.defineProperty(exports, "NumericField", {
  enumerable: true,
  get: function () {
    return _components.NumericField;
  }
});
Object.defineProperty(exports, "RedirectAppLinks", {
  enumerable: true,
  get: function () {
    return _sharedUxLinkRedirectApp.RedirectAppLinks;
  }
});
Object.defineProperty(exports, "SelectField", {
  enumerable: true,
  get: function () {
    return _components.SelectField;
  }
});
Object.defineProperty(exports, "SuperSelectField", {
  enumerable: true,
  get: function () {
    return _components.SuperSelectField;
  }
});
Object.defineProperty(exports, "TextField", {
  enumerable: true,
  get: function () {
    return _components.TextField;
  }
});
Object.defineProperty(exports, "ToggleField", {
  enumerable: true,
  get: function () {
    return _components.ToggleField;
  }
});
Object.defineProperty(exports, "UseMultiFields", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.UseMultiFields;
  }
});
Object.defineProperty(exports, "attemptToURIDecode", {
  enumerable: true,
  get: function () {
    return _public2.attemptToURIDecode;
  }
});
Object.defineProperty(exports, "fieldValidators", {
  enumerable: true,
  get: function () {
    return _helpers.fieldValidators;
  }
});
Object.defineProperty(exports, "getFieldValidityAndErrorMessage", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.getFieldValidityAndErrorMessage;
  }
});
Object.defineProperty(exports, "useExecutionContext", {
  enumerable: true,
  get: function () {
    return _public.useExecutionContext;
  }
});
Object.defineProperty(exports, "useForm", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.useForm;
  }
});
Object.defineProperty(exports, "useFormContext", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.useFormContext;
  }
});
Object.defineProperty(exports, "useFormData", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.useFormData;
  }
});
Object.defineProperty(exports, "useFormIsModified", {
  enumerable: true,
  get: function () {
    return _hook_form_lib.useFormIsModified;
  }
});
exports.useKibana = void 0;
var _public = require("@kbn/kibana-react-plugin/public");
var _hook_form_lib = require("@kbn/es-ui-shared-plugin/static/forms/hook_form_lib");
var _helpers = require("@kbn/es-ui-shared-plugin/static/forms/helpers");
var _components = require("@kbn/es-ui-shared-plugin/static/forms/components");
var _public2 = require("@kbn/es-ui-shared-plugin/public");
var _reactKibanaContextRender = require("@kbn/react-kibana-context-render");
var _sharedUxLinkRedirectApp = require("@kbn/shared-ux-link-redirect-app");
var _public3 = require("@kbn/core/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useKibana = () => (0, _public.useKibana)();
exports.useKibana = useKibana;
