// @kbn/canvas-plugin/public/routes/workpad/workpad_routing_context.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WorkpadRoutingContextComponent = exports.WorkpadRoutingContext = void 0;
var _react = _interopRequireWildcard(require("react"));
var _use_routing_context = require("./hooks/use_routing_context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const basicWorkpadRoutingContext = {
  gotoPage: page => undefined,
  getUrl: page => '',
  isFullscreen: false,
  setFullscreen: fullscreen => undefined,
  autoplayInterval: 0,
  setAutoplayInterval: interval => undefined,
  isAutoplayPaused: true,
  setIsAutoplayPaused: isPaused => undefined,
  nextPage: () => undefined,
  previousPage: () => undefined,
  refreshInterval: 0,
  setRefreshInterval: interval => undefined,
  undo: () => undefined,
  redo: () => undefined
};
const WorkpadRoutingContext = exports.WorkpadRoutingContext = /*#__PURE__*/(0, _react.createContext)(basicWorkpadRoutingContext);
const WorkpadRoutingContextComponent = ({
  children
}) => {
  const routingContext = (0, _use_routing_context.useRoutingContext)();
  return /*#__PURE__*/_react.default.createElement(WorkpadRoutingContext.Provider, {
    value: routingContext
  }, children);
};
exports.WorkpadRoutingContextComponent = WorkpadRoutingContextComponent;
