// @kbn/enterprise-search-plugin/public/applications/shared/kibana_chrome/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SetAiSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetAiSearchChrome;
  }
});
Object.defineProperty(exports, "SetAnalyticsChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetAnalyticsChrome;
  }
});
Object.defineProperty(exports, "SetAppSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetAppSearchChrome;
  }
});
Object.defineProperty(exports, "SetElasticsearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetElasticsearchChrome;
  }
});
Object.defineProperty(exports, "SetEnterpriseSearchApplicationsChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetEnterpriseSearchApplicationsChrome;
  }
});
Object.defineProperty(exports, "SetEnterpriseSearchContentChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetEnterpriseSearchContentChrome;
  }
});
Object.defineProperty(exports, "SetEnterpriseSearchRelevanceChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetEnterpriseSearchRelevanceChrome;
  }
});
Object.defineProperty(exports, "SetSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetSearchChrome;
  }
});
Object.defineProperty(exports, "SetSearchExperiencesChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetSearchExperiencesChrome;
  }
});
Object.defineProperty(exports, "SetSemanticSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetSemanticSearchChrome;
  }
});
Object.defineProperty(exports, "SetVectorSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetVectorSearchChrome;
  }
});
Object.defineProperty(exports, "SetWorkplaceSearchChrome", {
  enumerable: true,
  get: function () {
    return _set_chrome.SetWorkplaceSearchChrome;
  }
});
var _set_chrome = require("./set_chrome");
