// @kbn/esql-utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ENABLE_ESQL", {
  enumerable: true,
  get: function () {
    return _constants.ENABLE_ESQL;
  }
});
Object.defineProperty(exports, "TextBasedLanguages", {
  enumerable: true,
  get: function () {
    return _src.TextBasedLanguages;
  }
});
Object.defineProperty(exports, "appendToESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.appendToESQLQuery;
  }
});
Object.defineProperty(exports, "appendWhereClauseToESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.appendWhereClauseToESQLQuery;
  }
});
Object.defineProperty(exports, "getESQLAdHocDataview", {
  enumerable: true,
  get: function () {
    return _src.getESQLAdHocDataview;
  }
});
Object.defineProperty(exports, "getESQLQueryColumns", {
  enumerable: true,
  get: function () {
    return _src.getESQLQueryColumns;
  }
});
Object.defineProperty(exports, "getESQLQueryColumnsRaw", {
  enumerable: true,
  get: function () {
    return _src.getESQLQueryColumnsRaw;
  }
});
Object.defineProperty(exports, "getESQLResults", {
  enumerable: true,
  get: function () {
    return _src.getESQLResults;
  }
});
Object.defineProperty(exports, "getESQLWithSafeLimit", {
  enumerable: true,
  get: function () {
    return _src.getESQLWithSafeLimit;
  }
});
Object.defineProperty(exports, "getIndexForESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.getIndexForESQLQuery;
  }
});
Object.defineProperty(exports, "getIndexPatternFromESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.getIndexPatternFromESQLQuery;
  }
});
Object.defineProperty(exports, "getInitialESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.getInitialESQLQuery;
  }
});
Object.defineProperty(exports, "getLimitFromESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.getLimitFromESQLQuery;
  }
});
Object.defineProperty(exports, "hasTransformationalCommand", {
  enumerable: true,
  get: function () {
    return _src.hasTransformationalCommand;
  }
});
Object.defineProperty(exports, "removeDropCommandsFromESQLQuery", {
  enumerable: true,
  get: function () {
    return _src.removeDropCommandsFromESQLQuery;
  }
});
var _src = require("./src");
var _constants = require("./constants");
