// @kbn/lens-plugin/public/visualizations/datatable/components/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LENS_TOGGLE_ACTION = exports.LENS_EDIT_SORT_ACTION = exports.LENS_EDIT_RESIZE_ACTION = exports.LENS_EDIT_PAGESIZE_ACTION = exports.DEFAULT_ROW_HEIGHT_LINES = exports.DEFAULT_ROW_HEIGHT = exports.DEFAULT_HEADER_ROW_HEIGHT_LINES = exports.DEFAULT_HEADER_ROW_HEIGHT = void 0;
var _types = require("../../../../common/types");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LENS_EDIT_SORT_ACTION = exports.LENS_EDIT_SORT_ACTION = 'sort';
const LENS_EDIT_RESIZE_ACTION = exports.LENS_EDIT_RESIZE_ACTION = 'resize';
const LENS_TOGGLE_ACTION = exports.LENS_TOGGLE_ACTION = 'toggle';
const LENS_EDIT_PAGESIZE_ACTION = exports.LENS_EDIT_PAGESIZE_ACTION = 'pagesize';
const DEFAULT_HEADER_ROW_HEIGHT_LINES = exports.DEFAULT_HEADER_ROW_HEIGHT_LINES = 3;
const DEFAULT_HEADER_ROW_HEIGHT = exports.DEFAULT_HEADER_ROW_HEIGHT = _types.RowHeightMode.custom;
const DEFAULT_ROW_HEIGHT = exports.DEFAULT_ROW_HEIGHT = _types.RowHeightMode.single;
const DEFAULT_ROW_HEIGHT_LINES = exports.DEFAULT_ROW_HEIGHT_LINES = 2;
