// @kbn/security-solution-plugin/common/api/detection_engine/model/error_schema.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorSchema = void 0;
var _zod = require("zod");
var _common_attributes = require("./rule_schema/common_attributes.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Error Schema
 *   version: not applicable
 */

const ErrorSchema = exports.ErrorSchema = _zod.z.object({
  id: _zod.z.string().optional(),
  rule_id: _common_attributes.RuleSignatureId.optional(),
  list_id: _zod.z.string().min(1).optional(),
  item_id: _zod.z.string().min(1).optional(),
  error: _zod.z.object({
    status_code: _zod.z.number().int().min(400),
    message: _zod.z.string()
  })
}).strict();
