// @kbn/enterprise-search-plugin/public/applications/workplace_search/components/shared/source_row/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SOURCE_ROW_SEARCHABLE_TOGGLE_LABEL = exports.SOURCE_ROW_REMOTE_TOOLTIP = exports.SOURCE_ROW_REMOTE_LABEL = exports.SOURCE_ROW_REAUTHENTICATE_STATUS_LINK_LABEL = exports.SOURCE_ROW_DETAILS_LABEL = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SOURCE_ROW_REAUTHENTICATE_STATUS_LINK_LABEL = exports.SOURCE_ROW_REAUTHENTICATE_STATUS_LINK_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.sourceRow.reauthenticateStatusLinkLabel', {
  defaultMessage: 'Re-authenticate'
});
const SOURCE_ROW_REMOTE_LABEL = exports.SOURCE_ROW_REMOTE_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.sourceRow.remoteLabel', {
  defaultMessage: 'Remote'
});
const SOURCE_ROW_REMOTE_TOOLTIP = exports.SOURCE_ROW_REMOTE_TOOLTIP = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.sourceRow.remoteTooltip', {
  defaultMessage: "Remote sources rely on the source's search service directly, and no content is indexed with Workplace Search. Speed and integrity of results are functions of the third-party service's health and performance."
});
const SOURCE_ROW_SEARCHABLE_TOGGLE_LABEL = exports.SOURCE_ROW_SEARCHABLE_TOGGLE_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.sourceRow.searchableToggleLabel', {
  defaultMessage: 'Source searchable toggle'
});
const SOURCE_ROW_DETAILS_LABEL = exports.SOURCE_ROW_DETAILS_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.sourceRow.detailsLabel', {
  defaultMessage: 'Details'
});
