// @kbn/infra-plugin/public/containers/logs/log_analysis/modules/log_entry_categories/use_log_entry_categories_module.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLogEntryCategoriesModuleContext = exports.useLogEntryCategoriesModule = exports.LogEntryCategoriesModuleProvider = void 0;
var _constate = _interopRequireDefault(require("constate"));
var _react = require("react");
var _log_analysis_module = require("../../log_analysis_module");
var _log_analysis_module_configuration = require("../../log_analysis_module_configuration");
var _log_analysis_module_definition = require("../../log_analysis_module_definition");
var _module_descriptor = require("./module_descriptor");
var _use_log_entry_categories_quality = require("./use_log_entry_categories_quality");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useLogEntryCategoriesModule = ({
  indexPattern,
  logViewId,
  spaceId,
  idFormat,
  timestampField,
  runtimeMappings
}) => {
  const sourceConfiguration = (0, _react.useMemo)(() => ({
    indices: indexPattern.split(','),
    sourceId: logViewId,
    spaceId,
    timestampField,
    runtimeMappings
  }), [indexPattern, logViewId, spaceId, timestampField, runtimeMappings]);
  const logAnalysisModule = (0, _log_analysis_module.useLogAnalysisModule)({
    moduleDescriptor: _module_descriptor.logEntryCategoriesModule,
    idFormat,
    sourceConfiguration
  });
  const {
    getIsJobConfigurationOutdated
  } = (0, _log_analysis_module_configuration.useLogAnalysisModuleConfiguration)({
    sourceConfiguration,
    moduleDescriptor: _module_descriptor.logEntryCategoriesModule
  });
  const {
    fetchModuleDefinition,
    getIsJobDefinitionOutdated
  } = (0, _log_analysis_module_definition.useLogAnalysisModuleDefinition)({
    sourceConfiguration,
    idFormat,
    moduleDescriptor: _module_descriptor.logEntryCategoriesModule
  });
  const {
    categoryQualityWarnings
  } = (0, _use_log_entry_categories_quality.useLogEntryCategoriesQuality)({
    jobSummaries: logAnalysisModule.jobSummaries
  });
  const hasOutdatedJobConfigurations = (0, _react.useMemo)(() => logAnalysisModule.jobSummaries.some(getIsJobConfigurationOutdated), [getIsJobConfigurationOutdated, logAnalysisModule.jobSummaries]);
  const hasOutdatedJobDefinitions = (0, _react.useMemo)(() => logAnalysisModule.jobSummaries.some(getIsJobDefinitionOutdated), [getIsJobDefinitionOutdated, logAnalysisModule.jobSummaries]);
  const hasStoppedJobs = (0, _react.useMemo)(() => Object.values(logAnalysisModule.jobStatus).some(currentJobStatus => currentJobStatus === 'stopped'), [logAnalysisModule.jobStatus]);
  return {
    ...logAnalysisModule,
    categoryQualityWarnings,
    fetchModuleDefinition,
    hasOutdatedJobConfigurations,
    hasOutdatedJobDefinitions,
    hasStoppedJobs
  };
};
exports.useLogEntryCategoriesModule = useLogEntryCategoriesModule;
const [LogEntryCategoriesModuleProvider, useLogEntryCategoriesModuleContext] = (0, _constate.default)(useLogEntryCategoriesModule);
exports.useLogEntryCategoriesModuleContext = useLogEntryCategoriesModuleContext;
exports.LogEntryCategoriesModuleProvider = LogEntryCategoriesModuleProvider;
