// @kbn/security-solution-plugin/server/assistant/tools/knowledge_base/knowledge_base_retrieval_tool.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KNOWLEDGE_BASE_RETRIEVAL_TOOL = void 0;
var _tools = require("@langchain/core/tools");
var _zod = require("zod");
var _common = require("../../../../common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const toolDetails = {
  description: "Call this for fetching details from the user's knowledge base. The knowledge base contains useful information the user wants to store between conversation contexts. Call this function when the user asks for information about themself, like 'what is my favorite...' or 'using my saved....'. Input must always be the free-text query on a single line, with no other text. You are welcome to re-write the query to be a summary of items/things to search for in the knowledge base, as a vector search will be performed to return similar results when requested. If the results returned do not look relevant, disregard and tell the user you were unable to find the information they were looking for. All requests include a `knowledge history` section which includes some existing knowledge of the user. DO NOT CALL THIS FUNCTION if the `knowledge history` sections appears to be able to answer the user's query.",
  id: 'knowledge-base-retrieval-tool',
  name: 'KnowledgeBaseRetrievalTool'
};
const KNOWLEDGE_BASE_RETRIEVAL_TOOL = exports.KNOWLEDGE_BASE_RETRIEVAL_TOOL = {
  ...toolDetails,
  sourceRegister: _common.APP_UI_ID,
  isSupported: params => {
    const {
      kbDataClient,
      isEnabledKnowledgeBase,
      modelExists
    } = params;
    return isEnabledKnowledgeBase && modelExists && kbDataClient != null;
  },
  getTool(params) {
    if (!this.isSupported(params)) return null;
    const {
      kbDataClient,
      logger
    } = params;
    if (kbDataClient == null) return null;
    return new _tools.DynamicStructuredTool({
      name: toolDetails.name,
      description: toolDetails.description,
      schema: _zod.z.object({
        query: _zod.z.string().describe(`Summary of items/things to search for in the knowledge base`)
      }),
      func: async (input, _, cbManager) => {
        logger.debug(`KnowledgeBaseRetrievalToolParams:input\n ${JSON.stringify(input, null, 2)}`);
        const docs = await kbDataClient.getKnowledgeBaseDocuments({
          query: input.query,
          kbResource: 'user',
          required: false
        });
        return JSON.stringify(docs);
      },
      tags: ['knowledge-base']
      // TODO: Remove after ZodAny is fixed https://github.com/langchain-ai/langchainjs/blob/main/langchain-core/src/tools.ts
    });
  }
};
