// @kbn/saved-objects-tagging-plugin/public/components/assign_flyout/open_assign_flyout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAssignFlyoutOpener = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _reactKibanaMount = require("@kbn/react-kibana-mount");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const LoadingIndicator = () => /*#__PURE__*/_react.default.createElement(_eui.EuiDelayRender, null, /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, null));
const LazyAssignFlyout = /*#__PURE__*/_react.default.lazy(() => Promise.resolve().then(() => _interopRequireWildcard(require('./assign_flyout'))).then(({
  AssignFlyout
}) => ({
  default: AssignFlyout
})));
const getAssignFlyoutOpener = ({
  overlays,
  notifications,
  tagCache,
  assignmentService,
  assignableTypes,
  ...startServices
}) => async ({
  tagIds
}) => {
  const flyout = overlays.openFlyout((0, _reactKibanaMount.toMountPoint)( /*#__PURE__*/_react.default.createElement(_react.default.Suspense, {
    fallback: /*#__PURE__*/_react.default.createElement(LoadingIndicator, null)
  }, /*#__PURE__*/_react.default.createElement(LazyAssignFlyout, {
    tagIds: tagIds,
    tagCache: tagCache,
    notifications: notifications,
    allowedTypes: assignableTypes,
    assignmentService: assignmentService,
    onClose: () => flyout.close()
  })), startServices), {
    size: 'm',
    maxWidth: 600
  });
  return flyout;
};
exports.getAssignFlyoutOpener = getAssignFlyoutOpener;
