// @kbn/security-solution-plugin/server/lib/entity_analytics/asset_criticality/audit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AssetCriticalityAuditActions = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let AssetCriticalityAuditActions = exports.AssetCriticalityAuditActions = /*#__PURE__*/function (AssetCriticalityAuditActions) {
  AssetCriticalityAuditActions["ASSET_CRITICALITY_INITIALIZE"] = "asset_criticality_initialize";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_UNASSIGN"] = "asset_criticality_unassign";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_GET"] = "asset_criticality_get";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_PRIVILEGE_GET"] = "asset_criticality_privilege_get";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_STATUS_GET"] = "asset_criticality_status_get";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_UPDATE"] = "asset_criticality_update";
  AssetCriticalityAuditActions["ASSET_CRITICALITY_BULK_UPDATE"] = "asset_criticality_bulk_update";
  return AssetCriticalityAuditActions;
}({});
