// @kbn/osquery-plugin/common/api/saved_query/create_saved_query.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SuccessResponse = exports.CreateSavedQueryRequestBody = void 0;
var _zod = require("zod");
var _common_attributes = require("../model/schema/common_attributes.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Create Saved Query Schema
 *   version: 2023-10-31
 */

const CreateSavedQueryRequestBody = exports.CreateSavedQueryRequestBody = _zod.z.object({
  id: _common_attributes.SavedQueryId.optional(),
  description: _common_attributes.DescriptionOrUndefined.optional(),
  query: _common_attributes.QueryOrUndefined.optional(),
  ecs_mapping: _common_attributes.ECSMappingOrUndefined.optional(),
  version: _common_attributes.VersionOrUndefined.optional(),
  platform: _common_attributes.DescriptionOrUndefined.optional(),
  interval: _common_attributes.Interval.optional(),
  snapshot: _common_attributes.SnapshotOrUndefined.optional(),
  removed: _common_attributes.RemovedOrUndefined.optional()
});
const SuccessResponse = exports.SuccessResponse = _zod.z.object({});
