// @kbn/enterprise-search-plugin/public/applications/applications/components/search_applications/create_search_application_flyout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateSearchApplication = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactRouterDom = require("react-router-dom");
var _kea = require("kea");
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _api = require("../../../../../common/types/api");
var _constants = require("../../../shared/constants");
var _doc_links = require("../../../shared/doc_links");
var _handle_api_errors = require("../../../shared/flash_messages/handle_api_errors");
var _query_params = require("../../../shared/query_params");
var _indices_select_combobox = require("./components/indices_select_combobox");
var _create_search_application_logic = require("./create_search_application_logic");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CreateSearchApplication = ({
  onClose
}) => {
  const {
    createSearchApplication,
    setName,
    setSelectedIndices
  } = (0, _kea.useActions)(_create_search_application_logic.CreateSearchApplicationLogic);
  const {
    createDisabled,
    createSearchApplicationError,
    createSearchApplicationStatus,
    searchApplicationName,
    searchApplicationNameStatus,
    formDisabled,
    indicesStatus,
    selectedIndices
  } = (0, _kea.useValues)(_create_search_application_logic.CreateSearchApplicationLogic);
  const {
    search
  } = (0, _reactRouterDom.useLocation)();
  const {
    ...params
  } = (0, _query_params.parseQueryParams)(search);
  const indexName = params[_constants.ESINDEX_QUERY_PARAMETER];
  const onIndicesChange = selectedOptions => {
    setSelectedIndices(selectedOptions.map(option => option.label));
  };
  (0, _react.useEffect)(() => {
    if (indexName && typeof indexName === 'string') setSelectedIndices([indexName]);
  }, []);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyout, {
    onClose: onClose,
    size: "m"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "m"
  }, /*#__PURE__*/_react.default.createElement("h3", null, _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.headerTitle', {
    defaultMessage: 'Create a Search Application'
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    color: "subdued"
  }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.enterpriseSearch.searchApplications.createSearchApplication.headerSubTitle",
    defaultMessage: "Explore our {docsLink} to learn more!",
    values: {
      docsLink: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        href: _doc_links.docLinks.searchApplications,
        target: "_blank",
        "data-telemetry-id": "entSearchApplications-createSearchApplication-docsLink",
        external: true
      }, _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.header.docsLink', {
        defaultMessage: 'Search Applications documentation'
      }))
    }
  }))), createSearchApplicationStatus === _api.Status.ERROR && createSearchApplicationError && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    color: "danger",
    title: _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.header.createError.title', {
      defaultMessage: 'Error creating search application'
    })
  }, (0, _handle_api_errors.getErrorsFromHttpResponse)(createSearchApplicationError).map((errMessage, i) => /*#__PURE__*/_react.default.createElement("p", {
    id: `createErrorMsg.${i}`
  }, errMessage))))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.featureCallOut.title', {
      defaultMessage: 'Beta feature'
    }),
    color: "warning",
    iconType: "beaker"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.enterpriseSearch.searchApplications.createSearchApplication.featureCallOut.description",
    defaultMessage: "This functionality is in beta and is subject to change. The design and code is less mature than official GA features and is being provided as-is with no warranties. Beta features are not subject to the support SLA of official GA features."
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSteps, {
    steps: [{
      children: /*#__PURE__*/_react.default.createElement(_indices_select_combobox.IndicesSelectComboBox, {
        fullWidth: true,
        isDisabled: formDisabled,
        onChange: onIndicesChange,
        selectedOptions: selectedIndices.map(index => (0, _indices_select_combobox.indexToOption)(index))
      }),
      status: indicesStatus,
      title: _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.selectIndices.title', {
        defaultMessage: 'Select indices'
      })
    }, {
      children: /*#__PURE__*/_react.default.createElement(_eui.EuiFieldText, {
        fullWidth: true,
        disabled: formDisabled,
        placeholder: _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.searchApplicationName.placeholder', {
          defaultMessage: 'Search application name'
        }),
        value: searchApplicationName,
        onChange: e => setName(e.target.value)
      }),
      status: searchApplicationNameStatus,
      title: _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.searchApplicationName.title', {
        defaultMessage: 'Name your search application'
      })
    }]
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutFooter, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    disabled: formDisabled,
    "data-telemetry-id": "entSearchApplications-createSearchApplication-cancel",
    onClick: onClose
  }, _constants.CANCEL_BUTTON_LABEL)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    isDisabled: createDisabled || formDisabled,
    "data-telemetry-id": "entSearchApplications-createSearchApplication-submit",
    fill: true,
    iconType: "plusInCircle",
    onClick: () => createSearchApplication()
  }, _i18n.i18n.translate('xpack.enterpriseSearch.searchApplications.createSearchApplication.submit', {
    defaultMessage: 'Create'
  }))))));
};
exports.CreateSearchApplication = CreateSearchApplication;
