// @kbn/security-solution-plugin/public/common/components/event_details/summary_view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SummaryView = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _translations = _interopRequireWildcard(require("./translations"));
var i18n = _translations;
var _summary_table = require("./table/summary_table");
var _summary_value_cell = require("./table/summary_value_cell");
var _prevalence_cell = require("./table/prevalence_cell");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const baseColumns = [{
  field: 'title',
  truncateText: false,
  name: i18n.HIGHLIGHTED_FIELDS_FIELD,
  textOnly: true
}, {
  field: 'description',
  truncateText: false,
  render: _summary_value_cell.SummaryValueCell,
  name: i18n.HIGHLIGHTED_FIELDS_VALUE
}];
const allColumns = [...baseColumns, {
  field: 'description',
  truncateText: true,
  render: _prevalence_cell.PrevalenceCellRenderer,
  name: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, i18n.HIGHLIGHTED_FIELDS_ALERT_PREVALENCE, ' ', /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    type: "iInCircle",
    color: "subdued",
    title: i18n.HIGHLIGHTED_FIELDS_ALERT_PREVALENCE,
    content: /*#__PURE__*/_react.default.createElement("span", null, i18n.HIGHLIGHTED_FIELDS_ALERT_PREVALENCE_TOOLTIP)
  })),
  align: 'right',
  width: '130px'
}];
const rowProps = {
  // Class name for each row. On hover of a row, all actions for that row will be shown.
  className: 'flyoutTableHoverActions'
};
const SummaryViewComponent = ({
  goToTable,
  rows,
  title,
  isReadOnly
}) => {
  const columns = isReadOnly ? baseColumns : allColumns;
  return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xxxs"
  }, /*#__PURE__*/_react.default.createElement("h5", null, title))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    onClick: goToTable,
    "data-test-subj": "summary-view-go-to-table-link"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, _translations.VIEW_ALL_FIELDS)))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_summary_table.SummaryTable, {
    "data-test-subj": "summary-view",
    items: rows,
    columns: columns,
    rowProps: rowProps,
    compressed: true
  }));
};
const SummaryView = exports.SummaryView = /*#__PURE__*/_react.default.memo(SummaryViewComponent);
