// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/test_pipeline/test_pipeline_tabs/tab_documents/reset_documents_modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResetDocumentsModal = void 0;
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const i18nTexts = {
  modalTitle: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.testPipeline.resetDocumentsModal.title', {
    defaultMessage: 'Clear documents'
  }),
  modalDescription: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.testPipeline.resetDocumentsModal.description', {
    defaultMessage: 'This will reset the output.'
  }),
  cancelButtonLabel: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.testPipeline.resetDocumentsModal.cancelButtonLabel', {
    defaultMessage: 'Cancel'
  }),
  resetButtonLabel: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.testPipeline.resetDocumentsModal.resetButtonLabel', {
    defaultMessage: 'Clear documents'
  })
};
const ResetDocumentsModal = ({
  confirmResetTestOutput,
  closeModal
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiConfirmModal, {
    buttonColor: "danger",
    "data-test-subj": "resetDocumentsConfirmationModal",
    title: i18nTexts.modalTitle,
    onCancel: closeModal,
    onConfirm: confirmResetTestOutput,
    cancelButtonText: i18nTexts.cancelButtonLabel,
    confirmButtonText: i18nTexts.resetButtonLabel
  }, /*#__PURE__*/_react.default.createElement("p", null, i18nTexts.modalDescription));
};
exports.ResetDocumentsModal = ResetDocumentsModal;
