// @kbn/synthetics-plugin/server/saved_objects/synthetics_settings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.uptimeSettingsObjectType = exports.uptimeSettingsObjectId = exports.syntheticsSettingsObjectType = exports.syntheticsSettingsObjectId = exports.syntheticsSettings = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const uptimeSettingsObjectType = exports.uptimeSettingsObjectType = 'uptime-dynamic-settings';
const uptimeSettingsObjectId = exports.uptimeSettingsObjectId = 'uptime-dynamic-settings-singleton';
const syntheticsSettingsObjectType = exports.syntheticsSettingsObjectType = 'synthetics-dynamic-settings';
const syntheticsSettingsObjectId = exports.syntheticsSettingsObjectId = 'synthetics-dynamic-settings-singleton';
const syntheticsSettings = exports.syntheticsSettings = {
  name: syntheticsSettingsObjectType,
  hidden: false,
  namespaceType: 'single',
  mappings: {
    dynamic: false,
    properties: {}
  },
  management: {
    importableAndExportable: true,
    icon: 'uptimeApp',
    getTitle: () => _i18n.i18n.translate('xpack.synthetics.settings.index', {
      defaultMessage: 'Synthetics settings'
    })
  }
};
