// @kbn/fleet-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AGENTS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_INDEX;
  }
});
Object.defineProperty(exports, "AGENTS_PREFIX", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_PREFIX;
  }
});
Object.defineProperty(exports, "AGENTS_SETUP_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_SETUP_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_ACTIONS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_ACTIONS_INDEX;
  }
});
Object.defineProperty(exports, "AGENT_ACTIONS_RESULTS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_ACTIONS_RESULTS_INDEX;
  }
});
Object.defineProperty(exports, "AGENT_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_INDEX;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "AGENT_UPDATE_LAST_CHECKIN_INTERVAL_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_UPDATE_LAST_CHECKIN_INTERVAL_MS;
  }
});
Object.defineProperty(exports, "API_VERSIONS", {
  enumerable: true,
  get: function () {
    return _constants.API_VERSIONS;
  }
});
Object.defineProperty(exports, "APP_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.APP_API_ROUTES;
  }
});
Object.defineProperty(exports, "ASSETS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.ASSETS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "DASHBOARD_LOCATORS_IDS", {
  enumerable: true,
  get: function () {
    return _constants.DASHBOARD_LOCATORS_IDS;
  }
});
Object.defineProperty(exports, "ENDPOINT_PRIVILEGES", {
  enumerable: true,
  get: function () {
    return _constants.ENDPOINT_PRIVILEGES;
  }
});
Object.defineProperty(exports, "EPM_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.EPM_API_ROUTES;
  }
});
Object.defineProperty(exports, "ElasticsearchAssetType", {
  enumerable: true,
  get: function () {
    return _types.ElasticsearchAssetType;
  }
});
Object.defineProperty(exports, "FLEET_APM_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_APM_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_CLOUD_DEFEND_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_CLOUD_DEFEND_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_CLOUD_SECURITY_POSTURE_CNVM_POLICY_TEMPLATE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_CLOUD_SECURITY_POSTURE_CNVM_POLICY_TEMPLATE;
  }
});
Object.defineProperty(exports, "FLEET_CLOUD_SECURITY_POSTURE_CSPM_POLICY_TEMPLATE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_CLOUD_SECURITY_POSTURE_CSPM_POLICY_TEMPLATE;
  }
});
Object.defineProperty(exports, "FLEET_CLOUD_SECURITY_POSTURE_KSPM_POLICY_TEMPLATE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_CLOUD_SECURITY_POSTURE_KSPM_POLICY_TEMPLATE;
  }
});
Object.defineProperty(exports, "FLEET_CLOUD_SECURITY_POSTURE_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_CLOUD_SECURITY_POSTURE_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_ELASTIC_AGENT_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_ELASTIC_AGENT_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_ENDPOINT_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_ENDPOINT_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_ENROLLMENT_API_PREFIX", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_ENROLLMENT_API_PREFIX;
  }
});
Object.defineProperty(exports, "FLEET_KUBERNETES_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_KUBERNETES_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_SERVER_ARTIFACTS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SERVER_ARTIFACTS_INDEX;
  }
});
Object.defineProperty(exports, "FLEET_SERVER_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SERVER_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_SYNTHETICS_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SYNTHETICS_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_SYSTEM_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SYSTEM_PACKAGE;
  }
});
Object.defineProperty(exports, "FleetServerAgentComponentStatuses", {
  enumerable: true,
  get: function () {
    return _constants.FleetServerAgentComponentStatuses;
  }
});
Object.defineProperty(exports, "INTEGRATIONS_PLUGIN_ID", {
  enumerable: true,
  get: function () {
    return _constants.INTEGRATIONS_PLUGIN_ID;
  }
});
Object.defineProperty(exports, "INVALID_NAMESPACE_CHARACTERS", {
  enumerable: true,
  get: function () {
    return _services.INVALID_NAMESPACE_CHARACTERS;
  }
});
Object.defineProperty(exports, "MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "OUTPUT_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.OUTPUT_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PACKAGES_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGES_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_API_ROOT", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_API_ROOT;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_API_ROUTES;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PLUGIN_ID", {
  enumerable: true,
  get: function () {
    return _constants.PLUGIN_ID;
  }
});
Object.defineProperty(exports, "PRECONFIGURATION_DELETION_RECORD_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PRECONFIGURATION_DELETION_RECORD_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "SETUP_API_ROUTE", {
  enumerable: true,
  get: function () {
    return _constants.SETUP_API_ROUTE;
  }
});
Object.defineProperty(exports, "SO_SEARCH_LIMIT", {
  enumerable: true,
  get: function () {
    return _constants.SO_SEARCH_LIMIT;
  }
});
Object.defineProperty(exports, "UNINSTALL_TOKENS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.UNINSTALL_TOKENS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "agentPolicyRouteService", {
  enumerable: true,
  get: function () {
    return _services.agentPolicyRouteService;
  }
});
Object.defineProperty(exports, "agentPolicyStatuses", {
  enumerable: true,
  get: function () {
    return _constants.agentPolicyStatuses;
  }
});
Object.defineProperty(exports, "agentRouteService", {
  enumerable: true,
  get: function () {
    return _services.agentRouteService;
  }
});
Object.defineProperty(exports, "appRoutesService", {
  enumerable: true,
  get: function () {
    return _services.appRoutesService;
  }
});
Object.defineProperty(exports, "epmRouteService", {
  enumerable: true,
  get: function () {
    return _services.epmRouteService;
  }
});
Object.defineProperty(exports, "fleetSetupRouteService", {
  enumerable: true,
  get: function () {
    return _services.fleetSetupRouteService;
  }
});
Object.defineProperty(exports, "getFileDataIndexName", {
  enumerable: true,
  get: function () {
    return _services.getFileDataIndexName;
  }
});
Object.defineProperty(exports, "getFileMetadataIndexName", {
  enumerable: true,
  get: function () {
    return _services.getFileMetadataIndexName;
  }
});
Object.defineProperty(exports, "isValidDataset", {
  enumerable: true,
  get: function () {
    return _services.isValidDataset;
  }
});
Object.defineProperty(exports, "isValidNamespace", {
  enumerable: true,
  get: function () {
    return _services.isValidNamespace;
  }
});
Object.defineProperty(exports, "packagePolicyRouteService", {
  enumerable: true,
  get: function () {
    return _services.packagePolicyRouteService;
  }
});
Object.defineProperty(exports, "setupRouteService", {
  enumerable: true,
  get: function () {
    return _services.setupRouteService;
  }
});
var _constants = require("./constants");
var _services = require("./services");
var _types = require("./types");
