// @kbn/ml-plugin/public/application/components/model_snapshots/close_job_confirm/close_job_confirm.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CloseJobConfirm = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _model_snapshots_table = require("../model_snapshots_table");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CloseJobConfirm = ({
  combinedJobState,
  hideCloseJobModalVisible,
  forceCloseJob
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiConfirmModal, {
    title: combinedJobState === _model_snapshots_table.COMBINED_JOB_STATE.OPEN_AND_RUNNING ? _i18n.i18n.translate('xpack.ml.modelSnapshotTable.closeJobConfirm.stopAndClose.title', {
      defaultMessage: 'Stop datafeed and close job?'
    }) : _i18n.i18n.translate('xpack.ml.modelSnapshotTable.closeJobConfirm.close.title', {
      defaultMessage: 'Close job?'
    }),
    onCancel: hideCloseJobModalVisible,
    onConfirm: forceCloseJob,
    cancelButtonText: _i18n.i18n.translate('xpack.ml.modelSnapshotTable.closeJobConfirm.cancelButton', {
      defaultMessage: 'Cancel'
    }),
    confirmButtonText: combinedJobState === _model_snapshots_table.COMBINED_JOB_STATE.OPEN_AND_RUNNING ? _i18n.i18n.translate('xpack.ml.modelSnapshotTable.closeJobConfirm.stopAndClose.button', {
      defaultMessage: 'Stop and close'
    }) : _i18n.i18n.translate('xpack.ml.modelSnapshotTable.closeJobConfirm.close.button', {
      defaultMessage: 'Close'
    }),
    defaultFocusedButton: "confirm"
  }, /*#__PURE__*/_react.default.createElement("p", null, combinedJobState === _model_snapshots_table.COMBINED_JOB_STATE.OPEN_AND_RUNNING && /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.modelSnapshotTable.closeJobConfirm.contentOpenAndRunning",
    defaultMessage: "Job is currently open and running."
  }), combinedJobState === _model_snapshots_table.COMBINED_JOB_STATE.OPEN_AND_STOPPED && /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.modelSnapshotTable.closeJobConfirm.contentOpen",
    defaultMessage: "Job is currently open."
  }), /*#__PURE__*/_react.default.createElement("br", null), /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.modelSnapshotTable.closeJobConfirm.content",
    defaultMessage: "Snapshot revert can only happen on jobs which are closed."
  })));
};
exports.CloseJobConfirm = CloseJobConfirm;
