// @kbn/security-solution-plugin/public/common/components/progress_inline/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProgressInline = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Wrapper = _styledComponents.default.dl`
  ${({
  theme
}) => (0, _styledComponents.css)`
    align-items: center;
    display: inline-flex;

    & > * + * {
      margin-left: ${theme.eui.euiSizeS};
    }

    .siemProgressInline__bar {
      width: 100px;
    }
  `}
`;
Wrapper.displayName = 'Wrapper';
const ProgressInline = exports.ProgressInline = /*#__PURE__*/_react.default.memo(({
  children,
  current,
  max,
  unit
}) => /*#__PURE__*/_react.default.createElement(Wrapper, {
  className: "siemProgressInline"
}, /*#__PURE__*/_react.default.createElement("dt", {
  className: "siemProgressInline__title"
}, children), /*#__PURE__*/_react.default.createElement("dd", {
  className: "siemProgressInline__bar"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiProgress, {
  color: "success",
  max: max,
  value: current
})), /*#__PURE__*/_react.default.createElement("dd", {
  className: "siemProgressInline__ratio"
}, current.toLocaleString(), '/', max.toLocaleString(), " ", unit)));
ProgressInline.displayName = 'ProgressInline';
