// @kbn/entityManager-plugin/server/lib/auth/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "canEnableEntityDiscovery", {
  enumerable: true,
  get: function () {
    return _privileges.canEnableEntityDiscovery;
  }
});
Object.defineProperty(exports, "canRunEntityDiscovery", {
  enumerable: true,
  get: function () {
    return _privileges.canRunEntityDiscovery;
  }
});
Object.defineProperty(exports, "checkIfAPIKeysAreEnabled", {
  enumerable: true,
  get: function () {
    return _api_key.checkIfAPIKeysAreEnabled;
  }
});
Object.defineProperty(exports, "checkIfEntityDiscoveryAPIKeyIsValid", {
  enumerable: true,
  get: function () {
    return _api_key.checkIfEntityDiscoveryAPIKeyIsValid;
  }
});
Object.defineProperty(exports, "deleteEntityDiscoveryAPIKey", {
  enumerable: true,
  get: function () {
    return _saved_object.deleteEntityDiscoveryAPIKey;
  }
});
Object.defineProperty(exports, "generateEntityDiscoveryAPIKey", {
  enumerable: true,
  get: function () {
    return _api_key.generateEntityDiscoveryAPIKey;
  }
});
Object.defineProperty(exports, "readEntityDiscoveryAPIKey", {
  enumerable: true,
  get: function () {
    return _saved_object.readEntityDiscoveryAPIKey;
  }
});
Object.defineProperty(exports, "saveEntityDiscoveryAPIKey", {
  enumerable: true,
  get: function () {
    return _saved_object.saveEntityDiscoveryAPIKey;
  }
});
var _api_key = require("./api_key/api_key");
var _saved_object = require("./api_key/saved_object");
var _privileges = require("./privileges");
