// @kbn/security-solution-plugin/public/resolver/view/resolver_no_process_events.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResolverNoProcessEvents = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StyledEuiCodeBlock = (0, _styledComponents.default)(_eui.EuiCodeBlock)`
  align-self: flex-start;
  display: inline-block;
`;
const StyledEuiFlexGroup = (0, _styledComponents.default)(_eui.EuiFlexGroup)`
  max-width: 600px;
  margin: 60px auto 0;
`;
const ResolverNoProcessEvents = () => /*#__PURE__*/_react.default.createElement(StyledEuiFlexGroup, {
  "data-test-subj": "resolver:no-process-events",
  direction: "column"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, null, /*#__PURE__*/_react.default.createElement("h4", null, _i18n.i18n.translate('xpack.securitySolution.resolver.noProcessEvents.title', {
  defaultMessage: 'No Process Events Found'
}))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "s"
}, _i18n.i18n.translate('xpack.securitySolution.resolver.noProcessEvents.timeRange', {
  defaultMessage: `
            The Analyze Event tool creates graphs based on process events.
            If the analyzed event does not have an associated process in the current time range,
            or stored in Elasticsearch within any time range, a graph will not be created.
            You can check for associated processes by expanding your time range.
          `
})), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "s"
}, _i18n.i18n.translate('xpack.securitySolution.resolver.noProcessEvents.dataView', {
  defaultMessage: `In case you selected a different data view,
          make sure your data view contains all of the indices that are stored in the source event at "{field}".`,
  values: {
    field: 'kibana.alert.rule.parameters.index'
  }
})), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "s"
}, _i18n.i18n.translate('xpack.securitySolution.resolver.noProcessEvents.eventCategory', {
  defaultMessage: `You may also add the below to your timeline query to check for process events.
            If none are listed, a graph cannot be created from events found in that query.`
})), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(StyledEuiCodeBlock, {
  language: "html",
  paddingSize: "s",
  isCopyable: true
}, 'event.category: "process"'));
exports.ResolverNoProcessEvents = ResolverNoProcessEvents;
