// @kbn/transform-plugin/server/routes/api/transforms_single/register_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerRoute = registerRoute;
var _constants = require("../../../../common/constants");
var _common = require("../../../../common/api_schemas/common");
var _route_handler = require("./route_handler");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function registerRoute({
  router,
  getLicense
}) {
  /**
   * @apiGroup Transforms
   *
   * @api {get} /internal/transform/transforms/:transformId Get transform
   * @apiName GetTransform
   * @apiDescription Returns a single transform
   *
   * @apiSchema (params) transformIdParamSchema
   */
  router.versioned.get({
    path: (0, _constants.addInternalBasePath)('transforms/{transformId}'),
    access: 'internal'
  }).addVersion({
    version: '1',
    validate: {
      request: {
        params: _common.transformIdParamSchema
      }
    }
  }, async (ctx, request, response) => {
    const license = await getLicense();
    return license.guardApiRoute(_route_handler.routeHandler)(ctx, request, response);
  });
}
