// @kbn/apm-plugin/public/components/shared/stacktrace/frame_heading_renderers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CSharpFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _c_sharp_frame_heading_renderer.CSharpFrameHeadingRenderer;
  }
});
Object.defineProperty(exports, "DefaultFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _default_frame_heading_renderer.DefaultFrameHeadingRenderer;
  }
});
Object.defineProperty(exports, "JavaFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _java_frame_heading_renderer.JavaFrameHeadingRenderer;
  }
});
Object.defineProperty(exports, "JavaScriptFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _java_script_frame_heading_renderer.JavaScriptFrameHeadingRenderer;
  }
});
Object.defineProperty(exports, "PhpFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _php_frame_heading_renderer.PhpFrameHeadingRenderer;
  }
});
Object.defineProperty(exports, "RubyFrameHeadingRenderer", {
  enumerable: true,
  get: function () {
    return _ruby_frame_heading_renderer.RubyFrameHeadingRenderer;
  }
});
var _c_sharp_frame_heading_renderer = require("./c_sharp_frame_heading_renderer");
var _default_frame_heading_renderer = require("./default_frame_heading_renderer");
var _java_frame_heading_renderer = require("./java_frame_heading_renderer");
var _java_script_frame_heading_renderer = require("./java_script_frame_heading_renderer");
var _ruby_frame_heading_renderer = require("./ruby_frame_heading_renderer");
var _php_frame_heading_renderer = require("./php_frame_heading_renderer");
