// @kbn/security-solution-plugin/public/timelines/links.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.links = void 0;
var _i18n = require("@kbn/i18n");
var _constants = require("../../common/constants");
var _translations = require("../app/translations");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const links = exports.links = {
  id: _constants.SecurityPageName.timelines,
  title: _translations.TIMELINES,
  path: _constants.TIMELINES_PATH,
  globalNavPosition: 7,
  capabilities: [`${_constants.SERVER_APP_ID}.show`],
  globalSearchKeywords: [_i18n.i18n.translate('xpack.securitySolution.appLinks.timelines', {
    defaultMessage: 'Timelines'
  })],
  links: [{
    id: _constants.SecurityPageName.timelinesTemplates,
    title: _i18n.i18n.translate('xpack.securitySolution.appLinks.timeline.templates', {
      defaultMessage: 'Templates'
    }),
    path: `${_constants.TIMELINES_PATH}/template`,
    sideNavDisabled: true
  }, {
    id: _constants.SecurityPageName.notesManagement,
    title: _translations.NOTES,
    description: _i18n.i18n.translate('xpack.securitySolution.appLinks.notesManagementDescription', {
      defaultMessage: 'Visualize and delete notes.'
    }),
    path: `${_constants.TIMELINES_PATH}/notes`,
    experimentalKey: 'securitySolutionNotesEnabled'
  }]
};
