// @kbn/index-management-plugin/public/application/components/shared/components/details_panel/tab_mappings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TabMappings = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TabMappings = ({
  mappings
}) => {
  if (mappings && Object.keys(mappings).length) {
    return /*#__PURE__*/_react.default.createElement("div", {
      "data-test-subj": "mappingsTabContent"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
      isCopyable: true,
      language: "json"
    }, JSON.stringify(mappings, null, 2)));
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.idxMgmt.mappingsTab.noMappingsTitle",
      defaultMessage: "No mappings defined."
    }),
    iconType: "pin",
    "data-test-subj": "noMappingsCallout",
    size: "s"
  });
};
exports.TabMappings = TabMappings;
