// @kbn/synthetics-plugin/public/apps/synthetics/components/common/components/table_title.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableTitle = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TableTitle = ({
  pageIndex,
  pageSize,
  total,
  label
}) => {
  const start = Math.min(pageIndex * pageSize + 1, total);
  const end = Math.min(start + pageSize - 1, total);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.synthetics.tableTitle.showing",
    defaultMessage: "Showing {count} of {total} {label}",
    values: {
      count: /*#__PURE__*/_react.default.createElement("strong", null, start, "-", end),
      total,
      label: /*#__PURE__*/_react.default.createElement("strong", null, label)
    }
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    margin: "none",
    style: {
      height: 2
    }
  }));
};
exports.TableTitle = TableTitle;
