// @kbn/data-visualizer-plugin/public/application/index_data_visualizer/embeddables/grid_embeddable/embeddable_field_stats_no_results.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmbeddableNoResultsEmptyPrompt = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _react2 = require("@emotion/react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EmbeddableNoResultsEmptyPrompt = () => /*#__PURE__*/_react.default.createElement("div", {
  css: (0, _react2.css)({
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    flex: '1 0 100%',
    textAlign: 'center'
  })
}, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "xs",
  color: "subdued"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
  type: "visualizeApp",
  size: "m",
  color: "subdued"
}), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
  id: "xpack.dataVisualizer.index.embeddableNoResultsMessage",
  defaultMessage: "No results found"
})));
exports.EmbeddableNoResultsEmptyPrompt = EmbeddableNoResultsEmptyPrompt;
