// @kbn/osquery-plugin/public/routes/components/empty_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyPrompt = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _translations = require("../../shared_components/osquery_action/translations");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EmptyPromptComponent = () => /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  iconType: "logoOsquery",
  title: /*#__PURE__*/_react.default.createElement("h2", null, _translations.PERMISSION_DENIED),
  titleSize: "xs",
  body: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.results.permissionDenied",
    defaultMessage: "To access these results, ask your administrator for {osquery} Kibana privileges."
    // eslint-disable-next-line react-perf/jsx-no-new-object-as-prop
    ,
    values: {
      osquery: /*#__PURE__*/_react.default.createElement(_eui.EuiCode, null, "osquery")
    }
  })
});
const EmptyPrompt = exports.EmptyPrompt = /*#__PURE__*/_react.default.memo(EmptyPromptComponent);
