// @kbn/osquery-plugin/public/routes/components/missing_privileges.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MissingPrivileges = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const panelCss = {
  maxWidth: '500px',
  marginRight: 'auto',
  marginLeft: 'auto'
};
const MissingPrivilegesComponent = () => /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
  css: panelCss
}, /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  iconType: "securityApp",
  title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.permissionDeniedErrorTitle",
    defaultMessage: "Permission denied"
  })),
  body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.permissionDeniedErrorMessage",
    defaultMessage: "You are not authorized to access this page."
  }))
})), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null));
const MissingPrivileges = exports.MissingPrivileges = /*#__PURE__*/_react.default.memo(MissingPrivilegesComponent);
