// @kbn/canvas-plugin/public/components/element_wrapper/element_wrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElementWrapper = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _positionable = require("../positionable");
var _element_content = require("../element_content");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ElementWrapper = props => {
  const {
    renderable,
    transformMatrix,
    width,
    height,
    state,
    handlers
  } = props;
  return /*#__PURE__*/_react.default.createElement(_positionable.Positionable, {
    transformMatrix: transformMatrix,
    width: width,
    height: height
  }, /*#__PURE__*/_react.default.createElement(_element_content.ElementContent, {
    renderable: renderable,
    state: state,
    handlers: handlers,
    width: width,
    height: height
  }));
};
exports.ElementWrapper = ElementWrapper;
ElementWrapper.propTypes = {
  // positionable props (from element object)
  transformMatrix: _propTypes.default.arrayOf(_propTypes.default.number).isRequired,
  width: _propTypes.default.number.isRequired,
  height: _propTypes.default.number.isRequired,
  // ElementContent pass-through props
  renderable: _propTypes.default.object,
  state: _propTypes.default.string,
  handlers: _propTypes.default.object.isRequired
};
