// @kbn/security-solution-plugin/common/api/detection_engine/prebuilt_rules/get_prebuilt_rules_and_timelines_status/get_prebuilt_rules_and_timelines_status_route.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GetPrebuiltRulesAndTimelinesStatusResponse = void 0;
var _zod = require("zod");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Prebuilt Rules Status API endpoint
 *   version: 2023-10-31
 */

const GetPrebuiltRulesAndTimelinesStatusResponse = exports.GetPrebuiltRulesAndTimelinesStatusResponse = _zod.z.object({
  /**
   * The total number of custom rules
   */
  rules_custom_installed: _zod.z.number().int().min(0),
  /**
   * The total number of installed prebuilt rules
   */
  rules_installed: _zod.z.number().int().min(0),
  /**
   * The total number of available prebuilt rules that are not installed
   */
  rules_not_installed: _zod.z.number().int().min(0),
  /**
   * The total number of outdated prebuilt rules
   */
  rules_not_updated: _zod.z.number().int().min(0),
  /**
   * The total number of installed prebuilt timelines
   */
  timelines_installed: _zod.z.number().int().min(0),
  /**
   * The total number of available prebuilt timelines that are not installed
   */
  timelines_not_installed: _zod.z.number().int().min(0),
  /**
   * The total number of outdated prebuilt timelines
   */
  timelines_not_updated: _zod.z.number().int().min(0)
}).strict();
