// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/test_pipeline/test_pipeline_tabs/tab_output.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OutputTab = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const OutputTab = ({
  handleTestPipeline,
  isRunningTest,
  cachedVerbose,
  cachedDocuments,
  testOutput
}) => {
  const [isVerboseEnabled, setIsVerboseEnabled] = (0, _react.useState)(Boolean(cachedVerbose));
  let content;
  if (isRunningTest) {
    content = /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
      size: "m"
    });
  } else if (testOutput) {
    content = /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
      language: "json",
      isCopyable: true
    }, JSON.stringify(testOutput, null, 2));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-test-subj": "outputTabContent"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ingestPipelines.testPipelineFlyout.outputTab.descriptionText",
    defaultMessage: "View the output data, or see how each processor affects the document as it passes through the pipeline."
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "l"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    label: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.ingestPipelines.testPipelineFlyout.outputTab.verboseSwitchLabel",
      defaultMessage: "View verbose output"
    }),
    checked: isVerboseEnabled,
    "data-test-subj": "verboseOutputToggle",
    onChange: async e => {
      const isVerbose = e.target.checked;
      setIsVerboseEnabled(isVerbose);
      await handleTestPipeline({
        documents: cachedDocuments,
        verbose: isVerbose
      });
    }
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    size: "s",
    onClick: async () => await handleTestPipeline({
      documents: cachedDocuments,
      verbose: isVerboseEnabled
    }, true),
    iconType: "refresh",
    "data-test-subj": "refreshOutputButton"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ingestPipelines.testPipelineFlyout.outputTab.descriptionLinkLabel",
    defaultMessage: "Refresh output"
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), content);
};
exports.OutputTab = OutputTab;
