// @kbn/cloud-security-posture-plugin/public/pages/rules/rules_flyout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RuleFlyout = exports.RULES_FLYOUT_SWITCH_BUTTON = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _use_kibana = require("../../common/hooks/use_kibana");
var _detection_rules = require("../../../common/utils/detection_rules");
var _rule_tab = require("../configurations/findings_flyout/rule_tab");
var _overview_tab = require("../configurations/findings_flyout/overview_tab");
var TEST_SUBJECTS = _interopRequireWildcard(require("./test_subjects"));
var _use_change_csp_rule_state = require("./use_change_csp_rule_state");
var _take_action = require("../../components/take_action");
var _use_fetch_detection_rules_by_tags = require("../../common/api/use_fetch_detection_rules_by_tags");
var _create_detection_rule_from_benchmark = require("../configurations/utils/create_detection_rule_from_benchmark");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RULES_FLYOUT_SWITCH_BUTTON = exports.RULES_FLYOUT_SWITCH_BUTTON = 'rule-flyout-switch-button';
const tabs = [{
  label: _i18n.i18n.translate('xpack.csp.rules.ruleFlyout.overviewTabLabel', {
    defaultMessage: 'Overview'
  }),
  id: 'overview',
  disabled: false
}, {
  label: _i18n.i18n.translate('xpack.csp.rules.ruleFlyout.remediationTabLabel', {
    defaultMessage: 'Remediation'
  }),
  id: 'remediation',
  disabled: false
}];
const RuleFlyout = ({
  onClose,
  rule
}) => {
  const [tab, setTab] = (0, _react.useState)('overview');
  const {
    mutate: mutateRuleState
  } = (0, _use_change_csp_rule_state.useChangeCspRuleState)();
  const {
    data: rulesData
  } = (0, _use_fetch_detection_rules_by_tags.useFetchDetectionRulesByTags)((0, _detection_rules.getFindingsDetectionRuleSearchTags)(rule.metadata));
  const {
    notifications,
    analytics,
    i18n: i18nStart,
    theme
  } = (0, _use_kibana.useKibana)().services;
  const startServices = {
    notifications,
    analytics,
    i18n: i18nStart,
    theme
  };
  const isRuleMuted = (rule === null || rule === void 0 ? void 0 : rule.state) === 'muted';
  const switchRuleStates = async () => {
    if (rule.metadata.benchmark.rule_number) {
      const rulesObjectRequest = {
        benchmark_id: rule.metadata.benchmark.id,
        benchmark_version: rule.metadata.benchmark.version,
        rule_number: rule.metadata.benchmark.rule_number,
        rule_id: rule.metadata.id
      };
      const nextRuleStates = isRuleMuted ? 'unmute' : 'mute';
      await mutateRuleState({
        newState: nextRuleStates,
        ruleIds: [rulesObjectRequest]
      });
      (0, _take_action.showChangeBenchmarkRuleStatesSuccessToast)(startServices, isRuleMuted, {
        numberOfRules: 1,
        numberOfDetectionRules: (rulesData === null || rulesData === void 0 ? void 0 : rulesData.total) || 0
      });
    }
  };
  const createMisconfigurationRuleFn = async http => await (0, _create_detection_rule_from_benchmark.createDetectionRuleFromBenchmarkRule)(http, rule.metadata);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyout, {
    onClose: onClose,
    "data-test-subj": TEST_SUBJECTS.CSP_RULES_FLYOUT_CONTAINER,
    outsideClickCloses: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "l"
  }, /*#__PURE__*/_react.default.createElement("h2", null, rule.metadata.name)), /*#__PURE__*/_react.default.createElement(_eui.EuiTabs, null, tabs.map(item => /*#__PURE__*/_react.default.createElement(_eui.EuiTab, {
    key: item.id,
    isSelected: tab === item.id,
    onClick: () => setTab(item.id),
    disabled: item.disabled
  }, item.label)))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, tab === 'overview' && /*#__PURE__*/_react.default.createElement(RuleOverviewTab, {
    rule: rule.metadata,
    ruleData: rule,
    switchRuleStates: switchRuleStates
  }), tab === 'remediation' && /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    compressed: false,
    listItems: (0, _overview_tab.getRemediationList)(rule.metadata)
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutFooter, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none",
    direction: "rowReverse"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, isRuleMuted ? /*#__PURE__*/_react.default.createElement(_take_action.TakeAction, {
    enableBenchmarkRuleFn: switchRuleStates,
    createRuleFn: createMisconfigurationRuleFn,
    isCreateDetectionRuleDisabled: true
  }) : /*#__PURE__*/_react.default.createElement(_take_action.TakeAction, {
    disableBenchmarkRuleFn: switchRuleStates,
    createRuleFn: createMisconfigurationRuleFn,
    isCreateDetectionRuleDisabled: false
  })))));
};
exports.RuleFlyout = RuleFlyout;
const RuleOverviewTab = ({
  rule,
  ruleData,
  switchRuleStates
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  direction: "column"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
  listItems: [...ruleState(ruleData, switchRuleStates), ...(0, _rule_tab.getRuleList)(rule, ruleData.state)]
})));
const ruleState = (rule, switchRuleStates) => [{
  title: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "xs",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.csp.rules.rulesFlyout.ruleStateSwitchTitle",
    defaultMessage: "Enabled"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    css: {
      '.euiToolTipAnchor': {
        display: 'flex' // needed to align the icon with the title
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: _i18n.i18n.translate('xpack.csp.rules.rulesFlyout.ruleStateSwitchTooltip', {
      defaultMessage: `Disabling a rule will also disable its associated detection rules and alerts. Enabling it again does not automatically re-enable them`
    })
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    size: "m",
    color: "subdued",
    type: "iInCircle"
  })))),
  description: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    className: "eui-textTruncate",
    checked: (rule === null || rule === void 0 ? void 0 : rule.state) !== 'muted',
    onChange: switchRuleStates,
    "data-test-subj": RULES_FLYOUT_SWITCH_BUTTON,
    label: " "
  }))
}];
