// @kbn/maps-plugin/public/legacy_visualizations/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GEOHASH_GRID", {
  enumerable: true,
  get: function () {
    return _tile_map.GEOHASH_GRID;
  }
});
Object.defineProperty(exports, "createRegionMapFn", {
  enumerable: true,
  get: function () {
    return _region_map.createRegionMapFn;
  }
});
Object.defineProperty(exports, "createTileMapFn", {
  enumerable: true,
  get: function () {
    return _tile_map.createTileMapFn;
  }
});
Object.defineProperty(exports, "getGeoHashBucketAgg", {
  enumerable: true,
  get: function () {
    return _tile_map.getGeoHashBucketAgg;
  }
});
Object.defineProperty(exports, "regionMapRenderer", {
  enumerable: true,
  get: function () {
    return _region_map.regionMapRenderer;
  }
});
Object.defineProperty(exports, "regionMapVisType", {
  enumerable: true,
  get: function () {
    return _region_map.regionMapVisType;
  }
});
Object.defineProperty(exports, "tileMapRenderer", {
  enumerable: true,
  get: function () {
    return _tile_map.tileMapRenderer;
  }
});
Object.defineProperty(exports, "tileMapVisType", {
  enumerable: true,
  get: function () {
    return _tile_map.tileMapVisType;
  }
});
var _tile_map = require("./tile_map");
var _region_map = require("./region_map");
