// @kbn/vis-type-vega-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUsageCollectionStart = exports.setThemeService = exports.setNotifications = exports.setMapsEms = exports.setInjectedVars = exports.setDocLinks = exports.setDataViews = exports.setData = exports.getUsageCollectionStart = exports.getThemeService = exports.getNotifications = exports.getMapsEms = exports.getInjectedVars = exports.getEnableExternalUrls = exports.getDocLinks = exports.getDataViews = exports.getData = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getData, setData] = (0, _public.createGetterSetter)('Data');
exports.setData = setData;
exports.getData = getData;
const [getDataViews, setDataViews] = (0, _public.createGetterSetter)('DataViews');
exports.setDataViews = setDataViews;
exports.getDataViews = getDataViews;
const [getNotifications, setNotifications] = (0, _public.createGetterSetter)('Notifications');
exports.setNotifications = setNotifications;
exports.getNotifications = getNotifications;
const [getMapsEms, setMapsEms] = (0, _public.createGetterSetter)('mapsEms');
exports.setMapsEms = setMapsEms;
exports.getMapsEms = getMapsEms;
const [getInjectedVars, setInjectedVars] = (0, _public.createGetterSetter)('InjectedVars');
exports.setInjectedVars = setInjectedVars;
exports.getInjectedVars = getInjectedVars;
const getEnableExternalUrls = () => getInjectedVars().enableExternalUrls;
exports.getEnableExternalUrls = getEnableExternalUrls;
const [getDocLinks, setDocLinks] = (0, _public.createGetterSetter)('docLinks');
exports.setDocLinks = setDocLinks;
exports.getDocLinks = getDocLinks;
const [getUsageCollectionStart, setUsageCollectionStart] = (0, _public.createGetterSetter)('UsageCollection');
exports.setUsageCollectionStart = setUsageCollectionStart;
exports.getUsageCollectionStart = getUsageCollectionStart;
const [getThemeService, setThemeService] = (0, _public.createGetterSetter)('ThemeServiceStart');
exports.setThemeService = setThemeService;
exports.getThemeService = getThemeService;
