// @kbn/cross-cluster-replication-plugin/public/app/store/action_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FOLLOWER_INDEX_UNFOLLOW = exports.FOLLOWER_INDEX_SELECT_EDIT = exports.FOLLOWER_INDEX_SELECT_DETAIL = exports.FOLLOWER_INDEX_RESUME = exports.FOLLOWER_INDEX_PAUSE = exports.FOLLOWER_INDEX_LOAD = exports.FOLLOWER_INDEX_GET = exports.FOLLOWER_INDEX_CREATE = exports.AUTO_FOLLOW_STATS_LOAD = exports.AUTO_FOLLOW_PATTERN_UPDATE = exports.AUTO_FOLLOW_PATTERN_SELECT_EDIT = exports.AUTO_FOLLOW_PATTERN_SELECT_DETAIL = exports.AUTO_FOLLOW_PATTERN_RESUME = exports.AUTO_FOLLOW_PATTERN_PAUSE = exports.AUTO_FOLLOW_PATTERN_LOAD = exports.AUTO_FOLLOW_PATTERN_GET = exports.AUTO_FOLLOW_PATTERN_DELETE = exports.AUTO_FOLLOW_PATTERN_CREATE = exports.API_REQUEST_START = exports.API_REQUEST_END = exports.API_ERROR_SET = exports.API = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// Api
const API = exports.API = 'API';
const API_REQUEST_START = exports.API_REQUEST_START = 'API_REQUEST_START';
const API_REQUEST_END = exports.API_REQUEST_END = 'API_REQUEST_END';
const API_ERROR_SET = exports.API_ERROR_SET = 'API_ERROR_SET';

// Auto Follow Pattern
const AUTO_FOLLOW_PATTERN_SELECT_DETAIL = exports.AUTO_FOLLOW_PATTERN_SELECT_DETAIL = 'AUTO_FOLLOW_PATTERN_SELECT_DETAIL';
const AUTO_FOLLOW_PATTERN_SELECT_EDIT = exports.AUTO_FOLLOW_PATTERN_SELECT_EDIT = 'AUTO_FOLLOW_PATTERN_SELECT_EDIT';
const AUTO_FOLLOW_PATTERN_LOAD = exports.AUTO_FOLLOW_PATTERN_LOAD = 'AUTO_FOLLOW_PATTERN_LOAD';
const AUTO_FOLLOW_PATTERN_GET = exports.AUTO_FOLLOW_PATTERN_GET = 'AUTO_FOLLOW_PATTERN_GET';
const AUTO_FOLLOW_PATTERN_CREATE = exports.AUTO_FOLLOW_PATTERN_CREATE = 'AUTO_FOLLOW_PATTERN_CREATE';
const AUTO_FOLLOW_PATTERN_UPDATE = exports.AUTO_FOLLOW_PATTERN_UPDATE = 'AUTO_FOLLOW_PATTERN_UPDATE';
const AUTO_FOLLOW_PATTERN_DELETE = exports.AUTO_FOLLOW_PATTERN_DELETE = 'AUTO_FOLLOW_PATTERN_DELETE';
const AUTO_FOLLOW_PATTERN_PAUSE = exports.AUTO_FOLLOW_PATTERN_PAUSE = 'AUTO_FOLLOW_PATTERN_PAUSE';
const AUTO_FOLLOW_PATTERN_RESUME = exports.AUTO_FOLLOW_PATTERN_RESUME = 'AUTO_FOLLOW_PATTERN_RESUME';

// Follower index
const FOLLOWER_INDEX_SELECT_DETAIL = exports.FOLLOWER_INDEX_SELECT_DETAIL = 'FOLLOWER_INDEX_SELECT_DETAIL';
const FOLLOWER_INDEX_SELECT_EDIT = exports.FOLLOWER_INDEX_SELECT_EDIT = 'FOLLOWER_INDEX_SELECT_EDIT';
const FOLLOWER_INDEX_LOAD = exports.FOLLOWER_INDEX_LOAD = 'FOLLOWER_INDEX_LOAD';
const FOLLOWER_INDEX_GET = exports.FOLLOWER_INDEX_GET = 'FOLLOWER_INDEX_GET';
const FOLLOWER_INDEX_CREATE = exports.FOLLOWER_INDEX_CREATE = 'FOLLOWER_INDEX_CREATE';
const FOLLOWER_INDEX_PAUSE = exports.FOLLOWER_INDEX_PAUSE = 'FOLLOWER_INDEX_PAUSE';
const FOLLOWER_INDEX_RESUME = exports.FOLLOWER_INDEX_RESUME = 'FOLLOWER_INDEX_RESUME';
const FOLLOWER_INDEX_UNFOLLOW = exports.FOLLOWER_INDEX_UNFOLLOW = 'FOLLOWER_INDEX_UNFOLLOW';

// Stats
const AUTO_FOLLOW_STATS_LOAD = exports.AUTO_FOLLOW_STATS_LOAD = 'AUTO_FOLLOW_STATS_LOAD';
