// @kbn/security-solution-plugin/public/detection_engine/rule_creation/components/step_rule_actions/notification_action.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FrequencyDescription = void 0;
exports.NotificationAction = NotificationAction;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _helpers = require("../../../rule_creation_ui/pages/rule_creation/helpers");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DescriptionLine = ({
  children
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  size: "xs",
  color: "subdued"
}, children));
const FrequencyDescription = ({
  frequency
}) => {
  if (!frequency) {
    return null;
  }
  if (!frequency.summary) {
    return /*#__PURE__*/_react.default.createElement(DescriptionLine, null, i18n.FOR_EACH_ALERT_PER_RULE_RUN);
  }
  if (frequency.notifyWhen === 'onActiveAlert') {
    return /*#__PURE__*/_react.default.createElement(DescriptionLine, null, i18n.SUMMARY_OF_ALERTS_PER_RULE_RUN);
  }
  if (!frequency.throttle) {
    return null;
  }
  const {
    unit,
    value
  } = (0, _helpers.getTimeTypeValue)(frequency.throttle);
  const messagesByUnit = {
    s: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.detectionEngine.ruleDetails.onceInEverySecondsLabel",
      defaultMessage: "Once {secondsCount, plural, one {a} other {in every}} {secondsCount, plural, one {second} other {# seconds}}",
      values: {
        secondsCount: value
      }
    }),
    m: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.detectionEngine.ruleDetails.onceInEveryMinutesLabel",
      defaultMessage: "Once {minutesCount, plural, one {a} other {in every}} {minutesCount, plural, one {minute} other {# minutes}}",
      values: {
        minutesCount: value
      }
    }),
    h: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.detectionEngine.ruleDetails.onceInEveryHoursLabel",
      defaultMessage: "Once {hoursCount, plural, one {an} other {in every}} {hoursCount, plural, one {hour} other {# hours}}",
      values: {
        hoursCount: value
      }
    }),
    d: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.detectionEngine.ruleDetails.onceInEveryDaysLabel",
      defaultMessage: "Once {daysCount, plural, one {a} other {in every}} {daysCount, plural, one {day} other {# days}}",
      values: {
        daysCount: value
      }
    })
  };
  return /*#__PURE__*/_react.default.createElement(DescriptionLine, null, messagesByUnit[unit] || i18n.PERIODICALLY);
};
exports.FrequencyDescription = FrequencyDescription;
function NotificationAction({
  action,
  connectorTypes,
  connectors,
  actionTypeRegistry
}) {
  var _connectorType$name, _connector$name, _actionTypeRegistry$g, _actionTypeRegistry$g2;
  const connectorType = connectorTypes.find(({
    id
  }) => id === action.actionTypeId);
  const connectorTypeName = (_connectorType$name = connectorType === null || connectorType === void 0 ? void 0 : connectorType.name) !== null && _connectorType$name !== void 0 ? _connectorType$name : '';
  const connector = connectors.find(({
    id
  }) => id === action.id);
  const connectorName = (_connector$name = connector === null || connector === void 0 ? void 0 : connector.name) !== null && _connector$name !== void 0 ? _connector$name : '';
  const iconType = (_actionTypeRegistry$g = (_actionTypeRegistry$g2 = actionTypeRegistry.get(action.actionTypeId)) === null || _actionTypeRegistry$g2 === void 0 ? void 0 : _actionTypeRegistry$g2.iconClass) !== null && _actionTypeRegistry$g !== void 0 ? _actionTypeRegistry$g : 'apps';
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "s",
    component: "span",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: connectorTypeName,
    anchorClassName: "eui-textTruncate"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    size: "m",
    type: iconType
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, connectorName), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "xs",
    component: "span",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    size: "s",
    type: "bell",
    color: "subdued"
  })), /*#__PURE__*/_react.default.createElement(FrequencyDescription, {
    frequency: action.frequency
  })))));
}
