// @kbn/console-plugin/common/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "API_BASE_PATH", {
  enumerable: true,
  get: function () {
    return _api.API_BASE_PATH;
  }
});
Object.defineProperty(exports, "AUTOCOMPLETE_DEFINITIONS_FOLDER", {
  enumerable: true,
  get: function () {
    return _autocomplete_definitions.AUTOCOMPLETE_DEFINITIONS_FOLDER;
  }
});
Object.defineProperty(exports, "DEFAULT_INPUT_VALUE", {
  enumerable: true,
  get: function () {
    return _editor_input.DEFAULT_INPUT_VALUE;
  }
});
Object.defineProperty(exports, "DEFAULT_VARIABLES", {
  enumerable: true,
  get: function () {
    return _variables.DEFAULT_VARIABLES;
  }
});
Object.defineProperty(exports, "GENERATED_SUBFOLDER", {
  enumerable: true,
  get: function () {
    return _autocomplete_definitions.GENERATED_SUBFOLDER;
  }
});
Object.defineProperty(exports, "KIBANA_API_PREFIX", {
  enumerable: true,
  get: function () {
    return _api.KIBANA_API_PREFIX;
  }
});
Object.defineProperty(exports, "MAJOR_VERSION", {
  enumerable: true,
  get: function () {
    return _plugin.MAJOR_VERSION;
  }
});
Object.defineProperty(exports, "MANUAL_SUBFOLDER", {
  enumerable: true,
  get: function () {
    return _autocomplete_definitions.MANUAL_SUBFOLDER;
  }
});
Object.defineProperty(exports, "OVERRIDES_SUBFOLDER", {
  enumerable: true,
  get: function () {
    return _autocomplete_definitions.OVERRIDES_SUBFOLDER;
  }
});
var _plugin = require("./plugin");
var _api = require("./api");
var _variables = require("./variables");
var _autocomplete_definitions = require("./autocomplete_definitions");
var _editor_input = require("./editor_input");
