// @kbn/charts-plugin/common/static/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ColorSchemas: true,
  vislibColorMaps: true,
  colorSchemas: true,
  getHeatmapColors: true,
  ColorMode: true,
  LabelRotation: true,
  defaultCountLabel: true
};
Object.defineProperty(exports, "ColorMode", {
  enumerable: true,
  get: function () {
    return _components.ColorMode;
  }
});
Object.defineProperty(exports, "ColorSchemas", {
  enumerable: true,
  get: function () {
    return _color_maps.ColorSchemas;
  }
});
Object.defineProperty(exports, "LabelRotation", {
  enumerable: true,
  get: function () {
    return _components.LabelRotation;
  }
});
Object.defineProperty(exports, "colorSchemas", {
  enumerable: true,
  get: function () {
    return _color_maps.colorSchemas;
  }
});
Object.defineProperty(exports, "defaultCountLabel", {
  enumerable: true,
  get: function () {
    return _components.defaultCountLabel;
  }
});
Object.defineProperty(exports, "getHeatmapColors", {
  enumerable: true,
  get: function () {
    return _color_maps.getHeatmapColors;
  }
});
Object.defineProperty(exports, "vislibColorMaps", {
  enumerable: true,
  get: function () {
    return _color_maps.vislibColorMaps;
  }
});
var _color_maps = require("./color_maps");
var _components = require("./components");
var _styles = require("./styles");
Object.keys(_styles).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _styles[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _styles[key];
    }
  });
});
