// @kbn/infra-plugin/public/components/asset_details/tabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Anomalies", {
  enumerable: true,
  get: function () {
    return _anomalies.Anomalies;
  }
});
Object.defineProperty(exports, "Dashboards", {
  enumerable: true,
  get: function () {
    return _dashboards.Dashboards;
  }
});
Object.defineProperty(exports, "Logs", {
  enumerable: true,
  get: function () {
    return _logs.Logs;
  }
});
Object.defineProperty(exports, "Metadata", {
  enumerable: true,
  get: function () {
    return _metadata.Metadata;
  }
});
Object.defineProperty(exports, "Metrics", {
  enumerable: true,
  get: function () {
    return _metrics.Metrics;
  }
});
Object.defineProperty(exports, "Osquery", {
  enumerable: true,
  get: function () {
    return _osquery.Osquery;
  }
});
Object.defineProperty(exports, "Overview", {
  enumerable: true,
  get: function () {
    return _overview.Overview;
  }
});
Object.defineProperty(exports, "Processes", {
  enumerable: true,
  get: function () {
    return _processes.Processes;
  }
});
Object.defineProperty(exports, "Profiling", {
  enumerable: true,
  get: function () {
    return _profiling.Profiling;
  }
});
var _anomalies = require("./anomalies/anomalies");
var _metadata = require("./metadata/metadata");
var _processes = require("./processes/processes");
var _profiling = require("./profiling/profiling");
var _osquery = require("./osquery/osquery");
var _logs = require("./logs/logs");
var _overview = require("./overview/overview");
var _metrics = require("./metrics/metrics");
var _dashboards = require("./dashboards/dashboards");
