// @kbn/monitoring-plugin/public/components/metricbeat_migration/instruction_steps/apm/disable_internal_collection_instructions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getApmInstructionsForDisablingInternalCollection = getApmInstructionsForDisablingInternalCollection;
var _i18n = require("@kbn/i18n");
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _monospace = require("../components/monospace");
var _i18nReact = require("@kbn/i18n-react");
var _common_instructions = require("../common_instructions");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function getApmInstructionsForDisablingInternalCollection(product, meta) {
  const disableInternalCollectionStep = {
    title: _i18n.i18n.translate('xpack.monitoring.metricbeatMigration.apmInstructions.disableInternalCollection.title', {
      defaultMessage: "Disable self monitoring of the APM server's monitoring metrics"
    }),
    children: /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.monitoring.metricbeatMigration.apmInstructions.disableInternalCollection.description",
      defaultMessage: "Add the following setting in the APM server's configuration file ({file}):",
      values: {
        file: /*#__PURE__*/_react.default.createElement(_monospace.Monospace, null, "apm-server.yml")
      }
    }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
      isCopyable: true,
      language: "bash"
    }, "monitoring.enabled: false"), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.monitoring.metricbeatMigration.apmInstructions.disableInternalCollection.note",
      defaultMessage: "You'll need to restart the APM server after making this change."
    }))))
  };
  const migrationStatusStep = (0, _common_instructions.getDisableStatusStep)(product, meta);
  return [disableInternalCollectionStep, migrationStatusStep];
}
