// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/styles/card_item.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useCardItemStyles = exports.SHADOW_ANIMATION_DURATION = void 0;
var _eui = require("@elastic/eui");
var _css = require("@emotion/css");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SHADOW_ANIMATION_DURATION = exports.SHADOW_ANIMATION_DURATION = 350;
const useCardItemStyles = () => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const shadow = (0, _eui.useEuiShadow)('l');
  return (0, _css.css)({
    '&.card-item': {
      padding: euiTheme.size.base,
      borderRadius: euiTheme.size.s,
      border: `1px solid ${euiTheme.colors.lightShade}`,
      boxSizing: `content-box`
    },
    '&:hover, &.card-expanded': {
      boxShadow: shadow,
      transition: `box-shadow ${SHADOW_ANIMATION_DURATION}ms ease-out`
    },
    '&.card-expanded': {
      border: `2px solid #6092c0`
    }
  });
};
exports.useCardItemStyles = useCardItemStyles;
