// @kbn/security-solution-plugin/common/api/search_strategy/network/model/top_tables_fields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.topTablesFields = exports.NetworkTopTablesFields = void 0;
var _zod = require("zod");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let NetworkTopTablesFields = exports.NetworkTopTablesFields = /*#__PURE__*/function (NetworkTopTablesFields) {
  NetworkTopTablesFields["bytes_in"] = "bytes_in";
  NetworkTopTablesFields["bytes_out"] = "bytes_out";
  NetworkTopTablesFields["flows"] = "flows";
  NetworkTopTablesFields["destination_ips"] = "destination_ips";
  NetworkTopTablesFields["source_ips"] = "source_ips";
  return NetworkTopTablesFields;
}({});
const topTablesFields = exports.topTablesFields = _zod.z.enum([NetworkTopTablesFields.bytes_in, NetworkTopTablesFields.bytes_out, NetworkTopTablesFields.flows, NetworkTopTablesFields.destination_ips, NetworkTopTablesFields.source_ips]);
