// @kbn/security-solution-plugin/common/endpoint/service/artifacts/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY_TEXT = exports.PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY = exports.GLOBAL_ARTIFACT_TAG = exports.FILTER_PROCESS_DESCENDANTS_TAG = exports.DEFAULT_EXCEPTION_LIST_ITEM_SEARCHABLE_FIELDS = exports.BY_POLICY_ARTIFACT_TAG_PREFIX = exports.ALL_ENDPOINT_ARTIFACT_LIST_IDS = void 0;
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
var _securitysolutionListUtils = require("@kbn/securitysolution-list-utils");
var _EVENT_FILTERS_OPERAT;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const BY_POLICY_ARTIFACT_TAG_PREFIX = exports.BY_POLICY_ARTIFACT_TAG_PREFIX = 'policy:';
const GLOBAL_ARTIFACT_TAG = exports.GLOBAL_ARTIFACT_TAG = `${BY_POLICY_ARTIFACT_TAG_PREFIX}all`;
const FILTER_PROCESS_DESCENDANTS_TAG = exports.FILTER_PROCESS_DESCENDANTS_TAG = 'filter_process_descendants';
const PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY = exports.PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY = Object.freeze({
  field: 'event.category',
  operator: 'included',
  type: 'match',
  value: 'process'
});
const PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY_TEXT = exports.PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY_TEXT = `${PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY.field} ${(_EVENT_FILTERS_OPERAT = _securitysolutionListUtils.EVENT_FILTERS_OPERATORS.find(({
  type
}) => type === PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY.type)) === null || _EVENT_FILTERS_OPERAT === void 0 ? void 0 : _EVENT_FILTERS_OPERAT.message} ${PROCESS_DESCENDANT_EVENT_FILTER_EXTRA_ENTRY.value}`;

// TODO: refact all uses of `ALL_ENDPOINT_ARTIFACTS_LIST_IDS to sue new const from shared package
const ALL_ENDPOINT_ARTIFACT_LIST_IDS = exports.ALL_ENDPOINT_ARTIFACT_LIST_IDS = _securitysolutionListConstants.ENDPOINT_ARTIFACT_LIST_IDS;
const DEFAULT_EXCEPTION_LIST_ITEM_SEARCHABLE_FIELDS = exports.DEFAULT_EXCEPTION_LIST_ITEM_SEARCHABLE_FIELDS = [`name`, `description`, `entries.value`, `entries.entries.value`, `item_id`];
