// @kbn/uptime-plugin/public/legacy_uptime/components/fleet_package/synthetics_custom_assets_extension.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SyntheticsCustomAssetsExtension = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _public = require("@kbn/fleet-plugin/public");
var _public2 = require("@kbn/kibana-react-plugin/public");
var _plugin = require("../../../../common/constants/plugin");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SyntheticsCustomAssetsExtension = () => {
  var _http$basePath$prepen;
  const {
    http
  } = (0, _public2.useKibana)().services;
  const views = [{
    name: _i18n.i18n.translate('xpack.uptime.fleetIntegration.assets.name', {
      defaultMessage: 'Monitors'
    }),
    url: (_http$basePath$prepen = http === null || http === void 0 ? void 0 : http.basePath.prepend(`/app/${_plugin.PLUGIN.SYNTHETICS_PLUGIN_ID}`)) !== null && _http$basePath$prepen !== void 0 ? _http$basePath$prepen : '',
    description: _i18n.i18n.translate('xpack.uptime.fleetIntegration.assets.description', {
      defaultMessage: 'View monitors in Synthetics'
    })
  }];
  return /*#__PURE__*/_react.default.createElement(_public.CustomAssetsAccordion, {
    views: views,
    initialIsOpen: true
  });
};
exports.SyntheticsCustomAssetsExtension = SyntheticsCustomAssetsExtension;
