// @kbn/security-solution-plugin/public/explore/network/components/port/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Port = void 0;
var _react = _interopRequireDefault(require("react"));
var _draggables = require("../../../../common/components/draggables");
var _empty_value = require("../../../../common/components/empty_value");
var _links = require("../../../../common/components/links");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Port = exports.Port = /*#__PURE__*/_react.default.memo(({
  Component,
  contextId,
  eventId,
  fieldName,
  fieldType,
  isAggregatable,
  isDraggable,
  title,
  value
}) => isDraggable ? /*#__PURE__*/_react.default.createElement(_draggables.DefaultDraggable, {
  "data-test-subj": "port",
  field: fieldName,
  fieldType: fieldType,
  isAggregatable: isAggregatable,
  id: `port-default-draggable-${contextId}-${eventId}-${fieldName}-${value}`,
  isDraggable: isDraggable,
  tooltipContent: fieldName,
  value: value
}, /*#__PURE__*/_react.default.createElement(_links.PortOrServiceNameLink, {
  portOrServiceName: value || (0, _empty_value.getEmptyValue)(),
  Component: Component,
  title: title
})) : /*#__PURE__*/_react.default.createElement(_links.PortOrServiceNameLink, {
  portOrServiceName: value || (0, _empty_value.getEmptyValue)(),
  Component: Component,
  title: title
}));
Port.displayName = 'Port';
