// @kbn/observability-ai-assistant-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AssistantAvatar", {
  enumerable: true,
  get: function () {
    return _assistant_avatar.AssistantAvatar;
  }
});
Object.defineProperty(exports, "ChatActionClickType", {
  enumerable: true,
  get: function () {
    return _types.ChatActionClickType;
  }
});
Object.defineProperty(exports, "ChatItemControls", {
  enumerable: true,
  get: function () {
    return _chat_item_controls.ChatItemControls;
  }
});
Object.defineProperty(exports, "ChatState", {
  enumerable: true,
  get: function () {
    return _use_chat.ChatState;
  }
});
Object.defineProperty(exports, "ConnectorSelectorBase", {
  enumerable: true,
  get: function () {
    return _connector_selector_base.ConnectorSelectorBase;
  }
});
Object.defineProperty(exports, "DEFAULT_LANGUAGE_OPTION", {
  enumerable: true,
  get: function () {
    return _language_options.DEFAULT_LANGUAGE_OPTION;
  }
});
Object.defineProperty(exports, "FailedToLoadResponse", {
  enumerable: true,
  get: function () {
    return _failed_to_load_response.FailedToLoadResponse;
  }
});
Object.defineProperty(exports, "FeedbackButtons", {
  enumerable: true,
  get: function () {
    return _feedback_buttons.FeedbackButtons;
  }
});
Object.defineProperty(exports, "FunctionVisibility", {
  enumerable: true,
  get: function () {
    return _common.FunctionVisibility;
  }
});
Object.defineProperty(exports, "KnowledgeBaseEntryRole", {
  enumerable: true,
  get: function () {
    return _common.KnowledgeBaseEntryRole;
  }
});
Object.defineProperty(exports, "LANGUAGE_OPTIONS", {
  enumerable: true,
  get: function () {
    return _language_options.LANGUAGE_OPTIONS;
  }
});
Object.defineProperty(exports, "MessageRole", {
  enumerable: true,
  get: function () {
    return _common.MessageRole;
  }
});
Object.defineProperty(exports, "MessageText", {
  enumerable: true,
  get: function () {
    return _message_text.MessageText;
  }
});
Object.defineProperty(exports, "ObservabilityAIAssistantTelemetryEventType", {
  enumerable: true,
  get: function () {
    return _telemetry_event_type.ObservabilityAIAssistantTelemetryEventType;
  }
});
Object.defineProperty(exports, "StreamingChatResponseEventType", {
  enumerable: true,
  get: function () {
    return _common.StreamingChatResponseEventType;
  }
});
Object.defineProperty(exports, "VISUALIZE_ESQL_USER_INTENTIONS", {
  enumerable: true,
  get: function () {
    return _visualize_esql.VISUALIZE_ESQL_USER_INTENTIONS;
  }
});
Object.defineProperty(exports, "VisualizeESQLUserIntention", {
  enumerable: true,
  get: function () {
    return _visualize_esql.VisualizeESQLUserIntention;
  }
});
Object.defineProperty(exports, "aiAssistantCapabilities", {
  enumerable: true,
  get: function () {
    return _capabilities.aiAssistantCapabilities;
  }
});
Object.defineProperty(exports, "aiAssistantLogsIndexPattern", {
  enumerable: true,
  get: function () {
    return _settings_keys.aiAssistantLogsIndexPattern;
  }
});
Object.defineProperty(exports, "aiAssistantResponseLanguage", {
  enumerable: true,
  get: function () {
    return _settings_keys.aiAssistantResponseLanguage;
  }
});
Object.defineProperty(exports, "aiAssistantSearchConnectorIndexPattern", {
  enumerable: true,
  get: function () {
    return _settings_keys.aiAssistantSearchConnectorIndexPattern;
  }
});
Object.defineProperty(exports, "aiAssistantSimulatedFunctionCalling", {
  enumerable: true,
  get: function () {
    return _settings_keys.aiAssistantSimulatedFunctionCalling;
  }
});
Object.defineProperty(exports, "concatenateChatCompletionChunks", {
  enumerable: true,
  get: function () {
    return _common.concatenateChatCompletionChunks;
  }
});
Object.defineProperty(exports, "createFunctionRequestMessage", {
  enumerable: true,
  get: function () {
    return _create_function_request_message.createFunctionRequestMessage;
  }
});
Object.defineProperty(exports, "createFunctionResponseMessage", {
  enumerable: true,
  get: function () {
    return _create_function_response_message.createFunctionResponseMessage;
  }
});
Object.defineProperty(exports, "createScreenContextAction", {
  enumerable: true,
  get: function () {
    return _create_screen_context_action.createScreenContextAction;
  }
});
Object.defineProperty(exports, "createStorybookChatService", {
  enumerable: true,
  get: function () {
    return _storybook_mock.createStorybookChatService;
  }
});
Object.defineProperty(exports, "createStorybookService", {
  enumerable: true,
  get: function () {
    return _storybook_mock.createStorybookService;
  }
});
Object.defineProperty(exports, "isSupportedConnectorType", {
  enumerable: true,
  get: function () {
    return _common.isSupportedConnectorType;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "useAbortableAsync", {
  enumerable: true,
  get: function () {
    return _use_abortable_async.useAbortableAsync;
  }
});
Object.defineProperty(exports, "useGenAIConnectorsWithoutContext", {
  enumerable: true,
  get: function () {
    return _use_genai_connectors.useGenAIConnectorsWithoutContext;
  }
});
var _plugin = require("./plugin");
var _capabilities = require("../common/capabilities");
var _assistant_avatar = require("./components/assistant_avatar");
var _connector_selector_base = require("./components/connector_selector/connector_selector_base");
var _use_abortable_async = require("./hooks/use_abortable_async");
var _use_genai_connectors = require("./hooks/use_genai_connectors");
var _storybook_mock = require("./storybook_mock");
var _create_screen_context_action = require("./utils/create_screen_context_action");
var _use_chat = require("./hooks/use_chat");
var _feedback_buttons = require("./components/buttons/feedback_buttons");
var _chat_item_controls = require("./components/chat/chat_item_controls");
var _failed_to_load_response = require("./components/message_panel/failed_to_load_response");
var _message_text = require("./components/message_panel/message_text");
var _types = require("./components/chat/types");
var _visualize_esql = require("../common/functions/visualize_esql");
var _common = require("../common");
var _telemetry_event_type = require("./analytics/telemetry_event_type");
var _create_function_request_message = require("../common/utils/create_function_request_message");
var _create_function_response_message = require("../common/utils/create_function_response_message");
var _language_options = require("../common/ui_settings/language_options");
var _settings_keys = require("../common/ui_settings/settings_keys");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const plugin = pluginInitializerContext => new _plugin.ObservabilityAIAssistantPlugin(pluginInitializerContext);
exports.plugin = plugin;
