// @kbn/security-solution-plugin/public/common/hooks/use_fetch/request_names.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.REQUEST_NAMES = void 0;
var _constants = require("../../../../common/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const REQUEST_NAMES = exports.REQUEST_NAMES = {
  SECURITY_DASHBOARDS: `${_constants.APP_UI_ID} fetch security dashboards`,
  SECURITY_TAGS: `${_constants.APP_UI_ID} fetch security tags`,
  SECURITY_CREATE_TAG: `${_constants.APP_UI_ID} fetch security create tag`,
  CTI_TAGS: `${_constants.APP_UI_ID} fetch cti tags`,
  ANOMALIES_TABLE: `${_constants.APP_UI_ID} fetch anomalies table data`,
  GET_RISK_SCORE_DEPRECATED: `${_constants.APP_UI_ID} fetch is risk score deprecated`,
  ENABLE_RISK_SCORE: `${_constants.APP_UI_ID} fetch enable risk score`,
  REFRESH_RISK_SCORE: `${_constants.APP_UI_ID} fetch refresh risk score`,
  UPGRADE_RISK_SCORE: `${_constants.APP_UI_ID} fetch upgrade risk score`
};
