// @kbn/unified-search-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ACTION_GLOBAL_APPLY_FILTER", {
  enumerable: true,
  get: function () {
    return _actions.ACTION_GLOBAL_APPLY_FILTER;
  }
});
Object.defineProperty(exports, "DataViewPicker", {
  enumerable: true,
  get: function () {
    return _dataview_picker.DataViewPicker;
  }
});
Object.defineProperty(exports, "DataViewSelector", {
  enumerable: true,
  get: function () {
    return _dataview_picker.DataViewSelector;
  }
});
Object.defineProperty(exports, "DataViewsList", {
  enumerable: true,
  get: function () {
    return _dataview_picker.DataViewsList;
  }
});
Object.defineProperty(exports, "FilterBadgeGroup", {
  enumerable: true,
  get: function () {
    return _filter_badge.FilterBadgeGroup;
  }
});
Object.defineProperty(exports, "FilterItem", {
  enumerable: true,
  get: function () {
    return _filter_bar.FilterItem;
  }
});
Object.defineProperty(exports, "FilterItems", {
  enumerable: true,
  get: function () {
    return _filter_bar.FilterItems;
  }
});
Object.defineProperty(exports, "QueryStringInput", {
  enumerable: true,
  get: function () {
    return _query_string_input.QueryStringInput;
  }
});
Object.defineProperty(exports, "QuerySuggestionTypes", {
  enumerable: true,
  get: function () {
    return _query_suggestion_provider.QuerySuggestionTypes;
  }
});
Object.defineProperty(exports, "SearchBar", {
  enumerable: true,
  get: function () {
    return _search_bar.SearchBar;
  }
});
Object.defineProperty(exports, "UPDATE_FILTER_REFERENCES_ACTION", {
  enumerable: true,
  get: function () {
    return _actions.UPDATE_FILTER_REFERENCES_ACTION;
  }
});
Object.defineProperty(exports, "UPDATE_FILTER_REFERENCES_TRIGGER", {
  enumerable: true,
  get: function () {
    return _triggers.UPDATE_FILTER_REFERENCES_TRIGGER;
  }
});
Object.defineProperty(exports, "createFilterAction", {
  enumerable: true,
  get: function () {
    return _apply_filter_action.createFilterAction;
  }
});
Object.defineProperty(exports, "createSearchBar", {
  enumerable: true,
  get: function () {
    return _create_search_bar.createSearchBar;
  }
});
exports.plugin = plugin;
var _query_string_input = require("./query_string_input");
var _search_bar = require("./search_bar");
var _filter_bar = require("./filter_bar");
var _filter_badge = require("./filter_badge");
var _dataview_picker = require("./dataview_picker");
var _actions = require("./actions");
var _triggers = require("./triggers");
var _create_search_bar = require("./search_bar/create_search_bar");
var _apply_filter_action = require("./actions/apply_filter_action");
var _query_suggestion_provider = require("./autocomplete/providers/query_suggestion_provider");
var _plugin = require("./plugin");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/*
 * Autocomplete query suggestions:
 */

// This exports static code and TypeScript types,
// as well as, Kibana Platform `plugin()` initializer.
function plugin(initializerContext) {
  return new _plugin.UnifiedSearchPublicPlugin(initializerContext);
}
