// @kbn/unified-search-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setTheme = exports.setOverlays = exports.setIndexPatterns = exports.setI18n = exports.setAnalytics = exports.getTheme = exports.getOverlays = exports.getIndexPatterns = exports.getI18n = exports.getAnalytics = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getIndexPatterns, setIndexPatterns] = (0, _public.createGetterSetter)('IndexPatterns');
exports.setIndexPatterns = setIndexPatterns;
exports.getIndexPatterns = getIndexPatterns;
const [getAnalytics, setAnalytics] = (0, _public.createGetterSetter)('Analytics');
exports.setAnalytics = setAnalytics;
exports.getAnalytics = getAnalytics;
const [getI18n, setI18n] = (0, _public.createGetterSetter)('I18n');
exports.setI18n = setI18n;
exports.getI18n = getI18n;
const [getTheme, setTheme] = (0, _public.createGetterSetter)('Theme');
exports.setTheme = setTheme;
exports.getTheme = getTheme;
const [getOverlays, setOverlays] = (0, _public.createGetterSetter)('Overlays');
exports.setOverlays = setOverlays;
exports.getOverlays = getOverlays;
