// @kbn/security-solution-plugin/public/explore/components/stat_items/stat_item_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatItemHeader = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _utils = require("./utils");
var i18n = _interopRequireWildcard(require("../../../common/containers/query_toggle/translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StatItemHeaderComponent = ({
  onToggle,
  isToggleExpanded,
  description
}) => /*#__PURE__*/_react.default.createElement(_utils.FlexGroup, {
  gutterSize: 'none'
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  className: isToggleExpanded ? '' : 'no-margin'
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  gutterSize: 'none',
  responsive: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
  "aria-label": i18n.QUERY_BUTTON_TITLE(isToggleExpanded),
  "data-test-subj": "query-toggle-stat",
  color: "text",
  display: "empty",
  iconType: isToggleExpanded ? 'arrowDown' : 'arrowRight',
  onClick: onToggle,
  size: "xs",
  title: i18n.QUERY_BUTTON_TITLE(isToggleExpanded)
})), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
  size: "xxxs"
}, /*#__PURE__*/_react.default.createElement(_utils.StyledTitle, null, description))))));
const StatItemHeader = exports.StatItemHeader = /*#__PURE__*/_react.default.memo(StatItemHeaderComponent);
StatItemHeader.displayName = 'StatItemHeader';
