// @kbn/observability-onboarding-plugin/public/application/observability_onboarding_flow.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ObservabilityOnboardingFlow = ObservabilityOnboardingFlow;
var _reactQuery = require("@tanstack/react-query");
var _react = _interopRequireWildcard(require("react"));
var _sharedUxRouter = require("@kbn/shared-ux-router");
var _reactRouterDomV5Compat = require("react-router-dom-v5-compat");
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
var _background = _interopRequireDefault(require("./header/background.svg"));
var _footer = require("./footer/footer");
var _onboarding_flow_form = require("./onboarding_flow_form/onboarding_flow_form");
var _header = require("./header/header");
var _system_logs = require("./quickstart_flows/system_logs");
var _custom_logs = require("./quickstart_flows/custom_logs");
var _otel_logs = require("./quickstart_flows/otel_logs");
var _auto_detect = require("./quickstart_flows/auto_detect");
var _back_button = require("./shared/back_button");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const queryClient = new _reactQuery.QueryClient();
function ObservabilityOnboardingFlow() {
  const {
    pathname
  } = (0, _reactRouterDomV5Compat.useLocation)();
  (0, _react.useEffect)(() => {
    window.scrollTo(0, 0);
  }, [pathname]);
  return /*#__PURE__*/_react.default.createElement(_reactQuery.QueryClientProvider, {
    client: queryClient
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate, {
    css: (0, _react2.css)`
          padding-top: 0px !important;
        `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate.Section, {
    paddingSize: "xl",
    css: (0, _react2.css)`
            & > div {
              background-image: url(${_background.default});
              background-position: right center;
              background-repeat: no-repeat;
            }
          `,
    grow: false,
    restrictWidth: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xl"
  }), /*#__PURE__*/_react.default.createElement(_header.Header, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate.Section, {
    paddingSize: "xl",
    color: "subdued",
    restrictWidth: true
  }, /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Routes, null, /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, {
    path: "/auto-detect"
  }, /*#__PURE__*/_react.default.createElement(_back_button.BackButton, null), /*#__PURE__*/_react.default.createElement(_auto_detect.AutoDetectPanel, null)), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, {
    path: "/systemLogs"
  }, /*#__PURE__*/_react.default.createElement(_back_button.BackButton, null), /*#__PURE__*/_react.default.createElement(_system_logs.SystemLogsPanel, null)), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, {
    path: "/customLogs"
  }, /*#__PURE__*/_react.default.createElement(_back_button.BackButton, null), /*#__PURE__*/_react.default.createElement(_custom_logs.CustomLogsPanel, null)), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, {
    path: "/otel-logs"
  }, /*#__PURE__*/_react.default.createElement(_back_button.BackButton, null), /*#__PURE__*/_react.default.createElement(_otel_logs.OtelLogsPanel, null)), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, null, /*#__PURE__*/_react.default.createElement(_onboarding_flow_form.OnboardingFlowForm, null))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xl"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate.Section, {
    contentProps: {
      css: {
        paddingBlock: 0
      }
    },
    css: (0, _react2.css)`
            padding-inline: 0px;
          `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true,
    css: (0, _react2.css)`
              border-radius: 0px;
              border-left: none;
              border-bottom: none;
              border-right: none;
            `
  }, /*#__PURE__*/_react.default.createElement(_footer.Footer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xl"
  })))));
}
