// @kbn/synthetics-plugin/public/apps/synthetics/state/service_locations/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  serviceLocationsReducer: true
};
exports.serviceLocationsReducer = void 0;
var _toolkit = require("@reduxjs/toolkit");
var _runtime_types = require("../../../../../common/runtime_types");
var _actions = require("./actions");
Object.keys(_actions).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _actions[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _actions[key];
    }
  });
});
var _effects = require("./effects");
Object.keys(_effects).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _effects[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _effects[key];
    }
  });
});
var _selectors = require("./selectors");
Object.keys(_selectors).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _selectors[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _selectors[key];
    }
  });
});
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const initialState = {
  locations: [],
  throttling: _runtime_types.DEFAULT_THROTTLING,
  loading: false,
  locationsLoaded: false,
  error: null
};
const serviceLocationsReducer = exports.serviceLocationsReducer = (0, _toolkit.createReducer)(initialState, builder => {
  builder.addCase(_actions.getServiceLocations, state => {
    state.loading = true;
    state.locationsLoaded = true;
  }).addCase(_actions.getServiceLocationsSuccess, (state, action) => {
    state.loading = false;
    state.error = null;
    state.locations = action.payload.locations;
    state.throttling = action.payload.throttling || _runtime_types.DEFAULT_THROTTLING;
  }).addCase(_actions.getServiceLocationsFailure, (state, action) => {
    state.loading = false;
    state.error = action.payload;
  });
});
