// @kbn/triggers-actions-ui-plugin/public/application/sections/rule_details/components/rule_action_error_log_flyout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RuleActionErrorLogFlyout = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _rule_error_log = require("./rule_error_log");
var _rule_action_error_badge = require("./rule_action_error_badge");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RuleActionErrorLogFlyout = props => {
  const {
    runLog,
    refreshToken,
    onClose,
    activeSpaceId
  } = props;
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const {
    id,
    rule_id: ruleId,
    message,
    num_errored_actions: totalErrors,
    space_ids: spaceIds = []
  } = runLog;
  const isFlyoutPush = (0, _eui.useIsWithinBreakpoints)(['xl']);
  const logFromDifferentSpace = (0, _react.useMemo)(() => Boolean(activeSpaceId && !(spaceIds !== null && spaceIds !== void 0 && spaceIds.includes(activeSpaceId))), [activeSpaceId, spaceIds]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyout, {
    type: isFlyoutPush ? 'push' : 'overlay',
    onClose: onClose,
    size: isFlyoutPush ? 'm' : 'l',
    "data-test-subj": "ruleActionErrorLogFlyout"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "m"
  }, /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.triggersActionsUI.sections.ruleDetails.ruleActionErrorLogFlyout.actionErrors",
    defaultMessage: "Errored Actions"
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs",
    style: {
      fontWeight: euiTheme.font.weight.bold
    }
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.triggersActionsUI.sections.ruleDetails.ruleActionErrorLogFlyout.message",
    defaultMessage: "Message"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    "data-test-subj": "ruleActionErrorLogFlyoutMessageText"
  }, message), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    size: "full"
  }), /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_rule_action_error_badge.RuleActionErrorBadge, {
    totalErrors: totalErrors
  }), "\xA0", /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.triggersActionsUI.sections.ruleDetails.ruleActionErrorLogFlyout.actionErrorsPlural",
    defaultMessage: "{value, plural, one {errored action} other {errored actions}}",
    values: {
      value: totalErrors
    }
  })), /*#__PURE__*/_react.default.createElement(_rule_error_log.RuleErrorLogWithApi, {
    ruleId: ruleId,
    runId: id,
    spaceId: spaceIds[0],
    logFromDifferentSpace: logFromDifferentSpace,
    refreshToken: refreshToken
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutFooter, null, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "ruleActionErrorLogFlyoutCloseButton",
    onClick: onClose
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.triggersActionsUI.sections.ruleDetails.ruleActionErrorLogFlyout.close",
    defaultMessage: "Close"
  }))));
};
exports.RuleActionErrorLogFlyout = RuleActionErrorLogFlyout;
