// @kbn/security-solution-plugin/public/common/store/app/actions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateNote = exports.removeError = exports.deleteNote = exports.addNotes = exports.addErrorHash = exports.addError = void 0;
var _typescriptFsa = _interopRequireDefault(require("typescript-fsa"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const actionCreator = (0, _typescriptFsa.default)('x-pack/security_solution/local/app');
const updateNote = exports.updateNote = actionCreator('UPDATE_NOTE');
const addNotes = exports.addNotes = actionCreator('ADD_NOTE');
const deleteNote = exports.deleteNote = actionCreator('DELETE_NOTE');
const addError = exports.addError = actionCreator('ADD_ERRORS');
const removeError = exports.removeError = actionCreator('REMOVE_ERRORS');
const addErrorHash = exports.addErrorHash = actionCreator('ADD_ERROR_HASH');
