// @kbn/observability-shared-plugin/public/components/profiling/embeddables/embeddable_flamegraph.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmbeddableFlamegraph = EmbeddableFlamegraph;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _ = require(".");
var _component_registry = require("../helpers/component_registry");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function EmbeddableFlamegraph({
  height,
  ...props
}) {
  const EmbeddableFlamegraphComponent = (0, _component_registry.getProfilingComponent)(_.EMBEDDABLE_FLAMEGRAPH);
  return /*#__PURE__*/_react.default.createElement("div", {
    css: (0, _react2.css)`
        width: 100%;
        height: ${height};
        display: flex;
        flex: 1 1 100%;
        z-index: 1;
        min-height: 0;
      `
  }, EmbeddableFlamegraphComponent && /*#__PURE__*/_react.default.createElement(EmbeddableFlamegraphComponent, props));
}
