// @kbn/security-plugin-types-server/src/authentication/api_keys/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "crossClusterApiKeySchema", {
  enumerable: true,
  get: function () {
    return _api_keys.crossClusterApiKeySchema;
  }
});
Object.defineProperty(exports, "getRestApiKeyWithKibanaPrivilegesSchema", {
  enumerable: true,
  get: function () {
    return _api_keys.getRestApiKeyWithKibanaPrivilegesSchema;
  }
});
Object.defineProperty(exports, "getUpdateRestApiKeyWithKibanaPrivilegesSchema", {
  enumerable: true,
  get: function () {
    return _api_keys.getUpdateRestApiKeyWithKibanaPrivilegesSchema;
  }
});
Object.defineProperty(exports, "restApiKeySchema", {
  enumerable: true,
  get: function () {
    return _api_keys.restApiKeySchema;
  }
});
Object.defineProperty(exports, "updateCrossClusterApiKeySchema", {
  enumerable: true,
  get: function () {
    return _api_keys.updateCrossClusterApiKeySchema;
  }
});
Object.defineProperty(exports, "updateRestApiKeySchema", {
  enumerable: true,
  get: function () {
    return _api_keys.updateRestApiKeySchema;
  }
});
var _api_keys = require("./api_keys");
