// @kbn/stack-connectors-plugin/common/sentinelone/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SUB_ACTION = exports.SENTINELONE_TITLE = exports.SENTINELONE_CONNECTOR_ID = exports.API_MAX_RESULTS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SENTINELONE_TITLE = exports.SENTINELONE_TITLE = 'Sentinel One';
const SENTINELONE_CONNECTOR_ID = exports.SENTINELONE_CONNECTOR_ID = '.sentinelone';
const API_MAX_RESULTS = exports.API_MAX_RESULTS = 1000;
let SUB_ACTION = exports.SUB_ACTION = /*#__PURE__*/function (SUB_ACTION) {
  SUB_ACTION["EXECUTE_SCRIPT"] = "executeScript";
  SUB_ACTION["GET_AGENTS"] = "getAgents";
  SUB_ACTION["ISOLATE_HOST"] = "isolateHost";
  SUB_ACTION["RELEASE_HOST"] = "releaseHost";
  SUB_ACTION["GET_REMOTE_SCRIPTS"] = "getRemoteScripts";
  SUB_ACTION["GET_REMOTE_SCRIPT_STATUS"] = "getRemoteScriptStatus";
  SUB_ACTION["GET_REMOTE_SCRIPT_RESULTS"] = "getRemoteScriptResults";
  SUB_ACTION["FETCH_AGENT_FILES"] = "fetchAgentFiles";
  SUB_ACTION["DOWNLOAD_AGENT_FILE"] = "downloadAgentFile";
  SUB_ACTION["GET_ACTIVITIES"] = "getActivities";
  return SUB_ACTION;
}({});
