// @kbn/cases-plugin/public/components/user_actions/comment/show_alert.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserActionShowAlert = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var i18n = _interopRequireWildcard(require("../translations"));
var _navigation = require("../../../common/navigation");
var _types = require("../../../../common/types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UserActionShowAlertComponent = ({
  id,
  alertId,
  index,
  onShowAlertDetails
}) => {
  const {
    navigateToCaseView
  } = (0, _navigation.useCaseViewNavigation)();
  const {
    detailName
  } = (0, _navigation.useCaseViewParams)();
  const onClick = (0, _react.useCallback)(() => {
    if (onShowAlertDetails) {
      onShowAlertDetails(alertId, index);
    } else {
      navigateToCaseView({
        detailName,
        tabId: _types.CASE_VIEW_PAGE_TABS.ALERTS
      });
    }
  }, [alertId, detailName, index, navigateToCaseView, onShowAlertDetails]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    position: "top",
    content: /*#__PURE__*/_react.default.createElement("p", null, i18n.SHOW_ALERT_TOOLTIP)
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    "aria-label": i18n.SHOW_ALERT_TOOLTIP,
    "data-test-subj": `comment-action-show-alert-${id}`,
    onClick: onClick,
    iconType: "arrowRight",
    id: `${id}-show-alert`
  }));
};
UserActionShowAlertComponent.displayName = 'UserActionShowAlert';
const UserActionShowAlert = exports.UserActionShowAlert = /*#__PURE__*/(0, _react.memo)(UserActionShowAlertComponent);
