// @kbn/infra-plugin/public/components/asset_details/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INTEGRATION_NAME = exports.ContentTabIds = exports.ASSET_DETAILS_ASSET_TYPE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let ContentTabIds = exports.ContentTabIds = /*#__PURE__*/function (ContentTabIds) {
  ContentTabIds["OVERVIEW"] = "overview";
  ContentTabIds["METADATA"] = "metadata";
  ContentTabIds["METRICS"] = "metrics";
  ContentTabIds["PROCESSES"] = "processes";
  ContentTabIds["PROFILING"] = "profiling";
  ContentTabIds["ANOMALIES"] = "anomalies";
  ContentTabIds["OSQUERY"] = "osquery";
  ContentTabIds["LOGS"] = "logs";
  ContentTabIds["LINK_TO_APM"] = "linkToApm";
  ContentTabIds["DASHBOARDS"] = "dashboards";
  return ContentTabIds;
}({});
let INTEGRATION_NAME = exports.INTEGRATION_NAME = /*#__PURE__*/function (INTEGRATION_NAME) {
  INTEGRATION_NAME["kubernetesNode"] = "kubernetesNode";
  INTEGRATION_NAME["kubernetesContainer"] = "kubernetesContainer";
  INTEGRATION_NAME["docker"] = "docker";
  return INTEGRATION_NAME;
}({});
let ASSET_DETAILS_ASSET_TYPE = exports.ASSET_DETAILS_ASSET_TYPE = /*#__PURE__*/function (ASSET_DETAILS_ASSET_TYPE) {
  ASSET_DETAILS_ASSET_TYPE["container"] = "container";
  ASSET_DETAILS_ASSET_TYPE["host"] = "host";
  return ASSET_DETAILS_ASSET_TYPE;
}({});
