// @kbn/security-solution-plugin/public/detection_engine/rule_creation_ui/components/step_schedule_rule/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StepScheduleRuleReadOnly = exports.StepScheduleRule = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _react = _interopRequireWildcard(require("react"));
var _description_step = require("../description_step");
var _schedule_item_form = require("../../../rule_creation/components/schedule_item_form");
var _shared_imports = require("../../../../shared_imports");
var _step_content_wrapper = require("../../../rule_creation/components/step_content_wrapper");
var _schema = require("./schema");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StyledForm = (0, _styledComponents.default)(_shared_imports.Form)`
  max-width: 235px !important;
`;
const StepScheduleRuleComponent = ({
  isLoading,
  isUpdateView = false,
  form
}) => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_step_content_wrapper.StepContentWrapper, {
    addPadding: !isUpdateView
  }, /*#__PURE__*/_react.default.createElement(StyledForm, {
    form: form,
    "data-test-subj": "stepScheduleRule"
  }, /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    path: "interval",
    component: _schedule_item_form.ScheduleItem,
    componentProps: {
      idAria: 'detectionEngineStepScheduleRuleInterval',
      isDisabled: isLoading,
      dataTestSubj: 'detectionEngineStepScheduleRuleInterval',
      minimumValue: 1
    }
  }), /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    path: "from",
    component: _schedule_item_form.ScheduleItem,
    componentProps: {
      idAria: 'detectionEngineStepScheduleRuleFrom',
      isDisabled: isLoading,
      dataTestSubj: 'detectionEngineStepScheduleRuleFrom',
      minimumValue: 1
    }
  }))));
};
const StepScheduleRule = exports.StepScheduleRule = /*#__PURE__*/(0, _react.memo)(StepScheduleRuleComponent);
const StepScheduleRuleReadOnlyComponent = ({
  addPadding,
  defaultValues: data,
  descriptionColumns
}) => {
  return /*#__PURE__*/_react.default.createElement(_step_content_wrapper.StepContentWrapper, {
    addPadding: addPadding
  }, /*#__PURE__*/_react.default.createElement(_description_step.StepRuleDescription, {
    columns: descriptionColumns,
    schema: _schema.schema,
    data: data
  }));
};
const StepScheduleRuleReadOnly = exports.StepScheduleRuleReadOnly = /*#__PURE__*/(0, _react.memo)(StepScheduleRuleReadOnlyComponent);
