// @kbn/fleet-plugin/common/constants/agent_policy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.agentPolicyStatuses = exports.LICENSE_FOR_SCHEDULE_UPGRADE = exports.DEFAULT_MAX_AGENT_POLICIES_WITH_INACTIVITY_TIMEOUT = exports.DEFAULT_LOG_LEVEL = exports.AGENT_POLICY_SAVED_OBJECT_TYPE = exports.AGENT_POLICY_INDEX = exports.AGENT_POLICY_DEFAULT_MONITORING_DATASETS = exports.AGENT_LOG_LEVELS = exports.AGENTLESS_POLICY_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AGENT_POLICY_SAVED_OBJECT_TYPE = exports.AGENT_POLICY_SAVED_OBJECT_TYPE = 'ingest-agent-policies';
const AGENT_POLICY_INDEX = exports.AGENT_POLICY_INDEX = '.fleet-policies';
const agentPolicyStatuses = exports.agentPolicyStatuses = {
  Active: 'active',
  Inactive: 'inactive'
};
const AGENT_POLICY_DEFAULT_MONITORING_DATASETS = exports.AGENT_POLICY_DEFAULT_MONITORING_DATASETS = ['elastic_agent', 'elastic_agent.elastic_agent', 'elastic_agent.apm_server', 'elastic_agent.filebeat', 'elastic_agent.filebeat_input', 'elastic_agent.fleet_server', 'elastic_agent.metricbeat', 'elastic_agent.osquerybeat', 'elastic_agent.packetbeat', 'elastic_agent.endpoint_security', 'elastic_agent.auditbeat', 'elastic_agent.heartbeat', 'elastic_agent.cloudbeat', 'elastic_agent.cloud_defend', 'elastic_agent.pf_host_agent', 'elastic_agent.pf_elastic_collector', 'elastic_agent.pf_elastic_symbolizer'];
const LICENSE_FOR_SCHEDULE_UPGRADE = exports.LICENSE_FOR_SCHEDULE_UPGRADE = 'platinum';
const DEFAULT_MAX_AGENT_POLICIES_WITH_INACTIVITY_TIMEOUT = exports.DEFAULT_MAX_AGENT_POLICIES_WITH_INACTIVITY_TIMEOUT = 750;
const AGENTLESS_POLICY_ID = exports.AGENTLESS_POLICY_ID = 'agentless'; // the policy id defined here: https://github.com/elastic/project-controller/blob/main/internal/project/security/security_kibana_config.go#L86

const AGENT_LOG_LEVELS = exports.AGENT_LOG_LEVELS = ['error', 'warning', 'info', 'debug'];
const DEFAULT_LOG_LEVEL = exports.DEFAULT_LOG_LEVEL = 'info';
