// @kbn/security-solution-plugin/public/attack_discovery/pages/page_title/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PageTitle = void 0;
var _eui = require("@elastic/eui");
var _react = require("@emotion/react");
var _react2 = _interopRequireDefault(require("react"));
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PageTitleComponent = () => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    "data-test-subj": "pageTitle",
    gutterSize: "none"
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiTitle, {
    "data-test-subj": "attackDiscoveryPageTitle",
    size: "l"
  }, /*#__PURE__*/_react2.default.createElement("h1", null, i18n.ATTACK_DISCOVERY_PAGE_TITLE))), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    css: (0, _react.css)`
          vertical-align: middle;
          padding-left: ${euiTheme.size.m};
          * {
            vertical-align: middle;
          }
        `
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiBetaBadge, {
    iconType: 'beaker',
    label: i18n.BETA,
    tooltipContent: i18n.BETA_TOOLTIP,
    size: "m",
    color: "hollow",
    css: (0, _react.css)`
            margin-bottom: ${euiTheme.size.s};
          `
  })));
};
PageTitleComponent.displayName = 'PageTitle';
const PageTitle = exports.PageTitle = /*#__PURE__*/_react2.default.memo(PageTitleComponent);
