// @kbn/osquery-plugin/public/routes/packs/list/load_integration_assets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadIntegrationAssetsButton = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _use_import_assets = require("../../../assets/use_import_assets");
var _use_assets_status = require("../../../assets/use_assets_status");
var _translations = require("./translations");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LoadIntegrationAssetsButtonComponent = ({
  fill
}) => {
  var _data$upToDate;
  const {
    data
  } = (0, _use_assets_status.useAssetsStatus)();
  const {
    isLoading,
    mutateAsync
  } = (0, _use_import_assets.useImportAssets)({
    successToastText: data !== null && data !== void 0 && (_data$upToDate = data.upToDate) !== null && _data$upToDate !== void 0 && _data$upToDate.length ? _translations.UPDATE_PREBUILT_PACKS_SUCCESS_TEXT : _translations.LOAD_PREBUILT_PACKS_SUCCESS_TEXT
  });
  const handleClick = (0, _react.useCallback)(() => mutateAsync(), [mutateAsync]);
  if (data !== null && data !== void 0 && data.install.length || data !== null && data !== void 0 && data.update.length) {
    var _data$upToDate2;
    return /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      fill: !!fill,
      isLoading: isLoading,
      onClick: handleClick,
      iconType: "plusInCircle"
    }, data !== null && data !== void 0 && (_data$upToDate2 = data.upToDate) !== null && _data$upToDate2 !== void 0 && _data$upToDate2.length ? _translations.UPDATE_PREBUILT_PACKS_BUTTON : _translations.LOAD_PREBUILT_PACKS_BUTTON);
  }
  return null;
};
const LoadIntegrationAssetsButton = exports.LoadIntegrationAssetsButton = /*#__PURE__*/_react.default.memo(LoadIntegrationAssetsButtonComponent);
