// @kbn/cases-plugin/public/components/custom_fields/toggle/configure.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Configure = void 0;
var _react = _interopRequireDefault(require("react"));
var _hook_form_lib = require("@kbn/es-ui-shared-plugin/static/forms/hook_form_lib");
var _components = require("@kbn/es-ui-shared-plugin/static/forms/components");
var i18n = _interopRequireWildcard(require("../translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ConfigureComponent = () => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_hook_form_lib.UseField, {
    path: "required",
    component: _components.CheckBoxField,
    componentProps: {
      label: i18n.FIELD_OPTIONS,
      'data-test-subj': 'toggle-custom-field-required-wrapper',
      euiFieldProps: {
        label: i18n.FIELD_OPTION_REQUIRED,
        'data-test-subj': 'toggle-custom-field-required'
      }
    }
  }), /*#__PURE__*/_react.default.createElement(_hook_form_lib.UseField, {
    path: "defaultValue",
    component: _components.ToggleField,
    config: {
      defaultValue: false
    },
    componentProps: {
      label: i18n.DEFAULT_VALUE,
      euiFieldProps: {
        'data-test-subj': 'toggle-custom-field-default-value'
      }
    }
  }));
};
ConfigureComponent.displayName = 'Configure';
const Configure = exports.Configure = /*#__PURE__*/_react.default.memo(ConfigureComponent);
