// @kbn/cloud-defend-plugin/public/application/router.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CloudDefendRouter = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactQuery = require("@tanstack/react-query");
var _reactRouterDom = require("react-router-dom");
var _sharedUxRouter = require("@kbn/shared-ux-router");
var _constants = require("../common/navigation/constants");
var _security_solution_context = require("./security_solution_context");
var _policies = require("../pages/policies");
var _route = require("./route");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const queryClient = new _reactQuery.QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false
    }
  }
});
const CloudDefendRouter = ({
  securitySolutionContext
}) => {
  const routerElement = /*#__PURE__*/_react.default.createElement(_reactQuery.QueryClientProvider, {
    client: queryClient
  }, /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Routes, null, /*#__PURE__*/_react.default.createElement(_route.CloudDefendRoute, (0, _extends2.default)({}, _constants.cloudDefendPages.policies, {
    component: _policies.Policies
  })), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, null, /*#__PURE__*/_react.default.createElement(_reactRouterDom.Redirect, {
    to: _constants.cloudDefendPages.policies.path
  }))));
  if (securitySolutionContext) {
    return /*#__PURE__*/_react.default.createElement(_security_solution_context.SecuritySolutionContext.Provider, {
      value: securitySolutionContext
    }, routerElement);
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, routerElement);
};

// Using a default export for usage with `React.lazy`
// eslint-disable-next-line import/no-default-export
exports.default = exports.CloudDefendRouter = CloudDefendRouter;
