// @kbn/observability-ai-assistant-plugin/server/functions/kibana.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerKibanaFunction = registerKibanaFunction;
var _axios = _interopRequireDefault(require("axios"));
var _url = require("url");
var _lodash = require("lodash");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function registerKibanaFunction({
  functions,
  resources
}) {
  functions.registerFunction({
    name: 'kibana',
    description: 'Call Kibana APIs on behalf of the user. Only call this function when the user has explicitly requested it, and you know how to call it, for example by querying the knowledge base or having the user explain it to you. Assume that pathnames, bodies and query parameters may have changed since your knowledge cut off date.',
    descriptionForUser: 'Call Kibana APIs on behalf of the user',
    parameters: {
      type: 'object',
      properties: {
        method: {
          type: 'string',
          description: 'The HTTP method of the Kibana endpoint',
          enum: ['GET', 'PUT', 'POST', 'DELETE', 'PATCH']
        },
        pathname: {
          type: 'string',
          description: 'The pathname of the Kibana endpoint, excluding query parameters'
        },
        query: {
          type: 'object',
          description: 'The query parameters, as an object'
        },
        body: {
          type: 'object',
          description: 'The body of the request'
        }
      },
      required: ['method', 'pathname']
    }
  }, ({
    arguments: {
      method,
      pathname,
      body,
      query
    }
  }, signal) => {
    const {
      request
    } = resources;
    const {
      protocol,
      host,
      pathname: pathnameFromRequest
    } = request.rewrittenUrl || request.url;
    const origin = (0, _lodash.first)((0, _lodash.castArray)(request.headers.origin));
    const nextUrl = {
      host,
      protocol,
      ...(origin ? (0, _lodash.pick)((0, _url.parse)(origin), 'host', 'protocol') : {}),
      pathname: pathnameFromRequest.replace('/internal/observability_ai_assistant/chat/complete', pathname),
      query: query ? query : undefined
    };
    const copiedHeaderNames = ['accept-encoding', 'accept-language', 'accept', 'content-type', 'cookie', 'kbn-build-number', 'kbn-version', 'origin', 'referer', 'user-agent', 'x-elastic-internal-origin', 'x-kbn-context'];
    const headers = (0, _lodash.pickBy)(request.headers, (value, key) => {
      return copiedHeaderNames.includes(key.toLowerCase()) || key.toLowerCase().startsWith('sec-');
    });
    return (0, _axios.default)({
      method,
      headers,
      url: (0, _url.format)(nextUrl),
      data: body ? JSON.stringify(body) : undefined,
      signal
    }).then(response => {
      return {
        content: response.data
      };
    });
  });
}
