// @kbn/security-solution-plugin/public/common/components/toasters/modal_all_errors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ModalAllErrors = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * @deprecated Use x-pack/plugins/security_solution/public/common/hooks/use_app_toasts.ts instead
 */

/**
 * @deprecated Use x-pack/plugins/security_solution/public/common/hooks/use_app_toasts.ts instead
 */
const ModalAllErrorsComponent = ({
  isShowing,
  toast,
  toggle
}) => {
  const handleClose = (0, _react.useCallback)(() => toggle(toast), [toggle, toast]);
  if (!isShowing || toast == null) return null;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiModal, {
    onClose: handleClose
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiModalHeader, null, /*#__PURE__*/_react.default.createElement(_eui.EuiModalHeaderTitle, null, i18n.TITLE_ERROR_MODAL)), /*#__PURE__*/_react.default.createElement(_eui.EuiModalBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: toast.title,
    color: "danger",
    size: "s",
    iconType: "warning"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), Array.isArray(toast.errors) &&
  // FunFact: This can be a non-array in some rare cases
  toast.errors.map((error, index) => /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
    key: `${toast.id}-${index}`,
    id: "accordion1",
    initialIsOpen: index === 0 ? true : false,
    buttonContent: error.length > 100 ? `${error.substring(0, 100)} ...` : error,
    "data-test-subj": "modal-all-errors-accordion"
  }, /*#__PURE__*/_react.default.createElement(MyEuiCodeBlock, null, error)))), /*#__PURE__*/_react.default.createElement(_eui.EuiModalFooter, null, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    onClick: handleClose,
    fill: true,
    "data-test-subj": "modal-all-errors-close"
  }, i18n.CLOSE_ERROR_MODAL)));
};

/**
 * @deprecated Use x-pack/plugins/security_solution/public/common/hooks/use_app_toasts.ts instead
 */
const ModalAllErrors = exports.ModalAllErrors = /*#__PURE__*/_react.default.memo(ModalAllErrorsComponent);

/**
 * @deprecated Use x-pack/plugins/security_solution/public/common/hooks/use_app_toasts.ts instead
 */
const MyEuiCodeBlock = (0, _styledComponents.default)(_eui.EuiCodeBlock)`
  margin-top: 4px;
`;

/**
 * @deprecated Use x-pack/plugins/security_solution/public/common/hooks/use_app_toasts.ts instead
 */
MyEuiCodeBlock.displayName = 'MyEuiCodeBlock';
