// @kbn/aiops-plugin/public/components/change_point_detection/function_picker.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FunctionPicker = void 0;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _constants = require("./constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FunctionPicker = exports.FunctionPicker = /*#__PURE__*/_react.default.memo(({
  value,
  onChange
}) => {
  const options = Object.keys(_constants.fnOperationTypeMapping).map(v => {
    return {
      id: v,
      label: v
    };
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiButtonGroup, {
    legend: _i18n.i18n.translate('xpack.aiops.changePointDetection.selectFunctionLabel', {
      defaultMessage: 'Function'
    }),
    options: options,
    idSelected: value,
    onChange: id => onChange(id),
    isFullWidth: true,
    buttonSize: "compressed"
  });
});
