// @kbn/ml-plugin/public/application/data_frame_analytics/pages/analytics_exploration/components/data_view_prompt/data_view_prompt.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataViewPrompt = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _kibana = require("../../../../../contexts/kibana");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DataViewPrompt = ({
  destIndex,
  color
}) => {
  const {
    services: {
      http: {
        basePath
      },
      application: {
        capabilities
      }
    }
  } = (0, _kibana.useMlKibana)();
  const canCreateDataView = (0, _react.useMemo)(() => capabilities.savedObjectsManagement.edit === true || capabilities.indexPatterns.save === true, [capabilities]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs",
    color: color !== null && color !== void 0 ? color : 'warning'
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.dataframe.analytics.dataViewPromptMessage",
    defaultMessage: "No data view exists for index {destIndex}. ",
    values: {
      destIndex: destIndex !== null && destIndex !== void 0 ? destIndex : ''
    }
  }), canCreateDataView === true ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.dataframe.analytics.dataViewPromptLink",
    defaultMessage: "{linkToDataViewManagement}{destIndex}.",
    values: {
      destIndex: destIndex ? ` for ${destIndex}` : '',
      linkToDataViewManagement: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        href: `${basePath.get()}/app/management/kibana/dataViews/create`,
        target: "_blank"
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.ml.dataframe.analytics.dataViewPromptLinkText",
        defaultMessage: "Create a data view"
      }))
    }
  }) : null));
};
exports.DataViewPrompt = DataViewPrompt;
