// @kbn/discover-utils/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CONTEXT_DEFAULT_SIZE_SETTING: true,
  CONTEXT_STEP_SETTING: true,
  CONTEXT_TIE_BREAKER_FIELDS_SETTING: true,
  DEFAULT_ALLOWED_LOGS_BASE_PATTERNS: true,
  DEFAULT_COLUMNS_SETTING: true,
  DOC_HIDE_TIME_COLUMN_SETTING: true,
  DOC_TABLE_LEGACY: true,
  FIELDS_LIMIT_SETTING: true,
  HIDE_ANNOUNCEMENTS: true,
  MAX_DOC_FIELDS_DISPLAYED: true,
  MODIFY_COLUMNS_ON_SWITCH: true,
  ROW_HEIGHT_OPTION: true,
  SAMPLE_ROWS_PER_PAGE_SETTING: true,
  SAMPLE_SIZE_SETTING: true,
  SEARCH_EMBEDDABLE_TYPE: true,
  SEARCH_FIELDS_FROM_SOURCE: true,
  SEARCH_ON_PAGE_LOAD_SETTING: true,
  SHOW_FIELD_STATISTICS: true,
  SHOW_MULTIFIELDS: true,
  SORT_DEFAULT_ORDER_SETTING: true,
  TRUNCATE_MAX_HEIGHT: true,
  IgnoredReason: true,
  buildDataTableRecord: true,
  buildDataTableRecordList: true,
  createLogsContextService: true,
  fieldConstants: true,
  formatFieldValue: true,
  formatHit: true,
  getDocId: true,
  getLogDocumentOverview: true,
  getIgnoredReason: true,
  getMessageFieldWithFallbacks: true,
  getShouldShowFieldHandler: true,
  isNestedFieldParent: true,
  isLegacyTableEnabled: true,
  usePager: true
};
Object.defineProperty(exports, "CONTEXT_DEFAULT_SIZE_SETTING", {
  enumerable: true,
  get: function () {
    return _src.CONTEXT_DEFAULT_SIZE_SETTING;
  }
});
Object.defineProperty(exports, "CONTEXT_STEP_SETTING", {
  enumerable: true,
  get: function () {
    return _src.CONTEXT_STEP_SETTING;
  }
});
Object.defineProperty(exports, "CONTEXT_TIE_BREAKER_FIELDS_SETTING", {
  enumerable: true,
  get: function () {
    return _src.CONTEXT_TIE_BREAKER_FIELDS_SETTING;
  }
});
Object.defineProperty(exports, "DEFAULT_ALLOWED_LOGS_BASE_PATTERNS", {
  enumerable: true,
  get: function () {
    return _src.DEFAULT_ALLOWED_LOGS_BASE_PATTERNS;
  }
});
Object.defineProperty(exports, "DEFAULT_COLUMNS_SETTING", {
  enumerable: true,
  get: function () {
    return _src.DEFAULT_COLUMNS_SETTING;
  }
});
Object.defineProperty(exports, "DOC_HIDE_TIME_COLUMN_SETTING", {
  enumerable: true,
  get: function () {
    return _src.DOC_HIDE_TIME_COLUMN_SETTING;
  }
});
Object.defineProperty(exports, "DOC_TABLE_LEGACY", {
  enumerable: true,
  get: function () {
    return _src.DOC_TABLE_LEGACY;
  }
});
Object.defineProperty(exports, "FIELDS_LIMIT_SETTING", {
  enumerable: true,
  get: function () {
    return _src.FIELDS_LIMIT_SETTING;
  }
});
Object.defineProperty(exports, "HIDE_ANNOUNCEMENTS", {
  enumerable: true,
  get: function () {
    return _src.HIDE_ANNOUNCEMENTS;
  }
});
Object.defineProperty(exports, "IgnoredReason", {
  enumerable: true,
  get: function () {
    return _src.IgnoredReason;
  }
});
Object.defineProperty(exports, "MAX_DOC_FIELDS_DISPLAYED", {
  enumerable: true,
  get: function () {
    return _src.MAX_DOC_FIELDS_DISPLAYED;
  }
});
Object.defineProperty(exports, "MODIFY_COLUMNS_ON_SWITCH", {
  enumerable: true,
  get: function () {
    return _src.MODIFY_COLUMNS_ON_SWITCH;
  }
});
Object.defineProperty(exports, "ROW_HEIGHT_OPTION", {
  enumerable: true,
  get: function () {
    return _src.ROW_HEIGHT_OPTION;
  }
});
Object.defineProperty(exports, "SAMPLE_ROWS_PER_PAGE_SETTING", {
  enumerable: true,
  get: function () {
    return _src.SAMPLE_ROWS_PER_PAGE_SETTING;
  }
});
Object.defineProperty(exports, "SAMPLE_SIZE_SETTING", {
  enumerable: true,
  get: function () {
    return _src.SAMPLE_SIZE_SETTING;
  }
});
Object.defineProperty(exports, "SEARCH_EMBEDDABLE_TYPE", {
  enumerable: true,
  get: function () {
    return _src.SEARCH_EMBEDDABLE_TYPE;
  }
});
Object.defineProperty(exports, "SEARCH_FIELDS_FROM_SOURCE", {
  enumerable: true,
  get: function () {
    return _src.SEARCH_FIELDS_FROM_SOURCE;
  }
});
Object.defineProperty(exports, "SEARCH_ON_PAGE_LOAD_SETTING", {
  enumerable: true,
  get: function () {
    return _src.SEARCH_ON_PAGE_LOAD_SETTING;
  }
});
Object.defineProperty(exports, "SHOW_FIELD_STATISTICS", {
  enumerable: true,
  get: function () {
    return _src.SHOW_FIELD_STATISTICS;
  }
});
Object.defineProperty(exports, "SHOW_MULTIFIELDS", {
  enumerable: true,
  get: function () {
    return _src.SHOW_MULTIFIELDS;
  }
});
Object.defineProperty(exports, "SORT_DEFAULT_ORDER_SETTING", {
  enumerable: true,
  get: function () {
    return _src.SORT_DEFAULT_ORDER_SETTING;
  }
});
Object.defineProperty(exports, "TRUNCATE_MAX_HEIGHT", {
  enumerable: true,
  get: function () {
    return _src.TRUNCATE_MAX_HEIGHT;
  }
});
Object.defineProperty(exports, "buildDataTableRecord", {
  enumerable: true,
  get: function () {
    return _src.buildDataTableRecord;
  }
});
Object.defineProperty(exports, "buildDataTableRecordList", {
  enumerable: true,
  get: function () {
    return _src.buildDataTableRecordList;
  }
});
Object.defineProperty(exports, "createLogsContextService", {
  enumerable: true,
  get: function () {
    return _src.createLogsContextService;
  }
});
Object.defineProperty(exports, "fieldConstants", {
  enumerable: true,
  get: function () {
    return _src.fieldConstants;
  }
});
Object.defineProperty(exports, "formatFieldValue", {
  enumerable: true,
  get: function () {
    return _src.formatFieldValue;
  }
});
Object.defineProperty(exports, "formatHit", {
  enumerable: true,
  get: function () {
    return _src.formatHit;
  }
});
Object.defineProperty(exports, "getDocId", {
  enumerable: true,
  get: function () {
    return _src.getDocId;
  }
});
Object.defineProperty(exports, "getIgnoredReason", {
  enumerable: true,
  get: function () {
    return _src.getIgnoredReason;
  }
});
Object.defineProperty(exports, "getLogDocumentOverview", {
  enumerable: true,
  get: function () {
    return _src.getLogDocumentOverview;
  }
});
Object.defineProperty(exports, "getMessageFieldWithFallbacks", {
  enumerable: true,
  get: function () {
    return _src.getMessageFieldWithFallbacks;
  }
});
Object.defineProperty(exports, "getShouldShowFieldHandler", {
  enumerable: true,
  get: function () {
    return _src.getShouldShowFieldHandler;
  }
});
Object.defineProperty(exports, "isLegacyTableEnabled", {
  enumerable: true,
  get: function () {
    return _src.isLegacyTableEnabled;
  }
});
Object.defineProperty(exports, "isNestedFieldParent", {
  enumerable: true,
  get: function () {
    return _src.isNestedFieldParent;
  }
});
Object.defineProperty(exports, "usePager", {
  enumerable: true,
  get: function () {
    return _src.usePager;
  }
});
var _src = require("./src");
var _types = require("./src/types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
