// @kbn/slo-plugin/public/pages/slo_edit/components/common/runtime_field_used.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RunTimeFieldUsed = RunTimeFieldUsed;
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _use_find_runtime_usage = require("../../hooks/use_find_runtime_usage");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function RunTimeFieldUsed({
  dataView,
  name
}) {
  const fieldNames = (0, _use_find_runtime_usage.useRunTimeFieldBeingUsed)(name, dataView);
  if (fieldNames.length === 0) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: _i18n.i18n.translate('xpack.slo.runTimeFieldUsed.euiCallOut.runtimeFieldsBeingUsedLabel', {
      defaultMessage: 'Runtime fields being used'
    }),
    color: "warning",
    iconType: "warning"
  }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.slo.runTimeFieldUsed.p.theRuntimeFieldLabel",
    defaultMessage: "The runtime field(s) {fields} from kibana dataview are being used in the query. If you update the runtime field, the query will not be updated automatically. You must save the slo definition again to update the underlying transform query.",
    values: {
      fields: /*#__PURE__*/_react.default.createElement("strong", null, fieldNames.join(', '))
    }
  }))));
}
