// @kbn/osquery-plugin/public/fleet_integration/navigation_buttons.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NavigationButtons = void 0;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireWildcard(require("react"));
var _common = require("../../common");
var _kibana = require("../common/lib/kibana");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const NavigationButtonsComponent = ({
  isDisabled = false,
  agentPolicyId
}) => {
  const {
    application: {
      getUrlForApp,
      navigateToApp
    }
  } = (0, _kibana.useKibana)().services;
  const liveQueryHref = (0, _react.useMemo)(() => getUrlForApp(_common.PLUGIN_ID, {
    path: agentPolicyId ? `/live_queries/new?agentPolicyId=${agentPolicyId}` : '/live_queries/new'
  }), [agentPolicyId, getUrlForApp]);
  const liveQueryClick = (0, _react.useCallback)(event => {
    if (!(0, _kibana.isModifiedEvent)(event) && (0, _kibana.isLeftClickEvent)(event)) {
      event.preventDefault();
      navigateToApp(_common.PLUGIN_ID, {
        path: agentPolicyId ? `/live_queries/new?agentPolicyId=${agentPolicyId}` : '/live_queries/new'
      });
    }
  }, [agentPolicyId, navigateToApp]);
  const packsHref = getUrlForApp(_common.PLUGIN_ID, {
    path: `/packs`
  });
  const packsClick = (0, _react.useCallback)(event => {
    if (!(0, _kibana.isModifiedEvent)(event) && (0, _kibana.isLeftClickEvent)(event)) {
      event.preventDefault();
      navigateToApp(_common.PLUGIN_ID, {
        path: `/packs`
      });
    }
  }, [navigateToApp]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "l"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "console"
    }),
    title: _i18n.i18n.translate('xpack.osquery.fleetIntegration.runLiveQueriesButtonText', {
      defaultMessage: 'Run live queries'
    }),
    href: liveQueryHref,
    onClick: liveQueryClick,
    description: '',
    isDisabled: isDisabled
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "clock"
    }),
    title: _i18n.i18n.translate('xpack.osquery.fleetIntegration.packsButtonText', {
      defaultMessage: 'Packs'
    }),
    description: '',
    isDisabled: isDisabled,
    href: packsHref,
    onClick: packsClick
  })));
};
NavigationButtonsComponent.displayName = 'NavigationButtonsComponent';
const NavigationButtons = exports.NavigationButtons = /*#__PURE__*/_react.default.memo(NavigationButtonsComponent);
