// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/fields/create_field/required_parameters_forms/scaled_float_type.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScaledFloatTypeRequiredParameters = void 0;
var _react = _interopRequireDefault(require("react"));
var _shared_imports = require("../../../../../shared_imports");
var _lib = require("../../../../../lib");
var _constants = require("../../../../../constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ScaledFloatTypeRequiredParameters = () => {
  return /*#__PURE__*/_react.default.createElement(_shared_imports.FormRow, {
    title: /*#__PURE__*/_react.default.createElement("h3", null, _constants.PARAMETERS_DEFINITION.scaling_factor.title),
    description: _constants.PARAMETERS_DEFINITION.scaling_factor.description
  }, /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    path: "scaling_factor",
    config: (0, _lib.getFieldConfig)('scaling_factor'),
    component: _shared_imports.Field
  }));
};
exports.ScaledFloatTypeRequiredParameters = ScaledFloatTypeRequiredParameters;
