// @kbn/expression-metric-vis-plugin/public/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getFormatService", {
  enumerable: true,
  get: function () {
    return _format_service.getFormatService;
  }
});
Object.defineProperty(exports, "getPaletteService", {
  enumerable: true,
  get: function () {
    return _palette_service.getPaletteService;
  }
});
Object.defineProperty(exports, "getThemeService", {
  enumerable: true,
  get: function () {
    return _theme_service.getThemeService;
  }
});
Object.defineProperty(exports, "getUiSettingsService", {
  enumerable: true,
  get: function () {
    return _ui_settings.getUiSettingsService;
  }
});
Object.defineProperty(exports, "setFormatService", {
  enumerable: true,
  get: function () {
    return _format_service.setFormatService;
  }
});
Object.defineProperty(exports, "setPaletteService", {
  enumerable: true,
  get: function () {
    return _palette_service.setPaletteService;
  }
});
Object.defineProperty(exports, "setThemeService", {
  enumerable: true,
  get: function () {
    return _theme_service.setThemeService;
  }
});
Object.defineProperty(exports, "setUiSettingsService", {
  enumerable: true,
  get: function () {
    return _ui_settings.setUiSettingsService;
  }
});
var _format_service = require("./format_service");
var _theme_service = require("./theme_service");
var _palette_service = require("./palette_service");
var _ui_settings = require("./ui_settings");
