// @kbn/cloud-security-posture-plugin/public/application/csp_router.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.CspRouter = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _reactQuery = require("@tanstack/react-query");
var _reactRouterDom = require("react-router-dom");
var _sharedUxRouter = require("@kbn/shared-ux-router");
var _constants = require("../common/navigation/constants");
var _security_solution_context = require("./security_solution_context");
var pages = _interopRequireWildcard(require("../pages"));
var _csp_route = require("./csp_route");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const queryClient = new _reactQuery.QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false
    }
  }
});

/** Props for the cloud security posture router component */

const CspRouter = ({
  securitySolutionContext
}) => {
  const routerElement = /*#__PURE__*/_react.default.createElement(_reactQuery.QueryClientProvider, {
    client: queryClient
  }, /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Routes, null, /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, (0, _extends2.default)({}, _constants.cloudPosturePages.findings, {
    component: pages.Findings
  })), /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, (0, _extends2.default)({}, _constants.cloudPosturePages.dashboard, {
    component: pages.ComplianceDashboard
  })), /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, (0, _extends2.default)({}, _constants.cloudPosturePages.vulnerability_dashboard, {
    component: pages.VulnerabilityDashboard
  })), /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, _constants.cloudPosturePages.benchmarks, /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Routes, null, /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, (0, _extends2.default)({}, _constants.benchmarksNavigation.rules, {
    component: pages.Rules
  })), /*#__PURE__*/_react.default.createElement(_csp_route.CspRoute, (0, _extends2.default)({}, _constants.cloudPosturePages.benchmarks, {
    component: pages.Benchmarks
  })))), /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Route, null, /*#__PURE__*/_react.default.createElement(_reactRouterDom.Redirect, {
    to: _constants.cloudPosturePages.dashboard.path
  }))));
  if (securitySolutionContext) {
    return /*#__PURE__*/_react.default.createElement(_security_solution_context.SecuritySolutionContext.Provider, {
      value: securitySolutionContext
    }, routerElement);
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, routerElement);
};

// Using a default export for usage with `React.lazy`
// eslint-disable-next-line import/no-default-export
exports.default = exports.CspRouter = CspRouter;
