// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/step_links/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ManageProjectsStepLinkId = exports.IntegrationsPageName = exports.AddIntegrationCalloutStepLinkId = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let IntegrationsPageName = exports.IntegrationsPageName = /*#__PURE__*/function (IntegrationsPageName) {
  IntegrationsPageName["integrationsSecurity"] = "integrations:/browse/security";
  IntegrationsPageName["integrationsSecurityCloud"] = "integrations:/browse/security/cloudsecurity_cdr";
  IntegrationsPageName["integrationsSecurityEdrXrd"] = "integrations:/browse/security/edr_xdr";
  return IntegrationsPageName;
}({});
const AddIntegrationCalloutStepLinkId = exports.AddIntegrationCalloutStepLinkId = 'addIntegrationCallout';
const ManageProjectsStepLinkId = exports.ManageProjectsStepLinkId = 'manageProjects';
