// @kbn/search-playground/public/analytics/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnalyticsEvents = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let AnalyticsEvents = exports.AnalyticsEvents = /*#__PURE__*/function (AnalyticsEvents) {
  AnalyticsEvents["chatPageLoaded"] = "chat_page_loaded";
  AnalyticsEvents["chatCleared"] = "chat_cleared";
  AnalyticsEvents["chatQuestionSent"] = "chat_question_sent";
  AnalyticsEvents["chatRequestStopped"] = "chat_request_stopped";
  AnalyticsEvents["chatRegenerateMessages"] = "chat_regenerate_messages";
  AnalyticsEvents["citationDetailsExpanded"] = "citation_details_expanded";
  AnalyticsEvents["citationDetailsCollapsed"] = "citation_details_collapsed";
  AnalyticsEvents["editContextFieldToggled"] = "edit_context_field_toggled";
  AnalyticsEvents["editContextDocSizeChanged"] = "edit_context_doc_size_changed";
  AnalyticsEvents["genAiConnectorAdded"] = "gen_ai_connector_added";
  AnalyticsEvents["genAiConnectorCreated"] = "gen_ai_connector_created";
  AnalyticsEvents["genAiConnectorExists"] = "gen_ai_connector_exists";
  AnalyticsEvents["genAiConnectorSetup"] = "gen_ai_connector_setup";
  AnalyticsEvents["includeCitations"] = "include_citations";
  AnalyticsEvents["instructionsFieldChanged"] = "instructions_field_changed";
  AnalyticsEvents["queryFieldsUpdated"] = "view_query_fields_updated";
  AnalyticsEvents["queryModeLoaded"] = "query_mode_loaded";
  AnalyticsEvents["modelSelected"] = "model_selected";
  AnalyticsEvents["retrievalDocsFlyoutOpened"] = "retrieval_docs_flyout_opened";
  AnalyticsEvents["sourceFieldsLoaded"] = "source_fields_loaded";
  AnalyticsEvents["sourceIndexUpdated"] = "source_index_updated";
  AnalyticsEvents["setupChatPageLoaded"] = "start_new_chat_page_loaded";
  AnalyticsEvents["viewCodeFlyoutOpened"] = "view_code_flyout_opened";
  AnalyticsEvents["viewCodeLanguageChange"] = "view_code_language_change";
  return AnalyticsEvents;
}({});
