// @kbn/expression-partition-vis-plugin/common/types/expression_renderers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ValueFormats = exports.LegendDisplay = exports.LabelPositions = exports.EmptySizeRatios = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let EmptySizeRatios = exports.EmptySizeRatios = /*#__PURE__*/function (EmptySizeRatios) {
  EmptySizeRatios[EmptySizeRatios["SMALL"] = 0.3] = "SMALL";
  EmptySizeRatios[EmptySizeRatios["MEDIUM"] = 0.54] = "MEDIUM";
  EmptySizeRatios[EmptySizeRatios["LARGE"] = 0.7] = "LARGE";
  return EmptySizeRatios;
}({});
let LabelPositions = exports.LabelPositions = /*#__PURE__*/function (LabelPositions) {
  LabelPositions["INSIDE"] = "inside";
  LabelPositions["DEFAULT"] = "default";
  return LabelPositions;
}({});
let ValueFormats = exports.ValueFormats = /*#__PURE__*/function (ValueFormats) {
  ValueFormats["PERCENT"] = "percent";
  ValueFormats["VALUE"] = "value";
  return ValueFormats;
}({});
let LegendDisplay = exports.LegendDisplay = /*#__PURE__*/function (LegendDisplay) {
  LegendDisplay["SHOW"] = "show";
  LegendDisplay["HIDE"] = "hide";
  LegendDisplay["DEFAULT"] = "default";
  return LegendDisplay;
}({});
