// @kbn/investigate-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ChromeOption", {
  enumerable: true,
  get: function () {
    return _types2.ChromeOption;
  }
});
Object.defineProperty(exports, "ESQL_WIDGET_NAME", {
  enumerable: true,
  get: function () {
    return _constants.ESQL_WIDGET_NAME;
  }
});
Object.defineProperty(exports, "InvestigateWidgetColumnSpan", {
  enumerable: true,
  get: function () {
    return _types.InvestigateWidgetColumnSpan;
  }
});
Object.defineProperty(exports, "createEsqlWidget", {
  enumerable: true,
  get: function () {
    return _create_esql_widget.createEsqlWidget;
  }
});
Object.defineProperty(exports, "createWidgetFactory", {
  enumerable: true,
  get: function () {
    return _create_widget.createWidgetFactory;
  }
});
Object.defineProperty(exports, "getEsFilterFromGlobalParameters", {
  enumerable: true,
  get: function () {
    return _get_es_filters_from_global_parameters.getEsFilterFromGlobalParameters;
  }
});
exports.plugin = void 0;
var _plugin = require("./plugin");
var _types = require("../common/types");
var _types2 = require("./types");
var _create_widget = require("./create_widget");
var _get_es_filters_from_global_parameters = require("./util/get_es_filters_from_global_parameters");
var _constants = require("./esql_widget/constants");
var _create_esql_widget = require("./esql_widget/create_esql_widget");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const plugin = pluginInitializerContext => new _plugin.InvestigatePlugin(pluginInitializerContext);
exports.plugin = plugin;
