// @kbn/investigate-plugin/public/plugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InvestigatePlugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _widget_registry = require("./widget_registry");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class InvestigatePlugin {
  constructor(context) {
    (0, _defineProperty2.default)(this, "logger", void 0);
    (0, _defineProperty2.default)(this, "widgetRegistry", new _widget_registry.WidgetRegistry());
    this.logger = context.logger.get();
  }
  setup(coreSetup, pluginsSetup) {
    return {
      registerWidget: this.widgetRegistry.registerWidget
    };
  }
  start(coreStart, pluginsStart) {
    return {
      getWidgetDefinitions: this.widgetRegistry.getWidgetDefinitions
    };
  }
}
exports.InvestigatePlugin = InvestigatePlugin;
