// @kbn/rollup-plugin/public/crud_app/sections/components/job_details/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_HISTOGRAM", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_HISTOGRAM;
  }
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_JSON", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_JSON;
  }
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_METRICS", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_METRICS;
  }
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_REQUEST", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_REQUEST;
  }
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_SUMMARY", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_SUMMARY;
  }
});
Object.defineProperty(exports, "JOB_DETAILS_TAB_TERMS", {
  enumerable: true,
  get: function () {
    return _job_details.JOB_DETAILS_TAB_TERMS;
  }
});
Object.defineProperty(exports, "JobDetails", {
  enumerable: true,
  get: function () {
    return _job_details.JobDetails;
  }
});
Object.defineProperty(exports, "tabToHumanizedMap", {
  enumerable: true,
  get: function () {
    return _job_details.tabToHumanizedMap;
  }
});
var _job_details = require("./job_details");
