// @kbn/expression-heatmap-plugin/public/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getDatatableUtilities", {
  enumerable: true,
  get: function () {
    return _datatable_utilities.getDatatableUtilities;
  }
});
Object.defineProperty(exports, "getFormatService", {
  enumerable: true,
  get: function () {
    return _format_service.getFormatService;
  }
});
Object.defineProperty(exports, "getPaletteService", {
  enumerable: true,
  get: function () {
    return _palette_service.getPaletteService;
  }
});
Object.defineProperty(exports, "getUISettings", {
  enumerable: true,
  get: function () {
    return _ui_settings.getUISettings;
  }
});
Object.defineProperty(exports, "setDatatableUtilities", {
  enumerable: true,
  get: function () {
    return _datatable_utilities.setDatatableUtilities;
  }
});
Object.defineProperty(exports, "setFormatService", {
  enumerable: true,
  get: function () {
    return _format_service.setFormatService;
  }
});
Object.defineProperty(exports, "setPaletteService", {
  enumerable: true,
  get: function () {
    return _palette_service.setPaletteService;
  }
});
Object.defineProperty(exports, "setUISettings", {
  enumerable: true,
  get: function () {
    return _ui_settings.setUISettings;
  }
});
var _datatable_utilities = require("./datatable_utilities");
var _format_service = require("./format_service");
var _palette_service = require("./palette_service");
var _ui_settings = require("./ui_settings");
