// @kbn/spaces-plugin/public/space_selector/components/space_cards.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SpaceCards = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
require("./space_cards.scss");
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _space_card = require("./space_card");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class SpaceCards extends _react.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "renderSpace", space => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      key: space.id,
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_space_card.SpaceCard, {
      space: space,
      serverBasePath: this.props.serverBasePath
    })));
  }
  render() {
    return /*#__PURE__*/_react.default.createElement("div", {
      className: "spaceCards"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      gutterSize: "l",
      justifyContent: "center",
      wrap: true,
      responsive: false
    }, this.props.spaces.map(this.renderSpace)));
  }
}
exports.SpaceCards = SpaceCards;
