// @kbn/enterprise-search-plugin/public/applications/workplace_search/views/groups/components/source_option_item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SourceOptionItem = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _truncate = require("../../../../shared/truncate");
var _source_icon = require("../../../components/shared/source_icon");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const MAX_LENGTH = 28;
const SourceOptionItem = ({
  source
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  gutterSize: "m",
  justifyContent: "flexStart",
  alignItems: "center",
  responsive: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_source_icon.SourceIcon, (0, _extends2.default)({}, source, {
  size: "s"
}))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_truncate.TruncatedContent, {
  tooltipType: "title",
  content: source.name,
  length: MAX_LENGTH
})));
exports.SourceOptionItem = SourceOptionItem;
