// @kbn/unified-doc-viewer-plugin/public/components/doc_viewer_logs_overview/logs_overview.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogsOverview = LogsOverview;
var _react = _interopRequireDefault(require("react"));
var _discoverUtils = require("@kbn/discover-utils");
var _eui = require("@elastic/eui");
var _logs_overview_header = require("./logs_overview_header");
var _logs_overview_highlights = require("./logs_overview_highlights");
var _use_field_actions = require("../../hooks/use_field_actions");
var _plugin = require("../../plugin");
var _logs_overview_ai_assistant = require("./logs_overview_ai_assistant");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function LogsOverview({
  columns,
  dataView,
  hit,
  filter,
  onAddColumn,
  onRemoveColumn
}) {
  const {
    fieldFormats
  } = (0, _plugin.getUnifiedDocViewerServices)();
  const parsedDoc = (0, _discoverUtils.getLogDocumentOverview)(hit, {
    dataView,
    fieldFormats
  });
  return /*#__PURE__*/_react.default.createElement(_use_field_actions.FieldActionsProvider, {
    columns: columns,
    filter: filter,
    onAddColumn: onAddColumn,
    onRemoveColumn: onRemoveColumn
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_logs_overview_header.LogsOverviewHeader, {
    doc: parsedDoc
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    margin: "xs"
  }), /*#__PURE__*/_react.default.createElement(_logs_overview_highlights.LogsOverviewHighlights, {
    formattedDoc: parsedDoc,
    flattenedDoc: hit.flattened
  }), /*#__PURE__*/_react.default.createElement(_logs_overview_ai_assistant.LogsOverviewAIAssistant, {
    doc: hit
  }));
}
