// @kbn/securitysolution-io-ts-alerting-types/src/threat_technique/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.threat_techniques = exports.threat_technique_reference = exports.threat_technique_name = exports.threat_technique_id = exports.threat_technique = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _threat_subtechnique = require("../threat_subtechnique");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/* eslint-disable @typescript-eslint/naming-convention */

const threat_technique_id = exports.threat_technique_id = t.string;
const threat_technique_name = exports.threat_technique_name = t.string;
const threat_technique_reference = exports.threat_technique_reference = t.string;
const threat_technique = exports.threat_technique = t.intersection([t.exact(t.type({
  id: threat_technique_id,
  name: threat_technique_name,
  reference: threat_technique_reference
})), t.exact(t.partial({
  subtechnique: _threat_subtechnique.threat_subtechniques
}))]);
const threat_techniques = exports.threat_techniques = t.array(threat_technique);
