// @kbn/lens-plugin/public/shared_components/axis/ticks/axis_ticks_settings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AxisTicksSettings = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AxisTicksSettings = ({
  axis,
  isAxisLabelVisible,
  updateTicksVisibilityState
}) => {
  const onTicksStatusChange = (0, _react.useCallback)(visible => updateTicksVisibilityState(visible, axis), [axis, updateTicksVisibilityState]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    display: "columnCompressedSwitch",
    label: _i18n.i18n.translate('xpack.lens.shared.tickLabels', {
      defaultMessage: 'Tick labels'
    }),
    fullWidth: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    compressed: true,
    "data-test-subj": `lnsshow${axis}AxisTickLabels`,
    label: _i18n.i18n.translate('xpack.lens.shared.tickLabels', {
      defaultMessage: 'Tick labels'
    }),
    onChange: () => onTicksStatusChange(!isAxisLabelVisible),
    checked: isAxisLabelVisible,
    showLabel: false
  })));
};
exports.AxisTicksSettings = AxisTicksSettings;
