// @kbn/cases-plugin/public/components/all_cases/table_filter_config/more_filters_selectable.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoreFiltersSelectable = void 0;
var _react = _interopRequireDefault(require("react"));
var _multi_select_filter = require("../multi_select_filter");
var _translations = require("../translations");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const MoreFiltersSelectable = ({
  options,
  activeFilters,
  onChange,
  isLoading
}) => {
  return /*#__PURE__*/_react.default.createElement(_multi_select_filter.MultiSelectFilter, {
    buttonLabel: _translations.MORE_FILTERS_LABEL,
    buttonIconType: "plus",
    hideActiveOptionsNumber: true,
    id: "filters",
    onChange: onChange,
    options: options,
    selectedOptionKeys: activeFilters,
    transparentBackground: true,
    isLoading: isLoading
  });
};
exports.MoreFiltersSelectable = MoreFiltersSelectable;
MoreFiltersSelectable.displayName = 'MoreFiltersSelectable';
