// @kbn/cloud-security-posture-plugin/public/components/empty_state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmptyState = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _react2 = require("@emotion/react");
var _test_subjects = require("./test_subjects");
var _constants = require("../common/constants");
var _illustration_product_no_results_magnifying_glass = _interopRequireDefault(require("../assets/illustrations/illustration_product_no_results_magnifying_glass.svg"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EmptyState = ({
  onResetFilters,
  docsUrl = _constants.FINDINGS_DOCS_URL
}) => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    css: (0, _react2.css)`
        max-width: 734px;
        && > .euiEmptyPrompt__main {
          gap: ${euiTheme.size.xl};
        }
        && {
          margin-top: ${euiTheme.size.xxxl}};
        }
      `,
    "data-test-subj": _test_subjects.EMPTY_STATE_TEST_SUBJ,
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiImage, {
      url: _illustration_product_no_results_magnifying_glass.default,
      alt: _i18n.i18n.translate('xpack.csp.emptyState.illustrationAlt', {
        defaultMessage: 'No results'
      }),
      css: (0, _react2.css)`
            width: 290px;
          `
    }),
    title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.csp.emptyState.title",
      defaultMessage: "No results match your search criteria"
    })),
    layout: "horizontal",
    color: "plain",
    body: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.csp.emptyState.description",
      defaultMessage: "Try modifying your search or filter set"
    }))),
    actions: [/*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      color: "primary",
      fill: true,
      onClick: onResetFilters
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.csp.emptyState.resetFiltersButton",
      defaultMessage: "Reset filters"
    })), /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
      href: docsUrl,
      target: "_blank"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.csp.emptyState.readDocsLink",
      defaultMessage: "Read the docs"
    }))]
  });
};
exports.EmptyState = EmptyState;
