// @kbn/cloud-security-posture-plugin/public/components/test_subjects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VULNERABILITIES_CVSS_SCORE_BADGE_SUBJ = exports.VULNERABILITIES_CONTAINER_TEST_SUBJ = exports.TAKE_ACTION_SUBJ = exports.SETUP_TECHNOLOGY_SELECTOR_TEST_SUBJ = exports.SETUP_TECHNOLOGY_SELECTOR_ACCORDION_TEST_SUBJ = exports.NO_VULNERABILITIES_STATUS_TEST_SUBJ = exports.NO_FINDINGS_STATUS_TEST_SUBJ = exports.KSPM_NOT_INSTALLED_ACTION_SUBJ = exports.GCP_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = exports.EMPTY_STATE_TEST_SUBJ = exports.CSP_GROUPING_LOADING = exports.CSP_GROUPING = exports.CSP_FINDINGS_COMPLIANCE_SCORE = exports.CSP_FIELDS_SELECTOR_RESET_BUTTON = exports.CSP_FIELDS_SELECTOR_OPEN_BUTTON = exports.CSP_FIELDS_SELECTOR_MODAL = exports.CSP_FIELDS_SELECTOR_CLOSE_BUTTON = exports.CSPM_NOT_INSTALLED_ACTION_SUBJ = exports.CREATE_RULE_ACTION_SUBJ = exports.CNVM_NOT_INSTALLED_ACTION_SUBJ = exports.CIS_GCP_OPTION_TEST_SUBJ = exports.CIS_GCP_INPUT_FIELDS_TEST_SUBJECTS = exports.CIS_AZURE_SETUP_FORMAT_TEST_SUBJECTS = exports.CIS_AZURE_OPTION_TEST_SUBJ = exports.CIS_AZURE_INPUT_FIELDS_TEST_SUBJECTS = exports.CHART_PANEL_TEST_SUBJECTS = exports.AZURE_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = exports.AWS_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = exports.AWS_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CHART_PANEL_TEST_SUBJECTS = exports.CHART_PANEL_TEST_SUBJECTS = {
  LOADING: 'chart_is_loading',
  EMPTY: 'chart_is_empty',
  ERROR: 'chart_is_error',
  TEST_CHART: 'testing_chart'
};
const NO_FINDINGS_STATUS_TEST_SUBJ = exports.NO_FINDINGS_STATUS_TEST_SUBJ = {
  NO_AGENTS_DEPLOYED: 'status-api-no-agent-deployed',
  INDEXING: 'status-api-indexing',
  INDEX_TIMEOUT: 'status-api-index-timeout',
  UNPRIVILEGED: 'status-api-unprivileged',
  NO_FINDINGS: 'no-findings-found'
};
const EMPTY_STATE_TEST_SUBJ = exports.EMPTY_STATE_TEST_SUBJ = 'csp:empty-state';
const NO_VULNERABILITIES_STATUS_TEST_SUBJ = exports.NO_VULNERABILITIES_STATUS_TEST_SUBJ = {
  SCANNING_VULNERABILITIES: 'scanning-vulnerabilities-empty-prompt',
  NOT_INSTALLED: 'cnvm-integration-not-installed',
  NOT_DEPLOYED: 'agent-not-deployed-vuln-mgmt',
  UNPRIVILEGED: 'status-api-vuln-mgmt-unprivileged',
  NO_VULNERABILITIES: 'no-vulnerabilities-vuln-mgmt-found',
  INDEX_TIMEOUT: 'vulnerabilities-timeout'
};
const CNVM_NOT_INSTALLED_ACTION_SUBJ = exports.CNVM_NOT_INSTALLED_ACTION_SUBJ = 'cnvm-not-installed-action';
const CSPM_NOT_INSTALLED_ACTION_SUBJ = exports.CSPM_NOT_INSTALLED_ACTION_SUBJ = 'cspm-not-installed-action';
const KSPM_NOT_INSTALLED_ACTION_SUBJ = exports.KSPM_NOT_INSTALLED_ACTION_SUBJ = 'kspm-not-installed-action';
const VULNERABILITIES_CONTAINER_TEST_SUBJ = exports.VULNERABILITIES_CONTAINER_TEST_SUBJ = 'vulnerabilities_container';
const VULNERABILITIES_CVSS_SCORE_BADGE_SUBJ = exports.VULNERABILITIES_CVSS_SCORE_BADGE_SUBJ = 'vulnerabilities_cvss_score_badge';
const TAKE_ACTION_SUBJ = exports.TAKE_ACTION_SUBJ = 'csp:take_action';
const CREATE_RULE_ACTION_SUBJ = exports.CREATE_RULE_ACTION_SUBJ = 'csp:create_rule';
const CSP_GROUPING = exports.CSP_GROUPING = 'cloudSecurityGrouping';
const CSP_GROUPING_LOADING = exports.CSP_GROUPING_LOADING = 'cloudSecurityGroupingLoading';
const CSP_FINDINGS_COMPLIANCE_SCORE = exports.CSP_FINDINGS_COMPLIANCE_SCORE = 'cloudSecurityFindingsComplianceScore';
const CSP_FIELDS_SELECTOR_MODAL = exports.CSP_FIELDS_SELECTOR_MODAL = 'cloudSecurityFieldsSelectorModal';
const CSP_FIELDS_SELECTOR_OPEN_BUTTON = exports.CSP_FIELDS_SELECTOR_OPEN_BUTTON = 'cloudSecurityFieldsSelectorOpenButton';
const CSP_FIELDS_SELECTOR_RESET_BUTTON = exports.CSP_FIELDS_SELECTOR_RESET_BUTTON = 'cloudSecurityFieldsSelectorResetButton';
const CSP_FIELDS_SELECTOR_CLOSE_BUTTON = exports.CSP_FIELDS_SELECTOR_CLOSE_BUTTON = 'cloudSecurityFieldsSelectorCloseButton';
const AWS_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = exports.AWS_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = 'aws-credentials-type-selector';
const AWS_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = exports.AWS_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = {
  CLOUDFORMATION: 'aws-cloudformation-setup-option',
  MANUAL: 'aws-manual-setup-option'
};
const GCP_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = exports.GCP_CREDENTIALS_TYPE_OPTIONS_TEST_SUBJ = {
  CLOUD_SHELL: 'gcpGoogleCloudShellOptionTestId',
  MANUAL: 'gcpManualOptionTestId'
};
const CIS_GCP_OPTION_TEST_SUBJ = exports.CIS_GCP_OPTION_TEST_SUBJ = 'cisGcpTestId';
const CIS_AZURE_OPTION_TEST_SUBJ = exports.CIS_AZURE_OPTION_TEST_SUBJ = 'cisAzureTestId';
const SETUP_TECHNOLOGY_SELECTOR_ACCORDION_TEST_SUBJ = exports.SETUP_TECHNOLOGY_SELECTOR_ACCORDION_TEST_SUBJ = 'setup-technology-selector-accordion';
const SETUP_TECHNOLOGY_SELECTOR_TEST_SUBJ = exports.SETUP_TECHNOLOGY_SELECTOR_TEST_SUBJ = 'setup-technology-selector';
const AZURE_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = exports.AZURE_CREDENTIALS_TYPE_SELECTOR_TEST_SUBJ = 'azure-credentials-type-selector';
const CIS_AZURE_INPUT_FIELDS_TEST_SUBJECTS = exports.CIS_AZURE_INPUT_FIELDS_TEST_SUBJECTS = {
  TENANT_ID: 'cisAzureTenantId',
  CLIENT_ID: 'cisAzureClientId',
  CLIENT_SECRET: 'passwordInput-client-secret',
  CLIENT_CERTIFICATE_PATH: 'cisAzureClientCertificatePath',
  CLIENT_CERTIFICATE_PASSWORD: 'cisAzureClientCertificatePassword',
  CLIENT_USERNAME: 'cisAzureClientUsername',
  CLIENT_PASSWORD: 'cisAzureClientPassword'
};
const CIS_AZURE_SETUP_FORMAT_TEST_SUBJECTS = exports.CIS_AZURE_SETUP_FORMAT_TEST_SUBJECTS = {
  ARM_TEMPLATE: 'cisAzureArmTemplate',
  MANUAL: 'cisAzureManual'
};
const CIS_GCP_INPUT_FIELDS_TEST_SUBJECTS = exports.CIS_GCP_INPUT_FIELDS_TEST_SUBJECTS = {
  GOOGLE_CLOUD_SHELL_SETUP: 'google_cloud_shell_setup_test_id',
  PROJECT_ID: 'project_id_test_id',
  ORGANIZATION_ID: 'organization_id_test_id',
  CREDENTIALS_TYPE: 'credentials_type_test_id',
  CREDENTIALS_FILE: 'credentials_file_test_id',
  CREDENTIALS_JSON: 'credentials_json_test_id'
};
