// @elastic/eui/lib/components/markdown_editor/markdown_editor_drop_zone.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiMarkdownEditorDropZoneStyles = void 0;
var _react = require("@emotion/react");
var _services = require("../../services");
var _global_styling = require("../../global_styling");
var _markdown_editor = require("./markdown_editor.styles");
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
var _ref = process.env.NODE_ENV === "production" ? {
  name: "gwjnpv-euiMarkdownEditorDropZone__input",
  styles: "position:absolute;inset:0;opacity:0;overflow:hidden;&:hover{cursor:pointer;}&:hover:disabled{cursor:not-allowed;};label:euiMarkdownEditorDropZone__input;"
} : {
  name: "gwjnpv-euiMarkdownEditorDropZone__input",
  styles: "position:absolute;inset:0;opacity:0;overflow:hidden;&:hover{cursor:pointer;}&:hover:disabled{cursor:not-allowed;};label:euiMarkdownEditorDropZone__input;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var euiMarkdownEditorDropZoneStyles = exports.euiMarkdownEditorDropZoneStyles = function euiMarkdownEditorDropZoneStyles(euiThemeContext) {
  var euiTheme = euiThemeContext.euiTheme;
  var _euiMarkdownEditorVar = (0, _markdown_editor.euiMarkdownEditorVariables)(euiThemeContext),
    minHeight = _euiMarkdownEditorVar.minHeight;
  return {
    euiMarkdownEditorDropZone: /*#__PURE__*/(0, _react.css)("position:relative;display:flex;flex-direction:column;", (0, _global_styling.logicalCSS)('min-height', minHeight), ";;label:euiMarkdownEditorDropZone;"),
    isDragging: /*#__PURE__*/(0, _react.css)(".euiMarkdownEditorFooter,.euiMarkdownEditorTextArea{background-color:", (0, _services.transparentize)(euiTheme.colors.primary, 0.1), "!important;};label:isDragging;"),
    isDraggingError: /*#__PURE__*/(0, _react.css)(".euiMarkdownEditorFooter,.euiMarkdownEditorTextArea{background-color:", (0, _services.transparentize)(euiTheme.colors.danger, 0.1), "!important;};label:isDraggingError;"),
    hasError: /*#__PURE__*/(0, _react.css)(".euiMarkdownEditorTextArea{--euiFormControlStateColor:", euiTheme.colors.danger, "!important;background-size:100% 100%;};label:hasError;"),
    euiMarkdownEditorDropZone__input: _ref
  };
};
