// @kbn/observability-shared-plugin/common/field_names/synthetics.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYNTHETICS_WAIT_TIMINGS = exports.SYNTHETICS_TOTAL_TIMINGS = exports.SYNTHETICS_STEP_NAME = exports.SYNTHETICS_STEP_DURATION = exports.SYNTHETICS_SSL_TIMINGS = exports.SYNTHETICS_SEND_TIMINGS = exports.SYNTHETICS_RECEIVE_TIMINGS = exports.SYNTHETICS_LCP = exports.SYNTHETICS_FCP = exports.SYNTHETICS_DOCUMENT_ONLOAD = exports.SYNTHETICS_DNS_TIMINGS = exports.SYNTHETICS_DCL = exports.SYNTHETICS_CONNECT_TIMINGS = exports.SYNTHETICS_CLS = exports.SYNTHETICS_BLOCKED_TIMINGS = exports.NETWORK_TIMINGS_FIELDS = exports.MONITOR_DURATION_US = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const MONITOR_DURATION_US = exports.MONITOR_DURATION_US = 'monitor.duration.us';
const SYNTHETICS_CLS = exports.SYNTHETICS_CLS = 'browser.experience.cls';
const SYNTHETICS_LCP = exports.SYNTHETICS_LCP = 'browser.experience.lcp.us';
const SYNTHETICS_FCP = exports.SYNTHETICS_FCP = 'browser.experience.fcp.us';
const SYNTHETICS_DOCUMENT_ONLOAD = exports.SYNTHETICS_DOCUMENT_ONLOAD = 'browser.experience.load.us';
const SYNTHETICS_DCL = exports.SYNTHETICS_DCL = 'browser.experience.dcl.us';
const SYNTHETICS_STEP_NAME = exports.SYNTHETICS_STEP_NAME = 'synthetics.step.name.keyword';
const SYNTHETICS_STEP_DURATION = exports.SYNTHETICS_STEP_DURATION = 'synthetics.step.duration.us';
const SYNTHETICS_DNS_TIMINGS = exports.SYNTHETICS_DNS_TIMINGS = 'synthetics.payload.timings.dns';
const SYNTHETICS_SSL_TIMINGS = exports.SYNTHETICS_SSL_TIMINGS = 'synthetics.payload.timings.ssl';
const SYNTHETICS_BLOCKED_TIMINGS = exports.SYNTHETICS_BLOCKED_TIMINGS = 'synthetics.payload.timings.blocked';
const SYNTHETICS_CONNECT_TIMINGS = exports.SYNTHETICS_CONNECT_TIMINGS = 'synthetics.payload.timings.connect';
const SYNTHETICS_RECEIVE_TIMINGS = exports.SYNTHETICS_RECEIVE_TIMINGS = 'synthetics.payload.timings.receive';
const SYNTHETICS_SEND_TIMINGS = exports.SYNTHETICS_SEND_TIMINGS = 'synthetics.payload.timings.send';
const SYNTHETICS_WAIT_TIMINGS = exports.SYNTHETICS_WAIT_TIMINGS = 'synthetics.payload.timings.wait';
const SYNTHETICS_TOTAL_TIMINGS = exports.SYNTHETICS_TOTAL_TIMINGS = 'synthetics.payload.timings.total';
const NETWORK_TIMINGS_FIELDS = exports.NETWORK_TIMINGS_FIELDS = [SYNTHETICS_DNS_TIMINGS, SYNTHETICS_SSL_TIMINGS, SYNTHETICS_BLOCKED_TIMINGS, SYNTHETICS_CONNECT_TIMINGS, SYNTHETICS_RECEIVE_TIMINGS, SYNTHETICS_SEND_TIMINGS, SYNTHETICS_WAIT_TIMINGS, SYNTHETICS_TOTAL_TIMINGS];
