// @kbn/vis-type-table-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUsageCollectionStart = exports.setFormatService = exports.setDataViewsStart = exports.getUsageCollectionStart = exports.getFormatService = exports.getDataViewsStart = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getFormatService, setFormatService] = (0, _public.createGetterSetter)('FieldFormats');
exports.setFormatService = setFormatService;
exports.getFormatService = getFormatService;
const [getUsageCollectionStart, setUsageCollectionStart] = (0, _public.createGetterSetter)('UsageCollection', false);
exports.setUsageCollectionStart = setUsageCollectionStart;
exports.getUsageCollectionStart = getUsageCollectionStart;
const [getDataViewsStart, setDataViewsStart] = (0, _public.createGetterSetter)('dataViews');
exports.setDataViewsStart = setDataViewsStart;
exports.getDataViewsStart = getDataViewsStart;
