// @kbn/snapshot-restore-plugin/public/application/sections/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "PolicyAdd", {
  enumerable: true,
  get: function () {
    return _policy_add.PolicyAdd;
  }
});
Object.defineProperty(exports, "PolicyEdit", {
  enumerable: true,
  get: function () {
    return _policy_edit.PolicyEdit;
  }
});
Object.defineProperty(exports, "RepositoryAdd", {
  enumerable: true,
  get: function () {
    return _repository_add.RepositoryAdd;
  }
});
Object.defineProperty(exports, "RepositoryEdit", {
  enumerable: true,
  get: function () {
    return _repository_edit.RepositoryEdit;
  }
});
Object.defineProperty(exports, "RestoreSnapshot", {
  enumerable: true,
  get: function () {
    return _restore_snapshot.RestoreSnapshot;
  }
});
Object.defineProperty(exports, "SnapshotRestoreHome", {
  enumerable: true,
  get: function () {
    return _home.SnapshotRestoreHome;
  }
});
var _home = require("./home");
var _repository_add = require("./repository_add");
var _repository_edit = require("./repository_edit");
var _restore_snapshot = require("./restore_snapshot");
var _policy_add = require("./policy_add");
var _policy_edit = require("./policy_edit");
