// @kbn/ingest-pipelines-plugin/public/application/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ApiService", {
  enumerable: true,
  get: function () {
    return _api.ApiService;
  }
});
Object.defineProperty(exports, "BreadcrumbService", {
  enumerable: true,
  get: function () {
    return _breadcrumbs.BreadcrumbService;
  }
});
Object.defineProperty(exports, "DocumentationService", {
  enumerable: true,
  get: function () {
    return _documentation.DocumentationService;
  }
});
Object.defineProperty(exports, "FileReaderService", {
  enumerable: true,
  get: function () {
    return _file_reader.FileReaderService;
  }
});
Object.defineProperty(exports, "UiMetricService", {
  enumerable: true,
  get: function () {
    return _ui_metric.UiMetricService;
  }
});
Object.defineProperty(exports, "apiService", {
  enumerable: true,
  get: function () {
    return _api.apiService;
  }
});
Object.defineProperty(exports, "breadcrumbService", {
  enumerable: true,
  get: function () {
    return _breadcrumbs.breadcrumbService;
  }
});
Object.defineProperty(exports, "documentationService", {
  enumerable: true,
  get: function () {
    return _documentation.documentationService;
  }
});
Object.defineProperty(exports, "fileReaderService", {
  enumerable: true,
  get: function () {
    return _file_reader.fileReaderService;
  }
});
Object.defineProperty(exports, "uiMetricService", {
  enumerable: true,
  get: function () {
    return _ui_metric.uiMetricService;
  }
});
var _documentation = require("./documentation");
var _ui_metric = require("./ui_metric");
var _api = require("./api");
var _breadcrumbs = require("./breadcrumbs");
var _file_reader = require("./file_reader");
