// @kbn/remote-clusters-plugin/public/application/services/documentation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cloudSetupTrustUrl = void 0;
exports.init = init;
exports.transportPortUrl = exports.skippingDisconnectedClustersUrl = exports.remoteClustersUrl = exports.proxySettingsUrl = exports.proxyModeUrl = exports.onPremSetupTrustWithCertUrl = exports.onPremSetupTrustWithApiKeyUrl = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

let skippingDisconnectedClustersUrl = exports.skippingDisconnectedClustersUrl = void 0;
let remoteClustersUrl = exports.remoteClustersUrl = void 0;
let transportPortUrl = exports.transportPortUrl = void 0;
let proxyModeUrl = exports.proxyModeUrl = void 0;
let proxySettingsUrl = exports.proxySettingsUrl = void 0;
let onPremSetupTrustWithCertUrl = exports.onPremSetupTrustWithCertUrl = void 0;
let onPremSetupTrustWithApiKeyUrl = exports.onPremSetupTrustWithApiKeyUrl = void 0;
let cloudSetupTrustUrl = exports.cloudSetupTrustUrl = void 0;
function init({
  links
}) {
  exports.skippingDisconnectedClustersUrl = skippingDisconnectedClustersUrl = links.ccs.skippingDisconnectedClusters;
  exports.remoteClustersUrl = remoteClustersUrl = links.elasticsearch.remoteClusters;
  exports.transportPortUrl = transportPortUrl = links.elasticsearch.transportSettings;
  exports.proxyModeUrl = proxyModeUrl = links.elasticsearch.remoteClustersProxy;
  exports.proxySettingsUrl = proxySettingsUrl = links.elasticsearch.remoteClusersProxySettings;
  exports.onPremSetupTrustWithCertUrl = onPremSetupTrustWithCertUrl = links.elasticsearch.remoteClustersOnPremSetupTrustWithCert;
  exports.onPremSetupTrustWithApiKeyUrl = onPremSetupTrustWithApiKeyUrl = links.elasticsearch.remoteClustersOnPremSetupTrustWithApiKey;
  exports.cloudSetupTrustUrl = cloudSetupTrustUrl = links.elasticsearch.remoteClustersCloudSetupTrust;
}
