// @kbn/logs-explorer-plugin/public/state_machines/logs_explorer_controller/src/notifications.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createDatasetSelectionRestoreFailedNotifier = exports.createDataViewSelectionRestoreFailedNotifier = exports.createCreateDataViewFailedNotifier = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createDatasetSelectionRestoreFailedNotifier = toasts => () => toasts.addWarning({
  title: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.restoreDatasetSelectionFailedToastTitle', {
    defaultMessage: "We couldn't restore your datasets selection."
  }),
  text: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.restoreDatasetSelectionFailedToastMessage', {
    defaultMessage: 'We switched to "All log datasets" as the default selection.'
  })
});
exports.createDatasetSelectionRestoreFailedNotifier = createDatasetSelectionRestoreFailedNotifier;
const createDataViewSelectionRestoreFailedNotifier = toasts => () => toasts.addWarning({
  title: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.restoreDataViewSelectionFailedToastTitle', {
    defaultMessage: "We couldn't restore your data view selection."
  }),
  text: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.restoreDataViewSelectionFailedToastMessage', {
    defaultMessage: 'We switched to "All log datasets" as the default selection.'
  })
});
exports.createDataViewSelectionRestoreFailedNotifier = createDataViewSelectionRestoreFailedNotifier;
const createCreateDataViewFailedNotifier = toasts => () => toasts.addWarning({
  title: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.createAdHocDataViewFailedToastTitle', {
    defaultMessage: "We couldn't create a data view for your selection."
  }),
  text: _i18n.i18n.translate('xpack.logsExplorer.dataSourceSelection.createAdHocDataViewFailedToastMessage', {
    defaultMessage: 'We switched to "All log datasets" as the default selection.'
  })
});
exports.createCreateDataViewFailedNotifier = createCreateDataViewFailedNotifier;
