// @kbn/stack-connectors-plugin/public/connector_types/cases_webhook/action_variables.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.urlVarsExt = exports.urlVars = exports.commentVars = exports.casesVars = void 0;
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const casesVars = exports.casesVars = [{
  name: 'case.title',
  description: i18n.CASE_TITLE_DESC,
  useWithTripleBracesInTemplates: true
}, {
  name: 'case.description',
  description: i18n.CASE_DESCRIPTION_DESC,
  useWithTripleBracesInTemplates: true
}, {
  name: 'case.tags',
  description: i18n.CASE_TAGS_DESC,
  useWithTripleBracesInTemplates: true
}, {
  name: 'case.id',
  description: i18n.CASE_ID_DESC,
  useWithTripleBracesInTemplates: true
}, {
  name: 'case.status',
  description: i18n.CASE_SEVERITY_DESC,
  useWithTripleBracesInTemplates: true
}, {
  name: 'case.severity',
  description: i18n.CASE_STATUS_DESC,
  useWithTripleBracesInTemplates: true
}];
const commentVars = exports.commentVars = [{
  name: 'case.comment',
  description: i18n.CASE_COMMENT_DESC,
  useWithTripleBracesInTemplates: true
}];
const urlVars = exports.urlVars = [{
  name: 'external.system.id',
  description: i18n.EXTERNAL_ID_DESC,
  useWithTripleBracesInTemplates: true
}];
const urlVarsExt = exports.urlVarsExt = [...urlVars, {
  name: 'external.system.title',
  description: i18n.EXTERNAL_TITLE_DESC,
  useWithTripleBracesInTemplates: true
}];
