// @kbn/security-solution-plugin/public/attack_discovery/attack_discovery_panel/actions/actions_placeholder/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionsPlaceholder = void 0;
var _react = require("@emotion/react");
var _eui = require("@elastic/eui");
var _react2 = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ActionsPlaceholderComponent = () => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return /*#__PURE__*/_react2.default.createElement("div", {
    css: (0, _react.css)`
        margin-left: ${euiTheme.size.m};
        width: 400px;
      `,
    "data-test-subj": "actionsPlaceholder"
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "none"
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiSkeletonTitle, {
    css: (0, _react.css)`
              inline-size: 100%;
              width: 120px;
            `,
    "data-test-subj": "skeletonTitle1",
    isLoading: true,
    size: "s"
  })), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiSkeletonTitle, {
    css: (0, _react.css)`
              inline-size: 100%;
              width: 120px;
            `,
    "data-test-subj": "skeletonTitle2",
    isLoading: true,
    size: "s"
  })), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiSkeletonTitle, {
    css: (0, _react.css)`
              inline-size: 100%;
              width: 120px;
            `,
    "data-test-subj": "skeletonTitle3",
    isLoading: true,
    size: "s"
  }))));
};
ActionsPlaceholderComponent.displayName = 'ActionsPlaceholder';
const ActionsPlaceholder = exports.ActionsPlaceholder = /*#__PURE__*/_react2.default.memo(ActionsPlaceholderComponent);
