// @kbn/data-view-field-editor-plugin/public/components/field_editor/form_fields/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CustomDescriptionField", {
  enumerable: true,
  get: function () {
    return _custom_description_field.CustomDescriptionField;
  }
});
Object.defineProperty(exports, "CustomLabelField", {
  enumerable: true,
  get: function () {
    return _custom_label_field.CustomLabelField;
  }
});
Object.defineProperty(exports, "FormatField", {
  enumerable: true,
  get: function () {
    return _format_field.FormatField;
  }
});
Object.defineProperty(exports, "PopularityField", {
  enumerable: true,
  get: function () {
    return _popularity_field.PopularityField;
  }
});
Object.defineProperty(exports, "ScriptField", {
  enumerable: true,
  get: function () {
    return _script_field.ScriptField;
  }
});
Object.defineProperty(exports, "TypeField", {
  enumerable: true,
  get: function () {
    return _type_field.TypeField;
  }
});
var _type_field = require("./type_field");
var _custom_label_field = require("./custom_label_field");
var _custom_description_field = require("./custom_description_field");
var _popularity_field = require("./popularity_field");
var _script_field = require("./script_field");
var _format_field = require("./format_field");
