// @kbn/core-ui-settings-server-internal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CoreUiSettingsRouteHandlerContext", {
  enumerable: true,
  get: function () {
    return _src.CoreUiSettingsRouteHandlerContext;
  }
});
Object.defineProperty(exports, "UiSettingsClient", {
  enumerable: true,
  get: function () {
    return _src.UiSettingsClient;
  }
});
Object.defineProperty(exports, "UiSettingsGlobalClient", {
  enumerable: true,
  get: function () {
    return _src.UiSettingsGlobalClient;
  }
});
Object.defineProperty(exports, "UiSettingsService", {
  enumerable: true,
  get: function () {
    return _src.UiSettingsService;
  }
});
Object.defineProperty(exports, "createOrUpgradeSavedConfig", {
  enumerable: true,
  get: function () {
    return _create_or_upgrade_saved_config.createOrUpgradeSavedConfig;
  }
});
Object.defineProperty(exports, "uiSettingsConfig", {
  enumerable: true,
  get: function () {
    return _src.uiSettingsConfig;
  }
});
var _src = require("./src");
var _create_or_upgrade_saved_config = require("./src/create_or_upgrade_saved_config");
