// @kbn/core-rendering-server-internal/src/views/template.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Template = void 0;
var _react = _interopRequireWildcard(require("react"));
var _coreBaseCommon = require("@kbn/core-base-common");
var _fonts = require("./fonts");
var _logo = require("./logo");
var _styles = require("./styles");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const Template = ({
  metadata: {
    uiPublicUrl,
    locale,
    darkMode,
    stylesheetPaths,
    scriptPaths,
    injectedMetadata,
    i18n,
    bootstrapScriptUrl,
    strictCsp,
    customBranding
  }
}) => {
  var _customBranding$pageT, _customBranding$favic, _customBranding$favic2;
  const title = (_customBranding$pageT = customBranding.pageTitle) !== null && _customBranding$pageT !== void 0 ? _customBranding$pageT : 'Elastic';
  const favIcon = (_customBranding$favic = customBranding.faviconSVG) !== null && _customBranding$favic !== void 0 ? _customBranding$favic : `${uiPublicUrl}/favicons/favicon.svg`;
  const favIconPng = (_customBranding$favic2 = customBranding.faviconPNG) !== null && _customBranding$favic2 !== void 0 ? _customBranding$favic2 : `${uiPublicUrl}/favicons/favicon.png`;
  const logo = customBranding.logo ? /*#__PURE__*/_react.default.createElement("img", {
    src: customBranding.logo,
    width: "64",
    height: "64",
    alt: "logo"
  }) : /*#__PURE__*/_react.default.createElement(_logo.Logo, null);
  return /*#__PURE__*/_react.default.createElement("html", {
    lang: locale
  }, /*#__PURE__*/_react.default.createElement("head", null, /*#__PURE__*/_react.default.createElement("meta", {
    charSet: "utf-8"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    httpEquiv: "X-UA-Compatible",
    content: "IE=edge,chrome=1"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "viewport",
    content: "width=device-width"
  }), /*#__PURE__*/_react.default.createElement("title", null, title), /*#__PURE__*/_react.default.createElement(_fonts.Fonts, {
    url: uiPublicUrl
  }), /*#__PURE__*/_react.default.createElement("link", {
    rel: "alternate icon",
    type: "image/png",
    href: favIconPng
  }), /*#__PURE__*/_react.default.createElement("link", {
    rel: "icon",
    type: "image/svg+xml",
    href: favIcon
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "theme-color",
    content: "#ffffff"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "color-scheme",
    content: "light dark"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: _coreBaseCommon.EUI_STYLES_GLOBAL
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "emotion"
  }), /*#__PURE__*/_react.default.createElement(_styles.Styles, {
    darkMode: darkMode,
    stylesheetPaths: stylesheetPaths
  }), scriptPaths.map(path => /*#__PURE__*/_react.default.createElement("script", {
    key: path,
    src: path
  })), /*#__PURE__*/_react.default.createElement("meta", {
    name: "add-styles-here"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: "add-scripts-here"
  }), /*#__PURE__*/_react.default.createElement("meta", {
    name: _coreBaseCommon.EUI_STYLES_UTILS
  })), /*#__PURE__*/_react.default.createElement("body", null, /*#__PURE__*/(0, _react.createElement)('kbn-csp', {
    data: JSON.stringify({
      strictCsp
    })
  }), /*#__PURE__*/(0, _react.createElement)('kbn-injected-metadata', {
    data: JSON.stringify(injectedMetadata)
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnWelcomeView",
    id: "kbn_loading_message",
    style: {
      display: 'none'
    },
    "data-test-subj": "kbnLoadingMessage"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnLoaderWrap"
  }, logo, /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnWelcomeText",
    "data-error-message-title": i18n('core.ui.welcomeErrorMessageTitle', {
      defaultMessage: 'Elastic did not load properly'
    }),
    "data-error-message-text": i18n('core.ui.welcomeErrorMessageText', {
      defaultMessage: 'Please reload this page. If the issue persists, check the browser console and server logs.'
    }),
    "data-error-message-reload": i18n('core.ui.welcomeErrorReloadButton', {
      defaultMessage: 'Reload'
    })
  }, i18n('core.ui.welcomeMessage', {
    defaultMessage: 'Loading Elastic'
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnProgress"
  }))), /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnWelcomeView",
    id: "kbn_legacy_browser_error",
    style: {
      display: 'none'
    }
  }, logo, /*#__PURE__*/_react.default.createElement("h2", {
    className: "kbnWelcomeTitle"
  }, i18n('core.ui.legacyBrowserTitle', {
    defaultMessage: 'Please upgrade your browser'
  })), /*#__PURE__*/_react.default.createElement("div", {
    className: "kbnWelcomeText"
  }, i18n('core.ui.legacyBrowserMessage', {
    defaultMessage: 'This Elastic installation has strict security requirements enabled that your current browser does not meet.'
  }))), /*#__PURE__*/_react.default.createElement("script", null, `// kbnUnsafeInlineTest do not remove this comment:
            //   it is used for filtering out expected CSP failures, and must be the first piece of content in this script block.
            // Since this is an unsafe inline script, this code will not run
            // in browsers that support content security policy(CSP). This is
            // intentional as we check for the existence of __kbnCspNotEnforced__ in
            // bootstrap.
            window.__kbnCspNotEnforced__ = true;
          `), /*#__PURE__*/_react.default.createElement("script", {
    src: bootstrapScriptUrl
  })));
};
exports.Template = Template;
