// @kbn/monitoring-plugin/public/components/beats/overview/latest_versions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LatestVersions = LatestVersions;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function LatestVersions({
  latestVersions
}) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiBasicTable, {
    items: latestVersions,
    columns: [{
      field: 'version',
      name: ''
    }, {
      field: 'count',
      dataType: 'number',
      name: ''
    }]
  });
}
LatestVersions.propTypes = {
  latestVersions: _propTypes.default.arrayOf(_propTypes.default.shape({
    version: _propTypes.default.string.isRequired,
    count: _propTypes.default.number.isRequired
  })).isRequired
};
