// @kbn/rule-registry-plugin/server/routes/update_alert_by_id.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateAlertByIdRoute = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _securitysolutionEsUtils = require("@kbn/securitysolution-es-utils");
var _route_validation = require("./utils/route_validation");
var _constants = require("../../common/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const updateAlertByIdRoute = router => {
  router.post({
    path: _constants.BASE_RAC_ALERTS_API_PATH,
    validate: {
      body: (0, _route_validation.buildRouteValidation)(t.intersection([t.exact(t.type({
        status: t.string,
        ids: t.array(t.string),
        index: t.string
      })), t.exact(t.partial({
        _version: t.string
      }))]))
    },
    options: {
      tags: ['access:rac']
    }
  }, async (context, req, response) => {
    try {
      const racContext = await context.rac;
      const alertsClient = await racContext.getAlertsClient();
      const {
        status,
        ids,
        index,
        _version
      } = req.body;
      const updatedAlert = await alertsClient.update({
        id: ids[0],
        status,
        _version,
        index
      });
      if (updatedAlert == null) {
        return response.notFound({
          body: {
            message: `alerts with ids ${ids} and index ${index} not found`
          }
        });
      }
      return response.ok({
        body: {
          success: true,
          ...updatedAlert
        }
      });
    } catch (exc) {
      const err = (0, _securitysolutionEsUtils.transformError)(exc);
      const contentType = {
        'content-type': 'application/json'
      };
      const defaultedHeaders = {
        ...contentType
      };
      return response.customError({
        headers: defaultedHeaders,
        statusCode: err.statusCode,
        body: {
          message: err.message,
          attributes: {
            success: false
          }
        }
      });
    }
  });
};
exports.updateAlertByIdRoute = updateAlertByIdRoute;
