// @kbn/triggers-actions-ui-plugin/public/application/internal/shareable_components_sandbox/rules_list_sandbox.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RulesListSandbox = void 0;
var _react = _interopRequireDefault(require("react"));
var _get_rules_list = require("../../../common/get_rules_list");
var _use_connector_context = require("../../context/use_connector_context");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const style = {
  flex: 1
};
const RulesListSandbox = () => {
  const {
    services: {
      validateEmailAddresses
    }
  } = (0, _use_connector_context.useConnectorContext)();
  return /*#__PURE__*/_react.default.createElement("div", {
    style: style
  }, (0, _get_rules_list.getRulesListLazy)({
    connectorServices: {
      validateEmailAddresses
    },
    rulesListProps: {}
  }));
};
exports.RulesListSandbox = RulesListSandbox;
