// @kbn/security-solution-plugin/public/exceptions/components/manage_rules/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ManageRules = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _add_to_rules_table = require("../../../detection_engine/rule_exceptions/components/flyout_components/add_to_rules_table");
var i18n = _interopRequireWildcard(require("../../translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ManageRules = exports.ManageRules = /*#__PURE__*/(0, _react.memo)(({
  linkedRules,
  showButtonLoader,
  saveIsDisabled = true,
  onSave,
  onCancel,
  onRuleSelectionChange
}) => {
  const complicatedFlyoutTitleId = (0, _eui.useGeneratedHtmlId)({
    prefix: 'complicatedFlyoutTitle'
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyout, {
    hideCloseButton: true,
    ownFocus: true,
    onClose: onCancel,
    "aria-labelledby": complicatedFlyoutTitleId
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "m"
  }, /*#__PURE__*/_react.default.createElement("h2", {
    id: complicatedFlyoutTitleId
  }, i18n.LINK_RULES_HEADER)), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s",
    color: "subdued"
  }, i18n.MANAGE_RULES_DESCRIPTION)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_add_to_rules_table.ExceptionsAddToRulesTable, {
    initiallySelectedRules: linkedRules,
    onRuleSelectionChange: onRuleSelectionChange
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutFooter, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    onClick: onCancel,
    flush: "left"
  }, i18n.MANAGE_RULES_CANCEL)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "manageListRulesSaveButton",
    isLoading: showButtonLoader,
    disabled: saveIsDisabled,
    onClick: onSave,
    fill: true
  }, i18n.MANAGE_RULES_SAVE)))));
});
ManageRules.displayName = 'ManageRules';
