// @kbn/infra-plugin/public/components/logging/log_analysis_job_status/recreate_job_callout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RecreateJobCallout = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _create_job_button = require("../log_analysis_setup/create_job_button");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RecreateJobCallout = ({
  children,
  hasSetupCapabilities,
  onRecreateMlJob,
  title
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
  color: "warning",
  iconType: "warning",
  title: title
}, children, /*#__PURE__*/_react.default.createElement(_create_job_button.RecreateJobButton, {
  color: "warning",
  hasSetupCapabilities: hasSetupCapabilities,
  onClick: onRecreateMlJob
}));
exports.RecreateJobCallout = RecreateJobCallout;
