// @kbn/security-solution-plugin/public/common/components/event_details/cti_details/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.timelineDataToEnrichment = exports.parseExistingEnrichments = exports.isInvestigationTimeEnrichment = exports.getShimmedIndicatorValue = exports.getFirstSeen = exports.getEnrichmentValue = exports.getEnrichmentIdentifiers = exports.getEnrichmentFields = exports.filterDuplicateEnrichments = exports.buildThreatDetailsItems = void 0;
var _lodash = require("lodash");
var _field_formatters = require("../../../../../common/utils/field_formatters");
var _constants = require("../../../../../common/constants");
var _constants2 = require("../../../../../common/cti/constants");
var _cti = require("../../../../../common/search_strategy/security_solution/cti");
var _data_retrieval = require("../../../../../common/utils/data_retrieval");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const isInvestigationTimeEnrichment = type => type === _constants2.ENRICHMENT_TYPES.InvestigationTime;
exports.isInvestigationTimeEnrichment = isInvestigationTimeEnrichment;
const parseExistingEnrichments = data => {
  const threatIndicatorField = data.find(({
    field,
    originalValue
  }) => field === _constants.ENRICHMENT_DESTINATION_PATH && originalValue);
  if (!threatIndicatorField) {
    return [];
  }
  const {
    originalValue
  } = threatIndicatorField;
  const enrichmentStrings = Array.isArray(originalValue) ? originalValue : [originalValue];
  return enrichmentStrings.reduce((enrichments, enrichmentString) => {
    try {
      const enrichment = (0, _field_formatters.getDataFromFieldsHits)(JSON.parse(enrichmentString));
      enrichments.push(enrichment);
    } catch (e) {
      // omit failed parse
    }
    return enrichments;
  }, []);
};
exports.parseExistingEnrichments = parseExistingEnrichments;
const timelineDataToEnrichment = data => data.reduce((acc, item) => {
  acc[item.field] = item.originalValue;
  return acc;
}, {});
exports.timelineDataToEnrichment = timelineDataToEnrichment;
const getEnrichmentValue = (enrichment, field) => (0, _data_retrieval.getFirstElement)(enrichment[field]);

/**
 * These fields (e.g. 'indicator.ip') may be in one of three places depending on whether it's:
 *   * a queried, legacy filebeat indicator ('threatintel.indicator.ip')
 *   * a queried, ECS 1.11 filebeat indicator ('threat.indicator.ip')
 *   * an existing indicator from an enriched alert ('indicator.ip')
 */
exports.getEnrichmentValue = getEnrichmentValue;
const getShimmedIndicatorValue = (enrichment, field) => getEnrichmentValue(enrichment, field) || getEnrichmentValue(enrichment, `threatintel.${field}`) || getEnrichmentValue(enrichment, `threat.${field}`);
exports.getShimmedIndicatorValue = getShimmedIndicatorValue;
const getEnrichmentIdentifiers = enrichment => ({
  id: getEnrichmentValue(enrichment, _constants2.MATCHED_ID),
  field: getEnrichmentValue(enrichment, _constants2.MATCHED_FIELD),
  value: getEnrichmentValue(enrichment, _constants2.MATCHED_ATOMIC),
  type: getEnrichmentValue(enrichment, _constants2.MATCHED_TYPE),
  feedName: getShimmedIndicatorValue(enrichment, _constants2.FEED_NAME)
});
exports.getEnrichmentIdentifiers = getEnrichmentIdentifiers;
const buildEnrichmentId = enrichment => {
  const {
    id,
    field
  } = getEnrichmentIdentifiers(enrichment);
  return `${id}${field}`;
};

/**
 * This function receives an array of enrichments and removes
 * investigation-time enrichments if that exact indicator already exists
 * elsewhere in the list.
 *
 * @param enrichments {@type CtiEnrichment[]}
 */
const filterDuplicateEnrichments = enrichments => {
  if (enrichments.length < 2) {
    return enrichments;
  }
  const enrichmentsById = (0, _lodash.groupBy)(enrichments, buildEnrichmentId);
  return Object.values(enrichmentsById).map(enrichmentGroup => {
    var _enrichmentGroup$find;
    return (_enrichmentGroup$find = enrichmentGroup.find(enrichment => !isInvestigationTimeEnrichment(getEnrichmentValue(enrichment, _constants2.MATCHED_TYPE)))) !== null && _enrichmentGroup$find !== void 0 ? _enrichmentGroup$find : enrichmentGroup[0];
  });
};
exports.filterDuplicateEnrichments = filterDuplicateEnrichments;
const getEnrichmentFields = items => items.reduce((fields, item) => {
  if ((0, _cti.isValidEventField)(item.field)) {
    const value = (0, _data_retrieval.getFirstElement)(item.originalValue);
    if (value) {
      return {
        ...fields,
        [item.field]: value
      };
    }
  }
  return fields;
}, {});
exports.getEnrichmentFields = getEnrichmentFields;
const getFirstSeen = enrichment => {
  const firstSeenValue = getShimmedIndicatorValue(enrichment, _constants2.FIRST_SEEN);
  const firstSeenDate = Date.parse(firstSeenValue !== null && firstSeenValue !== void 0 ? firstSeenValue : 'no date');
  return Number.isInteger(firstSeenDate) ? firstSeenDate : new Date(-1).valueOf();
};
exports.getFirstSeen = getFirstSeen;
const buildThreatDetailsItems = enrichment => Object.keys(enrichment).sort().map(field => {
  const title = field.startsWith(_constants.DEFAULT_INDICATOR_SOURCE_PATH) ? field.replace(`${_constants.DEFAULT_INDICATOR_SOURCE_PATH}`, 'indicator') : field;
  let value = (0, _data_retrieval.getFirstElement)(enrichment[field]);
  if ((0, _lodash.isObject)(value)) {
    value = i18n.NESTED_OBJECT_VALUES_NOT_RENDERED;
  }
  return {
    title,
    description: {
      fieldName: field,
      value
    }
  };
});
exports.buildThreatDetailsItems = buildThreatDetailsItems;
