// @kbn/security-solution-plugin/public/explore/network/containers/network_dns/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useNetworkDns = exports.ID = void 0;
var _react = require("react");
var _fastDeepEqual = _interopRequireDefault(require("fast-deep-equal"));
var _use_selector = require("../../../../common/hooks/use_selector");
var _helpers = require("../../../../common/containers/helpers");
var _helpers2 = require("../../../components/paginated_table/helpers");
var _store = require("../../store");
var _search_strategy = require("../../../../../common/search_strategy");
var i18n = _interopRequireWildcard(require("./translations"));
var _use_search_strategy = require("../../../../common/containers/use_search_strategy");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ID = exports.ID = 'networkDnsQuery';
const useNetworkDns = ({
  endDate,
  filterQuery,
  id,
  indexNames,
  skip,
  startDate
}) => {
  const getNetworkDnsSelector = (0, _react.useMemo)(() => _store.networkSelectors.dnsSelector(), []);
  const {
    activePage,
    sort,
    isPtrIncluded,
    limit
  } = (0, _use_selector.useDeepEqualSelector)(getNetworkDnsSelector);
  const [networkDnsRequest, setNetworkDnsRequest] = (0, _react.useState)(null);
  const wrappedLoadMore = (0, _react.useCallback)(newActivePage => {
    setNetworkDnsRequest(prevRequest => {
      if (!prevRequest) {
        return prevRequest;
      }
      return {
        ...prevRequest,
        pagination: (0, _helpers2.generateTablePaginationOptions)(newActivePage, limit)
      };
    });
  }, [limit]);
  const {
    loading,
    result: response,
    search,
    refetch,
    inspect
  } = (0, _use_search_strategy.useSearchStrategy)({
    factoryQueryType: _search_strategy.NetworkQueries.dns,
    initialResult: {
      edges: [],
      totalCount: -1,
      pageInfo: {
        activePage: 0,
        fakeTotalCount: 0,
        showMorePagesIndicator: false
      }
    },
    errorMessage: i18n.ERROR_NETWORK_DNS,
    abort: skip
  });
  const networkDnsResponse = (0, _react.useMemo)(() => ({
    id,
    inspect,
    isInspected: false,
    loadPage: wrappedLoadMore,
    networkDns: response.edges,
    pageInfo: response.pageInfo,
    refetch,
    totalCount: response.totalCount
  }), [id, inspect, refetch, response.edges, response.pageInfo, response.totalCount, wrappedLoadMore]);
  (0, _react.useEffect)(() => {
    setNetworkDnsRequest(prevRequest => {
      const myRequest = {
        ...(prevRequest !== null && prevRequest !== void 0 ? prevRequest : {}),
        defaultIndex: indexNames,
        isPtrIncluded,
        factoryQueryType: _search_strategy.NetworkQueries.dns,
        filterQuery: (0, _helpers.createFilter)(filterQuery),
        pagination: (0, _helpers2.generateTablePaginationOptions)(activePage, limit, true),
        sort,
        timerange: {
          interval: '12h',
          from: startDate,
          to: endDate
        }
      };
      if (!(0, _fastDeepEqual.default)(prevRequest, myRequest)) {
        return myRequest;
      }
      return prevRequest;
    });
  }, [activePage, indexNames, endDate, filterQuery, limit, startDate, sort, isPtrIncluded]);
  (0, _react.useEffect)(() => {
    if (!skip && networkDnsRequest) {
      search(networkDnsRequest);
    }
  }, [networkDnsRequest, search, skip]);
  return [loading, networkDnsResponse];
};
exports.useNetworkDns = useNetworkDns;
