// @kbn/security-solution-plugin/public/management/components/endpoint_execute_action/execute_action_host_response.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExecuteActionHostResponse = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _translations = require("../endpoint_response_actions_list/translations");
var _response_action_file_download_link = require("../response_action_file_download_link");
var _execute_action_host_response_output = require("./execute_action_host_response_output");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ExecuteActionHostResponse = exports.ExecuteActionHostResponse = /*#__PURE__*/(0, _react.memo)(({
  action,
  agentId = action.agents[0],
  canAccessFileDownloadLink,
  textSize = 's',
  'data-test-subj': dataTestSubj
}) => {
  const outputContent = (0, _react.useMemo)(() => action.outputs && action.outputs[agentId] && action.outputs[agentId].content, [action.outputs, agentId]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_response_action_file_download_link.ResponseActionFileDownloadLink, {
    action: action,
    buttonTitle: _translations.EXECUTE_FILE_LINK_TITLE,
    canAccessFileDownloadLink: canAccessFileDownloadLink,
    "data-test-subj": `${dataTestSubj}-getExecuteLink`,
    textSize: textSize
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xxl"
  })), outputContent && /*#__PURE__*/_react.default.createElement(_execute_action_host_response_output.ExecuteActionHostResponseOutput, {
    outputContent: outputContent,
    "data-test-subj": `${dataTestSubj}-executeResponseOutput`,
    textSize: textSize
  }));
});
ExecuteActionHostResponse.displayName = 'ExecuteActionHostResponse';
