// @kbn/vis-type-table-plugin/public/components/table_vis_split.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableVisSplit = void 0;
var _react = _interopRequireWildcard(require("react"));
var _uiTheme = require("@kbn/ui-theme");
var _table_vis_basic = require("./table_vis_basic");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const TableVisSplit = exports.TableVisSplit = /*#__PURE__*/(0, _react.memo)(({
  fireEvent,
  tables,
  visConfig,
  uiStateProps,
  enforceMinWidth
}) => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, tables.map(({
    table,
    title
  }) => {
    // reserve minimum size per table
    const minTableWidth = table.columns.reduce((sum, column, index) => {
      var _uiStateProps$columns, _uiStateProps$columns2;
      return sum + ((_uiStateProps$columns = (_uiStateProps$columns2 = uiStateProps.columnsWidth.find(width => width.colIndex === index)) === null || _uiStateProps$columns2 === void 0 ? void 0 : _uiStateProps$columns2.width) !== null && _uiStateProps$columns !== void 0 ? _uiStateProps$columns : 25);
    }, 0);
    return /*#__PURE__*/_react.default.createElement("div", {
      key: title,
      className: "tbvChart__split",
      css: enforceMinWidth ? {
        minWidth: `calc(${minTableWidth}px + 2 * ${_uiTheme.euiThemeVars.euiSizeS})`
      } : {}
    }, /*#__PURE__*/_react.default.createElement(_table_vis_basic.TableVisBasic, {
      fireEvent: fireEvent,
      table: table,
      visConfig: visConfig,
      title: title,
      uiStateProps: uiStateProps
    }));
  }));
});
