// @kbn/coloring/src/shared_components/color_mapping/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  CategoricalColorMapping: true,
  SPECIAL_TOKENS_STRING_CONVERTION: true,
  DEFAULT_COLOR_MAPPING_CONFIG: true,
  DEFAULT_OTHER_ASSIGNMENT_INDEX: true,
  getPaletteColors: true,
  getColorsFromMapping: true
};
Object.defineProperty(exports, "CategoricalColorMapping", {
  enumerable: true,
  get: function () {
    return _categorical_color_mapping.CategoricalColorMapping;
  }
});
Object.defineProperty(exports, "DEFAULT_COLOR_MAPPING_CONFIG", {
  enumerable: true,
  get: function () {
    return _default_color_mapping.DEFAULT_COLOR_MAPPING_CONFIG;
  }
});
Object.defineProperty(exports, "DEFAULT_OTHER_ASSIGNMENT_INDEX", {
  enumerable: true,
  get: function () {
    return _default_color_mapping.DEFAULT_OTHER_ASSIGNMENT_INDEX;
  }
});
Object.defineProperty(exports, "SPECIAL_TOKENS_STRING_CONVERTION", {
  enumerable: true,
  get: function () {
    return _rule_matching.SPECIAL_TOKENS_STRING_CONVERTION;
  }
});
Object.defineProperty(exports, "getColorsFromMapping", {
  enumerable: true,
  get: function () {
    return _default_color_mapping.getColorsFromMapping;
  }
});
Object.defineProperty(exports, "getPaletteColors", {
  enumerable: true,
  get: function () {
    return _default_color_mapping.getPaletteColors;
  }
});
var _categorical_color_mapping = require("./categorical_color_mapping");
var _palettes = require("./palettes");
Object.keys(_palettes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _palettes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _palettes[key];
    }
  });
});
var _color_handling = require("./color/color_handling");
Object.keys(_color_handling).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _color_handling[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _color_handling[key];
    }
  });
});
var _rule_matching = require("./color/rule_matching");
var _default_color_mapping = require("./config/default_color_mapping");
