// @kbn/security-solution-plugin/public/flyout/document_details/right/tabs/table_tab.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getColumns = exports.TableTab = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _columns = require("../../../../common/components/event_details/columns");
var _field_value_cell = require("../../../../common/components/event_details/table/field_value_cell");
var _field_name_cell = require("../../../../common/components/event_details/table/field_name_cell");
var _cell_actions = require("../components/cell_actions");
var i18n = _interopRequireWildcard(require("../../../../common/components/event_details/translations"));
var _context = require("../../shared/context");
var _event_fields_browser = require("../../../../common/components/event_details/event_fields_browser");
var _types = require("../../../../../common/types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getColumns = ({
  browserFields,
  eventId,
  contextId,
  scopeId,
  getLinkValue,
  isDraggable
}) => [{
  field: 'field',
  name: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("strong", null, i18n.FIELD)),
  width: '30%',
  render: (field, data) => {
    return /*#__PURE__*/_react.default.createElement(_field_name_cell.FieldNameCell, {
      data: data,
      field: field,
      fieldMapping: undefined
    });
  }
}, {
  field: 'values',
  name: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("strong", null, i18n.VALUE)),
  width: '70%',
  render: (values, data) => {
    const fieldFromBrowserField = (0, _columns.getFieldFromBrowserField)(data.field, browserFields);
    return /*#__PURE__*/_react.default.createElement(_cell_actions.CellActions, {
      field: data.field,
      value: values,
      isObjectArray: data.isObjectArray
    }, /*#__PURE__*/_react.default.createElement(_field_value_cell.FieldValueCell, {
      contextId: contextId,
      data: data,
      eventId: eventId,
      fieldFromBrowserField: fieldFromBrowserField,
      getLinkValue: getLinkValue,
      isDraggable: isDraggable,
      values: values
    }));
  }
}];

/**
 * Table view displayed in the document details expandable flyout right section
 */
exports.getColumns = getColumns;
const TableTab = exports.TableTab = /*#__PURE__*/(0, _react.memo)(() => {
  const {
    browserFields,
    dataFormattedForFieldBrowser,
    eventId,
    scopeId
  } = (0, _context.useDocumentDetailsContext)();
  return /*#__PURE__*/_react.default.createElement(_event_fields_browser.EventFieldsBrowser, {
    browserFields: browserFields,
    data: dataFormattedForFieldBrowser,
    eventId: eventId,
    isDraggable: false,
    timelineTabType: _types.TimelineTabs.query,
    scopeId: scopeId,
    isReadOnly: false,
    columnsProvider: getColumns
  });
});
TableTab.displayName = 'TableTab';
