// @kbn/core-saved-objects-server-internal/src/routes/update.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.registerUpdateRoute = void 0;
var _configSchema = require("@kbn/config-schema");
var _utils = require("./utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const registerUpdateRoute = (router, {
  config,
  coreUsageData,
  logger,
  access
}) => {
  const {
    allowHttpApiAccess
  } = config;
  router.put({
    path: '/{type}/{id}',
    options: {
      access,
      summary: `Update a saved object`,
      deprecated: true
    },
    validate: {
      params: _configSchema.schema.object({
        type: _configSchema.schema.string(),
        id: _configSchema.schema.string()
      }),
      body: _configSchema.schema.object({
        attributes: _configSchema.schema.recordOf(_configSchema.schema.string(), _configSchema.schema.any()),
        version: _configSchema.schema.maybe(_configSchema.schema.string()),
        references: _configSchema.schema.maybe(_configSchema.schema.arrayOf(_configSchema.schema.object({
          name: _configSchema.schema.string(),
          type: _configSchema.schema.string(),
          id: _configSchema.schema.string()
        }))),
        upsert: _configSchema.schema.maybe(_configSchema.schema.recordOf(_configSchema.schema.string(), _configSchema.schema.any()))
      })
    }
  }, (0, _utils.catchAndReturnBoomErrors)(async (context, request, response) => {
    (0, _utils.logWarnOnExternalRequest)({
      method: 'get',
      path: '/api/saved_objects/{type}/{id}',
      request,
      logger
    });
    const {
      type,
      id
    } = request.params;
    const {
      attributes,
      version,
      references,
      upsert
    } = request.body;
    const options = {
      version,
      references,
      upsert,
      migrationVersionCompatibility: 'raw'
    };
    const usageStatsClient = coreUsageData.getClient();
    usageStatsClient.incrementSavedObjectsUpdate({
      request,
      types: [type]
    }).catch(() => {});
    const {
      savedObjects
    } = await context.core;
    if (!allowHttpApiAccess) {
      (0, _utils.throwIfTypeNotVisibleByAPI)(type, savedObjects.typeRegistry);
    }
    const result = await savedObjects.client.update(type, id, attributes, options);
    return response.ok({
      body: result
    });
  }));
};
exports.registerUpdateRoute = registerUpdateRoute;
