// @kbn/data-visualizer-plugin/public/application/index_data_visualizer/search_strategy/requests/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SAMPLER_TOP_TERMS_THRESHOLD = exports.SAMPLER_TOP_TERMS_SHARD_SIZE = exports.PERCENTILE_SPACING = exports.MAX_PERCENT = exports.MAX_EXAMPLES_DEFAULT = exports.MAX_CHART_COLUMNS = exports.FIELDS_REQUEST_BATCH_SIZE = exports.AGGREGATABLE_EXISTS_REQUEST_BATCH_SIZE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SAMPLER_TOP_TERMS_THRESHOLD = exports.SAMPLER_TOP_TERMS_THRESHOLD = 100000;
const SAMPLER_TOP_TERMS_SHARD_SIZE = exports.SAMPLER_TOP_TERMS_SHARD_SIZE = 5000;
const AGGREGATABLE_EXISTS_REQUEST_BATCH_SIZE = exports.AGGREGATABLE_EXISTS_REQUEST_BATCH_SIZE = 200;
const FIELDS_REQUEST_BATCH_SIZE = exports.FIELDS_REQUEST_BATCH_SIZE = 10;
const MAX_CHART_COLUMNS = exports.MAX_CHART_COLUMNS = 20;
const MAX_EXAMPLES_DEFAULT = exports.MAX_EXAMPLES_DEFAULT = 10;
const MAX_PERCENT = exports.MAX_PERCENT = 100;
const PERCENTILE_SPACING = exports.PERCENTILE_SPACING = 5;
