// @kbn/observability-ai-assistant-plugin/public/components/insight/insight_error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InsightError = InsightError;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function InsightError() {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: _i18n.i18n.translate('xpack.observabilityAiAssistant.insight.error.title', {
      defaultMessage: 'Error'
    }),
    color: "danger",
    iconType: "error"
  }, _i18n.i18n.translate('xpack.observabilityAiAssistant.insight.error.description', {
    defaultMessage: 'An error occured.'
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "observabilityAiAssistantInsightErrorRegenerateButton",
    fill: true,
    color: "danger"
  }, _i18n.i18n.translate('xpack.observabilityAiAssistant.insight.error.buttonLabel', {
    defaultMessage: 'Regenerate'
  })));
}
