// @kbn/search-notebooks/server/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  config: true
};
exports.config = void 0;
var _configSchema = require("@kbn/config-schema");
var _types = require("./types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const configSchema = _configSchema.schema.object({
  enabled: _configSchema.schema.boolean({
    defaultValue: true
  }),
  catalog: _configSchema.schema.maybe(_configSchema.schema.object({
    url: _configSchema.schema.uri(),
    ttl: _configSchema.schema.number({
      defaultValue: 900
    }),
    errorTTL: _configSchema.schema.number({
      defaultValue: 3600
    })
  }))
});
const config = exports.config = {
  schema: configSchema
};
