// @kbn/search-notebooks/server/plugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchNotebooksPlugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _routes = require("./routes");
var _utils = require("./utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class SearchNotebooksPlugin {
  constructor(initializerContext) {
    (0, _defineProperty2.default)(this, "config", void 0);
    (0, _defineProperty2.default)(this, "logger", void 0);
    (0, _defineProperty2.default)(this, "notebooksCache", void 0);
    this.config = initializerContext.config.get();
    this.logger = initializerContext.logger.get();
    this.notebooksCache = (0, _utils.createNotebooksCache)();
  }
  setup(core) {
    if (!this.config.enabled) return {};
    this.logger.debug('searchNotebooks: Setup');
    const router = core.http.createRouter();

    // Register server side APIs
    (0, _routes.defineRoutes)({
      config: this.config,
      logger: this.logger,
      notebooksCache: this.notebooksCache,
      router
    });
    return {};
  }
  start(core) {
    return {};
  }
  stop() {}
}
exports.SearchNotebooksPlugin = SearchNotebooksPlugin;
