// @kbn/vis-default-editor-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BasicOptions", {
  enumerable: true,
  get: function () {
    return _options.BasicOptions;
  }
});
Object.defineProperty(exports, "ColorRanges", {
  enumerable: true,
  get: function () {
    return _options.ColorRanges;
  }
});
Object.defineProperty(exports, "ColorSchemaOptions", {
  enumerable: true,
  get: function () {
    return _options.ColorSchemaOptions;
  }
});
Object.defineProperty(exports, "DefaultEditorController", {
  enumerable: true,
  get: function () {
    return _default_editor_controller.DefaultEditorController;
  }
});
Object.defineProperty(exports, "DefaultEditorSize", {
  enumerable: true,
  get: function () {
    return _editor_size.DefaultEditorSize;
  }
});
Object.defineProperty(exports, "LegendSizeSettings", {
  enumerable: true,
  get: function () {
    return _options.LegendSizeSettings;
  }
});
Object.defineProperty(exports, "LongLegendOptions", {
  enumerable: true,
  get: function () {
    return _options.LongLegendOptions;
  }
});
Object.defineProperty(exports, "NumberInputOption", {
  enumerable: true,
  get: function () {
    return _options.NumberInputOption;
  }
});
Object.defineProperty(exports, "PalettePicker", {
  enumerable: true,
  get: function () {
    return _palette_picker.PalettePicker;
  }
});
Object.defineProperty(exports, "PercentageModeOption", {
  enumerable: true,
  get: function () {
    return _options.PercentageModeOption;
  }
});
Object.defineProperty(exports, "RangeOption", {
  enumerable: true,
  get: function () {
    return _options.RangeOption;
  }
});
Object.defineProperty(exports, "RangesParamEditor", {
  enumerable: true,
  get: function () {
    return _ranges.RangesParamEditor;
  }
});
Object.defineProperty(exports, "RequiredNumberInputOption", {
  enumerable: true,
  get: function () {
    return _options.RequiredNumberInputOption;
  }
});
Object.defineProperty(exports, "SelectOption", {
  enumerable: true,
  get: function () {
    return _options.SelectOption;
  }
});
Object.defineProperty(exports, "SwitchOption", {
  enumerable: true,
  get: function () {
    return _options.SwitchOption;
  }
});
Object.defineProperty(exports, "TextInputOption", {
  enumerable: true,
  get: function () {
    return _options.TextInputOption;
  }
});
Object.defineProperty(exports, "groupAndSortBy", {
  enumerable: true,
  get: function () {
    return _utils2.groupAndSortBy;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "useValidation", {
  enumerable: true,
  get: function () {
    return _utils.useValidation;
  }
});
var _default_editor_controller = require("./default_editor_controller");
var _plugin = require("./plugin");
var _utils = require("./components/controls/utils");
var _palette_picker = require("./components/controls/palette_picker");
var _options = require("./components/options");
var _ranges = require("./components/controls/ranges");
var _utils2 = require("./utils");
var _editor_size = require("./editor_size");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const plugin = context => {
  return new _plugin.VisDefaultEditorPlugin();
};
exports.plugin = plugin;
