// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/hooks/use_setup_sections.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSetUpSections = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _css = require("@emotion/css");
var _card_item = require("../card_item");
var _sections = require("../sections");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useSetUpSections = ({
  euiTheme
}) => {
  const setUpCards = (0, _react.useCallback)(({
    activeSections,
    expandedCardSteps,
    finishedSteps,
    toggleTaskCompleteStatus,
    onStepClicked,
    sectionId
  }) => {
    var _Object$values;
    const section = activeSections === null || activeSections === void 0 ? void 0 : activeSections[sectionId];
    return section ? (_Object$values = Object.values(section)) === null || _Object$values === void 0 ? void 0 : _Object$values.map(cardItem => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      key: cardItem.id
    }, /*#__PURE__*/_react.default.createElement(_card_item.CardItem, {
      activeStepIds: cardItem.activeStepIds,
      cardId: cardItem.id,
      "data-test-subj": cardItem.id,
      expandedCardSteps: expandedCardSteps,
      finishedSteps: finishedSteps[cardItem.id],
      toggleTaskCompleteStatus: toggleTaskCompleteStatus,
      onStepClicked: onStepClicked,
      sectionId: sectionId
    }))) : null;
  }, []);
  const setUpSections = (0, _react.useCallback)(({
    activeSections,
    expandedCardSteps,
    finishedSteps,
    toggleTaskCompleteStatus,
    onStepClicked
  }) => (0, _sections.getSections)().reduce((acc, currentSection) => {
    const cardNodes = setUpCards({
      activeSections,
      expandedCardSteps,
      finishedSteps,
      toggleTaskCompleteStatus,
      onStepClicked,
      sectionId: currentSection.id
    });
    if (cardNodes && cardNodes.length > 0) {
      acc.push( /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
        color: "plain",
        element: "div",
        grow: false,
        paddingSize: "none",
        hasShadow: false,
        borderRadius: "none",
        css: (0, _css.css)`
                margin: ${euiTheme.size.l} 0;
                padding-top: 4px;
                background-color: ${euiTheme.colors.lightestShade};
              `,
        key: currentSection.id,
        id: currentSection.id,
        "data-test-subj": `section-${currentSection.id}`
      }, /*#__PURE__*/_react.default.createElement("h2", {
        css: (0, _css.css)`
                  font-size: ${euiTheme.base * 1.375}px;
                  font-weight: ${euiTheme.font.weight.bold};
                `
      }, currentSection.title), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
        size: "l"
      }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
        gutterSize: "none",
        direction: "column",
        css: (0, _css.css)`
                  gap: ${euiTheme.size.base};
                `
      }, cardNodes)));
    }
    return acc;
  }, []), [euiTheme.base, euiTheme.colors.lightestShade, euiTheme.font.weight.bold, euiTheme.size.base, euiTheme.size.l, setUpCards]);
  return {
    setUpSections
  };
};
exports.useSetUpSections = useSetUpSections;
