// @kbn/security-solution-plugin/public/common/components/events_viewer/default_event_headers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.defaultEventHeaders = void 0;
var _default_headers = require("../../../timelines/components/timeline/body/column_headers/default_headers");
var _constants = require("../../../timelines/components/timeline/body/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const defaultEventHeaders = exports.defaultEventHeaders = [{
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: '@timestamp',
  initialWidth: _constants.DEFAULT_DATE_COLUMN_MIN_WIDTH,
  esTypes: ['date'],
  type: 'date'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'message'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'host.name'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'event.module'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'agent.type'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'event.dataset'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'event.action'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'user.name'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'source.ip'
}, {
  columnHeaderType: _default_headers.defaultColumnHeaderType,
  id: 'destination.ip'
}];
