// @kbn/ux-plugin/public/components/app/rum_dashboard/impactful_metrics/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ImpactfulMetrics = ImpactfulMetrics;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _js_errors = require("./js_errors");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function ImpactfulMetrics() {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    wrap: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    style: {
      flexBasis: 650
    }
  }, /*#__PURE__*/_react.default.createElement(_js_errors.JSErrors, null))));
}
