// @kbn/uptime-plugin/public/legacy_uptime/components/monitor/ping_list/headers.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PingHeaders = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PingHeaders = ({
  headers
}) => {
  const headersList = Object.keys(headers).sort().map(header => ({
    title: header,
    description: headers[header]
  }));
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
    id: "responseHeaderAccord",
    buttonContent: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "s"
    }, /*#__PURE__*/_react.default.createElement("h3", null, _i18n.i18n.translate('xpack.uptime.pingList.headers.title', {
      defaultMessage: 'Response headers'
    })))
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    titleProps: {
      style: {
        width: '30%',
        paddingLeft: 30
      }
    },
    compressed: true,
    type: "responsiveColumn",
    listItems: headersList
  })));
};
exports.PingHeaders = PingHeaders;
