// @kbn/unified-search-plugin/public/filters_builder/filter_item/actions/action_strings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.strings = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const strings = exports.strings = {
  getDeleteFilterGroupButtonIconLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.deleteFilterGroupButtonIcon', {
    defaultMessage: 'Delete filter group'
  }),
  getAddOrFilterGroupButtonIconLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.addOrFilterGroupButtonIcon', {
    defaultMessage: 'Add filter group with OR'
  }),
  getAddOrFilterGroupButtonLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.addOrFilterGroupButtonLabel', {
    defaultMessage: 'OR'
  }),
  getAddAndFilterGroupButtonIconLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.addAndFilterGroupButtonIcon', {
    defaultMessage: 'Add filter group with AND'
  }),
  getAddAndFilterGroupButtonLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.addAndFilterGroupButtonLabel', {
    defaultMessage: 'AND'
  }),
  getDeleteButtonDisabled: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.deleteButtonDisabled', {
    defaultMessage: 'A minimum of one item is required.'
  }),
  getMoreActionsLabel: () => _i18n.i18n.translate('unifiedSearch.filter.filtersBuilder.moreActionsLabel', {
    defaultMessage: 'More actions'
  })
};
