// @kbn/security-solution-serverless/public/upselling/sections/cloud_security_posture/cloud_security_posture_integration_pli_block.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CloudSecurityPostureIntegrationPliBlock = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Component displayed when a given product tier is not allowed to use Cloud Security Posture Integrations installation forms.
 */
const CloudSecurityPostureIntegrationPliBlock = exports.CloudSecurityPostureIntegrationPliBlock = /*#__PURE__*/(0, _react.memo)(() => {
  // TODO: prefer to use getProductTypeByPLI(ProductFeatureKey.cloudSecurityPosture) after we change returned text to include "Protection"
  const requiredPLI = 'Cloud Protection Essentials';
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    "data-test-subj": "cloud-security-posture-integration-pli-auth-block",
    isDisabled: true,
    description: false,
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xl",
      type: "lock"
    }),
    betaBadgeProps: {
      label: _i18n.i18n.translate('xpack.securitySolutionServerless.cloudSecurityPostureIntegrationPliBlock.badgeText', {
        defaultMessage: 'Cloud Protection Essentials'
      })
    },
    title: /*#__PURE__*/_react.default.createElement("h3", null, /*#__PURE__*/_react.default.createElement("strong", null, _i18n.i18n.translate('xpack.securitySolutionServerless.cloudSecurityPostureIntegrationPliBlock.cardTitle', {
      defaultMessage: 'Protection updates'
    })))
  }, /*#__PURE__*/_react.default.createElement("div", null, _i18n.i18n.translate('xpack.securitySolutionServerless.cloudSecurityPostureIntegrationPliBlock.cardMessage', {
    defaultMessage: 'To turn on CSPM, KSPM or CNVM, view your Cloud Posture Dashboards and generate findings of misconfiguration or vulnerabilities in your cloud environment, you must add {requiredPLI} under Manage --> Project features.',
    values: {
      requiredPLI
    }
  }))));
});
CloudSecurityPostureIntegrationPliBlock.displayName = 'CloudSecurityPostureIntegrationPliBlock';
