// @kbn/cases-plugin/public/components/custom_fields/text/edit.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Edit = void 0;
var _lodash = require("lodash");
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _hook_form_lib = require("@kbn/es-ui-shared-plugin/static/forms/hook_form_lib");
var _components = require("@kbn/es-ui-shared-plugin/static/forms/components");
var _domain = require("../../../../common/types/domain");
var _view = require("./view");
var _translations = require("../translations");
var _config = require("./config");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FormWrapperComponent = ({
  initialValue,
  customFieldConfiguration,
  isLoading,
  onChange
}) => {
  const {
    form
  } = (0, _hook_form_lib.useForm)({
    defaultValue: {
      value: (customFieldConfiguration === null || customFieldConfiguration === void 0 ? void 0 : customFieldConfiguration.defaultValue) != null && (0, _lodash.isEmpty)(initialValue) ? String(customFieldConfiguration.defaultValue) : initialValue
    }
  });
  const [{
    value
  }] = (0, _hook_form_lib.useFormData)({
    form
  });
  const {
    submit,
    isValid
  } = form;
  const formFieldConfig = (0, _config.getTextFieldConfig)({
    required: customFieldConfiguration.required,
    label: customFieldConfiguration.label
  });
  const populatedWithDefault = value === (customFieldConfiguration === null || customFieldConfiguration === void 0 ? void 0 : customFieldConfiguration.defaultValue) && (0, _lodash.isEmpty)(initialValue);
  (0, _react.useEffect)(() => {
    onChange({
      value,
      isValid,
      submit
    });
  }, [isValid, onChange, submit, value]);
  return /*#__PURE__*/_react.default.createElement(_hook_form_lib.Form, {
    form: form
  }, /*#__PURE__*/_react.default.createElement(_hook_form_lib.UseField, {
    path: "value",
    config: formFieldConfig,
    component: _components.TextField,
    helpText: populatedWithDefault && _translations.POPULATED_WITH_DEFAULT,
    componentProps: {
      euiFieldProps: {
        fullWidth: true,
        disabled: isLoading,
        isLoading,
        'data-test-subj': `case-text-custom-field-form-field-${customFieldConfiguration.key}`
      }
    }
  }));
};
FormWrapperComponent.displayName = 'FormWrapper';
const EditComponent = ({
  customField,
  customFieldConfiguration,
  onSubmit,
  isLoading,
  canUpdate
}) => {
  var _customField$value;
  const initialValue = (_customField$value = customField === null || customField === void 0 ? void 0 : customField.value) !== null && _customField$value !== void 0 ? _customField$value : '';
  const [isEdit, setIsEdit] = (0, _react.useState)(false);
  const [formState, setFormState] = (0, _react.useState)({
    isValid: undefined,
    submit: async () => ({
      isValid: false,
      data: {
        value: ''
      }
    }),
    value: initialValue
  });
  const onEdit = () => {
    setIsEdit(true);
  };
  const onCancel = () => {
    setIsEdit(false);
  };
  const onSubmitCustomField = async () => {
    const {
      isValid,
      data
    } = await formState.submit();
    if (isValid) {
      var _customField$key;
      const value = (0, _lodash.isEmpty)(data.value) ? null : data.value;
      onSubmit({
        ...customField,
        key: (_customField$key = customField === null || customField === void 0 ? void 0 : customField.key) !== null && _customField$key !== void 0 ? _customField$key : customFieldConfiguration.key,
        type: _domain.CustomFieldTypes.TEXT,
        value
      });
    }
    setIsEdit(false);
  };
  const title = customFieldConfiguration.label;
  const isTextFieldValid = formState.isValid || formState.value === customFieldConfiguration.defaultValue && !initialValue;
  const isCustomFieldValueDefined = !(0, _lodash.isEmpty)(customField === null || customField === void 0 ? void 0 : customField.value);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "none",
    justifyContent: "spaceBetween",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("h4", null, title))), isLoading && /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
    "data-test-subj": `case-text-custom-field-loading-${customFieldConfiguration.key}`
  }), !isLoading && canUpdate && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    "data-test-subj": `case-text-custom-field-edit-button-${customFieldConfiguration.key}`,
    "aria-label": (0, _translations.EDIT_CUSTOM_FIELDS_ARIA_LABEL)(title),
    iconType: 'pencil',
    onClick: onEdit
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    margin: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m",
    "data-test-subj": `case-text-custom-field-${customFieldConfiguration.key}`,
    direction: "column"
  }, !isCustomFieldValueDefined && !isEdit && /*#__PURE__*/_react.default.createElement("p", {
    "data-test-subj": "no-custom-field-value"
  }, _translations.NO_CUSTOM_FIELD_SET), !isEdit && isCustomFieldValueDefined && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_view.View, {
    customField: customField
  })), isEdit && canUpdate && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m",
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(FormWrapperComponent, {
    initialValue: initialValue,
    isLoading: isLoading,
    onChange: setFormState,
    customFieldConfiguration: customFieldConfiguration
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    color: "success",
    "data-test-subj": `case-text-custom-field-submit-button-${customFieldConfiguration.key}`,
    fill: true,
    iconType: "save",
    onClick: onSubmitCustomField,
    size: "s",
    disabled: !isTextFieldValid || isLoading
  }, _translations.SAVE)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    "data-test-subj": `case-text-custom-field-cancel-button-${customFieldConfiguration.key}`,
    iconType: "cross",
    onClick: onCancel,
    size: "s"
  }, _translations.CANCEL)))))));
};
EditComponent.displayName = 'Edit';
const Edit = exports.Edit = /*#__PURE__*/_react.default.memo(EditComponent);
