// @kbn/core-elasticsearch-client-server-internal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AgentManager", {
  enumerable: true,
  get: function () {
    return _agent_manager.AgentManager;
  }
});
Object.defineProperty(exports, "ClusterClient", {
  enumerable: true,
  get: function () {
    return _cluster_client.ClusterClient;
  }
});
Object.defineProperty(exports, "DEFAULT_HEADERS", {
  enumerable: true,
  get: function () {
    return _headers.DEFAULT_HEADERS;
  }
});
Object.defineProperty(exports, "PRODUCT_ORIGIN_HEADER", {
  enumerable: true,
  get: function () {
    return _headers.PRODUCT_ORIGIN_HEADER;
  }
});
Object.defineProperty(exports, "PRODUCT_RESPONSE_HEADER", {
  enumerable: true,
  get: function () {
    return _headers.PRODUCT_RESPONSE_HEADER;
  }
});
Object.defineProperty(exports, "RESERVED_HEADERS", {
  enumerable: true,
  get: function () {
    return _headers.RESERVED_HEADERS;
  }
});
Object.defineProperty(exports, "ScopedClusterClient", {
  enumerable: true,
  get: function () {
    return _scoped_cluster_client.ScopedClusterClient;
  }
});
Object.defineProperty(exports, "USER_AGENT_HEADER", {
  enumerable: true,
  get: function () {
    return _headers.USER_AGENT_HEADER;
  }
});
Object.defineProperty(exports, "configureClient", {
  enumerable: true,
  get: function () {
    return _configure_client.configureClient;
  }
});
Object.defineProperty(exports, "getErrorMessage", {
  enumerable: true,
  get: function () {
    return _log_query_and_deprecation.getErrorMessage;
  }
});
Object.defineProperty(exports, "getRequestDebugMeta", {
  enumerable: true,
  get: function () {
    return _log_query_and_deprecation.getRequestDebugMeta;
  }
});
var _scoped_cluster_client = require("./src/scoped_cluster_client");
var _cluster_client = require("./src/cluster_client");
var _configure_client = require("./src/configure_client");
var _agent_manager = require("./src/agent_manager");
var _log_query_and_deprecation = require("./src/log_query_and_deprecation");
var _headers = require("./src/headers");
