// @kbn/enterprise-search-plugin/public/applications/workplace_search/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  DAYS_OF_WEEK_VALUES: true,
  FeatureIds: true
};
exports.FeatureIds = exports.DAYS_OF_WEEK_VALUES = void 0;
var _workplace_search = require("../../../common/types/workplace_search");
Object.keys(_workplace_search).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _workplace_search[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _workplace_search[key];
    }
  });
});
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DAYS_OF_WEEK_VALUES = exports.DAYS_OF_WEEK_VALUES = ['sunday', 'monday', 'tuesday', 'wednesday', 'thursday', 'friday', 'saturday'];
let FeatureIds = exports.FeatureIds = /*#__PURE__*/function (FeatureIds) {
  FeatureIds["SyncFrequency"] = "SyncFrequency";
  FeatureIds["SyncedItems"] = "SyncedItems";
  FeatureIds["SearchableContent"] = "SearchableContent";
  FeatureIds["Remote"] = "Remote";
  FeatureIds["Private"] = "Private";
  FeatureIds["GlobalAccessPermissions"] = "GlobalAccessPermissions";
  FeatureIds["DocumentLevelPermissions"] = "DocumentLevelPermissions";
  return FeatureIds;
}({}); // https://www.elastic.co/guide/en/workplace-search/current/workplace-search-custom-sources-api.html#_schema_data_types
