// @kbn/triggers-actions-ui-plugin/public/application/hooks/use_load_config_query.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useLoadConfigQuery = void 0;
var _reactQuery = require("@tanstack/react-query");
var _fetch_ui_config = require("@kbn/alerts-ui-shared/src/common/apis/fetch_ui_config");
var _kibana = require("../../common/lib/kibana");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useLoadConfigQuery = () => {
  const {
    http
  } = (0, _kibana.useKibana)().services;
  const {
    data
  } = (0, _reactQuery.useQuery)({
    queryKey: ['loadConfig'],
    queryFn: () => {
      return (0, _fetch_ui_config.fetchUiConfig)({
        http
      });
    },
    initialData: {
      isUsingSecurity: false
    },
    refetchOnWindowFocus: false
  });
  return {
    config: data
  };
};
exports.useLoadConfigQuery = useLoadConfigQuery;
