// @kbn/security-solution-plugin/public/detection_engine/rule_creation_ui/components/throttle_select_field/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThrottleSelectField = exports.THROTTLE_OPTIONS_FOR_RULE_CREATION_AND_EDITING = exports.THROTTLE_OPTIONS_FOR_BULK_RULE_ACTIONS = exports.DEFAULT_THROTTLE_OPTION = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _constants = require("../../../../../common/constants");
var _shared_imports = require("../../../../shared_imports");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const THROTTLE_OPTIONS_FOR_BULK_RULE_ACTIONS = exports.THROTTLE_OPTIONS_FOR_BULK_RULE_ACTIONS = [{
  value: _constants.NOTIFICATION_THROTTLE_RULE,
  text: 'On each rule execution'
}, {
  value: '1h',
  text: 'Hourly'
}, {
  value: '1d',
  text: 'Daily'
}, {
  value: '7d',
  text: 'Weekly'
}];
const THROTTLE_OPTIONS_FOR_RULE_CREATION_AND_EDITING = exports.THROTTLE_OPTIONS_FOR_RULE_CREATION_AND_EDITING = [{
  value: _constants.NOTIFICATION_THROTTLE_NO_ACTIONS,
  text: 'Perform no actions'
}, ...THROTTLE_OPTIONS_FOR_BULK_RULE_ACTIONS];
const DEFAULT_THROTTLE_OPTION = exports.DEFAULT_THROTTLE_OPTION = THROTTLE_OPTIONS_FOR_RULE_CREATION_AND_EDITING[0];
const ThrottleSelectField = props => {
  const {
    setValue
  } = props.field;
  const onChange = (0, _react.useCallback)(e => {
    const throttle = e.target.value;
    setValue(throttle);
  }, [setValue]);
  const newEuiFieldProps = {
    ...props.euiFieldProps,
    onChange
  };
  return /*#__PURE__*/_react.default.createElement(_shared_imports.SelectField, (0, _extends2.default)({}, props, {
    euiFieldProps: newEuiFieldProps
  }));
};
exports.ThrottleSelectField = ThrottleSelectField;
