// @kbn/apm-plugin/server/routes/services/get_service_instances/get_service_instances_system_metric_statistics.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getServiceInstancesSystemMetricStatistics = getServiceInstancesSystemMetricStatistics;
var _server = require("@kbn/observability-plugin/server");
var _common = require("@kbn/observability-plugin/common");
var _apm = require("../../../../common/es_fields/apm");
var _service_nodes = require("../../../../common/service_nodes");
var _environment_query = require("../../../../common/utils/environment_query");
var _get_bucket_size = require("../../../../common/utils/get_bucket_size");
var _memory = require("../../metrics/by_agent/shared/memory");
var _get_offset_in_ms = require("../../../../common/utils/get_offset_in_ms");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

async function getServiceInstancesSystemMetricStatistics({
  environment,
  kuery,
  apmEventClient,
  serviceName,
  size,
  start,
  end,
  serviceNodeIds,
  numBuckets,
  includeTimeseries,
  offset
}) {
  var _response$aggregation;
  const {
    startWithOffset,
    endWithOffset
  } = (0, _get_offset_in_ms.getOffsetInMs)({
    start,
    end,
    offset
  });
  const {
    intervalString
  } = (0, _get_bucket_size.getBucketSize)({
    start: startWithOffset,
    end: endWithOffset,
    numBuckets
  });
  const cpuUsageFilter = {
    exists: {
      field: _apm.METRIC_PROCESS_CPU_PERCENT
    }
  };
  function withTimeseries(agg) {
    return {
      ...(includeTimeseries ? {
        avg: {
          avg: agg
        },
        timeseries: {
          date_histogram: {
            field: '@timestamp',
            fixed_interval: intervalString,
            min_doc_count: 0,
            extended_bounds: {
              min: startWithOffset,
              max: endWithOffset
            }
          },
          aggs: {
            avg: {
              avg: agg
            }
          }
        }
      } : {
        avg: {
          avg: agg
        }
      })
    };
  }
  const subAggs = {
    memory_usage_cgroup: {
      filter: _memory.cgroupMemory.filter,
      aggs: withTimeseries({
        script: _memory.cgroupMemory.script
      })
    },
    memory_usage_system: {
      filter: _memory.systemMemory.filter,
      aggs: withTimeseries({
        script: _memory.systemMemory.script
      })
    },
    cpu_usage: {
      filter: cpuUsageFilter,
      aggs: withTimeseries({
        field: _apm.METRIC_PROCESS_CPU_PERCENT
      })
    }
  };
  const response = await apmEventClient.search('get_service_instances_system_metric_statistics', {
    apm: {
      events: [_common.ProcessorEvent.metric]
    },
    body: {
      track_total_hits: false,
      size: 0,
      query: {
        bool: {
          filter: [{
            term: {
              [_apm.SERVICE_NAME]: serviceName
            }
          }, ...(0, _server.rangeQuery)(startWithOffset, endWithOffset), ...(0, _environment_query.environmentQuery)(environment), ...(0, _server.kqlQuery)(kuery), ...(serviceNodeIds !== null && serviceNodeIds !== void 0 && serviceNodeIds.length ? [{
            terms: {
              [_apm.SERVICE_NODE_NAME]: serviceNodeIds
            }
          }] : []), {
            bool: {
              should: [_memory.cgroupMemory.filter, _memory.systemMemory.filter, cpuUsageFilter],
              minimum_should_match: 1
            }
          }]
        }
      },
      aggs: {
        [_apm.SERVICE_NODE_NAME]: {
          terms: {
            field: _apm.SERVICE_NODE_NAME,
            missing: _service_nodes.SERVICE_NODE_NAME_MISSING,
            ...(size ? {
              size
            } : {}),
            ...(serviceNodeIds !== null && serviceNodeIds !== void 0 && serviceNodeIds.length ? {
              include: serviceNodeIds
            } : {})
          },
          aggs: subAggs
        }
      }
    }
  });
  return ((_response$aggregation = response.aggregations) === null || _response$aggregation === void 0 ? void 0 : _response$aggregation[_apm.SERVICE_NODE_NAME].buckets.map(serviceNodeBucket => {
    const serviceNodeName = String(serviceNodeBucket.key);
    const hasCGroupData = serviceNodeBucket.memory_usage_cgroup.avg.value !== null;
    const memoryMetricsKey = hasCGroupData ? 'memory_usage_cgroup' : 'memory_usage_system';
    const cpuUsage =
    // Timeseries is available when includeTimeseries is true
    'timeseries' in serviceNodeBucket.cpu_usage ? serviceNodeBucket.cpu_usage.timeseries.buckets.map(dateBucket => ({
      x: dateBucket.key,
      y: dateBucket.avg.value
    })) : serviceNodeBucket.cpu_usage.avg.value;
    const memoryUsageValue = serviceNodeBucket[memoryMetricsKey];
    const memoryUsage =
    // Timeseries is available when includeTimeseries is true
    'timeseries' in memoryUsageValue ? memoryUsageValue.timeseries.buckets.map(dateBucket => ({
      x: dateBucket.key,
      y: dateBucket.avg.value
    })) : serviceNodeBucket[memoryMetricsKey].avg.value;
    return {
      serviceNodeName,
      cpuUsage,
      memoryUsage
    };
  })) || [];
}
