// @kbn/slo-plugin/server/services/create_slo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateSLO = void 0;
var _sloSchema = require("@kbn/slo-schema");
var _std = require("@kbn/std");
var _uuid = require("uuid");
var _constants = require("../../common/constants");
var _slo_summary_pipeline_template = require("../assets/ingest_templates/slo_summary_pipeline_template");
var _models = require("../domain/models");
var _services = require("../domain/services");
var _errors = require("../errors");
var _retry = require("../utils/retry");
var _create_temp_summary = require("./summary_transform_generator/helpers/create_temp_summary");
var _get_transform_compite_query = require("./utils/get_transform_compite_query");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class CreateSLO {
  constructor(esClient, repository, transformManager, summaryTransformManager, logger, spaceId, basePath) {
    this.esClient = esClient;
    this.repository = repository;
    this.transformManager = transformManager;
    this.summaryTransformManager = summaryTransformManager;
    this.logger = logger;
    this.spaceId = spaceId;
    this.basePath = basePath;
  }
  async execute(params) {
    const slo = this.toSLO(params);
    (0, _services.validateSLO)(slo);
    const rollbackOperations = [];
    await this.repository.save(slo, {
      throwOnConflict: true
    });
    rollbackOperations.push(() => this.repository.deleteById(slo.id));
    const rollupTransformId = (0, _constants.getSLOTransformId)(slo.id, slo.revision);
    const summaryTransformId = (0, _constants.getSLOSummaryTransformId)(slo.id, slo.revision);
    try {
      await this.transformManager.install(slo);
      rollbackOperations.push(() => this.transformManager.uninstall(rollupTransformId));
      await this.transformManager.start(rollupTransformId);
      rollbackOperations.push(() => this.transformManager.stop(rollupTransformId));
      await (0, _retry.retryTransientEsErrors)(() => this.esClient.ingest.putPipeline((0, _slo_summary_pipeline_template.getSLOSummaryPipelineTemplate)(slo, this.spaceId, this.basePath)), {
        logger: this.logger
      });
      rollbackOperations.push(() => this.esClient.ingest.deletePipeline({
        id: (0, _constants.getSLOSummaryPipelineId)(slo.id, slo.revision)
      }, {
        ignore: [404]
      }));
      await this.summaryTransformManager.install(slo);
      rollbackOperations.push(() => this.summaryTransformManager.uninstall(summaryTransformId));
      await this.summaryTransformManager.start(summaryTransformId);
      rollbackOperations.push(() => this.summaryTransformManager.stop(summaryTransformId));
      await (0, _retry.retryTransientEsErrors)(() => this.esClient.index({
        index: _constants.SLO_SUMMARY_TEMP_INDEX_NAME,
        id: `slo-${slo.id}`,
        document: (0, _create_temp_summary.createTempSummaryDocument)(slo, this.spaceId, this.basePath),
        refresh: true
      }), {
        logger: this.logger
      });
    } catch (err) {
      var _err$meta, _err$meta$body, _err$meta$body$error;
      this.logger.error(`Cannot install the SLO [id: ${slo.id}, revision: ${slo.revision}]. Rolling back.`);
      await (0, _std.asyncForEach)(rollbackOperations.reverse(), async operation => {
        try {
          await operation();
        } catch (rollbackErr) {
          this.logger.error('Rollback operation failed', rollbackErr);
        }
      });
      if (((_err$meta = err.meta) === null || _err$meta === void 0 ? void 0 : (_err$meta$body = _err$meta.body) === null || _err$meta$body === void 0 ? void 0 : (_err$meta$body$error = _err$meta$body.error) === null || _err$meta$body$error === void 0 ? void 0 : _err$meta$body$error.type) === 'security_exception') {
        throw new _errors.SecurityException(err.meta.body.error.reason);
      }
      throw err;
    }
    return this.toResponse(slo);
  }
  async inspect(params) {
    const slo = this.toSLO(params);
    (0, _services.validateSLO)(slo);
    const rollUpTransform = await this.transformManager.inspect(slo);
    const pipeline = (0, _slo_summary_pipeline_template.getSLOSummaryPipelineTemplate)(slo, this.spaceId, this.basePath);
    const summaryTransform = await this.summaryTransformManager.inspect(slo);
    const temporaryDoc = (0, _create_temp_summary.createTempSummaryDocument)(slo, this.spaceId, this.basePath);
    return {
      slo,
      pipeline,
      temporaryDoc,
      summaryTransform,
      rollUpTransform,
      // @ts-expect-error there is some issue with deprecated types being used in es types
      rollUpTransformCompositeQuery: (0, _get_transform_compite_query.getTransformQueryComposite)(rollUpTransform),
      // @ts-expect-error there is some issue with deprecated types being used in es types
      summaryTransformCompositeQuery: (0, _get_transform_compite_query.getTransformQueryComposite)(summaryTransform)
    };
  }
  toSLO(params) {
    var _params$id, _params$settings$sync, _params$settings, _params$settings$freq, _params$settings2, _params$settings$prev, _params$settings3, _params$revision, _params$tags;
    const now = new Date();
    return {
      ...params,
      id: (_params$id = params.id) !== null && _params$id !== void 0 ? _params$id : (0, _uuid.v4)(),
      settings: {
        syncDelay: (_params$settings$sync = (_params$settings = params.settings) === null || _params$settings === void 0 ? void 0 : _params$settings.syncDelay) !== null && _params$settings$sync !== void 0 ? _params$settings$sync : new _models.Duration(1, _models.DurationUnit.Minute),
        frequency: (_params$settings$freq = (_params$settings2 = params.settings) === null || _params$settings2 === void 0 ? void 0 : _params$settings2.frequency) !== null && _params$settings$freq !== void 0 ? _params$settings$freq : new _models.Duration(1, _models.DurationUnit.Minute),
        preventInitialBackfill: (_params$settings$prev = (_params$settings3 = params.settings) === null || _params$settings3 === void 0 ? void 0 : _params$settings3.preventInitialBackfill) !== null && _params$settings$prev !== void 0 ? _params$settings$prev : false
      },
      revision: (_params$revision = params.revision) !== null && _params$revision !== void 0 ? _params$revision : 1,
      enabled: true,
      tags: (_params$tags = params.tags) !== null && _params$tags !== void 0 ? _params$tags : [],
      createdAt: now,
      updatedAt: now,
      groupBy: !!params.groupBy ? params.groupBy : _sloSchema.ALL_VALUE,
      version: _constants.SLO_MODEL_VERSION
    };
  }
  toResponse(slo) {
    return {
      id: slo.id
    };
  }
}
exports.CreateSLO = CreateSLO;
