// @kbn/cases-plugin/server/routes/api/user_actions/find_user_actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.findUserActionsRoute = void 0;
var _configSchema = require("@kbn/config-schema");
var _constants = require("../../../../common/constants");
var _error = require("../../../common/error");
var _create_cases_route = require("../create_cases_route");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const findUserActionsRoute = exports.findUserActionsRoute = (0, _create_cases_route.createCasesRoute)({
  method: 'get',
  path: _constants.CASE_FIND_USER_ACTIONS_URL,
  params: {
    params: _configSchema.schema.object({
      case_id: _configSchema.schema.string()
    })
  },
  routerOptions: {
    access: 'public',
    summary: `Find case activity`,
    tags: ['oas-tag:cases'],
    description: 'Returns a paginated list of user activity for a case.'
    // You must have `read` privileges for the **Cases** feature in the
    // **Management**, **Observability**, or **Security** section of the Kibana
    // feature privileges, depending on the owner of the case you're seeking.
  },
  handler: async ({
    context,
    request,
    response
  }) => {
    try {
      const caseContext = await context.cases;
      const casesClient = await caseContext.getCasesClient();
      const caseId = request.params.case_id;
      const options = request.query;
      const res = await casesClient.userActions.find({
        caseId,
        params: options
      });
      return response.ok({
        body: res
      });
    } catch (error) {
      throw (0, _error.createCaseError)({
        message: `Failed to find user actions in route for case id: ${request.params.case_id}: ${error}`,
        error
      });
    }
  }
});
