// @kbn/core-saved-objects-server/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ALERTING_CASES_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.ALERTING_CASES_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "ALL_SAVED_OBJECT_INDICES", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.ALL_SAVED_OBJECT_INDICES;
  }
});
Object.defineProperty(exports, "ANALYTICS_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.ANALYTICS_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "ENCRYPTION_EXTENSION_ID", {
  enumerable: true,
  get: function () {
    return _extensions.ENCRYPTION_EXTENSION_ID;
  }
});
Object.defineProperty(exports, "INGEST_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.INGEST_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "MAIN_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.MAIN_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "SECURITY_EXTENSION_ID", {
  enumerable: true,
  get: function () {
    return _extensions.SECURITY_EXTENSION_ID;
  }
});
Object.defineProperty(exports, "SECURITY_SOLUTION_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.SECURITY_SOLUTION_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "SPACES_EXTENSION_ID", {
  enumerable: true,
  get: function () {
    return _extensions.SPACES_EXTENSION_ID;
  }
});
Object.defineProperty(exports, "SavedObjectsErrorHelpers", {
  enumerable: true,
  get: function () {
    return _saved_objects_error_helpers.SavedObjectsErrorHelpers;
  }
});
Object.defineProperty(exports, "TASK_MANAGER_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _saved_objects_index_pattern.TASK_MANAGER_SAVED_OBJECT_INDEX;
  }
});
var _saved_objects_index_pattern = require("./src/saved_objects_index_pattern");
var _extensions = require("./src/extensions/extensions");
var _saved_objects_error_helpers = require("./src/saved_objects_error_helpers");
