// @kbn/security-solution-plugin/public/resolver/store/camera/scaling_constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.zoomCurveRate = exports.unitsPerNudge = exports.panAnimationDuration = exports.nudgeAnimationDuration = exports.minimum = exports.maximum = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * The minimum allowed value for the camera scale. This is the least scale that we will ever render something at.
 */
const minimum = exports.minimum = 0.5;

/**
 * The maximum allowed value for the camera scale. This is greatest scale that we will ever render something at.
 */
const maximum = exports.maximum = 2;

/**
 * The curve of the zoom function growth rate. The higher the scale factor is, the higher the zoom rate will be.
 */
const zoomCurveRate = exports.zoomCurveRate = 2;

/**
 * The size, in world units, of a 'nudge' as caused by clicking the up, right, down, or left panning buttons.
 */
const unitsPerNudge = exports.unitsPerNudge = 50;

/**
 * The duration a nudge animation lasts.
 */
const nudgeAnimationDuration = exports.nudgeAnimationDuration = 300;

/**
 * The duration a panning animation lasts
 */
const panAnimationDuration = exports.panAnimationDuration = 1000;
