// @kbn/share-plugin/server/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CSV_QUOTE_VALUES_SETTING", {
  enumerable: true,
  get: function () {
    return _constants.CSV_QUOTE_VALUES_SETTING;
  }
});
Object.defineProperty(exports, "CSV_SEPARATOR_SETTING", {
  enumerable: true,
  get: function () {
    return _constants.CSV_SEPARATOR_SETTING;
  }
});
exports.config = void 0;
exports.plugin = plugin;
var _config = require("../common/config");
var _constants = require("../common/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
async function plugin(initializerContext) {
  const {
    SharePlugin
  } = await Promise.resolve().then(() => _interopRequireWildcard(require('./plugin')));
  return new SharePlugin(initializerContext);
}
const config = exports.config = {
  exposeToBrowser: {
    new_version: {
      enabled: true
    }
  },
  schema: _config.configSchema
};
