// @kbn/threat-intelligence-plugin/public/modules/indicators/components/flyout/test_ids.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TIMELINE_BUTTON_TEST_ID = exports.TAKE_ACTION_BUTTON_TEST_ID = exports.POPOVER_BUTTON_TEST_ID = exports.INVESTIGATE_IN_TIMELINE_TEST_ID = exports.INDICATORS_FLYOUT_TITLE_TEST_ID = exports.INDICATORS_FLYOUT_TABS_TEST_ID = exports.INDICATORS_FLYOUT_SUBTITLE_TEST_ID = exports.INDICATORS_FLYOUT_OVERVIEW_TITLE = exports.INDICATORS_FLYOUT_OVERVIEW_TABLE = exports.INDICATORS_FLYOUT_OVERVIEW_HIGH_LEVEL_BLOCKS = exports.FLYOUT_TABLE_TEST_ID = exports.FILTER_OUT_BUTTON_TEST_ID = exports.FILTER_IN_BUTTON_TEST_ID = exports.COPY_TO_CLIPBOARD_BUTTON_TEST_ID = exports.CODE_BLOCK_TEST_ID = exports.ADD_TO_NEW_CASE_TEST_ID = exports.ADD_TO_EXISTING_CASE_TEST_ID = exports.ADD_TO_BLOCK_LIST_TEST_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/* Flyout */
const INDICATORS_FLYOUT_TITLE_TEST_ID = exports.INDICATORS_FLYOUT_TITLE_TEST_ID = 'tiIndicatorFlyoutTitle';
const INDICATORS_FLYOUT_SUBTITLE_TEST_ID = exports.INDICATORS_FLYOUT_SUBTITLE_TEST_ID = 'tiIndicatorFlyoutSubtitle';
const INDICATORS_FLYOUT_TABS_TEST_ID = exports.INDICATORS_FLYOUT_TABS_TEST_ID = 'tiIndicatorFlyoutTabs';

/* Indicator Value Actions */
const TIMELINE_BUTTON_TEST_ID = exports.TIMELINE_BUTTON_TEST_ID = 'TimelineButton';
const FILTER_IN_BUTTON_TEST_ID = exports.FILTER_IN_BUTTON_TEST_ID = 'FilterInButton';
const FILTER_OUT_BUTTON_TEST_ID = exports.FILTER_OUT_BUTTON_TEST_ID = 'FilterOutButton';
const COPY_TO_CLIPBOARD_BUTTON_TEST_ID = exports.COPY_TO_CLIPBOARD_BUTTON_TEST_ID = 'CopyToClipboardButton';
const POPOVER_BUTTON_TEST_ID = exports.POPOVER_BUTTON_TEST_ID = 'PopoverButton';

/* Take Action */
const TAKE_ACTION_BUTTON_TEST_ID = exports.TAKE_ACTION_BUTTON_TEST_ID = 'tiIndicatorFlyoutTakeActionButton';
const INVESTIGATE_IN_TIMELINE_TEST_ID = exports.INVESTIGATE_IN_TIMELINE_TEST_ID = 'tiIndicatorFlyoutInvestigateInTimelineContextMenu';
const ADD_TO_EXISTING_CASE_TEST_ID = exports.ADD_TO_EXISTING_CASE_TEST_ID = 'tiIndicatorFlyoutAddToExistingCaseContextMenu';
const ADD_TO_NEW_CASE_TEST_ID = exports.ADD_TO_NEW_CASE_TEST_ID = 'tiIndicatorFlyoutAddToNewCaseContextMenu';
const ADD_TO_BLOCK_LIST_TEST_ID = exports.ADD_TO_BLOCK_LIST_TEST_ID = 'tiIndicatorFlyoutAddToBlockListContextMenu';

/* Json Tab */
const CODE_BLOCK_TEST_ID = exports.CODE_BLOCK_TEST_ID = 'tiFlyoutJsonCodeBlock';

/* Table Tab */
const FLYOUT_TABLE_TEST_ID = exports.FLYOUT_TABLE_TEST_ID = 'tiFlyoutTable';

/* Overview Tab */
const INDICATORS_FLYOUT_OVERVIEW_TITLE = exports.INDICATORS_FLYOUT_OVERVIEW_TITLE = 'tiFlyoutOverviewTitle';
const INDICATORS_FLYOUT_OVERVIEW_TABLE = exports.INDICATORS_FLYOUT_OVERVIEW_TABLE = 'tiFlyoutOverviewTableRow';
const INDICATORS_FLYOUT_OVERVIEW_HIGH_LEVEL_BLOCKS = exports.INDICATORS_FLYOUT_OVERVIEW_HIGH_LEVEL_BLOCKS = 'tiFlyoutOverviewHighLevelBlocks';
