// @kbn/infra-plugin/public/components/asset_details/tabs/anomalies/anomalies.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Anomalies = void 0;
var _react = _interopRequireWildcard(require("react"));
var _public = require("@kbn/observability-shared-plugin/public");
var _anomalies_table = require("../../../ml/anomaly_detection/anomalies_table/anomalies_table");
var _use_asset_details_render_props = require("../../hooks/use_asset_details_render_props");
var _use_date_picker = require("../../hooks/use_date_picker");
var _use_intersecting_state = require("../../hooks/use_intersecting_state");
var _use_request_observable = require("../../hooks/use_request_observable");
var _use_kibana = require("../../../../hooks/use_kibana");
var _flyout_home = require("../../../ml/anomaly_detection/flyout_home");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Anomalies = () => {
  var _overrides$anomalies;
  const ref = (0, _react.useRef)(null);
  const {
    request$
  } = (0, _use_request_observable.useRequestObservable)();
  const {
    getParsedDateRange
  } = (0, _use_date_picker.useDatePickerContext)();
  const {
    asset,
    overrides
  } = (0, _use_asset_details_render_props.useAssetDetailsRenderPropsContext)();
  const {
    kibanaVersion,
    isCloudEnv,
    isServerlessEnv
  } = (0, _react.useContext)(_use_kibana.KibanaEnvironmentContext);
  const {
    onClose = () => {}
  } = (_overrides$anomalies = overrides === null || overrides === void 0 ? void 0 : overrides.anomalies) !== null && _overrides$anomalies !== void 0 ? _overrides$anomalies : {};
  const parsedDateRange = (0, _react.useMemo)(() => getParsedDateRange(), [getParsedDateRange]);
  const state = (0, _use_intersecting_state.useIntersectingState)(ref, {
    parsedDateRange
  });
  return /*#__PURE__*/_react.default.createElement("div", {
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_anomalies_table.AnomaliesTable, {
    closeFlyout: onClose,
    hostName: asset.name,
    dateRange: state.parsedDateRange,
    hideDatePicker: true,
    request$: request$
  }), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      position: 'fixed',
      bottom: '24px',
      right: '24px',
      display: 'flex',
      justifyContent: 'flex-end',
      alignItems: 'flex-end'
    }
  }, /*#__PURE__*/_react.default.createElement(_public.FeatureFeedbackButton, {
    "data-test-subj": "infraMLHostFlyoutFeedbackLink",
    formUrl: _flyout_home.INFRA_ML_FLYOUT_FEEDBACK_LINK,
    kibanaVersion: kibanaVersion,
    isCloudEnv: isCloudEnv,
    isServerlessEnv: isServerlessEnv,
    nodeType: 'host'
  })));
};
exports.Anomalies = Anomalies;
