// @kbn/infra-plugin/server/routes/profiling/lib/fetch_profiling_flamegraph.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fetchProfilingFlamegraph = fetchProfilingFlamegraph;
var _server = require("@kbn/observability-plugin/server");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

async function fetchProfilingFlamegraph({
  kuery,
  from,
  to
}, profilingDataAccess, coreRequestContext) {
  const startSecs = from / 1000;
  const endSecs = to / 1000;
  return await profilingDataAccess.services.fetchFlamechartData({
    core: coreRequestContext,
    esClient: coreRequestContext.elasticsearch.client.asCurrentUser,
    totalSeconds: endSecs - startSecs,
    query: {
      bool: {
        filter: [...(0, _server.kqlQuery)(kuery), {
          range: {
            ['@timestamp']: {
              gte: String(startSecs),
              lt: String(endSecs),
              format: 'epoch_second'
            }
          }
        }]
      }
    }
  });
}
