// @kbn/metrics-data-access-plugin/public/components/infrastructure_node_metrics_tables/shared/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "averageOfValues", {
  enumerable: true,
  get: function () {
    return _helpers.averageOfValues;
  }
});
Object.defineProperty(exports, "createMetricByFieldLookup", {
  enumerable: true,
  get: function () {
    return _metrics_to_api_options.createMetricByFieldLookup;
  }
});
Object.defineProperty(exports, "makeUnpackMetric", {
  enumerable: true,
  get: function () {
    return _helpers.makeUnpackMetric;
  }
});
Object.defineProperty(exports, "metricsToApiOptions", {
  enumerable: true,
  get: function () {
    return _metrics_to_api_options.metricsToApiOptions;
  }
});
Object.defineProperty(exports, "scaleUpPercentage", {
  enumerable: true,
  get: function () {
    return _helpers.scaleUpPercentage;
  }
});
Object.defineProperty(exports, "useInfrastructureNodeMetrics", {
  enumerable: true,
  get: function () {
    return _use_infrastructure_node_metrics.useInfrastructureNodeMetrics;
  }
});
var _helpers = require("./helpers");
var _metrics_to_api_options = require("./metrics_to_api_options");
var _use_infrastructure_node_metrics = require("./use_infrastructure_node_metrics");
