// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/pick_fields_step/components/population_view/metric_selection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PopulationDetectors = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _job_creator_context = require("../../../job_creator_context");
var _new_job_capabilities_service = require("../../../../../../../services/new_job_capabilities/new_job_capabilities_service");
var _fields_utils = require("../../../../../../../../../common/util/fields_utils");
var _settings = require("../../../charts/common/settings");
var _metric_selector = require("./metric_selector");
var _population_field = require("../population_field");
var _chart_grid = require("./chart_grid");
var _toast_notification_service = require("../../../../../../../services/toast_notification_service");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PopulationDetectors = ({
  setIsValid
}) => {
  const {
    jobCreator: jc,
    jobCreatorUpdate,
    jobCreatorUpdated,
    chartLoader,
    chartInterval
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const jobCreator = jc;
  const fields = (0, _react.useMemo)(() => (0, _fields_utils.sortFields)([..._new_job_capabilities_service.newJobCapsService.fields, ...jobCreator.runtimeFields]),
  // eslint-disable-next-line react-hooks/exhaustive-deps
  []);
  const [selectedOptions, setSelectedOptions] = (0, _react.useState)([]);
  const [aggFieldPairList, setAggFieldPairList] = (0, _react.useState)(jobCreator.aggFieldPairs);
  const [lineChartsData, setLineChartsData] = (0, _react.useState)({});
  const [loadingData, setLoadingData] = (0, _react.useState)(false);
  const [start, setStart] = (0, _react.useState)(jobCreator.start);
  const [end, setEnd] = (0, _react.useState)(jobCreator.end);
  const [bucketSpanMs, setBucketSpanMs] = (0, _react.useState)(jobCreator.bucketSpanMs);
  const [chartSettings, setChartSettings] = (0, _react.useState)(_settings.defaultChartSettings);
  const [populationField, setPopulationField] = (0, _react.useState)(jobCreator.populationField);
  const [fieldValuesPerDetector, setFieldValuesPerDetector] = (0, _react.useState)({});
  const [byFieldsUpdated, setByFieldsUpdated] = (0, _react.useReducer)(s => s + 1, 0);
  const [pageReady, setPageReady] = (0, _react.useState)(false);
  const updateByFields = () => setByFieldsUpdated(0);
  function detectorChangeHandler(selectedOptionsIn) {
    addDetector(selectedOptionsIn);
  }
  function addDetector(selectedOptionsIn) {
    if (selectedOptionsIn !== null && selectedOptionsIn.length) {
      const option = selectedOptionsIn[0];
      if (typeof option !== 'undefined') {
        const newPair = {
          agg: option.agg,
          field: option.field,
          by: {
            field: null,
            value: null
          }
        };
        setAggFieldPairList([...aggFieldPairList, newPair]);
        setSelectedOptions([]);
      } else {
        setAggFieldPairList([]);
      }
    }
  }
  function deleteDetector(index) {
    aggFieldPairList.splice(index, 1);
    setAggFieldPairList([...aggFieldPairList]);
    updateByFields();
  }
  (0, _react.useEffect)(() => {
    setPageReady(true);
  }, []);

  // watch for changes in detector list length
  (0, _react.useEffect)(() => {
    jobCreator.removeAllDetectors();
    aggFieldPairList.forEach((pair, i) => {
      jobCreator.addDetector(pair.agg, pair.field);
      if (pair.by !== undefined) {
        // re-add by fields
        jobCreator.setByField(pair.by.field, i);
      }
    });
    jobCreatorUpdate();
    loadCharts();
    setIsValid(aggFieldPairList.length > 0);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [aggFieldPairList.length]);

  // watch for changes in by field values
  // redraw the charts if they change.
  // triggered when example fields have been loaded
  // if the split field or by fields have changed
  (0, _react.useEffect)(() => {
    loadCharts();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [JSON.stringify(fieldValuesPerDetector), populationField, pageReady]);

  // watch for change in jobCreator
  (0, _react.useEffect)(() => {
    if (jobCreator.start !== start || jobCreator.end !== end) {
      setStart(jobCreator.start);
      setEnd(jobCreator.end);
      loadCharts();
    }
    if (jobCreator.bucketSpanMs !== bucketSpanMs) {
      setBucketSpanMs(jobCreator.bucketSpanMs);
      loadCharts();
    }
    setPopulationField(jobCreator.populationField);

    // update by fields and their by fields
    let update = false;
    const newList = [...aggFieldPairList];
    newList.forEach((pair, i) => {
      const bf = jobCreator.getByField(i);
      if (pair.by !== undefined && pair.by.field !== bf) {
        pair.by.field = bf;
        update = true;
      }
    });
    if (update) {
      setAggFieldPairList(newList);
      updateByFields();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreatorUpdated]);

  // watch for changes in split field or by fields.
  // load example field values
  // changes to fieldValues here will trigger the card effect via setFieldValuesPerDetector
  (0, _react.useEffect)(() => {
    loadFieldExamples();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [populationField, byFieldsUpdated]);
  async function loadCharts() {
    if (allDataReady()) {
      setLoadingData(true);
      try {
        const cs = (0, _settings.getChartSettings)(jobCreator, chartInterval);
        setChartSettings(cs);
        const resp = await chartLoader.loadPopulationCharts(jobCreator.start, jobCreator.end, aggFieldPairList, jobCreator.populationField, cs.intervalMs, jobCreator.runtimeMappings, jobCreator.datafeedConfig.indices_options);
        setLineChartsData(resp);
      } catch (error) {
        (0, _toast_notification_service.getToastNotificationService)().displayErrorToast(error);
        setLineChartsData([]);
      }
      setLoadingData(false);
    }
  }
  async function loadFieldExamples() {
    const promises = [];
    aggFieldPairList.forEach((af, i) => {
      if (af.by !== undefined && af.by.field !== null) {
        promises.push((async (index, field) => {
          return {
            index,
            fields: await chartLoader.loadFieldExampleValues(field, jobCreator.runtimeMappings, jobCreator.datafeedConfig.indices_options)
          };
        })(i, af.by.field));
      }
    });
    const results = await Promise.all(promises);
    const fieldValues = results.reduce((p, c) => {
      p[c.index] = c.fields;
      return p;
    }, {});
    const newPairs = aggFieldPairList.map((pair, i) => ({
      ...pair,
      ...(pair.by === undefined || pair.by.field === null ? {} : {
        by: {
          ...pair.by,
          value: fieldValues[i][0]
        }
      })
    }));
    setAggFieldPairList([...newPairs]);
    setFieldValuesPerDetector(fieldValues);
  }
  function allDataReady() {
    let ready = aggFieldPairList.length > 0;
    aggFieldPairList.forEach(af => {
      if (af.by !== undefined && af.by.field !== null) {
        // if a by field is set, it's only ready when the value is loaded
        ready = ready && af.by.value !== null;
      }
    });
    return ready;
  }
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_population_field.PopulationFieldSelector, null), populationField !== null && /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    margin: "l"
  }), populationField !== null && /*#__PURE__*/_react.default.createElement(_chart_grid.ChartGrid, {
    aggFieldPairList: aggFieldPairList,
    chartSettings: chartSettings,
    splitField: populationField,
    lineChartsData: lineChartsData,
    modelData: [],
    anomalyData: [],
    deleteDetector: deleteDetector,
    jobType: jobCreator.type,
    fieldValuesPerDetector: fieldValuesPerDetector,
    loading: loadingData
  }), populationField !== null && /*#__PURE__*/_react.default.createElement(_metric_selector.MetricSelector, {
    fields: fields,
    detectorChangeHandler: detectorChangeHandler,
    selectedOptions: selectedOptions,
    removeOptions: []
  }));
};
exports.PopulationDetectors = PopulationDetectors;
