// @kbn/cases-plugin/common/api/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCustomFieldReplaceUrl = exports.getCasesFromAlertsUrl = exports.getCasesDeleteFileAttachmentsUrl = exports.getCaseUsersUrl = exports.getCaseUserActionUrl = exports.getCaseUserActionStatsUrl = exports.getCasePushUrl = exports.getCaseFindUserActionsUrl = exports.getCaseFindAttachmentsUrl = exports.getCaseDetailsUrl = exports.getCaseDetailsMetricsUrl = exports.getCaseConnectorsUrl = exports.getCaseConfigurationDetailsUrl = exports.getCaseCommentsUrl = exports.getCaseCommentDeleteUrl = exports.getCaseBulkGetAttachmentsUrl = void 0;
var _constants = require("../constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getCaseDetailsUrl = id => {
  return _constants.CASE_DETAILS_URL.replace('{case_id}', id);
};
exports.getCaseDetailsUrl = getCaseDetailsUrl;
const getCaseDetailsMetricsUrl = id => {
  return _constants.INTERNAL_CASE_METRICS_DETAILS_URL.replace('{case_id}', id);
};
exports.getCaseDetailsMetricsUrl = getCaseDetailsMetricsUrl;
const getCaseCommentsUrl = id => {
  return _constants.CASE_COMMENTS_URL.replace('{case_id}', id);
};
exports.getCaseCommentsUrl = getCaseCommentsUrl;
const getCaseFindAttachmentsUrl = caseId => {
  return _constants.CASE_FIND_ATTACHMENTS_URL.replace('{case_id}', caseId);
};
exports.getCaseFindAttachmentsUrl = getCaseFindAttachmentsUrl;
const getCaseCommentDeleteUrl = (caseId, commentId) => {
  return _constants.CASE_COMMENT_DELETE_URL.replace('{case_id}', caseId).replace('{comment_id}', commentId);
};
exports.getCaseCommentDeleteUrl = getCaseCommentDeleteUrl;
const getCaseUserActionUrl = id => {
  return _constants.CASE_USER_ACTIONS_URL.replace('{case_id}', id);
};
exports.getCaseUserActionUrl = getCaseUserActionUrl;
const getCaseUserActionStatsUrl = id => {
  return _constants.INTERNAL_GET_CASE_USER_ACTIONS_STATS_URL.replace('{case_id}', id);
};
exports.getCaseUserActionStatsUrl = getCaseUserActionStatsUrl;
const getCaseFindUserActionsUrl = id => {
  return _constants.CASE_FIND_USER_ACTIONS_URL.replace('{case_id}', id);
};
exports.getCaseFindUserActionsUrl = getCaseFindUserActionsUrl;
const getCasePushUrl = (caseId, connectorId) => {
  return _constants.CASE_PUSH_URL.replace('{case_id}', caseId).replace('{connector_id}', connectorId);
};
exports.getCasePushUrl = getCasePushUrl;
const getCaseConfigurationDetailsUrl = configureID => {
  return _constants.CASE_CONFIGURE_DETAILS_URL.replace('{configuration_id}', configureID);
};
exports.getCaseConfigurationDetailsUrl = getCaseConfigurationDetailsUrl;
const getCasesFromAlertsUrl = alertId => {
  return _constants.CASE_ALERTS_URL.replace('{alert_id}', alertId);
};
exports.getCasesFromAlertsUrl = getCasesFromAlertsUrl;
const getCaseBulkGetAttachmentsUrl = id => {
  return _constants.INTERNAL_BULK_GET_ATTACHMENTS_URL.replace('{case_id}', id);
};
exports.getCaseBulkGetAttachmentsUrl = getCaseBulkGetAttachmentsUrl;
const getCaseConnectorsUrl = id => {
  return _constants.INTERNAL_CONNECTORS_URL.replace('{case_id}', id);
};
exports.getCaseConnectorsUrl = getCaseConnectorsUrl;
const getCaseUsersUrl = id => {
  return _constants.INTERNAL_CASE_USERS_URL.replace('{case_id}', id);
};
exports.getCaseUsersUrl = getCaseUsersUrl;
const getCasesDeleteFileAttachmentsUrl = id => {
  return _constants.INTERNAL_DELETE_FILE_ATTACHMENTS_URL.replace('{case_id}', id);
};
exports.getCasesDeleteFileAttachmentsUrl = getCasesDeleteFileAttachmentsUrl;
const getCustomFieldReplaceUrl = (caseId, customFieldId) => {
  return _constants.INTERNAL_PUT_CUSTOM_FIELDS_URL.replace('{case_id}', caseId).replace('{custom_field_id}', customFieldId);
};
exports.getCustomFieldReplaceUrl = getCustomFieldReplaceUrl;
