// @kbn/cases-plugin/public/containers/user_profiles/use_bulk_get_user_profiles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useBulkGetUserProfiles = void 0;
var _reactQuery = require("@tanstack/react-query");
var i18n = _interopRequireWildcard(require("../translations"));
var _kibana = require("../../common/lib/kibana");
var _constants = require("../constants");
var _api = require("./api");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const profilesToMap = profiles => profiles.reduce((acc, profile) => {
  acc.set(profile.uid, profile);
  return acc;
}, new Map());
const useBulkGetUserProfiles = ({
  uids
}) => {
  const {
    security
  } = (0, _kibana.useKibana)().services;
  const toasts = (0, _kibana.useToasts)();
  return (0, _reactQuery.useQuery)(_constants.casesQueriesKeys.userProfilesList(uids), () => {
    return (0, _api.bulkGetUserProfiles)({
      security,
      uids
    });
  }, {
    select: profilesToMap,
    retry: false,
    keepPreviousData: true,
    staleTime: 60 * 1000,
    // one minute
    onError: error => {
      if (error.name !== 'AbortError') {
        toasts.addError(error.body && error.body.message ? new Error(error.body.message) : error, {
          title: i18n.ERROR_TITLE
        });
      }
    },
    enabled: uids.length > 0
  });
};
exports.useBulkGetUserProfiles = useBulkGetUserProfiles;
