// @kbn/uptime-plugin/public/legacy_uptime/components/overview/alerts/monitor_status_alert/old_alert_call_out.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OldAlertCallOut = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const OldAlertCallOut = ({
  isOldAlert
}) => {
  if (!isOldAlert) return null;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    size: "s",
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.uptime.alerts.monitorStatus.oldAlertCallout.title",
      defaultMessage: "You may be editing an older alert, some fields may not auto-populate."
    }),
    iconType: "warning"
  }));
};
exports.OldAlertCallOut = OldAlertCallOut;
