// @kbn/cases-plugin/public/components/edit_connector/push_button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PushButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PushButtonComponent = ({
  disabled,
  errorsMsg,
  isLoading,
  hasBeenPushed,
  connectorName,
  showTooltip,
  pushToService
}) => {
  const button = /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    "data-test-subj": "push-to-external-service",
    iconType: "importAction",
    onClick: pushToService,
    disabled: disabled,
    isLoading: isLoading
  }, hasBeenPushed ? i18n.UPDATE_INCIDENT(connectorName) : i18n.PUSH_INCIDENT(connectorName));
  return showTooltip ? /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    position: "top",
    title: errorsMsg.length > 0 ? errorsMsg[0].title : i18n.PUSH_LOCKED_TITLE(connectorName),
    content: /*#__PURE__*/_react.default.createElement("p", null, errorsMsg.length > 0 ? errorsMsg[0].description : i18n.PUSH_LOCKED_DESC),
    "data-test-subj": "push-button-tooltip"
  }, button) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, button);
};
PushButtonComponent.displayName = 'PushButton';
const PushButton = exports.PushButton = /*#__PURE__*/_react.default.memo(PushButtonComponent);
