// @kbn/securitysolution-lists-common/api/create_list_item/create_list_item.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateListItemResponse = exports.CreateListItemRequestBody = void 0;
var _zod = require("zod");
var _list_common = require("../model/list_common.gen");
var _list_schemas = require("../model/list_schemas.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Create list item API endpoint
 *   version: 2023-10-31
 */

const CreateListItemRequestBody = exports.CreateListItemRequestBody = _zod.z.object({
  id: _list_common.ListItemId.optional(),
  list_id: _list_common.ListId,
  value: _list_common.ListItemValue,
  meta: _list_common.ListItemMetadata.optional(),
  /**
   * Determines when changes made by the request are made visible to search
   */
  refresh: _zod.z.enum(['true', 'false', 'wait_for']).optional()
});
const CreateListItemResponse = exports.CreateListItemResponse = _list_schemas.ListItem;
