// @kbn/newsfeed-plugin/public/components/empty_news.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NewsEmptyPrompt = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const NewsEmptyPrompt = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    iconType: "documents",
    titleSize: "s",
    "data-test-subj": "emptyNewsfeed",
    title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "newsfeed.emptyPrompt.noNewsTitle",
      defaultMessage: "No news?"
    })),
    body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "newsfeed.emptyPrompt.noNewsText",
      defaultMessage: "If your Kibana instance doesn\u2019t have internet access, ask your administrator to disable this feature. Otherwise, we\u2019ll keep trying to fetch the news."
    }))
  });
};
exports.NewsEmptyPrompt = NewsEmptyPrompt;
