// @kbn/observability-shared-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AGENT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.AGENT;
  }
});
Object.defineProperty(exports, "AGENT_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.AGENT_NAME;
  }
});
Object.defineProperty(exports, "AGENT_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.AGENT_VERSION;
  }
});
Object.defineProperty(exports, "CLIENT_GEO", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLIENT_GEO;
  }
});
Object.defineProperty(exports, "CLIENT_GEO_COUNTRY_ISO_CODE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLIENT_GEO_COUNTRY_ISO_CODE;
  }
});
Object.defineProperty(exports, "CLIENT_GEO_COUNTRY_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLIENT_GEO_COUNTRY_NAME;
  }
});
Object.defineProperty(exports, "CLOUD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLOUD;
  }
});
Object.defineProperty(exports, "CLOUD_AVAILABILITY_ZONE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLOUD_AVAILABILITY_ZONE;
  }
});
Object.defineProperty(exports, "CLOUD_MACHINE_TYPE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLOUD_MACHINE_TYPE;
  }
});
Object.defineProperty(exports, "CLOUD_PROVIDER", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLOUD_PROVIDER;
  }
});
Object.defineProperty(exports, "CLOUD_REGION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLOUD_REGION;
  }
});
Object.defineProperty(exports, "CLS_FIELD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CLS_FIELD;
  }
});
Object.defineProperty(exports, "CONTAINER_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.CONTAINER_ID;
  }
});
Object.defineProperty(exports, "DESTINATION_ADDRESS", {
  enumerable: true,
  get: function () {
    return _elasticsearch.DESTINATION_ADDRESS;
  }
});
Object.defineProperty(exports, "ERROR_CULPRIT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_CULPRIT;
  }
});
Object.defineProperty(exports, "ERROR_EXC_HANDLED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_EXC_HANDLED;
  }
});
Object.defineProperty(exports, "ERROR_EXC_MESSAGE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_EXC_MESSAGE;
  }
});
Object.defineProperty(exports, "ERROR_EXC_TYPE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_EXC_TYPE;
  }
});
Object.defineProperty(exports, "ERROR_GROUP_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_GROUP_ID;
  }
});
Object.defineProperty(exports, "ERROR_LOG_LEVEL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_LOG_LEVEL;
  }
});
Object.defineProperty(exports, "ERROR_LOG_MESSAGE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_LOG_MESSAGE;
  }
});
Object.defineProperty(exports, "ERROR_PAGE_URL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.ERROR_PAGE_URL;
  }
});
Object.defineProperty(exports, "EVENT_OUTCOME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.EVENT_OUTCOME;
  }
});
Object.defineProperty(exports, "FCP_FIELD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.FCP_FIELD;
  }
});
Object.defineProperty(exports, "FID_FIELD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.FID_FIELD;
  }
});
Object.defineProperty(exports, "HOST", {
  enumerable: true,
  get: function () {
    return _elasticsearch.HOST;
  }
});
Object.defineProperty(exports, "HOST_HOSTNAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.HOST_HOSTNAME;
  }
});
Object.defineProperty(exports, "HOST_OS_PLATFORM", {
  enumerable: true,
  get: function () {
    return _elasticsearch.HOST_OS_PLATFORM;
  }
});
Object.defineProperty(exports, "HTTP_REQUEST_METHOD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.HTTP_REQUEST_METHOD;
  }
});
Object.defineProperty(exports, "HTTP_RESPONSE_STATUS_CODE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.HTTP_RESPONSE_STATUS_CODE;
  }
});
Object.defineProperty(exports, "KUBERNETES", {
  enumerable: true,
  get: function () {
    return _elasticsearch.KUBERNETES;
  }
});
Object.defineProperty(exports, "LABEL_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.LABEL_NAME;
  }
});
Object.defineProperty(exports, "LCP_FIELD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.LCP_FIELD;
  }
});
exports.LOGS_ONBOARDING_FEEDBACK_LINK = exports.LOGS_EXPLORER_FEEDBACK_LINK = void 0;
Object.defineProperty(exports, "METRIC_CGROUP_MEMORY_LIMIT_BYTES", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_CGROUP_MEMORY_LIMIT_BYTES;
  }
});
Object.defineProperty(exports, "METRIC_CGROUP_MEMORY_USAGE_BYTES", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_CGROUP_MEMORY_USAGE_BYTES;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_GC_COUNT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_GC_COUNT;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_GC_TIME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_GC_TIME;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_HEAP_MEMORY_COMMITTED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_HEAP_MEMORY_COMMITTED;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_HEAP_MEMORY_MAX", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_HEAP_MEMORY_MAX;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_HEAP_MEMORY_USED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_HEAP_MEMORY_USED;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_NON_HEAP_MEMORY_COMMITTED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_NON_HEAP_MEMORY_COMMITTED;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_NON_HEAP_MEMORY_MAX", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_NON_HEAP_MEMORY_MAX;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_NON_HEAP_MEMORY_USED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_NON_HEAP_MEMORY_USED;
  }
});
Object.defineProperty(exports, "METRIC_JAVA_THREAD_COUNT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_JAVA_THREAD_COUNT;
  }
});
Object.defineProperty(exports, "METRIC_PROCESS_CPU_PERCENT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_PROCESS_CPU_PERCENT;
  }
});
Object.defineProperty(exports, "METRIC_SYSTEM_CPU_PERCENT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_SYSTEM_CPU_PERCENT;
  }
});
Object.defineProperty(exports, "METRIC_SYSTEM_CPU_USAGE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_SYSTEM_CPU_USAGE;
  }
});
Object.defineProperty(exports, "METRIC_SYSTEM_FREE_MEMORY", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_SYSTEM_FREE_MEMORY;
  }
});
Object.defineProperty(exports, "METRIC_SYSTEM_MEMORY_USAGE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_SYSTEM_MEMORY_USAGE;
  }
});
Object.defineProperty(exports, "METRIC_SYSTEM_TOTAL_MEMORY", {
  enumerable: true,
  get: function () {
    return _elasticsearch.METRIC_SYSTEM_TOTAL_MEMORY;
  }
});
Object.defineProperty(exports, "NETWORK_TIMINGS_FIELDS", {
  enumerable: true,
  get: function () {
    return _synthetics.NETWORK_TIMINGS_FIELDS;
  }
});
Object.defineProperty(exports, "OBSERVER_HOSTNAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.OBSERVER_HOSTNAME;
  }
});
Object.defineProperty(exports, "OBSERVER_LISTENING", {
  enumerable: true,
  get: function () {
    return _elasticsearch.OBSERVER_LISTENING;
  }
});
Object.defineProperty(exports, "OBSERVER_VERSION_MAJOR", {
  enumerable: true,
  get: function () {
    return _elasticsearch.OBSERVER_VERSION_MAJOR;
  }
});
Object.defineProperty(exports, "ObservabilityTriggerId", {
  enumerable: true,
  get: function () {
    return _trigger_ids.ObservabilityTriggerId;
  }
});
Object.defineProperty(exports, "PARENT_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PARENT_ID;
  }
});
Object.defineProperty(exports, "POD_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.POD_NAME;
  }
});
Object.defineProperty(exports, "PROCESSOR_EVENT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROCESSOR_EVENT;
  }
});
Object.defineProperty(exports, "PROFILE_ALLOC_OBJECTS", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_ALLOC_OBJECTS;
  }
});
Object.defineProperty(exports, "PROFILE_ALLOC_SPACE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_ALLOC_SPACE;
  }
});
Object.defineProperty(exports, "PROFILE_CPU_NS", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_CPU_NS;
  }
});
Object.defineProperty(exports, "PROFILE_DURATION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_DURATION;
  }
});
Object.defineProperty(exports, "PROFILE_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_ID;
  }
});
Object.defineProperty(exports, "PROFILE_INUSE_OBJECTS", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_INUSE_OBJECTS;
  }
});
Object.defineProperty(exports, "PROFILE_INUSE_SPACE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_INUSE_SPACE;
  }
});
Object.defineProperty(exports, "PROFILE_SAMPLES_COUNT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_SAMPLES_COUNT;
  }
});
Object.defineProperty(exports, "PROFILE_STACK", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_STACK;
  }
});
Object.defineProperty(exports, "PROFILE_TOP_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_TOP_ID;
  }
});
Object.defineProperty(exports, "PROFILE_WALL_US", {
  enumerable: true,
  get: function () {
    return _elasticsearch.PROFILE_WALL_US;
  }
});
Object.defineProperty(exports, "SERVICE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE;
  }
});
Object.defineProperty(exports, "SERVICE_ENVIRONMENT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_ENVIRONMENT;
  }
});
Object.defineProperty(exports, "SERVICE_FRAMEWORK_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_FRAMEWORK_NAME;
  }
});
Object.defineProperty(exports, "SERVICE_FRAMEWORK_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_FRAMEWORK_VERSION;
  }
});
Object.defineProperty(exports, "SERVICE_LANGUAGE_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_LANGUAGE_NAME;
  }
});
Object.defineProperty(exports, "SERVICE_LANGUAGE_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_LANGUAGE_VERSION;
  }
});
Object.defineProperty(exports, "SERVICE_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_NAME;
  }
});
Object.defineProperty(exports, "SERVICE_NODE_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_NODE_NAME;
  }
});
Object.defineProperty(exports, "SERVICE_RUNTIME_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_RUNTIME_NAME;
  }
});
Object.defineProperty(exports, "SERVICE_RUNTIME_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_RUNTIME_VERSION;
  }
});
Object.defineProperty(exports, "SERVICE_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SERVICE_VERSION;
  }
});
exports.SLO_ALERTS_TABLE_ID = exports.SLO_ALERTS_TABLE_CONFIG_ID = void 0;
Object.defineProperty(exports, "SPAN_ACTION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_ACTION;
  }
});
Object.defineProperty(exports, "SPAN_DESTINATION_SERVICE_RESOURCE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_DESTINATION_SERVICE_RESOURCE;
  }
});
Object.defineProperty(exports, "SPAN_DESTINATION_SERVICE_RESPONSE_TIME_COUNT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_DESTINATION_SERVICE_RESPONSE_TIME_COUNT;
  }
});
Object.defineProperty(exports, "SPAN_DESTINATION_SERVICE_RESPONSE_TIME_SUM", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_DESTINATION_SERVICE_RESPONSE_TIME_SUM;
  }
});
Object.defineProperty(exports, "SPAN_DURATION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_DURATION;
  }
});
Object.defineProperty(exports, "SPAN_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_ID;
  }
});
Object.defineProperty(exports, "SPAN_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_NAME;
  }
});
Object.defineProperty(exports, "SPAN_SELF_TIME_SUM", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_SELF_TIME_SUM;
  }
});
Object.defineProperty(exports, "SPAN_SUBTYPE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_SUBTYPE;
  }
});
Object.defineProperty(exports, "SPAN_TYPE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.SPAN_TYPE;
  }
});
Object.defineProperty(exports, "SYNTHETICS_BLOCKED_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_BLOCKED_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_CONNECT_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_CONNECT_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_DNS_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_DNS_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_RECEIVE_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_RECEIVE_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_SEND_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_SEND_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_SSL_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_SSL_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_TOTAL_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_TOTAL_TIMINGS;
  }
});
Object.defineProperty(exports, "SYNTHETICS_WAIT_TIMINGS", {
  enumerable: true,
  get: function () {
    return _synthetics.SYNTHETICS_WAIT_TIMINGS;
  }
});
Object.defineProperty(exports, "TBT_FIELD", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TBT_FIELD;
  }
});
Object.defineProperty(exports, "TRACE_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRACE_ID;
  }
});
Object.defineProperty(exports, "TRANSACTION_DOM_INTERACTIVE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_DOM_INTERACTIVE;
  }
});
Object.defineProperty(exports, "TRANSACTION_DURATION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_DURATION;
  }
});
Object.defineProperty(exports, "TRANSACTION_DURATION_HISTOGRAM", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_DURATION_HISTOGRAM;
  }
});
Object.defineProperty(exports, "TRANSACTION_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_ID;
  }
});
Object.defineProperty(exports, "TRANSACTION_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_NAME;
  }
});
Object.defineProperty(exports, "TRANSACTION_PAGE_URL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_PAGE_URL;
  }
});
Object.defineProperty(exports, "TRANSACTION_RESULT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_RESULT;
  }
});
Object.defineProperty(exports, "TRANSACTION_ROOT", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_ROOT;
  }
});
Object.defineProperty(exports, "TRANSACTION_SAMPLED", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_SAMPLED;
  }
});
Object.defineProperty(exports, "TRANSACTION_TIME_TO_FIRST_BYTE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_TIME_TO_FIRST_BYTE;
  }
});
Object.defineProperty(exports, "TRANSACTION_TYPE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_TYPE;
  }
});
Object.defineProperty(exports, "TRANSACTION_URL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.TRANSACTION_URL;
  }
});
Object.defineProperty(exports, "URL_FULL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.URL_FULL;
  }
});
Object.defineProperty(exports, "USER_AGENT_DEVICE", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_DEVICE;
  }
});
Object.defineProperty(exports, "USER_AGENT_NAME", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_NAME;
  }
});
Object.defineProperty(exports, "USER_AGENT_ORIGINAL", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_ORIGINAL;
  }
});
Object.defineProperty(exports, "USER_AGENT_OS", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_OS;
  }
});
Object.defineProperty(exports, "USER_AGENT_OS_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_OS_VERSION;
  }
});
Object.defineProperty(exports, "USER_AGENT_VERSION", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_AGENT_VERSION;
  }
});
Object.defineProperty(exports, "USER_ID", {
  enumerable: true,
  get: function () {
    return _elasticsearch.USER_ID;
  }
});
exports.casesFeatureId = void 0;
Object.defineProperty(exports, "colorTransformer", {
  enumerable: true,
  get: function () {
    return _color_palette.colorTransformer;
  }
});
Object.defineProperty(exports, "getInspectResponse", {
  enumerable: true,
  get: function () {
    return _get_inspect_response.getInspectResponse;
  }
});
exports.sloFeatureId = exports.observabilityFeatureId = exports.observabilityAppId = void 0;
var _ruleDataUtils = require("@kbn/rule-data-utils");
var _elasticsearch = require("./field_names/elasticsearch");
var _synthetics = require("./field_names/synthetics");
var _color_palette = require("./color_palette");
var _trigger_ids = require("./trigger_ids");
var _get_inspect_response = require("./utils/get_inspect_response");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const observabilityFeatureId = exports.observabilityFeatureId = 'observability';
const observabilityAppId = exports.observabilityAppId = 'observability-overview';
const casesFeatureId = exports.casesFeatureId = 'observabilityCases';
const sloFeatureId = exports.sloFeatureId = 'slo';

// SLO alerts table in slo detail page
const SLO_ALERTS_TABLE_ID = exports.SLO_ALERTS_TABLE_ID = 'xpack.observability.slo.sloDetails.alertTable';
// Emebeddable SLO alerts table
const SLO_ALERTS_TABLE_CONFIG_ID = exports.SLO_ALERTS_TABLE_CONFIG_ID = `${_ruleDataUtils.AlertConsumers.SLO}-embeddable-alerts-table`;
const LOGS_ONBOARDING_FEEDBACK_LINK = exports.LOGS_ONBOARDING_FEEDBACK_LINK = 'https://ela.st/logs-onboarding-feedback';
const LOGS_EXPLORER_FEEDBACK_LINK = exports.LOGS_EXPLORER_FEEDBACK_LINK = 'https://ela.st/explorer-feedback';
