// @kbn/observability-logs-explorer-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AllDatasetsLocatorDefinition", {
  enumerable: true,
  get: function () {
    return _locators.AllDatasetsLocatorDefinition;
  }
});
Object.defineProperty(exports, "OBSERVABILITY_LOGS_EXPLORER_URL_STATE_KEY", {
  enumerable: true,
  get: function () {
    return _url_schema.OBSERVABILITY_LOGS_EXPLORER_URL_STATE_KEY;
  }
});
Object.defineProperty(exports, "SingleDatasetLocatorDefinition", {
  enumerable: true,
  get: function () {
    return _locators.SingleDatasetLocatorDefinition;
  }
});
Object.defineProperty(exports, "deepCompactObject", {
  enumerable: true,
  get: function () {
    return _deep_compact_object.deepCompactObject;
  }
});
Object.defineProperty(exports, "logsExplorerUrlSchemaV1", {
  enumerable: true,
  get: function () {
    return _url_schema.logsExplorerUrlSchemaV1;
  }
});
Object.defineProperty(exports, "logsExplorerUrlSchemaV2", {
  enumerable: true,
  get: function () {
    return _url_schema.logsExplorerUrlSchemaV2;
  }
});
var _locators = require("./locators");
var _url_schema = require("./url_schema");
var _deep_compact_object = require("./utils/deep_compact_object");
