// @kbn/controls-plugin/public/__stories__/decorators.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.decorators = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const bar = '#c5ced8';
const panel = '#ffff';
const background = '#FAFBFD';
const minHeight = 60;
const panelStyle = {
  height: 165,
  width: 400,
  background: panel
};
const kqlBarStyle = {
  background: bar,
  padding: 16,
  minHeight,
  fontStyle: 'italic'
};
const layout = OptionStory => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  style: {
    background
  },
  direction: "column"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: kqlBarStyle
}, "KQL Bar"), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(OptionStory, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  style: panelStyle
}))));
const decorators = exports.decorators = [layout];
