// @kbn/event-annotation-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ANNOTATIONS_LISTING_VIEW_ID", {
  enumerable: true,
  get: function () {
    return _constants.ANNOTATIONS_LISTING_VIEW_ID;
  }
});
Object.defineProperty(exports, "CONTENT_ID", {
  enumerable: true,
  get: function () {
    return _content_management.CONTENT_ID;
  }
});
Object.defineProperty(exports, "eventAnnotationGroup", {
  enumerable: true,
  get: function () {
    return _event_annotation_group.eventAnnotationGroup;
  }
});
Object.defineProperty(exports, "manualPointEventAnnotation", {
  enumerable: true,
  get: function () {
    return _manual_event_annotation.manualPointEventAnnotation;
  }
});
Object.defineProperty(exports, "manualRangeEventAnnotation", {
  enumerable: true,
  get: function () {
    return _manual_event_annotation.manualRangeEventAnnotation;
  }
});
Object.defineProperty(exports, "queryPointEventAnnotation", {
  enumerable: true,
  get: function () {
    return _query_point_event_annotation.queryPointEventAnnotation;
  }
});
var _manual_event_annotation = require("./manual_event_annotation");
var _query_point_event_annotation = require("./query_point_event_annotation");
var _event_annotation_group = require("./event_annotation_group");
var _content_management = require("./content_management");
var _constants = require("./constants");
