// @kbn/index-management-plugin/public/application/components/component_templates/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIM_COMPONENT_TEMPLATE_UPDATE = exports.UIM_COMPONENT_TEMPLATE_LIST_LOAD = exports.UIM_COMPONENT_TEMPLATE_DETAILS = exports.UIM_COMPONENT_TEMPLATE_DELETE_MANY = exports.UIM_COMPONENT_TEMPLATE_DELETE = exports.UIM_COMPONENT_TEMPLATE_CREATE = exports.MANAGED_BY_FLEET = exports.APP_CLUSTER_REQUIRED_PRIVILEGES = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// ui metric constants
const UIM_COMPONENT_TEMPLATE_LIST_LOAD = exports.UIM_COMPONENT_TEMPLATE_LIST_LOAD = 'component_template_list_load';
const UIM_COMPONENT_TEMPLATE_DELETE = exports.UIM_COMPONENT_TEMPLATE_DELETE = 'component_template_delete';
const UIM_COMPONENT_TEMPLATE_DELETE_MANY = exports.UIM_COMPONENT_TEMPLATE_DELETE_MANY = 'component_template_delete_many';
const UIM_COMPONENT_TEMPLATE_DETAILS = exports.UIM_COMPONENT_TEMPLATE_DETAILS = 'component_template_details';
const UIM_COMPONENT_TEMPLATE_CREATE = exports.UIM_COMPONENT_TEMPLATE_CREATE = 'component_template_create';
const UIM_COMPONENT_TEMPLATE_UPDATE = exports.UIM_COMPONENT_TEMPLATE_UPDATE = 'component_template_update';

// privileges
const APP_CLUSTER_REQUIRED_PRIVILEGES = exports.APP_CLUSTER_REQUIRED_PRIVILEGES = ['manage_index_templates'];
const MANAGED_BY_FLEET = exports.MANAGED_BY_FLEET = 'fleet';
