// @kbn/fleet-plugin/server/types/rest_spec/debug.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FetchSavedObjectsRequestSchema = exports.FetchSavedObjectNamesRequestSchema = exports.FetchIndexRequestSchema = void 0;
var _configSchema = require("@kbn/config-schema");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FetchIndexRequestSchema = exports.FetchIndexRequestSchema = {
  body: _configSchema.schema.object({
    index: _configSchema.schema.string()
  })
};
const FetchSavedObjectsRequestSchema = exports.FetchSavedObjectsRequestSchema = {
  body: _configSchema.schema.object({
    type: _configSchema.schema.string(),
    name: _configSchema.schema.string()
  })
};
const FetchSavedObjectNamesRequestSchema = exports.FetchSavedObjectNamesRequestSchema = {
  body: _configSchema.schema.object({
    type: _configSchema.schema.string()
  })
};
