// @kbn/security-solution-plugin/public/common/components/search_bar/selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "endSelector", {
  enumerable: true,
  get: function () {
    return _selectors.endSelector;
  }
});
exports.filterQuerySelector = void 0;
Object.defineProperty(exports, "fromStrSelector", {
  enumerable: true,
  get: function () {
    return _selectors.fromStrSelector;
  }
});
exports.getSavedQuery = exports.getFilterQuery = void 0;
Object.defineProperty(exports, "isLoadingSelector", {
  enumerable: true,
  get: function () {
    return _selectors.isLoadingSelector;
  }
});
Object.defineProperty(exports, "kindSelector", {
  enumerable: true,
  get: function () {
    return _selectors.kindSelector;
  }
});
Object.defineProperty(exports, "queriesSelector", {
  enumerable: true,
  get: function () {
    return _selectors.queriesSelector;
  }
});
exports.savedQuerySelector = void 0;
Object.defineProperty(exports, "startSelector", {
  enumerable: true,
  get: function () {
    return _selectors.startSelector;
  }
});
Object.defineProperty(exports, "toStrSelector", {
  enumerable: true,
  get: function () {
    return _selectors.toStrSelector;
  }
});
var _reselect = require("reselect");
var _selectors = require("../super_date_picker/selectors");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getFilterQuery = inputState => inputState.query;
exports.getFilterQuery = getFilterQuery;
const getSavedQuery = inputState => inputState.savedQuery;
exports.getSavedQuery = getSavedQuery;
const filterQuerySelector = () => (0, _reselect.createSelector)(getFilterQuery, filterQuery => filterQuery);
exports.filterQuerySelector = filterQuerySelector;
const savedQuerySelector = () => (0, _reselect.createSelector)(getSavedQuery, savedQuery => savedQuery);
exports.savedQuerySelector = savedQuerySelector;
