// @kbn/content-management-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ContentClient", {
  enumerable: true,
  get: function () {
    return _content_client.ContentClient;
  }
});
Object.defineProperty(exports, "ContentClientProvider", {
  enumerable: true,
  get: function () {
    return _content_client.ContentClientProvider;
  }
});
exports.plugin = plugin;
Object.defineProperty(exports, "useContentClient", {
  enumerable: true,
  get: function () {
    return _content_client.useContentClient;
  }
});
Object.defineProperty(exports, "useCreateContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client.useCreateContentMutation;
  }
});
Object.defineProperty(exports, "useDeleteContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client.useDeleteContentMutation;
  }
});
Object.defineProperty(exports, "useGetContentQuery", {
  enumerable: true,
  get: function () {
    return _content_client.useGetContentQuery;
  }
});
Object.defineProperty(exports, "useSearchContentQuery", {
  enumerable: true,
  get: function () {
    return _content_client.useSearchContentQuery;
  }
});
Object.defineProperty(exports, "useUpdateContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client.useUpdateContentMutation;
  }
});
var _plugin = require("./plugin");
var _content_client = require("./content_client");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function plugin() {
  return new _plugin.ContentManagementPlugin();
}
