// @kbn/monitoring-plugin/public/application/pages/kibana/kibana_template.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KibanaTemplate = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _page_template = require("../page_template");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const KibanaTemplate = ({
  ...props
}) => {
  const tabs = [{
    id: 'overview',
    label: _i18n.i18n.translate('xpack.monitoring.kibanaNavigation.overviewLinkText', {
      defaultMessage: 'Overview'
    }),
    route: '/kibana',
    testSubj: 'kibanaOverviewPage'
  }, {
    id: 'instances',
    label: _i18n.i18n.translate('xpack.monitoring.kibanaNavigation.instancesLinkText', {
      defaultMessage: 'Instances'
    }),
    route: '/kibana/instances',
    testSubj: 'kibanaInstancesPage'
  }];
  return /*#__PURE__*/_react.default.createElement(_page_template.PageTemplate, (0, _extends2.default)({}, props, {
    tabs: tabs,
    product: "kibana"
  }));
};
exports.KibanaTemplate = KibanaTemplate;
