// @kbn/search-playground/public/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  ChatFormFields: true
};
exports.ChatFormFields = void 0;
var _types = require("../common/types");
Object.keys(_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _types[key];
    }
  });
});
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
// eslint-disable-next-line @typescript-eslint/no-empty-interface
let ChatFormFields = exports.ChatFormFields = /*#__PURE__*/function (ChatFormFields) {
  ChatFormFields["question"] = "question";
  ChatFormFields["citations"] = "citations";
  ChatFormFields["prompt"] = "prompt";
  ChatFormFields["indices"] = "indices";
  ChatFormFields["elasticsearchQuery"] = "elasticsearch_query";
  ChatFormFields["summarizationModel"] = "summarization_model";
  ChatFormFields["sourceFields"] = "source_fields";
  ChatFormFields["docSize"] = "doc_size";
  ChatFormFields["queryFields"] = "query_fields";
  return ChatFormFields;
}({});
