// @kbn/transform-plugin/public/app/sections/transform_management/components/action_reauthorize/reauthorize_action_modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReauthorizeActionModal = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ReauthorizeActionModal = ({
  closeModal,
  items,
  reauthorizeAndCloseModal
}) => {
  const isBulkAction = items.length > 1;
  const bulkReauthorizeModalTitle = _i18n.i18n.translate('xpack.transform.transformList.bulkReauthorizeModalTitle', {
    defaultMessage: 'Reauthorize {count} {count, plural, one {transform} other {transforms}}?',
    values: {
      count: items && items.length
    }
  });
  const reauthorizeModalTitle = _i18n.i18n.translate('xpack.transform.transformList.reauthorizeModalTitle', {
    defaultMessage: 'Reauthorize {transformId}?',
    values: {
      transformId: items[0] && items[0].config.id
    }
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiConfirmModal, {
    "data-test-subj": "transformReauthorizeModal",
    title: isBulkAction === true ? bulkReauthorizeModalTitle : reauthorizeModalTitle,
    onCancel: closeModal,
    onConfirm: reauthorizeAndCloseModal,
    cancelButtonText: _i18n.i18n.translate('xpack.transform.transformList.reauthorizeModalCancelButton', {
      defaultMessage: 'Cancel'
    }),
    confirmButtonText: _i18n.i18n.translate('xpack.transform.transformList.reauthorizeModalConfirmButton', {
      defaultMessage: 'Reauthorize'
    }),
    defaultFocusedButton: _eui.EUI_MODAL_CONFIRM_BUTTON,
    buttonColor: "primary"
  }, /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('xpack.transform.transformList.reauthorizeModalBody', {
    defaultMessage: 'Your current roles are used to update and start the transform. Starting a transform increases search and indexing load in your cluster. If excessive load is experienced, stop the transform.'
  })));
};
exports.ReauthorizeActionModal = ReauthorizeActionModal;
