// @kbn/security-solution-plugin/public/detection_engine/rule_creation_ui/components/step_about_rule/default_value.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.threatDefault = exports.stepAboutDefaultValue = void 0;
var _constants = require("../../../../../common/constants");
var _helpers = require("../../../../detections/pages/detection_engine/rules/helpers");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const threatDefault = exports.threatDefault = [{
  framework: 'MITRE ATT&CK',
  tactic: {
    id: 'none',
    name: 'none',
    reference: 'none'
  },
  technique: []
}];
const stepAboutDefaultValue = exports.stepAboutDefaultValue = {
  author: [],
  name: '',
  description: '',
  isAssociatedToEndpointList: false,
  isBuildingBlock: false,
  severity: {
    value: 'low',
    mapping: (0, _helpers.fillEmptySeverityMappings)([]),
    isMappingChecked: false
  },
  riskScore: {
    value: 21,
    mapping: [],
    isMappingChecked: false
  },
  references: [''],
  falsePositives: [''],
  investigationFields: [],
  license: '',
  ruleNameOverride: '',
  tags: [],
  timestampOverride: '',
  threat: threatDefault,
  note: '',
  maxSignals: _constants.DEFAULT_MAX_SIGNALS,
  setup: ''
};
