// @kbn/upgrade-assistant-plugin/server/lib/reindexing/error_symbols.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ReindexTaskFailed = exports.ReindexTaskCannotBeDeleted = exports.ReindexIsNotInQueue = exports.ReindexCannotBeCancelled = exports.ReindexAlreadyInProgress = exports.MultipleReindexJobsFound = exports.IndexNotFound = exports.CannotCreateIndex = exports.AccessForbidden = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AccessForbidden = exports.AccessForbidden = Symbol('AccessForbidden');
const IndexNotFound = exports.IndexNotFound = Symbol('IndexNotFound');
const CannotCreateIndex = exports.CannotCreateIndex = Symbol('CannotCreateIndex');
const ReindexTaskFailed = exports.ReindexTaskFailed = Symbol('ReindexTaskFailed');
const ReindexTaskCannotBeDeleted = exports.ReindexTaskCannotBeDeleted = Symbol('ReindexTaskCannotBeDeleted');
const ReindexAlreadyInProgress = exports.ReindexAlreadyInProgress = Symbol('ReindexAlreadyInProgress');
const ReindexIsNotInQueue = exports.ReindexIsNotInQueue = Symbol('ReindexIsNotInQueue');
const ReindexCannotBeCancelled = exports.ReindexCannotBeCancelled = Symbol('ReindexCannotBeCancelled');
const MultipleReindexJobsFound = exports.MultipleReindexJobsFound = Symbol('MultipleReindexJobsFound');
