// @kbn/data-visualizer-plugin/public/application/common/components/stats_table/components/field_count_stats/total_fields_count.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TotalFieldsCount = void 0;
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireDefault(require("react"));
var _styles = require("./styles");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TotalFieldsCount = ({
  fieldsCountStats
}) => {
  if (!fieldsCountStats || fieldsCountStats.visibleFieldsCount === undefined || fieldsCountStats.totalFieldsCount === undefined) return null;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    css: _styles.dvFieldCountItemCss,
    "data-test-subj": "dataVisualizerFieldsSummary",
    gutterSize: "s",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("h5", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.searchPanel.allFieldsLabel",
    defaultMessage: "All fields"
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiNotificationBadge, {
    color: "subdued",
    size: "m",
    "data-test-subj": "dataVisualizerVisibleFieldsCount"
  }, /*#__PURE__*/_react.default.createElement("strong", null, fieldsCountStats.visibleFieldsCount))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    color: "subdued",
    size: "s",
    "data-test-subj": "dataVisualizerTotalFieldsCount"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.searchPanel.ofFieldsTotal",
    defaultMessage: "of {totalCount} total",
    values: {
      totalCount: fieldsCountStats.totalFieldsCount
    }
  }))));
};
exports.TotalFieldsCount = TotalFieldsCount;
