// @kbn/ml-plugin/public/embeddables/job_creation/aiops/show_flyout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.showPatternAnalysisToADJobFlyout = showPatternAnalysisToADJobFlyout;
var _react = _interopRequireDefault(require("react"));
var _create_flyout = require("../common/create_flyout");
var _flyout = require("./flyout");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

async function showPatternAnalysisToADJobFlyout(dataView, field, query, timeRange, coreStart, share, data, dashboardService, lens) {
  const Comp = ({
    onClose
  }) => /*#__PURE__*/_react.default.createElement(_flyout.CreateCategorizationJobFlyout, {
    dataView: dataView,
    field: field,
    query: query,
    timeRange: timeRange,
    onClose: onClose
  });
  return (0, _create_flyout.createFlyout)(Comp, coreStart, share, data, dashboardService, lens);
}
