// @kbn/maps-plugin/public/connected_components/edit_layer_panel/flyout_footer/flyout_footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutFooter = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _panel_strings = require("../../panel_strings");
var _remove_layer_confirm_modal = require("../../../components/remove_layer_confirm_modal");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class FlyoutFooter extends _react.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      showRemoveModal: false
    });
    (0, _defineProperty2.default)(this, "_showRemoveModal", () => {
      this.setState({
        showRemoveModal: true
      });
    });
  }
  render() {
    const removeModal = this.props.selectedLayer && this.state.showRemoveModal ? /*#__PURE__*/_react.default.createElement(_remove_layer_confirm_modal.RemoveLayerConfirmModal, {
      layer: this.props.selectedLayer,
      onCancel: () => {
        this.setState({
          showRemoveModal: false
        });
      },
      onConfirm: () => {
        this.setState({
          showRemoveModal: false
        });
        this.props.removeLayer();
      }
    }) : null;
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, removeModal, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      responsive: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
      onClick: this.props.cancelLayerPanel,
      flush: "left",
      "data-test-subj": "layerPanelCancelButton"
    }, this.props.hasStateChanged ? _panel_strings.panelStrings.discardChanges : _panel_strings.panelStrings.close)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
      color: "danger",
      onClick: this._showRemoveModal,
      flush: "right",
      "data-test-subj": "mapRemoveLayerButton"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.maps.layerPanel.footer.removeLayerButtonLabel",
      defaultMessage: "Remove layer"
    }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      disabled: !this.props.hasStateChanged,
      iconType: "check",
      onClick: this.props.saveLayerEdits,
      fill: true
    }, _panel_strings.panelStrings.keepChanges))));
  }
}
exports.FlyoutFooter = FlyoutFooter;
