// @kbn/monitoring-plugin/public/components/elasticsearch/node/advanced.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdvancedNode = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _node_detail_status = require("../node_detail_status");
var _chart = require("../../chart");
var _i18nReact = require("@kbn/i18n-react");
var _callout = require("../../../alerts/callout");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AdvancedNode = ({
  nodeSummary,
  metrics,
  alerts,
  ...props
}) => {
  const metricsToShow = [metrics.node_gc, metrics.node_gc_time, metrics.node_jvm_mem, metrics.node_cpu_utilization, metrics.node_index_3, metrics.node_index_4, metrics.node_index_time, metrics.node_request_total, metrics.node_index_threads, metrics.node_read_threads, metrics.node_cgroup_cpu, metrics.node_cgroup_stats, metrics.node_latency];
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPage, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPageBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiScreenReaderOnly, null, /*#__PURE__*/_react.default.createElement("h1", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.monitoring.elasticsearch.nodes.advanced.heading",
    defaultMessage: "Elasticsearch node advanced"
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, null, /*#__PURE__*/_react.default.createElement(_node_detail_status.NodeDetailStatus, {
    stats: nodeSummary,
    alerts: alerts
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_callout.AlertsCallout, {
    alerts: alerts
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiPageSection, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGrid, {
    columns: 2,
    gutterSize: "s"
  }, metricsToShow.map((metric, index) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    key: index
  }, /*#__PURE__*/_react.default.createElement(_chart.MonitoringTimeseriesContainer, (0, _extends2.default)({
    series: metric
  }, props)), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null)))))));
};
exports.AdvancedNode = AdvancedNode;
