// @kbn/slo-plugin/public/hooks/use_reset_slo.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useResetSlo = useResetSlo;
var _i18n = require("@kbn/i18n");
var _reactQuery = require("@tanstack/react-query");
var _kibana_react = require("../utils/kibana_react");
var _query_key_factory = require("./query_key_factory");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function useResetSlo() {
  const {
    http,
    notifications: {
      toasts
    }
  } = (0, _kibana_react.useKibana)().services;
  const queryClient = (0, _reactQuery.useQueryClient)();
  return (0, _reactQuery.useMutation)(['resetSlo'], ({
    id,
    name
  }) => {
    try {
      return http.post(`/api/observability/slos/${id}/_reset`);
    } catch (error) {
      return Promise.reject(_i18n.i18n.translate('xpack.slo.slo.reset.errorMessage', {
        defaultMessage: 'Failed to reset {name} (id: {id}), something went wrong: {error}',
        values: {
          error: String(error),
          name,
          id
        }
      }));
    }
  }, {
    onError: (error, {
      name,
      id
    }) => {
      var _error$body$message, _error$body;
      toasts.addError(new Error((_error$body$message = (_error$body = error.body) === null || _error$body === void 0 ? void 0 : _error$body.message) !== null && _error$body$message !== void 0 ? _error$body$message : error.message), {
        title: _i18n.i18n.translate('xpack.slo.slo.reset.errorNotification', {
          defaultMessage: 'Failed to reset {name} (id: {id})',
          values: {
            name,
            id
          }
        })
      });
    },
    onSuccess: (_data, {
      name,
      id
    }) => {
      queryClient.invalidateQueries({
        queryKey: _query_key_factory.sloKeys.lists(),
        exact: false
      });
      queryClient.invalidateQueries({
        queryKey: _query_key_factory.sloKeys.historicalSummaries(),
        exact: false
      });
      queryClient.invalidateQueries({
        queryKey: _query_key_factory.sloKeys.details(),
        exact: false
      });
      toasts.addSuccess(_i18n.i18n.translate('xpack.slo.slo.reset.successNotification', {
        defaultMessage: '{name} reset successfully',
        values: {
          name
        }
      }));
    }
  });
}
