// @kbn/security-solution-plugin/public/explore/users/store/model.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UsersType = exports.UsersTableType = exports.UsersDetailsTableType = exports.UserAssetTableType = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let UsersType = exports.UsersType = /*#__PURE__*/function (UsersType) {
  UsersType["page"] = "page";
  UsersType["details"] = "details";
  UsersType["flyout"] = "flyout";
  return UsersType;
}({});
let UsersTableType = exports.UsersTableType = /*#__PURE__*/function (UsersTableType) {
  UsersTableType["allUsers"] = "allUsers";
  UsersTableType["authentications"] = "authentications";
  UsersTableType["anomalies"] = "anomalies";
  UsersTableType["risk"] = "userRisk";
  UsersTableType["events"] = "events";
  return UsersTableType;
}({});
let UsersDetailsTableType = exports.UsersDetailsTableType = /*#__PURE__*/function (UsersDetailsTableType) {
  UsersDetailsTableType["authentications"] = "authentications";
  UsersDetailsTableType["anomalies"] = "anomalies";
  UsersDetailsTableType["risk"] = "userRisk";
  UsersDetailsTableType["events"] = "events";
  return UsersDetailsTableType;
}({});
let UserAssetTableType = exports.UserAssetTableType = /*#__PURE__*/function (UserAssetTableType) {
  UserAssetTableType["assetEntra"] = "userAssetEntra";
  UserAssetTableType["assetOkta"] = "userAssetOkta";
  return UserAssetTableType;
}({});
