// @kbn/observability-ai-assistant-plugin/common/functions/visualize_esql.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VisualizeESQLUserIntention = exports.VISUALIZE_ESQL_USER_INTENTIONS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let VisualizeESQLUserIntention = exports.VisualizeESQLUserIntention = /*#__PURE__*/function (VisualizeESQLUserIntention) {
  VisualizeESQLUserIntention["generateQueryOnly"] = "generateQueryOnly";
  VisualizeESQLUserIntention["executeAndReturnResults"] = "executeAndReturnResults";
  VisualizeESQLUserIntention["visualizeAuto"] = "visualizeAuto";
  VisualizeESQLUserIntention["visualizeXy"] = "visualizeXy";
  VisualizeESQLUserIntention["visualizeBar"] = "visualizeBar";
  VisualizeESQLUserIntention["visualizeLine"] = "visualizeLine";
  VisualizeESQLUserIntention["visualizeArea"] = "visualizeArea";
  VisualizeESQLUserIntention["visualizeTable"] = "visualizeTable";
  VisualizeESQLUserIntention["visualizeDonut"] = "visualizeDonut";
  VisualizeESQLUserIntention["visualizeTreemap"] = "visualizeTreemap";
  VisualizeESQLUserIntention["visualizeHeatmap"] = "visualizeHeatmap";
  VisualizeESQLUserIntention["visualizeTagcloud"] = "visualizeTagcloud";
  VisualizeESQLUserIntention["visualizeWaffle"] = "visualizeWaffle";
  return VisualizeESQLUserIntention;
}({});
const VISUALIZE_ESQL_USER_INTENTIONS = exports.VISUALIZE_ESQL_USER_INTENTIONS = Object.values(VisualizeESQLUserIntention);
