// @kbn/slo-plugin/public/pages/slos_welcome/slos_welcome.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SlosWelcomePage = SlosWelcomePage;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireWildcard(require("react"));
var _paths = require("../../../common/locators/paths");
var _header_menu = require("../../components/header_menu/header_menu");
var _slo_outdated_callout = require("../../components/slo/slo_outdated_callout");
var _slo_permissions_callout = require("../../components/slo/slo_permissions_callout");
var _use_fetch_slo_list = require("../../hooks/use_fetch_slo_list");
var _use_license = require("../../hooks/use_license");
var _use_permissions = require("../../hooks/use_permissions");
var _use_plugin_context = require("../../hooks/use_plugin_context");
var _kibana_react = require("../../utils/kibana_react");
var _illustration = _interopRequireDefault(require("./assets/illustration.svg"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function SlosWelcomePage() {
  const {
    application: {
      navigateToUrl
    },
    http: {
      basePath
    }
  } = (0, _kibana_react.useKibana)().services;
  const {
    ObservabilityPageTemplate
  } = (0, _use_plugin_context.usePluginContext)();
  const {
    data: permissions
  } = (0, _use_permissions.usePermissions)();
  const {
    hasAtLeast
  } = (0, _use_license.useLicense)();
  const hasRightLicense = hasAtLeast('platinum');
  const {
    data: sloList
  } = (0, _use_fetch_slo_list.useFetchSloList)();
  const {
    total
  } = sloList !== null && sloList !== void 0 ? sloList : {
    total: 0
  };
  const hasSlosAndPermissions = total > 0 && hasRightLicense && (permissions === null || permissions === void 0 ? void 0 : permissions.hasAllReadRequested) === true;
  const handleClickCreateSlo = () => {
    navigateToUrl(basePath.prepend(_paths.paths.sloCreate));
  };
  (0, _react.useEffect)(() => {
    if (hasSlosAndPermissions) {
      navigateToUrl(basePath.prepend(_paths.paths.slos));
    }
  }, [basePath, navigateToUrl, hasSlosAndPermissions]);
  return /*#__PURE__*/_react.default.createElement(ObservabilityPageTemplate, {
    "data-test-subj": "slosPageWelcomePrompt"
  }, /*#__PURE__*/_react.default.createElement(_header_menu.HeaderMenu, null), /*#__PURE__*/_react.default.createElement(_slo_outdated_callout.SloOutdatedCallout, null), /*#__PURE__*/_react.default.createElement(_slo_permissions_callout.SloPermissionsCallout, null), /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate.EmptyPrompt, {
    title: /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "l"
    }, /*#__PURE__*/_react.default.createElement("h1", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.title', {
      defaultMessage: 'Track and deliver on your SLOs'
    }))),
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiImage, {
      size: "fullWidth",
      src: _illustration.default,
      alt: ""
    }),
    color: "transparent",
    layout: "horizontal",
    hasBorder: false,
    body: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.messageParagraph1', {
      defaultMessage: 'Measure key metrics important to the business, such as service-level indicators and service-level objectives (SLIs/SLOs) to deliver on SLAs.'
    })), /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.messageParagraph2', {
      defaultMessage: 'Easily report the uptime and reliability of your services to stakeholders with real-time insights.'
    })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    })),
    actions: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, hasRightLicense ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      direction: "column"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "xxs"
    }, /*#__PURE__*/_react.default.createElement("span", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.getStartedMessage', {
      defaultMessage: 'To get started, create your first SLO.'
    })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement("span", null, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      "data-test-subj": "o11ySloListWelcomePromptCreateSloButton",
      fill: true,
      color: "primary",
      onClick: handleClickCreateSlo,
      disabled: !(permissions !== null && permissions !== void 0 && permissions.hasAllWriteRequested)
    }, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.buttonLabel', {
      defaultMessage: 'Create SLO'
    }))))) : /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      direction: "column"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "xxs"
    }, /*#__PURE__*/_react.default.createElement("span", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.needLicenseMessage', {
      defaultMessage: 'You need an Elastic Cloud subscription or Platinum license to use SLOs.'
    })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      direction: "row"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      fill: true,
      href: "https://www.elastic.co/cloud/elasticsearch-service/signup",
      target: "_blank",
      "data-test-subj": "slosPageWelcomePromptSignupForCloudButton"
    }, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.signupForCloud', {
      defaultMessage: 'Sign up for Elastic Cloud'
    }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      href: "https://www.elastic.co/subscriptions",
      target: "_blank",
      "data-test-subj": "slosPageWelcomePromptSignupForLicenseButton"
    }, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.signupForLicense', {
      defaultMessage: 'Sign up for license'
    }))))))),
    footer: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "xxs"
    }, /*#__PURE__*/_react.default.createElement("span", null, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.learnMore', {
      defaultMessage: 'Want to learn more?'
    }))), "\xA0", /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
      "data-test-subj": "o11ySloListWelcomePromptReadTheDocsLink",
      href: "https://www.elastic.co/guide/en/observability/current/slo.html",
      target: "_blank"
    }, _i18n.i18n.translate('xpack.slo.sloList.welcomePrompt.learnMoreLink', {
      defaultMessage: 'Read the docs'
    })))
  }));
}
