// @kbn/security-solution-plugin/common/api/entity_analytics/risk_engine/engine_status_route.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RiskEngineStatusResponse = exports.RiskEngineStatusEnum = exports.RiskEngineStatus = void 0;
var _zod = require("zod");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Risk Scoring API
 *   version: 1
 */

const RiskEngineStatus = exports.RiskEngineStatus = _zod.z.enum(['NOT_INSTALLED', 'DISABLED', 'ENABLED']);
const RiskEngineStatusEnum = exports.RiskEngineStatusEnum = RiskEngineStatus.enum;
const RiskEngineStatusResponse = exports.RiskEngineStatusResponse = _zod.z.object({
  legacy_risk_engine_status: RiskEngineStatus,
  risk_engine_status: RiskEngineStatus,
  /**
   * Indicates whether the maximum amount of risk engines has been reached
   */
  is_max_amount_of_risk_engines_reached: _zod.z.boolean()
});
