// @kbn/search-connectors-plugin/common/doc_links.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.docLinks = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class ESDocLinks {
  constructor() {
    (0, _defineProperty2.default)(this, "connectors", '');
    (0, _defineProperty2.default)(this, "connectorsAzureBlobStorage", '');
    (0, _defineProperty2.default)(this, "connectorsBox", '');
    (0, _defineProperty2.default)(this, "connectorsClients", '');
    (0, _defineProperty2.default)(this, "connectorsConfluence", '');
    (0, _defineProperty2.default)(this, "connectorsDropbox", '');
    (0, _defineProperty2.default)(this, "connectorsGithub", '');
    (0, _defineProperty2.default)(this, "connectorsGmail", '');
    (0, _defineProperty2.default)(this, "connectorsGoogleCloudStorage", '');
    (0, _defineProperty2.default)(this, "connectorsGoogleDrive", '');
    (0, _defineProperty2.default)(this, "connectorsJira", '');
    (0, _defineProperty2.default)(this, "connectorsMicrosoftSQL", '');
    (0, _defineProperty2.default)(this, "connectorsMongoDB", '');
    (0, _defineProperty2.default)(this, "connectorsMySQL", '');
    (0, _defineProperty2.default)(this, "connectorsNative", '');
    (0, _defineProperty2.default)(this, "connectorsNetworkDrive", '');
    (0, _defineProperty2.default)(this, "connectorsNotion", '');
    (0, _defineProperty2.default)(this, "connectorsOneDrive", '');
    (0, _defineProperty2.default)(this, "connectorsOracle", '');
    (0, _defineProperty2.default)(this, "connectorsOutlook", '');
    (0, _defineProperty2.default)(this, "connectorsPostgreSQL", '');
    (0, _defineProperty2.default)(this, "connectorsRedis", '');
    (0, _defineProperty2.default)(this, "connectorsS3", '');
    (0, _defineProperty2.default)(this, "connectorsSalesforce", '');
    (0, _defineProperty2.default)(this, "connectorsServiceNow", '');
    (0, _defineProperty2.default)(this, "connectorsSharepoint", '');
    (0, _defineProperty2.default)(this, "connectorsSharepointOnline", '');
    (0, _defineProperty2.default)(this, "connectorsSlack", '');
    (0, _defineProperty2.default)(this, "connectorsTeams", '');
    (0, _defineProperty2.default)(this, "connectorsZoom", '');
  }
  setDocLinks(docLinks) {
    this.connectors = docLinks.enterpriseSearch.connectors;
    this.connectorsAzureBlobStorage = docLinks.enterpriseSearch.connectorsAzureBlobStorage;
    this.connectorsBox = docLinks.enterpriseSearch.connectorsBox;
    this.connectorsConfluence = docLinks.enterpriseSearch.connectorsConfluence;
    this.connectorsClients = docLinks.enterpriseSearch.connectorsClients;
    this.connectorsDropbox = docLinks.enterpriseSearch.connectorsDropbox;
    this.connectorsGithub = docLinks.enterpriseSearch.connectorsGithub;
    this.connectorsGoogleCloudStorage = docLinks.enterpriseSearch.connectorsGoogleCloudStorage;
    this.connectorsGoogleDrive = docLinks.enterpriseSearch.connectorsGoogleDrive;
    this.connectorsGmail = docLinks.enterpriseSearch.connectorsGmail;
    this.connectorsJira = docLinks.enterpriseSearch.connectorsJira;
    this.connectorsMicrosoftSQL = docLinks.enterpriseSearch.connectorsMicrosoftSQL;
    this.connectorsMongoDB = docLinks.enterpriseSearch.connectorsMongoDB;
    this.connectorsMySQL = docLinks.enterpriseSearch.connectorsMySQL;
    this.connectorsNative = docLinks.enterpriseSearch.connectorsNative;
    this.connectorsNetworkDrive = docLinks.enterpriseSearch.connectorsNetworkDrive;
    this.connectorsNotion = docLinks.enterpriseSearch.connectorsNotion;
    this.connectorsOneDrive = docLinks.enterpriseSearch.connectorsOneDrive;
    this.connectorsOracle = docLinks.enterpriseSearch.connectorsOracle;
    this.connectorsOutlook = docLinks.enterpriseSearch.connectorsOutlook;
    this.connectorsPostgreSQL = docLinks.enterpriseSearch.connectorsPostgreSQL;
    this.connectorsRedis = docLinks.enterpriseSearch.connectorsRedis;
    this.connectorsS3 = docLinks.enterpriseSearch.connectorsS3;
    this.connectorsSalesforce = docLinks.enterpriseSearch.connectorsSalesforce;
    this.connectorsServiceNow = docLinks.enterpriseSearch.connectorsServiceNow;
    this.connectorsSharepoint = docLinks.enterpriseSearch.connectorsSharepoint;
    this.connectorsSharepointOnline = docLinks.enterpriseSearch.connectorsSharepointOnline;
    this.connectorsSlack = docLinks.enterpriseSearch.connectorsSlack;
    this.connectorsTeams = docLinks.enterpriseSearch.connectorsTeams;
    this.connectorsZoom = docLinks.enterpriseSearch.connectorsZoom;
  }
}
const docLinks = exports.docLinks = new ESDocLinks();
