// @kbn/observability-plugin/public/pages/overview/components/sections/ux/core_web_vitals/service_name.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServiceName = ServiceName;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SERVICE_LABEL = _i18n.i18n.translate('xpack.observability.ux.coreWebVitals.service', {
  defaultMessage: 'Service'
});
const SERVICE_LABEL_HELP = _i18n.i18n.translate('xpack.observability.ux.service.help', {
  defaultMessage: 'The RUM service with the most traffic is selected'
});
function ServiceName({
  name
}) {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, SERVICE_LABEL, /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    color: "text",
    "aria-label": SERVICE_LABEL_HELP,
    type: "questionInCircle",
    content: SERVICE_LABEL_HELP
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("h3", null, name)));
}
