// @kbn/ml-plugin/public/application/management/jobs_list/components/access_denied_page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AccessDeniedPage = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AccessDeniedPage = () => /*#__PURE__*/_react.default.createElement(_eui.EuiPageTemplate.EmptyPrompt, {
  "data-test-subj": "mlPageAccessDenied",
  color: 'danger',
  iconType: "warning",
  title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.management.jobsList.noPermissionToAccessLabel",
    defaultMessage: "Access denied"
  })),
  body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.management.jobsList.noGrantedPrivilegesDescription",
    defaultMessage: "You don\u2019t have permission to manage Machine Learning jobs. Access to the plugin requires the Machine Learning feature to be visible in this space."
  }))
});
exports.AccessDeniedPage = AccessDeniedPage;
