// @kbn/infra-plugin/public/components/logging/log_analysis_setup/initial_configuration_step/index_setup_dataset_filter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IndexSetupDatasetFilter = void 0;
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireWildcard(require("react"));
var _use_visibility_state = require("../../../../utils/use_visibility_state");
var _quality_warning_notices = require("../../log_analysis_job_status/quality_warning_notices");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const IndexSetupDatasetFilter = ({
  availableDatasets,
  datasetFilter,
  isDisabled,
  onChangeDatasetFilter
}) => {
  const {
    isVisible,
    hide,
    show
  } = (0, _use_visibility_state.useVisibilityState)(false);
  const changeDatasetFilter = (0, _react.useCallback)(options => {
    const selectedDatasets = options.filter(({
      checked
    }) => checked === 'on').map(({
      label
    }) => label);
    onChangeDatasetFilter(selectedDatasets.length === 0 ? {
      type: 'includeAll'
    } : {
      type: 'includeSome',
      datasets: selectedDatasets
    });
  }, [onChangeDatasetFilter]);
  const selectableOptions = (0, _react.useMemo)(() => availableDatasets.map(({
    dataset,
    warnings
  }) => ({
    label: dataset,
    append: warnings.length > 0 ? /*#__PURE__*/_react.default.createElement(DatasetWarningMarker, {
      warnings: warnings
    }) : null,
    checked: datasetFilter.type === 'includeSome' && datasetFilter.datasets.includes(dataset) ? 'on' : undefined
  })), [availableDatasets, datasetFilter]);
  const datasetFilterButton = /*#__PURE__*/_react.default.createElement(_eui.EuiFilterButton, {
    disabled: isDisabled,
    isSelected: isVisible,
    onClick: show,
    iconType: "arrowDown"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.infra.analysisSetup.indexDatasetFilterIncludeAllButtonLabel",
    defaultMessage: "{includeType, select, includeAll {All datasets} includeSome {{includedDatasetCount, plural, one {# dataset} other {# datasets}}} other {No dataset available}}",
    values: {
      includeType: datasetFilter.type,
      includedDatasetCount: datasetFilter.type === 'includeSome' ? datasetFilter.datasets.length : 0
    }
  }));
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFilterGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
    button: datasetFilterButton,
    closePopover: hide,
    isOpen: isVisible,
    panelPaddingSize: "none"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSelectable, {
    onChange: changeDatasetFilter,
    options: selectableOptions,
    searchable: true
  }, (list, search) => /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_eui.EuiPopoverTitle, null, search), list))));
};
exports.IndexSetupDatasetFilter = IndexSetupDatasetFilter;
const DatasetWarningMarker = ({
  warnings
}) => {
  const warningDescriptions = warnings.flatMap(warning => warning.type === 'categoryQualityWarning' ? warning.reasons.map(reason => /*#__PURE__*/_react.default.createElement(_quality_warning_notices.CategoryQualityWarningReasonDescription, {
    key: reason.type,
    reason: reason
  })) : []);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    content: warningDescriptions,
    type: "warning",
    color: "warning"
  });
};
