// @kbn/cloud-security-posture-plugin/public/pages/configurations/test_subjects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getResourceFindingsTableRowTestId = exports.getFindingsTableRowTestId = exports.getFindingsTableCellTestId = exports.getFindingsByResourceTableRowTestId = exports.RESOURCES_FINDINGS_TABLE = exports.RESOURCES_FINDINGS_CONTAINER = exports.LATEST_FINDINGS_TABLE = exports.LATEST_FINDINGS_CONTAINER = exports.FINDINGS_TABLE_EXPAND_COLUMN = exports.FINDINGS_TABLE_CELL_ADD_NEGATED_FILTER = exports.FINDINGS_TABLE_CELL_ADD_FILTER = exports.FINDINGS_TABLE = exports.FINDINGS_MISCONFIGS_FLYOUT_DESCRIPTION_LIST = exports.FINDINGS_GROUP_BY_SELECTOR = exports.FINDINGS_GROUPING_COUNTER = exports.FINDINGS_FLYOUT = exports.FINDINGS_CONTAINER = exports.FINDINGS_BY_RESOURCE_TABLE_RESOURCE_ID_COLUMN = exports.FINDINGS_BY_RESOURCE_TABLE = exports.FINDINGS_BY_RESOURCE_CONTAINER = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FINDINGS_FLYOUT = exports.FINDINGS_FLYOUT = 'findings_flyout';
const FINDINGS_TABLE_EXPAND_COLUMN = exports.FINDINGS_TABLE_EXPAND_COLUMN = 'findings_table_expand_column';
const FINDINGS_TABLE = exports.FINDINGS_TABLE = 'findings_table';
const FINDINGS_CONTAINER = exports.FINDINGS_CONTAINER = 'findings_container';
const FINDINGS_BY_RESOURCE_CONTAINER = exports.FINDINGS_BY_RESOURCE_CONTAINER = 'findings_by_resource_container';
const FINDINGS_BY_RESOURCE_TABLE_RESOURCE_ID_COLUMN = exports.FINDINGS_BY_RESOURCE_TABLE_RESOURCE_ID_COLUMN = 'findings_by_resource_table_resource_id_column';
const FINDINGS_BY_RESOURCE_TABLE = exports.FINDINGS_BY_RESOURCE_TABLE = 'findings_by_resource_table';
const getFindingsByResourceTableRowTestId = id => `findings_resource_table_row_${id}`;
exports.getFindingsByResourceTableRowTestId = getFindingsByResourceTableRowTestId;
const LATEST_FINDINGS_CONTAINER = exports.LATEST_FINDINGS_CONTAINER = 'latest_findings_container';
const LATEST_FINDINGS_TABLE = exports.LATEST_FINDINGS_TABLE = 'latest_findings_table';
const FINDINGS_GROUP_BY_SELECTOR = exports.FINDINGS_GROUP_BY_SELECTOR = 'findings_group_by_selector';
const FINDINGS_GROUPING_COUNTER = exports.FINDINGS_GROUPING_COUNTER = 'findings_grouping_counter';
const getFindingsTableRowTestId = id => `findings_table_row_${id}`;
exports.getFindingsTableRowTestId = getFindingsTableRowTestId;
const getFindingsTableCellTestId = (columnId, rowId) => `findings_table_cell_${columnId}_${rowId}`;
exports.getFindingsTableCellTestId = getFindingsTableCellTestId;
const FINDINGS_TABLE_CELL_ADD_FILTER = exports.FINDINGS_TABLE_CELL_ADD_FILTER = 'findings_table_cell_add_filter';
const FINDINGS_TABLE_CELL_ADD_NEGATED_FILTER = exports.FINDINGS_TABLE_CELL_ADD_NEGATED_FILTER = 'findings_table_cell_add_negated_filter';
const RESOURCES_FINDINGS_CONTAINER = exports.RESOURCES_FINDINGS_CONTAINER = 'resources_findings_container';
const RESOURCES_FINDINGS_TABLE = exports.RESOURCES_FINDINGS_TABLE = 'resource_findings_table';
const getResourceFindingsTableRowTestId = id => `resource_findings_table_row_${id}`;
exports.getResourceFindingsTableRowTestId = getResourceFindingsTableRowTestId;
const FINDINGS_MISCONFIGS_FLYOUT_DESCRIPTION_LIST = exports.FINDINGS_MISCONFIGS_FLYOUT_DESCRIPTION_LIST = 'misconfigs-findings-flyout-description-list';
