// @kbn/alerting-plugin/common/routes/rule/response/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "actionParamsSchema", {
  enumerable: true,
  get: function () {
    return _latest.actionParamsSchema;
  }
});
Object.defineProperty(exports, "actionParamsSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.actionParamsSchema;
  }
});
Object.defineProperty(exports, "alertDelaySchemaV1", {
  enumerable: true,
  get: function () {
    return _v.alertDelaySchema;
  }
});
Object.defineProperty(exports, "mappedParamsSchema", {
  enumerable: true,
  get: function () {
    return _latest.mappedParamsSchema;
  }
});
Object.defineProperty(exports, "mappedParamsSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.mappedParamsSchema;
  }
});
Object.defineProperty(exports, "monitoringSchema", {
  enumerable: true,
  get: function () {
    return _latest.monitoringSchema;
  }
});
Object.defineProperty(exports, "monitoringSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.monitoringSchema;
  }
});
Object.defineProperty(exports, "notifyWhenSchema", {
  enumerable: true,
  get: function () {
    return _latest.notifyWhenSchema;
  }
});
Object.defineProperty(exports, "notifyWhenSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.notifyWhenSchema;
  }
});
Object.defineProperty(exports, "ruleExecutionStatusSchema", {
  enumerable: true,
  get: function () {
    return _latest.ruleExecutionStatusSchema;
  }
});
Object.defineProperty(exports, "ruleExecutionStatusSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.ruleExecutionStatusSchema;
  }
});
Object.defineProperty(exports, "ruleLastRunSchema", {
  enumerable: true,
  get: function () {
    return _latest.ruleLastRunSchema;
  }
});
Object.defineProperty(exports, "ruleLastRunSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.ruleLastRunSchema;
  }
});
Object.defineProperty(exports, "ruleParamsSchema", {
  enumerable: true,
  get: function () {
    return _latest.ruleParamsSchema;
  }
});
Object.defineProperty(exports, "ruleParamsSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.ruleParamsSchema;
  }
});
Object.defineProperty(exports, "ruleResponseSchema", {
  enumerable: true,
  get: function () {
    return _latest.ruleResponseSchema;
  }
});
Object.defineProperty(exports, "ruleResponseSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.ruleResponseSchema;
  }
});
Object.defineProperty(exports, "ruleSnoozeScheduleSchema", {
  enumerable: true,
  get: function () {
    return _latest.ruleSnoozeScheduleSchema;
  }
});
Object.defineProperty(exports, "ruleSnoozeScheduleSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.ruleSnoozeScheduleSchema;
  }
});
Object.defineProperty(exports, "scheduleIdsSchema", {
  enumerable: true,
  get: function () {
    return _latest.scheduleIdsSchema;
  }
});
Object.defineProperty(exports, "scheduleIdsSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.scheduleIdsSchema;
  }
});
var _latest = require("./schemas/latest");
var _v = require("./schemas/v1");
