// @kbn/ml-plugin/public/application/data_frame_analytics/pages/analytics_creation/components/back_to_list_panel/back_to_list_panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BackToListPanel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _kibana = require("../../../../../contexts/kibana");
var _locator = require("../../../../../../../common/constants/locator");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const BackToListPanel = () => {
  const analyticsManagementPageLink = (0, _kibana.useMlLink)({
    page: _locator.ML_PAGES.DATA_FRAME_ANALYTICS_JOBS_MANAGE
  });
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCard, {
    className: "dfAnalyticsCreationWizard__card",
    icon: /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      size: "xxl",
      type: "list"
    }),
    title: _i18n.i18n.translate('xpack.ml.dataframe.analytics.create.analyticsListCardTitle', {
      defaultMessage: 'Data Frame Analytics'
    }),
    description: _i18n.i18n.translate('xpack.ml.dataframe.analytics.create.analyticsListCardDescription', {
      defaultMessage: 'Return to the analytics management page.'
    }),
    href: analyticsManagementPageLink,
    "data-test-subj": "analyticsWizardCardManagement"
  }));
};
exports.BackToListPanel = BackToListPanel;
