// @kbn/infra-plugin/public/pages/metrics/inventory_view/components/waffle/gradient_legend.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GradientLegend = void 0;
var _react = _interopRequireDefault(require("react"));
var _common = require("@kbn/kibana-react-plugin/common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createTickRender = (bounds, formatter) => (rule, index) => {
  const value = rule.value === 0 ? bounds.min : bounds.max * rule.value;
  const style = {
    left: `${rule.value * 100}%`
  };
  const label = formatter(value);
  return /*#__PURE__*/_react.default.createElement(GradientLegendTick, {
    style: style,
    key: `legend-rule-${index}`
  }, /*#__PURE__*/_react.default.createElement(GradientLegendTickLine, null), /*#__PURE__*/_react.default.createElement(GradientLegendTickLabel, null, label));
};
const GradientLegend = ({
  legend,
  bounds,
  formatter
}) => {
  const maxValue = legend.rules.reduce((acc, rule) => {
    return acc < rule.value ? rule.value : acc;
  }, 0);
  const colorStops = legend.rules.map(rule => {
    const percent = rule.value / maxValue * 100;
    return `${rule.color} ${percent}%`;
  });
  const style = {
    background: `linear-gradient(to right, ${colorStops})`
  };
  return /*#__PURE__*/_react.default.createElement(GradientLegendContainer, {
    style: style
  }, legend.rules.map(createTickRender(bounds, formatter)));
};
exports.GradientLegend = GradientLegend;
const GradientLegendContainer = _common.euiStyled.div`
  position: absolute;
  height: 10px;
  bottom: 0;
  left: 0;
  right: 40px;
`;
const GradientLegendTick = _common.euiStyled.div`
  position: absolute;
  bottom: 0;
  top: -18px;
`;
const GradientLegendTickLine = _common.euiStyled.div`
  position: absolute;
  background-color: ${props => props.theme.eui.euiBorderColor};
  width: 1px;
  left: 0;
  top: 15px;
  bottom: 0;
  ${GradientLegendTick}:first-child {
    top: 2px;
  }
  ${GradientLegendTick}:last-child {
    top: 2px;
  }
`;
const GradientLegendTickLabel = _common.euiStyled.div`
  position: absolute;
  font-size: 11px;
  text-align: center;
  top: 0;
  left: 0;
  white-space: nowrap;
  transform: translate(-50%, 0);
  ${GradientLegendTick}:first-child & {
    padding-left: 5px;
    transform: translate(0, 0);
  }
  ${GradientLegendTick}:last-child & {
    padding-right: 5px;
    transform: translate(-100%, 0);
  }
`;
