// @kbn/enterprise-search-plugin/public/applications/app_search/components/analytics/analytics_layout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AnalyticsLayout = void 0;
var _react = _interopRequireWildcard(require("react"));
var _reactRouterDom = require("react-router-dom");
var _kea = require("kea");
var _kibana = require("../../../shared/kibana");
var _engine = require("../engine");
var _layout = require("../layout");
var _log_retention = require("../log_retention");
var _components = require("./components");
var _constants = require("./constants");
var _ = require(".");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AnalyticsLayout = ({
  title,
  breadcrumbs = [],
  isQueryView,
  isAnalyticsView,
  children
}) => {
  const {
    history
  } = (0, _kea.useValues)(_kibana.KibanaLogic);
  const {
    query
  } = (0, _reactRouterDom.useParams)();
  const {
    dataLoading
  } = (0, _kea.useValues)(_.AnalyticsLogic);
  const {
    loadAnalyticsData,
    loadQueryData
  } = (0, _kea.useActions)(_.AnalyticsLogic);
  (0, _react.useEffect)(() => {
    if (isQueryView) loadQueryData(query);
    if (isAnalyticsView) loadAnalyticsData();
  }, [history.location.search]);
  return /*#__PURE__*/_react.default.createElement(_layout.AppSearchPageTemplate, {
    pageChrome: (0, _engine.getEngineBreadcrumbs)([_constants.ANALYTICS_TITLE, ...breadcrumbs]),
    pageHeader: {
      pageTitle: title,
      rightSideItems: [/*#__PURE__*/_react.default.createElement(_log_retention.LogRetentionTooltip, {
        type: _log_retention.LogRetentionOptions.Analytics,
        position: "left"
      })],
      children: /*#__PURE__*/_react.default.createElement(_components.AnalyticsFilters, null),
      responsive: false
    },
    isLoading: dataLoading
  }, /*#__PURE__*/_react.default.createElement(_log_retention.LogRetentionCallout, {
    type: _log_retention.LogRetentionOptions.Analytics
  }), children);
};
exports.AnalyticsLayout = AnalyticsLayout;
