// @kbn/apm-plugin/public/components/app/error_group_details/error_sampler/error_ui_actions_context_menu.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ErrorUiActionsContextMenu = ErrorUiActionsContextMenu;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function ErrorUiActionsContextMenu({
  items
}) {
  const [isContextMenuOpen, setIsContextMenuOpen] = (0, _react.useState)(false);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
    id: "errorContextMenu",
    button: /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      "data-test-subj": "ErrorSampleDetailsButton",
      onClick: () => setIsContextMenuOpen(isOpen => !isOpen),
      iconType: "arrowDown",
      iconSide: "right"
    }, _i18n.i18n.translate('xpack.apm.errorSampleDetails.investigateMenu', {
      defaultMessage: 'Investigate'
    })),
    isOpen: isContextMenuOpen,
    closePopover: () => setIsContextMenuOpen(() => false),
    panelPaddingSize: "none",
    anchorPosition: "downLeft"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiContextMenu, {
    initialPanelId: "mainMenu",
    panels: [{
      id: 'mainMenu',
      items: items
    }]
  }));
}
