// @kbn/fleet-plugin/server/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AgentNotFoundError", {
  enumerable: true,
  get: function () {
    return _errors.AgentNotFoundError;
  }
});
Object.defineProperty(exports, "FleetUnauthorizedError", {
  enumerable: true,
  get: function () {
    return _errors.FleetUnauthorizedError;
  }
});
Object.defineProperty(exports, "NamingCollisionError", {
  enumerable: true,
  get: function () {
    return _check_naming_collision.NamingCollisionError;
  }
});
Object.defineProperty(exports, "buildAgentStatusRuntimeField", {
  enumerable: true,
  get: function () {
    return _build_status_runtime_field.buildAgentStatusRuntimeField;
  }
});
Object.defineProperty(exports, "config", {
  enumerable: true,
  get: function () {
    return _config.config;
  }
});
Object.defineProperty(exports, "getRegistryUrl", {
  enumerable: true,
  get: function () {
    return _services.getRegistryUrl;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "relativeDownloadUrlFromArtifact", {
  enumerable: true,
  get: function () {
    return _mappings.relativeDownloadUrlFromArtifact;
  }
});
var _build_status_runtime_field = require("./services/agents/build_status_runtime_field");
var _services = require("./services");
var _check_naming_collision = require("./services/epm/packages/custom_integrations/validation/check_naming_collision");
var _errors = require("./errors");
var _config = require("./config");
var _mappings = require("./services/artifacts/mappings");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const plugin = async initializerContext => {
  const {
    FleetPlugin
  } = await Promise.resolve().then(() => _interopRequireWildcard(require('./plugin')));
  return new FleetPlugin(initializerContext);
};
exports.plugin = plugin;
