// @kbn/slo-plugin/public/pages/slos/components/card_view/slos_card_view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SloListCardView = SloListCardView;
var _eui = require("@elastic/eui");
var _sloSchema = require("@kbn/slo-schema");
var _react = _interopRequireDefault(require("react"));
var _use_fetch_active_alerts = require("../../../../hooks/use_fetch_active_alerts");
var _use_fetch_historical_summary = require("../../../../hooks/use_fetch_historical_summary");
var _use_fetch_rules_for_slo = require("../../../../hooks/use_fetch_rules_for_slo");
var _slo_card_item = require("./slo_card_item");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useColumns = () => {
  const isMobile = (0, _eui.useIsWithinBreakpoints)(['xs', 's']);
  const isMedium = (0, _eui.useIsWithinBreakpoints)(['m']);
  const isXLarge = (0, _eui.useIsWithinBreakpoints)(['xl']);
  switch (true) {
    case isMobile:
      return 1;
    case isMedium:
      return 3;
    case isXLarge:
      return 4;
    default:
      return 3;
  }
};
function SloListCardView({
  sloList,
  loading,
  error
}) {
  const sloIdsAndInstanceIds = sloList.map(slo => {
    var _slo$instanceId;
    return [slo.id, (_slo$instanceId = slo.instanceId) !== null && _slo$instanceId !== void 0 ? _slo$instanceId : _sloSchema.ALL_VALUE];
  });
  const {
    data: activeAlertsBySlo
  } = (0, _use_fetch_active_alerts.useFetchActiveAlerts)({
    sloIdsAndInstanceIds
  });
  const {
    data: rulesBySlo,
    refetchRules
  } = (0, _use_fetch_rules_for_slo.useFetchRulesForSlo)({
    sloIds: sloIdsAndInstanceIds.map(item => item[0])
  });
  const {
    isLoading: historicalSummaryLoading,
    data: historicalSummaries = []
  } = (0, _use_fetch_historical_summary.useFetchHistoricalSummary)({
    sloList
  });
  const columns = useColumns();
  if (loading && sloList.length === 0) {
    return /*#__PURE__*/_react.default.createElement(LoadingSloGrid, {
      gridSize: columns
    });
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGrid, {
    columns: columns,
    gutterSize: "m"
  }, sloList.filter(slo => slo.summary).map(slo => {
    var _slo$instanceId2, _historicalSummaries$;
    return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      key: `${slo.id}-${(_slo$instanceId2 = slo.instanceId) !== null && _slo$instanceId2 !== void 0 ? _slo$instanceId2 : 'ALL_VALUE'}`
    }, /*#__PURE__*/_react.default.createElement(_slo_card_item.SloCardItem, {
      slo: slo,
      loading: loading,
      error: error,
      activeAlerts: activeAlertsBySlo.get(slo),
      rules: rulesBySlo === null || rulesBySlo === void 0 ? void 0 : rulesBySlo[slo.id],
      historicalSummary: (_historicalSummaries$ = historicalSummaries.find(historicalSummary => {
        var _slo$instanceId3;
        return historicalSummary.sloId === slo.id && historicalSummary.instanceId === ((_slo$instanceId3 = slo.instanceId) !== null && _slo$instanceId3 !== void 0 ? _slo$instanceId3 : _sloSchema.ALL_VALUE);
      })) === null || _historicalSummaries$ === void 0 ? void 0 : _historicalSummaries$.data,
      historicalSummaryLoading: historicalSummaryLoading,
      refetchRules: refetchRules
    }));
  }));
}
function LoadingSloGrid({
  gridSize
}) {
  const ROWS = 4;
  const COLUMNS = gridSize;
  const loaders = Array(ROWS * COLUMNS).fill(null);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGrid, {
    gutterSize: "m",
    columns: COLUMNS
  }, loaders.map((_, i) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    key: i
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    style: {
      height: '200px'
    },
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSkeletonText, {
    lines: 2
  })), ' '))));
}
