// @kbn/security-solution-plugin/public/common/components/ml_popover/hooks/use_fetch_modules_query.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useInvalidateFetchModulesQuery = exports.useFetchModulesQuery = exports.GET_MODULES_QUERY_KEY = void 0;
var _reactQuery = require("@tanstack/react-query");
var _react = require("react");
var _api = require("../api");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ONE_MINUTE = 60000;
const GET_MODULES_QUERY_KEY = exports.GET_MODULES_QUERY_KEY = ['GET', '/internal/ml/modules/get_module/:moduleId'];
const useFetchModulesQuery = (queryArgs, options) => {
  return (0, _reactQuery.useQuery)([GET_MODULES_QUERY_KEY, queryArgs], async ({
    signal
  }) => (0, _api.getModules)({
    signal,
    ...queryArgs
  }), {
    refetchIntervalInBackground: false,
    staleTime: ONE_MINUTE * 5,
    retry: false,
    ...options
  });
};
exports.useFetchModulesQuery = useFetchModulesQuery;
const useInvalidateFetchModulesQuery = () => {
  const queryClient = (0, _reactQuery.useQueryClient)();
  return (0, _react.useCallback)(() => {
    queryClient.invalidateQueries(GET_MODULES_QUERY_KEY, {
      refetchType: 'active'
    });
  }, [queryClient]);
};
exports.useInvalidateFetchModulesQuery = useInvalidateFetchModulesQuery;
