// @kbn/lists-plugin/server/schemas/saved_objects/exceptions_list_so_schema.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.exceptionListSoSchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _securitysolutionIoTsListTypes = require("@kbn/securitysolution-io-ts-list-types");
var _securitysolutionIoTsTypes = require("@kbn/securitysolution-io-ts-types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Superset saved object of both lists and list items since they share the same saved object type.
 */
const exceptionListSoSchema = exports.exceptionListSoSchema = t.exact(t.type({
  comments: _securitysolutionIoTsListTypes.commentsArrayOrUndefined,
  created_at: _securitysolutionIoTsListTypes.created_at,
  created_by: _securitysolutionIoTsListTypes.created_by,
  description: _securitysolutionIoTsListTypes.description,
  entries: _securitysolutionIoTsListTypes.entriesArrayOrUndefined,
  expire_time: _securitysolutionIoTsListTypes.expireTimeOrUndefined,
  immutable: _securitysolutionIoTsListTypes.immutableOrUndefined,
  item_id: _securitysolutionIoTsListTypes.itemIdOrUndefined,
  list_id: _securitysolutionIoTsListTypes.list_id,
  list_type: _securitysolutionIoTsListTypes.list_type,
  meta: _securitysolutionIoTsListTypes.metaOrUndefined,
  name: _securitysolutionIoTsListTypes.name,
  os_types: _securitysolutionIoTsListTypes.osTypeArray,
  tags: _securitysolutionIoTsListTypes.tags,
  tie_breaker_id: _securitysolutionIoTsListTypes.tie_breaker_id,
  type: t.union([_securitysolutionIoTsListTypes.exceptionListType, _securitysolutionIoTsListTypes.exceptionListItemType]),
  updated_by: _securitysolutionIoTsListTypes.updated_by,
  version: _securitysolutionIoTsTypes.versionOrUndefined
}));
