// @kbn/apm-plugin/public/components/app/onboarding/instructions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createDjangoAgentInstructions", {
  enumerable: true,
  get: function () {
    return _django_agent.createDjangoAgentInstructions;
  }
});
Object.defineProperty(exports, "createDotNetAgentInstructions", {
  enumerable: true,
  get: function () {
    return _dotnet_agent.createDotNetAgentInstructions;
  }
});
Object.defineProperty(exports, "createFlaskAgentInstructions", {
  enumerable: true,
  get: function () {
    return _flask_agent.createFlaskAgentInstructions;
  }
});
Object.defineProperty(exports, "createGoAgentInstructions", {
  enumerable: true,
  get: function () {
    return _go_agent.createGoAgentInstructions;
  }
});
Object.defineProperty(exports, "createJavaAgentInstructions", {
  enumerable: true,
  get: function () {
    return _java_agent.createJavaAgentInstructions;
  }
});
Object.defineProperty(exports, "createNodeAgentInstructions", {
  enumerable: true,
  get: function () {
    return _node_agent.createNodeAgentInstructions;
  }
});
Object.defineProperty(exports, "createOpenTelemetryAgentInstructions", {
  enumerable: true,
  get: function () {
    return _otel_agent.createOpenTelemetryAgentInstructions;
  }
});
Object.defineProperty(exports, "createPhpAgentInstructions", {
  enumerable: true,
  get: function () {
    return _php_agent.createPhpAgentInstructions;
  }
});
Object.defineProperty(exports, "createRackAgentInstructions", {
  enumerable: true,
  get: function () {
    return _rack_agent.createRackAgentInstructions;
  }
});
Object.defineProperty(exports, "createRailsAgentInstructions", {
  enumerable: true,
  get: function () {
    return _rails_agent.createRailsAgentInstructions;
  }
});
var _java_agent = require("./java_agent");
var _node_agent = require("./node_agent");
var _django_agent = require("./django_agent");
var _flask_agent = require("./flask_agent");
var _rails_agent = require("./rails_agent");
var _rack_agent = require("./rack_agent");
var _go_agent = require("./go_agent");
var _dotnet_agent = require("./dotnet_agent");
var _php_agent = require("./php_agent");
var _otel_agent = require("./otel_agent");
