// @kbn/cases-plugin/server/services/user_actions/builders/status.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusUserActionBuilder = void 0;
var _constants = require("../../../../common/constants");
var _domain = require("../../../../common/types/domain");
var _abstract_builder = require("../abstract_builder");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class StatusUserActionBuilder extends _abstract_builder.UserActionBuilder {
  build(args) {
    const action = _domain.UserActionActions.update;
    const parameters = this.buildCommonUserAction({
      ...args,
      action,
      valueKey: 'status',
      value: args.payload.status,
      type: _domain.UserActionTypes.status
    });
    const getMessage = id => `User updated the status for case id: ${args.caseId} - user action id: ${id}`;
    const eventDetails = {
      getMessage,
      action,
      descriptiveAction: 'case_user_action_update_case_status',
      savedObjectId: args.caseId,
      savedObjectType: _constants.CASE_SAVED_OBJECT
    };
    return {
      parameters,
      eventDetails
    };
  }
}
exports.StatusUserActionBuilder = StatusUserActionBuilder;
