// @kbn/infra-plugin/public/components/logging/log_entry_examples/log_entry_examples.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogEntryExampleMessages = void 0;
var _react = _interopRequireDefault(require("react"));
var _common = require("@kbn/kibana-react-plugin/common");
var _log_entry_examples_empty_indicator = require("./log_entry_examples_empty_indicator");
var _log_entry_examples_failure_indicator = require("./log_entry_examples_failure_indicator");
var _log_entry_examples_loading_indicator = require("./log_entry_examples_loading_indicator");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LogEntryExampleMessages = ({
  isLoading,
  hasFailedLoading,
  exampleCount,
  hasResults,
  onReload,
  children
}) => {
  return /*#__PURE__*/_react.default.createElement(Wrapper, null, isLoading ? /*#__PURE__*/_react.default.createElement(_log_entry_examples_loading_indicator.LogEntryExampleMessagesLoadingIndicator, {
    exampleCount: exampleCount
  }) : hasFailedLoading ? /*#__PURE__*/_react.default.createElement(_log_entry_examples_failure_indicator.LogEntryExampleMessagesFailureIndicator, {
    onRetry: onReload
  }) : !hasResults ? /*#__PURE__*/_react.default.createElement(_log_entry_examples_empty_indicator.LogEntryExampleMessagesEmptyIndicator, {
    onReload: onReload
  }) : children);
};
exports.LogEntryExampleMessages = LogEntryExampleMessages;
const Wrapper = _common.euiStyled.div`
  align-items: stretch;
  flex-direction: column;
  flex: 1 0 0%;
  overflow: hidden;
  padding-top: 1px; // Buffer for the "Reload" buttons' hover state
`;
