// @kbn/enterprise-search-plugin/public/applications/enterprise_search_content/components/search_index/connector/sync_rules/edit_sync_rules_tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditSyncRulesTab = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _connector_sync_form = require("./connector_sync_form");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EditSyncRulesTab = ({
  children,
  revertAction
}) => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "flexEnd"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "enterpriseSearchEditSyncRulesTabRevertToActiveRulesButton",
    "data-telemetry-id": "entSearchContent-connector-syncRules-editRules-revert",
    onClick: revertAction
  }, _i18n.i18n.translate('xpack.enterpriseSearch.content.index.connector.syncRules.flyout.revertButtonTitle', {
    defaultMessage: 'Revert to active rules'
  }))))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_connector_sync_form.ConnectorSyncRulesForm, null, children))));
};
exports.EditSyncRulesTab = EditSyncRulesTab;
