// @kbn/security-solution-plugin/public/detections/components/alerts_table/alerts_filter_group/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertsTableFilterGroup = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _types = require("../../../../../common/types");
var i18n = _interopRequireWildcard(require("../translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AlertsTableFilterGroupComponent = ({
  status = _types.FILTER_OPEN,
  onFilterGroupChanged
}) => {
  const options = [{
    id: 'open',
    label: i18n.OPEN_ALERTS,
    'data-test-subj': 'openAlerts'
  }, {
    id: 'acknowledged',
    label: i18n.ACKNOWLEDGED_ALERTS,
    'data-test-subj': 'acknowledgedAlerts'
  }, {
    id: 'closed',
    label: i18n.CLOSED_ALERTS,
    'data-test-subj': 'closedAlerts'
  }];
  const onChange = (0, _react.useCallback)(id => {
    onFilterGroupChanged(id);
  }, [onFilterGroupChanged]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiButtonGroup, {
    legend: "filter status",
    color: "primary",
    options: options,
    idSelected: status,
    "data-test-subj": "alerts-table-filter-group",
    onChange: onChange
  });
};
const AlertsTableFilterGroup = exports.AlertsTableFilterGroup = /*#__PURE__*/_react.default.memo(AlertsTableFilterGroupComponent);
