// @kbn/maps-plugin/public/connected_components/add_layer_panel/flyout_body/flyout_body.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutBody = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _layer_wizard_select = require("./layer_wizard_select");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FlyoutBody = props => {
  function renderContent() {
    if (!props.layerWizard || !props.currentStepId) {
      return /*#__PURE__*/_react.default.createElement(_layer_wizard_select.LayerWizardSelect, {
        onSelect: props.onWizardSelect
      });
    }
    const renderWizardArgs = {
      previewLayers: props.previewLayers,
      mapColors: props.mapColors,
      mostCommonDataViewId: props.mostCommonDataViewId,
      currentStepId: props.currentStepId,
      isOnFinalStep: props.isOnFinalStep,
      enableNextBtn: props.enableNextBtn,
      disableNextBtn: props.disableNextBtn,
      startStepLoading: props.startStepLoading,
      stopStepLoading: props.stopStepLoading,
      advanceToNextStep: props.advanceToNextStep
    };
    const backButton = props.showBackButton ? /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
      size: "xs",
      flush: "left",
      onClick: props.onClear,
      iconType: "arrowLeft"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.maps.addLayerPanel.changeDataSourceButtonLabel",
      defaultMessage: "Change layer"
    })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    })) : null;
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, backButton, props.layerWizard.renderWizard(renderWizardArgs));
  }
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "mapLayerPanel__body",
    "data-test-subj": "layerAddForm"
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: "mapLayerPanel__bodyOverflow"
  }, renderContent()));
};
exports.FlyoutBody = FlyoutBody;
