// @kbn/visualizations-plugin/public/convert_to_lens/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "convertToFiltersColumn", {
  enumerable: true,
  get: function () {
    return _lib.convertToFiltersColumn;
  }
});
Object.defineProperty(exports, "createStaticValueColumn", {
  enumerable: true,
  get: function () {
    return _lib.createStaticValueColumn;
  }
});
Object.defineProperty(exports, "getColumnsFromVis", {
  enumerable: true,
  get: function () {
    return _schemas.getColumnsFromVis;
  }
});
Object.defineProperty(exports, "getPalette", {
  enumerable: true,
  get: function () {
    return _lib.getPalette;
  }
});
Object.defineProperty(exports, "getPaletteFromStopsWithColors", {
  enumerable: true,
  get: function () {
    return _lib.getPaletteFromStopsWithColors;
  }
});
Object.defineProperty(exports, "getPercentageColumnFormulaColumn", {
  enumerable: true,
  get: function () {
    return _lib.getPercentageColumnFormulaColumn;
  }
});
Object.defineProperty(exports, "getPercentageModeConfig", {
  enumerable: true,
  get: function () {
    return _lib.getPercentageModeConfig;
  }
});
var _schemas = require("./schemas");
var _lib = require("../../common/convert_to_lens/lib");
