// @kbn/screenshotting-plugin/server/browsers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ChromiumArchivePaths", {
  enumerable: true,
  get: function () {
    return _chromium.ChromiumArchivePaths;
  }
});
Object.defineProperty(exports, "DEFAULT_VIEWPORT", {
  enumerable: true,
  get: function () {
    return _chromium.DEFAULT_VIEWPORT;
  }
});
Object.defineProperty(exports, "HeadlessChromiumDriver", {
  enumerable: true,
  get: function () {
    return _chromium.HeadlessChromiumDriver;
  }
});
Object.defineProperty(exports, "HeadlessChromiumDriverFactory", {
  enumerable: true,
  get: function () {
    return _chromium.HeadlessChromiumDriverFactory;
  }
});
Object.defineProperty(exports, "download", {
  enumerable: true,
  get: function () {
    return _download.download;
  }
});
Object.defineProperty(exports, "getChromiumDisconnectedError", {
  enumerable: true,
  get: function () {
    return _chromium.getChromiumDisconnectedError;
  }
});
Object.defineProperty(exports, "install", {
  enumerable: true,
  get: function () {
    return _install.install;
  }
});
var _download = require("./download");
var _install = require("./install");
var _chromium = require("./chromium");
