// @kbn/rollup-plugin/public/crud_app/sections/job_list/deprecated_prompt.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DeprecatedPrompt = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _documentation_links = require("../../services/documentation_links");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DeprecatedPrompt = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPageSection, {
    alignment: "center",
    grow: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    color: "primary",
    "data-test-subj": "jobListDeprecatedPrompt",
    iconType: "iInCircle",
    title: /*#__PURE__*/_react.default.createElement("h1", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.rollupJobs.deprecatedPromptTitle",
      defaultMessage: "Deprecated in 8.11.0"
    })),
    body: /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.rollupJobs.deprecatedPromptDescription",
      defaultMessage: "Rollups are deprecated and will be removed in a future version. Use downsampling instead."
    }))),
    actions: /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      href: _documentation_links.documentationLinks.fleet.datastreamsDownsampling,
      target: "_blank",
      fill: true,
      iconType: "help",
      "data-test-subj": "rollupDeprecatedPromptDocsLink"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.rollupJobs.deprecatedPrompt.downsamplingDocsButtonLabel",
      defaultMessage: "Downsampling docs"
    }))
  }));
};
exports.DeprecatedPrompt = DeprecatedPrompt;
