// @kbn/vis-type-xy-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AxisType", {
  enumerable: true,
  get: function () {
    return _types.AxisType;
  }
});
Object.defineProperty(exports, "ScaleType", {
  enumerable: true,
  get: function () {
    return _types.ScaleType;
  }
});
Object.defineProperty(exports, "TruncateLabelsOption", {
  enumerable: true,
  get: function () {
    return _truncate_labels.TruncateLabelsOption;
  }
});
Object.defineProperty(exports, "getPositions", {
  enumerable: true,
  get: function () {
    return _positions.getPositions;
  }
});
Object.defineProperty(exports, "getScaleTypes", {
  enumerable: true,
  get: function () {
    return _scale_types.getScaleTypes;
  }
});
exports.plugin = plugin;
var _plugin = require("./plugin");
var _types = require("./types");
var _truncate_labels = require("./editor/components/common/truncate_labels");
var _positions = require("./editor/positions");
var _scale_types = require("./editor/scale_types");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

// TODO: Remove when vis_type_vislib is removed
// https://github.com/elastic/kibana/issues/56143

function plugin(initializerContext) {
  return new _plugin.VisTypeXyPlugin(initializerContext);
}
