// @kbn/vis-type-xy-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUISettings = exports.setPalettesService = exports.setDataViewsStart = exports.getUISettings = exports.getPalettesService = exports.getDataViewsStart = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getUISettings, setUISettings] = (0, _public.createGetterSetter)('xy core.uiSettings');
exports.setUISettings = setUISettings;
exports.getUISettings = getUISettings;
const [getPalettesService, setPalettesService] = (0, _public.createGetterSetter)('xy charts.palette');
exports.setPalettesService = setPalettesService;
exports.getPalettesService = getPalettesService;
const [getDataViewsStart, setDataViewsStart] = (0, _public.createGetterSetter)('dataViews');
exports.setDataViewsStart = setDataViewsStart;
exports.getDataViewsStart = getDataViewsStart;
