// @kbn/enterprise-search-plugin/public/applications/analytics/components/analytics_collection_view/analytics_collection_explore_table_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExploreTables = exports.ExploreTableColumns = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let ExploreTables = exports.ExploreTables = /*#__PURE__*/function (ExploreTables) {
  ExploreTables[ExploreTables["SearchTerms"] = 0] = "SearchTerms";
  ExploreTables[ExploreTables["WorsePerformers"] = 1] = "WorsePerformers";
  ExploreTables[ExploreTables["Clicked"] = 2] = "Clicked";
  ExploreTables[ExploreTables["Referrers"] = 3] = "Referrers";
  ExploreTables[ExploreTables["Locations"] = 4] = "Locations";
  return ExploreTables;
}({});
let ExploreTableColumns = exports.ExploreTableColumns = /*#__PURE__*/function (ExploreTableColumns) {
  ExploreTableColumns["count"] = "count";
  ExploreTableColumns["searchTerms"] = "searchTerms";
  ExploreTableColumns["query"] = "query";
  ExploreTableColumns["location"] = "location";
  ExploreTableColumns["page"] = "page";
  ExploreTableColumns["sessions"] = "sessions";
  return ExploreTableColumns;
}({});
