// @kbn/esql-ast/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ESQLErrorListener", {
  enumerable: true,
  get: function () {
    return _antlr_error_listener.ESQLErrorListener;
  }
});
Object.defineProperty(exports, "ROOT_STATEMENT", {
  enumerable: true,
  get: function () {
    return _antlr_facade.ROOT_STATEMENT;
  }
});
Object.defineProperty(exports, "Walker", {
  enumerable: true,
  get: function () {
    return _walker.Walker;
  }
});
Object.defineProperty(exports, "getAstAndSyntaxErrors", {
  enumerable: true,
  get: function () {
    return _ast_parser.getAstAndSyntaxErrors;
  }
});
Object.defineProperty(exports, "getLexer", {
  enumerable: true,
  get: function () {
    return _antlr_facade.getLexer;
  }
});
Object.defineProperty(exports, "getParser", {
  enumerable: true,
  get: function () {
    return _antlr_facade.getParser;
  }
});
Object.defineProperty(exports, "walk", {
  enumerable: true,
  get: function () {
    return _walker.walk;
  }
});
var _antlr_facade = require("./src/antlr_facade");
var _ast_parser = require("./src/ast_parser");
var _antlr_error_listener = require("./src/antlr_error_listener");
var _walker = require("./src/walker");
