// @kbn/security-plugin/public/management/roles/roles_grid/permission_denied/permission_denied.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PermissionDenied = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PermissionDenied = () => /*#__PURE__*/_react.default.createElement(_eui.EuiPageSection, {
  alignment: "center",
  color: "danger"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  iconType: "securityApp",
  title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.security.management.roles.deniedPermissionTitle",
    defaultMessage: "You need permission to manage roles"
  })),
  body: /*#__PURE__*/_react.default.createElement("p", {
    "data-test-subj": "permissionDeniedMessage"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.security.management.roles.noPermissionToManageRolesDescription",
    defaultMessage: "Contact your system administrator."
  }))
}));
exports.PermissionDenied = PermissionDenied;
