// @kbn/infra-plugin/public/components/loading/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InfraLoadingStaticPanel = exports.InfraLoadingStaticContentPanel = exports.InfraLoadingPanel = void 0;
var _eui = require("@elastic/eui");
var React = _interopRequireWildcard(require("react"));
var _common = require("@kbn/kibana-react-plugin/common");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class InfraLoadingPanel extends React.PureComponent {
  render() {
    const {
      height,
      text,
      width
    } = this.props;
    return /*#__PURE__*/React.createElement(InfraLoadingStaticPanel, {
      style: {
        height,
        width
      }
    }, /*#__PURE__*/React.createElement(InfraLoadingStaticContentPanel, null, /*#__PURE__*/React.createElement(_eui.EuiPanel, null, /*#__PURE__*/React.createElement(_eui.EuiLoadingChart, {
      size: "m"
    }), /*#__PURE__*/React.createElement(_eui.EuiText, null, /*#__PURE__*/React.createElement("p", null, text)))));
  }
}
exports.InfraLoadingPanel = InfraLoadingPanel;
const InfraLoadingStaticPanel = exports.InfraLoadingStaticPanel = _common.euiStyled.div`
  position: relative;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
`;
const InfraLoadingStaticContentPanel = exports.InfraLoadingStaticContentPanel = _common.euiStyled.div`
  flex: 0 0 auto;
  align-self: center;
  text-align: center;
`;
