// @kbn/cases-plugin/public/components/user_actions/property_actions/alert_property_actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertPropertyActions = void 0;
var _react = _interopRequireWildcard(require("react"));
var _types = require("../../../client/attachment_framework/types");
var _use_cases_context = require("../../cases_context/use_cases_context");
var _delete_attachment_confirmation_modal = require("../delete_attachment_confirmation_modal");
var _property_actions = require("./property_actions");
var i18n = _interopRequireWildcard(require("./translations"));
var _use_delete_property_action = require("./use_delete_property_action");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AlertPropertyActionsComponent = ({
  isLoading,
  totalAlerts,
  onDelete
}) => {
  const {
    permissions
  } = (0, _use_cases_context.useCasesContext)();
  const {
    showDeletionModal,
    onModalOpen,
    onConfirm,
    onCancel
  } = (0, _use_delete_property_action.useDeletePropertyAction)({
    onDelete
  });
  const propertyActions = (0, _react.useMemo)(() => {
    const showRemoveAlertIcon = permissions.delete;
    return [...(showRemoveAlertIcon ? [{
      type: _types.AttachmentActionType.BUTTON,
      color: 'danger',
      disabled: false,
      iconType: 'minusInCircle',
      label: i18n.REMOVE_ALERTS(totalAlerts),
      onClick: onModalOpen
    }] : [])];
  }, [permissions.delete, totalAlerts, onModalOpen]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_property_actions.UserActionPropertyActions, {
    isLoading: isLoading,
    propertyActions: propertyActions
  }), showDeletionModal ? /*#__PURE__*/_react.default.createElement(_delete_attachment_confirmation_modal.DeleteAttachmentConfirmationModal, {
    title: i18n.REMOVE_ALERTS(totalAlerts),
    confirmButtonText: i18n.REMOVE,
    onCancel: onCancel,
    onConfirm: onConfirm
  }) : null);
};
AlertPropertyActionsComponent.displayName = 'AlertPropertyActions';
const AlertPropertyActions = exports.AlertPropertyActions = /*#__PURE__*/_react.default.memo(AlertPropertyActionsComponent);
