// @kbn/index-lifecycle-management-plugin/public/application/constants/ui_metric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIM_POLICY_UPDATE = exports.UIM_POLICY_DETACH_INDEX = exports.UIM_POLICY_DELETE = exports.UIM_POLICY_CREATE = exports.UIM_POLICY_ATTACH_INDEX_TEMPLATE = exports.UIM_POLICY_ATTACH_INDEX = exports.UIM_INDEX_RETRY_STEP = exports.UIM_EDIT_CLICK = exports.UIM_CONFIG_WARM_PHASE = exports.UIM_CONFIG_SET_PRIORITY = exports.UIM_CONFIG_COLD_PHASE = exports.UIM_APP_NAME = exports.UIM_APP_LOAD = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UIM_APP_NAME = exports.UIM_APP_NAME = 'index_lifecycle_management';
const UIM_APP_LOAD = exports.UIM_APP_LOAD = 'app_load';
const UIM_POLICY_CREATE = exports.UIM_POLICY_CREATE = 'policy_create';
const UIM_POLICY_UPDATE = exports.UIM_POLICY_UPDATE = 'policy_update';
const UIM_POLICY_DELETE = exports.UIM_POLICY_DELETE = 'policy_delete';
const UIM_POLICY_ATTACH_INDEX = exports.UIM_POLICY_ATTACH_INDEX = 'policy_attach_index';
const UIM_POLICY_ATTACH_INDEX_TEMPLATE = exports.UIM_POLICY_ATTACH_INDEX_TEMPLATE = 'policy_attach_index_template';
const UIM_POLICY_DETACH_INDEX = exports.UIM_POLICY_DETACH_INDEX = 'policy_detach_index';
const UIM_CONFIG_COLD_PHASE = exports.UIM_CONFIG_COLD_PHASE = 'config_cold_phase';
const UIM_CONFIG_WARM_PHASE = exports.UIM_CONFIG_WARM_PHASE = 'config_warm_phase';
const UIM_CONFIG_SET_PRIORITY = exports.UIM_CONFIG_SET_PRIORITY = 'config_set_priority';
const UIM_INDEX_RETRY_STEP = exports.UIM_INDEX_RETRY_STEP = 'index_retry_step';
const UIM_EDIT_CLICK = exports.UIM_EDIT_CLICK = 'edit_click';
