// @kbn/ui-shared-deps-src/src/entry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Uuid = exports.TsLib = exports.StyledComponents = exports.SaferLodashSet = exports.Rxjs = exports.Reselect = exports.ReduxjsToolkit = exports.Redux = exports.ReactRouterDomV5Compat = exports.ReactRouterDom = exports.ReactRouter = exports.ReactRedux = exports.ReactQueryDevtools = exports.ReactQuery = exports.ReactDomServer = exports.ReactDom = exports.React = exports.MonacoBarePluginApi = exports.MomentTimezone = exports.Moment = exports.LodashFp = exports.Lodash = exports.KbnUiTheme = exports.KbnStd = exports.KbnSharedUxErrorBoundary = exports.KbnSearchErrors = exports.KbnRison = exports.KbnMonaco = exports.KbnI18nReact = exports.KbnI18n = exports.KbnEsqlAst = exports.KbnEsQuery = exports.KbnDatemath = exports.KbnCodeEditor = exports.KbnAnalytics = exports.Jquery = exports.IoTs = exports.Immer = exports.History = exports.HelloPangeaDnd = exports.Fflate = exports.EmotionReact = exports.EmotionCache = exports.ElasticNumeral = exports.ElasticEuiLibServicesFormat = exports.ElasticEuiLibServices = exports.ElasticEuiLibComponentsUseIsNestedEuiProvider = exports.ElasticEui = exports.ElasticCharts = exports.Classnames = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

require('./polyfills');
const Jquery = exports.Jquery = require('jquery');
window.$ = window.jQuery = Jquery;

// stateful deps
const KbnUiTheme = exports.KbnUiTheme = require('@kbn/ui-theme');
const KbnI18n = exports.KbnI18n = require('@kbn/i18n');
const KbnI18nReact = exports.KbnI18nReact = require('@kbn/i18n-react');
const EmotionCache = exports.EmotionCache = require('@emotion/cache');
const EmotionReact = exports.EmotionReact = require('@emotion/react');
const Moment = exports.Moment = require('moment');
const MomentTimezone = exports.MomentTimezone = require('moment-timezone/moment-timezone');
const IoTs = exports.IoTs = require('io-ts');
const KbnMonaco = exports.KbnMonaco = require('@kbn/monaco');
const MonacoBarePluginApi = exports.MonacoBarePluginApi = require('@kbn/monaco').BarePluginApi;
const React = exports.React = require('react');
const ReactDom = exports.ReactDom = require('react-dom');
const ReactDomServer = exports.ReactDomServer = require('react-dom/server');
// eslint-disable-next-line @kbn/eslint/module_migration
const ReactRouter = exports.ReactRouter = require('react-router');
const ReactRouterDom = exports.ReactRouterDom = require('react-router-dom');
const ReactRouterDomV5Compat = exports.ReactRouterDomV5Compat = require('react-router-dom-v5-compat');
const StyledComponents = exports.StyledComponents = require('styled-components');
Moment.tz.load(require('moment-timezone/data/packed/latest.json'));

// big deps which are locked to a single version
const Rxjs = exports.Rxjs = require('rxjs');
const ElasticNumeral = exports.ElasticNumeral = require('@elastic/numeral');
const ElasticCharts = exports.ElasticCharts = require('@elastic/charts');
const ElasticEui = exports.ElasticEui = require('@elastic/eui');
const ElasticEuiLibComponentsUseIsNestedEuiProvider = exports.ElasticEuiLibComponentsUseIsNestedEuiProvider = require('@elastic/eui/optimize/es/components/provider/nested');
const ElasticEuiLibServices = exports.ElasticEuiLibServices = require('@elastic/eui/optimize/es/services');
const ElasticEuiLibServicesFormat = exports.ElasticEuiLibServicesFormat = require('@elastic/eui/optimize/es/services/format');
const KbnDatemath = exports.KbnDatemath = require('@kbn/datemath');
const HelloPangeaDnd = exports.HelloPangeaDnd = require('@hello-pangea/dnd/dist/dnd');
const ReduxjsToolkit = exports.ReduxjsToolkit = require('@reduxjs/toolkit');
const ReactRedux = exports.ReactRedux = require('react-redux');
const Redux = exports.Redux = require('redux');
const Immer = exports.Immer = require('immer');
const Reselect = exports.Reselect = require('reselect');
const Lodash = exports.Lodash = require('lodash');
const LodashFp = exports.LodashFp = require('lodash/fp');
const {
  unzlibSync,
  strFromU8
} = require('fflate');
const Fflate = exports.Fflate = {
  unzlibSync,
  strFromU8
};

// runtime deps which don't need to be copied across all bundles
const TsLib = exports.TsLib = require('tslib');
const Uuid = exports.Uuid = require('uuid');
const KbnAnalytics = exports.KbnAnalytics = require('@kbn/analytics');
const KbnEsQuery = exports.KbnEsQuery = require('@kbn/es-query');
const KbnSearchErrors = exports.KbnSearchErrors = require('@kbn/search-errors');
const KbnStd = exports.KbnStd = require('@kbn/std');
const SaferLodashSet = exports.SaferLodashSet = require('@kbn/safer-lodash-set');
const KbnSharedUxErrorBoundary = exports.KbnSharedUxErrorBoundary = require('@kbn/shared-ux-error-boundary');
const KbnRison = exports.KbnRison = require('@kbn/rison');
const History = exports.History = require('history');
const Classnames = exports.Classnames = require('classnames');
const ReactQuery = exports.ReactQuery = require('@tanstack/react-query');
const ReactQueryDevtools = exports.ReactQueryDevtools = require('@tanstack/react-query-devtools');
const KbnCodeEditor = exports.KbnCodeEditor = require('@kbn/code-editor');
const KbnEsqlAst = exports.KbnEsqlAst = require('@kbn/esql-ast');
