// @kbn/fleet-plugin/server/constants/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AGENT_TYPE_PERMANENT: true,
  AGENT_TYPE_EPHEMERAL: true,
  AGENT_TYPE_TEMPORARY: true,
  AGENT_POLLING_THRESHOLD_MS: true,
  AGENT_POLLING_REQUEST_TIMEOUT_MARGIN_MS: true,
  AGENT_POLLING_INTERVAL: true,
  AGENT_UPDATE_LAST_CHECKIN_INTERVAL_MS: true,
  AGENT_POLICY_ROLLOUT_RATE_LIMIT_REQUEST_PER_INTERVAL: true,
  AGENT_POLICY_ROLLOUT_RATE_LIMIT_INTERVAL_MS: true,
  AGENT_UPDATE_ACTIONS_INTERVAL_MS: true,
  UNPRIVILEGED_AGENT_KUERY: true,
  PRIVILEGED_AGENT_KUERY: true,
  MAX_TIME_COMPLETE_INSTALL: true,
  REQUEST_DIAGNOSTICS_TIMEOUT_MS: true,
  LIMITED_CONCURRENCY_ROUTE_TAG: true,
  PLUGIN_ID: true,
  EPM_API_ROUTES: true,
  DATA_STREAM_API_ROUTES: true,
  PACKAGE_POLICY_API_ROUTES: true,
  AGENT_API_ROUTES: true,
  AGENT_POLICY_API_ROUTES: true,
  AGENTS_SETUP_API_ROUTES: true,
  ENROLLMENT_API_KEY_ROUTES: true,
  INSTALL_SCRIPT_API_ROUTES: true,
  OUTPUT_API_ROUTES: true,
  SETUP_API_ROUTE: true,
  SETTINGS_API_ROUTES: true,
  APP_API_ROUTES: true,
  PRECONFIGURATION_API_ROUTES: true,
  DOWNLOAD_SOURCE_API_ROOT: true,
  DOWNLOAD_SOURCE_API_ROUTES: true,
  CREATE_STANDALONE_AGENT_API_KEY_ROUTE: true,
  FLEET_DEBUG_ROUTES: true,
  INGEST_SAVED_OBJECT_INDEX: true,
  SO_SEARCH_LIMIT: true,
  AGENTS_PREFIX: true,
  AGENT_POLICY_SAVED_OBJECT_TYPE: true,
  PACKAGE_POLICY_SAVED_OBJECT_TYPE: true,
  OUTPUT_SAVED_OBJECT_TYPE: true,
  PACKAGES_SAVED_OBJECT_TYPE: true,
  ASSETS_SAVED_OBJECT_TYPE: true,
  GLOBAL_SETTINGS_SAVED_OBJECT_TYPE: true,
  MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE: true,
  UNINSTALL_TOKENS_SAVED_OBJECT_TYPE: true,
  DEFAULT_OUTPUT: true,
  DEFAULT_OUTPUT_ID: true,
  SERVERLESS_DEFAULT_OUTPUT_ID: true,
  PACKAGE_POLICY_DEFAULT_INDEX_PRIVILEGES: true,
  AGENT_POLICY_DEFAULT_MONITORING_DATASETS: true,
  ENROLLMENT_API_KEYS_INDEX: true,
  AGENTS_INDEX: true,
  PRECONFIGURATION_DELETION_RECORD_SAVED_OBJECT_TYPE: true,
  PRECONFIGURATION_LATEST_KEYWORD: true,
  AUTO_UPDATE_PACKAGES: true,
  USER_SETTINGS_TEMPLATE_SUFFIX: true,
  PACKAGE_TEMPLATE_SUFFIX: true,
  DEFAULT_DOWNLOAD_SOURCE_URI: true,
  DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE: true,
  DEFAULT_DOWNLOAD_SOURCE_ID: true,
  DEFAULT_FLEET_SERVER_HOST_ID: true,
  FLEET_SERVER_HOST_SAVED_OBJECT_TYPE: true,
  SERVERLESS_DEFAULT_FLEET_SERVER_HOST_ID: true,
  FLEET_SERVER_PACKAGE: true,
  FLEET_PROXY_SAVED_OBJECT_TYPE: true,
  ENDPOINT_PRIVILEGES: true,
  ENDPOINT_EXCEPTIONS_PRIVILEGES: true,
  MESSAGE_SIGNING_SERVICE_API_ROUTES: true,
  SECRETS_ENDPOINT_PATH: true,
  SECRETS_MINIMUM_FLEET_SERVER_VERSION: true,
  OUTPUT_SECRETS_MINIMUM_FLEET_SERVER_VERSION: true,
  OUTPUT_HEALTH_DATA_STREAM: true,
  FLEET_SETUP_LOCK_TYPE: true,
  STACK_COMPONENT_TEMPLATE_ECS_MAPPINGS: true,
  FLEET_GLOBALS_COMPONENT_TEMPLATE_NAME: true,
  FLEET_GLOBALS_COMPONENT_TEMPLATE_CONTENT: true,
  FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_NAME: true,
  FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_CONTENT: true,
  FLEET_COMPONENT_TEMPLATES: true,
  FLEET_FINAL_PIPELINE_ID: true,
  FLEET_FINAL_PIPELINE_CONTENT: true,
  FLEET_FINAL_PIPELINE_VERSION: true,
  FLEET_INSTALL_FORMAT_VERSION: true,
  FLEET_AGENT_POLICIES_SCHEMA_VERSION: true,
  STACK_COMPONENT_TEMPLATE_LOGS_SETTINGS: true,
  STACK_COMPONENT_TEMPLATE_METRICS_SETTINGS: true,
  STACK_COMPONENT_TEMPLATE_METRICS_TSDB_SETTINGS: true,
  STACK_COMPONENT_TEMPLATES: true,
  FILE_STORAGE_DATA_AGENT_INDEX: true,
  FILE_STORAGE_METADATA_AGENT_INDEX: true
};
Object.defineProperty(exports, "AGENTS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_INDEX;
  }
});
Object.defineProperty(exports, "AGENTS_PREFIX", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_PREFIX;
  }
});
Object.defineProperty(exports, "AGENTS_SETUP_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENTS_SETUP_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_API_ROUTES;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_DEFAULT_MONITORING_DATASETS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_DEFAULT_MONITORING_DATASETS;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_ROLLOUT_RATE_LIMIT_INTERVAL_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_ROLLOUT_RATE_LIMIT_INTERVAL_MS;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_ROLLOUT_RATE_LIMIT_REQUEST_PER_INTERVAL", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_ROLLOUT_RATE_LIMIT_REQUEST_PER_INTERVAL;
  }
});
Object.defineProperty(exports, "AGENT_POLICY_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLICY_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "AGENT_POLLING_INTERVAL", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLLING_INTERVAL;
  }
});
Object.defineProperty(exports, "AGENT_POLLING_REQUEST_TIMEOUT_MARGIN_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLLING_REQUEST_TIMEOUT_MARGIN_MS;
  }
});
Object.defineProperty(exports, "AGENT_POLLING_THRESHOLD_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_POLLING_THRESHOLD_MS;
  }
});
Object.defineProperty(exports, "AGENT_TYPE_EPHEMERAL", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_TYPE_EPHEMERAL;
  }
});
Object.defineProperty(exports, "AGENT_TYPE_PERMANENT", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_TYPE_PERMANENT;
  }
});
Object.defineProperty(exports, "AGENT_TYPE_TEMPORARY", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_TYPE_TEMPORARY;
  }
});
Object.defineProperty(exports, "AGENT_UPDATE_ACTIONS_INTERVAL_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_UPDATE_ACTIONS_INTERVAL_MS;
  }
});
Object.defineProperty(exports, "AGENT_UPDATE_LAST_CHECKIN_INTERVAL_MS", {
  enumerable: true,
  get: function () {
    return _constants.AGENT_UPDATE_LAST_CHECKIN_INTERVAL_MS;
  }
});
Object.defineProperty(exports, "APP_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.APP_API_ROUTES;
  }
});
Object.defineProperty(exports, "ASSETS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.ASSETS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "AUTO_UPDATE_PACKAGES", {
  enumerable: true,
  get: function () {
    return _constants.AUTO_UPDATE_PACKAGES;
  }
});
Object.defineProperty(exports, "CREATE_STANDALONE_AGENT_API_KEY_ROUTE", {
  enumerable: true,
  get: function () {
    return _constants.CREATE_STANDALONE_AGENT_API_KEY_ROUTE;
  }
});
Object.defineProperty(exports, "DATA_STREAM_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.DATA_STREAM_API_ROUTES;
  }
});
Object.defineProperty(exports, "DEFAULT_DOWNLOAD_SOURCE_ID", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_DOWNLOAD_SOURCE_ID;
  }
});
Object.defineProperty(exports, "DEFAULT_DOWNLOAD_SOURCE_URI", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_DOWNLOAD_SOURCE_URI;
  }
});
Object.defineProperty(exports, "DEFAULT_FLEET_SERVER_HOST_ID", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_FLEET_SERVER_HOST_ID;
  }
});
Object.defineProperty(exports, "DEFAULT_OUTPUT", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_OUTPUT;
  }
});
Object.defineProperty(exports, "DEFAULT_OUTPUT_ID", {
  enumerable: true,
  get: function () {
    return _constants.DEFAULT_OUTPUT_ID;
  }
});
Object.defineProperty(exports, "DOWNLOAD_SOURCE_API_ROOT", {
  enumerable: true,
  get: function () {
    return _constants.DOWNLOAD_SOURCE_API_ROOT;
  }
});
Object.defineProperty(exports, "DOWNLOAD_SOURCE_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.DOWNLOAD_SOURCE_API_ROUTES;
  }
});
Object.defineProperty(exports, "DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.DOWNLOAD_SOURCE_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "ENDPOINT_EXCEPTIONS_PRIVILEGES", {
  enumerable: true,
  get: function () {
    return _constants.ENDPOINT_EXCEPTIONS_PRIVILEGES;
  }
});
Object.defineProperty(exports, "ENDPOINT_PRIVILEGES", {
  enumerable: true,
  get: function () {
    return _constants.ENDPOINT_PRIVILEGES;
  }
});
Object.defineProperty(exports, "ENROLLMENT_API_KEYS_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.ENROLLMENT_API_KEYS_INDEX;
  }
});
Object.defineProperty(exports, "ENROLLMENT_API_KEY_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.ENROLLMENT_API_KEY_ROUTES;
  }
});
Object.defineProperty(exports, "EPM_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.EPM_API_ROUTES;
  }
});
Object.defineProperty(exports, "FILE_STORAGE_DATA_AGENT_INDEX", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FILE_STORAGE_DATA_AGENT_INDEX;
  }
});
Object.defineProperty(exports, "FILE_STORAGE_METADATA_AGENT_INDEX", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FILE_STORAGE_METADATA_AGENT_INDEX;
  }
});
Object.defineProperty(exports, "FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_CONTENT", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_CONTENT;
  }
});
Object.defineProperty(exports, "FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_NAME", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_AGENT_ID_VERIFY_COMPONENT_TEMPLATE_NAME;
  }
});
Object.defineProperty(exports, "FLEET_AGENT_POLICIES_SCHEMA_VERSION", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_AGENT_POLICIES_SCHEMA_VERSION;
  }
});
Object.defineProperty(exports, "FLEET_COMPONENT_TEMPLATES", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_COMPONENT_TEMPLATES;
  }
});
Object.defineProperty(exports, "FLEET_DEBUG_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_DEBUG_ROUTES;
  }
});
Object.defineProperty(exports, "FLEET_FINAL_PIPELINE_CONTENT", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_FINAL_PIPELINE_CONTENT;
  }
});
Object.defineProperty(exports, "FLEET_FINAL_PIPELINE_ID", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_FINAL_PIPELINE_ID;
  }
});
Object.defineProperty(exports, "FLEET_FINAL_PIPELINE_VERSION", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_FINAL_PIPELINE_VERSION;
  }
});
Object.defineProperty(exports, "FLEET_GLOBALS_COMPONENT_TEMPLATE_CONTENT", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_GLOBALS_COMPONENT_TEMPLATE_CONTENT;
  }
});
Object.defineProperty(exports, "FLEET_GLOBALS_COMPONENT_TEMPLATE_NAME", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_GLOBALS_COMPONENT_TEMPLATE_NAME;
  }
});
Object.defineProperty(exports, "FLEET_INSTALL_FORMAT_VERSION", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.FLEET_INSTALL_FORMAT_VERSION;
  }
});
Object.defineProperty(exports, "FLEET_PROXY_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_PROXY_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "FLEET_SERVER_HOST_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SERVER_HOST_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "FLEET_SERVER_PACKAGE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SERVER_PACKAGE;
  }
});
Object.defineProperty(exports, "FLEET_SETUP_LOCK_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.FLEET_SETUP_LOCK_TYPE;
  }
});
Object.defineProperty(exports, "GLOBAL_SETTINGS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.GLOBAL_SETTINGS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "INGEST_SAVED_OBJECT_INDEX", {
  enumerable: true,
  get: function () {
    return _constants.INGEST_SAVED_OBJECT_INDEX;
  }
});
Object.defineProperty(exports, "INSTALL_SCRIPT_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.INSTALL_SCRIPT_API_ROUTES;
  }
});
Object.defineProperty(exports, "LIMITED_CONCURRENCY_ROUTE_TAG", {
  enumerable: true,
  get: function () {
    return _constants.LIMITED_CONCURRENCY_ROUTE_TAG;
  }
});
Object.defineProperty(exports, "MAX_TIME_COMPLETE_INSTALL", {
  enumerable: true,
  get: function () {
    return _constants.MAX_TIME_COMPLETE_INSTALL;
  }
});
Object.defineProperty(exports, "MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.MESSAGE_SIGNING_KEYS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "MESSAGE_SIGNING_SERVICE_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.MESSAGE_SIGNING_SERVICE_API_ROUTES;
  }
});
Object.defineProperty(exports, "OUTPUT_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.OUTPUT_API_ROUTES;
  }
});
Object.defineProperty(exports, "OUTPUT_HEALTH_DATA_STREAM", {
  enumerable: true,
  get: function () {
    return _constants.OUTPUT_HEALTH_DATA_STREAM;
  }
});
Object.defineProperty(exports, "OUTPUT_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.OUTPUT_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "OUTPUT_SECRETS_MINIMUM_FLEET_SERVER_VERSION", {
  enumerable: true,
  get: function () {
    return _constants.OUTPUT_SECRETS_MINIMUM_FLEET_SERVER_VERSION;
  }
});
Object.defineProperty(exports, "PACKAGES_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGES_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_API_ROUTES;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_DEFAULT_INDEX_PRIVILEGES", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_DEFAULT_INDEX_PRIVILEGES;
  }
});
Object.defineProperty(exports, "PACKAGE_POLICY_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_POLICY_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PACKAGE_TEMPLATE_SUFFIX", {
  enumerable: true,
  get: function () {
    return _constants.PACKAGE_TEMPLATE_SUFFIX;
  }
});
Object.defineProperty(exports, "PLUGIN_ID", {
  enumerable: true,
  get: function () {
    return _constants.PLUGIN_ID;
  }
});
Object.defineProperty(exports, "PRECONFIGURATION_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.PRECONFIGURATION_API_ROUTES;
  }
});
Object.defineProperty(exports, "PRECONFIGURATION_DELETION_RECORD_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.PRECONFIGURATION_DELETION_RECORD_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "PRECONFIGURATION_LATEST_KEYWORD", {
  enumerable: true,
  get: function () {
    return _constants.PRECONFIGURATION_LATEST_KEYWORD;
  }
});
Object.defineProperty(exports, "PRIVILEGED_AGENT_KUERY", {
  enumerable: true,
  get: function () {
    return _constants.PRIVILEGED_AGENT_KUERY;
  }
});
Object.defineProperty(exports, "REQUEST_DIAGNOSTICS_TIMEOUT_MS", {
  enumerable: true,
  get: function () {
    return _constants.REQUEST_DIAGNOSTICS_TIMEOUT_MS;
  }
});
Object.defineProperty(exports, "SECRETS_ENDPOINT_PATH", {
  enumerable: true,
  get: function () {
    return _constants.SECRETS_ENDPOINT_PATH;
  }
});
Object.defineProperty(exports, "SECRETS_MINIMUM_FLEET_SERVER_VERSION", {
  enumerable: true,
  get: function () {
    return _constants.SECRETS_MINIMUM_FLEET_SERVER_VERSION;
  }
});
Object.defineProperty(exports, "SERVERLESS_DEFAULT_FLEET_SERVER_HOST_ID", {
  enumerable: true,
  get: function () {
    return _constants.SERVERLESS_DEFAULT_FLEET_SERVER_HOST_ID;
  }
});
Object.defineProperty(exports, "SERVERLESS_DEFAULT_OUTPUT_ID", {
  enumerable: true,
  get: function () {
    return _constants.SERVERLESS_DEFAULT_OUTPUT_ID;
  }
});
Object.defineProperty(exports, "SETTINGS_API_ROUTES", {
  enumerable: true,
  get: function () {
    return _constants.SETTINGS_API_ROUTES;
  }
});
Object.defineProperty(exports, "SETUP_API_ROUTE", {
  enumerable: true,
  get: function () {
    return _constants.SETUP_API_ROUTE;
  }
});
Object.defineProperty(exports, "SO_SEARCH_LIMIT", {
  enumerable: true,
  get: function () {
    return _constants.SO_SEARCH_LIMIT;
  }
});
Object.defineProperty(exports, "STACK_COMPONENT_TEMPLATES", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.STACK_COMPONENT_TEMPLATES;
  }
});
Object.defineProperty(exports, "STACK_COMPONENT_TEMPLATE_ECS_MAPPINGS", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.STACK_COMPONENT_TEMPLATE_ECS_MAPPINGS;
  }
});
Object.defineProperty(exports, "STACK_COMPONENT_TEMPLATE_LOGS_SETTINGS", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.STACK_COMPONENT_TEMPLATE_LOGS_SETTINGS;
  }
});
Object.defineProperty(exports, "STACK_COMPONENT_TEMPLATE_METRICS_SETTINGS", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.STACK_COMPONENT_TEMPLATE_METRICS_SETTINGS;
  }
});
Object.defineProperty(exports, "STACK_COMPONENT_TEMPLATE_METRICS_TSDB_SETTINGS", {
  enumerable: true,
  get: function () {
    return _fleet_es_assets.STACK_COMPONENT_TEMPLATE_METRICS_TSDB_SETTINGS;
  }
});
Object.defineProperty(exports, "UNINSTALL_TOKENS_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.UNINSTALL_TOKENS_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "UNPRIVILEGED_AGENT_KUERY", {
  enumerable: true,
  get: function () {
    return _constants.UNPRIVILEGED_AGENT_KUERY;
  }
});
Object.defineProperty(exports, "USER_SETTINGS_TEMPLATE_SUFFIX", {
  enumerable: true,
  get: function () {
    return _constants.USER_SETTINGS_TEMPLATE_SUFFIX;
  }
});
var _constants = require("../../common/constants");
var _fleet_es_assets = require("./fleet_es_assets");
var _mappings = require("../../common/constants/mappings");
Object.keys(_mappings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mappings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mappings[key];
    }
  });
});
