// @kbn/security-solution-plugin/server/lib/timeline/routes/timelines/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "copyTimelineRoute", {
  enumerable: true,
  get: function () {
    return _copy_timeline.copyTimelineRoute;
  }
});
Object.defineProperty(exports, "createTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _create_timelines.createTimelinesRoute;
  }
});
Object.defineProperty(exports, "deleteTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _delete_timelines.deleteTimelinesRoute;
  }
});
Object.defineProperty(exports, "exportTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _export_timelines.exportTimelinesRoute;
  }
});
Object.defineProperty(exports, "getTimelineRoute", {
  enumerable: true,
  get: function () {
    return _get_timeline.getTimelineRoute;
  }
});
Object.defineProperty(exports, "getTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _get_timelines.getTimelinesRoute;
  }
});
Object.defineProperty(exports, "importTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _import_timelines.importTimelinesRoute;
  }
});
Object.defineProperty(exports, "patchTimelinesRoute", {
  enumerable: true,
  get: function () {
    return _patch_timelines.patchTimelinesRoute;
  }
});
Object.defineProperty(exports, "persistFavoriteRoute", {
  enumerable: true,
  get: function () {
    return _persist_favorite.persistFavoriteRoute;
  }
});
Object.defineProperty(exports, "resolveTimelineRoute", {
  enumerable: true,
  get: function () {
    return _resolve_timeline.resolveTimelineRoute;
  }
});
var _create_timelines = require("./create_timelines");
var _delete_timelines = require("./delete_timelines");
var _export_timelines = require("./export_timelines");
var _get_timeline = require("./get_timeline");
var _get_timelines = require("./get_timelines");
var _import_timelines = require("./import_timelines");
var _patch_timelines = require("./patch_timelines");
var _persist_favorite = require("./persist_favorite");
var _resolve_timeline = require("./resolve_timeline");
var _copy_timeline = require("./copy_timeline");
