// @kbn/security-solution-plugin/public/common/components/chart_settings_popover/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartSettingsPopover = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _inspect = require("../inspect");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ChartSettingsPopoverComponent = ({
  initialPanelId,
  isPopoverOpen,
  panels,
  setIsPopoverOpen
}) => {
  const onButtonClick = (0, _react.useCallback)(() => setIsPopoverOpen(isOpen => !isOpen), [setIsPopoverOpen]);
  const closePopover = (0, _react.useCallback)(() => setIsPopoverOpen(false), [setIsPopoverOpen]);
  const button = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    "aria-label": i18n.CHART_SETTINGS_POPOVER_ARIA_LABEL,
    color: "text",
    iconType: "boxesHorizontal",
    onClick: onButtonClick,
    size: "xs"
  }), [onButtonClick]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
    anchorPosition: "downCenter",
    button: button,
    className: _inspect.BUTTON_CLASS,
    closePopover: closePopover,
    isOpen: isPopoverOpen,
    panelPaddingSize: "none"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiContextMenu, {
    initialPanelId: initialPanelId,
    panels: panels
  }));
};
ChartSettingsPopoverComponent.displayName = 'ChartSettingsPopoverComponent';
const ChartSettingsPopover = exports.ChartSettingsPopover = /*#__PURE__*/_react.default.memo(ChartSettingsPopoverComponent);
