// @kbn/data-visualizer-plugin/public/application/common/components/stats_table/components/expanded_row_field_header/expanded_row_field_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandedRowFieldHeader = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ExpandedRowFieldHeader = ({
  children
}) => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const dvExpandedRowFieldHeader = (0, _react2.css)({
    textTransform: 'uppercase',
    textAlign: 'left',
    color: euiTheme.colors.darkShade,
    fontWeight: 'bold',
    paddingBottom: euiTheme.size.s
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs",
    color: 'subdued',
    css: dvExpandedRowFieldHeader,
    textAlign: 'center'
  }, children);
};
exports.ExpandedRowFieldHeader = ExpandedRowFieldHeader;
