// @kbn/data-view-field-editor-plugin/public/components/field_format_editor/samples/samples.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormatEditorSamples = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
require("./samples.scss");
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

class FormatEditorSamples extends _react.PureComponent {
  render() {
    const {
      samples,
      sampleType
    } = this.props;
    const columns = [{
      field: 'input',
      name: _i18n.i18n.translate('indexPatternFieldEditor.samples.inputHeader', {
        defaultMessage: 'Input'
      }),
      render: input => {
        return typeof input === 'object' ? JSON.stringify(input) : input;
      }
    }, {
      field: 'output',
      name: _i18n.i18n.translate('indexPatternFieldEditor.samples.outputHeader', {
        defaultMessage: 'Output'
      }),
      render: output => {
        return sampleType === 'html' ? /*#__PURE__*/_react.default.createElement("div", {
          /*
           * Justification for dangerouslySetInnerHTML:
           * Sample output may contain HTML tags, like URL image/audio format.
           */
          dangerouslySetInnerHTML: {
            __html: output
          } // eslint-disable-line react/no-danger
        }) : /*#__PURE__*/_react.default.createElement("div", null, output);
      }
    }];
    return samples.length ? /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
      label: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternFieldEditor.samplesHeader",
        defaultMessage: "Samples"
      })
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiBasicTable, {
      className: "kbnFieldFormatEditor__samples",
      compressed: true,
      items: samples,
      columns: columns
    })) : null;
  }
}
exports.FormatEditorSamples = FormatEditorSamples;
(0, _defineProperty2.default)(FormatEditorSamples, "defaultProps", {
  sampleType: 'text'
});
