// @kbn/cases-plugin/public/components/connectors/fields_preview_form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConnectorFieldsPreviewForm = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _ = require(".");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ConnectorFieldsFormPreviewComponent = ({
  connector,
  fields
}) => {
  const {
    caseConnectorsRegistry
  } = (0, _.getCaseConnectors)();
  if (connector == null || connector.actionTypeId == null || connector.actionTypeId === '.none') {
    return null;
  }
  const {
    previewComponent: PreviewComponent
  } = caseConnectorsRegistry.get(connector.actionTypeId);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, PreviewComponent != null ? /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      justifyContent: "center"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
      size: "m"
    })))
  }, /*#__PURE__*/_react.default.createElement("div", {
    "data-test-subj": 'connector-fields-preview'
  }, /*#__PURE__*/_react.default.createElement(PreviewComponent, {
    connector: connector,
    fields: fields,
    key: connector.id
  }))) : null);
};
ConnectorFieldsFormPreviewComponent.displayName = 'ConnectorFieldsFormPreview';
const ConnectorFieldsPreviewForm = exports.ConnectorFieldsPreviewForm = /*#__PURE__*/(0, _react.memo)(ConnectorFieldsFormPreviewComponent);
