// @kbn/data-visualizer-plugin/public/application/index_data_visualizer/constants/index_data_visualizer_viewer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MAX_CONCURRENT_REQUESTS = exports.DATA_VISUALIZER_INDEX_VIEWER_ID = exports.DATA_VISUALIZER_INDEX_VIEWER = void 0;
exports.getDefaultPageState = getDefaultPageState;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DATA_VISUALIZER_INDEX_VIEWER = exports.DATA_VISUALIZER_INDEX_VIEWER = 'DATA_VISUALIZER_INDEX_VIEWER';
const DATA_VISUALIZER_INDEX_VIEWER_ID = exports.DATA_VISUALIZER_INDEX_VIEWER_ID = 'index_data_visualizer';
const MAX_CONCURRENT_REQUESTS = exports.MAX_CONCURRENT_REQUESTS = 10;
function getDefaultPageState() {
  return {
    overallStats: {
      totalCount: 0,
      aggregatableExistsFields: [],
      aggregatableNotExistsFields: [],
      nonAggregatableExistsFields: [],
      nonAggregatableNotExistsFields: []
    },
    metricConfigs: [],
    totalMetricFieldCount: 0,
    populatedMetricFieldCount: 0,
    metricsLoaded: false,
    nonMetricConfigs: [],
    nonMetricsLoaded: false,
    documentCountStats: undefined
  };
}
