// @kbn/stack-connectors-plugin/public/connector_types/lib/servicenow/selection_row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServiceNowSelectableRowIcon = exports.default = ServiceNowSelectableRowIcon;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _public = require("@kbn/triggers-actions-ui-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// eslint-disable-next-line import/no-default-export

function ServiceNowSelectableRowIcon({
  actionConnector
}) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    "aria-label": deprecatedTooltipTitle,
    type: "warning",
    color: "warning",
    content: _public.connectorDeprecatedMessage,
    "data-test-subj": `deprecated-connector-icon-${actionConnector.id}`,
    anchorClassName: 'euiFormControlLayout__prepend'
  });
}
const deprecatedTooltipTitle = _i18n.i18n.translate('xpack.stackConnectors.components.serviceNow.deprecatedTooltipTitle', {
  defaultMessage: 'Deprecated connector'
});
