// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/pick_fields_step/components/categorization_detector/categorization_detector.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CategorizationDetector = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _mlAnomalyUtils = require("@kbn/ml-anomaly-utils");
var _job_creator_context = require("../../../job_creator_context");
var _detector_cards = require("./detector_cards");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CategorizationDetector = () => {
  const {
    jobCreator: jc,
    jobCreatorUpdate,
    jobCreatorUpdated
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const jobCreator = jc;
  const [categorizationDetectorType, setCategorizationDetectorType] = (0, _react.useState)(jobCreator.selectedDetectorType);
  (0, _react.useEffect)(() => {
    if (categorizationDetectorType !== jobCreator.selectedDetectorType) {
      jobCreator.setDetectorType(categorizationDetectorType);
      jobCreatorUpdate();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [categorizationDetectorType]);
  (0, _react.useEffect)(() => {
    setCategorizationDetectorType(jobCreator.selectedDetectorType);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreatorUpdated]);
  function onCountSelection() {
    setCategorizationDetectorType(_mlAnomalyUtils.ML_JOB_AGGREGATION.COUNT);
  }
  function onHighCountSelection() {
    setCategorizationDetectorType(_mlAnomalyUtils.ML_JOB_AGGREGATION.HIGH_COUNT);
  }
  function onRareSelection() {
    setCategorizationDetectorType(_mlAnomalyUtils.ML_JOB_AGGREGATION.RARE);
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("h3", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.newJob.wizard.pickFieldsStep.categorizationDetectorSelect.title",
    defaultMessage: "Categorization detector"
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "l",
    css: {
      maxWidth: '1100px'
    }
  }, /*#__PURE__*/_react.default.createElement(_detector_cards.CountCard, {
    onClick: onCountSelection,
    isSelected: categorizationDetectorType === _mlAnomalyUtils.ML_JOB_AGGREGATION.COUNT
  }), /*#__PURE__*/_react.default.createElement(_detector_cards.HighCountCard, {
    onClick: onHighCountSelection,
    isSelected: categorizationDetectorType === _mlAnomalyUtils.ML_JOB_AGGREGATION.HIGH_COUNT
  }), /*#__PURE__*/_react.default.createElement(_detector_cards.RareCard, {
    onClick: onRareSelection,
    isSelected: categorizationDetectorType === _mlAnomalyUtils.ML_JOB_AGGREGATION.RARE
  })));
};
exports.CategorizationDetector = CategorizationDetector;
