// @kbn/exploratory-view-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ALL_VALUES_SELECTED", {
  enumerable: true,
  get: function () {
    return _url_constants.ALL_VALUES_SELECTED;
  }
});
Object.defineProperty(exports, "ENVIRONMENT_ALL", {
  enumerable: true,
  get: function () {
    return _constants2.ENVIRONMENT_ALL;
  }
});
Object.defineProperty(exports, "EXPLORATORY_VIEW_APP_URL", {
  enumerable: true,
  get: function () {
    return _constants.APP_ROUTE;
  }
});
Object.defineProperty(exports, "ExploratoryView", {
  enumerable: true,
  get: function () {
    return _shared.ExploratoryView;
  }
});
Object.defineProperty(exports, "FILTER_RECORDS", {
  enumerable: true,
  get: function () {
    return _constants2.FILTER_RECORDS;
  }
});
Object.defineProperty(exports, "FilterValueLabel", {
  enumerable: true,
  get: function () {
    return _shared.FilterValueLabel;
  }
});
Object.defineProperty(exports, "OPERATION_COLUMN", {
  enumerable: true,
  get: function () {
    return _constants2.OPERATION_COLUMN;
  }
});
Object.defineProperty(exports, "RECORDS_FIELD", {
  enumerable: true,
  get: function () {
    return _constants2.RECORDS_FIELD;
  }
});
Object.defineProperty(exports, "RECORDS_PERCENTAGE_FIELD", {
  enumerable: true,
  get: function () {
    return _constants2.RECORDS_PERCENTAGE_FIELD;
  }
});
Object.defineProperty(exports, "REPORT_METRIC_FIELD", {
  enumerable: true,
  get: function () {
    return _constants2.REPORT_METRIC_FIELD;
  }
});
Object.defineProperty(exports, "ReportTypes", {
  enumerable: true,
  get: function () {
    return _constants2.ReportTypes;
  }
});
Object.defineProperty(exports, "SelectableUrlList", {
  enumerable: true,
  get: function () {
    return _shared.SelectableUrlList;
  }
});
Object.defineProperty(exports, "TERMS_COLUMN", {
  enumerable: true,
  get: function () {
    return _constants2.TERMS_COLUMN;
  }
});
Object.defineProperty(exports, "USE_BREAK_DOWN_COLUMN", {
  enumerable: true,
  get: function () {
    return _constants2.USE_BREAK_DOWN_COLUMN;
  }
});
Object.defineProperty(exports, "createExploratoryViewUrl", {
  enumerable: true,
  get: function () {
    return _exploratory_view_url.createExploratoryViewUrl;
  }
});
Object.defineProperty(exports, "fromQuery", {
  enumerable: true,
  get: function () {
    return _url.fromQuery;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "toQuery", {
  enumerable: true,
  get: function () {
    return _url.toQuery;
  }
});
var _plugin = require("./plugin");
var _url_constants = require("./components/shared/exploratory_view/configurations/constants/url_constants");
var _constants = require("./constants");
var _shared = require("./components/shared");
var _exploratory_view_url = require("./components/shared/exploratory_view/configurations/exploratory_view_url");
var _constants2 = require("./components/shared/exploratory_view/configurations/constants");
var _url = require("./utils/url");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const plugin = initializerContext => {
  return new _plugin.Plugin(initializerContext);
};
exports.plugin = plugin;
