// @kbn/transform-plugin/public/app/sections/transform_management/components/transform_list/expanded_row_json_pane.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExpandedRowJsonPane = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ExpandedRowJsonPane = ({
  json
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-test-subj": "transformJsonTabContent"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
    "aria-label": _i18n.i18n.translate('xpack.transform.transformList.transformDetails.expandedRowJsonPane', {
      defaultMessage: 'JSON of transform configuration'
    }),
    fontSize: "s",
    language: "json",
    paddingSize: "s",
    style: {
      width: '100%'
    },
    isCopyable: true
  }, JSON.stringify(json, null, 2))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, "\xA0")));
};
exports.ExpandedRowJsonPane = ExpandedRowJsonPane;
