// @kbn/security-solution-plugin/public/explore/network/components/network_top_n_flow_table/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NetworkTopNFlowTableId = exports.NetworkTopNFlowTable = void 0;
var _fp = require("lodash/fp");
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _fastDeepEqual = _interopRequireDefault(require("fast-deep-equal"));
var _search_strategy = require("../../../../../common/search_strategy");
var _use_selector = require("../../../../common/hooks/use_selector");
var _paginated_table = require("../../../components/paginated_table");
var _store = require("../../store");
var _columns = require("./columns");
var i18n = _interopRequireWildcard(require("./translations"));
var _helpers = require("../../../components/paginated_table/helpers");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const rowItems = [{
  text: i18n.ROWS_5,
  numberOfRow: 5
}, {
  text: i18n.ROWS_10,
  numberOfRow: 10
}];
const NetworkTopNFlowTableId = exports.NetworkTopNFlowTableId = 'networkTopSourceFlow-top-talkers';
const NetworkTopNFlowTableComponent = ({
  data,
  flowTargeted,
  id,
  isInspect,
  loading,
  loadPage,
  setQuerySkip,
  totalCount,
  type
}) => {
  const dispatch = (0, _reactRedux.useDispatch)();
  const getTopNFlowSelector = (0, _react.useMemo)(() => _store.networkSelectors.topNFlowSelector(), []);
  const {
    activePage,
    limit,
    sort
  } = (0, _use_selector.useDeepEqualSelector)(state => getTopNFlowSelector(state, type, flowTargeted));
  const columns = (0, _react.useMemo)(() => (0, _columns.getNFlowColumnsCurated)(flowTargeted, type, NetworkTopNFlowTableId), [flowTargeted, type]);
  let tableType;
  const headerTitle = flowTargeted === _search_strategy.FlowTargetSourceDest.source ? i18n.SOURCE_IP : i18n.DESTINATION_IP;
  if (type === _store.networkModel.NetworkType.page) {
    tableType = flowTargeted === _search_strategy.FlowTargetSourceDest.source ? _store.networkModel.NetworkTableType.topNFlowSource : _store.networkModel.NetworkTableType.topNFlowDestination;
  } else {
    tableType = flowTargeted === _search_strategy.FlowTargetSourceDest.source ? _store.networkModel.NetworkDetailsTableType.topNFlowSource : _store.networkModel.NetworkDetailsTableType.topNFlowDestination;
  }
  const onChange = (0, _react.useCallback)(criteria => {
    if (criteria.sort != null) {
      const splitField = criteria.sort.field.split('.');
      const field = (0, _fp.last)(splitField);
      const newSortDirection = field !== sort.field ? _search_strategy.Direction.desc : criteria.sort.direction; // sort by desc on init click
      const newTopNFlowSort = {
        field: field,
        direction: newSortDirection
      };
      if (!(0, _fastDeepEqual.default)(newTopNFlowSort, sort)) {
        dispatch(_store.networkActions.updateNetworkTable({
          networkType: type,
          tableType,
          updates: {
            sort: newTopNFlowSort
          }
        }));
      }
    }
  }, [sort, dispatch, type, tableType]);
  const sorting = (0, _react.useMemo)(() => ({
    field: sort.field === _search_strategy.NetworkTopTablesFields.bytes_out || sort.field === _search_strategy.NetworkTopTablesFields.bytes_in ? `node.network.${sort.field}` : `node.${flowTargeted}.${sort.field}`,
    direction: sort.direction
  }), [flowTargeted, sort]);
  const updateActivePage = (0, _react.useCallback)(newPage => dispatch(_store.networkActions.updateNetworkTable({
    networkType: type,
    tableType,
    updates: {
      activePage: newPage
    }
  })), [dispatch, type, tableType]);
  const updateLimitPagination = (0, _react.useCallback)(newLimit => dispatch(_store.networkActions.updateNetworkTable({
    networkType: type,
    tableType,
    updates: {
      limit: newLimit
    }
  })), [dispatch, type, tableType]);

  // limits pagination to 5 pages with 1 page increments to prevent expensive queries
  const limitedPaginationTotalCount = (0, _react.useMemo)(() => (0, _helpers.getLimitedPaginationTotalCount)({
    activePage,
    totalCount,
    limit
  }), [activePage, limit, totalCount]);
  const showMorePagesIndicator = (0, _react.useMemo)(() => {
    return totalCount / limit > limitedPaginationTotalCount / limit;
  }, [totalCount, limitedPaginationTotalCount, limit]);
  return /*#__PURE__*/_react.default.createElement(_paginated_table.PaginatedTable, {
    activePage: activePage,
    columns: columns,
    dataTestSubj: `table-${tableType}`,
    headerCount: totalCount,
    headerTitle: headerTitle,
    headerUnit: i18n.UNIT(totalCount),
    id: id,
    isInspect: isInspect,
    itemsPerRow: rowItems,
    limit: limit,
    loading: loading,
    loadPage: loadPage,
    onChange: onChange,
    pageOfItems: data,
    setQuerySkip: setQuerySkip,
    showMorePagesIndicator: showMorePagesIndicator,
    sorting: sorting,
    totalCount: limitedPaginationTotalCount,
    updateActivePage: updateActivePage,
    updateLimitPagination: updateLimitPagination
  });
};
const NetworkTopNFlowTable = exports.NetworkTopNFlowTable = /*#__PURE__*/_react.default.memo(NetworkTopNFlowTableComponent);
