// @kbn/cloud-security-posture-plugin/public/pages/vulnerabilities/latest_vulnerabilities_container.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LatestVulnerabilitiesContainer = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _use_latest_vulnerabilities_grouping = require("./hooks/use_latest_vulnerabilities_grouping");
var _latest_vulnerabilities_table = require("./latest_vulnerabilities_table");
var _latest_vulnerabilities_group_renderer = require("./latest_vulnerabilities_group_renderer");
var _findings_search_bar = require("../configurations/layout/findings_search_bar");
var _error_callout = require("../configurations/layout/error_callout");
var _empty_state = require("../../components/empty_state");
var _cloud_security_grouping = require("../../components/cloud_security_grouping");
var _constants = require("../../common/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LatestVulnerabilitiesContainer = () => {
  const SubGrouping = ({
    renderChildComponent,
    groupingLevel,
    parentGroupFilters,
    selectedGroup,
    groupSelectorComponent
  }) => {
    const {
      groupData,
      grouping,
      isFetching,
      activePageIndex,
      pageSize,
      onChangeGroupsItemsPerPage,
      onChangeGroupsPage,
      isGroupLoading,
      setActivePageIndex
    } = (0, _use_latest_vulnerabilities_grouping.useLatestVulnerabilitiesGrouping)({
      groupPanelRenderer: _latest_vulnerabilities_group_renderer.groupPanelRenderer,
      groupStatsRenderer: _latest_vulnerabilities_group_renderer.groupStatsRenderer,
      groupingLevel,
      selectedGroup,
      groupFilters: parentGroupFilters ? JSON.parse(parentGroupFilters) : []
    });

    /**
     * This is used to reset the active page index when the selected group changes
     * It is needed because the grouping number of pages can change according to the selected group
     */
    (0, _react.useEffect)(() => {
      setActivePageIndex(0);
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [selectedGroup]);
    return /*#__PURE__*/_react.default.createElement(_cloud_security_grouping.CloudSecurityGrouping, {
      data: groupData,
      grouping: grouping,
      renderChildComponent: renderChildComponent,
      onChangeGroupsItemsPerPage: onChangeGroupsItemsPerPage,
      onChangeGroupsPage: onChangeGroupsPage,
      activePageIndex: activePageIndex,
      isFetching: isFetching,
      pageSize: pageSize,
      selectedGroup: selectedGroup,
      isGroupLoading: isGroupLoading,
      groupingLevel: groupingLevel,
      groupSelectorComponent: groupSelectorComponent
    });
  };
  const renderChildComponent = ({
    level,
    currentSelectedGroup,
    selectedGroupOptions,
    parentGroupFilters,
    groupSelectorComponent
  }) => {
    let getChildComponent;
    if (currentSelectedGroup === 'none') {
      return /*#__PURE__*/_react.default.createElement(_latest_vulnerabilities_table.LatestVulnerabilitiesTable, {
        groupSelectorComponent: groupSelectorComponent,
        nonPersistedFilters: [...(parentGroupFilters ? JSON.parse(parentGroupFilters) : [])]
      });
    }
    if (level < selectedGroupOptions.length - 1 && !selectedGroupOptions.includes('none')) {
      getChildComponent = currentGroupFilters => {
        const nextGroupingLevel = level + 1;
        return renderChildComponent({
          level: nextGroupingLevel,
          currentSelectedGroup: selectedGroupOptions[nextGroupingLevel],
          selectedGroupOptions,
          parentGroupFilters: JSON.stringify([...currentGroupFilters, ...(parentGroupFilters ? JSON.parse(parentGroupFilters) : [])]),
          groupSelectorComponent
        });
      };
    } else {
      getChildComponent = currentGroupFilters => {
        return /*#__PURE__*/_react.default.createElement(_latest_vulnerabilities_table.LatestVulnerabilitiesTable, {
          nonPersistedFilters: [...currentGroupFilters, ...(parentGroupFilters ? JSON.parse(parentGroupFilters) : [])],
          height: _constants.DEFAULT_GROUPING_TABLE_HEIGHT
        });
      };
    }
    return /*#__PURE__*/_react.default.createElement(SubGrouping, {
      renderChildComponent: getChildComponent,
      selectedGroup: selectedGroupOptions[level],
      groupingLevel: level,
      parentGroupFilters: parentGroupFilters,
      groupSelectorComponent: groupSelectorComponent
    });
  };
  const {
    grouping,
    isFetching,
    urlQuery,
    setUrlQuery,
    onResetFilters,
    error,
    isEmptyResults
  } = (0, _use_latest_vulnerabilities_grouping.useLatestVulnerabilitiesGrouping)({
    groupPanelRenderer: _latest_vulnerabilities_group_renderer.groupPanelRenderer,
    groupStatsRenderer: _latest_vulnerabilities_group_renderer.groupStatsRenderer
  });
  if (error || isEmptyResults) {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_findings_search_bar.FindingsSearchBar, {
      query: urlQuery,
      setQuery: setUrlQuery,
      loading: isFetching
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "m"
    }), error && /*#__PURE__*/_react.default.createElement(_error_callout.ErrorCallout, {
      error: error
    }), isEmptyResults && /*#__PURE__*/_react.default.createElement(_empty_state.EmptyState, {
      onResetFilters: onResetFilters
    }));
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_findings_search_bar.FindingsSearchBar, {
    query: urlQuery,
    setQuery: setUrlQuery,
    loading: isFetching
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement("div", null, renderChildComponent({
    level: 0,
    currentSelectedGroup: grouping.selectedGroups[0],
    selectedGroupOptions: grouping.selectedGroups,
    groupSelectorComponent: grouping.groupSelector
  })));
};
exports.LatestVulnerabilitiesContainer = LatestVulnerabilitiesContainer;
