// @kbn/cloud-security-posture-plugin/public/pages/vulnerabilities/test_subjects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VULNERABILITIES_GROUPING_COUNTER = exports.TAB_ID_VULNERABILITY_FLYOUT = exports.SEVERITY_STATUS_VULNERABILITY_FLYOUT = exports.OVERVIEW_TAB_VULNERABILITY_FLYOUT = exports.LATEST_VULNERABILITIES_TABLE = exports.JSON_TAB_VULNERABILITY_FLYOUT = exports.FINDINGS_VULNERABILITY_FLYOUT_DESCRIPTION_LIST = exports.FINDINGS_VULNERABILITY_FLYOUT = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FINDINGS_VULNERABILITY_FLYOUT = exports.FINDINGS_VULNERABILITY_FLYOUT = 'vulnerability_findings_flyout';
const FINDINGS_VULNERABILITY_FLYOUT_DESCRIPTION_LIST = exports.FINDINGS_VULNERABILITY_FLYOUT_DESCRIPTION_LIST = 'vulnerability-flyout-description-list';
const JSON_TAB_VULNERABILITY_FLYOUT = exports.JSON_TAB_VULNERABILITY_FLYOUT = 'vulnerability_json_tab_flyout';
const OVERVIEW_TAB_VULNERABILITY_FLYOUT = exports.OVERVIEW_TAB_VULNERABILITY_FLYOUT = 'vulnerability_overview_tab_flyout';
const SEVERITY_STATUS_VULNERABILITY_FLYOUT = exports.SEVERITY_STATUS_VULNERABILITY_FLYOUT = 'vulnerability_severity_status_flyout';
const TAB_ID_VULNERABILITY_FLYOUT = tabId => `vulnerability-finding-flyout-tab-${tabId}`;
exports.TAB_ID_VULNERABILITY_FLYOUT = TAB_ID_VULNERABILITY_FLYOUT;
const LATEST_VULNERABILITIES_TABLE = exports.LATEST_VULNERABILITIES_TABLE = 'latest_vulnerabilities_table';
const VULNERABILITIES_GROUPING_COUNTER = exports.VULNERABILITIES_GROUPING_COUNTER = 'vulnerabilities_grouping_counter';
