// @kbn/spaces-plugin/public/management/components/unauthorized_prompt/unauthorized_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UnauthorizedPrompt = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UnauthorizedPrompt = () => /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  iconType: "spacesApp",
  iconColor: 'danger',
  title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.spaces.management.unauthorizedPrompt.permissionDeniedTitle",
    defaultMessage: "Permission denied"
  })),
  body: /*#__PURE__*/_react.default.createElement("p", {
    "data-test-subj": "permissionDeniedMessage"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.spaces.management.unauthorizedPrompt.permissionDeniedDescription",
    defaultMessage: "You don't have permission to manage spaces."
  }))
});
exports.UnauthorizedPrompt = UnauthorizedPrompt;
