// @kbn/canvas-plugin/public/components/arg_add/arg_add.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ArgAdd = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ArgAdd = ({
  onValueAdd = () => {},
  displayName,
  help
}) => {
  return /*#__PURE__*/_react.default.createElement("button", {
    className: "canvasArg__add",
    onClick: onValueAdd
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    compressed: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionListTitle, null, displayName), /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionListDescription, null, /*#__PURE__*/_react.default.createElement("small", null, help))));
};
exports.ArgAdd = ArgAdd;
ArgAdd.propTypes = {
  displayName: _propTypes.default.string.isRequired,
  help: _propTypes.default.string.isRequired,
  onValueAdd: _propTypes.default.func
};
