// @kbn/spaces-plugin/public/copy_saved_objects_to_space/components/copy_status_summary_indicator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CopyStatusSummaryIndicator = void 0;
require("./copy_status_summary_indicator.scss");
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _resolve_all_conflicts = require("./resolve_all_conflicts");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const renderIcon = props => {
  const {
    space,
    summarizedCopyResult,
    conflictResolutionInProgress,
    retries,
    onRetriesChange,
    onDestinationMapChange
  } = props;
  const getDataTestSubj = status => `cts-summary-indicator-${status}-${space.id}`;
  if (summarizedCopyResult.processing || conflictResolutionInProgress) {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
      "data-test-subj": getDataTestSubj('loading')
    });
  }
  const {
    successful,
    hasUnresolvableErrors,
    hasMissingReferences,
    hasConflicts
  } = summarizedCopyResult;
  if (successful) {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
      type: 'checkInCircleFilled',
      color: 'success',
      iconProps: {
        'data-test-subj': getDataTestSubj('success')
      },
      content: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.spaces.management.copyToSpace.copyStatusSummary.successMessage",
        defaultMessage: "Copy to the {space} space was successful.",
        values: {
          space: space.name
        }
      })
    });
  }
  if (hasUnresolvableErrors) {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
      type: 'alert',
      color: 'danger',
      iconProps: {
        'data-test-subj': getDataTestSubj('failed')
      },
      content: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.spaces.management.copyToSpace.copyStatusSummary.failedMessage",
        defaultMessage: "Copy to the {space} space failed. Expand this section for details.",
        values: {
          space: space.name
        }
      })
    });
  }
  const missingReferences = hasMissingReferences ? /*#__PURE__*/_react.default.createElement("span", {
    className: "spcCopyToSpace__missingReferencesIcon"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    type: 'link',
    color: 'warning',
    iconProps: {
      'data-test-subj': getDataTestSubj('missingReferences')
    },
    content: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.spaces.management.copyToSpace.copyStatusSummary.missingReferencesMessage",
      defaultMessage: "Missing references detected in the {space} space. Expand this section for details.",
      values: {
        space: space.name
      }
    })
  })) : null;
  if (hasConflicts) {
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_resolve_all_conflicts.ResolveAllConflicts, {
      summarizedCopyResult: summarizedCopyResult,
      retries: retries,
      onRetriesChange: onRetriesChange,
      onDestinationMapChange: onDestinationMapChange
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
      type: 'alert',
      color: 'warning',
      iconProps: {
        'data-test-subj': getDataTestSubj('conflicts')
      },
      content: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.spaces.management.copyToSpace.copyStatusSummary.conflictsMessage",
        defaultMessage: "Conflicts detected in the {space} space. Expand this section to resolve.",
        values: {
          space: space.name
        }
      })
    }), missingReferences);
  }
  return missingReferences;
};
const CopyStatusSummaryIndicator = props => {
  const {
    summarizedCopyResult
  } = props;
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, renderIcon(props), /*#__PURE__*/_react.default.createElement(_eui.EuiBadge, {
    color: "#DDD",
    className: "spcCopyToSpace__summaryCountBadge"
  }, summarizedCopyResult.objects.length));
};
exports.CopyStatusSummaryIndicator = CopyStatusSummaryIndicator;
