// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/pipeline_processors_editor_item_tooltip/processor_information.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProcessorInformation = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _shared = require("../shared");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ProcessorInformation = exports.ProcessorInformation = /*#__PURE__*/(0, _react.memo)(({
  processor
}) => {
  var _processorDescriptor$, _processor$options$de;
  const processorDescriptor = (0, _shared.getProcessorDescriptor)(processor.type);
  const label = (_processorDescriptor$ = processorDescriptor === null || processorDescriptor === void 0 ? void 0 : processorDescriptor.label) !== null && _processorDescriptor$ !== void 0 ? _processorDescriptor$ : processor.type;
  const description = (_processor$options$de = processor.options.description) !== null && _processor$options$de !== void 0 ? _processor$options$de : processorDescriptor === null || processorDescriptor === void 0 ? void 0 : processorDescriptor.getDefaultDescription(processor.options);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "center",
    alignItems: "center",
    responsive: false,
    gutterSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("strong", null, label))), description ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    color: "subdued",
    size: "s"
  }, description)) : undefined));
});
