// @kbn/ml-plugin/public/application/settings/filter_lists/components/filter_list_usage_popover/filter_list_usage_popover.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilterListUsagePopover = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * React popover listing the jobs or detectors using a particular filter list in a custom rule.
 */

class FilterListUsagePopover extends _react.Component {
  constructor(props) {
    super(props);
    (0, _defineProperty2.default)(this, "onButtonClick", () => {
      this.setState({
        isPopoverOpen: !this.state.isPopoverOpen
      });
    });
    (0, _defineProperty2.default)(this, "closePopover", () => {
      this.setState({
        isPopoverOpen: false
      });
    });
    this.state = {
      isPopoverOpen: false
    };
  }
  render() {
    const {
      entityType,
      entityValues
    } = this.props;
    const linkText = `${entityValues.length} ${entityType}${entityValues.length !== 1 ? 's' : ''}`;
    const listItems = entityValues.map(value => /*#__PURE__*/_react.default.createElement("li", {
      key: value
    }, value));
    const button = /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
      size: "s",
      color: "primary",
      onClick: this.onButtonClick
    }, linkText);
    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
      id: `${entityType}_filter_list_usage`,
      panelClassName: "ml-filter-list-usage-popover",
      ownFocus: true,
      button: button,
      isOpen: this.state.isPopoverOpen,
      closePopover: this.closePopover
    }, /*#__PURE__*/_react.default.createElement("ul", null, listItems)));
  }
}
exports.FilterListUsagePopover = FilterListUsagePopover;
FilterListUsagePopover.propTypes = {
  entityType: _propTypes.default.oneOf(['job', 'detector']),
  entityValues: _propTypes.default.array.isRequired
};
