// @kbn/slo-plugin/public/pages/slos/components/slos_overview/overview_item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverViewItem = OverViewItem;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _use_url_search_state = require("../../hooks/use_url_search_state");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function OverViewItem({
  title,
  description,
  titleColor,
  isLoading,
  query,
  tooltip,
  onClick
}) {
  const {
    onStateChange
  } = (0, _use_url_search_state.useUrlSearchState)();
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: tooltip
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiStat, {
    title: title,
    description: description,
    titleColor: titleColor,
    reverse: true,
    isLoading: isLoading,
    onClick: () => {
      if (onClick) {
        onClick();
        return;
      }
      onStateChange({
        kqlQuery: query
      });
    },
    css: {
      cursor: 'pointer'
    }
  })));
}
