// @kbn/apm-plugin/public/components/shared/kuery_bar/typeahead/click_outside.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class ClickOutside extends _react.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "setNodeRef", node => {
      this.nodeRef = node;
    });
    (0, _defineProperty2.default)(this, "onClick", event => {
      if (this.nodeRef && !this.nodeRef.contains(event.target)) {
        this.props.onClickOutside();
      }
    });
  }
  componentDidMount() {
    document.addEventListener('mousedown', this.onClick);
  }
  componentWillUnmount() {
    document.removeEventListener('mousedown', this.onClick);
  }
  render() {
    // eslint-disable-next-line no-unused-vars
    const {
      onClickOutside,
      ...restProps
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
      ref: this.setNodeRef
    }, restProps), this.props.children);
  }
}
exports.default = ClickOutside;
ClickOutside.propTypes = {
  onClickOutside: _propTypes.default.func.isRequired
};
module.exports = exports.default;
