// @kbn/index-management/src/home_sections.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Section = exports.IndexDetailsSection = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let Section = exports.Section = /*#__PURE__*/function (Section) {
  Section["Indices"] = "indices";
  Section["DataStreams"] = "data_streams";
  Section["IndexTemplates"] = "templates";
  Section["ComponentTemplates"] = "component_templates";
  Section["EnrichPolicies"] = "enrich_policies";
  return Section;
}({});
let IndexDetailsSection = exports.IndexDetailsSection = /*#__PURE__*/function (IndexDetailsSection) {
  IndexDetailsSection["Overview"] = "overview";
  IndexDetailsSection["Mappings"] = "mappings";
  IndexDetailsSection["Settings"] = "settings";
  IndexDetailsSection["Stats"] = "stats";
  return IndexDetailsSection;
}({});
