// @kbn/ml-data-view-utils/components/create_data_view_switch.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateDataViewSwitch = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CreateDataViewSwitch = ({
  canCreateDataView,
  createDataView,
  setCreateDataView
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
  name: "mlCreateDataView",
  disabled: canCreateDataView === false,
  label: _i18n.i18n.translate('xpack.ml.dataViewUtils.createDataViewSwitchLabel', {
    defaultMessage: 'Create data view'
  }),
  checked: createDataView === true,
  onChange: () => setCreateDataView(!createDataView),
  "data-test-subj": "mlCreateDataViewSwitch"
});
exports.CreateDataViewSwitch = CreateDataViewSwitch;
