// @kbn/maps-plugin/public/connected_components/panel_strings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.panelStrings = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const panelStrings = exports.panelStrings = {
  apply: _i18n.i18n.translate('xpack.maps.panel.applyLabel', {
    defaultMessage: 'Apply'
  }),
  clear: _i18n.i18n.translate('xpack.maps.panel.clearLabel', {
    defaultMessage: 'Clear'
  }),
  close: _i18n.i18n.translate('xpack.maps.panel.closeLabel', {
    defaultMessage: 'Close'
  }),
  discardChanges: _i18n.i18n.translate('xpack.maps.panel.discardChangesLabel', {
    defaultMessage: 'Discard changes'
  }),
  edit: _i18n.i18n.translate('xpack.maps.panel.editLabel', {
    defaultMessage: 'Edit'
  }),
  keepChanges: _i18n.i18n.translate('xpack.maps.panel.keepChangesLabel', {
    defaultMessage: 'Keep changes'
  })
};
