// @kbn/console-plugin/public/services/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AutocompleteInfo", {
  enumerable: true,
  get: function () {
    return _autocomplete.AutocompleteInfo;
  }
});
Object.defineProperty(exports, "DEFAULT_SETTINGS", {
  enumerable: true,
  get: function () {
    return _settings.DEFAULT_SETTINGS;
  }
});
Object.defineProperty(exports, "ENTITIES", {
  enumerable: true,
  get: function () {
    return _autocomplete.ENTITIES;
  }
});
Object.defineProperty(exports, "EmbeddableConsoleInfo", {
  enumerable: true,
  get: function () {
    return _embeddable_console.EmbeddableConsoleInfo;
  }
});
Object.defineProperty(exports, "History", {
  enumerable: true,
  get: function () {
    return _history.History;
  }
});
Object.defineProperty(exports, "Settings", {
  enumerable: true,
  get: function () {
    return _settings.Settings;
  }
});
Object.defineProperty(exports, "Storage", {
  enumerable: true,
  get: function () {
    return _storage.Storage;
  }
});
Object.defineProperty(exports, "StorageKeys", {
  enumerable: true,
  get: function () {
    return _storage.StorageKeys;
  }
});
Object.defineProperty(exports, "createHistory", {
  enumerable: true,
  get: function () {
    return _history.createHistory;
  }
});
Object.defineProperty(exports, "createSettings", {
  enumerable: true,
  get: function () {
    return _settings.createSettings;
  }
});
Object.defineProperty(exports, "createStorage", {
  enumerable: true,
  get: function () {
    return _storage.createStorage;
  }
});
Object.defineProperty(exports, "getAutocompleteInfo", {
  enumerable: true,
  get: function () {
    return _autocomplete.getAutocompleteInfo;
  }
});
Object.defineProperty(exports, "getStorage", {
  enumerable: true,
  get: function () {
    return _storage.getStorage;
  }
});
Object.defineProperty(exports, "setAutocompleteInfo", {
  enumerable: true,
  get: function () {
    return _autocomplete.setAutocompleteInfo;
  }
});
Object.defineProperty(exports, "setStorage", {
  enumerable: true,
  get: function () {
    return _storage.setStorage;
  }
});
var _history = require("./history");
var _storage = require("./storage");
var _settings = require("./settings");
var _autocomplete = require("./autocomplete");
var _embeddable_console = require("./embeddable_console");
