// @kbn/security-solution-plugin/public/management/pages/policy/view/artifacts/delete_modal/policy_artifacts_delete_modal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PolicyArtifactsDeleteModal = void 0;
var _eui = require("@elastic/eui");
var _reactQuery = require("@tanstack/react-query");
var _react = _interopRequireWildcard(require("react"));
var _artifacts = require("../../../../../hooks/artifacts");
var _kibana = require("../../../../../../common/lib/kibana");
var _artifacts2 = require("../../../../../../../common/endpoint/service/artifacts");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// eslint-disable-next-line react/display-name
const PolicyArtifactsDeleteModal = exports.PolicyArtifactsDeleteModal = /*#__PURE__*/_react.default.memo(({
  policyId,
  policyName,
  apiClient,
  exception,
  onClose,
  labels
}) => {
  const toasts = (0, _kibana.useToasts)();
  const queryClient = (0, _reactQuery.useQueryClient)();
  const {
    mutate: updateArtifact,
    isLoading: isUpdateArtifactLoading
  } = (0, _artifacts.useBulkUpdateArtifact)(apiClient, {
    onSuccess: () => {
      toasts.addSuccess({
        title: labels.deleteModalSuccessMessageTitle,
        text: labels.deleteModalSuccessMessageText(exception, policyName)
      });
      queryClient.invalidateQueries(['list', apiClient]);
      onClose();
    },
    onError: error => {
      toasts.addError(error, {
        title: labels.deleteModalErrorMessage
      });
    }
  });
  const handleModalConfirm = (0, _react.useCallback)(() => {
    const modifiedException = {
      ...exception,
      tags: exception.tags.filter(tag => tag !== `${_artifacts2.BY_POLICY_ARTIFACT_TAG_PREFIX}${policyId}`)
    };
    updateArtifact([modifiedException]);
  }, [exception, policyId, updateArtifact]);
  const handleOnClose = (0, _react.useCallback)(() => {
    if (!isUpdateArtifactLoading) {
      onClose();
    }
  }, [isUpdateArtifactLoading, onClose]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiConfirmModal, {
    onCancel: handleOnClose,
    onConfirm: handleModalConfirm,
    title: labels.deleteModalTitle,
    cancelButtonText: labels.deleteModalCancelButtonTitle,
    confirmButtonText: labels.deleteModalSubmitButtonTitle,
    isLoading: isUpdateArtifactLoading,
    "data-test-subj": 'remove-from-policy-dialog'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    color: "warning",
    iconType: "help"
  }, /*#__PURE__*/_react.default.createElement("p", null, labels.deleteModalImpactInfo)), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("p", null, labels.deleteModalConfirmInfo)));
});
