// @kbn/security-solution-plugin/public/entity_analytics/components/risk_score_onboarding/risk_score_restart_button.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RiskScoreRestartButton = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _use_space_id = require("../../../common/hooks/use_space_id");
var _kibana = require("../../../common/lib/kibana");
var _use_fetch = require("../../../common/hooks/use_fetch");
var _use_risk_score_toast_content = require("./use_risk_score_toast_content");
var _utils = require("./utils");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RiskScoreRestartButtonComponent = ({
  refetch,
  riskScoreEntity
}) => {
  const {
    fetch,
    isLoading
  } = (0, _use_fetch.useFetch)(_use_fetch.REQUEST_NAMES.REFRESH_RISK_SCORE, _utils.restartRiskScoreTransforms);
  const spaceId = (0, _use_space_id.useSpaceId)();
  const {
    renderDocLink
  } = (0, _use_risk_score_toast_content.useRiskScoreToastContent)();
  const {
    http,
    ...startServices
  } = (0, _kibana.useKibana)().services;
  const onClick = (0, _react.useCallback)(async () => {
    fetch({
      http,
      refetch,
      renderDocLink,
      riskScoreEntity,
      spaceId,
      startServices
    });
  }, [fetch, http, refetch, renderDocLink, riskScoreEntity, spaceId, startServices]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    color: "primary",
    fill: true,
    onClick: onClick,
    isLoading: isLoading,
    "data-test-subj": `restart_${riskScoreEntity}_risk_score`
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.securitySolution.riskScore.restartButtonTitle",
    defaultMessage: "Restart"
  }));
};
const RiskScoreRestartButton = exports.RiskScoreRestartButton = /*#__PURE__*/_react.default.memo(RiskScoreRestartButtonComponent);
RiskScoreRestartButton.displayName = 'RiskScoreRestartButton';
