// @kbn/slo-schema/src/rest_specs/routes/fetch_historical_summary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.historicalSummarySchema = exports.fetchHistoricalSummaryResponseSchema = exports.fetchHistoricalSummaryParamsSchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _schema = require("../../schema");
var _common = require("../../schema/common");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const fetchHistoricalSummaryParamsSchema = exports.fetchHistoricalSummaryParamsSchema = t.type({
  body: t.type({
    list: t.array(t.intersection([t.type({
      sloId: _schema.sloIdSchema,
      instanceId: t.string,
      timeWindow: _schema.timeWindowSchema,
      budgetingMethod: _schema.budgetingMethodSchema,
      objective: _schema.objectiveSchema,
      groupBy: _common.allOrAnyStringOrArray,
      revision: t.number
    }), t.partial({
      remoteName: t.string,
      range: _common.dateRangeSchema
    })]))
  })
});
const historicalSummarySchema = exports.historicalSummarySchema = t.type({
  date: _common.dateType,
  status: _common.statusSchema,
  sliValue: t.number,
  errorBudget: _common.errorBudgetSchema
});
const fetchHistoricalSummaryResponseSchema = exports.fetchHistoricalSummaryResponseSchema = t.array(t.type({
  sloId: _schema.sloIdSchema,
  instanceId: _common.allOrAnyString,
  data: t.array(historicalSummarySchema)
}));
