// @kbn/slo-schema/src/rest_specs/routes/find_group.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sloGroupWithSummaryResponseSchema = exports.findSLOGroupsResponseSchema = exports.findSLOGroupsParamsSchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _common = require("../../schema/common");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const groupBySchema = t.union([t.literal('ungrouped'), t.literal('slo.tags'), t.literal('status'), t.literal('slo.indicator.type'), t.literal('slo.instanceId'), t.literal('_index')]);
const findSLOGroupsParamsSchema = exports.findSLOGroupsParamsSchema = t.partial({
  query: t.partial({
    page: t.string,
    perPage: t.string,
    groupBy: groupBySchema,
    groupsFilter: t.union([t.array(t.string), t.string]),
    kqlQuery: t.string,
    filters: t.string
  })
});
const sloGroupWithSummaryResponseSchema = exports.sloGroupWithSummaryResponseSchema = t.type({
  group: t.string,
  groupBy: groupBySchema,
  summary: _common.groupSummarySchema
});
const findSLOGroupsResponseSchema = exports.findSLOGroupsResponseSchema = t.type({
  page: t.number,
  perPage: t.number,
  total: t.number,
  results: t.array(sloGroupWithSummaryResponseSchema)
});
