// @kbn/data-view-field-editor-plugin/public/components/field_format_editor/editors/truncate/truncate.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TruncateFormatEditor = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _default = require("../default/default");
var _samples = require("../../samples");
var _sample = require("./sample");
var _constants = require("./constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

class TruncateFormatEditor extends _default.DefaultFormatEditor {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {
      ..._default.defaultState,
      sampleInputs: [_sample.sample]
    });
  }
  render() {
    var _formatParams$fieldLe;
    const {
      formatParams,
      onError
    } = this.props;
    const {
      error,
      samples
    } = this.state;
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
      label: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "indexPatternFieldEditor.truncate.lengthLabel",
        defaultMessage: "Field length"
      }),
      isInvalid: !!error,
      error: error
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFieldNumber, {
      defaultValue: (_formatParams$fieldLe = formatParams.fieldLength) !== null && _formatParams$fieldLe !== void 0 ? _formatParams$fieldLe : undefined,
      min: 1,
      "data-test-subj": 'truncateEditorLength',
      onChange: e => {
        if (e.target.checkValidity()) {
          this.onChange({
            fieldLength: e.target.value ? Number(e.target.value) : null
          });
        } else {
          onError(e.target.validationMessage);
        }
      },
      isInvalid: !!error
    })), /*#__PURE__*/_react.default.createElement(_samples.FormatEditorSamples, {
      samples: samples
    }));
  }
}
exports.TruncateFormatEditor = TruncateFormatEditor;
(0, _defineProperty2.default)(TruncateFormatEditor, "formatId", _constants.formatId);
