// @kbn/apm-plugin/public/components/app/mobile/transaction_overview/transaction_overview_tabs/transactions_tab.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.transactionsTab = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _transactions_table = require("../../../../shared/transactions_table");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function TransactionsTab({
  environment,
  kuery,
  start,
  end
}) {
  return /*#__PURE__*/_react.default.createElement(_transactions_table.TransactionsTable, {
    hideTitle: true,
    hideViewTransactionsLink: true,
    numberOfTransactionsPerPage: 25,
    showMaxTransactionGroupsExceededWarning: true,
    environment: environment,
    kuery: kuery,
    start: start,
    end: end,
    saveTableOptionsToUrl: true
  });
}
const transactionsTab = exports.transactionsTab = {
  dataTestSubj: 'apmTransactionsTab',
  key: 'transactions',
  label: _i18n.i18n.translate('xpack.apm.mobile.transactions.overview.tabs.transactions', {
    defaultMessage: 'Transactions'
  }),
  component: TransactionsTab
};
