// @kbn/cases-plugin/public/components/actions/severity/use_severity_action.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useSeverityAction = void 0;
var _react = require("react");
var _domain = require("../../../../common/types/domain");
var _use_bulk_update_case = require("../../../containers/use_bulk_update_case");
var i18n = _interopRequireWildcard(require("./translations"));
var _use_cases_context = require("../../cases_context/use_cases_context");
var _config = require("../../severity/config");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getSeverityToasterMessage = (severity, cases) => {
  const totalCases = cases.length;
  const caseTitle = totalCases === 1 ? cases[0].title : '';
  switch (severity) {
    case _domain.CaseSeverity.LOW:
      return i18n.SET_SEVERITY_LOW({
        totalCases,
        caseTitle
      });
    case _domain.CaseSeverity.MEDIUM:
      return i18n.SET_SEVERITY_MEDIUM({
        totalCases,
        caseTitle
      });
    case _domain.CaseSeverity.HIGH:
      return i18n.SET_SEVERITY_HIGH({
        totalCases,
        caseTitle
      });
    case _domain.CaseSeverity.CRITICAL:
      return i18n.SET_SEVERITY_CRITICAL({
        totalCases,
        caseTitle
      });
    default:
      return '';
  }
};
const shouldDisableSeverity = (cases, severity) => cases.every(theCase => theCase.severity === severity);
const useSeverityAction = ({
  onAction,
  onActionSuccess,
  isDisabled,
  selectedSeverity
}) => {
  const {
    mutate: updateCases
  } = (0, _use_bulk_update_case.useUpdateCases)();
  const {
    permissions
  } = (0, _use_cases_context.useCasesContext)();
  const canUpdateSeverity = permissions.update;
  const isActionDisabled = isDisabled || !canUpdateSeverity;
  const handleUpdateCaseSeverity = (0, _react.useCallback)((selectedCases, severity) => {
    onAction();
    const casesToUpdate = selectedCases.map(theCase => ({
      severity,
      id: theCase.id,
      version: theCase.version
    }));
    updateCases({
      cases: casesToUpdate,
      successToasterTitle: getSeverityToasterMessage(severity, selectedCases)
    }, {
      onSuccess: onActionSuccess
    });
  }, [onAction, updateCases, onActionSuccess]);
  const getSeverityIcon = severity => selectedSeverity && selectedSeverity === severity ? 'check' : 'empty';
  const getActions = selectedCases => {
    return [{
      name: _config.severities[_domain.CaseSeverity.LOW].label,
      icon: getSeverityIcon(_domain.CaseSeverity.LOW),
      onClick: () => handleUpdateCaseSeverity(selectedCases, _domain.CaseSeverity.LOW),
      disabled: isActionDisabled || shouldDisableSeverity(selectedCases, _domain.CaseSeverity.LOW),
      'data-test-subj': 'cases-bulk-action-severity-low',
      key: 'cases-bulk-action-severity-low'
    }, {
      name: _config.severities[_domain.CaseSeverity.MEDIUM].label,
      icon: getSeverityIcon(_domain.CaseSeverity.MEDIUM),
      onClick: () => handleUpdateCaseSeverity(selectedCases, _domain.CaseSeverity.MEDIUM),
      disabled: isActionDisabled || shouldDisableSeverity(selectedCases, _domain.CaseSeverity.MEDIUM),
      'data-test-subj': 'cases-bulk-action-severity-medium',
      key: 'cases-bulk-action-severity-medium'
    }, {
      name: _config.severities[_domain.CaseSeverity.HIGH].label,
      icon: getSeverityIcon(_domain.CaseSeverity.HIGH),
      onClick: () => handleUpdateCaseSeverity(selectedCases, _domain.CaseSeverity.HIGH),
      disabled: isActionDisabled || shouldDisableSeverity(selectedCases, _domain.CaseSeverity.HIGH),
      'data-test-subj': 'cases-bulk-action-severity-high',
      key: 'cases-bulk-action-severity-high'
    }, {
      name: _config.severities[_domain.CaseSeverity.CRITICAL].label,
      icon: getSeverityIcon(_domain.CaseSeverity.CRITICAL),
      onClick: () => handleUpdateCaseSeverity(selectedCases, _domain.CaseSeverity.CRITICAL),
      disabled: isActionDisabled || shouldDisableSeverity(selectedCases, _domain.CaseSeverity.CRITICAL),
      'data-test-subj': 'cases-bulk-action-severity-critical',
      key: 'cases-bulk-action-severity-critical'
    }];
  };
  return {
    getActions,
    canUpdateSeverity
  };
};
exports.useSeverityAction = useSeverityAction;
