// @kbn/cases-plugin/public/components/templates/template_fields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TemplateFields = void 0;
var _react = _interopRequireWildcard(require("react"));
var _hook_form_lib = require("@kbn/es-ui-shared-plugin/static/forms/hook_form_lib");
var _components = require("@kbn/es-ui-shared-plugin/static/forms/components");
var _eui = require("@elastic/eui");
var _optional_field_label = require("../optional_field_label");
var _template_tags = require("./template_tags");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TemplateFieldsComponent = ({
  isLoading = false,
  configurationTemplateTags
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  "data-test-subj": "template-fields",
  direction: "column",
  gutterSize: "none"
}, /*#__PURE__*/_react.default.createElement(_hook_form_lib.UseField, {
  path: "name",
  component: _components.TextField,
  componentProps: {
    euiFieldProps: {
      'data-test-subj': 'template-name-input',
      fullWidth: true,
      autoFocus: true,
      isLoading
    }
  }
}), /*#__PURE__*/_react.default.createElement(_template_tags.TemplateTags, {
  isLoading: isLoading,
  tagOptions: configurationTemplateTags
}), /*#__PURE__*/_react.default.createElement(_hook_form_lib.UseField, {
  path: "templateDescription",
  component: _components.TextAreaField,
  componentProps: {
    labelAppend: _optional_field_label.OptionalFieldLabel,
    euiFieldProps: {
      'data-test-subj': 'template-description-input',
      fullWidth: true,
      isLoading
    }
  }
}));
TemplateFieldsComponent.displayName = 'TemplateFields';
const TemplateFields = exports.TemplateFields = /*#__PURE__*/(0, _react.memo)(TemplateFieldsComponent);
