// @kbn/securitysolution-lists-common/api/model/list_common.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ListTypeEnum = exports.ListType = exports.ListName = exports.ListMetadata = exports.ListItemValue = exports.ListItemMetadata = exports.ListItemId = exports.ListItemDescription = exports.ListId = exports.ListDescription = void 0;
var _zod = require("zod");
var _primitives = require("@kbn/openapi-common/schemas/primitives.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Common List Attributes
 *   version: not applicable
 */

const ListId = exports.ListId = _primitives.NonEmptyString;
const ListType = exports.ListType = _zod.z.enum(['binary', 'boolean', 'byte', 'date', 'date_nanos', 'date_range', 'double', 'double_range', 'float', 'float_range', 'geo_point', 'geo_shape', 'half_float', 'integer', 'integer_range', 'ip', 'ip_range', 'keyword', 'long', 'long_range', 'shape', 'short', 'text']);
const ListTypeEnum = exports.ListTypeEnum = ListType.enum;
const ListName = exports.ListName = _primitives.NonEmptyString;
const ListDescription = exports.ListDescription = _primitives.NonEmptyString;
const ListMetadata = exports.ListMetadata = _zod.z.object({}).catchall(_zod.z.unknown());
const ListItemId = exports.ListItemId = _primitives.NonEmptyString;
const ListItemValue = exports.ListItemValue = _primitives.NonEmptyString;
const ListItemDescription = exports.ListItemDescription = _primitives.NonEmptyString;
const ListItemMetadata = exports.ListItemMetadata = _zod.z.object({}).catchall(_zod.z.unknown());
