// @kbn/canvas-plugin/public/components/workpad_page/workpad_static_page/static_workpad_page.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StaticWorkpadPage = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _element_wrapper = require("../../element_wrapper");
var _prop_types = require("../prop_types");
var _workpad = require("../../../lib/workpad");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class StaticWorkpadPage extends _react.PureComponent {
  render() {
    const {
      pageId,
      pageStyle,
      className,
      animationStyle,
      elements,
      height,
      width
    } = this.props;
    return /*#__PURE__*/_react.default.createElement("div", {
      key: pageId,
      id: pageId,
      "data-test-subj": "canvasWorkpadPage",
      className: `canvasPage kbn-resetFocusState canvasStaticPage ${className}`,
      "data-shared-items-container": true,
      style: {
        ...pageStyle,
        ...animationStyle,
        height,
        width
      }
    }, elements.filter(node => !(0, _workpad.isGroupId)(node.id)).map(element => /*#__PURE__*/_react.default.createElement(_element_wrapper.ElementWrapper, {
      key: element.id,
      element: element
    })));
  }
}
exports.StaticWorkpadPage = StaticWorkpadPage;
(0, _defineProperty2.default)(StaticWorkpadPage, "propTypes", _prop_types.staticWorkpadPagePropTypes);
