// @kbn/security-plugin/public/management/api_keys/api_keys_grid/not_enabled/not_enabled.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotEnabled = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _public = require("@kbn/kibana-react-plugin/public");
var _sharedUxPageKibanaTemplate = require("@kbn/shared-ux-page-kibana-template");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const NotEnabled = () => {
  const docLinks = (0, _public.useKibana)().services.docLinks;
  return /*#__PURE__*/_react.default.createElement(_sharedUxPageKibanaTemplate.KibanaPageTemplate.EmptyPrompt, {
    title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.security.management.apiKeys.table.apiKeysDisabledErrorTitle",
      defaultMessage: "API keys not enabled in Elasticsearch"
    })),
    color: "danger",
    iconType: "warning",
    body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.security.management.apiKeys.table.apiKeysDisabledErrorDescription",
      defaultMessage: "Contact your system administrator and refer to the {link} to enable API keys.",
      values: {
        link: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
          href: `${docLinks.links.security.apiKeyServiceSettings}`,
          target: "_blank"
        }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
          id: "xpack.security.management.apiKeys.table.apiKeysDisabledErrorLinkText",
          defaultMessage: "docs"
        }))
      }
    }))
  });
};
exports.NotEnabled = NotEnabled;
