// @kbn/ml-plugin/server/shared_services/providers/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getAnomalyDetectorsProvider", {
  enumerable: true,
  get: function () {
    return _anomaly_detectors.getAnomalyDetectorsProvider;
  }
});
Object.defineProperty(exports, "getJobServiceProvider", {
  enumerable: true,
  get: function () {
    return _job_service.getJobServiceProvider;
  }
});
Object.defineProperty(exports, "getMlSystemProvider", {
  enumerable: true,
  get: function () {
    return _system.getMlSystemProvider;
  }
});
Object.defineProperty(exports, "getModulesProvider", {
  enumerable: true,
  get: function () {
    return _modules.getModulesProvider;
  }
});
Object.defineProperty(exports, "getResultsServiceProvider", {
  enumerable: true,
  get: function () {
    return _results_service.getResultsServiceProvider;
  }
});
Object.defineProperty(exports, "getTrainedModelsProvider", {
  enumerable: true,
  get: function () {
    return _trained_models.getTrainedModelsProvider;
  }
});
var _system = require("./system");
var _job_service = require("./job_service");
var _modules = require("./modules");
var _results_service = require("./results_service");
var _trained_models = require("./trained_models");
var _anomaly_detectors = require("./anomaly_detectors");
