// @kbn/threat-intelligence-plugin/public/modules/indicators/components/table/more_actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MoreActions = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _add_to_block_list = require("../../../block_list/components/add_to_block_list");
var _add_to_new_case = require("../../../cases/components/add_to_new_case");
var _add_to_existing_case = require("../../../cases/components/add_to_existing_case");
var _can_add_to_block_list = require("../../../block_list/utils/can_add_to_block_list");
var _test_ids = require("./test_ids");
var _translations = require("./translations");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Component rendered in the action column.
 * Renders a ... icon button, with a dropdown.
 */
const MoreActions = ({
  indicator
}) => {
  const [isPopoverOpen, setPopover] = (0, _react.useState)(false);
  const smallContextMenuPopoverId = (0, _eui.useGeneratedHtmlId)({
    prefix: 'smallContextMenuPopover'
  });
  const closePopover = () => {
    setPopover(false);
  };
  const items = [/*#__PURE__*/_react.default.createElement(_add_to_existing_case.AddToExistingCase, {
    indicator: indicator,
    onClick: closePopover,
    "data-test-subj": _test_ids.ADD_TO_EXISTING_TEST_ID
  }), /*#__PURE__*/_react.default.createElement(_add_to_new_case.AddToNewCase, {
    indicator: indicator,
    onClick: closePopover,
    "data-test-subj": _test_ids.ADD_TO_NEW_CASE_TEST_ID
  }), /*#__PURE__*/_react.default.createElement(_add_to_block_list.AddToBlockListContextMenu, {
    data: (0, _can_add_to_block_list.canAddToBlockList)(indicator),
    onClick: closePopover,
    "data-test-subj": _test_ids.ADD_TO_BLOCK_LIST_TEST_ID
  })];
  const button = /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: _translations.MORE_ACTIONS_BUTTON_LABEL
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    "aria-label": _translations.MORE_ACTIONS_BUTTON_LABEL,
    iconType: "boxesHorizontal",
    iconSize: "s",
    size: "xs",
    onClick: () => setPopover(prevIsPopoverOpen => !prevIsPopoverOpen),
    style: {
      height: '100%'
    },
    "data-test-subj": _test_ids.MORE_ACTIONS_TEST_ID
  }));
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
    id: smallContextMenuPopoverId,
    button: button,
    isOpen: isPopoverOpen,
    closePopover: closePopover,
    panelPaddingSize: "none",
    anchorPosition: "downLeft"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiContextMenuPanel, {
    size: "s",
    items: items
  }));
};
exports.MoreActions = MoreActions;
