// @kbn/aiops-log-rate-analysis/state/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "LogRateAnalysisReduxProvider", {
  enumerable: true,
  get: function () {
    return _store.LogRateAnalysisReduxProvider;
  }
});
Object.defineProperty(exports, "brushSelectionUpdate", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.brushSelectionUpdate;
  }
});
Object.defineProperty(exports, "clearAllRowState", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_table_row_slice.clearAllRowState;
  }
});
Object.defineProperty(exports, "clearSelection", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.clearSelection;
  }
});
Object.defineProperty(exports, "setAnalysisType", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setAnalysisType;
  }
});
Object.defineProperty(exports, "setAutoRunAnalysis", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setAutoRunAnalysis;
  }
});
Object.defineProperty(exports, "setDocumentCountChartData", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setDocumentCountChartData;
  }
});
Object.defineProperty(exports, "setInitialAnalysisStart", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setInitialAnalysisStart;
  }
});
Object.defineProperty(exports, "setIsBrushCleared", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setIsBrushCleared;
  }
});
Object.defineProperty(exports, "setPinnedGroup", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_table_row_slice.setPinnedGroup;
  }
});
Object.defineProperty(exports, "setPinnedSignificantItem", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_table_row_slice.setPinnedSignificantItem;
  }
});
Object.defineProperty(exports, "setSelectedGroup", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_table_row_slice.setSelectedGroup;
  }
});
Object.defineProperty(exports, "setSelectedSignificantItem", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_table_row_slice.setSelectedSignificantItem;
  }
});
Object.defineProperty(exports, "setStickyHistogram", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setStickyHistogram;
  }
});
Object.defineProperty(exports, "setWindowParameters", {
  enumerable: true,
  get: function () {
    return _log_rate_analysis_slice.setWindowParameters;
  }
});
Object.defineProperty(exports, "useAppDispatch", {
  enumerable: true,
  get: function () {
    return _hooks.useAppDispatch;
  }
});
Object.defineProperty(exports, "useAppSelector", {
  enumerable: true,
  get: function () {
    return _hooks.useAppSelector;
  }
});
Object.defineProperty(exports, "useAppStore", {
  enumerable: true,
  get: function () {
    return _hooks.useAppStore;
  }
});
Object.defineProperty(exports, "useCurrentSelectedGroup", {
  enumerable: true,
  get: function () {
    return _use_current_selected_group.useCurrentSelectedGroup;
  }
});
Object.defineProperty(exports, "useCurrentSelectedSignificantItem", {
  enumerable: true,
  get: function () {
    return _use_current_selected_significant_item.useCurrentSelectedSignificantItem;
  }
});
var _log_rate_analysis_slice = require("./log_rate_analysis_slice");
var _log_rate_analysis_table_row_slice = require("./log_rate_analysis_table_row_slice");
var _store = require("./store");
var _hooks = require("./hooks");
var _use_current_selected_group = require("./use_current_selected_group");
var _use_current_selected_significant_item = require("./use_current_selected_significant_item");
