// @kbn/infra-plugin/common/http_api/log_analysis/results/v1/log_entry_anomalies_datasets.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getLogEntryAnomaliesDatasetsSuccessReponsePayloadRT = exports.getLogEntryAnomaliesDatasetsResponsePayloadRT = exports.getLogEntryAnomaliesDatasetsRequestPayloadRT = exports.LOG_ANALYSIS_GET_LOG_ENTRY_ANOMALIES_DATASETS_PATH = void 0;
var rt = _interopRequireWildcard(require("io-ts"));
var _common = require("@kbn/logs-shared-plugin/common");
var _shared = require("../../../shared");
var _id_formats = require("../../id_formats/v1/id_formats");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LOG_ANALYSIS_GET_LOG_ENTRY_ANOMALIES_DATASETS_PATH = exports.LOG_ANALYSIS_GET_LOG_ENTRY_ANOMALIES_DATASETS_PATH = '/api/infra/log_analysis/results/log_entry_anomalies_datasets';

/**
 * request
 */

const getLogEntryAnomaliesDatasetsRequestPayloadRT = exports.getLogEntryAnomaliesDatasetsRequestPayloadRT = rt.type({
  data: rt.type({
    // log view
    logView: _common.persistedLogViewReferenceRT,
    idFormats: _id_formats.idFormatByJobTypeRT,
    // the time range to fetch the anomalies datasets from
    timeRange: _shared.timeRangeRT
  })
});
/**
 * response
 */

const getLogEntryAnomaliesDatasetsSuccessReponsePayloadRT = exports.getLogEntryAnomaliesDatasetsSuccessReponsePayloadRT = rt.intersection([rt.type({
  data: rt.type({
    datasets: rt.array(rt.string)
  })
}), rt.partial({
  timing: _shared.routeTimingMetadataRT
})]);
const getLogEntryAnomaliesDatasetsResponsePayloadRT = exports.getLogEntryAnomaliesDatasetsResponsePayloadRT = rt.union([getLogEntryAnomaliesDatasetsSuccessReponsePayloadRT, _shared.badRequestErrorRT, _shared.forbiddenErrorRT]);
