// @kbn/core-usage-data-base-server-internal/src/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.REPOSITORY_RESOLVE_OUTCOME_STATS = exports.CORE_USAGE_STATS_TYPE = exports.CORE_USAGE_STATS_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/** @internal */
const CORE_USAGE_STATS_TYPE = exports.CORE_USAGE_STATS_TYPE = 'core-usage-stats';

/** @internal */
const CORE_USAGE_STATS_ID = exports.CORE_USAGE_STATS_ID = 'core-usage-stats';

/** @internal */
const REPOSITORY_RESOLVE_OUTCOME_STATS = exports.REPOSITORY_RESOLVE_OUTCOME_STATS = {
  EXACT_MATCH: 'savedObjectsRepository.resolvedOutcome.exactMatch',
  ALIAS_MATCH: 'savedObjectsRepository.resolvedOutcome.aliasMatch',
  CONFLICT: 'savedObjectsRepository.resolvedOutcome.conflict',
  NOT_FOUND: 'savedObjectsRepository.resolvedOutcome.notFound',
  TOTAL: 'savedObjectsRepository.resolvedOutcome.total'
};
