// @kbn/cloud-defend-plugin/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.STATUS_ROUTE_PATH = exports.PROCESS_INDEX_PATTERN_DEFAULT_NS = exports.PROCESS_INDEX_PATTERN = exports.PROCESS_DATASET = exports.POLICIES_ROUTE_PATH = exports.POLICIES_PACKAGE_POLICY_PREFIX = exports.PLUGIN_NAME = exports.PLUGIN_ID = exports.LOGS_CLOUD_DEFEND_PATTERN = exports.INTEGRATION_PACKAGE_NAME = exports.INPUT_CONTROL = exports.FILE_INDEX_PATTERN_DEFAULT_NS = exports.FILE_INDEX_PATTERN = exports.FILE_DATASET = exports.DEFAULT_POLICIES_PER_PAGE = exports.CURRENT_API_VERSION = exports.CLOUD_DEFEND_FLEET_PACKAGE_KUERY = exports.ALERTS_INDEX_PATTERN_DEFAULT_NS = exports.ALERTS_INDEX_PATTERN = exports.ALERTS_DATASET = void 0;
var _common = require("@kbn/fleet-plugin/common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PLUGIN_ID = exports.PLUGIN_ID = 'cloudDefend';
const PLUGIN_NAME = exports.PLUGIN_NAME = 'Cloud Defend';
const INTEGRATION_PACKAGE_NAME = exports.INTEGRATION_PACKAGE_NAME = 'cloud_defend';
const INPUT_CONTROL = exports.INPUT_CONTROL = 'cloud_defend/control';
const LOGS_CLOUD_DEFEND_PATTERN = exports.LOGS_CLOUD_DEFEND_PATTERN = 'logs-cloud_defend.*';
const ALERTS_DATASET = exports.ALERTS_DATASET = 'cloud_defend.alerts';
const ALERTS_INDEX_PATTERN = exports.ALERTS_INDEX_PATTERN = 'logs-cloud_defend.alerts*';
const ALERTS_INDEX_PATTERN_DEFAULT_NS = exports.ALERTS_INDEX_PATTERN_DEFAULT_NS = 'logs-cloud_defend.alerts-default';
const FILE_DATASET = exports.FILE_DATASET = 'cloud_defend.file';
const FILE_INDEX_PATTERN = exports.FILE_INDEX_PATTERN = 'logs-cloud_defend.file*';
const FILE_INDEX_PATTERN_DEFAULT_NS = exports.FILE_INDEX_PATTERN_DEFAULT_NS = 'logs-cloud_defend.file-default';
const PROCESS_DATASET = exports.PROCESS_DATASET = 'cloud_defend.process';
const PROCESS_INDEX_PATTERN = exports.PROCESS_INDEX_PATTERN = 'logs-cloud_defend.process*';
const PROCESS_INDEX_PATTERN_DEFAULT_NS = exports.PROCESS_INDEX_PATTERN_DEFAULT_NS = 'logs-cloud_defend.process-default';
const CURRENT_API_VERSION = exports.CURRENT_API_VERSION = '1';
const POLICIES_ROUTE_PATH = exports.POLICIES_ROUTE_PATH = '/internal/cloud_defend/policies';
const STATUS_ROUTE_PATH = exports.STATUS_ROUTE_PATH = '/internal/cloud_defend/status';
const CLOUD_DEFEND_FLEET_PACKAGE_KUERY = exports.CLOUD_DEFEND_FLEET_PACKAGE_KUERY = `${_common.PACKAGE_POLICY_SAVED_OBJECT_TYPE}.package.name:${INTEGRATION_PACKAGE_NAME}`;
const DEFAULT_POLICIES_PER_PAGE = exports.DEFAULT_POLICIES_PER_PAGE = 20;
const POLICIES_PACKAGE_POLICY_PREFIX = exports.POLICIES_PACKAGE_POLICY_PREFIX = 'package_policy.';
