// @kbn/core-saved-objects-migration-server-internal/src/core/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "REMOVED_TYPES", {
  enumerable: true,
  get: function () {
    return _unused_types.REMOVED_TYPES;
  }
});
Object.defineProperty(exports, "TransformSavedObjectDocumentError", {
  enumerable: true,
  get: function () {
    return _transform_saved_object_document_error.TransformSavedObjectDocumentError;
  }
});
Object.defineProperty(exports, "buildActiveMappings", {
  enumerable: true,
  get: function () {
    return _build_active_mappings.buildActiveMappings;
  }
});
Object.defineProperty(exports, "buildTypesMappings", {
  enumerable: true,
  get: function () {
    return _build_types_mappings.buildTypesMappings;
  }
});
Object.defineProperty(exports, "createIndexMap", {
  enumerable: true,
  get: function () {
    return _build_index_map.createIndexMap;
  }
});
Object.defineProperty(exports, "deterministicallyRegenerateObjectId", {
  enumerable: true,
  get: function () {
    return _regenerate_object_id.deterministicallyRegenerateObjectId;
  }
});
Object.defineProperty(exports, "excludeUnusedTypesQuery", {
  enumerable: true,
  get: function () {
    return _unused_types.excludeUnusedTypesQuery;
  }
});
Object.defineProperty(exports, "getBaseMappings", {
  enumerable: true,
  get: function () {
    return _build_active_mappings.getBaseMappings;
  }
});
var _build_active_mappings = require("./build_active_mappings");
var _unused_types = require("./unused_types");
var _transform_saved_object_document_error = require("./transform_saved_object_document_error");
var _regenerate_object_id = require("./regenerate_object_id");
var _build_types_mappings = require("./build_types_mappings");
var _build_index_map = require("./build_index_map");
