// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/pick_fields_step/components/summary_count_field/summary_count_field.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SummaryCountField = void 0;
var _react = _interopRequireWildcard(require("react"));
var _summary_count_field_select = require("./summary_count_field_select");
var _job_creator_context = require("../../../job_creator_context");
var _new_job_capabilities_service = require("../../../../../../../services/new_job_capabilities/new_job_capabilities_service");
var _description = require("./description");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SummaryCountField = () => {
  const {
    jobCreator: jc,
    jobCreatorUpdate,
    jobCreatorUpdated,
    jobValidator,
    jobValidatorUpdated
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const jobCreator = jc;
  const {
    fields
  } = _new_job_capabilities_service.newJobCapsService;
  const [summaryCountFieldName, setSummaryCountFieldName] = (0, _react.useState)(jobCreator.summaryCountFieldName);
  const [validation, setValidation] = (0, _react.useState)(jobValidator.summaryCountField);
  (0, _react.useEffect)(() => {
    setValidation(jobValidator.summaryCountField);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobValidatorUpdated]);
  (0, _react.useEffect)(() => {
    jobCreator.summaryCountFieldName = summaryCountFieldName;
    jobCreatorUpdate();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [summaryCountFieldName]);
  (0, _react.useEffect)(() => {
    setSummaryCountFieldName(jobCreator.summaryCountFieldName);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreatorUpdated]);
  return /*#__PURE__*/_react.default.createElement(_description.Description, {
    validation: validation
  }, /*#__PURE__*/_react.default.createElement(_summary_count_field_select.SummaryCountFieldSelect, {
    fields: fields,
    changeHandler: setSummaryCountFieldName,
    selectedField: summaryCountFieldName
  }));
};
exports.SummaryCountField = SummaryCountField;
