// @kbn/security-solution-plugin/public/entity_analytics/pages/entity_analytics_dashboard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EntityAnalyticsPage = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _search_strategy = require("../../../common/search_strategy");
var _translations = require("../../app/translations");
var _spy_routes = require("../../common/utils/route/spy_routes");
var _types = require("../../app/types");
var _containers = require("../../sourcerer/containers");
var _page_wrapper = require("../../common/components/page_wrapper");
var _header_page = require("../../common/components/header_page");
var _empty_prompt = require("../../common/components/empty_prompt");
var _search_bar = require("../../common/components/search_bar");
var _constants = require("../../common/store/inputs/constants");
var _filters_global = require("../../common/components/filters_global");
var _use_risk_engine_status = require("../api/hooks/use_risk_engine_status");
var _risk_score_update_panel = require("../components/risk_score_update_panel");
var _helper_hooks = require("../../helper_hooks");
var _entity_analytics_header = require("../components/entity_analytics_header");
var _entity_analytics_anomalies = require("../components/entity_analytics_anomalies");
var _entity_analytics_risk_score = require("../components/entity_analytics_risk_score");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EntityAnalyticsComponent = () => {
  const {
    data: riskScoreEngineStatus
  } = (0, _use_risk_engine_status.useRiskEngineStatus)();
  const {
    indicesExist,
    loading: isSourcererLoading,
    sourcererDataView
  } = (0, _containers.useSourcererDataView)();
  const isRiskScoreModuleLicenseAvailable = (0, _helper_hooks.useHasSecurityCapability)('entity-analytics');
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, indicesExist ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_filters_global.FiltersGlobal, null, /*#__PURE__*/_react.default.createElement(_search_bar.SiemSearchBar, {
    id: _constants.InputsModelId.global,
    sourcererDataView: sourcererDataView
  })), /*#__PURE__*/_react.default.createElement(_page_wrapper.SecuritySolutionPageWrapper, {
    "data-test-subj": "entityAnalyticsPage"
  }, /*#__PURE__*/_react.default.createElement(_header_page.HeaderPage, {
    title: _translations.ENTITY_ANALYTICS
  }), isSourcererLoading ? /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
    size: "l",
    "data-test-subj": "entityAnalyticsLoader"
  }) : /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    "data-test-subj": "entityAnalyticsSections"
  }, (riskScoreEngineStatus === null || riskScoreEngineStatus === void 0 ? void 0 : riskScoreEngineStatus.isUpdateAvailable) && isRiskScoreModuleLicenseAvailable && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_risk_score_update_panel.RiskScoreUpdatePanel, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_entity_analytics_header.EntityAnalyticsHeader, null)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_entity_analytics_risk_score.EntityAnalyticsRiskScores, {
    riskEntity: _search_strategy.RiskScoreEntity.host
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_entity_analytics_risk_score.EntityAnalyticsRiskScores, {
    riskEntity: _search_strategy.RiskScoreEntity.user
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_entity_analytics_anomalies.EntityAnalyticsAnomalies, null))))) : /*#__PURE__*/_react.default.createElement(_empty_prompt.EmptyPrompt, null), /*#__PURE__*/_react.default.createElement(_spy_routes.SpyRoute, {
    pageName: _types.SecurityPageName.entityAnalytics
  }));
};
const EntityAnalyticsPage = exports.EntityAnalyticsPage = /*#__PURE__*/_react.default.memo(EntityAnalyticsComponent);
