// @kbn/elastic-assistant-common/impl/schemas/attack_discovery/post_attack_discovery_route.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AttackDiscoveryPostResponse = exports.AttackDiscoveryPostRequestBody = void 0;
var _zod = require("zod");
var _bulk_crud_anonymization_fields_route = require("../anonymization_fields/bulk_crud_anonymization_fields_route.gen");
var _common_attributes = require("../conversations/common_attributes.gen");
var _common_attributes2 = require("./common_attributes.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Post Attack discovery API endpoint
 *   version: 1
 */

const AttackDiscoveryPostRequestBody = exports.AttackDiscoveryPostRequestBody = _zod.z.object({
  alertsIndexPattern: _zod.z.string(),
  anonymizationFields: _zod.z.array(_bulk_crud_anonymization_fields_route.AnonymizationFieldResponse),
  /**
   * LLM API configuration.
   */
  apiConfig: _common_attributes.ApiConfig,
  langSmithProject: _zod.z.string().optional(),
  langSmithApiKey: _zod.z.string().optional(),
  model: _zod.z.string().optional(),
  replacements: _common_attributes.Replacements.optional(),
  size: _zod.z.number(),
  subAction: _zod.z.enum(['invokeAI', 'invokeStream'])
});
const AttackDiscoveryPostResponse = exports.AttackDiscoveryPostResponse = _common_attributes2.AttackDiscoveryResponse;
