// @kbn/security-solution-plugin/public/management/pages/blocklist/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SEARCHABLE_FIELDS = exports.BLOCKLISTS_LIST_TYPE = exports.BLOCKLISTS_LIST_DEFINITION = void 0;
var _securitysolutionIoTsListTypes = require("@kbn/securitysolution-io-ts-list-types");
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const BLOCKLISTS_LIST_TYPE = exports.BLOCKLISTS_LIST_TYPE = _securitysolutionIoTsListTypes.ExceptionListTypeEnum.ENDPOINT_BLOCKLISTS;
const BLOCKLISTS_LIST_DEFINITION = exports.BLOCKLISTS_LIST_DEFINITION = {
  name: _securitysolutionListConstants.ENDPOINT_BLOCKLISTS_LIST_NAME,
  namespace_type: 'agnostic',
  description: _securitysolutionListConstants.ENDPOINT_BLOCKLISTS_LIST_DESCRIPTION,
  list_id: _securitysolutionListConstants.ENDPOINT_BLOCKLISTS_LIST_ID,
  type: BLOCKLISTS_LIST_TYPE
};
const SEARCHABLE_FIELDS = exports.SEARCHABLE_FIELDS = [`name`, `description`, 'item_id', `entries.value`, `entries.entries.value`, `comments.comment`];
