// @kbn/securitysolution-io-ts-list-types/src/request/create_rule_exception_item_schema/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createRuleExceptionListItemSchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _securitysolutionIoTsTypes = require("@kbn/securitysolution-io-ts-types");
var _common = require("../../common");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const createRuleExceptionListItemSchema = exports.createRuleExceptionListItemSchema = t.intersection([t.exact(t.type({
  description: _common.description,
  entries: _common.nonEmptyEntriesArray,
  name: _common.name,
  type: _common.exceptionListItemType
})), t.exact(t.partial({
  comments: _common.DefaultCreateCommentsArray,
  // defaults to empty array if not set during decode
  item_id: _securitysolutionIoTsTypes.DefaultUuid,
  // defaults to GUID (uuid v4) if not set during decode
  list_id: t.undefined,
  meta: _common.meta,
  // defaults to undefined if not set during decode
  namespace_type: _common.namespaceType,
  // defaults to 'single' if not set during decode
  os_types: _common.osTypeArrayOrUndefined,
  // defaults to empty array if not set during decode
  tags: _common.tags,
  // defaults to empty array if not set during decode
  expire_time: _common.expireTimeOrUndefined
}))]);

// This type is used after a decode since some things are defaults after a decode.
