// @kbn/security-solution-plugin/public/flyout/document_details/left/components/related_alerts_by_session.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RelatedAlertsBySession = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _correlations_details_alerts_table = require("./correlations_details_alerts_table");
var _use_fetch_related_alerts_by_session = require("../../shared/hooks/use_fetch_related_alerts_by_session");
var _test_ids = require("./test_ids");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Show related alerts by session in an expandable panel with a table
 */
const RelatedAlertsBySession = ({
  entityId,
  scopeId,
  eventId
}) => {
  const {
    loading,
    error,
    data,
    dataCount
  } = (0, _use_fetch_related_alerts_by_session.useFetchRelatedAlertsBySession)({
    entityId,
    scopeId
  });
  if (error) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_correlations_details_alerts_table.CorrelationsDetailsAlertsTable, {
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.flyout.left.insights.correlations.sessionAlertsTitle",
      defaultMessage: "{count} {count, plural, one {alert} other {alerts}} related by session",
      values: {
        count: dataCount
      }
    }),
    loading: loading,
    alertIds: data,
    scopeId: scopeId,
    eventId: eventId,
    noItemsMessage: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.securitySolution.flyout.left.insights.correlations.sessionAlertsNoDataDescription",
      defaultMessage: "No alerts related by session."
    }),
    "data-test-subj": _test_ids.CORRELATIONS_DETAILS_BY_SESSION_SECTION_TEST_ID
  });
};
exports.RelatedAlertsBySession = RelatedAlertsBySession;
RelatedAlertsBySession.displayName = 'RelatedAlertsBySession';
