// @kbn/vis-type-timelion-plugin/public/timelion_options.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = TimelionOptions;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _public = require("@kbn/kibana-react-plugin/public");
var _components = require("./components");
require("./timelion_options.scss");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function TimelionOptions({
  services,
  stateParams,
  setValue,
  setValidity
}) {
  const setInterval = (0, _react.useCallback)(value => setValue('interval', value), [setValue]);
  const setExpressionInput = (0, _react.useCallback)(value => setValue('expression', value), [setValue]);
  return /*#__PURE__*/_react.default.createElement(_public.KibanaContextProvider, {
    services: services
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    className: "visEditorSidebar__timelionOptions",
    paddingSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_components.TimelionInterval, {
    value: stateParams.interval,
    setValue: setInterval,
    setValidity: setValidity
  }), /*#__PURE__*/_react.default.createElement(_components.TimelionExpressionInput, {
    value: stateParams.expression,
    setValue: setExpressionInput
  })));
}

// default export required for React.Lazy
// eslint-disable-next-line import/no-default-export
module.exports = exports.default;
