// @kbn/observability-onboarding-plugin/public/application/quickstart_flows/shared/troubleshooting_link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TroubleshootingLink = TroubleshootingLink;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function TroubleshootingLink() {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    justifyContent: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    "data-test-subj": "observabilityOnboardingTroubleshootingLinkTroubleshootingButton",
    iconType: "help",
    color: "primary",
    href: "https://www.elastic.co/guide/en/observability/current/logs-troubleshooting.html",
    target: "_blank"
  }, _i18n.i18n.translate('xpack.observability_onboarding.installElasticAgent.troubleshooting', {
    defaultMessage: 'Troubleshooting'
  })));
}
