// @kbn/remote-clusters-plugin/public/application/sections/components/configured_by_node_warning/configured_by_node_warning.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfiguredByNodeWarning = ConfiguredByNodeWarning;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function ConfiguredByNodeWarning() {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.remoteClusters.configuredByNodeWarningTitle",
      defaultMessage: "You can't edit or delete this remote cluster because it's defined in a node's elasticsearch.yml configuration file."
    }),
    color: "primary",
    iconType: "iInCircle",
    "data-test-subj": "remoteClusterConfiguredByNodeWarning"
  });
}
