// @kbn/ml-plugin/public/application/components/custom_urls/custom_urls_wrapper.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CustomUrlsWrapper = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _dashboard_service = require("../../services/dashboard_service");
var _kibana = require("../../contexts/kibana");
var _custom_urls = require("./custom_urls");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CustomUrlsWrapper = props => {
  const {
    services: {
      data: {
        query: {
          timefilter: {
            timefilter
          }
        }
      }
    }
  } = (0, _kibana.useMlKibana)();
  const dashboardService = (0, _dashboard_service.useDashboardService)();
  return /*#__PURE__*/_react.default.createElement(_custom_urls.CustomUrls, (0, _extends2.default)({}, props, {
    currentTimeFilter: timefilter.getTime(),
    dashboardService: dashboardService
  }));
};
exports.CustomUrlsWrapper = CustomUrlsWrapper;
