// @kbn/security-solution-plugin/public/overview/components/link_panel/disabled_link_panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DisabledLinkPanel = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _inner_link_panel = require("./inner_link_panel");
var _links = require("../../../common/components/links");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DisabledLinkPanelComponent = ({
  bodyCopy,
  buttonCopy,
  dataTestSubjPrefix,
  docLink,
  learnMoreUrl,
  LinkPanelViewComponent,
  listItems,
  moreButtons,
  titleCopy
}) => {
  return /*#__PURE__*/_react.default.createElement(LinkPanelViewComponent, {
    listItems: listItems,
    splitPanel: /*#__PURE__*/_react.default.createElement(_inner_link_panel.InnerLinkPanel, {
      body: bodyCopy,
      button: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, buttonCopy && docLink && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_links.LinkButton, {
        color: "warning",
        href: docLink,
        target: "_blank",
        "data-test-subj": `${dataTestSubjPrefix}-enable-module-button`
      }, buttonCopy)), moreButtons && moreButtons),
      color: "warning",
      dataTestSubj: `${dataTestSubjPrefix}-inner-panel-danger`,
      learnMoreLink: learnMoreUrl,
      title: titleCopy
    })
  });
};
const DisabledLinkPanel = exports.DisabledLinkPanel = /*#__PURE__*/(0, _react.memo)(DisabledLinkPanelComponent);
DisabledLinkPanel.displayName = 'DisabledLinkPanel';
