// @kbn/aiops-change-point-detection/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EMBEDDABLE_CHANGE_POINT_CHART_TYPE = exports.CHANGE_POINT_DETECTION_VIEW_TYPE = exports.CHANGE_POINT_DETECTION_ENABLED = exports.CHANGE_POINT_CHART_DATA_VIEW_REF_NAME = exports.CASES_ATTACHMENT_CHANGE_POINT_CHART = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * This is an internal hard coded feature flag so we can easily turn on/off the
 * "Change Point Detection UI" during development until the first release.
 */
const CHANGE_POINT_DETECTION_ENABLED = exports.CHANGE_POINT_DETECTION_ENABLED = true;
const CASES_ATTACHMENT_CHANGE_POINT_CHART = exports.CASES_ATTACHMENT_CHANGE_POINT_CHART = 'aiopsChangePointChart';
const EMBEDDABLE_CHANGE_POINT_CHART_TYPE = exports.EMBEDDABLE_CHANGE_POINT_CHART_TYPE = 'aiopsChangePointChart';
const CHANGE_POINT_CHART_DATA_VIEW_REF_NAME = exports.CHANGE_POINT_CHART_DATA_VIEW_REF_NAME = 'aiopsChangePointChartDataViewId';
const CHANGE_POINT_DETECTION_VIEW_TYPE = exports.CHANGE_POINT_DETECTION_VIEW_TYPE = {
  CHARTS: 'charts',
  TABLE: 'table'
};
