// @kbn/security-solution-plugin/public/explore/network/components/source_destination/field_names.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NETWORK_TRANSPORT_FIELD_NAME = exports.NETWORK_PROTOCOL_FIELD_NAME = exports.NETWORK_PACKETS_FIELD_NAME = exports.NETWORK_DIRECTION_FIELD_NAME = exports.NETWORK_COMMUNITY_ID_FIELD_NAME = exports.NETWORK_BYTES_FIELD_NAME = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const NETWORK_BYTES_FIELD_NAME = exports.NETWORK_BYTES_FIELD_NAME = 'network.bytes';
const NETWORK_COMMUNITY_ID_FIELD_NAME = exports.NETWORK_COMMUNITY_ID_FIELD_NAME = 'network.community_id';
const NETWORK_DIRECTION_FIELD_NAME = exports.NETWORK_DIRECTION_FIELD_NAME = 'network.direction';
const NETWORK_PACKETS_FIELD_NAME = exports.NETWORK_PACKETS_FIELD_NAME = 'network.packets';
const NETWORK_PROTOCOL_FIELD_NAME = exports.NETWORK_PROTOCOL_FIELD_NAME = 'network.protocol';
const NETWORK_TRANSPORT_FIELD_NAME = exports.NETWORK_TRANSPORT_FIELD_NAME = 'network.transport';
