// @kbn/data-view-field-editor-plugin/public/components/preview/field_preview_empty_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FieldPreviewEmptyPrompt = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const FieldPreviewEmptyPrompt = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    style: {
      height: '100%'
    },
    "data-test-subj": "emptyPrompt"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    iconType: "inspect",
    title: /*#__PURE__*/_react.default.createElement("h2", null, _i18n.i18n.translate('indexPatternFieldEditor.fieldPreview.emptyPromptTitle', {
      defaultMessage: 'Preview'
    })),
    titleSize: "s",
    body: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "s"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
      color: "subdued"
    }, /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('indexPatternFieldEditor.fieldPreview.emptyPromptDescription', {
      defaultMessage: 'Enter the name of an existing field or define a script to view a preview of the calculated output.'
    }))))
  })));
};
exports.FieldPreviewEmptyPrompt = FieldPreviewEmptyPrompt;
