// @kbn/data-plugin/public/utils/table_inspector_view/components/data_view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _data_table = require("./data_table");
var _data_table_selector = require("./data_table_selector");
var _download_options = require("./download_options");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

class DataViewComponent extends _react.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "state", {});
    (0, _defineProperty2.default)(this, "onUpdateData", tables => {
      const keys = Object.keys(tables);
      const datatable = keys.length ? tables[keys[0]] : undefined;
      if (datatable) {
        this.setState({
          datatable
        });
      }
    });
    (0, _defineProperty2.default)(this, "selectTable", datatable => {
      if (datatable !== this.state.datatable) {
        this.setState({
          datatable
        });
      }
    });
  }
  static getDerivedStateFromProps(nextProps, state) {
    if (state && nextProps.adapters === state.adapters) {
      return null;
    }
    const {
      tables
    } = nextProps.adapters.tables;
    const keys = Object.keys(tables);
    const datatable = keys.length ? tables[keys[0]] : undefined;
    return {
      adapters: nextProps.adapters,
      datatable
    };
  }
  componentDidMount() {
    this.props.adapters.tables.on('change', this.onUpdateData);
  }
  componentWillUnmount() {
    this.props.adapters.tables.removeListener('change', this.onUpdateData);
  }
  static renderNoData() {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
      title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "data.inspector.table.noDataAvailableTitle",
        defaultMessage: "No data available"
      })),
      body: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "data.inspector.table.noDataAvailableDescription",
        defaultMessage: "The element did not provide any data."
      })))
    });
  }
  render() {
    var _this$props$options;
    if (!this.state.datatable) {
      return DataViewComponent.renderNoData();
    }
    const datatables = Object.values(this.state.adapters.tables.tables);
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: true
    }, datatables.length > 1 ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "xs"
    }, /*#__PURE__*/_react.default.createElement("p", {
      role: "status",
      "aria-live": "polite",
      "aria-atomic": "true"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      "data-test-subj": "inspectorDataViewSelectorLabel",
      id: "data.inspector.table.tablesDescription",
      defaultMessage: "There are {tablesCount, plural, one {# table} other {# tables} } in total",
      values: {
        tablesCount: datatables.length
      }
    }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "xs"
    }), /*#__PURE__*/_react.default.createElement(_data_table_selector.TableSelector, {
      tables: datatables,
      selectedTable: this.state.datatable,
      onTableChanged: this.selectTable
    })) : null), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_download_options.DataDownloadOptions, {
      title: ((_this$props$options = this.props.options) === null || _this$props$options === void 0 ? void 0 : _this$props$options.fileName) || this.props.title,
      uiSettings: this.props.uiSettings,
      datatables: datatables,
      fieldFormats: this.props.fieldFormats
    }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    }), /*#__PURE__*/_react.default.createElement(_data_table.DataTableFormat, {
      data: this.state.datatable,
      uiSettings: this.props.uiSettings,
      fieldFormats: this.props.fieldFormats,
      uiActions: this.props.uiActions,
      isFilterable: this.props.isFilterable
    }));
  }
}

// default export required for React.Lazy
// eslint-disable-next-line import/no-default-export
(0, _defineProperty2.default)(DataViewComponent, "propTypes", {
  adapters: _propTypes.default.object.isRequired,
  title: _propTypes.default.string.isRequired,
  uiSettings: _propTypes.default.object,
  uiActions: _propTypes.default.object.isRequired,
  fieldFormats: _propTypes.default.object.isRequired,
  isFilterable: _propTypes.default.func.isRequired
});
var _default = exports.default = DataViewComponent;
module.exports = exports.default;
