// @kbn/synthetics-plugin/public/apps/synthetics/state/ui/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.toggleIntegrationsPopover = exports.setSelectedMonitorId = exports.setSearchTextAction = exports.setRefreshPausedAction = exports.setRefreshIntervalAction = exports.setEsKueryString = exports.setBasePath = exports.setAlertFlyoutVisible = void 0;
var _toolkit = require("@reduxjs/toolkit");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const setAlertFlyoutVisible = exports.setAlertFlyoutVisible = (0, _toolkit.createAction)('[UI] TOGGLE ALERT FLYOUT');
const setBasePath = exports.setBasePath = (0, _toolkit.createAction)('[UI] SET BASE PATH');
const setEsKueryString = exports.setEsKueryString = (0, _toolkit.createAction)('[UI] SET ES KUERY STRING');
const setSearchTextAction = exports.setSearchTextAction = (0, _toolkit.createAction)('[UI] SET SEARCH');
const toggleIntegrationsPopover = exports.toggleIntegrationsPopover = (0, _toolkit.createAction)('[UI] TOGGLE INTEGRATION POPOVER STATE');
const setSelectedMonitorId = exports.setSelectedMonitorId = (0, _toolkit.createAction)('[UI] SET MONITOR ID');
const setRefreshPausedAction = exports.setRefreshPausedAction = (0, _toolkit.createAction)('[UI] SET REFRESH PAUSED');
const setRefreshIntervalAction = exports.setRefreshIntervalAction = (0, _toolkit.createAction)('[UI] SET REFRESH INTERVAL');
