// @kbn/discover-plugin/public/context_awareness/profiles/root_profile.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SolutionType = exports.RootProfileService = void 0;
var _profile_service = require("../profile_service");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let SolutionType = exports.SolutionType = /*#__PURE__*/function (SolutionType) {
  SolutionType["Observability"] = "oblt";
  SolutionType["Security"] = "security";
  SolutionType["Search"] = "search";
  SolutionType["Default"] = "default";
  return SolutionType;
}({});
class RootProfileService extends _profile_service.AsyncProfileService {
  constructor() {
    super({
      profileId: 'default-root-profile',
      solutionType: SolutionType.Default
    });
  }
}
exports.RootProfileService = RootProfileService;
