// @kbn/enterprise-search-plugin/public/applications/app_search/components/api_logs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "API_LOGS_TITLE", {
  enumerable: true,
  get: function () {
    return _constants.API_LOGS_TITLE;
  }
});
Object.defineProperty(exports, "ApiLogFlyout", {
  enumerable: true,
  get: function () {
    return _api_log.ApiLogFlyout;
  }
});
Object.defineProperty(exports, "ApiLogs", {
  enumerable: true,
  get: function () {
    return _api_logs.ApiLogs;
  }
});
Object.defineProperty(exports, "ApiLogsLogic", {
  enumerable: true,
  get: function () {
    return _api_logs_logic.ApiLogsLogic;
  }
});
Object.defineProperty(exports, "ApiLogsTable", {
  enumerable: true,
  get: function () {
    return _components.ApiLogsTable;
  }
});
Object.defineProperty(exports, "NewApiEventsPrompt", {
  enumerable: true,
  get: function () {
    return _components.NewApiEventsPrompt;
  }
});
var _constants = require("./constants");
var _components = require("./components");
var _api_log = require("./api_log");
var _api_logs = require("./api_logs");
var _api_logs_logic = require("./api_logs_logic");
