// @kbn/uptime-plugin/public/legacy_uptime/lib/helper/observability_integration/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getInfraContainerHref", {
  enumerable: true,
  get: function () {
    return _get_infra_href.getInfraContainerHref;
  }
});
Object.defineProperty(exports, "getInfraIpHref", {
  enumerable: true,
  get: function () {
    return _get_infra_href.getInfraIpHref;
  }
});
Object.defineProperty(exports, "getInfraKubernetesHref", {
  enumerable: true,
  get: function () {
    return _get_infra_href.getInfraKubernetesHref;
  }
});
Object.defineProperty(exports, "getLegacyApmHref", {
  enumerable: true,
  get: function () {
    return _get_apm_href.getLegacyApmHref;
  }
});
Object.defineProperty(exports, "getLoggingContainerHref", {
  enumerable: true,
  get: function () {
    return _get_logging_href.getLoggingContainerHref;
  }
});
Object.defineProperty(exports, "getLoggingIpHref", {
  enumerable: true,
  get: function () {
    return _get_logging_href.getLoggingIpHref;
  }
});
Object.defineProperty(exports, "getLoggingKubernetesHref", {
  enumerable: true,
  get: function () {
    return _get_logging_href.getLoggingKubernetesHref;
  }
});
var _get_apm_href = require("./get_apm_href");
var _get_infra_href = require("./get_infra_href");
var _get_logging_href = require("./get_logging_href");
