// @kbn/threat-intelligence-plugin/public/modules/indicators/components/barchart/test_ids.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TIMELINE_BUTTON_TEST_ID = exports.POPOVER_BUTTON_TEST_ID = exports.LOADING_TEST_ID = exports.FILTER_OUT_BUTTON_TEST_ID = exports.FILTER_IN_BUTTON_TEST_ID = exports.DROPDOWN_TEST_ID = exports.COPY_TO_CLIPBOARD_BUTTON_TEST_ID = exports.CHART_UPDATE_PROGRESS_TEST_ID = exports.BARCHART_WRAPPER_TEST_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/* Wrapper */
const CHART_UPDATE_PROGRESS_TEST_ID = exports.CHART_UPDATE_PROGRESS_TEST_ID = 'tiBarchartWrapper-updating';
const LOADING_TEST_ID = exports.LOADING_TEST_ID = 'tiBarchartWrapper-loading';
const BARCHART_WRAPPER_TEST_ID = exports.BARCHART_WRAPPER_TEST_ID = 'tiBarchartWrapper';

/* Field Selector */
const DROPDOWN_TEST_ID = exports.DROPDOWN_TEST_ID = 'tiIndicatorFieldSelectorDropdown';

/* Legend Action */
const POPOVER_BUTTON_TEST_ID = exports.POPOVER_BUTTON_TEST_ID = 'tiBarchartPopoverButton';
const TIMELINE_BUTTON_TEST_ID = exports.TIMELINE_BUTTON_TEST_ID = 'tiBarchartTimelineButton';
const FILTER_IN_BUTTON_TEST_ID = exports.FILTER_IN_BUTTON_TEST_ID = 'tiBarchartFilterInButton';
const FILTER_OUT_BUTTON_TEST_ID = exports.FILTER_OUT_BUTTON_TEST_ID = 'tiBarchartFilterOutButton';
const COPY_TO_CLIPBOARD_BUTTON_TEST_ID = exports.COPY_TO_CLIPBOARD_BUTTON_TEST_ID = 'tiBarchartCopyToClipboardButton';
