// @kbn/security-solution-plugin/common/api/timeline/persist_note/persist_note_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.persistNoteWithoutRefSchema = exports.persistNoteWithRefSchema = exports.persistNoteSchema = void 0;
var runtimeTypes = _interopRequireWildcard(require("io-ts"));
var _utility_types = require("../../../utility_types");
var _api = require("../model/api");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const persistNoteWithRefSchema = exports.persistNoteWithRefSchema = runtimeTypes.intersection([runtimeTypes.type({
  note: _api.BareNoteSchema
}), runtimeTypes.partial({
  overrideOwner: (0, _utility_types.unionWithNullType)(runtimeTypes.boolean),
  noteId: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  version: (0, _utility_types.unionWithNullType)(runtimeTypes.string)
})]);
const persistNoteWithoutRefSchema = exports.persistNoteWithoutRefSchema = runtimeTypes.intersection([runtimeTypes.type({
  note: _api.BareNoteWithoutExternalRefsSchema
}), runtimeTypes.partial({
  overrideOwner: (0, _utility_types.unionWithNullType)(runtimeTypes.boolean),
  noteId: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  version: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  eventIngested: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  eventTimestamp: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  eventDataView: (0, _utility_types.unionWithNullType)(runtimeTypes.string)
})]);
const persistNoteSchema = exports.persistNoteSchema = runtimeTypes.union([persistNoteWithRefSchema, persistNoteWithoutRefSchema]);
