// @kbn/es-ui-shared-plugin/static/forms/components/fields/file_picker_field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FilePickerField = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _hook_form_lib = require("../../hook_form_lib");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const ONE_MEGABYTE = 1048576;
const FilePickerField = ({
  field,
  euiFieldProps = {},
  idAria,
  maxFileSize = ONE_MEGABYTE,
  ...rest
}) => {
  const {
    isInvalid,
    errorMessage
  } = (0, _hook_form_lib.getFieldValidityAndErrorMessage)(field);
  const onChange = (0, _react.useCallback)(async files => {
    if (files) {
      const filesArr = [];
      try {
        for (const file of files) {
          if (file.size > maxFileSize) throw new Error(`${file.name} is too large, maximum size is ${Math.floor(maxFileSize) / 1024}kb`);
          const fileBuffer = await file.arrayBuffer();
          // Base64 encode the file
          const fileData = window.btoa(String.fromCharCode(...new Uint8Array(fileBuffer)));
          filesArr.push(fileData);
        }
        if (!euiFieldProps.multiple) field.setValue(filesArr[0]);else field.setValue(filesArr);
      } catch (e) {
        field.setErrors([e]);
      }
    } else {
      field.setValue(null);
    }
  }, [field, maxFileSize, euiFieldProps]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, (0, _extends2.default)({
    label: field.label,
    labelAppend: field.labelAppend,
    helpText: typeof field.helpText === 'function' ? field.helpText() : field.helpText,
    error: errorMessage,
    isInvalid: isInvalid,
    fullWidth: true,
    describedByIds: idAria ? [idAria] : undefined
  }, rest), /*#__PURE__*/_react.default.createElement(_eui.EuiFilePicker, (0, _extends2.default)({
    isInvalid: isInvalid,
    onChange: onChange,
    isLoading: field.isValidating,
    "data-test-subj": "input"
  }, euiFieldProps)));
};
exports.FilePickerField = FilePickerField;
