// @kbn/observability-ai-assistant-app-plugin/public/application.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Application = Application;
var _typedReactRouterConfig = require("@kbn/typed-react-router-config");
var _react = _interopRequireDefault(require("react"));
var _config = require("./routes/config");
var _shared_providers = require("./utils/shared_providers");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// This is the Conversation application.

function Application({
  coreStart,
  history,
  pluginsStart,
  service,
  theme$
}) {
  return /*#__PURE__*/_react.default.createElement(_shared_providers.SharedProviders, {
    coreStart: coreStart,
    pluginsStart: pluginsStart,
    service: service,
    theme$: theme$
  }, /*#__PURE__*/_react.default.createElement(_typedReactRouterConfig.RouterProvider, {
    history: history,
    router: _config.observabilityAIAssistantRouter
  }, /*#__PURE__*/_react.default.createElement(_typedReactRouterConfig.RouteRenderer, null)));
}
