// @kbn/cross-cluster-replication-plugin/public/shared_imports.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "APP_WRAPPER_CLASS", {
  enumerable: true,
  get: function () {
    return _public3.APP_WRAPPER_CLASS;
  }
});
Object.defineProperty(exports, "KibanaRenderContextProvider", {
  enumerable: true,
  get: function () {
    return _reactKibanaContextRender.KibanaRenderContextProvider;
  }
});
Object.defineProperty(exports, "PageError", {
  enumerable: true,
  get: function () {
    return _public.PageError;
  }
});
Object.defineProperty(exports, "PageLoading", {
  enumerable: true,
  get: function () {
    return _public.PageLoading;
  }
});
Object.defineProperty(exports, "SectionLoading", {
  enumerable: true,
  get: function () {
    return _public.SectionLoading;
  }
});
Object.defineProperty(exports, "extractQueryParams", {
  enumerable: true,
  get: function () {
    return _public.extractQueryParams;
  }
});
Object.defineProperty(exports, "indices", {
  enumerable: true,
  get: function () {
    return _public.indices;
  }
});
Object.defineProperty(exports, "useExecutionContext", {
  enumerable: true,
  get: function () {
    return _public2.useExecutionContext;
  }
});
var _public = require("@kbn/es-ui-shared-plugin/public");
var _public2 = require("@kbn/kibana-react-plugin/public");
var _reactKibanaContextRender = require("@kbn/react-kibana-context-render");
var _public3 = require("@kbn/core/public");
