// @kbn/interactive-setup-plugin/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VERIFICATION_CODE_LENGTH = exports.ERROR_PING_FAILURE = exports.ERROR_OUTSIDE_PREBOOT_STAGE = exports.ERROR_KIBANA_CONFIG_NOT_WRITABLE = exports.ERROR_KIBANA_CONFIG_FAILURE = exports.ERROR_ENROLL_FAILURE = exports.ERROR_ELASTICSEARCH_CONNECTION_CONFIGURED = exports.ERROR_CONFIGURE_FAILURE = exports.ERROR_COMPATIBILITY_FAILURE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const VERIFICATION_CODE_LENGTH = exports.VERIFICATION_CODE_LENGTH = 6;
const ERROR_OUTSIDE_PREBOOT_STAGE = exports.ERROR_OUTSIDE_PREBOOT_STAGE = 'outside_preboot_stage';
const ERROR_ELASTICSEARCH_CONNECTION_CONFIGURED = exports.ERROR_ELASTICSEARCH_CONNECTION_CONFIGURED = 'elasticsearch_connection_configured';
const ERROR_KIBANA_CONFIG_NOT_WRITABLE = exports.ERROR_KIBANA_CONFIG_NOT_WRITABLE = 'kibana_config_not_writable';
const ERROR_KIBANA_CONFIG_FAILURE = exports.ERROR_KIBANA_CONFIG_FAILURE = 'kibana_config_failure';
const ERROR_ENROLL_FAILURE = exports.ERROR_ENROLL_FAILURE = 'enroll_failure';
const ERROR_CONFIGURE_FAILURE = exports.ERROR_CONFIGURE_FAILURE = 'configure_failure';
const ERROR_PING_FAILURE = exports.ERROR_PING_FAILURE = 'ping_failure';
const ERROR_COMPATIBILITY_FAILURE = exports.ERROR_COMPATIBILITY_FAILURE = 'compatibility_failure';
