// @kbn/ml-plugin/public/application/data_frame_analytics/pages/analytics_management/hooks/use_create_analytics_form/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ACTION = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let ACTION = exports.ACTION = /*#__PURE__*/function (ACTION) {
  ACTION[ACTION["ADD_REQUEST_MESSAGE"] = 0] = "ADD_REQUEST_MESSAGE";
  ACTION[ACTION["RESET_REQUEST_MESSAGES"] = 1] = "RESET_REQUEST_MESSAGES";
  ACTION[ACTION["CLOSE_MODAL"] = 2] = "CLOSE_MODAL";
  ACTION[ACTION["OPEN_MODAL"] = 3] = "OPEN_MODAL";
  ACTION[ACTION["RESET_ADVANCED_EDITOR_MESSAGES"] = 4] = "RESET_ADVANCED_EDITOR_MESSAGES";
  ACTION[ACTION["RESET_FORM"] = 5] = "RESET_FORM";
  ACTION[ACTION["SET_ADVANCED_EDITOR_RAW_STRING"] = 6] = "SET_ADVANCED_EDITOR_RAW_STRING";
  ACTION[ACTION["SET_FORM_STATE"] = 7] = "SET_FORM_STATE";
  ACTION[ACTION["SET_DATA_VIEW_TITLES"] = 8] = "SET_DATA_VIEW_TITLES";
  ACTION[ACTION["SET_IS_JOB_CREATED"] = 9] = "SET_IS_JOB_CREATED";
  ACTION[ACTION["SET_IS_JOB_STARTED"] = 10] = "SET_IS_JOB_STARTED";
  ACTION[ACTION["SET_IS_MODAL_BUTTON_DISABLED"] = 11] = "SET_IS_MODAL_BUTTON_DISABLED";
  ACTION[ACTION["SET_IS_MODAL_VISIBLE"] = 12] = "SET_IS_MODAL_VISIBLE";
  ACTION[ACTION["SET_JOB_CONFIG"] = 13] = "SET_JOB_CONFIG";
  ACTION[ACTION["SET_JOB_IDS"] = 14] = "SET_JOB_IDS";
  ACTION[ACTION["SWITCH_TO_ADVANCED_EDITOR"] = 15] = "SWITCH_TO_ADVANCED_EDITOR";
  ACTION[ACTION["SWITCH_TO_FORM"] = 16] = "SWITCH_TO_FORM";
  ACTION[ACTION["SET_ESTIMATED_MODEL_MEMORY_LIMIT"] = 17] = "SET_ESTIMATED_MODEL_MEMORY_LIMIT";
  ACTION[ACTION["SET_JOB_CLONE"] = 18] = "SET_JOB_CLONE";
  return ACTION;
}({}); // Actions wrapping the dispatcher exposed by the custom hook
