// @kbn/session-view-plugin/public/components/tty_player_controls/styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStyles = void 0;
var _react = require("react");
var _hooks = require("../../hooks");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useStyles = () => {
  const {
    euiTheme,
    euiVars
  } = (0, _hooks.useEuiTheme)();
  const cached = (0, _react.useMemo)(() => {
    const {
      size,
      border
    } = euiTheme;
    const controlsPanel = {
      paddingTop: size.s,
      paddingBottom: size.s,
      borderBottomLeftRadius: border.radius.medium,
      borderBottomRightRadius: border.radius.medium,
      backgroundColor: euiVars.terminalOutputBackground
    };
    const controlButton = {
      width: size.l
    };
    return {
      controlsPanel,
      controlButton
    };
  }, [euiTheme, euiVars.terminalOutputBackground]);
  return cached;
};
exports.useStyles = useStyles;
