// @kbn/transform-plugin/public/app/sections/transform_management/components/action_start/use_start_action.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useStartAction = void 0;
var _react = _interopRequireWildcard(require("react"));
var _transform_list = require("../../../../common/transform_list");
var _constants = require("../../../../../../common/constants");
var _hooks = require("../../../../hooks");
var _start_action_name = require("./start_action_name");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const useStartAction = (forceDisable, transformNodes) => {
  const {
    canStartStopTransform
  } = (0, _hooks.useTransformCapabilities)();
  const startTransforms = (0, _hooks.useStartTransforms)();
  const [isModalVisible, setModalVisible] = (0, _react.useState)(false);
  const [items, setItems] = (0, _react.useState)([]);
  const closeModal = () => setModalVisible(false);
  const startAndCloseModal = () => {
    setModalVisible(false);
    startTransforms(items.map(i => ({
      id: i.id
    })));
  };
  const openModal = newItems => {
    if (Array.isArray(newItems)) {
      setItems(newItems);
      setModalVisible(true);
    }
  };
  const action = (0, _react.useMemo)(() => ({
    name: item => /*#__PURE__*/_react.default.createElement(_start_action_name.StartActionName, {
      items: [item],
      forceDisable: forceDisable,
      transformNodes: transformNodes
    }),
    available: item => (0, _transform_list.isTransformListRowWithStats)(item) ? item.stats.state === _constants.TRANSFORM_STATE.STOPPED : true,
    enabled: item => !(0, _start_action_name.isStartActionDisabled)([item], canStartStopTransform, transformNodes),
    description: _start_action_name.startActionNameText,
    icon: 'play',
    type: 'icon',
    onClick: item => openModal([item]),
    'data-test-subj': 'transformActionStart'
  }), [canStartStopTransform, forceDisable, transformNodes]);
  return {
    action,
    closeModal,
    isModalVisible,
    items,
    openModal,
    startAndCloseModal
  };
};
exports.useStartAction = useStartAction;
