// @kbn/logs-explorer-plugin/common/locators/logs_explorer/logs_explorer_locator.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogsExplorerLocatorDefinition = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _locators = require("@kbn/deeplinks-observability/locators");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class LogsExplorerLocatorDefinition {
  constructor(deps) {
    (0, _defineProperty2.default)(this, "id", _locators.LOGS_EXPLORER_LOCATOR_ID);
    (0, _defineProperty2.default)(this, "getLocation", params => {
      var _this$deps$discoverAp;
      const {
        dataset,
        columns
      } = params;
      const dataViewSpec = dataset ? {
        id: dataset,
        title: dataset
      } : undefined;
      const discoverColumns = columns === null || columns === void 0 ? void 0 : columns.map(column => {
        return column.type === 'document-field' ? column.field : column.smartField;
      });
      return (_this$deps$discoverAp = this.deps.discoverAppLocator) === null || _this$deps$discoverAp === void 0 ? void 0 : _this$deps$discoverAp.getLocation({
        ...params,
        columns: discoverColumns,
        dataViewId: dataset,
        dataViewSpec
      });
    });
    this.deps = deps;
  }
}
exports.LogsExplorerLocatorDefinition = LogsExplorerLocatorDefinition;
