// @kbn/deeplinks-search/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SERVERLESS_ES_SEARCH_PLAYGROUND_ID = exports.SERVERLESS_ES_SEARCH_INFERENCE_ENDPOINTS_ID = exports.SERVERLESS_ES_CONNECTORS_ID = exports.SERVERLESS_ES_APP_ID = exports.SEARCH_HOMEPAGE = exports.ENTERPRISE_SEARCH_WORKPLACESEARCH_APP_ID = exports.ENTERPRISE_SEARCH_RELEVANCE_APP_ID = exports.ENTERPRISE_SEARCH_CONTENT_APP_ID = exports.ENTERPRISE_SEARCH_APP_ID = exports.ENTERPRISE_SEARCH_APPSEARCH_APP_ID = exports.ENTERPRISE_SEARCH_APPLICATIONS_APP_ID = exports.ENTERPRISE_SEARCH_ANALYTICS_APP_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const ENTERPRISE_SEARCH_APP_ID = exports.ENTERPRISE_SEARCH_APP_ID = 'enterpriseSearch';
const ENTERPRISE_SEARCH_CONTENT_APP_ID = exports.ENTERPRISE_SEARCH_CONTENT_APP_ID = 'enterpriseSearchContent';
const ENTERPRISE_SEARCH_RELEVANCE_APP_ID = exports.ENTERPRISE_SEARCH_RELEVANCE_APP_ID = 'enterpriseSearchRelevance';
const ENTERPRISE_SEARCH_APPLICATIONS_APP_ID = exports.ENTERPRISE_SEARCH_APPLICATIONS_APP_ID = 'enterpriseSearchApplications';
const ENTERPRISE_SEARCH_ANALYTICS_APP_ID = exports.ENTERPRISE_SEARCH_ANALYTICS_APP_ID = 'enterpriseSearchAnalytics';
const ENTERPRISE_SEARCH_APPSEARCH_APP_ID = exports.ENTERPRISE_SEARCH_APPSEARCH_APP_ID = 'appSearch';
const ENTERPRISE_SEARCH_WORKPLACESEARCH_APP_ID = exports.ENTERPRISE_SEARCH_WORKPLACESEARCH_APP_ID = 'workplaceSearch';
const SERVERLESS_ES_APP_ID = exports.SERVERLESS_ES_APP_ID = 'serverlessElasticsearch';
const SERVERLESS_ES_CONNECTORS_ID = exports.SERVERLESS_ES_CONNECTORS_ID = 'serverlessConnectors';
const SERVERLESS_ES_SEARCH_PLAYGROUND_ID = exports.SERVERLESS_ES_SEARCH_PLAYGROUND_ID = 'searchPlayground';
const SERVERLESS_ES_SEARCH_INFERENCE_ENDPOINTS_ID = exports.SERVERLESS_ES_SEARCH_INFERENCE_ENDPOINTS_ID = 'searchInferenceEndpoints';
const SEARCH_HOMEPAGE = exports.SEARCH_HOMEPAGE = 'searchHomepage';
