// @kbn/data-quality-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DATA_QUALITY_LOCATOR_ID", {
  enumerable: true,
  get: function () {
    return _locators.DATA_QUALITY_LOCATOR_ID;
  }
});
Object.defineProperty(exports, "DATA_QUALITY_URL_STATE_KEY", {
  enumerable: true,
  get: function () {
    return _url_schema.DATA_QUALITY_URL_STATE_KEY;
  }
});
exports.PLUGIN_NAME = exports.PLUGIN_ID = void 0;
Object.defineProperty(exports, "datasetQualityUrlSchemaV1", {
  enumerable: true,
  get: function () {
    return _url_schema.datasetQualityUrlSchemaV1;
  }
});
var _i18n = require("@kbn/i18n");
var _url_schema = require("./url_schema");
var _locators = require("./locators");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PLUGIN_ID = exports.PLUGIN_ID = 'data_quality';
const PLUGIN_NAME = exports.PLUGIN_NAME = _i18n.i18n.translate('xpack.dataQuality.name', {
  defaultMessage: 'Data Set Quality'
});
