// @kbn/lens-plugin/public/shared_components/collapse_setting.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CollapseSetting = CollapseSetting;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const options = [{
  text: _i18n.i18n.translate('xpack.lens.collapse.none', {
    defaultMessage: 'None'
  }),
  value: ''
}, {
  text: _i18n.i18n.translate('xpack.lens.collapse.sum', {
    defaultMessage: 'Sum'
  }),
  value: 'sum'
}, {
  text: _i18n.i18n.translate('xpack.lens.collapse.min', {
    defaultMessage: 'Min'
  }),
  value: 'min'
}, {
  text: _i18n.i18n.translate('xpack.lens.collapse.max', {
    defaultMessage: 'Max'
  }),
  value: 'max'
}, {
  text: _i18n.i18n.translate('xpack.lens.collapse.avg', {
    defaultMessage: 'Average'
  }),
  value: 'avg'
}];
function CollapseSetting({
  value,
  onChange
}) {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    id: "lns-indexPattern-collapse-by",
    label: /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
      delay: "long",
      position: "top",
      content: _i18n.i18n.translate('xpack.lens.collapse.infoIcon', {
        defaultMessage: 'Do not show this dimension in the visualization and aggregate all metric values which have the same value for this dimension into a single number.'
      })
    }, /*#__PURE__*/_react.default.createElement("span", null, _i18n.i18n.translate('xpack.lens.collapse.label', {
      defaultMessage: 'Collapse by'
    }), '', /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      type: "questionInCircle",
      color: "subdued",
      size: "s",
      className: "eui-alignTop"
    }))),
    display: "rowCompressed",
    fullWidth: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSelect, {
    fullWidth: true,
    compressed: true,
    "data-test-subj": "indexPattern-collapse-by",
    options: options,
    value: value,
    onChange: e => {
      onChange(e.target.value);
    }
  })));
}
