// @kbn/lens-plugin/public/shared_components/range_input_field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RangeInputField = RangeInputField;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function RangeInputField({
  isInvalid,
  label,
  helpText,
  error,
  lowerValue,
  onLowerValueChange,
  onLowerValueBlur,
  upperValue,
  onUpperValueChange,
  onUpperValueBlur,
  testSubjLayout,
  testSubjLower,
  testSubjUpper
}) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    display: "columnCompressed",
    fullWidth: true,
    isInvalid: isInvalid,
    label: label,
    helpText: helpText,
    error: error
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFormControlLayoutDelimited, {
    "data-test-subj": testSubjLayout,
    compressed: true,
    startControl: /*#__PURE__*/_react.default.createElement(_eui.EuiFieldNumber, {
      compressed: true,
      value: lowerValue,
      isInvalid: isInvalid,
      "data-test-subj": testSubjLower,
      onChange: onLowerValueChange,
      onBlur: onLowerValueBlur,
      step: "any",
      controlOnly: true
    }),
    endControl: /*#__PURE__*/_react.default.createElement(_eui.EuiFieldNumber, {
      compressed: true,
      value: upperValue,
      "data-test-subj": testSubjUpper,
      onChange: onUpperValueChange,
      onBlur: onUpperValueBlur,
      step: "any",
      controlOnly: true
    })
  }));
}
