// @kbn/osquery-plugin/public/saved_queries/form/playground_flyout.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PlaygroundFlyout = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _reactHookForm = require("react-hook-form");
var _constants = require("../../../common/constants");
var _live_queries = require("../../live_queries");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const euiFlyoutHeaderCss = {
  '&.euiFlyoutHeader': {
    paddingTop: '21px',
    paddingBottom: '20px'
  }
};
const PlaygroundFlyoutComponent = ({
  enabled,
  onClose
}) => {
  // @ts-expect-error update types
  const {
    serializer,
    watch
  } = (0, _reactHookForm.useFormContext)();
  const watchedValues = watch();
  const {
    query,
    ecs_mapping: ecsMapping,
    id,
    timeout
  } = watchedValues;
  /* recalculate the form data when ecs_mapping changes */
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const serializedFormData = (0, _react.useMemo)(() => serializer(watchedValues), [ecsMapping]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyout, {
    type: "push",
    size: "m",
    onClose: onClose,
    "data-test-subj": 'osquery-save-query-flyout'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    css: euiFlyoutHeaderCss,
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("h5", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.queryPlaygroundFlyout.title",
    defaultMessage: "Test query"
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_live_queries.LiveQuery, {
    enabled: enabled && query !== '',
    formType: "simple",
    query: query,
    ecs_mapping: serializedFormData.ecs_mapping,
    savedQueryId: id,
    timeout: timeout || _constants.QUERY_TIMEOUT.DEFAULT,
    queryField: false,
    ecsMappingField: false
  })));
};
const PlaygroundFlyout = exports.PlaygroundFlyout = /*#__PURE__*/_react.default.memo(PlaygroundFlyoutComponent);
