// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/card_step/content/create_project_step_image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CreateProjectImage = void 0;
var _react = _interopRequireDefault(require("react"));
var _create_projects = _interopRequireDefault(require("../../images/create_projects.png"));
var _translations = require("../../translations");
var _content_wrapper = require("./content_wrapper");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CreateProjectImageComponent = () => /*#__PURE__*/_react.default.createElement(_content_wrapper.ContentWrapper, {
  shadow: false
}, /*#__PURE__*/_react.default.createElement("img", {
  src: _create_projects.default,
  alt: _translations.CREATE_PROJECT_TITLE,
  height: "100%",
  width: "100%"
}));
const CreateProjectImage = exports.CreateProjectImage = /*#__PURE__*/_react.default.memo(CreateProjectImageComponent);
