// @kbn/cases-plugin/public/components/all_cases/utility_bar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CasesTableUtilityBar = void 0;
var _react = require("@emotion/react");
var _react2 = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var i18n = _interopRequireWildcard(require("./translations"));
var _constants = require("../../../common/constants");
var _use_on_refresh_cases = require("./use_on_refresh_cases");
var _use_bulk_actions = require("./use_bulk_actions");
var _use_cases_context = require("../cases_context/use_cases_context");
var _columns_popover = require("./columns_popover");
var _use_cases_local_storage = require("../../common/use_cases_local_storage");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CasesTableUtilityBar = exports.CasesTableUtilityBar = /*#__PURE__*/_react2.default.memo(({
  isSelectorView,
  totalCases,
  selectedCases,
  deselectCases,
  pagination,
  selectedColumns,
  onSelectedColumnsChange,
  onClearFilters,
  showClearFiltersButton
}) => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const refreshCases = (0, _use_on_refresh_cases.useRefreshCases)();
  const {
    permissions
  } = (0, _use_cases_context.useCasesContext)();
  const [isPopoverOpen, setIsPopoverOpen] = (0, _react2.useState)(false);
  const [isMessageDismissed, setIsMessageDismissed] = (0, _react2.useState)(false);
  const localStorageKey = `cases.utilityBar.hideMaxLimitWarning`;
  const [doNotShowAgain, setDoNotShowAgain] = (0, _use_cases_local_storage.useCasesLocalStorage)(localStorageKey, false);
  const togglePopover = (0, _react2.useCallback)(() => setIsPopoverOpen(!isPopoverOpen), [isPopoverOpen]);
  const closePopover = (0, _react2.useCallback)(() => setIsPopoverOpen(false), []);
  const toggleWarning = (0, _react2.useCallback)(() => setIsMessageDismissed(!isMessageDismissed), [isMessageDismissed]);
  const onRefresh = (0, _react2.useCallback)(() => {
    deselectCases();
    refreshCases();
  }, [deselectCases, refreshCases]);
  const {
    panels,
    modals,
    flyouts
  } = (0, _use_bulk_actions.useBulkActions)({
    selectedCases,
    onAction: closePopover,
    onActionSuccess: onRefresh
  });
  const handleNotShowAgain = () => {
    setDoNotShowAgain(true);
  };

  /**
   * At least update or delete permissions needed to show bulk actions.
   * Granular permission check for each action is performed
   * in the useBulkActions hook.
   */
  const showBulkActions = (permissions.update || permissions.delete) && selectedCases.length > 0;
  const visibleCases = pagination !== null && pagination !== void 0 && pagination.pageSize && totalCases > pagination.pageSize ? pagination.pageSize : totalCases;
  const hasReachedMaxCases = pagination.pageSize && totalCases >= _constants.MAX_DOCS_PER_PAGE && pagination.pageSize * (pagination.pageIndex + 1) >= _constants.MAX_DOCS_PER_PAGE;
  const renderMaxLimitWarning = () => /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m"
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiText, {
    color: "default",
    size: "m",
    css: (0, _react.css)`
              margin-top: ${euiTheme.size.xs};
            `
  }, i18n.MAX_CASES(_constants.MAX_DOCS_PER_PAGE))), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiButtonEmpty, {
    size: "s",
    color: "warning",
    "data-test-subj": "dismiss-warning",
    onClick: toggleWarning
  }, i18n.DISMISS)), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiButtonEmpty, {
    size: "s",
    color: "warning",
    "data-test-subj": "do-not-show-warning",
    onClick: handleNotShowAgain
  }, i18n.NOT_SHOW_AGAIN)));
  return /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween",
    alignItems: "center",
    gutterSize: "s",
    css: (0, _react.css)`
            border-bottom: ${euiTheme.border.thin};
            padding-top: ${euiTheme.size.s};
            padding-bottom: ${euiTheme.size.s};
          `
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "flexStart",
    gutterSize: "s",
    alignItems: "center"
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "case-table-case-count",
    grow: false,
    css: (0, _react.css)`
                  border-right: ${euiTheme.border.thin};
                  padding-right: ${euiTheme.size.s};
                `
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiText, {
    size: "xs",
    color: "subdued"
  }, i18n.SHOWING_CASES(totalCases, visibleCases))), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "case-table-utility-bar-actions",
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    justifyContent: "flexStart",
    gutterSize: "s"
  }, !isSelectorView && showBulkActions && /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "case-table-selected-case-count",
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiText, {
    size: "xs",
    color: "subdued"
  }, i18n.SHOWING_SELECTED_CASES(selectedCases.length))), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiPopover, {
    isOpen: isPopoverOpen,
    closePopover: closePopover,
    panelPaddingSize: "none",
    "data-test-subj": "case-table-bulk-actions-popover",
    button: /*#__PURE__*/_react2.default.createElement(_eui.EuiButtonEmpty, {
      onClick: togglePopover,
      size: "xs",
      iconSide: "right",
      iconType: "arrowDown",
      flush: "left",
      "data-test-subj": "case-table-bulk-actions-link-icon"
    }, i18n.BULK_ACTIONS)
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiContextMenu, {
    panels: panels,
    initialPanelId: 0,
    "data-test-subj": "case-table-bulk-actions-context-menu"
  })))), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiButtonEmpty, {
    onClick: onRefresh,
    size: "xs",
    iconSide: "left",
    iconType: "refresh",
    flush: "left",
    "data-test-subj": "all-cases-refresh-link-icon"
  }, i18n.REFRESH)), showClearFiltersButton ? /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_eui.EuiButtonEmpty, {
    onClick: onClearFilters,
    size: "xs",
    iconSide: "left",
    iconType: "cross",
    flush: "left",
    "data-test-subj": "all-cases-clear-filters-link-icon"
  }, i18n.CLEAR_FILTERS)) : null)))), !isSelectorView && /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react2.default.createElement(_columns_popover.ColumnsPopover, {
    selectedColumns: selectedColumns,
    onSelectedColumnsChange: onSelectedColumnsChange
  }))), modals, flyouts, hasReachedMaxCases && !isMessageDismissed && !doNotShowAgain && /*#__PURE__*/_react2.default.createElement(_react2.default.Fragment, null, /*#__PURE__*/_react2.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react2.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react2.default.createElement(_eui.EuiCallOut, {
    title: renderMaxLimitWarning(),
    color: "warning",
    size: "s",
    "data-test-subj": "all-cases-maximum-limit-warning"
  }))), /*#__PURE__*/_react2.default.createElement(_eui.EuiSpacer, {
    size: "m"
  })));
});
CasesTableUtilityBar.displayName = 'CasesTableUtilityBar';
