// @kbn/ml-category-validator/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CATEGORY_EXAMPLES_ERROR_LIMIT", {
  enumerable: true,
  get: function () {
    return _categorization.CATEGORY_EXAMPLES_ERROR_LIMIT;
  }
});
Object.defineProperty(exports, "CATEGORY_EXAMPLES_SAMPLE_SIZE", {
  enumerable: true,
  get: function () {
    return _categorization.CATEGORY_EXAMPLES_SAMPLE_SIZE;
  }
});
Object.defineProperty(exports, "CATEGORY_EXAMPLES_VALIDATION_STATUS", {
  enumerable: true,
  get: function () {
    return _categorization.CATEGORY_EXAMPLES_VALIDATION_STATUS;
  }
});
Object.defineProperty(exports, "CATEGORY_EXAMPLES_WARNING_LIMIT", {
  enumerable: true,
  get: function () {
    return _categorization.CATEGORY_EXAMPLES_WARNING_LIMIT;
  }
});
Object.defineProperty(exports, "MEDIAN_LINE_LENGTH_LIMIT", {
  enumerable: true,
  get: function () {
    return _categorization.MEDIAN_LINE_LENGTH_LIMIT;
  }
});
Object.defineProperty(exports, "NULL_COUNT_PERCENT_LIMIT", {
  enumerable: true,
  get: function () {
    return _categorization.NULL_COUNT_PERCENT_LIMIT;
  }
});
Object.defineProperty(exports, "VALIDATION_CHECK_DESCRIPTION", {
  enumerable: true,
  get: function () {
    return _categorization.VALIDATION_CHECK_DESCRIPTION;
  }
});
Object.defineProperty(exports, "VALIDATION_RESULT", {
  enumerable: true,
  get: function () {
    return _categorization.VALIDATION_RESULT;
  }
});
Object.defineProperty(exports, "VALID_TOKEN_COUNT", {
  enumerable: true,
  get: function () {
    return _categorization.VALID_TOKEN_COUNT;
  }
});
Object.defineProperty(exports, "categorizationExamplesProvider", {
  enumerable: true,
  get: function () {
    return _examples.categorizationExamplesProvider;
  }
});
var _examples = require("./src/examples");
var _categorization = require("./common/constants/categorization");
