// @kbn/security-solution-plugin/public/detection_engine/rule_creation/components/related_integrations/related_integration_field_row.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RelatedIntegrationFieldRow = RelatedIntegrationFieldRow;
var _react = _interopRequireWildcard(require("react"));
var _shared_imports = require("../../../../shared_imports");
var _default_related_integration = require("./default_related_integration");
var _related_integration_field = require("./related_integration_field");
var _validate_related_integration = require("./validate_related_integration");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function RelatedIntegrationFieldRow({
  item,
  relatedIntegrations,
  removeItem
}) {
  const handleRemove = (0, _react.useCallback)(() => removeItem(item.id), [removeItem, item.id]);
  return /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    key: item.id,
    path: item.path,
    config: RELATED_INTEGRATION_FIELD_CONFIG,
    component: _related_integration_field.RelatedIntegrationField,
    readDefaultValueOnForm: !item.isNew,
    componentProps: {
      relatedIntegrations,
      onRemove: handleRemove
    }
  });
}
const RELATED_INTEGRATION_FIELD_CONFIG = {
  type: _shared_imports.FIELD_TYPES.JSON,
  validations: [{
    validator: _validate_related_integration.validateRelatedIntegration
  }],
  defaultValue: _default_related_integration.DEFAULT_RELATED_INTEGRATION
};
