// @kbn/vis-default-editor-plugin/public/components/sidebar/state/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EditorStateActionTypes = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let EditorStateActionTypes = exports.EditorStateActionTypes = /*#__PURE__*/function (EditorStateActionTypes) {
  EditorStateActionTypes["ADD_NEW_AGG"] = "ADD_NEW_AGG";
  EditorStateActionTypes["DISCARD_CHANGES"] = "DISCARD_CHANGES";
  EditorStateActionTypes["CHANGE_AGG_TYPE"] = "CHANGE_AGG_TYPE";
  EditorStateActionTypes["SET_AGG_PARAM_VALUE"] = "SET_AGG_PARAM_VALUE";
  EditorStateActionTypes["SET_STATE_PARAM_VALUE"] = "SET_STATE_PARAM_VALUE";
  EditorStateActionTypes["TOGGLE_ENABLED_AGG"] = "TOGGLE_ENABLED_AGG";
  EditorStateActionTypes["REMOVE_AGG"] = "REMOVE_AGG";
  EditorStateActionTypes["REORDER_AGGS"] = "REORDER_AGGS";
  EditorStateActionTypes["UPDATE_STATE_PARAMS"] = "UPDATE_STATE_PARAMS";
  return EditorStateActionTypes;
}({});
