// @kbn/profiling-plugin/public/components/stack_frame_summary/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StackFrameSummary = StackFrameSummary;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _profilingUtils = require("@kbn/profiling-utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function CalleeFunctionText({
  calleeFunctionName
}) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s",
    style: {
      fontWeight: 'bold',
      overflowWrap: 'anywhere'
    }
  }, calleeFunctionName);
}
function StackFrameSummary({
  frame,
  onFrameClick
}) {
  const calleeFunctionName = (0, _profilingUtils.getCalleeFunction)(frame);
  function handleOnClick() {
    if (onFrameClick) {
      onFrameClick(calleeFunctionName);
    }
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "xs"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement("div", null, onFrameClick ? /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    "data-test-subj": "profilingStackFrameSummaryLink",
    onClick: handleOnClick
  }, /*#__PURE__*/_react.default.createElement(CalleeFunctionText, {
    calleeFunctionName: calleeFunctionName
  })) : /*#__PURE__*/_react.default.createElement(CalleeFunctionText, {
    calleeFunctionName: calleeFunctionName
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    style: {
      overflowWrap: 'anywhere'
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, (0, _profilingUtils.getCalleeSource)(frame) || '‎')));
}
