// @kbn/cases-plugin/public/components/configure_cases/closure_options_radio.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ClosureOptionsRadio = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const radios = [{
  id: 'close-by-user',
  label: i18n.CASE_CLOSURE_OPTIONS_MANUAL
}, {
  id: 'close-by-pushing',
  label: i18n.CASE_CLOSURE_OPTIONS_NEW_INCIDENT
}];
const ClosureOptionsRadioComponent = ({
  closureTypeSelected,
  disabled,
  onChangeClosureType
}) => {
  const onChangeLocal = (0, _react.useCallback)(id => {
    onChangeClosureType(id);
  }, [onChangeClosureType]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiRadioGroup, {
    disabled: disabled,
    options: radios,
    idSelected: closureTypeSelected,
    onChange: onChangeLocal,
    name: "closure_options",
    "data-test-subj": "closure-options-radio-group"
  });
};
ClosureOptionsRadioComponent.displayName = 'ClosureOptionsRadio';
const ClosureOptionsRadio = exports.ClosureOptionsRadio = /*#__PURE__*/_react.default.memo(ClosureOptionsRadioComponent);
