// @kbn/remote-clusters-plugin/public/application/store/actions/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "addCluster", {
  enumerable: true,
  get: function () {
    return _add_cluster.addCluster;
  }
});
Object.defineProperty(exports, "clearAddClusterErrors", {
  enumerable: true,
  get: function () {
    return _add_cluster.clearAddClusterErrors;
  }
});
Object.defineProperty(exports, "clearEditClusterErrors", {
  enumerable: true,
  get: function () {
    return _edit_cluster.clearEditClusterErrors;
  }
});
Object.defineProperty(exports, "closeDetailPanel", {
  enumerable: true,
  get: function () {
    return _detail_panel.closeDetailPanel;
  }
});
Object.defineProperty(exports, "editCluster", {
  enumerable: true,
  get: function () {
    return _edit_cluster.editCluster;
  }
});
Object.defineProperty(exports, "loadClusters", {
  enumerable: true,
  get: function () {
    return _load_clusters.loadClusters;
  }
});
Object.defineProperty(exports, "openDetailPanel", {
  enumerable: true,
  get: function () {
    return _detail_panel.openDetailPanel;
  }
});
Object.defineProperty(exports, "refreshClusters", {
  enumerable: true,
  get: function () {
    return _refresh_clusters.refreshClusters;
  }
});
Object.defineProperty(exports, "removeClusters", {
  enumerable: true,
  get: function () {
    return _remove_clusters.removeClusters;
  }
});
Object.defineProperty(exports, "startEditingCluster", {
  enumerable: true,
  get: function () {
    return _edit_cluster.startEditingCluster;
  }
});
Object.defineProperty(exports, "stopEditingCluster", {
  enumerable: true,
  get: function () {
    return _edit_cluster.stopEditingCluster;
  }
});
var _add_cluster = require("./add_cluster");
var _detail_panel = require("./detail_panel");
var _remove_clusters = require("./remove_clusters");
var _edit_cluster = require("./edit_cluster");
var _load_clusters = require("./load_clusters");
var _refresh_clusters = require("./refresh_clusters");
