// @kbn/security-solution-plugin/public/management/components/administration_list_page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdministrationListPage = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _use_test_id_generator = require("../hooks/use_test_id_generator");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AdministrationListPage = exports.AdministrationListPage = /*#__PURE__*/(0, _react.memo)(({
  title,
  subtitle,
  actions,
  children,
  restrictWidth = false,
  hasBottomBorder = true,
  hideHeader = false,
  headerBackComponent,
  ...otherProps
}) => {
  const header = (0, _react.useMemo)(() => {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      direction: "column",
      gutterSize: "none",
      alignItems: "flexStart"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, headerBackComponent && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, headerBackComponent)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "l"
    }, /*#__PURE__*/_react.default.createElement("span", {
      "data-test-subj": "header-page-title"
    }, title))));
  }, [headerBackComponent, title]);
  const description = (0, _react.useMemo)(() => {
    return subtitle ? /*#__PURE__*/_react.default.createElement("span", {
      "data-test-subj": "header-panel-subtitle"
    }, subtitle) : undefined;
  }, [subtitle]);
  const getTestId = (0, _use_test_id_generator.useTestIdGenerator)(otherProps['data-test-subj']);
  return /*#__PURE__*/_react.default.createElement("div", otherProps, !hideHeader && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPageHeader, {
    pageTitle: header,
    description: description,
    bottomBorder: hasBottomBorder,
    rightSideItems: actions ? [actions] : undefined,
    restrictWidth: restrictWidth,
    "data-test-subj": getTestId('header')
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "l"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiPageSection, {
    paddingSize: "none",
    color: "transparent",
    restrictWidth: restrictWidth
  }, children));
});
AdministrationListPage.displayName = 'AdministrationListPage';
