// @kbn/rollup-plugin/public/crud_app/sections/components/job_details/tabs/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "TabHistogram", {
  enumerable: true,
  get: function () {
    return _tab_histogram.TabHistogram;
  }
});
Object.defineProperty(exports, "TabJson", {
  enumerable: true,
  get: function () {
    return _tab_json.TabJson;
  }
});
Object.defineProperty(exports, "TabMetrics", {
  enumerable: true,
  get: function () {
    return _tab_metrics.TabMetrics;
  }
});
Object.defineProperty(exports, "TabRequest", {
  enumerable: true,
  get: function () {
    return _tab_request.TabRequest;
  }
});
Object.defineProperty(exports, "TabSummary", {
  enumerable: true,
  get: function () {
    return _tab_summary.TabSummary;
  }
});
Object.defineProperty(exports, "TabTerms", {
  enumerable: true,
  get: function () {
    return _tab_terms.TabTerms;
  }
});
var _tab_summary = require("./tab_summary");
var _tab_terms = require("./tab_terms");
var _tab_histogram = require("./tab_histogram");
var _tab_metrics = require("./tab_metrics");
var _tab_json = require("./tab_json");
var _tab_request = require("./tab_request");
