// @kbn/security-solution-plugin/common/api/detection_engine/signals_migration/delete_signals_migration/delete_signals_migration.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MigrationCleanupResult = exports.AlertsMigrationCleanupResponse = exports.AlertsMigrationCleanupRequestBody = void 0;
var _zod = require("zod");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Alerts migration cleanup API endpoint
 *   version: 2023-10-31
 */

const MigrationCleanupResult = exports.MigrationCleanupResult = _zod.z.object({
  id: _zod.z.string(),
  destinationIndex: _zod.z.string(),
  status: _zod.z.enum(['success', 'failure', 'pending']),
  sourceIndex: _zod.z.string(),
  version: _zod.z.string(),
  updated: _zod.z.string().datetime(),
  error: _zod.z.object({
    message: _zod.z.string(),
    status_code: _zod.z.number().int()
  }).optional()
});
const AlertsMigrationCleanupRequestBody = exports.AlertsMigrationCleanupRequestBody = _zod.z.object({
  migration_ids: _zod.z.array(_zod.z.string()).min(1)
});
const AlertsMigrationCleanupResponse = exports.AlertsMigrationCleanupResponse = _zod.z.array(MigrationCleanupResult);
