// @kbn/fleet-plugin/server/services/epm/packages/install_errors_helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.addErrorToLatestFailedAttempts = addErrorToLatestFailedAttempts;
exports.clearLatestFailedAttempts = clearLatestFailedAttempts;
var _semver = require("semver");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const MAX_ATTEMPTS_TO_KEEP = 5;
function clearLatestFailedAttempts(installedVersion, latestAttempts = []) {
  return latestAttempts.filter(attempt => (0, _semver.lt)(installedVersion, attempt.target_version));
}
function addErrorToLatestFailedAttempts({
  error,
  createdAt,
  targetVersion,
  latestAttempts = []
}) {
  return [{
    created_at: createdAt,
    target_version: targetVersion,
    error: {
      name: error.name,
      message: error.message,
      stack: error.stack
    }
  }, ...latestAttempts].slice(0, MAX_ATTEMPTS_TO_KEEP);
}
