// @kbn/slo-plugin/public/pages/slo_details/components/history/slo_details_history.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SLODetailsHistory = SLODetailsHistory;
var _eui = require("@elastic/eui");
var _datemath = _interopRequireDefault(require("@kbn/datemath"));
var _react = _interopRequireWildcard(require("react"));
var _burn_rates = require("../../../../components/slo/burn_rate/burn_rates");
var _kibana_react = require("../../../../utils/kibana_react");
var _use_burn_rate_options = require("../../hooks/use_burn_rate_options");
var _events_chart_panel = require("../events_chart_panel");
var _historical_data_charts = require("../historical_data_charts");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function SLODetailsHistory({
  slo,
  isAutoRefreshing,
  selectedTabId
}) {
  const {
    uiSettings
  } = (0, _kibana_react.useKibana)().services;
  const {
    burnRateOptions
  } = (0, _use_burn_rate_options.useBurnRateOptions)(slo);
  const [start, setStart] = (0, _react.useState)(`now-${slo.timeWindow.duration}`);
  const [end, setEnd] = (0, _react.useState)('now');
  const onTimeChange = val => {
    setStart(val.start);
    setEnd(val.end);
  };
  const onRefresh = val => {};
  const range = (0, _react.useMemo)(() => {
    return {
      from: new Date(_datemath.default.parse(start).valueOf()),
      to: new Date(_datemath.default.parse(end, {
        roundUp: true
      }).valueOf())
    };
  }, [start, end]);
  const onBrushed = ({
    from,
    to
  }) => {
    setStart(from.toISOString());
    setEnd(to.toISOString());
  };
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "flexEnd"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true,
    css: {
      maxWidth: 500
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSuperDatePicker, {
    isLoading: false,
    start: start,
    end: end,
    onTimeChange: onTimeChange,
    onRefresh: onRefresh,
    width: "full",
    commonlyUsedRanges: uiSettings.get('timepicker:quickRanges').map(({
      from,
      to,
      display
    }) => {
      return {
        start: from,
        end: to,
        label: display
      };
    })
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "l"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "l"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_burn_rates.BurnRates, {
    slo: slo,
    isAutoRefreshing: isAutoRefreshing,
    burnRateOptions: burnRateOptions,
    selectedTabId: selectedTabId,
    range: range,
    onBrushed: onBrushed
  })), /*#__PURE__*/_react.default.createElement(_historical_data_charts.HistoricalDataCharts, {
    slo: slo,
    selectedTabId: selectedTabId,
    isAutoRefreshing: isAutoRefreshing,
    range: range,
    onBrushed: onBrushed
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_events_chart_panel.EventsChartPanel, {
    slo: slo,
    range: range,
    selectedTabId: selectedTabId,
    onBrushed: onBrushed
  }))));
}
