// @kbn/discover-plugin/public/components/doc_table/doc_table_context.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocTableContext = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
require("./index.scss");
var _skip_bottom_button = require("../../application/main/components/skip_bottom_button");
var _doc_table_wrapper = require("./doc_table_wrapper");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const DocTableWrapperMemoized = /*#__PURE__*/_react.default.memo(_doc_table_wrapper.DocTableWrapper);
const renderDocTable = tableProps => {
  return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_skip_bottom_button.SkipBottomButton, {
    onClick: tableProps.onSkipBottomButtonClick
  }), /*#__PURE__*/_react.default.createElement("table", {
    className: "kbn-table table",
    "data-test-subj": "docTable"
  }, /*#__PURE__*/_react.default.createElement("thead", null, tableProps.renderHeader()), /*#__PURE__*/_react.default.createElement("tbody", null, tableProps.renderRows(tableProps.rows))), /*#__PURE__*/_react.default.createElement("span", {
    tabIndex: -1,
    id: "discoverBottomMarker"
  }, "\u200B"));
};
const DocTableContext = props => {
  return /*#__PURE__*/_react.default.createElement(DocTableWrapperMemoized, (0, _extends2.default)({}, props, {
    render: renderDocTable
  }));
};
exports.DocTableContext = DocTableContext;
