// @kbn/snapshot-restore-plugin/public/application/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.renderApp = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactDom = require("react-dom");
var _sharedUxRouter = require("@kbn/shared-ux-router");
var _public = require("@kbn/kibana-react-plugin/public");
var _app = require("./app");
var _app_providers = require("./app_providers");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AppWithRouter = ({
  history
}) => /*#__PURE__*/_react.default.createElement(_sharedUxRouter.Router, {
  history: history
}, /*#__PURE__*/_react.default.createElement(_app.App, null));
const renderApp = (elem, dependencies) => {
  (0, _reactDom.render)( /*#__PURE__*/_react.default.createElement(_public.KibanaContextProvider, {
    services: {
      uiSettings: dependencies.services.uiSettings,
      settings: dependencies.services.settings,
      theme: dependencies.core.theme
    }
  }, /*#__PURE__*/_react.default.createElement(_app_providers.AppProviders, {
    appDependencies: dependencies
  }, /*#__PURE__*/_react.default.createElement(AppWithRouter, {
    history: dependencies.services.history
  }))), elem);
  return () => {
    (0, _reactDom.unmountComponentAtNode)(elem);
  };
};
exports.renderApp = renderApp;
