// @kbn/security-solution-plugin/public/common/components/markdown_editor/plugins/osquery/not_available_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PERMISSION_DENIED = exports.OsqueryNotAvailablePrompt = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PERMISSION_DENIED = exports.PERMISSION_DENIED = _i18n.i18n.translate('xpack.securitySolution.markdown.osquery.permissionDenied', {
  defaultMessage: 'Permission denied'
});
const OsqueryNotAvailablePrompt = () => /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  iconType: "logoOsquery",
  title: /*#__PURE__*/_react.default.createElement("h2", null, PERMISSION_DENIED),
  titleSize: "xs",
  body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.securitySolution.markdown.osquery.missingPrivileges",
    defaultMessage: "To access this page, ask your administrator for {osquery} Kibana privileges.",
    values: {
      osquery: /*#__PURE__*/_react.default.createElement(_eui.EuiCode, null, 'osquery')
    }
  }))
});
exports.OsqueryNotAvailablePrompt = OsqueryNotAvailablePrompt;
