// @kbn/text-based-languages/public/triggers/update_esql_query_trigger.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateESQLQueryTrigger = exports.UPDATE_ESQL_QUERY_TRIGGER = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const UPDATE_ESQL_QUERY_TRIGGER = exports.UPDATE_ESQL_QUERY_TRIGGER = 'UPDATE_ESQL_QUERY_TRIGGER';
const updateESQLQueryTrigger = exports.updateESQLQueryTrigger = {
  id: UPDATE_ESQL_QUERY_TRIGGER,
  title: _i18n.i18n.translate('textBasedLanguages.triggers.updateEsqlQueryTrigger', {
    defaultMessage: 'Update ES|QL query'
  }),
  description: _i18n.i18n.translate('textBasedLanguages.triggers.updateEsqlQueryTriggerDescription', {
    defaultMessage: 'Update ES|QL query with a new one'
  })
};
