// @kbn/lens-plugin/public/shared_components/axis/extent/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AxisBoundsControl", {
  enumerable: true,
  get: function () {
    return _axis_extent_settings.AxisBoundsControl;
  }
});
Object.defineProperty(exports, "axisExtentConfigToExpression", {
  enumerable: true,
  get: function () {
    return _to_expression.axisExtentConfigToExpression;
  }
});
Object.defineProperty(exports, "getDataBounds", {
  enumerable: true,
  get: function () {
    return _helpers.getDataBounds;
  }
});
Object.defineProperty(exports, "hasNumericHistogramDimension", {
  enumerable: true,
  get: function () {
    return _helpers.hasNumericHistogramDimension;
  }
});
Object.defineProperty(exports, "validateAxisDomain", {
  enumerable: true,
  get: function () {
    return _helpers.validateAxisDomain;
  }
});
Object.defineProperty(exports, "validateZeroInclusivityExtent", {
  enumerable: true,
  get: function () {
    return _helpers.validateZeroInclusivityExtent;
  }
});
var _axis_extent_settings = require("./axis_extent_settings");
var _helpers = require("./helpers");
var _to_expression = require("./to_expression");
