// @kbn/cases-plugin/common/types/api/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  configureApiV1: true,
  userActionApiV1: true,
  alertApiV1: true,
  statsApiV1: true,
  caseApiV1: true,
  externalServiceApiV1: true,
  userApiV1: true,
  connectorApiV1: true,
  attachmentApiV1: true,
  metricsApiV1: true,
  customFieldsApiV1: true
};
exports.userApiV1 = exports.userActionApiV1 = exports.statsApiV1 = exports.metricsApiV1 = exports.externalServiceApiV1 = exports.customFieldsApiV1 = exports.connectorApiV1 = exports.configureApiV1 = exports.caseApiV1 = exports.attachmentApiV1 = exports.alertApiV1 = void 0;
var _latest = require("./configure/latest");
Object.keys(_latest).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest[key];
    }
  });
});
var _latest2 = require("./user_action/latest");
Object.keys(_latest2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest2[key];
    }
  });
});
var _latest3 = require("./alert/latest");
Object.keys(_latest3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest3[key];
    }
  });
});
var _latest4 = require("./case/latest");
Object.keys(_latest4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest4[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest4[key];
    }
  });
});
var _latest5 = require("./external_service/latest");
Object.keys(_latest5).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest5[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest5[key];
    }
  });
});
var _latest6 = require("./stats/latest");
Object.keys(_latest6).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest6[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest6[key];
    }
  });
});
var _latest7 = require("./user/latest");
Object.keys(_latest7).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest7[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest7[key];
    }
  });
});
var _latest8 = require("./connector/latest");
Object.keys(_latest8).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest8[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest8[key];
    }
  });
});
var _latest9 = require("./attachment/latest");
Object.keys(_latest9).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest9[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest9[key];
    }
  });
});
var _latest10 = require("./metrics/latest");
Object.keys(_latest10).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest10[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest10[key];
    }
  });
});
var _latest11 = require("./custom_field/latest");
Object.keys(_latest11).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _latest11[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _latest11[key];
    }
  });
});
var _configureApiV = _interopRequireWildcard(require("./configure/v1"));
exports.configureApiV1 = _configureApiV;
var _userActionApiV = _interopRequireWildcard(require("./user_action/v1"));
exports.userActionApiV1 = _userActionApiV;
var _alertApiV = _interopRequireWildcard(require("./alert/v1"));
exports.alertApiV1 = _alertApiV;
var _statsApiV = _interopRequireWildcard(require("./stats/v1"));
exports.statsApiV1 = _statsApiV;
var _caseApiV = _interopRequireWildcard(require("./case/v1"));
exports.caseApiV1 = _caseApiV;
var _externalServiceApiV = _interopRequireWildcard(require("./external_service/v1"));
exports.externalServiceApiV1 = _externalServiceApiV;
var _userApiV = _interopRequireWildcard(require("./user/v1"));
exports.userApiV1 = _userApiV;
var _connectorApiV = _interopRequireWildcard(require("./connector/v1"));
exports.connectorApiV1 = _connectorApiV;
var _attachmentApiV = _interopRequireWildcard(require("./attachment/v1"));
exports.attachmentApiV1 = _attachmentApiV;
var _metricsApiV = _interopRequireWildcard(require("./metrics/v1"));
exports.metricsApiV1 = _metricsApiV;
var _customFieldsApiV = _interopRequireWildcard(require("./custom_field/v1"));
exports.customFieldsApiV1 = _customFieldsApiV;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
