// @kbn/monitoring-plugin/common/enums.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SetupModeFeature = exports.AlertSeverity = exports.AlertParamType = exports.AlertMessageTokenType = exports.AlertClusterHealthType = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let AlertClusterHealthType = exports.AlertClusterHealthType = /*#__PURE__*/function (AlertClusterHealthType) {
  AlertClusterHealthType["Green"] = "green";
  AlertClusterHealthType["Red"] = "red";
  AlertClusterHealthType["Yellow"] = "yellow";
  return AlertClusterHealthType;
}({});
let AlertSeverity = exports.AlertSeverity = /*#__PURE__*/function (AlertSeverity) {
  AlertSeverity["Success"] = "success";
  AlertSeverity["Danger"] = "danger";
  AlertSeverity["Warning"] = "warning";
  return AlertSeverity;
}({});
let AlertMessageTokenType = exports.AlertMessageTokenType = /*#__PURE__*/function (AlertMessageTokenType) {
  AlertMessageTokenType["Time"] = "time";
  AlertMessageTokenType["Link"] = "link";
  AlertMessageTokenType["DocLink"] = "docLink";
  return AlertMessageTokenType;
}({});
let AlertParamType = exports.AlertParamType = /*#__PURE__*/function (AlertParamType) {
  AlertParamType["TextField"] = "textfield";
  AlertParamType["Duration"] = "duration";
  AlertParamType["Percentage"] = "percentage";
  AlertParamType["Number"] = "number";
  return AlertParamType;
}({});
let SetupModeFeature = exports.SetupModeFeature = /*#__PURE__*/function (SetupModeFeature) {
  SetupModeFeature["MetricbeatMigration"] = "metricbeatMigration";
  SetupModeFeature["Alerts"] = "alerts";
  return SetupModeFeature;
}({});
