// @kbn/apm-plugin/common/agent_name.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AGENT_NAMES", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.AGENT_NAMES;
  }
});
Object.defineProperty(exports, "JAVA_AGENT_NAMES", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.JAVA_AGENT_NAMES;
  }
});
Object.defineProperty(exports, "OPEN_TELEMETRY_AGENT_NAMES", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.OPEN_TELEMETRY_AGENT_NAMES;
  }
});
Object.defineProperty(exports, "RUM_AGENT_NAMES", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.RUM_AGENT_NAMES;
  }
});
Object.defineProperty(exports, "isAWSLambdaAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isAWSLambdaAgentName;
  }
});
Object.defineProperty(exports, "isAndroidAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isAndroidAgentName;
  }
});
Object.defineProperty(exports, "isAzureFunctionsAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isAzureFunctionsAgentName;
  }
});
Object.defineProperty(exports, "isIosAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isIosAgentName;
  }
});
Object.defineProperty(exports, "isJRubyAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isJRubyAgentName;
  }
});
Object.defineProperty(exports, "isJavaAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isJavaAgentName;
  }
});
Object.defineProperty(exports, "isMobileAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isMobileAgentName;
  }
});
Object.defineProperty(exports, "isOpenTelemetryAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isOpenTelemetryAgentName;
  }
});
Object.defineProperty(exports, "isRumAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isRumAgentName;
  }
});
Object.defineProperty(exports, "isRumOrMobileAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isRumOrMobileAgentName;
  }
});
Object.defineProperty(exports, "isServerlessAgentName", {
  enumerable: true,
  get: function () {
    return _elasticAgentUtils.isServerlessAgentName;
  }
});
var _elasticAgentUtils = require("@kbn/elastic-agent-utils");
