// @kbn/apm-plugin/common/latency_distribution_chart_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.latencyDistributionChartTypeRt = exports.LatencyDistributionChartType = void 0;
var t = _interopRequireWildcard(require("io-ts"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let LatencyDistributionChartType = exports.LatencyDistributionChartType = /*#__PURE__*/function (LatencyDistributionChartType) {
  LatencyDistributionChartType["transactionLatency"] = "transactionLatency";
  LatencyDistributionChartType["latencyCorrelations"] = "latencyCorrelations";
  LatencyDistributionChartType["failedTransactionsCorrelations"] = "failedTransactionsCorrelations";
  LatencyDistributionChartType["dependencyLatency"] = "dependencyLatency";
  return LatencyDistributionChartType;
}({});
const latencyDistributionChartTypeRt = exports.latencyDistributionChartTypeRt = t.union([t.literal(LatencyDistributionChartType.transactionLatency), t.literal(LatencyDistributionChartType.latencyCorrelations), t.literal(LatencyDistributionChartType.failedTransactionsCorrelations), t.literal(LatencyDistributionChartType.dependencyLatency)]);
