// @kbn/apm-plugin/common/latency_aggregation_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.latencyAggregationTypeRt = exports.getLatencyAggregationType = exports.LatencyAggregationType = void 0;
var t = _interopRequireWildcard(require("io-ts"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let LatencyAggregationType = exports.LatencyAggregationType = /*#__PURE__*/function (LatencyAggregationType) {
  LatencyAggregationType["avg"] = "avg";
  LatencyAggregationType["p99"] = "p99";
  LatencyAggregationType["p95"] = "p95";
  return LatencyAggregationType;
}({});
const latencyAggregationTypeRt = exports.latencyAggregationTypeRt = t.union([t.literal(LatencyAggregationType.avg), t.literal(LatencyAggregationType.p95), t.literal(LatencyAggregationType.p99)]);
const getLatencyAggregationType = latencyAggregationType => {
  return latencyAggregationType !== null && latencyAggregationType !== void 0 ? latencyAggregationType : LatencyAggregationType.avg;
};
exports.getLatencyAggregationType = getLatencyAggregationType;
