// @kbn/canvas-plugin/public/components/layout_annotations/dragbox_annotation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DragBoxAnnotation = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _dom = require("../../lib/dom");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DragBoxAnnotation = ({
  transformMatrix,
  width,
  height
}) => /*#__PURE__*/_react.default.createElement("div", {
  className: "canvasDragBoxAnnotation canvasLayoutAnnotation",
  style: {
    height,
    marginLeft: -width / 2,
    marginTop: -height / 2,
    transform: (0, _dom.matrixToCSS)(transformMatrix),
    width
  }
});
exports.DragBoxAnnotation = DragBoxAnnotation;
DragBoxAnnotation.propTypes = {
  transformMatrix: _propTypes.default.arrayOf(_propTypes.default.number).isRequired,
  width: _propTypes.default.number.isRequired,
  height: _propTypes.default.number.isRequired
};
