// @kbn/canvas-plugin/public/components/layout_annotations/tooltip_annotation.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TooltipAnnotation = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _dom = require("../../lib/dom");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TooltipAnnotation = ({
  transformMatrix,
  text
}) => {
  const newStyle = {
    transform: `${(0, _dom.matrixToCSS)(transformMatrix)} translate(1em, -1em)`
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "tooltipAnnotation canvasLayoutAnnotation",
    style: newStyle
  }, /*#__PURE__*/_react.default.createElement("p", null, text, "\xB0"));
};
exports.TooltipAnnotation = TooltipAnnotation;
TooltipAnnotation.propTypes = {
  transformMatrix: _propTypes.default.arrayOf(_propTypes.default.number).isRequired,
  text: _propTypes.default.string.isRequired
};
