// @kbn/security-solution-plugin/public/management/components/endpoint_response_actions_list/components/actions_log_empty_state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionsLogEmptyState = void 0;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _eui = require("@elastic/eui");
var _public = require("@kbn/kibana-react-plugin/public");
var _i18n = require("@kbn/i18n");
var _management_empty_state_wrapper = require("../../management_empty_state_wrapper");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EmptyPrompt = (0, _styledComponents.default)(_eui.EuiEmptyPrompt)`
  max-width: 100%;
`;
const ActionsLogEmptyState = exports.ActionsLogEmptyState = /*#__PURE__*/(0, _react.memo)(({
  'data-test-subj': dataTestSubj
}) => {
  const {
    docLinks
  } = (0, _public.useKibana)().services;
  return /*#__PURE__*/_react.default.createElement(_management_empty_state_wrapper.ManagementEmptyStateWrapper, {
    "data-test-subj": dataTestSubj
  }, /*#__PURE__*/_react.default.createElement(EmptyPrompt, {
    iconType: "editorUnorderedList",
    title: /*#__PURE__*/_react.default.createElement("h2", null, _i18n.i18n.translate('xpack.securitySolution.responseActionsHistory.empty.title', {
      defaultMessage: 'Response actions history is empty'
    })),
    body: /*#__PURE__*/_react.default.createElement("div", null, _i18n.i18n.translate('xpack.securitySolution.responseActionsHistory.empty.content', {
      defaultMessage: 'No response actions performed'
    })),
    actions: [/*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
      external: true,
      href: docLinks === null || docLinks === void 0 ? void 0 : docLinks.links.securitySolution.responseActions,
      target: "_blank"
    }, _i18n.i18n.translate('xpack.securitySolution.responseActionsHistory.empty.link', {
      defaultMessage: 'Read more about response actions'
    }))]
  }));
});
ActionsLogEmptyState.displayName = 'ActionsLogEmptyState';
