// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/footer/footer.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getFooter = void 0;
var _documentation = _interopRequireDefault(require("../images/documentation.png"));
var _forum = _interopRequireDefault(require("../images/forum.png"));
var _demo = _interopRequireDefault(require("../images/demo.png"));
var _labs = _interopRequireDefault(require("../images/labs.png"));
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const footer = [{
  icon: _documentation.default,
  key: 'documentation',
  title: i18n.FOOTER_DOCUMENTATION_TITLE,
  description: i18n.FOOTER_DOCUMENTATION_DESCRIPTION,
  link: {
    title: i18n.FOOTER_DOCUMENTATION_LINK_TITLE,
    href: 'https://docs.elastic.co/integrations/elastic-security-intro'
  }
}, {
  icon: _forum.default,
  key: 'forum',
  title: i18n.FOOTER_FORUM_TITLE,
  description: i18n.FOOTER_FORUM_DESCRIPTION,
  link: {
    title: i18n.FOOTER_FORUM_LINK_TITLE,
    href: 'https://discuss.elastic.co/c/security/83'
  }
}, {
  icon: _demo.default,
  key: 'demo',
  title: i18n.FOOTER_DEMO_TITLE,
  description: i18n.FOOTER_DEMO_DESCRIPTION,
  link: {
    title: i18n.FOOTER_DEMO_LINK_TITLE,
    href: 'https://www.elastic.co/demo-gallery?solutions=security&features=null'
  }
}, {
  icon: _labs.default,
  key: 'labs',
  title: i18n.FOOTER_LABS_TITLE,
  description: i18n.FOOTER_LABS_DESCRIPTION,
  link: {
    title: i18n.FOOTER_LABS_LINK_TITLE,
    href: 'https://www.elastic.co/security-labs'
  }
}];
const getFooter = () => footer;
exports.getFooter = getFooter;
