// @kbn/fleet-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AvailablePackagesHook = void 0;
Object.defineProperty(exports, "CustomAssetsAccordion", {
  enumerable: true,
  get: function () {
    return _custom_assets_accordion.CustomAssetsAccordion;
  }
});
exports.LazyPackagePolicyInputVarField = void 0;
Object.defineProperty(exports, "PackageIcon", {
  enumerable: true,
  get: function () {
    return _package_icon.PackageIcon;
  }
});
exports.PackageList = void 0;
Object.defineProperty(exports, "PackagePolicyEditorDatastreamMappings", {
  enumerable: true,
  get: function () {
    return _datastream_mappings.PackagePolicyEditorDatastreamMappings;
  }
});
Object.defineProperty(exports, "PackagePolicyEditorDatastreamPipelines", {
  enumerable: true,
  get: function () {
    return _datastream_pipelines.PackagePolicyEditorDatastreamPipelines;
  }
});
Object.defineProperty(exports, "SetupTechnology", {
  enumerable: true,
  get: function () {
    return _types.SetupTechnology;
  }
});
Object.defineProperty(exports, "pagePathGetters", {
  enumerable: true,
  get: function () {
    return _constants.pagePathGetters;
  }
});
Object.defineProperty(exports, "pkgKeyFromPackageInfo", {
  enumerable: true,
  get: function () {
    return _services.pkgKeyFromPackageInfo;
  }
});
exports.plugin = void 0;
var _react = require("react");
var _plugin = require("./plugin");
var _types = require("./types");
var _constants = require("./constants");
var _services = require("./services");
var _custom_assets_accordion = require("./components/custom_assets_accordion");
var _package_icon = require("./components/package_icon");
var _datastream_pipelines = require("./applications/fleet/sections/agent_policy/create_package_policy_page/components/datastream_pipelines");
var _datastream_mappings = require("./applications/fleet/sections/agent_policy/create_package_policy_page/components/datastream_mappings");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const plugin = initializerContext => {
  return new _plugin.FleetPlugin(initializerContext);
};

// Export Package editor components for custom editors
exports.plugin = plugin;
// This Type export is added to prevent error TS4023

const LazyPackagePolicyInputVarField = exports.LazyPackagePolicyInputVarField = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./applications/fleet/sections/agent_policy/create_package_policy_page/components/steps/components/package_policy_input_var_field'))).then(module => ({
  default: module.PackagePolicyInputVarField
})));
const PackageList = () => {
  return Promise.resolve().then(() => _interopRequireWildcard(require('./applications/integrations/sections/epm/components/package_list_grid')));
};
exports.PackageList = PackageList;
const AvailablePackagesHook = () => {
  return Promise.resolve().then(() => _interopRequireWildcard(require('./applications/integrations/sections/epm/screens/home/hooks/use_available_packages')));
};
exports.AvailablePackagesHook = AvailablePackagesHook;
