// @kbn/index-management-plugin/public/application/sections/home/index_list/details_page/details_page_overview/overview_card.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OverviewCard = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _uiTheme = require("@kbn/ui-theme");
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const OverviewCard = ({
  title,
  content: {
    left: contentLeft,
    right: contentRight
  },
  footer: {
    left: footerLeft,
    right: footerRight
  } = {},
  'data-test-subj': dataTestSubj
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSplitPanel.Outer, {
    grow: true,
    hasBorder: true,
    "data-test-subj": dataTestSubj
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSplitPanel.Inner, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xxxs"
  }, /*#__PURE__*/_react.default.createElement("h4", null, title)), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none",
    justifyContent: "spaceBetween",
    wrap: true,
    alignItems: "center",
    css: (0, _react2.css)`
              min-height: ${_uiTheme.euiThemeVars.euiButtonHeightSmall};
            `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, contentLeft), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, contentRight))), /*#__PURE__*/_react.default.createElement(_eui.EuiSplitPanel.Inner, {
    grow: false,
    color: "subdued"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none",
    justifyContent: "spaceBetween",
    alignItems: "center",
    wrap: true
  }, footerLeft && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, footerLeft), footerRight && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, footerRight)))));
};
exports.OverviewCard = OverviewCard;
