// @kbn/index-management-plugin/public/application/sections/home/index_list/details_page/details_page_overview/storage_details.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StorageDetails = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _i18n = require("@kbn/i18n");
var _uiTheme = require("@kbn/ui-theme");
var _eui = require("@elastic/eui");
var _app_context = require("../../../../../app_context");
var _overview_card = require("./overview_card");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StorageDetails = ({
  primarySize,
  size,
  primary,
  replica
}) => {
  const {
    config
  } = (0, _app_context.useAppContext)();
  if (!config.enableIndexStats) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_overview_card.OverviewCard, {
    "data-test-subj": "indexDetailsStorage",
    title: _i18n.i18n.translate('xpack.idxMgmt.indexDetails.overviewTab.storage.cardTitle', {
      defaultMessage: 'Storage'
    }),
    content: {
      left: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
        gutterSize: "xs",
        alignItems: "baseline"
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
        grow: false
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
        css: (0, _react2.css)`
                  font-size: ${_uiTheme.euiThemeVars.euiFontSizeL};
                `
      }, primarySize)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
        color: "subdued"
      }, _i18n.i18n.translate('xpack.idxMgmt.indexDetails.overviewTab.storage.primarySizeLabel', {
        defaultMessage: 'Primary'
      })))),
      right: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
        gutterSize: "xs",
        alignItems: "baseline"
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
        grow: false
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
        css: (0, _react2.css)`
                  font-size: ${_uiTheme.euiThemeVars.euiFontSizeL};
                `
      }, size)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
        color: "subdued"
      }, _i18n.i18n.translate('xpack.idxMgmt.indexDetails.overviewTab.storage.totalSizeLabel', {
        defaultMessage: 'Total'
      }))))
    },
    footer: {
      left: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
        gutterSize: "xs"
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
        grow: false
      }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
        type: "shard",
        color: "subdued"
      })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
        color: "subdued"
      }, _i18n.i18n.translate('xpack.idxMgmt.indexDetails.overviewTab.storage.shardsLabel', {
        defaultMessage: 'Shards'
      })))),
      right: /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
        color: "subdued"
      }, _i18n.i18n.translate('xpack.idxMgmt.indexDetails.overviewTab.storage.primariesReplicasLabel', {
        defaultMessage: '{primary, plural, one {# Primary} other {# Primaries}} / {replica, plural, one {# Replica} other {# Replicas}} ',
        values: {
          primary,
          replica
        }
      }))
    }
  });
};
exports.StorageDetails = StorageDetails;
