// @kbn/index-management-plugin/public/application/components/mappings_editor/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  getUniqueId: true,
  getChildFieldsName: true,
  getFieldMeta: true,
  getTypeLabelFromField: true,
  getFieldConfig: true,
  getTypeMetaFromSource: true,
  normalize: true,
  deNormalize: true,
  updateFieldsPathAfterFieldNameChange: true,
  getAllChildFields: true,
  getAllDescendantAliases: true,
  getFieldAncestors: true,
  filterTypesForMultiField: true,
  filterTypesForNonRootFields: true,
  getMaxNestedDepth: true,
  buildFieldTreeFromIds: true,
  shouldDeleteChildFieldsAfterTypeChange: true,
  canUseMappingsEditor: true,
  stripUndefinedValues: true,
  normalizeRuntimeFields: true,
  deNormalizeRuntimeFields: true,
  getAllFieldTypesFromState: true,
  getFieldsFromState: true,
  getFieldsMatchingFilterFromState: true
};
Object.defineProperty(exports, "buildFieldTreeFromIds", {
  enumerable: true,
  get: function () {
    return _utils.buildFieldTreeFromIds;
  }
});
Object.defineProperty(exports, "canUseMappingsEditor", {
  enumerable: true,
  get: function () {
    return _utils.canUseMappingsEditor;
  }
});
Object.defineProperty(exports, "deNormalize", {
  enumerable: true,
  get: function () {
    return _utils.deNormalize;
  }
});
Object.defineProperty(exports, "deNormalizeRuntimeFields", {
  enumerable: true,
  get: function () {
    return _utils.deNormalizeRuntimeFields;
  }
});
Object.defineProperty(exports, "filterTypesForMultiField", {
  enumerable: true,
  get: function () {
    return _utils.filterTypesForMultiField;
  }
});
Object.defineProperty(exports, "filterTypesForNonRootFields", {
  enumerable: true,
  get: function () {
    return _utils.filterTypesForNonRootFields;
  }
});
Object.defineProperty(exports, "getAllChildFields", {
  enumerable: true,
  get: function () {
    return _utils.getAllChildFields;
  }
});
Object.defineProperty(exports, "getAllDescendantAliases", {
  enumerable: true,
  get: function () {
    return _utils.getAllDescendantAliases;
  }
});
Object.defineProperty(exports, "getAllFieldTypesFromState", {
  enumerable: true,
  get: function () {
    return _utils.getAllFieldTypesFromState;
  }
});
Object.defineProperty(exports, "getChildFieldsName", {
  enumerable: true,
  get: function () {
    return _utils.getChildFieldsName;
  }
});
Object.defineProperty(exports, "getFieldAncestors", {
  enumerable: true,
  get: function () {
    return _utils.getFieldAncestors;
  }
});
Object.defineProperty(exports, "getFieldConfig", {
  enumerable: true,
  get: function () {
    return _utils.getFieldConfig;
  }
});
Object.defineProperty(exports, "getFieldMeta", {
  enumerable: true,
  get: function () {
    return _utils.getFieldMeta;
  }
});
Object.defineProperty(exports, "getFieldsFromState", {
  enumerable: true,
  get: function () {
    return _utils.getFieldsFromState;
  }
});
Object.defineProperty(exports, "getFieldsMatchingFilterFromState", {
  enumerable: true,
  get: function () {
    return _utils.getFieldsMatchingFilterFromState;
  }
});
Object.defineProperty(exports, "getMaxNestedDepth", {
  enumerable: true,
  get: function () {
    return _utils.getMaxNestedDepth;
  }
});
Object.defineProperty(exports, "getTypeLabelFromField", {
  enumerable: true,
  get: function () {
    return _utils.getTypeLabelFromField;
  }
});
Object.defineProperty(exports, "getTypeMetaFromSource", {
  enumerable: true,
  get: function () {
    return _utils.getTypeMetaFromSource;
  }
});
Object.defineProperty(exports, "getUniqueId", {
  enumerable: true,
  get: function () {
    return _utils.getUniqueId;
  }
});
Object.defineProperty(exports, "normalize", {
  enumerable: true,
  get: function () {
    return _utils.normalize;
  }
});
Object.defineProperty(exports, "normalizeRuntimeFields", {
  enumerable: true,
  get: function () {
    return _utils.normalizeRuntimeFields;
  }
});
Object.defineProperty(exports, "shouldDeleteChildFieldsAfterTypeChange", {
  enumerable: true,
  get: function () {
    return _utils.shouldDeleteChildFieldsAfterTypeChange;
  }
});
Object.defineProperty(exports, "stripUndefinedValues", {
  enumerable: true,
  get: function () {
    return _utils.stripUndefinedValues;
  }
});
Object.defineProperty(exports, "updateFieldsPathAfterFieldNameChange", {
  enumerable: true,
  get: function () {
    return _utils.updateFieldsPathAfterFieldNameChange;
  }
});
var _utils = require("./utils");
var _serializers = require("./serializers");
Object.keys(_serializers).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _serializers[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _serializers[key];
    }
  });
});
var _validators = require("./validators");
Object.keys(_validators).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _validators[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _validators[key];
    }
  });
});
var _mappings_validator = require("./mappings_validator");
Object.keys(_mappings_validator).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mappings_validator[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mappings_validator[key];
    }
  });
});
var _search_fields = require("./search_fields");
Object.keys(_search_fields).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _search_fields[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _search_fields[key];
    }
  });
});
var _extract_mappings_definition = require("./extract_mappings_definition");
Object.keys(_extract_mappings_definition).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _extract_mappings_definition[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _extract_mappings_definition[key];
    }
  });
});
