// @kbn/stack-connectors-plugin/common/d3security/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SUB_ACTION = exports.D3_SECURITY_TITLE = exports.D3_SECURITY_CONNECTOR_ID = exports.D3SecuritySeverity = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const D3_SECURITY_TITLE = exports.D3_SECURITY_TITLE = _i18n.i18n.translate('xpack.stackConnectors.components.d3Security.connectorTypeTitle', {
  defaultMessage: 'D3 Security'
});
const D3_SECURITY_CONNECTOR_ID = exports.D3_SECURITY_CONNECTOR_ID = '.d3security';
let SUB_ACTION = exports.SUB_ACTION = /*#__PURE__*/function (SUB_ACTION) {
  SUB_ACTION["RUN"] = "run";
  SUB_ACTION["TEST"] = "test";
  return SUB_ACTION;
}({});
let D3SecuritySeverity = exports.D3SecuritySeverity = /*#__PURE__*/function (D3SecuritySeverity) {
  D3SecuritySeverity["EMPTY"] = "";
  D3SecuritySeverity["HIGH"] = "high";
  D3SecuritySeverity["MEDIUM"] = "medium";
  D3SecuritySeverity["LOW"] = "low";
  return D3SecuritySeverity;
}({});
