// @kbn/index-lifecycle-management-plugin/public/application/sections/edit_policy/components/phases/shared_fields/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DataTierAllocationField", {
  enumerable: true,
  get: function () {
    return _data_tier_allocation_field.DataTierAllocationField;
  }
});
Object.defineProperty(exports, "DownsampleField", {
  enumerable: true,
  get: function () {
    return _downsample_field.DownsampleField;
  }
});
Object.defineProperty(exports, "ForcemergeField", {
  enumerable: true,
  get: function () {
    return _forcemerge_field.ForcemergeField;
  }
});
Object.defineProperty(exports, "IndexPriorityField", {
  enumerable: true,
  get: function () {
    return _index_priority_field.IndexPriorityField;
  }
});
Object.defineProperty(exports, "MinAgeField", {
  enumerable: true,
  get: function () {
    return _min_age_field.MinAgeField;
  }
});
Object.defineProperty(exports, "ReadonlyField", {
  enumerable: true,
  get: function () {
    return _readonly_field.ReadonlyField;
  }
});
Object.defineProperty(exports, "ReplicasField", {
  enumerable: true,
  get: function () {
    return _replicas_field.ReplicasField;
  }
});
Object.defineProperty(exports, "SearchableSnapshotField", {
  enumerable: true,
  get: function () {
    return _searchable_snapshot_field.SearchableSnapshotField;
  }
});
Object.defineProperty(exports, "ShrinkField", {
  enumerable: true,
  get: function () {
    return _shrink_field.ShrinkField;
  }
});
Object.defineProperty(exports, "SnapshotPoliciesField", {
  enumerable: true,
  get: function () {
    return _snapshot_policies_field.SnapshotPoliciesField;
  }
});
var _data_tier_allocation_field = require("./data_tier_allocation_field");
var _forcemerge_field = require("./forcemerge_field");
var _min_age_field = require("./min_age_field");
var _snapshot_policies_field = require("./snapshot_policies_field");
var _shrink_field = require("./shrink_field");
var _searchable_snapshot_field = require("./searchable_snapshot_field");
var _readonly_field = require("./readonly_field");
var _replicas_field = require("./replicas_field");
var _index_priority_field = require("./index_priority_field");
var _downsample_field = require("./downsample_field");
