// @kbn/apm-plugin/public/components/alerting/ui_components/apm_rule_kql_filter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ApmRuleKqlFilter = ApmRuleKqlFilter;
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _apm_rule_unified_search_bar = require("./apm_rule_unified_search_bar");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function ApmRuleKqlFilter({
  ruleParams,
  setRuleParams,
  onToggleKqlFilter
}) {
  var _ruleParams$useKqlFil;
  const kqlFilterToggle = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    label: _i18n.i18n.translate('xpack.apm.rules.transactionDuration.kqlFilterToggle', {
      defaultMessage: 'Use KQL Filter'
    }),
    checked: (_ruleParams$useKqlFil = ruleParams.useKqlFilter) !== null && _ruleParams$useKqlFil !== void 0 ? _ruleParams$useKqlFil : false,
    onChange: onToggleKqlFilter
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: 'm'
  }));
  const kqlFilter = ruleParams.useKqlFilter ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    label: _i18n.i18n.translate('xpack.apm.rules.ruleFlyout.filterLabel', {
      defaultMessage: 'Filter'
    }),
    helpText: _i18n.i18n.translate('xpack.apm.rules.ruleFlyout.filterHelpText', {
      defaultMessage: 'Use a KQL expression to limit the scope of your alert trigger.'
    }),
    fullWidth: true,
    display: "rowCompressed"
  }, /*#__PURE__*/_react.default.createElement(_apm_rule_unified_search_bar.ApmRuleUnifiedSearchBar, {
    ruleParams: ruleParams,
    setRuleParams: setRuleParams
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: 'm'
  })) : null;
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, kqlFilterToggle, kqlFilter);
}
