// @kbn/security-solution-plugin/public/detection_engine/rule_creation/components/ml_jobs_description/ml_job_item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MlJobItem = void 0;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _eui = require("@elastic/eui");
var i18n = _interopRequireWildcard(require("./translations"));
var _helpers = require("../../../../../common/machine_learning/helpers");
var _ml_job_link = require("../ml_job_link/ml_job_link");
var _ml_audit_icon = require("../ml_audit_icon");
var _ml_job_status_badge = require("../ml_job_status_badge");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Wrapper = _styledComponents.default.div`
  overflow: hidden;
`;
const MlJobItemComponent = ({
  job,
  switchComponent,
  ...props
}) => {
  var _job$customSettings;
  const isStarted = (0, _helpers.isJobStarted)(job.jobState, job.datafeedState);
  return /*#__PURE__*/_react.default.createElement(Wrapper, props, /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement(_ml_job_link.MlJobLink, {
    jobId: job.id,
    jobName: (_job$customSettings = job.customSettings) === null || _job$customSettings === void 0 ? void 0 : _job$customSettings.security_app_display_name
  }), /*#__PURE__*/_react.default.createElement(_ml_audit_icon.MlAuditIcon, {
    message: job.auditMessage
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "flexStart"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    style: {
      marginRight: '0'
    }
  }, /*#__PURE__*/_react.default.createElement(_ml_job_status_badge.MlJobStatusBadge, {
    job: job
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, switchComponent), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    style: {
      marginLeft: '0'
    },
    "data-test-subj": "mlJobActionLabel"
  }, isStarted ? i18n.ML_STOP_JOB_LABEL : i18n.ML_RUN_JOB_LABEL)));
};
const MlJobItem = exports.MlJobItem = /*#__PURE__*/(0, _react.memo)(MlJobItemComponent);
