// @kbn/monaco/src/esql/lib/esql_state.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ESQLState = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/** @internal **/
class ESQLState {
  constructor() {
    (0, _defineProperty2.default)(this, "lastLine", 0);
  }
  setLineNumber(n) {
    this.lastLine = n;
  }
  getLineNumber() {
    return this.lastLine;
  }
  clone() {
    const newState = new ESQLState();
    newState.setLineNumber(this.lastLine);
    return newState;
  }
  equals(other) {
    return false;
  }
}
exports.ESQLState = ESQLState;
