// @kbn/security-solution-plugin/public/common/components/links_to_docs/links_components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SecuritySolutionRequirementsLink = exports.MlJobCompatibilityLink = exports.DetectionsRequirementsLink = exports.CoverageOverviewLink = void 0;
var _react = _interopRequireDefault(require("react"));
var _doc_link = require("./doc_link");
var i18n = _interopRequireWildcard(require("./links_translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SecuritySolutionRequirementsLink = () => /*#__PURE__*/_react.default.createElement(_doc_link.DocLink, {
  docPath: i18n.SOLUTION_REQUIREMENTS_LINK_PATH,
  linkText: i18n.SOLUTION_REQUIREMENTS_LINK_TEXT
});
exports.SecuritySolutionRequirementsLink = SecuritySolutionRequirementsLink;
const DetectionsRequirementsLink = () => /*#__PURE__*/_react.default.createElement(_doc_link.DocLink, {
  docPath: i18n.DETECTIONS_REQUIREMENTS_LINK_PATH,
  linkText: i18n.DETECTIONS_REQUIREMENTS_LINK_TEXT
});
exports.DetectionsRequirementsLink = DetectionsRequirementsLink;
const MlJobCompatibilityLink = () => /*#__PURE__*/_react.default.createElement(_doc_link.DocLink, {
  docPath: i18n.ML_JOB_COMPATIBILITY_LINK_PATH,
  linkText: i18n.ML_JOB_COMPATIBILITY_LINK_TEXT
});
exports.MlJobCompatibilityLink = MlJobCompatibilityLink;
const CoverageOverviewLink = () => /*#__PURE__*/_react.default.createElement(_doc_link.DocLink, {
  docPath: i18n.COVERAGE_OVERVIEW_LINK_PATH,
  linkText: i18n.COVERAGE_OVERVIEW_LINK_TEXT
});
exports.CoverageOverviewLink = CoverageOverviewLink;
