// @kbn/synthetics-plugin/public/apps/synthetics/state/monitor_list/actions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updateManagementPageStateAction = exports.quietFetchMonitorListAction = exports.fetchUpsertSuccessAction = exports.fetchUpsertMonitorAction = exports.fetchUpsertFailureAction = exports.fetchMonitorListAction = exports.fetchMonitorFiltersAction = exports.enableMonitorAlertAction = exports.clearMonitorUpsertStatus = exports.cleanMonitorListState = void 0;
var _toolkit = require("@reduxjs/toolkit");
var _actions = require("../utils/actions");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const fetchMonitorListAction = exports.fetchMonitorListAction = (0, _actions.createAsyncAction)('fetchMonitorListAction');
const quietFetchMonitorListAction = exports.quietFetchMonitorListAction = (0, _toolkit.createAction)('quietFetchMonitorListAction');
const fetchUpsertMonitorAction = exports.fetchUpsertMonitorAction = (0, _toolkit.createAction)('fetchUpsertMonitor');
const fetchUpsertSuccessAction = exports.fetchUpsertSuccessAction = (0, _toolkit.createAction)('fetchUpsertMonitorSuccess');
const fetchUpsertFailureAction = exports.fetchUpsertFailureAction = (0, _toolkit.createAction)('fetchUpsertMonitorFailure');
const enableMonitorAlertAction = exports.enableMonitorAlertAction = (0, _actions.createAsyncAction)('enableMonitorAlertAction');
const clearMonitorUpsertStatus = exports.clearMonitorUpsertStatus = (0, _toolkit.createAction)('clearMonitorUpsertStatus');
const updateManagementPageStateAction = exports.updateManagementPageStateAction = (0, _toolkit.createAction)('updateManagementPageState');
const cleanMonitorListState = exports.cleanMonitorListState = (0, _toolkit.createAction)('cleanMonitorListState');
const fetchMonitorFiltersAction = exports.fetchMonitorFiltersAction = (0, _actions.createAsyncAction)('fetchMonitorFiltersAction');
