// @kbn/console-plugin/public/application/contexts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EditorContextProvider", {
  enumerable: true,
  get: function () {
    return _editor_context.EditorContextProvider;
  }
});
Object.defineProperty(exports, "PanelContextProvider", {
  enumerable: true,
  get: function () {
    return _split_panel.PanelContextProvider;
  }
});
Object.defineProperty(exports, "PanelRegistry", {
  enumerable: true,
  get: function () {
    return _split_panel.PanelRegistry;
  }
});
Object.defineProperty(exports, "RequestContextProvider", {
  enumerable: true,
  get: function () {
    return _request_context.RequestContextProvider;
  }
});
Object.defineProperty(exports, "ServicesContextProvider", {
  enumerable: true,
  get: function () {
    return _services_context.ServicesContextProvider;
  }
});
Object.defineProperty(exports, "useEditorActionContext", {
  enumerable: true,
  get: function () {
    return _editor_context.useEditorActionContext;
  }
});
Object.defineProperty(exports, "useEditorReadContext", {
  enumerable: true,
  get: function () {
    return _editor_context.useEditorReadContext;
  }
});
Object.defineProperty(exports, "usePanelContext", {
  enumerable: true,
  get: function () {
    return _split_panel.usePanelContext;
  }
});
Object.defineProperty(exports, "useRequestActionContext", {
  enumerable: true,
  get: function () {
    return _request_context.useRequestActionContext;
  }
});
Object.defineProperty(exports, "useRequestReadContext", {
  enumerable: true,
  get: function () {
    return _request_context.useRequestReadContext;
  }
});
Object.defineProperty(exports, "useServicesContext", {
  enumerable: true,
  get: function () {
    return _services_context.useServicesContext;
  }
});
var _services_context = require("./services_context");
var _request_context = require("./request_context");
var _editor_context = require("./editor_context");
var _split_panel = require("./split_panel");
