// @kbn/fleet-plugin/public/components/confirm_force_install_modal.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ConfirmForceInstallModal = void 0;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ConfirmForceInstallModal = ({
  onCancel,
  onConfirm,
  pkg,
  docLinks
}) => {
  const title = pkg && pkg.name && pkg.version ? _i18n.i18n.translate('xpack.fleet.ConfirmForceInstallModal.calloutTitleWithPkg', {
    defaultMessage: 'Integration {pkgName}-{pkgVersion} has failed verification',
    values: {
      pkgName: pkg.name,
      pkgVersion: pkg.version
    }
  }) : _i18n.i18n.translate('xpack.fleet.ConfirmForceInstallModal.calloutTitleNoPkg', {
    defaultMessage: 'The integration has failed verification'
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiConfirmModal, {
    title: /*#__PURE__*/_react.default.createElement("span", {
      className: "eui-textBreakWord"
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.ConfirmForceInstallModal.title",
      defaultMessage: "Install unverified integration?"
    })),
    onCancel: onCancel,
    onConfirm: onConfirm,
    cancelButtonText: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.ConfirmForceInstallModal.cancelButtonLabel",
      defaultMessage: "Cancel"
    }),
    confirmButtonText: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.ConfirmForceInstallModal.confirmButtonLabel",
      defaultMessage: "Install anyway"
    }),
    buttonColor: "danger",
    "data-test-subj": "confirmForceInstallModal"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: title,
    color: "warning",
    iconType: "warning",
    children: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.ConfirmForceInstallModal.calloutBody",
      defaultMessage: "This integration contains an unsigned package of unknown authenticity and could contain malicious files. Learn more about {learnMoreLink}.",
      values: {
        learnMoreLink: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
          target: "_blank",
          external: true,
          href: docLinks.links.fleet.packageSignatures
        }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
          id: "xpack.fleet.ConfirmForceInstallModal.learnMoreLink",
          defaultMessage: "package signatures"
        }))
      }
    })
  }));
};
exports.ConfirmForceInstallModal = ConfirmForceInstallModal;
