// @kbn/logstash-plugin/public/application/components/pipeline_list/alert_call_out.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertCallOut = AlertCallOut;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _constants = require("./constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function AlertCallOut(props) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: _constants.PIPELINE_LIST.INFO_ALERTS.CALL_OUT_TITLE,
    color: "warning",
    iconType: "help"
  }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.logstash.alertCallOut.howToSeeAdditionalPipelinesDescription",
    defaultMessage: "How can I see additional pipelines?"
  })), props.children);
}
