// @kbn/vis-type-timelion-plugin/public/helpers/plugin_services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUsageCollection = exports.setIndexPatterns = exports.setFieldFormats = exports.setDataSearch = exports.setCoreStart = exports.setCharts = exports.getUsageCollection = exports.getIndexPatterns = exports.getFieldFormats = exports.getDataSearch = exports.getCoreStart = exports.getCharts = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getIndexPatterns, setIndexPatterns] = (0, _public.createGetterSetter)('dataViews');
exports.setIndexPatterns = setIndexPatterns;
exports.getIndexPatterns = getIndexPatterns;
const [getDataSearch, setDataSearch] = (0, _public.createGetterSetter)('Search');
exports.setDataSearch = setDataSearch;
exports.getDataSearch = getDataSearch;
const [getCharts, setCharts] = (0, _public.createGetterSetter)('Charts');
exports.setCharts = setCharts;
exports.getCharts = getCharts;
const [getCoreStart, setCoreStart] = (0, _public.createGetterSetter)('CoreStart');
exports.setCoreStart = setCoreStart;
exports.getCoreStart = getCoreStart;
const [getFieldFormats, setFieldFormats] = (0, _public.createGetterSetter)('FieldFormats');
exports.setFieldFormats = setFieldFormats;
exports.getFieldFormats = getFieldFormats;
const [getUsageCollection, setUsageCollection] = (0, _public.createGetterSetter)('UsageCollection', false);
exports.setUsageCollection = setUsageCollection;
exports.getUsageCollection = getUsageCollection;
