// @kbn/security-solution-plugin/public/management/pages/host_isolation_exceptions/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SEARCHABLE_FIELDS = exports.HOST_ISOLATION_EXCEPTIONS_LIST_DEFINITION = void 0;
var _securitysolutionIoTsListTypes = require("@kbn/securitysolution-io-ts-list-types");
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SEARCHABLE_FIELDS = exports.SEARCHABLE_FIELDS = [`item_id`, `name`, `description`, `entries.value`];
const HOST_ISOLATION_EXCEPTIONS_LIST_DEFINITION = exports.HOST_ISOLATION_EXCEPTIONS_LIST_DEFINITION = {
  name: _securitysolutionListConstants.ENDPOINT_HOST_ISOLATION_EXCEPTIONS_LIST_NAME,
  namespace_type: 'agnostic',
  description: _securitysolutionListConstants.ENDPOINT_HOST_ISOLATION_EXCEPTIONS_LIST_DESCRIPTION,
  list_id: _securitysolutionListConstants.ENDPOINT_HOST_ISOLATION_EXCEPTIONS_LIST_ID,
  type: _securitysolutionIoTsListTypes.ExceptionListTypeEnum.ENDPOINT_HOST_ISOLATION_EXCEPTIONS
};
