// @elastic/eui/lib/components/flyout/flyout_resizable.styles.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.euiFlyoutResizableButtonStyles = void 0;
var _react = require("@emotion/react");
var _global_styling = require("../../global_styling");
function _EMOTION_STRINGIFIED_CSS_ERROR__() { return "You have tried to stringify object returned from `css` function. It isn't supposed to be used directly (e.g. as value of the `className` prop), but rather handed to emotion so it can handle it (e.g. as value of `css` prop)."; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
var _ref = process.env.NODE_ENV === "production" ? {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;"
} : {
  name: "1gfwqnd-euiFlyoutResizableButton",
  styles: "position:absolute;label:euiFlyoutResizableButton;",
  toString: _EMOTION_STRINGIFIED_CSS_ERROR__
};
var euiFlyoutResizableButtonStyles = exports.euiFlyoutResizableButtonStyles = function euiFlyoutResizableButtonStyles(_ref2) {
  var euiTheme = _ref2.euiTheme;
  return {
    euiFlyoutResizableButton: _ref,
    overlay: {
      left: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('right', 0), ";;label:left;"),
      right: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('left', 0), ";;label:right;")
    },
    push: {
      left: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('right', "-".concat(euiTheme.border.width.thin)), ";;label:left;"),
      right: /*#__PURE__*/(0, _react.css)((0, _global_styling.logicalCSS)('left', "-".concat(euiTheme.border.width.thin)), ";;label:right;")
    }
  };
};
