// @kbn/ui-actions-enhanced-plugin/public/drilldowns/drilldown_manager/containers/tabs/tabs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.txtManage = exports.txtCreateNew = exports.Tabs = void 0;
var React = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _context = require("../context");
var _form_drilldown_wizard = require("../form_drilldown_wizard");
var _drilldown_list = require("../drilldown_list");
var _template_picker = require("../template_picker");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const txtCreateNew = exports.txtCreateNew = _i18n.i18n.translate('uiActionsEnhanced.drilldowns.containers.DrilldownManager.createNew', {
  defaultMessage: 'Create new'
});
const txtManage = exports.txtManage = _i18n.i18n.translate('uiActionsEnhanced.drilldowns.containers.DrilldownManager.manage', {
  defaultMessage: 'Manage'
});
const tabs = [{
  id: 'create',
  name: txtCreateNew,
  content: /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_eui.EuiSpacer, null), /*#__PURE__*/React.createElement(_form_drilldown_wizard.FormDrilldownWizard, null), /*#__PURE__*/React.createElement(_eui.EuiSpacer, {
    size: 'l'
  }), /*#__PURE__*/React.createElement(_template_picker.TemplatePicker, null))
}, {
  id: 'manage',
  name: txtManage,
  content: /*#__PURE__*/React.createElement(React.Fragment, null, /*#__PURE__*/React.createElement(_eui.EuiSpacer, null), /*#__PURE__*/React.createElement(_drilldown_list.DrilldownList, null))
}];
const Tabs = ({}) => {
  const drilldowns = (0, _context.useDrilldownManager)();
  const route = drilldowns.useRoute();
  return /*#__PURE__*/React.createElement(_eui.EuiTabbedContent, {
    tabs: tabs,
    selectedTab: tabs.find(({
      id
    }) => id === route[0]),
    onTabClick: ({
      id
    }) => drilldowns.setRoute([id])
  });
};
exports.Tabs = Tabs;
