// @kbn/security-solution-plugin/public/common/mock/router.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Router", {
  enumerable: true,
  get: function () {
    return _reactRouterDom.Router;
  }
});
exports.mockHistory = void 0;
Object.defineProperty(exports, "routeData", {
  enumerable: true,
  get: function () {
    return _reactRouter.default;
  }
});
var _reactRouterDom = require("react-router-dom");
var _reactRouter = _interopRequireDefault(require("react-router"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// eslint-disable-next-line @kbn/eslint/module_migration

const pop = 'POP';
const location = {
  pathname: '/network',
  search: '',
  state: '',
  hash: ''
};
const mockHistory = exports.mockHistory = {
  length: 2,
  location,
  action: pop,
  push: jest.fn(),
  replace: jest.fn(),
  go: jest.fn(),
  goBack: jest.fn(),
  goForward: jest.fn(),
  block: jest.fn(),
  createHref: jest.fn(),
  listen: jest.fn()
};
