// @kbn/serverless-search/common/i18n_string.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPDATE_LABEL = exports.SCHEDULING_LABEL = exports.SAVE_LABEL = exports.REQUIRED_LABEL = exports.OVERVIEW_LABEL = exports.OPTIONAL_LABEL = exports.NEXT_LABEL = exports.LEARN_MORE_LABEL = exports.INVALID_JSON_ERROR = exports.ENABLED_LABEL = exports.EDIT_LABEL = exports.DISABLED_LABEL = exports.DELETE_CONNECTOR_LABEL = exports.COPY_CONNECTOR_ID_LABEL = exports.CONNECTOR_LABEL = exports.CONNECTORS_LABEL = exports.CONFIGURATION_LABEL = exports.CANCEL_LABEL = exports.BACK_LABEL = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LEARN_MORE_LABEL = exports.LEARN_MORE_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.learnMore', {
  defaultMessage: 'Learn more'
});
const CANCEL_LABEL = exports.CANCEL_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.cancel', {
  defaultMessage: 'Cancel'
});
const EDIT_LABEL = exports.EDIT_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.edit', {
  defaultMessage: 'Edit'
});
const SAVE_LABEL = exports.SAVE_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.save', {
  defaultMessage: 'Save'
});
const UPDATE_LABEL = exports.UPDATE_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.update', {
  defaultMessage: 'Update'
});
const BACK_LABEL = exports.BACK_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.back', {
  defaultMessage: 'Back'
});
const NEXT_LABEL = exports.NEXT_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.next', {
  defaultMessage: 'Next'
});
const REQUIRED_LABEL = exports.REQUIRED_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.required', {
  defaultMessage: 'Required'
});
const OPTIONAL_LABEL = exports.OPTIONAL_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.optional', {
  defaultMessage: 'Optional'
});
const ENABLED_LABEL = exports.ENABLED_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.enabled', {
  defaultMessage: 'Enabled'
});
const DISABLED_LABEL = exports.DISABLED_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.disabled', {
  defaultMessage: 'Disabled'
});
const INVALID_JSON_ERROR = exports.INVALID_JSON_ERROR = _i18n.i18n.translate('xpack.serverlessSearch.invalidJsonError', {
  defaultMessage: 'Invalid JSON'
});
const CONNECTORS_LABEL = exports.CONNECTORS_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors', {
  defaultMessage: 'Connectors'
});
const CONNECTOR_LABEL = exports.CONNECTOR_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connector', {
  defaultMessage: 'Connector'
});
const DELETE_CONNECTOR_LABEL = exports.DELETE_CONNECTOR_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors.deleteConnectorLabel', {
  defaultMessage: 'Delete connector'
});
const COPY_CONNECTOR_ID_LABEL = exports.COPY_CONNECTOR_ID_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors.copyConnectorIdLabel', {
  defaultMessage: 'Copy connector id'
});
const OVERVIEW_LABEL = exports.OVERVIEW_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors.overviewLabel', {
  defaultMessage: 'Overview'
});
const CONFIGURATION_LABEL = exports.CONFIGURATION_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors.configurationLabel', {
  defaultMessage: 'Configuration'
});
const SCHEDULING_LABEL = exports.SCHEDULING_LABEL = _i18n.i18n.translate('xpack.serverlessSearch.connectors.schedulingLabel', {
  defaultMessage: 'Scheduling'
});
