// @kbn/home-plugin/server/services/sample_data/lib/sample_dataset_registry_types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EmbeddableTypes = exports.DatasetStatusTypes = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let DatasetStatusTypes = exports.DatasetStatusTypes = /*#__PURE__*/function (DatasetStatusTypes) {
  DatasetStatusTypes["NOT_INSTALLED"] = "not_installed";
  DatasetStatusTypes["INSTALLED"] = "installed";
  DatasetStatusTypes["UNKNOWN"] = "unknown";
  return DatasetStatusTypes;
}({});
let EmbeddableTypes = exports.EmbeddableTypes = /*#__PURE__*/function (EmbeddableTypes) {
  EmbeddableTypes["MAP_SAVED_OBJECT_TYPE"] = "map";
  EmbeddableTypes["SEARCH_EMBEDDABLE_TYPE"] = "search";
  EmbeddableTypes["VISUALIZE_EMBEDDABLE_TYPE"] = "visualization";
  return EmbeddableTypes;
}({});
/** This type is used to identify an object in a sample dataset. */
/**
 * This type is used by consumers to register a new app link for a sample dataset.
 */
