// @kbn/index-lifecycle-management-plugin/public/application/sections/edit_policy/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LearnMoreLink: true,
  OptionalLabel: true,
  PolicyJsonFlyout: true,
  DescribedFormRow: true,
  ToggleFieldWithDescribedFormRow: true,
  FieldLoadingError: true,
  Timeline: true,
  FormErrorsCallout: true,
  PhaseFooter: true,
  InfinityIcon: true,
  EditWarning: true
};
Object.defineProperty(exports, "DescribedFormRow", {
  enumerable: true,
  get: function () {
    return _described_form_row.DescribedFormRow;
  }
});
Object.defineProperty(exports, "EditWarning", {
  enumerable: true,
  get: function () {
    return _edit_warning.EditWarning;
  }
});
Object.defineProperty(exports, "FieldLoadingError", {
  enumerable: true,
  get: function () {
    return _field_loading_error.FieldLoadingError;
  }
});
Object.defineProperty(exports, "FormErrorsCallout", {
  enumerable: true,
  get: function () {
    return _form_errors_callout.FormErrorsCallout;
  }
});
Object.defineProperty(exports, "InfinityIcon", {
  enumerable: true,
  get: function () {
    return _infinity_icon.InfinityIcon;
  }
});
Object.defineProperty(exports, "LearnMoreLink", {
  enumerable: true,
  get: function () {
    return _learn_more_link.LearnMoreLink;
  }
});
Object.defineProperty(exports, "OptionalLabel", {
  enumerable: true,
  get: function () {
    return _optional_label.OptionalLabel;
  }
});
Object.defineProperty(exports, "PhaseFooter", {
  enumerable: true,
  get: function () {
    return _phase_footer.PhaseFooter;
  }
});
Object.defineProperty(exports, "PolicyJsonFlyout", {
  enumerable: true,
  get: function () {
    return _policy_json_flyout.PolicyJsonFlyout;
  }
});
Object.defineProperty(exports, "Timeline", {
  enumerable: true,
  get: function () {
    return _timeline.Timeline;
  }
});
Object.defineProperty(exports, "ToggleFieldWithDescribedFormRow", {
  enumerable: true,
  get: function () {
    return _described_form_row.ToggleFieldWithDescribedFormRow;
  }
});
var _learn_more_link = require("./learn_more_link");
var _optional_label = require("./optional_label");
var _policy_json_flyout = require("./policy_json_flyout");
var _described_form_row = require("./described_form_row");
var _field_loading_error = require("./field_loading_error");
var _timeline = require("./timeline");
var _form_errors_callout = require("./form_errors_callout");
var _phase_footer = require("./phase_footer");
var _infinity_icon = require("./infinity_icon");
var _edit_warning = require("./edit_warning");
var _phases = require("./phases");
Object.keys(_phases).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _phases[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _phases[key];
    }
  });
});
