// @kbn/index-lifecycle-management-plugin/public/application/sections/edit_policy/components/learn_more_link.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LearnMoreLink = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LearnMoreLink = ({
  docPath,
  text
}) => {
  const content = text ? text : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.indexLifecycleMgmt.learnMore",
    defaultMessage: "Learn more"
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    href: docPath,
    target: "_blank",
    external: true
  }, content);
};
exports.LearnMoreLink = LearnMoreLink;
