// @kbn/entityManager-plugin/common/errors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ERROR_USER_NOT_AUTHORIZED = exports.ERROR_PARTIAL_BUILTIN_INSTALLATION = exports.ERROR_DEFINITION_STOPPED = exports.ERROR_API_KEY_SERVICE_DISABLED = exports.ERROR_API_KEY_NOT_VALID = exports.ERROR_API_KEY_NOT_FOUND = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ERROR_API_KEY_NOT_FOUND = exports.ERROR_API_KEY_NOT_FOUND = 'api_key_not_found';
const ERROR_API_KEY_NOT_VALID = exports.ERROR_API_KEY_NOT_VALID = 'api_key_not_valid';
const ERROR_USER_NOT_AUTHORIZED = exports.ERROR_USER_NOT_AUTHORIZED = 'user_not_authorized';
const ERROR_API_KEY_SERVICE_DISABLED = exports.ERROR_API_KEY_SERVICE_DISABLED = 'api_key_service_disabled';
const ERROR_PARTIAL_BUILTIN_INSTALLATION = exports.ERROR_PARTIAL_BUILTIN_INSTALLATION = 'partial_builtin_installation';
const ERROR_DEFINITION_STOPPED = exports.ERROR_DEFINITION_STOPPED = 'error_definition_stopped';
