// @kbn/security-solution-plugin/common/endpoint/service/response_actions/is_response_action_supported.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isActionSupportedByAgentType = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/** @private */
const RESPONSE_ACTIONS_SUPPORT_MAP = {
  isolate: {
    automated: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: true,
      crowdstrike: true
    }
  },
  unisolate: {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: true,
      crowdstrike: true
    }
  },
  upload: {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  },
  'get-file': {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: true,
      crowdstrike: false
    }
  },
  'kill-process': {
    automated: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  },
  execute: {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  },
  'suspend-process': {
    automated: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  },
  'running-processes': {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  },
  scan: {
    automated: {
      endpoint: false,
      sentinel_one: false,
      crowdstrike: false
    },
    manual: {
      endpoint: true,
      sentinel_one: false,
      crowdstrike: false
    }
  }
};

/**
 * Check if a given Response action is supported (implemented) for a given agent type and action type
 * @param agentType
 * @param actionName
 * @param actionType
 */
const isActionSupportedByAgentType = (agentType, actionName, actionType) => {
  return RESPONSE_ACTIONS_SUPPORT_MAP[actionName][actionType][agentType];
};
exports.isActionSupportedByAgentType = isActionSupportedByAgentType;
