// @kbn/discover-plugin/public/components/doc_table/components/table_header/score_sort_warning.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocViewTableScoreSortWarning = DocViewTableScoreSortWarning;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function DocViewTableScoreSortWarning() {
  const tooltipContent = _i18n.i18n.translate('discover.docViews.table.scoreSortWarningTooltip', {
    defaultMessage: 'In order to retrieve values for _score, you must sort by it.'
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    content: tooltipContent,
    color: "warning",
    size: "s",
    type: "warning"
  });
}
