// @kbn/discover-plugin/public/components/doc_table/components/table_header/table_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableHeader = TableHeader;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _common = require("@kbn/field-formats-plugin/common");
var _discoverUtils = require("@kbn/discover-utils");
var _table_header_column = require("./table_header_column");
var _helpers = require("./helpers");
var _sorting = require("../../../../utils/sorting");
var _use_discover_services = require("../../../../hooks/use_discover_services");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function TableHeader({
  columns,
  dataView,
  onChangeSortOrder,
  onMoveColumn,
  onRemoveColumn,
  sortOrder
}) {
  const {
    uiSettings
  } = (0, _use_discover_services.useDiscoverServices)();
  const [defaultSortOrder, hideTimeColumn, isShortDots] = (0, _react.useMemo)(() => [uiSettings.get(_discoverUtils.SORT_DEFAULT_ORDER_SETTING, 'desc'), uiSettings.get(_discoverUtils.DOC_HIDE_TIME_COLUMN_SETTING, false), uiSettings.get(_common.FORMATS_UI_SETTINGS.SHORT_DOTS_ENABLE)], [uiSettings]);
  const displayedColumns = (0, _helpers.getDisplayedColumns)(columns, dataView, hideTimeColumn, isShortDots);
  return /*#__PURE__*/_react.default.createElement("tr", {
    "data-test-subj": "docTableHeader",
    className: "kbnDocTableHeader"
  }, /*#__PURE__*/_react.default.createElement("th", {
    style: {
      width: '24px'
    }
  }), displayedColumns.map((col, index) => {
    var _dataView$getFieldByN;
    return /*#__PURE__*/_react.default.createElement(_table_header_column.TableHeaderColumn, (0, _extends2.default)({
      key: `${col.name}-${index}`
    }, col, {
      customLabel: (_dataView$getFieldByN = dataView.getFieldByName(col.name)) === null || _dataView$getFieldByN === void 0 ? void 0 : _dataView$getFieldByN.customLabel,
      isTimeColumn: dataView.timeFieldName === col.name,
      sortOrder: sortOrder.length ? sortOrder : (0, _sorting.getDefaultSort)(dataView, defaultSortOrder, hideTimeColumn, false),
      onMoveColumn: onMoveColumn,
      onRemoveColumn: onRemoveColumn,
      onChangeSortOrder: onChangeSortOrder
    }));
  }));
}
