// @kbn/infra-plugin/public/pages/logs/log_entry_rate/page_setup_content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogEntryRateSetupContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _sharedUxPageKibanaTemplate = require("@kbn/shared-ux-page-kibana-template");
var _eui = require("@elastic/eui");
var _public = require("@kbn/observability-shared-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LogEntryRateSetupContent = ({
  onOpenSetup
}) => {
  (0, _public.useTrackPageview)({
    app: 'infra_logs',
    path: 'log_entry_rate_setup'
  });
  (0, _public.useTrackPageview)({
    app: 'infra_logs',
    path: 'log_entry_rate_setup',
    delay: 15000
  });
  return /*#__PURE__*/_react.default.createElement(_sharedUxPageKibanaTemplate.KibanaPageTemplate.EmptyPrompt, {
    "data-test-subj": "logEntryRateSetupPage",
    title: /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.infra.logs.logEntryRate.setupTitle",
      defaultMessage: "Set up log anomaly analysis"
    })),
    body: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "s"
    }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.infra.logs.logEntryRate.setupDescription",
      defaultMessage: "To enable log anomalies, set up a machine learning job"
    }))),
    actions: /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      "data-test-subj": "infraLogEntryRateSetupContentMlSetupButton",
      fill: true,
      onClick: onOpenSetup
    }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.infra.logs.logEntryRate.showAnalysisSetupButtonLabel",
      defaultMessage: "ML Setup"
    }))
  });
};
exports.LogEntryRateSetupContent = LogEntryRateSetupContent;
