// @kbn/apm-plugin/public/components/app/mobile/service_overview/stats/metric_item.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MetricItem = MetricItem;
var _react = _interopRequireDefault(require("react"));
var _charts = require("@elastic/charts");
var _eui = require("@elastic/eui");
var _lodash = require("lodash");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function MetricItem({
  data,
  id,
  isLoading,
  height = '124px'
}) {
  const hasData = !(0, _lodash.isEmpty)(data);
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      resize: 'none',
      padding: '0px',
      overflow: 'auto',
      height,
      borderRadius: '6px'
    }
  }, !hasData && isLoading ? /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSkeletonText, {
    lines: 3
  })) : /*#__PURE__*/_react.default.createElement(_charts.Chart, null, /*#__PURE__*/_react.default.createElement(_charts.Metric, {
    id: `metric_${id}`,
    data: [data]
  })));
}
