// @kbn/dashboard-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DASHBOARD_APP_ID", {
  enumerable: true,
  get: function () {
    return _dashboard_constants.DASHBOARD_APP_ID;
  }
});
Object.defineProperty(exports, "DASHBOARD_CONTAINER_TYPE", {
  enumerable: true,
  get: function () {
    return _dashboard_container.DASHBOARD_CONTAINER_TYPE;
  }
});
Object.defineProperty(exports, "DASHBOARD_GRID_COLUMN_COUNT", {
  enumerable: true,
  get: function () {
    return _dashboard_constants.DASHBOARD_GRID_COLUMN_COUNT;
  }
});
Object.defineProperty(exports, "DashboardListingTable", {
  enumerable: true,
  get: function () {
    return _dashboard_listing.DashboardListingTable;
  }
});
Object.defineProperty(exports, "DashboardRenderer", {
  enumerable: true,
  get: function () {
    return _dashboard_container.DashboardRenderer;
  }
});
Object.defineProperty(exports, "DashboardTopNav", {
  enumerable: true,
  get: function () {
    return _dashboard_top_nav.DashboardTopNav;
  }
});
Object.defineProperty(exports, "LEGACY_DASHBOARD_APP_ID", {
  enumerable: true,
  get: function () {
    return _dashboard_constants.LEGACY_DASHBOARD_APP_ID;
  }
});
Object.defineProperty(exports, "PanelPlacementStrategy", {
  enumerable: true,
  get: function () {
    return _dashboard_constants.PanelPlacementStrategy;
  }
});
Object.defineProperty(exports, "cleanEmptyKeys", {
  enumerable: true,
  get: function () {
    return _locator.cleanEmptyKeys;
  }
});
Object.defineProperty(exports, "createDashboardEditUrl", {
  enumerable: true,
  get: function () {
    return _dashboard_constants.createDashboardEditUrl;
  }
});
Object.defineProperty(exports, "getDashboardLocatorParamsFromEmbeddable", {
  enumerable: true,
  get: function () {
    return _get_dashboard_locator_params.getDashboardLocatorParamsFromEmbeddable;
  }
});
exports.plugin = plugin;
var _plugin = require("./plugin");
var _dashboard_constants = require("./dashboard_constants");
var _dashboard_container = require("./dashboard_container");
var _dashboard_listing = require("./dashboard_listing");
var _dashboard_top_nav = require("./dashboard_top_nav");
var _locator = require("./dashboard_app/locator/locator");
var _get_dashboard_locator_params = require("./dashboard_app/locator/get_dashboard_locator_params");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function plugin(initializerContext) {
  return new _plugin.DashboardPlugin(initializerContext);
}
