// @kbn/security-solution-plugin/common/api/entity_analytics/asset_criticality/get_asset_criticality_privileges.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EntityAnalyticsPrivileges = void 0;
var _zod = require("zod");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Get Asset Criticality Privileges Schema
 *   version: 1
 */

const EntityAnalyticsPrivileges = exports.EntityAnalyticsPrivileges = _zod.z.object({
  has_all_required: _zod.z.boolean(),
  has_read_permissions: _zod.z.boolean().optional(),
  has_write_permissions: _zod.z.boolean().optional(),
  privileges: _zod.z.object({
    elasticsearch: _zod.z.object({
      cluster: _zod.z.object({
        manage_index_templates: _zod.z.boolean().optional(),
        manage_transform: _zod.z.boolean().optional()
      }).optional(),
      index: _zod.z.object({}).catchall(_zod.z.object({
        read: _zod.z.boolean().optional(),
        write: _zod.z.boolean().optional()
      })).optional()
    })
  })
});
