// @kbn/security-solution-features/src/security/product_feature_config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.securityDefaultProductFeaturesConfig = void 0;
var _product_features_keys = require("../product_features_keys");
var _constants = require("../constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * App features privileges configuration for the Security Solution Kibana Feature app.
 * These are the configs that are shared between both offering types (ess and serverless).
 * They can be extended on each offering plugin to register privileges using different way on each offering type.
 *
 * Privileges can be added in different ways:
 * - `privileges`: the privileges that will be added directly into the main Security feature.
 * - `subFeatureIds`: the ids of the sub-features that will be added into the Security subFeatures entry.
 * - `subFeaturesPrivileges`: the privileges that will be added into the existing Security subFeature with the privilege `id` specified.
 */
const securityDefaultProductFeaturesConfig = exports.securityDefaultProductFeaturesConfig = {
  [_product_features_keys.ProductFeatureSecurityKey.advancedInsights]: {
    privileges: {
      all: {
        ui: ['entity-analytics'],
        api: [`${_constants.APP_ID}-entity-analytics`]
      },
      read: {
        ui: ['entity-analytics'],
        api: [`${_constants.APP_ID}-entity-analytics`]
      }
    }
  },
  [_product_features_keys.ProductFeatureSecurityKey.investigationGuide]: {
    privileges: {
      all: {
        ui: ['investigation-guide']
      },
      read: {
        ui: ['investigation-guide']
      }
    }
  },
  [_product_features_keys.ProductFeatureSecurityKey.investigationGuideInteractions]: {
    privileges: {
      all: {
        ui: ['investigation-guide-interactions']
      },
      read: {
        ui: ['investigation-guide-interactions']
      }
    }
  },
  [_product_features_keys.ProductFeatureSecurityKey.threatIntelligence]: {
    privileges: {
      all: {
        ui: ['threat-intelligence'],
        api: [`${_constants.APP_ID}-threat-intelligence`]
      },
      read: {
        ui: ['threat-intelligence'],
        api: [`${_constants.APP_ID}-threat-intelligence`]
      }
    }
  },
  [_product_features_keys.ProductFeatureSecurityKey.endpointHostManagement]: {
    subFeatureIds: [_product_features_keys.SecuritySubFeatureId.endpointList]
  },
  [_product_features_keys.ProductFeatureSecurityKey.endpointPolicyManagement]: {
    subFeatureIds: [_product_features_keys.SecuritySubFeatureId.policyManagement]
  },
  // Adds no additional kibana feature controls
  [_product_features_keys.ProductFeatureSecurityKey.endpointPolicyProtections]: {},
  [_product_features_keys.ProductFeatureSecurityKey.endpointArtifactManagement]: {
    subFeatureIds: [_product_features_keys.SecuritySubFeatureId.trustedApplications, _product_features_keys.SecuritySubFeatureId.blocklist, _product_features_keys.SecuritySubFeatureId.eventFilters],
    subFeaturesPrivileges: [{
      id: 'host_isolation_exceptions_all',
      api: [`${_constants.APP_ID}-accessHostIsolationExceptions`, `${_constants.APP_ID}-writeHostIsolationExceptions`],
      ui: ['accessHostIsolationExceptions', 'writeHostIsolationExceptions']
    }, {
      id: 'host_isolation_exceptions_read',
      api: [`${_constants.APP_ID}-accessHostIsolationExceptions`],
      ui: ['accessHostIsolationExceptions']
    }]
  },
  [_product_features_keys.ProductFeatureSecurityKey.endpointResponseActions]: {
    subFeatureIds: [_product_features_keys.SecuritySubFeatureId.hostIsolationExceptions, _product_features_keys.SecuritySubFeatureId.responseActionsHistory, _product_features_keys.SecuritySubFeatureId.hostIsolation, _product_features_keys.SecuritySubFeatureId.processOperations, _product_features_keys.SecuritySubFeatureId.fileOperations, _product_features_keys.SecuritySubFeatureId.executeAction, _product_features_keys.SecuritySubFeatureId.scanAction],
    subFeaturesPrivileges: [{
      id: 'host_isolation_all',
      api: [`${_constants.APP_ID}-writeHostIsolation`],
      ui: ['writeHostIsolation']
    }]
  },
  // Product features without RBAC
  [_product_features_keys.ProductFeatureSecurityKey.osqueryAutomatedResponseActions]: {},
  [_product_features_keys.ProductFeatureSecurityKey.endpointProtectionUpdates]: {},
  [_product_features_keys.ProductFeatureSecurityKey.endpointAgentTamperProtection]: {},
  [_product_features_keys.ProductFeatureSecurityKey.externalRuleActions]: {},
  [_product_features_keys.ProductFeatureSecurityKey.cloudSecurityPosture]: {},
  [_product_features_keys.ProductFeatureSecurityKey.integrationAssistant]: {}
};
