// @kbn/alerting-state-types/src/task_state/v1/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "alertStateSchema", {
  enumerable: true,
  get: function () {
    return _schema.alertStateSchema;
  }
});
Object.defineProperty(exports, "lastScheduledActionsSchema", {
  enumerable: true,
  get: function () {
    return _schema.lastScheduledActionsSchema;
  }
});
Object.defineProperty(exports, "metaSchema", {
  enumerable: true,
  get: function () {
    return _schema.metaSchema;
  }
});
Object.defineProperty(exports, "rawAlertInstanceSchema", {
  enumerable: true,
  get: function () {
    return _schema.rawAlertInstanceSchema;
  }
});
Object.defineProperty(exports, "throttledActionSchema", {
  enumerable: true,
  get: function () {
    return _schema.throttledActionSchema;
  }
});
exports.versionDefinition = void 0;
Object.defineProperty(exports, "versionSchema", {
  enumerable: true,
  get: function () {
    return _schema.versionSchema;
  }
});
var _migration = require("./migration");
var _schema = require("./schema");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const versionDefinition = exports.versionDefinition = {
  up: _migration.upMigration,
  schema: _schema.versionSchema
};
