// @kbn/aiops-plugin/server/routes/log_rate_analysis/response_stream_utils/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PROGRESS_STEP_P_VALUES = exports.PROGRESS_STEP_HISTOGRAMS_GROUPS = exports.PROGRESS_STEP_HISTOGRAMS = exports.PROGRESS_STEP_GROUPING = exports.MAX_CONCURRENT_QUERIES = exports.LOADED_FIELD_CANDIDATES = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// Overall progress is a float from 0 to 1.
const LOADED_FIELD_CANDIDATES = exports.LOADED_FIELD_CANDIDATES = 0.2;
const PROGRESS_STEP_P_VALUES = exports.PROGRESS_STEP_P_VALUES = 0.5;
const PROGRESS_STEP_GROUPING = exports.PROGRESS_STEP_GROUPING = 0.1;
const PROGRESS_STEP_HISTOGRAMS = exports.PROGRESS_STEP_HISTOGRAMS = 0.1;
const PROGRESS_STEP_HISTOGRAMS_GROUPS = exports.PROGRESS_STEP_HISTOGRAMS_GROUPS = 0.1;

// Don't use more than 10 here otherwise Kibana will emit an error
// regarding a limit of abort signal listeners of more than 10.
const MAX_CONCURRENT_QUERIES = exports.MAX_CONCURRENT_QUERIES = 10;
