// @kbn/data-views-plugin/common/data_view.stub.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "createStubDataView", {
  enumerable: true,
  get: function () {
    return _data_view.createStubDataView;
  }
});
Object.defineProperty(exports, "createStubIndexPattern", {
  enumerable: true,
  get: function () {
    return _data_view.createStubDataView;
  }
});
exports.stubLogstashDataView = exports.stubIndexPattern = exports.stubDataViewWithoutTimeField = exports.stubDataView = void 0;
exports.stubbedSavedObjectDataView = stubbedSavedObjectDataView;
exports.stubbedSavedObjectIndexPattern = void 0;
var _field = require("./field.stub");
var _data_view = require("./data_views/data_view.stub");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const stubDataView = exports.stubDataView = (0, _data_view.createStubDataView)({
  spec: {
    id: 'logstash-*',
    fields: _field.stubFieldSpecMap,
    title: 'logstash-*',
    timeFieldName: '@timestamp'
  }
});
const stubIndexPattern = exports.stubIndexPattern = stubDataView;
const stubDataViewWithoutTimeField = exports.stubDataViewWithoutTimeField = (0, _data_view.createStubDataView)({
  spec: {
    id: 'logstash-*',
    fields: _field.stubFieldSpecMap,
    title: 'logstash-*'
  }
});
const stubLogstashDataView = exports.stubLogstashDataView = (0, _data_view.createStubDataView)({
  spec: {
    id: 'logstash-*',
    title: 'logstash-*',
    timeFieldName: 'time',
    fields: _field.stubLogstashFieldSpecMap
  }
});
function stubbedSavedObjectDataView(id = null) {
  return {
    id: id !== null && id !== void 0 ? id : '',
    type: 'index-pattern',
    attributes: {
      timeFieldName: 'time',
      fields: JSON.stringify(_field.stubLogstashFieldSpecMap),
      title: 'title',
      name: 'Name'
    },
    version: '2',
    references: []
  };
}
const stubbedSavedObjectIndexPattern = exports.stubbedSavedObjectIndexPattern = stubbedSavedObjectDataView;
