// @kbn/event-annotation-listing-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AnnotationEditorControls", {
  enumerable: true,
  get: function () {
    return _eventAnnotationComponents.AnnotationEditorControls;
  }
});
Object.defineProperty(exports, "annotationsIconSet", {
  enumerable: true,
  get: function () {
    return _eventAnnotationComponents.annotationsIconSet;
  }
});
Object.defineProperty(exports, "defaultAnnotationColor", {
  enumerable: true,
  get: function () {
    return _eventAnnotationCommon.defaultAnnotationColor;
  }
});
Object.defineProperty(exports, "defaultAnnotationRangeColor", {
  enumerable: true,
  get: function () {
    return _eventAnnotationCommon.defaultAnnotationRangeColor;
  }
});
Object.defineProperty(exports, "getAnnotationAccessor", {
  enumerable: true,
  get: function () {
    return _eventAnnotationComponents.getAnnotationAccessor;
  }
});
Object.defineProperty(exports, "isManualPointAnnotationConfig", {
  enumerable: true,
  get: function () {
    return _eventAnnotationCommon.isManualPointAnnotationConfig;
  }
});
Object.defineProperty(exports, "isQueryAnnotationConfig", {
  enumerable: true,
  get: function () {
    return _eventAnnotationCommon.isQueryAnnotationConfig;
  }
});
Object.defineProperty(exports, "isRangeAnnotationConfig", {
  enumerable: true,
  get: function () {
    return _eventAnnotationCommon.isRangeAnnotationConfig;
  }
});
exports.plugin = void 0;
var _plugin = require("./plugin");
var _eventAnnotationCommon = require("@kbn/event-annotation-common");
var _eventAnnotationComponents = require("@kbn/event-annotation-components");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const plugin = () => new _plugin.EventAnnotationListingPlugin();
exports.plugin = plugin;
