// @kbn/cases-plugin/public/components/create/templates.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TemplateSelectorComponent = exports.TemplateSelector = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
var _experimental_badge = require("../experimental_badge/experimental_badge");
var _optional_field_label = require("../optional_field_label");
var _translations = require("./translations");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TemplateSelectorComponent = ({
  isLoading,
  templates,
  onTemplateChange
}) => {
  const [selectedTemplate, onSelectTemplate] = (0, _react.useState)();
  const isSmallScreen = (0, _eui.useIsWithinMaxBreakpoint)('s');
  const options = templates.map(template => ({
    text: template.name,
    value: template.key
  }));
  const onChange = (0, _react.useCallback)(e => {
    const selectedTemplated = templates.find(template => template.key === e.target.value);
    if (selectedTemplated) {
      onSelectTemplate(selectedTemplated.key);
      onTemplateChange(selectedTemplated.caseFields);
    }
  }, [onTemplateChange, templates]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    id: "createCaseTemplate",
    fullWidth: true,
    label: _translations.TEMPLATE_LABEL,
    labelAppend: /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      alignItems: "center",
      gutterSize: "s",
      css: (0, _react2.css)`
            flex-grow: 0;
          `,
      responsive: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false,
      css: (0, _react2.css)`
              line-height: 0;
            `
    }, /*#__PURE__*/_react.default.createElement(_experimental_badge.ExperimentalBadge, {
      compact: isSmallScreen
    })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      grow: false
    }, _optional_field_label.OptionalFieldLabel)),
    helpText: _translations.TEMPLATE_HELP_TEXT
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSelect, {
    onChange: onChange,
    options: options,
    disabled: isLoading,
    isLoading: isLoading,
    "data-test-subj": "create-case-template-select",
    fullWidth: true,
    hasNoInitialSelection: true,
    value: selectedTemplate
  }));
};
exports.TemplateSelectorComponent = TemplateSelectorComponent;
TemplateSelectorComponent.displayName = 'TemplateSelector';
const TemplateSelector = exports.TemplateSelector = /*#__PURE__*/_react.default.memo(TemplateSelectorComponent);
