// @kbn/security-solution-plugin/public/timelines/components/timeline/tabs/eql/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.EqlTabContentComponent = void 0;
var _eui = require("@elastic/eui");
var _fp = require("lodash/fp");
var _react = _interopRequireWildcard(require("react"));
var _reactRedux = require("react-redux");
var _fastDeepEqual = _interopRequireDefault(require("fast-deep-equal"));
var _reactReversePortal = require("react-reverse-portal");
var _unifiedDataTable = require("@kbn/unified-data-table");
var _expandableFlyout = require("@kbn/expandable-flyout");
var _kibana = require("../../../../../common/lib/kibana");
var _panel_keys = require("../../../../../flyout/document_details/shared/constants/panel_keys");
var _constants = require("../../../../../common/store/inputs/constants");
var _use_selector = require("../../../../../common/hooks/use_selector");
var _use_experimental_features = require("../../../../../common/hooks/use_experimental_features");
var _store = require("../../../../store");
var _containers = require("../../../../containers");
var _body = require("../../body");
var _footer = require("../../footer");
var _helpers = require("../../helpers");
var _refetch_timeline = require("../../refetch_timeline");
var _timeline = require("../../../../../../common/types/timeline");
var _event_details_width_context = require("../../../../../common/components/events_viewer/event_details_width_context");
var _store2 = require("../../../../../common/store");
var _model = require("../../../../../sourcerer/store/model");
var _defaults = require("../../../../store/defaults");
var _containers2 = require("../../../../../sourcerer/containers");
var _use_timeline_events_count = require("../../../../../common/hooks/use_timeline_events_count");
var _use_full_screen = require("../../../../../common/containers/use_full_screen");
var _side_panel = require("../../../side_panel");
var _layout = require("../shared/layout");
var _utils = require("../shared/utils");
var _unified_timeline_body = require("../../body/unified_timeline_body");
var _header = require("./header");
var _use_timeline_columns = require("../shared/use_timeline_columns");
var _use_timeline_control_columns = require("../shared/use_timeline_control_columns");
var _left = require("../../../../../flyout/document_details/left");
var _use_notes_in_flyout = require("../../properties/use_notes_in_flyout");
var _notes_flyout = require("../../properties/notes_flyout");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EqlTabContentComponent = ({
  activeTab,
  columns,
  end,
  eqlOptions,
  expandedDetail,
  timelineId,
  isLive,
  itemsPerPage,
  itemsPerPageOptions,
  onEventClosed,
  renderCellValue,
  rowRenderers,
  showExpandedDetails,
  start,
  timerangeKind,
  pinnedEventIds,
  eventIdToNoteIds
}) => {
  var _pageInfo$activePage;
  const {
    telemetry
  } = (0, _kibana.useKibana)().services;
  const dispatch = (0, _reactRedux.useDispatch)();
  const {
    query: eqlQuery = '',
    ...restEqlOption
  } = eqlOptions;
  const {
    portalNode: eqlEventsCountPortalNode
  } = (0, _use_timeline_events_count.useEqlEventsCountPortal)();
  const {
    setTimelineFullScreen,
    timelineFullScreen
  } = (0, _use_full_screen.useTimelineFullScreen)();
  const {
    browserFields,
    dataViewId,
    loading: loadingSourcerer,
    runtimeMappings,
    selectedPatterns
  } = (0, _containers2.useSourcererDataView)(_model.SourcererScopeName.timeline);
  const {
    augmentedColumnHeaders,
    timelineQueryFieldsFromColumns
  } = (0, _use_timeline_columns.useTimelineColumns)(columns);
  const unifiedComponentsInTimelineDisabled = (0, _use_experimental_features.useIsExperimentalFeatureEnabled)('unifiedComponentsInTimelineDisabled');
  const getManageTimeline = (0, _react.useMemo)(() => _store.timelineSelectors.getTimelineByIdSelector(), []);
  const currentTimeline = (0, _use_selector.useDeepEqualSelector)(state => getManageTimeline(state, timelineId !== null && timelineId !== void 0 ? timelineId : _timeline.TimelineId.active));
  const {
    sampleSize
  } = currentTimeline;
  const isBlankTimeline = (0, _fp.isEmpty)(eqlQuery);
  const canQueryTimeline = (0, _react.useCallback)(() => loadingSourcerer != null && !loadingSourcerer && !(0, _fp.isEmpty)(start) && !(0, _fp.isEmpty)(end) && !isBlankTimeline, [end, isBlankTimeline, loadingSourcerer, start]);
  const [dataLoadingState, {
    events,
    inspect,
    totalCount,
    pageInfo,
    loadPage,
    refreshedAt,
    refetch
  }] = (0, _containers.useTimelineEvents)({
    dataViewId,
    endDate: end,
    eqlOptions: restEqlOption,
    fields: timelineQueryFieldsFromColumns,
    filterQuery: eqlQuery !== null && eqlQuery !== void 0 ? eqlQuery : '',
    id: timelineId,
    indexNames: selectedPatterns,
    language: 'eql',
    limit: !unifiedComponentsInTimelineDisabled ? sampleSize : itemsPerPage,
    runtimeMappings,
    skip: !canQueryTimeline(),
    startDate: start,
    timerangeKind
  });
  const expandableFlyoutDisabled = (0, _use_experimental_features.useIsExperimentalFeatureEnabled)('expandableFlyoutDisabled');
  const {
    openFlyout
  } = (0, _expandableFlyout.useExpandableFlyoutApi)();
  const securitySolutionNotesEnabled = (0, _use_experimental_features.useIsExperimentalFeatureEnabled)('securitySolutionNotesEnabled');
  const {
    associateNote,
    notes,
    isNotesFlyoutVisible,
    closeNotesFlyout,
    showNotesFlyout,
    eventId: noteEventId,
    setNotesEventId
  } = (0, _use_notes_in_flyout.useNotesInFlyout)({
    eventIdToNoteIds,
    refetch,
    timelineId,
    activeTab: _timeline.TimelineTabs.eql
  });
  const onToggleShowNotes = (0, _react.useCallback)(eventId => {
    const indexName = selectedPatterns.join(',');
    if (eventId && !expandableFlyoutDisabled && securitySolutionNotesEnabled) {
      openFlyout({
        right: {
          id: _panel_keys.DocumentDetailsRightPanelKey,
          params: {
            id: eventId,
            indexName,
            scopeId: timelineId
          }
        },
        left: {
          id: _panel_keys.DocumentDetailsLeftPanelKey,
          path: {
            tab: _left.LeftPanelNotesTab
          },
          params: {
            id: eventId,
            indexName,
            scopeId: timelineId
          }
        }
      });
      telemetry.reportOpenNoteInExpandableFlyoutClicked({
        location: timelineId
      });
      telemetry.reportDetailsFlyoutOpened({
        location: timelineId,
        panel: 'left'
      });
    } else {
      if (eventId) {
        setNotesEventId(eventId);
        showNotesFlyout();
      }
    }
  }, [expandableFlyoutDisabled, openFlyout, securitySolutionNotesEnabled, selectedPatterns, telemetry, timelineId, setNotesEventId, showNotesFlyout]);
  const leadingControlColumns = (0, _use_timeline_control_columns.useTimelineControlColumn)({
    columns,
    sort: _utils.TIMELINE_NO_SORTING,
    timelineId,
    activeTab: _timeline.TimelineTabs.eql,
    refetch,
    events,
    pinnedEventIds,
    eventIdToNoteIds,
    onToggleShowNotes
  });
  const isQueryLoading = (0, _react.useMemo)(() => dataLoadingState === _unifiedDataTable.DataLoadingState.loading || dataLoadingState === _unifiedDataTable.DataLoadingState.loadingMore, [dataLoadingState]);
  const handleOnPanelClosed = (0, _react.useCallback)(() => {
    onEventClosed({
      tabType: _timeline.TimelineTabs.eql,
      id: timelineId
    });
  }, [onEventClosed, timelineId]);
  (0, _react.useEffect)(() => {
    dispatch(_store.timelineActions.updateIsLoading({
      id: timelineId,
      isLoading: isQueryLoading || loadingSourcerer
    }));
  }, [loadingSourcerer, timelineId, isQueryLoading, dispatch]);
  const unifiedHeader = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_header.EqlTabHeader, {
    activeTab: activeTab,
    setTimelineFullScreen: setTimelineFullScreen,
    timelineFullScreen: timelineFullScreen,
    timelineId: timelineId
  })), [activeTab, setTimelineFullScreen, timelineFullScreen, timelineId]);
  const NotesFlyoutMemo = (0, _react.useMemo)(() => {
    return /*#__PURE__*/_react.default.createElement(_notes_flyout.NotesFlyout, {
      associateNote: associateNote,
      eventId: noteEventId,
      show: isNotesFlyoutVisible,
      notes: notes,
      onClose: closeNotesFlyout,
      onCancel: closeNotesFlyout,
      timelineId: timelineId
    });
  }, [associateNote, closeNotesFlyout, isNotesFlyoutVisible, noteEventId, notes, timelineId]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, !unifiedComponentsInTimelineDisabled ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactReversePortal.InPortal, {
    node: eqlEventsCountPortalNode
  }, totalCount >= 0 ? /*#__PURE__*/_react.default.createElement(_layout.EventsCountBadge, {
    "data-test-subj": "eql-events-count"
  }, totalCount) : null), NotesFlyoutMemo, /*#__PURE__*/_react.default.createElement(_layout.FullWidthFlexGroup, null, /*#__PURE__*/_react.default.createElement(_unified_timeline_body.UnifiedTimelineBody, {
    header: unifiedHeader,
    columns: augmentedColumnHeaders,
    isSortEnabled: false,
    rowRenderers: rowRenderers,
    timelineId: timelineId,
    itemsPerPage: itemsPerPage,
    itemsPerPageOptions: itemsPerPageOptions,
    sort: _utils.TIMELINE_NO_SORTING,
    events: events,
    refetch: refetch,
    dataLoadingState: dataLoadingState,
    totalCount: isBlankTimeline ? 0 : totalCount,
    onEventClosed: onEventClosed,
    expandedDetail: expandedDetail,
    showExpandedDetails: showExpandedDetails,
    onChangePage: loadPage,
    activeTab: activeTab,
    updatedAt: refreshedAt,
    isTextBasedQuery: false,
    pageInfo: pageInfo,
    leadingControlColumns: leadingControlColumns
  }))) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_reactReversePortal.InPortal, {
    node: eqlEventsCountPortalNode
  }, totalCount >= 0 ? /*#__PURE__*/_react.default.createElement(_layout.EventsCountBadge, null, totalCount) : null), NotesFlyoutMemo, /*#__PURE__*/_react.default.createElement(_refetch_timeline.TimelineRefetch, {
    id: `${timelineId}-${_timeline.TimelineTabs.eql}`,
    inputId: _constants.InputsModelId.timeline,
    inspect: inspect,
    loading: isQueryLoading,
    refetch: refetch
  }), /*#__PURE__*/_react.default.createElement(_layout.FullWidthFlexGroup, {
    gutterSize: "s",
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_layout.ScrollableFlexItem, {
    grow: false
  }, unifiedHeader), /*#__PURE__*/_react.default.createElement(_layout.ScrollableFlexItem, {
    grow: true
  }, /*#__PURE__*/_react.default.createElement(_event_details_width_context.EventDetailsWidthProvider, null, /*#__PURE__*/_react.default.createElement(_layout.StyledEuiFlyoutBody, {
    "data-test-subj": `${_timeline.TimelineTabs.eql}-tab-flyout-body`,
    className: "timeline-flyout-body"
  }, /*#__PURE__*/_react.default.createElement(_body.StatefulBody, {
    activePage: pageInfo.activePage,
    browserFields: browserFields,
    data: isBlankTimeline ? _utils.TIMELINE_EMPTY_EVENTS : events,
    id: timelineId,
    refetch: refetch,
    renderCellValue: renderCellValue,
    rowRenderers: rowRenderers,
    sort: _utils.TIMELINE_NO_SORTING,
    tabType: _timeline.TimelineTabs.eql,
    totalPages: (0, _helpers.calculateTotalPages)({
      itemsCount: totalCount,
      itemsPerPage
    }),
    leadingControlColumns: leadingControlColumns,
    trailingControlColumns: _utils.timelineEmptyTrailingControlColumns
  })), /*#__PURE__*/_react.default.createElement(_layout.StyledEuiFlyoutFooter, {
    "data-test-subj": `${_timeline.TimelineTabs.eql}-tab-flyout-footer`,
    className: "timeline-flyout-footer"
  }, !isBlankTimeline && /*#__PURE__*/_react.default.createElement(_footer.Footer, {
    activePage: (_pageInfo$activePage = pageInfo === null || pageInfo === void 0 ? void 0 : pageInfo.activePage) !== null && _pageInfo$activePage !== void 0 ? _pageInfo$activePage : 0,
    "data-test-subj": "timeline-footer",
    updatedAt: refreshedAt,
    height: _footer.footerHeight,
    id: timelineId,
    isLive: isLive,
    isLoading: isQueryLoading || loadingSourcerer,
    itemsCount: isBlankTimeline ? 0 : events.length,
    itemsPerPage: itemsPerPage,
    itemsPerPageOptions: itemsPerPageOptions,
    onChangePage: loadPage,
    totalCount: isBlankTimeline ? 0 : totalCount
  })))), showExpandedDetails && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_layout.VerticalRule, null), /*#__PURE__*/_react.default.createElement(_layout.ScrollableFlexItem, {
    grow: 1
  }, /*#__PURE__*/_react.default.createElement(_side_panel.DetailsPanel, {
    browserFields: browserFields,
    runtimeMappings: runtimeMappings,
    tabType: _timeline.TimelineTabs.eql,
    scopeId: timelineId,
    handleOnPanelClosed: handleOnPanelClosed
  }))))));
};
exports.EqlTabContentComponent = EqlTabContentComponent;
const makeMapStateToProps = () => {
  const getTimeline = _store.timelineSelectors.getTimelineByIdSelector();
  const getInputsTimeline = _store2.inputsSelectors.getTimelineSelector();
  const mapStateToProps = (state, {
    timelineId
  }) => {
    var _getTimeline, _expandedDetail$Timel;
    const timeline = (_getTimeline = getTimeline(state, timelineId)) !== null && _getTimeline !== void 0 ? _getTimeline : _defaults.timelineDefaults;
    const input = getInputsTimeline(state);
    const {
      activeTab,
      columns,
      eqlOptions,
      expandedDetail,
      itemsPerPage,
      itemsPerPageOptions,
      pinnedEventIds,
      eventIdToNoteIds
    } = timeline;
    return {
      activeTab,
      columns,
      eqlOptions,
      end: input.timerange.to,
      expandedDetail,
      timelineId,
      isLive: input.policy.kind === 'interval',
      itemsPerPage,
      itemsPerPageOptions,
      pinnedEventIds,
      eventIdToNoteIds,
      showExpandedDetails: !!expandedDetail[_timeline.TimelineTabs.eql] && !!((_expandedDetail$Timel = expandedDetail[_timeline.TimelineTabs.eql]) !== null && _expandedDetail$Timel !== void 0 && _expandedDetail$Timel.panelView),
      start: input.timerange.from,
      timerangeKind: input.timerange.kind
    };
  };
  return mapStateToProps;
};
const mapDispatchToProps = dispatch => ({
  onEventClosed: args => {
    dispatch(_store.timelineActions.toggleDetailPanel(args));
  }
});
const connector = (0, _reactRedux.connect)(makeMapStateToProps, mapDispatchToProps);
const EqlTabContent = exports.default = connector( /*#__PURE__*/_react.default.memo(EqlTabContentComponent, (prevProps, nextProps) => prevProps.activeTab === nextProps.activeTab && (0, _utils.isTimerangeSame)(prevProps, nextProps) && (0, _fastDeepEqual.default)(prevProps.eqlOptions, nextProps.eqlOptions) && prevProps.isLive === nextProps.isLive && prevProps.itemsPerPage === nextProps.itemsPerPage && prevProps.onEventClosed === nextProps.onEventClosed && prevProps.showExpandedDetails === nextProps.showExpandedDetails && prevProps.timelineId === nextProps.timelineId && (0, _fastDeepEqual.default)(prevProps.columns, nextProps.columns) && (0, _fastDeepEqual.default)(prevProps.pinnedEventIds, nextProps.pinnedEventIds) && (0, _fastDeepEqual.default)(prevProps.eventIdToNoteIds, nextProps.eventIdToNoteIds) && (0, _fastDeepEqual.default)(prevProps.itemsPerPageOptions, nextProps.itemsPerPageOptions)));

// eslint-disable-next-line import/no-default-export
