// @kbn/fleet-plugin/public/mock/plugin_dependencies.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createStartDepsMock = exports.createSetupDepsMock = void 0;
var _mocks = require("@kbn/unified-search-plugin/public/mocks");
var _mocks2 = require("@kbn/data-plugin/public/mocks");
var _mocks3 = require("@kbn/data-views-plugin/public/mocks");
var _mocks4 = require("@kbn/licensing-plugin/public/mocks");
var _mocks5 = require("@kbn/cloud-plugin/public/mocks");
var _mocks6 = require("@kbn/home-plugin/public/mocks");
var _mocks7 = require("@kbn/navigation-plugin/public/mocks");
var _mocks8 = require("@kbn/custom-integrations-plugin/public/mocks");
var _mocks9 = require("@kbn/share-plugin/public/mocks");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createSetupDepsMock = () => {
  const cloud = _mocks5.cloudMock.createSetup();
  return {
    data: _mocks2.dataPluginMock.createSetupContract(),
    home: _mocks6.homePluginMock.createSetupContract(),
    customIntegrations: _mocks8.customIntegrationsMock.createSetup(),
    cloud
  };
};
exports.createSetupDepsMock = createSetupDepsMock;
const createStartDepsMock = () => {
  return {
    licensing: _mocks4.licensingMock.createStart(),
    data: _mocks2.dataPluginMock.createStartContract(),
    dataViews: _mocks3.dataViewPluginMocks.createStartContract(),
    unifiedSearch: _mocks.unifiedSearchPluginMock.createStartContract(),
    navigation: _mocks7.navigationPluginMock.createStartContract(),
    customIntegrations: _mocks8.customIntegrationsMock.createStart(),
    share: _mocks9.sharePluginMock.createStartContract(),
    cloud: _mocks5.cloudMock.createStart()
  };
};
exports.createStartDepsMock = createStartDepsMock;
