// @kbn/alerting-plugin/common/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LEGACY_BASE_ALERT_API_PATH: true,
  BASE_ALERTING_API_PATH: true,
  INTERNAL_BASE_ALERTING_API_PATH: true,
  INTERNAL_ALERTING_SNOOZE_RULE: true,
  INTERNAL_ALERTING_API_FIND_RULES_PATH: true,
  INTERNAL_ALERTING_API_MAINTENANCE_WINDOW_PATH: true,
  INTERNAL_ALERTING_API_GET_ACTIVE_MAINTENANCE_WINDOWS_PATH: true,
  INTERNAL_ALERTING_BACKFILL_API_PATH: true,
  INTERNAL_ALERTING_BACKFILL_FIND_API_PATH: true,
  INTERNAL_ALERTING_BACKFILL_SCHEDULE_API_PATH: true,
  ALERTING_FEATURE_ID: true,
  MONITORING_HISTORY_LIMIT: true,
  ENABLE_MAINTENANCE_WINDOWS: true,
  MaintenanceWindowStatus: true,
  MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE: true,
  MAINTENANCE_WINDOW_FEATURE_ID: true,
  MAINTENANCE_WINDOW_API_PRIVILEGES: true,
  MAINTENANCE_WINDOWS_APP_ID: true,
  MANAGEMENT_APP_ID: true,
  MAINTENANCE_WINDOW_PATHS: true,
  MAINTENANCE_WINDOW_DEEP_LINK_IDS: true,
  MAINTENANCE_WINDOW_DATE_FORMAT: true,
  mappingFromFieldMap: true,
  getComponentTemplateFromFieldMap: true,
  contextToSchemaName: true
};
exports.LEGACY_BASE_ALERT_API_PATH = exports.INTERNAL_BASE_ALERTING_API_PATH = exports.INTERNAL_ALERTING_SNOOZE_RULE = exports.INTERNAL_ALERTING_BACKFILL_SCHEDULE_API_PATH = exports.INTERNAL_ALERTING_BACKFILL_FIND_API_PATH = exports.INTERNAL_ALERTING_BACKFILL_API_PATH = exports.INTERNAL_ALERTING_API_MAINTENANCE_WINDOW_PATH = exports.INTERNAL_ALERTING_API_GET_ACTIVE_MAINTENANCE_WINDOWS_PATH = exports.INTERNAL_ALERTING_API_FIND_RULES_PATH = exports.ENABLE_MAINTENANCE_WINDOWS = exports.BASE_ALERTING_API_PATH = exports.ALERTING_FEATURE_ID = void 0;
Object.defineProperty(exports, "MAINTENANCE_WINDOWS_APP_ID", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOWS_APP_ID;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_API_PRIVILEGES", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_API_PRIVILEGES;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_DATE_FORMAT", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_DATE_FORMAT;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_DEEP_LINK_IDS", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_DEEP_LINK_IDS;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_FEATURE_ID", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_FEATURE_ID;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_PATHS", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_PATHS;
  }
});
Object.defineProperty(exports, "MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MAINTENANCE_WINDOW_SAVED_OBJECT_TYPE;
  }
});
Object.defineProperty(exports, "MANAGEMENT_APP_ID", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MANAGEMENT_APP_ID;
  }
});
exports.MONITORING_HISTORY_LIMIT = void 0;
Object.defineProperty(exports, "MaintenanceWindowStatus", {
  enumerable: true,
  get: function () {
    return _maintenance_window.MaintenanceWindowStatus;
  }
});
Object.defineProperty(exports, "contextToSchemaName", {
  enumerable: true,
  get: function () {
    return _alert_schema.contextToSchemaName;
  }
});
Object.defineProperty(exports, "getComponentTemplateFromFieldMap", {
  enumerable: true,
  get: function () {
    return _alert_schema.getComponentTemplateFromFieldMap;
  }
});
Object.defineProperty(exports, "mappingFromFieldMap", {
  enumerable: true,
  get: function () {
    return _alert_schema.mappingFromFieldMap;
  }
});
var _rule = require("./rule");
Object.keys(_rule).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule[key];
    }
  });
});
var _rules_settings = require("./rules_settings");
Object.keys(_rules_settings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rules_settings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rules_settings[key];
    }
  });
});
var _rule_type = require("./rule_type");
Object.keys(_rule_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule_type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule_type[key];
    }
  });
});
var _lib = require("./lib");
Object.keys(_lib).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _lib[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _lib[key];
    }
  });
});
var _alert_summary = require("./alert_summary");
Object.keys(_alert_summary).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _alert_summary[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _alert_summary[key];
    }
  });
});
var _builtin_action_groups = require("./builtin_action_groups");
Object.keys(_builtin_action_groups).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _builtin_action_groups[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _builtin_action_groups[key];
    }
  });
});
var _bulk_edit = require("./bulk_edit");
Object.keys(_bulk_edit).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bulk_edit[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bulk_edit[key];
    }
  });
});
var _disabled_action_groups = require("./disabled_action_groups");
Object.keys(_disabled_action_groups).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _disabled_action_groups[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _disabled_action_groups[key];
    }
  });
});
var _rule_notify_when_type = require("./rule_notify_when_type");
Object.keys(_rule_notify_when_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule_notify_when_type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule_notify_when_type[key];
    }
  });
});
var _parse_duration = require("./parse_duration");
Object.keys(_parse_duration).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _parse_duration[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _parse_duration[key];
    }
  });
});
var _execution_log_types = require("./execution_log_types");
Object.keys(_execution_log_types).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _execution_log_types[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _execution_log_types[key];
    }
  });
});
var _rule_snooze_type = require("./rule_snooze_type");
Object.keys(_rule_snooze_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule_snooze_type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule_snooze_type[key];
    }
  });
});
var _rrule_type = require("./rrule_type");
Object.keys(_rrule_type).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rrule_type[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rrule_type[key];
    }
  });
});
var _rule_tags_aggregation = require("./rule_tags_aggregation");
Object.keys(_rule_tags_aggregation).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule_tags_aggregation[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule_tags_aggregation[key];
    }
  });
});
var _iso_weekdays = require("./iso_weekdays");
Object.keys(_iso_weekdays).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _iso_weekdays[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _iso_weekdays[key];
    }
  });
});
var _mappings = require("./saved_objects/rules/mappings");
Object.keys(_mappings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _mappings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _mappings[key];
    }
  });
});
var _rule_circuit_breaker_error_message = require("./rule_circuit_breaker_error_message");
Object.keys(_rule_circuit_breaker_error_message).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _rule_circuit_breaker_error_message[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _rule_circuit_breaker_error_message[key];
    }
  });
});
var _maintenance_window_scoped_query_error_message = require("./maintenance_window_scoped_query_error_message");
Object.keys(_maintenance_window_scoped_query_error_message).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _maintenance_window_scoped_query_error_message[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _maintenance_window_scoped_query_error_message[key];
    }
  });
});
var _action_ref_prefix = require("./action_ref_prefix");
Object.keys(_action_ref_prefix).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _action_ref_prefix[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _action_ref_prefix[key];
    }
  });
});
var _maintenance_window = require("./maintenance_window");
var _alert_schema = require("./alert_schema");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// TODO: https://github.com/elastic/kibana/issues/110895
/* eslint-disable @kbn/eslint/no_export_all */

const LEGACY_BASE_ALERT_API_PATH = exports.LEGACY_BASE_ALERT_API_PATH = '/api/alerts';
const BASE_ALERTING_API_PATH = exports.BASE_ALERTING_API_PATH = '/api/alerting';
const INTERNAL_BASE_ALERTING_API_PATH = exports.INTERNAL_BASE_ALERTING_API_PATH = '/internal/alerting';
const INTERNAL_ALERTING_SNOOZE_RULE = exports.INTERNAL_ALERTING_SNOOZE_RULE = `${INTERNAL_BASE_ALERTING_API_PATH}/rule/{id}/_snooze`;
const INTERNAL_ALERTING_API_FIND_RULES_PATH = exports.INTERNAL_ALERTING_API_FIND_RULES_PATH = `${INTERNAL_BASE_ALERTING_API_PATH}/rules/_find`;
const INTERNAL_ALERTING_API_MAINTENANCE_WINDOW_PATH = exports.INTERNAL_ALERTING_API_MAINTENANCE_WINDOW_PATH = `${INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window`;
const INTERNAL_ALERTING_API_GET_ACTIVE_MAINTENANCE_WINDOWS_PATH = exports.INTERNAL_ALERTING_API_GET_ACTIVE_MAINTENANCE_WINDOWS_PATH = `${INTERNAL_ALERTING_API_MAINTENANCE_WINDOW_PATH}/_active`;
const INTERNAL_ALERTING_BACKFILL_API_PATH = exports.INTERNAL_ALERTING_BACKFILL_API_PATH = `${INTERNAL_BASE_ALERTING_API_PATH}/rules/backfill`;
const INTERNAL_ALERTING_BACKFILL_FIND_API_PATH = exports.INTERNAL_ALERTING_BACKFILL_FIND_API_PATH = `${INTERNAL_ALERTING_BACKFILL_API_PATH}/_find`;
const INTERNAL_ALERTING_BACKFILL_SCHEDULE_API_PATH = exports.INTERNAL_ALERTING_BACKFILL_SCHEDULE_API_PATH = `${INTERNAL_ALERTING_BACKFILL_API_PATH}/_schedule`;
const ALERTING_FEATURE_ID = exports.ALERTING_FEATURE_ID = 'alerts';
const MONITORING_HISTORY_LIMIT = exports.MONITORING_HISTORY_LIMIT = 200;
const ENABLE_MAINTENANCE_WINDOWS = exports.ENABLE_MAINTENANCE_WINDOWS = true;
