// @kbn/alerting-plugin/common/rules_settings.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RULES_SETTINGS_SAVED_OBJECT_TYPE = exports.RULES_SETTINGS_QUERY_DELAY_SAVED_OBJECT_ID = exports.RULES_SETTINGS_FLAPPING_SAVED_OBJECT_ID = exports.RULES_SETTINGS_FEATURE_ID = exports.READ_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = exports.READ_FLAPPING_SETTINGS_SUB_FEATURE_ID = exports.MIN_STATUS_CHANGE_THRESHOLD = exports.MIN_QUERY_DELAY = exports.MIN_LOOK_BACK_WINDOW = exports.MAX_STATUS_CHANGE_THRESHOLD = exports.MAX_QUERY_DELAY = exports.MAX_LOOK_BACK_WINDOW = exports.DISABLE_FLAPPING_SETTINGS = exports.DEFAULT_STATUS_CHANGE_THRESHOLD = exports.DEFAULT_SERVERLESS_QUERY_DELAY_SETTINGS = exports.DEFAULT_SERVERLESS_QUERY_DELAY = exports.DEFAULT_QUERY_DELAY_SETTINGS = exports.DEFAULT_QUERY_DELAY = exports.DEFAULT_LOOK_BACK_WINDOW = exports.DEFAULT_FLAPPING_SETTINGS = exports.API_PRIVILEGES = exports.ALL_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = exports.ALL_FLAPPING_SETTINGS_SUB_FEATURE_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const MIN_LOOK_BACK_WINDOW = exports.MIN_LOOK_BACK_WINDOW = 2;
const MAX_LOOK_BACK_WINDOW = exports.MAX_LOOK_BACK_WINDOW = 20;
const MIN_STATUS_CHANGE_THRESHOLD = exports.MIN_STATUS_CHANGE_THRESHOLD = 2;
const MAX_STATUS_CHANGE_THRESHOLD = exports.MAX_STATUS_CHANGE_THRESHOLD = 20;
const MIN_QUERY_DELAY = exports.MIN_QUERY_DELAY = 0;
const MAX_QUERY_DELAY = exports.MAX_QUERY_DELAY = 60;
const RULES_SETTINGS_FEATURE_ID = exports.RULES_SETTINGS_FEATURE_ID = 'rulesSettings';
const ALL_FLAPPING_SETTINGS_SUB_FEATURE_ID = exports.ALL_FLAPPING_SETTINGS_SUB_FEATURE_ID = 'allFlappingSettings';
const READ_FLAPPING_SETTINGS_SUB_FEATURE_ID = exports.READ_FLAPPING_SETTINGS_SUB_FEATURE_ID = 'readFlappingSettings';
const ALL_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = exports.ALL_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = 'allQueryDelaySettings';
const READ_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = exports.READ_QUERY_DELAY_SETTINGS_SUB_FEATURE_ID = 'readQueryDelaySettings';
const API_PRIVILEGES = exports.API_PRIVILEGES = {
  READ_FLAPPING_SETTINGS: 'read-flapping-settings',
  WRITE_FLAPPING_SETTINGS: 'write-flapping-settings',
  READ_QUERY_DELAY_SETTINGS: 'read-query-delay-settings',
  WRITE_QUERY_DELAY_SETTINGS: 'write-query-delay-settings'
};
const RULES_SETTINGS_SAVED_OBJECT_TYPE = exports.RULES_SETTINGS_SAVED_OBJECT_TYPE = 'rules-settings';
const RULES_SETTINGS_FLAPPING_SAVED_OBJECT_ID = exports.RULES_SETTINGS_FLAPPING_SAVED_OBJECT_ID = 'rules-settings';
const RULES_SETTINGS_QUERY_DELAY_SAVED_OBJECT_ID = exports.RULES_SETTINGS_QUERY_DELAY_SAVED_OBJECT_ID = 'query-delay-settings';
const DEFAULT_LOOK_BACK_WINDOW = exports.DEFAULT_LOOK_BACK_WINDOW = 20;
const DEFAULT_STATUS_CHANGE_THRESHOLD = exports.DEFAULT_STATUS_CHANGE_THRESHOLD = 4;
const DEFAULT_QUERY_DELAY = exports.DEFAULT_QUERY_DELAY = 0;
const DEFAULT_SERVERLESS_QUERY_DELAY = exports.DEFAULT_SERVERLESS_QUERY_DELAY = 15;
const DEFAULT_FLAPPING_SETTINGS = exports.DEFAULT_FLAPPING_SETTINGS = {
  enabled: true,
  lookBackWindow: DEFAULT_LOOK_BACK_WINDOW,
  statusChangeThreshold: DEFAULT_STATUS_CHANGE_THRESHOLD
};
const DISABLE_FLAPPING_SETTINGS = exports.DISABLE_FLAPPING_SETTINGS = {
  ...DEFAULT_FLAPPING_SETTINGS,
  enabled: false
};
const DEFAULT_QUERY_DELAY_SETTINGS = exports.DEFAULT_QUERY_DELAY_SETTINGS = {
  delay: DEFAULT_QUERY_DELAY
};
const DEFAULT_SERVERLESS_QUERY_DELAY_SETTINGS = exports.DEFAULT_SERVERLESS_QUERY_DELAY_SETTINGS = {
  delay: DEFAULT_SERVERLESS_QUERY_DELAY
};
