// @kbn/security-solution-plugin/common/api/timeline/pinned_events/pinned_events_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pinnedEventIds = exports.persistPinnedEventSchema = exports.PinnedEventRuntimeType = exports.BarePinnedEventType = void 0;
var runtimeTypes = _interopRequireWildcard(require("io-ts"));
var _utility_types = require("../../../utility_types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/* eslint-disable @typescript-eslint/no-empty-interface */

const pinnedEventIds = exports.pinnedEventIds = (0, _utility_types.unionWithNullType)(runtimeTypes.array(runtimeTypes.string));
const persistPinnedEventSchema = exports.persistPinnedEventSchema = runtimeTypes.intersection([runtimeTypes.type({
  eventId: runtimeTypes.string,
  timelineId: runtimeTypes.string
}), runtimeTypes.partial({
  pinnedEventId: (0, _utility_types.unionWithNullType)(runtimeTypes.string)
})]);

/*
 *  Note Types
 */
const BarePinnedEventType = exports.BarePinnedEventType = runtimeTypes.intersection([runtimeTypes.type({
  timelineId: runtimeTypes.string,
  eventId: runtimeTypes.string
}), runtimeTypes.partial({
  created: (0, _utility_types.unionWithNullType)(runtimeTypes.number),
  createdBy: (0, _utility_types.unionWithNullType)(runtimeTypes.string),
  updated: (0, _utility_types.unionWithNullType)(runtimeTypes.number),
  updatedBy: (0, _utility_types.unionWithNullType)(runtimeTypes.string)
})]);

/**
 * This type represents a pinned event type stored in a saved object that does not include any fields that reference
 * other saved objects.
 */

const PinnedEventRuntimeType = exports.PinnedEventRuntimeType = runtimeTypes.intersection([runtimeTypes.type({
  pinnedEventId: runtimeTypes.string,
  version: runtimeTypes.string
}), BarePinnedEventType]);
