// @kbn/data-visualizer-plugin/public/application/common/components/top_values/top_values.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TopValues = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _classnames = _interopRequireDefault(require("classnames"));
var _i18n = require("@kbn/i18n");
var _fieldTypes = require("@kbn/field-types");
var _react2 = require("@emotion/react");
var _mlNumberUtils = require("@kbn/ml-number-utils");
var _kibana_context = require("../../../kibana_context");
var _utils = require("../utils");
var _expanded_row_field_header = require("../stats_table/components/expanded_row_field_header");
var _expanded_row_panel = require("../stats_table/components/field_data_expanded_row/expanded_row_panel");
var _examples_list = require("../examples_list/examples_list");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function getPercentLabel(percent) {
  if (percent >= 0.1) {
    return `${(0, _mlNumberUtils.roundToDecimalPlace)(percent, 1)}%`;
  } else if (percent === 0) {
    return '0%';
  } else {
    return '< 0.1%';
  }
}
const TopValues = ({
  stats,
  fieldFormat,
  barColor,
  compressed,
  onAddFilter,
  /** Top values by default show % of time a value exist in sampled records/rows (i.e. value A exists in 10% of sampled records)
   * showSampledValues: true shows % of times a value exist in all arrays of values that have been flattened
   * Example for 4 records: ["a", "a", "b"], ["b", "b", "c"], "d", "e"
   * "a" exists in 1/4 records (50% - showSampledValues: false),
   * "a" exists in 2/8 sampled values (25% - showSampledValues: true).
   */
  showSampledValues = false
}) => {
  var _ref, _stats$topValuesSampl, _stats$totalDocuments;
  const {
    services: {
      data: {
        fieldFormats
      }
    }
  } = (0, _kibana_context.useDataVisualizerKibana)();
  if (stats === undefined || !stats.topValues) return null;
  const {
    fieldName,
    sampleCount,
    approximate
  } = stats;
  const originalTopValues = (_ref = showSampledValues ? stats.sampledValues : stats.topValues) !== null && _ref !== void 0 ? _ref : [];
  if ((originalTopValues === null || originalTopValues === void 0 ? void 0 : originalTopValues.length) === 0) return null;
  const totalDocuments = showSampledValues ? (_stats$topValuesSampl = stats.topValuesSampleSize) !== null && _stats$topValuesSampl !== void 0 ? _stats$topValuesSampl : 0 : Math.min(sampleCount !== null && sampleCount !== void 0 ? sampleCount : Infinity, (_stats$totalDocuments = stats.totalDocuments) !== null && _stats$totalDocuments !== void 0 ? _stats$totalDocuments : Infinity);
  const getMessage = () => {
    if (showSampledValues && stats.topValuesSampleSize !== undefined) {
      return /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.dataVisualizer.dataGrid.field.topValues.calculatedFromSampleValuesLabel",
        defaultMessage: "Calculated from {sampledDocumentsFormatted} sample {sampledDocuments, plural, one {value} other {values}}.",
        values: {
          sampledDocuments: stats.topValuesSampleSize,
          sampledDocumentsFormatted: /*#__PURE__*/_react.default.createElement("strong", null, fieldFormats.getDefaultInstance(_fieldTypes.KBN_FIELD_TYPES.NUMBER, [_fieldTypes.ES_FIELD_TYPES.INTEGER]).convert(stats.topValuesSampleSize))
        }
      });
    }
    /**
     * For ES|QL, where are randomly sampling a subset from source data, then query is excuted on top of that data
     * So the terms we get might not get the initial count
     */
    const method = approximate ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.dataVisualizer.dataGrid.field.topValues.estimatedMsg",
      defaultMessage: "Estimated"
    }) : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.dataVisualizer.dataGrid.field.topValues.calculatedMsg",
      defaultMessage: "Calculated"
    });
    return totalDocuments > (sampleCount !== null && sampleCount !== void 0 ? sampleCount : 0) ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.dataVisualizer.dataGrid.field.topValues.calculatedFromSampleRecordsLabel",
      defaultMessage: "{method} from {sampledDocumentsFormatted} sample {sampledDocuments, plural, one {record} other {records}}.",
      values: {
        method,
        sampledDocuments: sampleCount,
        sampledDocumentsFormatted: /*#__PURE__*/_react.default.createElement("strong", null, fieldFormats.getDefaultInstance(_fieldTypes.KBN_FIELD_TYPES.NUMBER, [_fieldTypes.ES_FIELD_TYPES.INTEGER]).convert(sampleCount))
      }
    }) : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.dataVisualizer.dataGrid.field.topValues.calculatedFromTotalRecordsLabel",
      defaultMessage: "{method} from {totalDocumentsFormatted} {totalDocuments, plural, one {record} other {records}}.",
      values: {
        method,
        totalDocuments,
        totalDocumentsFormatted: /*#__PURE__*/_react.default.createElement("strong", null, fieldFormats.getDefaultInstance(_fieldTypes.KBN_FIELD_TYPES.NUMBER, [_fieldTypes.ES_FIELD_TYPES.INTEGER]).convert(totalDocuments !== null && totalDocuments !== void 0 ? totalDocuments : 0))
      }
    });
  };
  const countsElement = /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    color: "subdued",
    size: "xs"
  }, getMessage());
  const topValues = originalTopValues.map(bucket => ({
    ...bucket,
    percent: typeof bucket.percent === 'number' ? bucket.percent : bucket.doc_count / totalDocuments
  }));
  const shouldShowOtherCount = approximate !== true;
  const topValuesOtherCountPercent = 1 - (topValues ? topValues.reduce((acc, bucket) => acc + bucket.percent, 0) : 0);
  const topValuesOtherCount = Math.floor(topValuesOtherCountPercent * (sampleCount !== null && sampleCount !== void 0 ? sampleCount : 0));
  return /*#__PURE__*/_react.default.createElement(_expanded_row_panel.ExpandedRowPanel, {
    dataTestSubj: 'dataVisualizerFieldDataTopValues',
    className: (0, _classnames.default)('dvPanel__wrapper', compressed ? 'dvPanel--compressed' : undefined)
  }, /*#__PURE__*/_react.default.createElement(_expanded_row_field_header.ExpandedRowFieldHeader, null, showSampledValues ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.dataGrid.field.topSampledValuesLabel",
    defaultMessage: "Top sampled values"
  }) : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.dataGrid.field.topValuesLabel",
    defaultMessage: "Top values"
  })), /*#__PURE__*/_react.default.createElement("div", {
    "data-test-subj": "dataVisualizerFieldDataTopValuesContent",
    className: (0, _classnames.default)('fieldDataTopValuesContainer', 'dvTopValues__wrapper')
  }, Array.isArray(topValues) ? topValues.map(value => {
    var _value$key_as_string;
    const fieldValue = (_value$key_as_string = value.key_as_string) !== null && _value$key_as_string !== void 0 ? _value$key_as_string : value.key ? value.key.toString() : '';
    const displayValue = fieldValue === '' ? _examples_list.EMPTY_EXAMPLE : fieldValue;
    return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
      gutterSize: "xs",
      alignItems: "center",
      key: displayValue
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      "data-test-subj": "dataVisualizerFieldDataTopValueBar"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiProgress, {
      value: value.percent,
      max: 1,
      color: barColor,
      size: "xs",
      label: value.key ? (0, _utils.kibanaFieldFormat)(value.key, fieldFormat) : displayValue,
      className: (0, _classnames.default)('eui-textTruncate', 'topValuesValueLabelContainer'),
      valueText: `${value.doc_count}${totalDocuments !== undefined ? ` (${getPercentLabel(value.percent * 100)})` : ''}`
    })), fieldName !== undefined && displayValue !== undefined && onAddFilter !== undefined ? /*#__PURE__*/_react.default.createElement("div", {
      css: (0, _react2.css)`
                        width: 48px;
                      `
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
      iconSize: "s",
      iconType: "plusInCircle",
      onClick: () => onAddFilter(fieldName, fieldValue, '+'),
      "aria-label": _i18n.i18n.translate('xpack.dataVisualizer.dataGrid.field.addFilterAriaLabel', {
        defaultMessage: 'Filter for {fieldName}: "{value}"',
        values: {
          fieldName,
          value: displayValue
        }
      }),
      "data-test-subj": `dvFieldDataTopValuesAddFilterButton-${fieldName}-${displayValue}`,
      style: {
        minHeight: 'auto',
        minWidth: 'auto',
        paddingRight: 2,
        paddingLeft: 2,
        paddingTop: 0,
        paddingBottom: 0
      }
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
      iconSize: "s",
      iconType: "minusInCircle",
      onClick: () => onAddFilter(fieldName, fieldValue, '-'),
      "aria-label": _i18n.i18n.translate('xpack.dataVisualizer.dataGrid.field.removeFilterAriaLabel', {
        defaultMessage: 'Filter out {fieldName}: "{value}"',
        values: {
          fieldName,
          value: displayValue
        }
      }),
      "data-test-subj": `dvFieldDataTopValuesExcludeFilterButton-${fieldName}-${displayValue}`,
      style: {
        minHeight: 'auto',
        minWidth: 'auto',
        paddingTop: 0,
        paddingBottom: 0,
        paddingRight: 2,
        paddingLeft: 2
      }
    })) : null);
  }) : null, shouldShowOtherCount && topValuesOtherCount > 0 ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "xs",
    alignItems: "center",
    key: "other"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "dataVisualizerFieldDataTopValueBar"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiProgress, {
    value: topValuesOtherCount,
    max: totalDocuments,
    color: barColor,
    size: "xs",
    label: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.dataVisualizer.dataGrid.field.topValuesOtherLabel",
      defaultMessage: "Other"
    }),
    className: (0, _classnames.default)('eui-textTruncate', 'topValuesValueLabelContainer'),
    valueText: `${topValuesOtherCount}${totalDocuments !== undefined ? ` (${getPercentLabel(topValuesOtherCountPercent * 100)})` : ''}`
  })), onAddFilter ? /*#__PURE__*/_react.default.createElement("div", {
    css: (0, _react2.css)`
                  width: 48px;
                `
  }) : null) : null, /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), countsElement)));
};
exports.TopValues = TopValues;
