// @kbn/observability-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  plugin: true,
  LazyAlertsFlyout: true,
  enableLegacyUptimeApp: true,
  syntheticsThrottlingEnabled: true,
  enableInspectEsQueries: true,
  enableComparisonByDefault: true,
  apmServiceGroupMaxNumberOfServices: true,
  enableInfrastructureHostsView: true,
  enableInfrastructureContainerAssetView: true,
  enableAgentExplorerView: true,
  apmEnableTableSearchBar: true,
  alertsLocatorID: true,
  ruleDetailsLocatorID: true,
  rulesLocatorID: true,
  sloDetailsLocatorID: true,
  sloEditLocatorID: true,
  uptimeOverviewLocatorID: true,
  observabilityFeatureId: true,
  observabilityAppId: true,
  getCoreVitalsComponent: true,
  ObservabilityAlertSearchBar: true,
  DatePicker: true,
  useFetchDataViews: true,
  useTimeBuckets: true,
  createUseRulesLink: true,
  useSummaryTimeRange: true,
  getApmTraceUrl: true,
  buildEsQuery: true,
  createObservabilityRuleTypeRegistryMock: true,
  DatePickerContextProvider: true,
  fromQuery: true,
  toQuery: true,
  getAlertSummaryTimeRange: true,
  calculateTimeRangeBucketSize: true,
  convertTo: true,
  formatAlertEvaluationValue: true,
  RuleFlyoutKueryBar: true,
  AutocompleteField: true,
  WithKueryAutocompletion: true,
  RuleConditionChart: true,
  getGroupFilters: true
};
Object.defineProperty(exports, "AutocompleteField", {
  enumerable: true,
  get: function () {
    return _rule_kql_filter.AutocompleteField;
  }
});
Object.defineProperty(exports, "DatePicker", {
  enumerable: true,
  get: function () {
    return _date_picker.DatePicker;
  }
});
Object.defineProperty(exports, "DatePickerContextProvider", {
  enumerable: true,
  get: function () {
    return _date_picker_context.DatePickerContextProvider;
  }
});
exports.LazyAlertsFlyout = void 0;
Object.defineProperty(exports, "ObservabilityAlertSearchBar", {
  enumerable: true,
  get: function () {
    return _get_alert_search_bar_lazy.ObservabilityAlertSearchBar;
  }
});
Object.defineProperty(exports, "RuleConditionChart", {
  enumerable: true,
  get: function () {
    return _rule_condition_chart.RuleConditionChart;
  }
});
Object.defineProperty(exports, "RuleFlyoutKueryBar", {
  enumerable: true,
  get: function () {
    return _rule_kql_filter.RuleFlyoutKueryBar;
  }
});
Object.defineProperty(exports, "WithKueryAutocompletion", {
  enumerable: true,
  get: function () {
    return _rule_kql_filter.WithKueryAutocompletion;
  }
});
Object.defineProperty(exports, "alertsLocatorID", {
  enumerable: true,
  get: function () {
    return _common.alertsLocatorID;
  }
});
Object.defineProperty(exports, "apmEnableTableSearchBar", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.apmEnableTableSearchBar;
  }
});
Object.defineProperty(exports, "apmServiceGroupMaxNumberOfServices", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.apmServiceGroupMaxNumberOfServices;
  }
});
Object.defineProperty(exports, "buildEsQuery", {
  enumerable: true,
  get: function () {
    return _build_es_query.buildEsQuery;
  }
});
Object.defineProperty(exports, "calculateTimeRangeBucketSize", {
  enumerable: true,
  get: function () {
    return _calculate_bucket_size.calculateTimeRangeBucketSize;
  }
});
Object.defineProperty(exports, "convertTo", {
  enumerable: true,
  get: function () {
    return _duration.convertTo;
  }
});
Object.defineProperty(exports, "createObservabilityRuleTypeRegistryMock", {
  enumerable: true,
  get: function () {
    return _observability_rule_type_registry_mock.createObservabilityRuleTypeRegistryMock;
  }
});
Object.defineProperty(exports, "createUseRulesLink", {
  enumerable: true,
  get: function () {
    return _create_use_rules_link.createUseRulesLink;
  }
});
Object.defineProperty(exports, "enableAgentExplorerView", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableAgentExplorerView;
  }
});
Object.defineProperty(exports, "enableComparisonByDefault", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableComparisonByDefault;
  }
});
Object.defineProperty(exports, "enableInfrastructureContainerAssetView", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableInfrastructureContainerAssetView;
  }
});
Object.defineProperty(exports, "enableInfrastructureHostsView", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableInfrastructureHostsView;
  }
});
Object.defineProperty(exports, "enableInspectEsQueries", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableInspectEsQueries;
  }
});
Object.defineProperty(exports, "enableLegacyUptimeApp", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.enableLegacyUptimeApp;
  }
});
Object.defineProperty(exports, "formatAlertEvaluationValue", {
  enumerable: true,
  get: function () {
    return _format_alert_evaluation_value.formatAlertEvaluationValue;
  }
});
Object.defineProperty(exports, "fromQuery", {
  enumerable: true,
  get: function () {
    return _url.fromQuery;
  }
});
Object.defineProperty(exports, "getAlertSummaryTimeRange", {
  enumerable: true,
  get: function () {
    return _alert_summary_widget.getAlertSummaryTimeRange;
  }
});
Object.defineProperty(exports, "getApmTraceUrl", {
  enumerable: true,
  get: function () {
    return _get_apm_trace_url.getApmTraceUrl;
  }
});
Object.defineProperty(exports, "getCoreVitalsComponent", {
  enumerable: true,
  get: function () {
    return _get_core_web_vitals_lazy.getCoreVitalsComponent;
  }
});
Object.defineProperty(exports, "getGroupFilters", {
  enumerable: true,
  get: function () {
    return _get_group.getGroupFilters;
  }
});
Object.defineProperty(exports, "observabilityAppId", {
  enumerable: true,
  get: function () {
    return _common.observabilityAppId;
  }
});
Object.defineProperty(exports, "observabilityFeatureId", {
  enumerable: true,
  get: function () {
    return _common.observabilityFeatureId;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "ruleDetailsLocatorID", {
  enumerable: true,
  get: function () {
    return _common.ruleDetailsLocatorID;
  }
});
Object.defineProperty(exports, "rulesLocatorID", {
  enumerable: true,
  get: function () {
    return _common.rulesLocatorID;
  }
});
Object.defineProperty(exports, "sloDetailsLocatorID", {
  enumerable: true,
  get: function () {
    return _common.sloDetailsLocatorID;
  }
});
Object.defineProperty(exports, "sloEditLocatorID", {
  enumerable: true,
  get: function () {
    return _common.sloEditLocatorID;
  }
});
Object.defineProperty(exports, "syntheticsThrottlingEnabled", {
  enumerable: true,
  get: function () {
    return _ui_settings_keys.syntheticsThrottlingEnabled;
  }
});
Object.defineProperty(exports, "toQuery", {
  enumerable: true,
  get: function () {
    return _url.toQuery;
  }
});
Object.defineProperty(exports, "uptimeOverviewLocatorID", {
  enumerable: true,
  get: function () {
    return _common.uptimeOverviewLocatorID;
  }
});
Object.defineProperty(exports, "useFetchDataViews", {
  enumerable: true,
  get: function () {
    return _use_fetch_data_views.useFetchDataViews;
  }
});
Object.defineProperty(exports, "useSummaryTimeRange", {
  enumerable: true,
  get: function () {
    return _use_summary_time_range.useSummaryTimeRange;
  }
});
Object.defineProperty(exports, "useTimeBuckets", {
  enumerable: true,
  get: function () {
    return _use_time_buckets.useTimeBuckets;
  }
});
var _react = require("react");
var _plugin = require("./plugin");
var _ui_settings_keys = require("../common/ui_settings_keys");
var _common = require("../common");
var _get_core_web_vitals_lazy = require("./pages/overview/components/sections/ux/core_web_vitals/get_core_web_vitals_lazy");
var _get_alert_search_bar_lazy = require("./components/alert_search_bar/get_alert_search_bar_lazy");
var _date_picker = require("./pages/overview/components/date_picker");
var _typings = require("./typings");
Object.keys(_typings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _typings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _typings[key];
    }
  });
});
var _use_fetch_data_views = require("./hooks/use_fetch_data_views");
var _use_time_buckets = require("./hooks/use_time_buckets");
var _create_use_rules_link = require("./hooks/create_use_rules_link");
var _use_summary_time_range = require("./hooks/use_summary_time_range");
var _get_apm_trace_url = require("./utils/get_apm_trace_url");
var _build_es_query = require("./utils/build_es_query");
var _observability_rule_type_registry_mock = require("./rules/observability_rule_type_registry_mock");
var _date_picker_context = require("./context/date_picker_context/date_picker_context");
var _url = require("./utils/url");
var _alert_summary_widget = require("./utils/alert_summary_widget");
var _calculate_bucket_size = require("./pages/overview/helpers/calculate_bucket_size");
var _duration = require("../common/utils/formatters/duration");
var _format_alert_evaluation_value = require("./utils/format_alert_evaluation_value");
var _rule_kql_filter = require("./components/rule_kql_filter");
var _rule_condition_chart = require("./components/rule_condition_chart");
var _get_group = require("../common/custom_threshold_rule/helpers/get_group");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */ // TODO: https://github.com/elastic/kibana/issues/110905
/* eslint-disable @kbn/eslint/no_export_all */
const plugin = initializerContext => {
  return new _plugin.Plugin(initializerContext);
};
exports.plugin = plugin;
const LazyAlertsFlyout = exports.LazyAlertsFlyout = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./components/alerts_flyout/alerts_flyout'))));
