// @kbn/apm-plugin/public/components/shared/charts/transaction_breakdown_chart/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TransactionBreakdownChart = TransactionBreakdownChart;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _use_annotations_context = require("../../../../context/annotations/use_annotations_context");
var _use_transaction_breakdown = require("./use_transaction_breakdown");
var _breakdown_chart = require("../breakdown_chart");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function TransactionBreakdownChart({
  height,
  showAnnotations = true,
  environment,
  kuery
}) {
  const {
    data,
    status
  } = (0, _use_transaction_breakdown.useTransactionBreakdown)({
    environment,
    kuery
  });
  const {
    annotations
  } = (0, _use_annotations_context.useAnnotationsContext)();
  const {
    timeseries
  } = data;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "s",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("h3", null, _i18n.i18n.translate('xpack.apm.transactionBreakdown.chartTitle', {
    defaultMessage: 'Time spent by span type'
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    content: _i18n.i18n.translate('xpack.apm.transactionBreakdown.chartHelp', {
      defaultMessage: 'The average duration of each span type. "app" indicates something was happening within the service. This could mean that the time was spent in application code and not in database or external requests, or that APM agent auto-instrumentation doesn\'t cover the executed code.'
    }),
    position: "right"
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_breakdown_chart.BreakdownChart, {
    fetchStatus: status,
    height: height,
    annotations: annotations,
    showAnnotations: showAnnotations,
    timeseries: timeseries,
    yAxisType: "percentage",
    id: "transactionBreakdownChart"
  }))));
}
