// @kbn/security-solution-plugin/public/management/pages/policy/view/ingest_manager_integration/endpoint_package_custom_extension/components/styled_components.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StyledEuiFlexItem = exports.StyledEuiFlexGridItem = exports.StyledEuiFlexGridGroup = void 0;
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StyledEuiFlexGridGroup = exports.StyledEuiFlexGridGroup = (0, _styledComponents.default)(_eui.EuiFlexGroup)`
  display: grid;
  grid-template-columns: 33% 45% 22%;
  grid-template-areas: 'title summary link';
`;
const StyledEuiFlexGridItem = exports.StyledEuiFlexGridItem = (0, _styledComponents.default)(_eui.EuiFlexItem)`
  grid-area: ${({
  gridarea
}) => gridarea};
  align-items: ${({
  alignitems
}) => alignitems !== null && alignitems !== void 0 ? alignitems : 'center'};
  margin: 0px;
  padding: 12px;
`;
const StyledEuiFlexItem = exports.StyledEuiFlexItem = (0, _styledComponents.default)(_eui.EuiFlexItem)`
  flex-direction: row-reverse;
`;
