// @kbn/unified-doc-viewer-plugin/public/components/doc_viewer_table/legacy/table_row_btn_filter_add.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocViewTableRowBtnFilterAdd = DocViewTableRowBtnFilterAdd;
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function DocViewTableRowBtnFilterAdd({
  onClick,
  disabled = false
}) {
  const tooltipContent = disabled ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "unifiedDocViewer.docViews.table.unindexedFieldsCanNotBeSearchedTooltip",
    defaultMessage: "Unindexed fields or ignored values cannot be searched"
  }) : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "unifiedDocViewer.docViews.table.filterForValueButtonTooltip",
    defaultMessage: "Filter for value"
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: tooltipContent
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    "aria-label": _i18n.i18n.translate('unifiedDocViewer.docViews.table.filterForValueButtonAriaLabel', {
      defaultMessage: 'Filter for value'
    }),
    className: "kbnDocViewer__actionButton",
    "data-test-subj": "addInclusiveFilterButton",
    disabled: disabled,
    onClick: onClick,
    iconType: 'plusInCircle',
    iconSize: 's'
  }));
}
