// @kbn/cases-plugin/common/types/api/user/v1.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SuggestUserProfilesRequestRt = exports.GetCaseUsersResponseRt = void 0;
var rt = _interopRequireWildcard(require("io-ts"));
var _constants = require("../../../constants");
var _schema = require("../../../schema");
var _v = require("../../domain/user/v1");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const GetCaseUsersResponseRt = exports.GetCaseUsersResponseRt = rt.strict({
  assignees: rt.array(_v.UserWithProfileInfoRt),
  unassignedUsers: rt.array(_v.UserWithProfileInfoRt),
  participants: rt.array(_v.UserWithProfileInfoRt),
  reporter: _v.UserWithProfileInfoRt
});
/**
 * User Profiles
 */
const SuggestUserProfilesRequestRt = exports.SuggestUserProfilesRequestRt = rt.intersection([rt.strict({
  name: rt.string,
  owners: rt.array(rt.string)
}), rt.exact(rt.partial({
  size: (0, _schema.limitedNumberSchema)({
    fieldName: 'size',
    min: 1,
    max: _constants.MAX_SUGGESTED_PROFILES
  })
}))]);
