// @kbn/triggers-actions-ui-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ALERT_HISTORY_PREFIX", {
  enumerable: true,
  get: function () {
    return _types.ALERT_HISTORY_PREFIX;
  }
});
Object.defineProperty(exports, "ActionConnectorMode", {
  enumerable: true,
  get: function () {
    return _types.ActionConnectorMode;
  }
});
Object.defineProperty(exports, "ActionForm", {
  enumerable: true,
  get: function () {
    return _action_connector_form.ActionForm;
  }
});
Object.defineProperty(exports, "AlertConditions", {
  enumerable: true,
  get: function () {
    return _sections.AlertConditions;
  }
});
Object.defineProperty(exports, "AlertConditionsGroup", {
  enumerable: true,
  get: function () {
    return _sections.AlertConditionsGroup;
  }
});
Object.defineProperty(exports, "AlertHistoryDefaultIndexName", {
  enumerable: true,
  get: function () {
    return _types.AlertHistoryDefaultIndexName;
  }
});
Object.defineProperty(exports, "AlertHistoryDocumentTemplate", {
  enumerable: true,
  get: function () {
    return _types.AlertHistoryDocumentTemplate;
  }
});
Object.defineProperty(exports, "AlertHistoryEsIndexConnectorId", {
  enumerable: true,
  get: function () {
    return _types.AlertHistoryEsIndexConnectorId;
  }
});
Object.defineProperty(exports, "AlertProvidedActionVariables", {
  enumerable: true,
  get: function () {
    return _lib.AlertProvidedActionVariables;
  }
});
Object.defineProperty(exports, "AlertsTableContext", {
  enumerable: true,
  get: function () {
    return _alerts_table_context.AlertsTableContext;
  }
});
Object.defineProperty(exports, "CreateConnectorFlyout", {
  enumerable: true,
  get: function () {
    return _action_connector_form.CreateConnectorFlyout;
  }
});
Object.defineProperty(exports, "EditConnectorFlyout", {
  enumerable: true,
  get: function () {
    return _action_connector_form.EditConnectorFlyout;
  }
});
Object.defineProperty(exports, "ForLastExpression", {
  enumerable: true,
  get: function () {
    return _common.ForLastExpression;
  }
});
Object.defineProperty(exports, "GroupByExpression", {
  enumerable: true,
  get: function () {
    return _common.GroupByExpression;
  }
});
Object.defineProperty(exports, "JsonEditorWithMessageVariables", {
  enumerable: true,
  get: function () {
    return _components.JsonEditorWithMessageVariables;
  }
});
Object.defineProperty(exports, "JsonFieldWrapper", {
  enumerable: true,
  get: function () {
    return _components.JsonFieldWrapper;
  }
});
Object.defineProperty(exports, "MustacheTextFieldWrapper", {
  enumerable: true,
  get: function () {
    return _components.MustacheTextFieldWrapper;
  }
});
Object.defineProperty(exports, "OfExpression", {
  enumerable: true,
  get: function () {
    return _common.OfExpression;
  }
});
Object.defineProperty(exports, "Plugin", {
  enumerable: true,
  get: function () {
    return _plugin.Plugin;
  }
});
Object.defineProperty(exports, "SectionLoading", {
  enumerable: true,
  get: function () {
    return _components.SectionLoading;
  }
});
Object.defineProperty(exports, "SimpleConnectorForm", {
  enumerable: true,
  get: function () {
    return _components.SimpleConnectorForm;
  }
});
Object.defineProperty(exports, "TIME_UNITS", {
  enumerable: true,
  get: function () {
    return _constants.TIME_UNITS;
  }
});
Object.defineProperty(exports, "TextAreaWithMessageVariables", {
  enumerable: true,
  get: function () {
    return _components.TextAreaWithMessageVariables;
  }
});
Object.defineProperty(exports, "TextFieldWithMessageVariables", {
  enumerable: true,
  get: function () {
    return _components.TextFieldWithMessageVariables;
  }
});
Object.defineProperty(exports, "ThresholdExpression", {
  enumerable: true,
  get: function () {
    return _common.ThresholdExpression;
  }
});
Object.defineProperty(exports, "ValueExpression", {
  enumerable: true,
  get: function () {
    return _common.ValueExpression;
  }
});
Object.defineProperty(exports, "WhenExpression", {
  enumerable: true,
  get: function () {
    return _common.WhenExpression;
  }
});
Object.defineProperty(exports, "builtInAggregationTypes", {
  enumerable: true,
  get: function () {
    return _common.builtInAggregationTypes;
  }
});
Object.defineProperty(exports, "builtInComparators", {
  enumerable: true,
  get: function () {
    return _common.builtInComparators;
  }
});
Object.defineProperty(exports, "builtInGroupByTypes", {
  enumerable: true,
  get: function () {
    return _common.builtInGroupByTypes;
  }
});
Object.defineProperty(exports, "bulkDeleteRules", {
  enumerable: true,
  get: function () {
    return _bulk_delete.bulkDeleteRules;
  }
});
Object.defineProperty(exports, "connectorDeprecatedMessage", {
  enumerable: true,
  get: function () {
    return _common.connectorDeprecatedMessage;
  }
});
Object.defineProperty(exports, "deprecatedMessage", {
  enumerable: true,
  get: function () {
    return _common.deprecatedMessage;
  }
});
Object.defineProperty(exports, "executeAction", {
  enumerable: true,
  get: function () {
    return _lib.executeAction;
  }
});
Object.defineProperty(exports, "firstFieldOption", {
  enumerable: true,
  get: function () {
    return _common.firstFieldOption;
  }
});
Object.defineProperty(exports, "getCategory", {
  enumerable: true,
  get: function () {
    return _helpers.getCategory;
  }
});
Object.defineProperty(exports, "getFields", {
  enumerable: true,
  get: function () {
    return _common.getFields;
  }
});
Object.defineProperty(exports, "getIndexOptions", {
  enumerable: true,
  get: function () {
    return _common.getIndexOptions;
  }
});
exports.getNotifyWhenOptions = void 0;
Object.defineProperty(exports, "getTimeFieldOptions", {
  enumerable: true,
  get: function () {
    return _common.getTimeFieldOptions;
  }
});
Object.defineProperty(exports, "getTimeOptions", {
  enumerable: true,
  get: function () {
    return _common.getTimeOptions;
  }
});
Object.defineProperty(exports, "getTimeUnitLabel", {
  enumerable: true,
  get: function () {
    return _get_time_unit_label.getTimeUnitLabel;
  }
});
Object.defineProperty(exports, "hasMustacheTokens", {
  enumerable: true,
  get: function () {
    return _lib.hasMustacheTokens;
  }
});
Object.defineProperty(exports, "loadActionTypes", {
  enumerable: true,
  get: function () {
    return _connector_types.loadActionTypes;
  }
});
Object.defineProperty(exports, "loadRule", {
  enumerable: true,
  get: function () {
    return _get_rule.loadRule;
  }
});
Object.defineProperty(exports, "loadRuleAggregations", {
  enumerable: true,
  get: function () {
    return _aggregate.loadRuleAggregations;
  }
});
Object.defineProperty(exports, "loadRuleSummary", {
  enumerable: true,
  get: function () {
    return _rule_summary.loadRuleSummary;
  }
});
Object.defineProperty(exports, "loadRuleTypes", {
  enumerable: true,
  get: function () {
    return _rule_types.loadRuleTypes;
  }
});
exports.plugin = plugin;
Object.defineProperty(exports, "suspendedComponentWithProps", {
  enumerable: true,
  get: function () {
    return _suspended_component_with_props.suspendedComponentWithProps;
  }
});
Object.defineProperty(exports, "templateActionVariable", {
  enumerable: true,
  get: function () {
    return _lib.templateActionVariable;
  }
});
Object.defineProperty(exports, "transformRule", {
  enumerable: true,
  get: function () {
    return _common_transformations.transformRule;
  }
});
Object.defineProperty(exports, "updateActionConnector", {
  enumerable: true,
  get: function () {
    return _lib.updateActionConnector;
  }
});
Object.defineProperty(exports, "useBulkUntrackAlerts", {
  enumerable: true,
  get: function () {
    return _use_bulk_untrack_alerts.useBulkUntrackAlerts;
  }
});
Object.defineProperty(exports, "useConnectorContext", {
  enumerable: true,
  get: function () {
    return _use_connector_context.useConnectorContext;
  }
});
Object.defineProperty(exports, "useKibana", {
  enumerable: true,
  get: function () {
    return _common.useKibana;
  }
});
Object.defineProperty(exports, "useLoadRuleTypesQuery", {
  enumerable: true,
  get: function () {
    return _hooks.useLoadRuleTypesQuery;
  }
});
Object.defineProperty(exports, "useSubAction", {
  enumerable: true,
  get: function () {
    return _hooks.useSubAction;
  }
});
Object.defineProperty(exports, "validateActionFilterQuery", {
  enumerable: true,
  get: function () {
    return _value_validators.validateActionFilterQuery;
  }
});
var _plugin = require("./plugin");
var _types = require("./types");
var _use_connector_context = require("./application/context/use_connector_context");
var _action_connector_form = require("./application/sections/action_connector_form");
var _helpers = require("./application/sections/field_browser/helpers");
var _components = require("./application/components");
var _lib = require("./application/lib");
var _sections = require("./application/sections");
var _common = require("./common");
var _hooks = require("./application/hooks");
var _rule_summary = require("./application/lib/rule_api/rule_summary");
var _bulk_delete = require("./application/lib/rule_api/bulk_delete");
var _aggregate = require("./application/lib/rule_api/aggregate");
var _get_rule = require("./application/lib/rule_api/get_rule");
var _rule_types = require("./application/lib/rule_api/rule_types");
var _suspended_component_with_props = require("./application/lib/suspended_component_with_props");
var _connector_types = require("./application/lib/action_connector_api/connector_types");
var _constants = require("./application/constants");
var _get_time_unit_label = require("./common/lib/get_time_unit_label");
var _common_transformations = require("./application/lib/rule_api/common_transformations");
var _value_validators = require("./application/lib/value_validators");
var _use_bulk_untrack_alerts = require("./application/sections/alerts_table/hooks/use_bulk_untrack_alerts");
var _alerts_table_context = require("./application/sections/alerts_table/contexts/alerts_table_context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */ // TODO: https://github.com/elastic/kibana/issues/110895
function plugin(context) {
  return new _plugin.Plugin(context);
}

// TODO remove this import when we expose the Rules tables as a component

const getNotifyWhenOptions = async () => {
  const {
    NOTIFY_WHEN_OPTIONS
  } = await Promise.resolve().then(() => _interopRequireWildcard(require('./application/sections/rule_form/rule_notify_when')));
  return NOTIFY_WHEN_OPTIONS;
};
exports.getNotifyWhenOptions = getNotifyWhenOptions;
