// @kbn/ux-plugin/public/components/app/rum_dashboard/visitor_breakdown_map/map_tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MapToolTip = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _use_layer_list = require("./use_layer_list");
var _translations = require("../translations");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DescriptionItem = (0, _styledComponents.default)(_eui.EuiDescriptionListDescription)`
  &&& {
    width: 25%;
  }
`;
const TitleItem = (0, _styledComponents.default)(_eui.EuiDescriptionListTitle)`
  &&& {
    width: 75%;
  }
`;
function MapToolTipComponent({
  closeTooltip,
  features = [],
  loadFeatureProperties
}) {
  var _features$;
  const {
    id: featureId,
    layerId,
    mbProperties
  } = (_features$ = features[0]) !== null && _features$ !== void 0 ? _features$ : {};
  const [regionName, setRegionName] = (0, _react.useState)(featureId);
  const [pageLoadDuration, setPageLoadDuration] = (0, _react.useState)('');
  const formatPageLoadValue = val => {
    const valInMs = val / 1000;
    if (valInMs > 1000) {
      return (valInMs / 1000).toFixed(2) + ' sec';
    }
    return (valInMs / 1000).toFixed(0) + ' ms';
  };
  (0, _react.useEffect)(() => {
    const loadRegionInfo = async () => {
      if (loadFeatureProperties) {
        const items = await loadFeatureProperties({
          layerId,
          properties: mbProperties
        });
        items.forEach(item => {
          if (item.getPropertyKey() === _use_layer_list.COUNTRY_NAME || item.getPropertyKey() === _use_layer_list.REGION_NAME) {
            setRegionName(item.getRawValue());
          }
          if (item.getPropertyKey() === _use_layer_list.TRANSACTION_DURATION_REGION || item.getPropertyKey() === _use_layer_list.TRANSACTION_DURATION_COUNTRY) {
            setPageLoadDuration(formatPageLoadValue(+item.getRawValue()));
          }
        });
      }
    };
    loadRegionInfo();
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiOutsideClickDetector, {
    onOutsideClick: () => {
      if (closeTooltip != null) {
        closeTooltip();
      }
    }
  }, /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPopoverTitle, null, regionName), /*#__PURE__*/_react.default.createElement(_eui.EuiDescriptionList, {
    type: "column",
    textStyle: "reverse",
    compressed: true,
    style: {
      width: 300
    }
  }, /*#__PURE__*/_react.default.createElement(TitleItem, {
    className: "eui-textNoWrap"
  }, _translations.I18LABELS.avgPageLoadDuration), /*#__PURE__*/_react.default.createElement(DescriptionItem, null, pageLoadDuration))));
}
const MapToolTip = exports.MapToolTip = /*#__PURE__*/_react.default.memo(MapToolTipComponent);
