// @kbn/security-solution-plugin/public/detection_engine/rule_monitoring/components/basic/filters/execution_status_filter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ExecutionStatusFilter = void 0;
var _react = _interopRequireWildcard(require("react"));
var _execution_status_indicator = require("../../indicators/execution_status_indicator");
var _multiselect_filter = require("../multiselect_filter");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ExecutionStatusFilterComponent = ({
  items,
  selectedItems,
  onChange
}) => {
  const renderItem = (0, _react.useCallback)(item => {
    return /*#__PURE__*/_react.default.createElement(_execution_status_indicator.ExecutionStatusIndicator, {
      status: item
    });
  }, []);
  return /*#__PURE__*/_react.default.createElement(_multiselect_filter.MultiselectFilter, {
    dataTestSubj: "ExecutionStatusFilter",
    title: i18n.FILTER_TITLE,
    items: items,
    selectedItems: selectedItems,
    onSelectionChange: onChange,
    renderItem: renderItem
  });
};
const ExecutionStatusFilter = exports.ExecutionStatusFilter = /*#__PURE__*/_react.default.memo(ExecutionStatusFilterComponent);
ExecutionStatusFilter.displayName = 'ExecutionStatusFilter';
