// @kbn/stack-connectors-plugin/public/connector_types/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.XmattersSeverityOptions = exports.XmattersAuthenticationType = exports.SeverityActionOptions = exports.ServerLogLevelOptions = exports.EventActionOptions = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let EventActionOptions = exports.EventActionOptions = /*#__PURE__*/function (EventActionOptions) {
  EventActionOptions["TRIGGER"] = "trigger";
  EventActionOptions["RESOLVE"] = "resolve";
  EventActionOptions["ACKNOWLEDGE"] = "acknowledge";
  return EventActionOptions;
}({});
let SeverityActionOptions = exports.SeverityActionOptions = /*#__PURE__*/function (SeverityActionOptions) {
  SeverityActionOptions["CRITICAL"] = "critical";
  SeverityActionOptions["ERROR"] = "error";
  SeverityActionOptions["WARNING"] = "warning";
  SeverityActionOptions["INFO"] = "info";
  return SeverityActionOptions;
}({});
let ServerLogLevelOptions = exports.ServerLogLevelOptions = /*#__PURE__*/function (ServerLogLevelOptions) {
  ServerLogLevelOptions["TRACE"] = "trace";
  ServerLogLevelOptions["DEBUG"] = "debug";
  ServerLogLevelOptions["INFO"] = "info";
  ServerLogLevelOptions["WARN"] = "warn";
  ServerLogLevelOptions["ERROR"] = "error";
  ServerLogLevelOptions["FATAL"] = "fatal";
  return ServerLogLevelOptions;
}({});
let XmattersSeverityOptions = exports.XmattersSeverityOptions = /*#__PURE__*/function (XmattersSeverityOptions) {
  XmattersSeverityOptions["CRITICAL"] = "critical";
  XmattersSeverityOptions["HIGH"] = "high";
  XmattersSeverityOptions["MEDIUM"] = "medium";
  XmattersSeverityOptions["LOW"] = "low";
  XmattersSeverityOptions["MINIMAL"] = "minimal";
  return XmattersSeverityOptions;
}({});
let XmattersAuthenticationType = exports.XmattersAuthenticationType = /*#__PURE__*/function (XmattersAuthenticationType) {
  XmattersAuthenticationType["Basic"] = "Basic Authentication";
  XmattersAuthenticationType["URL"] = "URL Authentication";
  return XmattersAuthenticationType;
}({});
