// @kbn/core-saved-objects-migration-server-internal/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DocumentMigrator", {
  enumerable: true,
  get: function () {
    return _src.DocumentMigrator;
  }
});
Object.defineProperty(exports, "KibanaMigrator", {
  enumerable: true,
  get: function () {
    return _src.KibanaMigrator;
  }
});
Object.defineProperty(exports, "REMOVED_TYPES", {
  enumerable: true,
  get: function () {
    return _core.REMOVED_TYPES;
  }
});
Object.defineProperty(exports, "addExcludedTypesToBoolQuery", {
  enumerable: true,
  get: function () {
    return _helpers.addExcludedTypesToBoolQuery;
  }
});
Object.defineProperty(exports, "buildActiveMappings", {
  enumerable: true,
  get: function () {
    return _src.buildActiveMappings;
  }
});
Object.defineProperty(exports, "buildTypesMappings", {
  enumerable: true,
  get: function () {
    return _src.buildTypesMappings;
  }
});
Object.defineProperty(exports, "bulkOverwriteTransformedDocuments", {
  enumerable: true,
  get: function () {
    return _actions.bulkOverwriteTransformedDocuments;
  }
});
Object.defineProperty(exports, "calculateExcludeFilters", {
  enumerable: true,
  get: function () {
    return _actions.calculateExcludeFilters;
  }
});
Object.defineProperty(exports, "checkClusterRoutingAllocationEnabled", {
  enumerable: true,
  get: function () {
    return _actions.checkClusterRoutingAllocationEnabled;
  }
});
Object.defineProperty(exports, "checkForUnknownDocs", {
  enumerable: true,
  get: function () {
    return _actions.checkForUnknownDocs;
  }
});
Object.defineProperty(exports, "cloneIndex", {
  enumerable: true,
  get: function () {
    return _actions.cloneIndex;
  }
});
Object.defineProperty(exports, "closePit", {
  enumerable: true,
  get: function () {
    return _actions.closePit;
  }
});
Object.defineProperty(exports, "createBulkDeleteOperationBody", {
  enumerable: true,
  get: function () {
    return _helpers.createBulkDeleteOperationBody;
  }
});
Object.defineProperty(exports, "createBulkIndexOperationTuple", {
  enumerable: true,
  get: function () {
    return _helpers.createBulkIndexOperationTuple;
  }
});
Object.defineProperty(exports, "createIndex", {
  enumerable: true,
  get: function () {
    return _actions.createIndex;
  }
});
Object.defineProperty(exports, "deterministicallyRegenerateObjectId", {
  enumerable: true,
  get: function () {
    return _core.deterministicallyRegenerateObjectId;
  }
});
Object.defineProperty(exports, "fetchIndices", {
  enumerable: true,
  get: function () {
    return _actions.fetchIndices;
  }
});
Object.defineProperty(exports, "getAggregatedTypesDocuments", {
  enumerable: true,
  get: function () {
    return _check_for_unknown_docs.getAggregatedTypesDocuments;
  }
});
Object.defineProperty(exports, "isClusterShardLimitExceeded", {
  enumerable: true,
  get: function () {
    return _es_errors.isClusterShardLimitExceeded;
  }
});
Object.defineProperty(exports, "isIncompatibleMappingException", {
  enumerable: true,
  get: function () {
    return _es_errors.isIncompatibleMappingException;
  }
});
Object.defineProperty(exports, "isIndexNotFoundException", {
  enumerable: true,
  get: function () {
    return _es_errors.isIndexNotFoundException;
  }
});
Object.defineProperty(exports, "isWriteBlockException", {
  enumerable: true,
  get: function () {
    return _es_errors.isWriteBlockException;
  }
});
Object.defineProperty(exports, "openPit", {
  enumerable: true,
  get: function () {
    return _actions.openPit;
  }
});
Object.defineProperty(exports, "pickupUpdatedMappings", {
  enumerable: true,
  get: function () {
    return _actions.pickupUpdatedMappings;
  }
});
Object.defineProperty(exports, "readWithPit", {
  enumerable: true,
  get: function () {
    return _actions.readWithPit;
  }
});
Object.defineProperty(exports, "refreshIndex", {
  enumerable: true,
  get: function () {
    return _actions.refreshIndex;
  }
});
Object.defineProperty(exports, "reindex", {
  enumerable: true,
  get: function () {
    return _actions.reindex;
  }
});
Object.defineProperty(exports, "removeWriteBlock", {
  enumerable: true,
  get: function () {
    return _actions.removeWriteBlock;
  }
});
Object.defineProperty(exports, "setWriteBlock", {
  enumerable: true,
  get: function () {
    return _actions.setWriteBlock;
  }
});
Object.defineProperty(exports, "transformDocs", {
  enumerable: true,
  get: function () {
    return _actions.transformDocs;
  }
});
Object.defineProperty(exports, "updateAliases", {
  enumerable: true,
  get: function () {
    return _actions.updateAliases;
  }
});
Object.defineProperty(exports, "updateAndPickupMappings", {
  enumerable: true,
  get: function () {
    return _actions.updateAndPickupMappings;
  }
});
Object.defineProperty(exports, "updateMappings", {
  enumerable: true,
  get: function () {
    return _actions.updateMappings;
  }
});
Object.defineProperty(exports, "waitForIndexStatus", {
  enumerable: true,
  get: function () {
    return _actions.waitForIndexStatus;
  }
});
Object.defineProperty(exports, "waitForPickupUpdatedMappingsTask", {
  enumerable: true,
  get: function () {
    return _actions.waitForPickupUpdatedMappingsTask;
  }
});
Object.defineProperty(exports, "waitForReindexTask", {
  enumerable: true,
  get: function () {
    return _actions.waitForReindexTask;
  }
});
Object.defineProperty(exports, "waitForTask", {
  enumerable: true,
  get: function () {
    return _actions.waitForTask;
  }
});
var _src = require("./src");
var _check_for_unknown_docs = require("./src/actions/check_for_unknown_docs");
var _helpers = require("./src/model/helpers");
var _actions = require("./src/actions");
var _es_errors = require("./src/actions/es_errors");
var _core = require("./src/core");
