// @kbn/discover-utils/src/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TRUNCATE_MAX_HEIGHT = exports.SORT_DEFAULT_ORDER_SETTING = exports.SHOW_MULTIFIELDS = exports.SHOW_FIELD_STATISTICS = exports.SEARCH_ON_PAGE_LOAD_SETTING = exports.SEARCH_FIELDS_FROM_SOURCE = exports.SEARCH_EMBEDDABLE_TYPE = exports.SAMPLE_SIZE_SETTING = exports.SAMPLE_ROWS_PER_PAGE_SETTING = exports.ROW_HEIGHT_OPTION = exports.MODIFY_COLUMNS_ON_SWITCH = exports.MAX_DOC_FIELDS_DISPLAYED = exports.HIDE_ANNOUNCEMENTS = exports.FIELDS_LIMIT_SETTING = exports.DOC_TABLE_LEGACY = exports.DOC_HIDE_TIME_COLUMN_SETTING = exports.DEFAULT_COLUMNS_SETTING = exports.CONTEXT_TIE_BREAKER_FIELDS_SETTING = exports.CONTEXT_STEP_SETTING = exports.CONTEXT_DEFAULT_SIZE_SETTING = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const CONTEXT_DEFAULT_SIZE_SETTING = exports.CONTEXT_DEFAULT_SIZE_SETTING = 'context:defaultSize';
const CONTEXT_STEP_SETTING = exports.CONTEXT_STEP_SETTING = 'context:step';
const CONTEXT_TIE_BREAKER_FIELDS_SETTING = exports.CONTEXT_TIE_BREAKER_FIELDS_SETTING = 'context:tieBreakerFields';
const DEFAULT_COLUMNS_SETTING = exports.DEFAULT_COLUMNS_SETTING = 'defaultColumns';
const DOC_HIDE_TIME_COLUMN_SETTING = exports.DOC_HIDE_TIME_COLUMN_SETTING = 'doc_table:hideTimeColumn';
const DOC_TABLE_LEGACY = exports.DOC_TABLE_LEGACY = 'doc_table:legacy';
const FIELDS_LIMIT_SETTING = exports.FIELDS_LIMIT_SETTING = 'fields:popularLimit';
const HIDE_ANNOUNCEMENTS = exports.HIDE_ANNOUNCEMENTS = 'hideAnnouncements';
const MAX_DOC_FIELDS_DISPLAYED = exports.MAX_DOC_FIELDS_DISPLAYED = 'discover:maxDocFieldsDisplayed';
const MODIFY_COLUMNS_ON_SWITCH = exports.MODIFY_COLUMNS_ON_SWITCH = 'discover:modifyColumnsOnSwitch';
const ROW_HEIGHT_OPTION = exports.ROW_HEIGHT_OPTION = 'discover:rowHeightOption';
const SAMPLE_ROWS_PER_PAGE_SETTING = exports.SAMPLE_ROWS_PER_PAGE_SETTING = 'discover:sampleRowsPerPage';
const SAMPLE_SIZE_SETTING = exports.SAMPLE_SIZE_SETTING = 'discover:sampleSize';
const SEARCH_EMBEDDABLE_TYPE = exports.SEARCH_EMBEDDABLE_TYPE = 'search';
const SEARCH_FIELDS_FROM_SOURCE = exports.SEARCH_FIELDS_FROM_SOURCE = 'discover:searchFieldsFromSource';
const SEARCH_ON_PAGE_LOAD_SETTING = exports.SEARCH_ON_PAGE_LOAD_SETTING = 'discover:searchOnPageLoad';
const SHOW_FIELD_STATISTICS = exports.SHOW_FIELD_STATISTICS = 'discover:showFieldStatistics';
const SHOW_MULTIFIELDS = exports.SHOW_MULTIFIELDS = 'discover:showMultiFields';
const SORT_DEFAULT_ORDER_SETTING = exports.SORT_DEFAULT_ORDER_SETTING = 'discover:sort:defaultOrder';
const TRUNCATE_MAX_HEIGHT = exports.TRUNCATE_MAX_HEIGHT = 'truncate:maxHeight';
