// @kbn/uptime-plugin/public/legacy_uptime/components/overview/status_panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusPanel = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _monitor = require("../monitor");
var _snapshot = require("./snapshot");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const STATUS_CHART_HEIGHT = '160px';
const StatusPanel = ({}) => /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
  hasBorder: true
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  gutterSize: "l"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: 2
}, /*#__PURE__*/_react.default.createElement(_snapshot.SnapshotComponent, {
  height: STATUS_CHART_HEIGHT
})), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: 10
}, /*#__PURE__*/_react.default.createElement(_monitor.PingHistogram, {
  height: STATUS_CHART_HEIGHT,
  isResponsive: true
}))));
exports.StatusPanel = StatusPanel;
