// @kbn/enterprise-search-plugin/public/applications/enterprise_search_content/components/connector_detail/components/run_from_source_step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RunFromSourceStep = void 0;
var _react = _interopRequireWildcard(require("react"));
var _dedent = _interopRequireDefault(require("dedent"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _searchApiPanels = require("@kbn/search-api-panels");
var _cloud_details = require("../../../../shared/cloud_details/cloud_details");
var _constants = require("../../search_index/connector/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RunFromSourceStep = ({
  apiKeyData,
  connectorId,
  isWaitingForConnector,
  serviceType
}) => {
  const [isOpen, setIsOpen] = _react.default.useState('open');
  (0, _react.useEffect)(() => {
    if (!isWaitingForConnector) {
      setIsOpen('closed');
    }
  }, [isWaitingForConnector]);
  const {
    elasticsearchUrl
  } = (0, _cloud_details.useCloudDetails)();
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "m"
  }, /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('xpack.enterpriseSearch.connectorDeployment.p.addTheFollowingConfigurationLabel', {
    defaultMessage: 'Clone or download the repo to your local machine'
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCode, null, "git clone https://github.com/elastic/connectors"), "\xA0\xA0\xA0", _i18n.i18n.translate('xpack.enterpriseSearch.connectorDeployment.orLabel', {
    defaultMessage: 'or'
  }), "\xA0\xA0\xA0", /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "enterpriseSearchConnectorDeploymentGoToSourceButton",
    iconType: "logoGithub",
    href: "https://github.com/elastic/connectors",
    target: "_blank"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    responsive: false,
    gutterSize: "xs"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, _i18n.i18n.translate('xpack.enterpriseSearch.connectorDeployment.goToSourceButtonLabel', {
    defaultMessage: 'Go to Source'
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "popout"
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
    id: "collapsibleAccordion",
    onToggle: () => setIsOpen(isOpen === 'closed' ? 'open' : 'closed'),
    forceState: isOpen,
    buttonContent: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "s"
    }, /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.enterpriseSearch.connectorDeployment.p.editConfigLabel",
      defaultMessage: "Edit the {configYaml} file and provide the following configuration",
      values: {
        configYaml: /*#__PURE__*/_react.default.createElement(_eui.EuiCode, null, _i18n.i18n.translate('xpack.enterpriseSearch.connectorDeployment.configYamlCodeBlockLabel', {
          defaultMessage: 'config.yml'
        }))
      }
    })))
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_searchApiPanels.CodeBox, {
    showTopBar: false,
    languageType: "yaml",
    codeSnippet: (0, _constants.getConnectorTemplate)({
      apiKeyData,
      connectorData: {
        id: connectorId !== null && connectorId !== void 0 ? connectorId : '',
        service_type: serviceType
      },
      host: elasticsearchUrl
    })
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("p", null, _i18n.i18n.translate('xpack.enterpriseSearch.connectorDeployment.p.compileAndRunLabel', {
    defaultMessage: 'Compile and run'
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_searchApiPanels.CodeBox, {
    showTopBar: false,
    languageType: "bash",
    codeSnippet: (0, _dedent.default)`
                              make install
                              make run
                              `
  })));
};
exports.RunFromSourceStep = RunFromSourceStep;
