// @kbn/security-solution-plugin/public/detection_engine/rule_response_actions/osquery/osquery_investigation_guide_panel.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OsqueryInvestigationGuidePanel = void 0;
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const panelCss = {
  marginBottom: '16px'
};
const flexGroupCss = {
  padding: `0 24px`
};
const OsqueryInvestigationGuidePanel = exports.OsqueryInvestigationGuidePanel = /*#__PURE__*/_react.default.memo(({
  onClick,
  queriesLength
}) => {
  const [hideInvestigationGuideSuggestion, setHideInvestigationGuideSuggestion] = (0, _react.useState)(false);
  const handleClick = (0, _react.useCallback)(() => {
    onClick();
    setHideInvestigationGuideSuggestion(true);
  }, [onClick]);
  if (hideInvestigationGuideSuggestion) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    color: 'primary',
    paddingSize: 'xs',
    css: panelCss
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: 'row',
    alignItems: 'center',
    css: flexGroupCss
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true,
    "data-test-subj": 'osquery-investigation-guide-text'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.securitySolution.responseActionsList.investigationGuideSuggestion",
    defaultMessage: "You have {queriesLength, plural, one {a query} other {queries}} in the investigation guide. Add {queriesLength, plural, one {it as a response action} other {them as response actions}}?",
    values: {
      queriesLength
    }
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
    size: 'm',
    color: 'primary',
    onClick: handleClick,
    "data-test-subj": 'osqueryAddInvestigationGuideQueries'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.securitySolution.responseActionsList.addButton",
    defaultMessage: "Add"
  }))))));
});
OsqueryInvestigationGuidePanel.displayName = 'OsqueryInvestigationGuidePanel';
