// @kbn/infra-plugin/public/components/asset_details/header/flyout_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutHeader = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FlyoutHeader = ({
  title,
  tabs = [],
  rightSideItems = []
}) => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m",
    justifyContent: "spaceBetween",
    direction: "row"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true,
    css: (0, _react2.css)`
            overflow: hidden;
            & h4 {
              text-overflow: ellipsis;
              overflow: hidden;
              white-space: nowrap;
              width: calc(100%);
            }
          `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("h4", null, title))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    css: (0, _react2.css)`
            align-items: flex-start;
            ${(0, _eui.useEuiMinBreakpoint)('m')} {
              align-items: flex-end;
            }
          `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m",
    responsive: false,
    justifyContent: "flexEnd",
    alignItems: "center",
    css: (0, _react2.css)`
              margin-right: ${euiTheme.size.l};
            `
  }, rightSideItems === null || rightSideItems === void 0 ? void 0 : rightSideItems.map((item, index) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    key: index,
    grow: false
  }, item))))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiTabs, {
    bottomBorder: true,
    css: (0, _react2.css)`
          margin-bottom: calc(-1 * (${euiTheme.size.l} + 1px));
        `,
    size: "s"
  }, tabs.map(({
    label,
    ...tab
  }, index) => /*#__PURE__*/_react.default.createElement(_eui.EuiTab, (0, _extends2.default)({
    key: index
  }, tab), label))));
};
exports.FlyoutHeader = FlyoutHeader;
