// @kbn/osquery-plugin/public/shared_components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "getExternalReferenceAttachmentRegular", {
  enumerable: true,
  get: function () {
    return _external_reference.getExternalReferenceAttachmentRegular;
  }
});
Object.defineProperty(exports, "getLazyLiveQueryField", {
  enumerable: true,
  get: function () {
    return _lazy_live_query_field.getLazyLiveQueryField;
  }
});
Object.defineProperty(exports, "getLazyOsqueryAction", {
  enumerable: true,
  get: function () {
    return _lazy_osquery_action.getLazyOsqueryAction;
  }
});
Object.defineProperty(exports, "getLazyOsqueryResponseActionTypeForm", {
  enumerable: true,
  get: function () {
    return _lazy_osquery_action_params_form.getLazyOsqueryResponseActionTypeForm;
  }
});
Object.defineProperty(exports, "getLazyOsqueryResult", {
  enumerable: true,
  get: function () {
    return _lazy_osquery_result.getLazyOsqueryResult;
  }
});
Object.defineProperty(exports, "getLazyOsqueryResults", {
  enumerable: true,
  get: function () {
    return _lazy_osquery_results.getLazyOsqueryResults;
  }
});
Object.defineProperty(exports, "useIsOsqueryAvailableSimple", {
  enumerable: true,
  get: function () {
    return _use_is_osquery_available_simple.useIsOsqueryAvailableSimple;
  }
});
var _lazy_osquery_result = require("./lazy_osquery_result");
var _lazy_osquery_results = require("./lazy_osquery_results");
var _lazy_osquery_action = require("./lazy_osquery_action");
var _lazy_live_query_field = require("./lazy_live_query_field");
var _use_is_osquery_available_simple = require("./osquery_action/use_is_osquery_available_simple");
var _lazy_osquery_action_params_form = require("./lazy_osquery_action_params_form");
var _external_reference = require("./attachments/external_reference");
