// @kbn/cloud-security-posture-plugin/public/pages/benchmarks/test_subjects.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BENCHMARKS_TABLE_DATA_TEST_SUBJ = exports.BENCHMARKS_TABLE_COLUMNS = exports.BENCHMARKS_PAGE_HEADER = exports.ADD_INTEGRATION_TEST_SUBJ = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const BENCHMARKS_PAGE_HEADER = exports.BENCHMARKS_PAGE_HEADER = 'benchmarks-page-header';
const BENCHMARKS_TABLE_DATA_TEST_SUBJ = exports.BENCHMARKS_TABLE_DATA_TEST_SUBJ = 'csp_benchmarks_table';
const ADD_INTEGRATION_TEST_SUBJ = exports.ADD_INTEGRATION_TEST_SUBJ = 'csp_add_integration';
const BENCHMARKS_TABLE_COLUMNS = exports.BENCHMARKS_TABLE_COLUMNS = {
  CIS_NAME: 'benchmark-table-column-cis-name',
  VERSION: 'benchmark-table-column-version',
  APPLICABLE_TO: 'benchmark-table-column-applicable-to',
  EVALUATED: 'benchmark-table-column-evaluated',
  COMPLIANCE: 'benchmark-table-column-compliance'
};
