// @kbn/synthetics-plugin/public/apps/synthetics/components/step_details_page/step_objects/object_count_list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ObjectCountList = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _color_palette = require("./color_palette");
var _use_object_metrics = require("../hooks/use_object_metrics");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ObjectCountList = () => {
  const objectMetrics = (0, _use_object_metrics.useObjectMetrics)();
  const hasAnyThresholdBreach = objectMetrics.items.some(({
    countDelta
  }) => Math.abs(Number(countDelta)) > 5);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement("h3", null, OBJECT_COUNT_LABEL))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, TOTAL_LABEL, ": ", /*#__PURE__*/_react.default.createElement("span", {
    style: {
      fontWeight: 'bold'
    }
  }, " ", objectMetrics.totalObjects)))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement("div", null, objectMetrics.items.map(({
    label,
    mimeType,
    percent,
    count,
    countDelta
  }) => /*#__PURE__*/_react.default.createElement(_react.Fragment, {
    key: mimeType
  }, /*#__PURE__*/_react.default.createElement(_color_palette.ColorPalette, {
    hasAnyThresholdBreach: hasAnyThresholdBreach,
    label: label,
    mimeType: mimeType,
    percent: percent,
    value: String(count),
    valueWidth: 30,
    loading: objectMetrics.loading,
    delta: Number(countDelta)
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  })))));
};
exports.ObjectCountList = ObjectCountList;
const OBJECT_COUNT_LABEL = _i18n.i18n.translate('xpack.synthetics.stepDetails.objectCount', {
  defaultMessage: 'Object count'
});
const TOTAL_LABEL = _i18n.i18n.translate('xpack.synthetics.stepDetails.total', {
  defaultMessage: 'Total'
});
