// @kbn/security-solution-plugin/public/timelines/components/timeline/properties/notes_flyout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NotesFlyout = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _note_cards = require("../../notes/note_cards");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * z-index override is needed because otherwise NotesFlyout appears below
 * Timeline Modal as they both have same z-index of 1000
 */
const NotesFlyoutContainer = (0, _styledComponents.default)(_eui.EuiFlyoutResizable)`
  /*
  * We want the width of flyout to be less than 50% of screen because
  * otherwise it interferes with the delete notes modal
  * */
  width: 30%;
  z-index: ${props => {
  var _eui$euiZFlyout$toFix;
  return ((_eui$euiZFlyout$toFix = props.theme.eui.euiZFlyout.toFixed()) !== null && _eui$euiZFlyout$toFix !== void 0 ? _eui$euiZFlyout$toFix : 1000) + 2;
}} !important;
`;
const NotesFlyout = exports.NotesFlyout = /*#__PURE__*/_react.default.memo(function NotesFlyout(props) {
  const {
    eventId,
    toggleShowAddNote,
    show,
    onClose,
    associateNote,
    notes,
    timelineId,
    onCancel
  } = props;
  const notesFlyoutTitleId = (0, _eui.useGeneratedHtmlId)({
    prefix: 'notesFlyoutTitle'
  });
  if (!show || !eventId) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiOutsideClickDetector, {
    onOutsideClick: onClose
  }, /*#__PURE__*/_react.default.createElement(NotesFlyoutContainer, {
    ownFocus: false,
    className: "timeline-notes-flyout",
    "data-test-subj": "timeline-notes-flyout",
    onClose: onClose,
    "aria-labelledby": notesFlyoutTitleId,
    minWidth: 500,
    maxWidth: 1400
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "m"
  }, /*#__PURE__*/_react.default.createElement("h2", null, i18n.NOTES))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_note_cards.NoteCards, {
    ariaRowindex: 0,
    associateNote: associateNote,
    className: "notes-in-flyout",
    "data-test-subj": "note-cards",
    notes: notes,
    showAddNote: true,
    toggleShowAddNote: toggleShowAddNote,
    eventId: eventId,
    timelineId: timelineId,
    onCancel: onCancel,
    showToggleEventDetailsAction: false
  }))));
});
