// @kbn/core-saved-objects-migration-server-internal/src/zdt/state/create_initial_state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createInitialState = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

/**
 * Create the initial state to be used for the ZDT migrator.
 */
const createInitialState = context => {
  const nodeRoles = getNodeRoles(context.nodeRoles);
  const enabledRoles = new Set(context.migrationConfig.zdt.runOnRoles);
  const runDocumentMigration = nodeRoles.some(role => enabledRoles.has(role));
  const initialState = {
    controlState: 'INIT',
    logs: [],
    retryCount: 0,
    retryDelay: 0,
    skipDocumentMigration: !runDocumentMigration
  };
  return initialState;
};
exports.createInitialState = createInitialState;
const getNodeRoles = roles => {
  return Object.entries(roles).filter(([_, enabled]) => enabled).map(([key]) => key);
};
