// @kbn/infra-plugin/public/components/asset_details/components/section.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Section = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const Section = exports.Section = /*#__PURE__*/_react.default.forwardRef(({
  title,
  closedSectionContent,
  extraAction,
  children,
  collapsible = false,
  ['data-test-subj']: dataTestSubj,
  id,
  initialTriggerValue
}, ref) => {
  const [trigger, setTrigger] = (0, _react.useState)('open');
  (0, _react.useEffect)(() => {
    setTrigger(initialTriggerValue !== null && initialTriggerValue !== void 0 ? initialTriggerValue : 'open');
  }, [initialTriggerValue]);
  const ButtonContent = () => closedSectionContent && trigger === 'closed' ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, title), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, closedSectionContent)) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, title);
  const collapsibleSectionAccordionId = (0, _eui.useGeneratedHtmlId)({
    prefix: id
  });
  const onToggle = isOpen => {
    setTrigger(isOpen ? 'open' : 'closed');
  };
  return collapsible ? /*#__PURE__*/_react.default.createElement("div", {
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
    id: collapsibleSectionAccordionId,
    "data-section-id": id,
    buttonElement: "div",
    element: "fieldset",
    buttonContent: /*#__PURE__*/_react.default.createElement(ButtonContent, null),
    buttonProps: {
      'data-test-subj': dataTestSubj
    },
    paddingSize: "s",
    initialIsOpen: true,
    extraAction: extraAction,
    forceState: trigger,
    onToggle: onToggle,
    "data-section-state": trigger,
    "data-test-subj": "infraAssetDetailsCollapseExpandSection"
  }, children)) : /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    id: collapsibleSectionAccordionId,
    "data-section-id": id,
    "data-test-subj": dataTestSubj,
    gutterSize: "m",
    direction: "column",
    ref: ref
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "m",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: true
  }, title), extraAction && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, extraAction)), _react.default.Children.toArray(children).filter(Boolean).length > 0 ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, children) : null);
});
