// @kbn/infra-plugin/public/components/asset_details/components/services_tooltip_content.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServicesTooltipContent = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _public = require("@kbn/observability-shared-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ServicesTooltipContent = () => {
  const linkProps = (0, _public.useLinkProps)({
    app: 'home',
    hash: '/tutorial/apm'
  });
  return /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.infra.assetDetails.services.tooltip.servicesLabel",
    defaultMessage: "Showing {apmTutorialLink} services detected on this host.",
    values: {
      apmTutorialLink: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        style: {
          display: 'inline-block'
        },
        "data-test-subj": "assetDetailsTooltipApmTutorialLink",
        href: linkProps.href
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.infra.assetDetails.table.services.tooltip.tutorialLink",
        defaultMessage: "APM-instrumented"
      }))
    }
  }));
};
exports.ServicesTooltipContent = ServicesTooltipContent;
