// @kbn/osquery-plugin/common/api/packs/update_packs.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UpdatePacksRequestParams = exports.UpdatePacksRequestBody = exports.SuccessResponse = void 0;
var _zod = require("zod");
var _common_attributes = require("../model/schema/common_attributes.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Update Saved Query Schema
 *   version: 2023-10-31
 */

const UpdatePacksRequestParams = exports.UpdatePacksRequestParams = _zod.z.object({
  id: _common_attributes.PackId.optional()
});
const UpdatePacksRequestBody = exports.UpdatePacksRequestBody = _zod.z.object({
  id: _common_attributes.PackId.optional(),
  description: _common_attributes.DescriptionOrUndefined.optional(),
  enabled: _common_attributes.EnabledOrUndefined.optional(),
  policy_ids: _common_attributes.PolicyIdsOrUndefined.optional(),
  shards: _common_attributes.Shards.optional(),
  queries: _common_attributes.ObjectQueries.optional()
});
const SuccessResponse = exports.SuccessResponse = _zod.z.object({});
