// @kbn/vis-type-vislib-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUsageCollectionStart = exports.setTheme = exports.setI18n = exports.setFormatService = exports.setDataActions = exports.setAnalytics = exports.getUsageCollectionStart = exports.getTheme = exports.getI18n = exports.getFormatService = exports.getDataActions = exports.getAnalytics = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getDataActions, setDataActions] = (0, _public.createGetterSetter)('vislib data.actions');
exports.setDataActions = setDataActions;
exports.getDataActions = getDataActions;
const [getFormatService, setFormatService] = (0, _public.createGetterSetter)('FieldFormats');
exports.setFormatService = setFormatService;
exports.getFormatService = getFormatService;
const [getAnalytics, setAnalytics] = (0, _public.createGetterSetter)('vislib theme service');
exports.setAnalytics = setAnalytics;
exports.getAnalytics = getAnalytics;
const [getI18n, setI18n] = (0, _public.createGetterSetter)('vislib theme service');
exports.setI18n = setI18n;
exports.getI18n = getI18n;
const [getTheme, setTheme] = (0, _public.createGetterSetter)('vislib theme service');
exports.setTheme = setTheme;
exports.getTheme = getTheme;
const [getUsageCollectionStart, setUsageCollectionStart] = (0, _public.createGetterSetter)('UsageCollection', false);
exports.setUsageCollectionStart = setUsageCollectionStart;
exports.getUsageCollectionStart = getUsageCollectionStart;
