// @kbn/cases-plugin/public/components/user_actions/markdown_form.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UserActionMarkdown = void 0;
var _react = _interopRequireWildcard(require("react"));
var _schema = require("./schema");
var _markdown_editor = require("../markdown_editor");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UserActionMarkdownComponent = /*#__PURE__*/(0, _react.forwardRef)(({
  id,
  content,
  caseId,
  isEditable,
  onChangeEditable,
  onSaveContent
}, ref) => {
  const editorRef = (0, _react.useRef)();
  const fieldName = 'content';
  return isEditable ? /*#__PURE__*/_react.default.createElement(_markdown_editor.EditableMarkdown, {
    id: id,
    content: content,
    caseId: caseId,
    editorRef: editorRef,
    onChangeEditable: onChangeEditable,
    onSaveContent: onSaveContent,
    ref: ref,
    fieldName: fieldName,
    formSchema: _schema.schema
  }) : /*#__PURE__*/_react.default.createElement(_markdown_editor.ScrollableMarkdown, {
    content: content
  });
});
UserActionMarkdownComponent.displayName = 'UserActionMarkdownComponent';
const UserActionMarkdown = exports.UserActionMarkdown = /*#__PURE__*/_react.default.memo(UserActionMarkdownComponent);
