// @kbn/security-solution-plugin/public/timelines/components/timeline/body/renderers/system/auth_ssh.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AuthSsh = void 0;
var _react = _interopRequireDefault(require("react"));
var _draggables = require("../../../../../../common/components/draggables");
var _helpers = require("../helpers");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AuthSsh = exports.AuthSsh = /*#__PURE__*/_react.default.memo(({
  contextId,
  eventId,
  isDraggable,
  sshSignature,
  sshMethod
}) => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, sshSignature != null && /*#__PURE__*/_react.default.createElement(_helpers.TokensFlexItem, {
  grow: false,
  component: "span"
}, /*#__PURE__*/_react.default.createElement(_draggables.DraggableBadge, {
  contextId: contextId,
  eventId: eventId,
  field: "system.audit.package.name",
  isDraggable: isDraggable,
  value: sshSignature,
  iconType: "document",
  isAggregatable: true,
  fieldType: "keyword"
})), sshMethod != null && /*#__PURE__*/_react.default.createElement(_helpers.TokensFlexItem, {
  grow: false,
  component: "span"
}, /*#__PURE__*/_react.default.createElement(_draggables.DraggableBadge, {
  contextId: contextId,
  eventId: eventId,
  field: "system.audit.package.version",
  isDraggable: isDraggable,
  value: sshMethod,
  iconType: "document",
  isAggregatable: true,
  fieldType: "keyword"
}))));
AuthSsh.displayName = 'AuthSsh';
