// @kbn/observability-ai-assistant-app-plugin/public/hooks/__storybook_mocks__/use_kibana.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useKibana = useKibana;
var _react = _interopRequireDefault(require("react"));
var _rxjs = require("rxjs");
var _use_chat = require("./use_chat");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ObservabilityAIAssistantMultipaneFlyoutContext = /*#__PURE__*/_react.default.createContext(undefined);
function useKibana() {
  return {
    services: {
      application: {
        navigateToApp: () => {}
      },
      http: {
        basePath: {
          prepend: () => ''
        }
      },
      notifications: {
        toasts: {
          addSuccess: () => {},
          addError: () => {}
        }
      },
      plugins: {
        start: {
          licensing: {
            license$: new _rxjs.Subject()
          },
          observabilityAIAssistant: {
            useChat: _use_chat.useChat,
            ObservabilityAIAssistantMultipaneFlyoutContext
          },
          share: {
            url: {
              locators: {
                get: () => {}
              }
            }
          },
          triggersActionsUi: {
            getAddRuleFlyout: {},
            getAddConnectorFlyout: {}
          }
        }
      },
      uiSettings: {
        get: setting => {
          if (setting === 'dateFormat') {
            return 'MMM D, YYYY HH:mm';
          }
        }
      }
    }
  };
}
