// @kbn/expression-xy-plugin/public/components/legend_color_picker.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LegendColorPickerWrapperContext = exports.LegendColorPickerWrapper = void 0;
var _react = _interopRequireWildcard(require("react"));
var _charts = require("@elastic/charts");
var _eui = require("@elastic/eui");
var _public = require("@kbn/charts-plugin/public");
var _helpers = require("../helpers");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const KEY_CODE_ENTER = 13;
function getAnchorPosition(legendPosition) {
  switch (legendPosition) {
    case _charts.Position.Bottom:
      return 'upCenter';
    case _charts.Position.Top:
      return 'downCenter';
    case _charts.Position.Left:
      return 'rightCenter';
    default:
      return 'leftCenter';
  }
}
const LegendColorPickerWrapperContext = exports.LegendColorPickerWrapperContext = /*#__PURE__*/(0, _react.createContext)(undefined);
const LegendColorPickerWrapper = ({
  anchor,
  color,
  onClose,
  onChange,
  seriesIdentifiers: [seriesIdentifier]
}) => {
  var _getSeriesName, _layer$splitAccessors, _uiState$get;
  const colorPickerWrappingContext = (0, _react.useContext)(LegendColorPickerWrapperContext);
  const handleOutsideClick = (0, _react.useCallback)(() => {
    onClose === null || onClose === void 0 ? void 0 : onClose();
  }, [onClose]);
  if (!colorPickerWrappingContext) {
    return null;
  }
  const {
    legendPosition,
    setColor,
    uiState,
    dataLayers,
    titles,
    formattedDatatables,
    fieldFormats,
    singleTable
  } = colorPickerWrappingContext;
  const {
    layerId
  } = (0, _helpers.getMetaFromSeriesId)(seriesIdentifier.specId);
  const layer = dataLayers.find(dataLayer => dataLayer.layerId === layerId);
  const allYAccessors = dataLayers.flatMap(dataLayer => dataLayer.accessors);
  const seriesName = layer ? ((_getSeriesName = (0, _helpers.getSeriesName)(seriesIdentifier, {
    splitAccessors: (_layer$splitAccessors = layer.splitAccessors) !== null && _layer$splitAccessors !== void 0 ? _layer$splitAccessors : [],
    accessorsCount: singleTable ? allYAccessors.length : layer.accessors.length,
    columns: formattedDatatables[layer.layerId].table.columns,
    splitAccessorsFormats: fieldFormats[layer.layerId].splitSeriesAccessors,
    alreadyFormattedColumns: formattedDatatables[layer.layerId].formattedColumns,
    columnToLabelMap: layer.columnToLabel ? JSON.parse(layer.columnToLabel) : {},
    multipleLayersWithSplits: (0, _helpers.hasMultipleLayersWithSplits)(dataLayers)
  }, titles[layer.layerId])) === null || _getSeriesName === void 0 ? void 0 : _getSeriesName.toString()) || '' : '';
  const overwriteColors = (_uiState$get = uiState === null || uiState === void 0 ? void 0 : uiState.get('vis.colors', {})) !== null && _uiState$get !== void 0 ? _uiState$get : {};
  const colorIsOverwritten = (seriesName.toString() in overwriteColors);
  let keyDownEventOn = false;
  const handleChange = newColor => {
    if (newColor) {
      onChange(newColor);
    }
    setColor(newColor, seriesName);
    // close the popover if no color is applied or the user has clicked a color
    if (!newColor || !keyDownEventOn) {
      onClose();
    }
  };
  const onKeyDown = e => {
    if (e.keyCode === KEY_CODE_ENTER) {
      onClose === null || onClose === void 0 ? void 0 : onClose();
    }
    keyDownEventOn = true;
  };
  return /*#__PURE__*/_react.default.createElement(_eui.EuiOutsideClickDetector, {
    onOutsideClick: handleOutsideClick
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiWrappingPopover, {
    isOpen: true,
    ownFocus: true,
    display: "block",
    button: anchor,
    anchorPosition: getAnchorPosition(legendPosition),
    closePopover: onClose,
    panelPaddingSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_public.ColorPicker, {
    color: (layer === null || layer === void 0 ? void 0 : layer.palette.name) === 'kibana_palette' ? color : color.toLowerCase(),
    onChange: handleChange,
    label: seriesName,
    useLegacyColors: (layer === null || layer === void 0 ? void 0 : layer.palette.name) === 'kibana_palette',
    colorIsOverwritten: colorIsOverwritten,
    onKeyDown: onKeyDown
  })));
};
exports.LegendColorPickerWrapper = LegendColorPickerWrapper;
