// @kbn/ml-plugin/public/application/components/custom_urls/custom_url_editor/interval_time_range_selector.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IntervalTimerangeSelector = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * React component for the form for adding a custom time range.
 */
const IntervalTimerangeSelector = ({
  setAddIntervalTimerange,
  addIntervalTimerange,
  disabled
}) => {
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    "data-test-subj": "mlJobCustomUrlIntervalTimeRangeSwitch",
    disabled: disabled,
    showLabel: true,
    label: _i18n.i18n.translate('xpack.ml.customUrlsEditor.addIntervalTimeRangeSwitchLabel', {
      defaultMessage: 'Add interval time range'
    }),
    checked: addIntervalTimerange,
    onChange: e => setAddIntervalTimerange(e.target.checked)
  }));
};
exports.IntervalTimerangeSelector = IntervalTimerangeSelector;
