// @kbn/maps-plugin/public/classes/styles/vector/components/legend/right_aligned_text.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RightAlignedText = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class RightAlignedText extends _react.Component {
  constructor(...args) {
    super(...args);
    (0, _defineProperty2.default)(this, "_textRef", /*#__PURE__*/_react.default.createRef());
  }
  componentDidMount() {
    this._setWidth();
  }
  componentDidUpdate() {
    this._setWidth();
  }
  _setWidth() {
    if (this._textRef.current) {
      this.props.setWidth(this._textRef.current.getBBox().width);
    }
  }
  render() {
    return /*#__PURE__*/_react.default.createElement("text", {
      ref: this._textRef,
      style: this.props.style,
      textAnchor: "end",
      x: this.props.x,
      y: this.props.y
    }, this.props.value);
  }
}
exports.RightAlignedText = RightAlignedText;
