// @kbn/security-solution-plugin/public/detection_engine/rule_creation_ui/components/description_step/alert_suppression_label.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertSuppressionLabel = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _utils = require("../../../../../common/detection_engine/utils");
var _technical_preview_badge = require("../../../../common/components/technical_preview_badge");
var _use_upselling = require("../../../../common/hooks/use_upselling");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AlertSuppressionLabel = ({
  label,
  ruleType
}) => {
  const alertSuppressionUpsellingMessage = (0, _use_upselling.useUpsellingMessage)('alert_suppression_rule_details');
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ruleType && (0, _utils.isSuppressionRuleInGA)(ruleType) ? label : /*#__PURE__*/_react.default.createElement(_technical_preview_badge.TechnicalPreviewBadge, {
    label: label
  }), alertSuppressionUpsellingMessage && /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    position: "top",
    content: alertSuppressionUpsellingMessage
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: 'warning',
    size: "l",
    color: "#BD271E",
    style: {
      marginLeft: '8px'
    },
    "data-test-subj": "alertSuppressionInsufficientLicensingIcon"
  })));
};
exports.AlertSuppressionLabel = AlertSuppressionLabel;
