// @kbn/charts-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  plugin: true,
  lightenColor: true,
  useActiveCursor: true,
  paletteIds: true,
  ColorSchemas: true,
  vislibColorMaps: true,
  colorSchemas: true,
  getHeatmapColors: true,
  ColorMode: true,
  LabelRotation: true,
  defaultCountLabel: true
};
Object.defineProperty(exports, "ColorMode", {
  enumerable: true,
  get: function () {
    return _common.ColorMode;
  }
});
Object.defineProperty(exports, "ColorSchemas", {
  enumerable: true,
  get: function () {
    return _common.ColorSchemas;
  }
});
Object.defineProperty(exports, "LabelRotation", {
  enumerable: true,
  get: function () {
    return _common.LabelRotation;
  }
});
Object.defineProperty(exports, "colorSchemas", {
  enumerable: true,
  get: function () {
    return _common.colorSchemas;
  }
});
Object.defineProperty(exports, "defaultCountLabel", {
  enumerable: true,
  get: function () {
    return _common.defaultCountLabel;
  }
});
Object.defineProperty(exports, "getHeatmapColors", {
  enumerable: true,
  get: function () {
    return _common.getHeatmapColors;
  }
});
Object.defineProperty(exports, "lightenColor", {
  enumerable: true,
  get: function () {
    return _lighten_color.lightenColor;
  }
});
Object.defineProperty(exports, "paletteIds", {
  enumerable: true,
  get: function () {
    return _common.paletteIds;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "useActiveCursor", {
  enumerable: true,
  get: function () {
    return _active_cursor.useActiveCursor;
  }
});
Object.defineProperty(exports, "vislibColorMaps", {
  enumerable: true,
  get: function () {
    return _common.vislibColorMaps;
  }
});
var _plugin = require("./plugin");
var _static = require("./static");
Object.keys(_static).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _static[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _static[key];
    }
  });
});
var _lighten_color = require("./services/palettes/lighten_color");
var _active_cursor = require("./services/active_cursor");
var _common = require("../common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

// TODO: https://github.com/elastic/kibana/issues/110891
/* eslint-disable @kbn/eslint/no_export_all */

const plugin = () => new _plugin.ChartsPlugin();
exports.plugin = plugin;
