// @kbn/security-plugin-types-server/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "GLOBAL_RESOURCE", {
  enumerable: true,
  get: function () {
    return _authorization.GLOBAL_RESOURCE;
  }
});
Object.defineProperty(exports, "crossClusterApiKeySchema", {
  enumerable: true,
  get: function () {
    return _authentication.crossClusterApiKeySchema;
  }
});
Object.defineProperty(exports, "elasticsearchRoleSchema", {
  enumerable: true,
  get: function () {
    return _authorization.elasticsearchRoleSchema;
  }
});
Object.defineProperty(exports, "getKibanaRoleSchema", {
  enumerable: true,
  get: function () {
    return _authorization.getKibanaRoleSchema;
  }
});
Object.defineProperty(exports, "getRestApiKeyWithKibanaPrivilegesSchema", {
  enumerable: true,
  get: function () {
    return _authentication.getRestApiKeyWithKibanaPrivilegesSchema;
  }
});
Object.defineProperty(exports, "getUpdateRestApiKeyWithKibanaPrivilegesSchema", {
  enumerable: true,
  get: function () {
    return _authentication.getUpdateRestApiKeyWithKibanaPrivilegesSchema;
  }
});
Object.defineProperty(exports, "restApiKeySchema", {
  enumerable: true,
  get: function () {
    return _authentication.restApiKeySchema;
  }
});
Object.defineProperty(exports, "updateCrossClusterApiKeySchema", {
  enumerable: true,
  get: function () {
    return _authentication.updateCrossClusterApiKeySchema;
  }
});
Object.defineProperty(exports, "updateRestApiKeySchema", {
  enumerable: true,
  get: function () {
    return _authentication.updateRestApiKeySchema;
  }
});
var _authentication = require("./src/authentication");
var _authorization = require("./src/authorization");
