// @kbn/security-solution-plugin/public/timelines/components/timeline/body/renderers/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.USER_NAME_FIELD_NAME = exports.SIGNAL_STATUS_FIELD_NAME = exports.SIGNAL_RULE_NAME_FIELD_NAME = exports.SIGNAL_ASSIGNEE_IDS_FIELD_NAME = exports.RULE_REFERENCE_FIELD_NAME = exports.REFERENCE_URL_FIELD_NAME = exports.REASON_FIELD_NAME = exports.QUARANTINED_PATH_FIELD_NAME = exports.MESSAGE_FIELD_NAME = exports.IP_FIELD_TYPE = exports.HOST_NAME_FIELD_NAME = exports.GEO_FIELD_TYPE = exports.EVENT_URL_FIELD_NAME = exports.EVENT_SUMMARY_FIELD_NAME = exports.EVENT_MODULE_FIELD_NAME = exports.DATE_FIELD_TYPE = exports.AGENT_STATUS_FIELD_NAME = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DATE_FIELD_TYPE = exports.DATE_FIELD_TYPE = 'date';
const HOST_NAME_FIELD_NAME = exports.HOST_NAME_FIELD_NAME = 'host.name';
const USER_NAME_FIELD_NAME = exports.USER_NAME_FIELD_NAME = 'user.name';
const IP_FIELD_TYPE = exports.IP_FIELD_TYPE = 'ip';
const GEO_FIELD_TYPE = exports.GEO_FIELD_TYPE = 'geo_point';
const MESSAGE_FIELD_NAME = exports.MESSAGE_FIELD_NAME = 'message';
const EVENT_MODULE_FIELD_NAME = exports.EVENT_MODULE_FIELD_NAME = 'event.module';
const RULE_REFERENCE_FIELD_NAME = exports.RULE_REFERENCE_FIELD_NAME = 'rule.reference';
const REFERENCE_URL_FIELD_NAME = exports.REFERENCE_URL_FIELD_NAME = 'reference.url';
const EVENT_URL_FIELD_NAME = exports.EVENT_URL_FIELD_NAME = 'event.url';
const SIGNAL_RULE_NAME_FIELD_NAME = exports.SIGNAL_RULE_NAME_FIELD_NAME = 'kibana.alert.rule.name';
const SIGNAL_STATUS_FIELD_NAME = exports.SIGNAL_STATUS_FIELD_NAME = 'kibana.alert.workflow_status';
const SIGNAL_ASSIGNEE_IDS_FIELD_NAME = exports.SIGNAL_ASSIGNEE_IDS_FIELD_NAME = 'kibana.alert.workflow_assignee_ids';
const AGENT_STATUS_FIELD_NAME = exports.AGENT_STATUS_FIELD_NAME = 'agent.status';
const QUARANTINED_PATH_FIELD_NAME = exports.QUARANTINED_PATH_FIELD_NAME = 'quarantined.path';
const REASON_FIELD_NAME = exports.REASON_FIELD_NAME = 'kibana.alert.reason';
const EVENT_SUMMARY_FIELD_NAME = exports.EVENT_SUMMARY_FIELD_NAME = 'eventSummary';
