// @kbn/security-solution-features/src/product_features_privileges.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProductFeaturesPrivileges = exports.ProductFeaturesPrivilegeId = void 0;
var _constants = require("./constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let ProductFeaturesPrivilegeId = exports.ProductFeaturesPrivilegeId = /*#__PURE__*/function (ProductFeaturesPrivilegeId) {
  ProductFeaturesPrivilegeId["endpointExceptions"] = "endpoint_exceptions";
  return ProductFeaturesPrivilegeId;
}({});
/**
 * This is the mapping of the privileges that are registered
 * using a different Kibana feature configuration (sub-feature, main feature privilege, etc)
 * in each offering type (ess, serverless)
 */
const ProductFeaturesPrivileges = exports.ProductFeaturesPrivileges = {
  [ProductFeaturesPrivilegeId.endpointExceptions]: {
    all: {
      ui: ['showEndpointExceptions', 'crudEndpointExceptions'],
      api: [`${_constants.APP_ID}-showEndpointExceptions`, `${_constants.APP_ID}-crudEndpointExceptions`]
    },
    read: {
      ui: ['showEndpointExceptions'],
      api: [`${_constants.APP_ID}-showEndpointExceptions`]
    }
  }
};
