// @kbn/monitoring-plugin/common/http_api/elasticsearch/post_elasticsearch_ccr.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.postElasticsearchCcrResponsePayloadRT = exports.postElasticsearchCcrRequestPayloadRT = exports.postElasticsearchCcrRequestParamsRT = exports.CcrShardRT = void 0;
var rt = _interopRequireWildcard(require("io-ts"));
var _shared = require("../shared");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const postElasticsearchCcrRequestParamsRT = exports.postElasticsearchCcrRequestParamsRT = rt.type({
  clusterUuid: _shared.clusterUuidRT
});
const postElasticsearchCcrRequestPayloadRT = exports.postElasticsearchCcrRequestPayloadRT = rt.intersection([rt.partial({
  ccs: _shared.ccsRT
}), rt.type({
  timeRange: _shared.timeRangeRT
})]);
const errorRt = rt.partial({
  error: rt.union([rt.string, rt.undefined])
});
const CcrShardRT = exports.CcrShardRT = rt.type({
  shardId: rt.number,
  error: errorRt,
  opsSynced: rt.number,
  syncLagTime: rt.number,
  syncLagOps: rt.number,
  syncLagOpsLeader: rt.number,
  syncLagOpsFollower: rt.number
});
const postElasticsearchCcrResponsePayloadRT = exports.postElasticsearchCcrResponsePayloadRT = rt.array(rt.type({
  id: rt.string,
  index: rt.string,
  follows: rt.string,
  shards: rt.array(CcrShardRT),
  error: errorRt,
  opsSynced: rt.number,
  syncLagTime: rt.number,
  syncLagOps: rt.number
}));
