// @kbn/osquery-plugin/public/packs/queries/ecs_field_css.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.semicolonWrapperCss = exports.resultComboBoxCss = exports.fieldSpanCss = exports.fieldIconCss = exports.euiSuperSelectCss = exports.descriptionWrapperCss = exports.buttonWrapperCss = exports.ECSFieldWrapperCss = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const resultComboBoxCss = exports.resultComboBoxCss = {
  '&.euiComboBox': {
    position: 'relative',
    left: '-1px',
    '.euiComboBox__inputWrap': {
      borderRadius: '0 6px 6px 0'
    }
  }
};
const euiSuperSelectCss = exports.euiSuperSelectCss = {
  minWidth: '70px',
  borderRadius: '6px 0 0 6px',
  '.euiIcon': {
    padding: 0,
    width: '18px',
    background: 'none'
  }
};
const fieldIconCss = exports.fieldIconCss = {
  width: '32px',
  '> svg': {
    padding: '0 6px !important'
  }
};
const fieldSpanCss = exports.fieldSpanCss = {
  paddingTop: '0 !important',
  paddingBottom: '0 !important',
  paddingLeft: '5px'
};
const descriptionWrapperCss = exports.descriptionWrapperCss = {
  overflow: 'hidden'
};
const semicolonWrapperCss = exports.semicolonWrapperCss = {
  marginTop: '28px'
};

// align the icon to the inputs
const buttonWrapperCss = exports.buttonWrapperCss = {
  marginTop: '28px',
  width: '24px'
};
const ECSFieldWrapperCss = exports.ECSFieldWrapperCss = {
  maxWidth: '100%'
};
