// @kbn/vis-type-timeseries-plugin/public/application/visualizations/constants/chart.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Y_ACCESSOR_INDEXES = exports.Y0_ACCESSOR_INDEXES = exports.X_ACCESSOR_INDEX = exports.STACK_ACCESSORS = exports.STACKED_OPTIONS = exports.GRID_LINE_CONFIG = exports.COLORS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const COLORS = exports.COLORS = {
  LINE_COLOR: 'rgba(105,112,125,0.2)',
  TEXT_COLOR: 'rgba(0,0,0,0.4)',
  TEXT_COLOR_REVERSED: 'rgba(255,255,255,0.5)',
  VALUE_COLOR: 'rgba(0,0,0,0.7)',
  VALUE_COLOR_REVERSED: 'rgba(255,255,255,0.8)'
};
const GRID_LINE_CONFIG = exports.GRID_LINE_CONFIG = {
  stroke: 'rgba(125,125,125,0.1)'
};
const X_ACCESSOR_INDEX = exports.X_ACCESSOR_INDEX = 0;
const STACK_ACCESSORS = exports.STACK_ACCESSORS = [0];
const Y_ACCESSOR_INDEXES = exports.Y_ACCESSOR_INDEXES = [1];
const Y0_ACCESSOR_INDEXES = exports.Y0_ACCESSOR_INDEXES = [2];
const STACKED_OPTIONS = exports.STACKED_OPTIONS = {
  NONE: 'none',
  PERCENT: 'percent',
  STACKED: 'stacked',
  STACKED_WITHIN_SERIES: 'stacked_within_series'
};
