// @kbn/security-plugin/public/management/role_mappings/role_mappings_api_client.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RoleMappingsAPIClient = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class RoleMappingsAPIClient {
  constructor(http) {
    this.http = http;
  }
  async getRoleMappings() {
    return this.http.get(`/internal/security/role_mapping`);
  }
  async getRoleMapping(name) {
    return this.http.get(`/internal/security/role_mapping/${encodeURIComponent(name)}`);
  }
  async saveRoleMapping(roleMapping) {
    const {
      name,
      ...payload
    } = roleMapping;
    return this.http.post(`/internal/security/role_mapping/${encodeURIComponent(name)}`, {
      body: JSON.stringify(payload)
    });
  }
  async deleteRoleMappings(names) {
    return Promise.all(names.map(name => this.http.delete(`/internal/security/role_mapping/${encodeURIComponent(name)}`).then(() => ({
      success: true,
      name
    })).catch(error => ({
      success: false,
      name,
      error
    }))));
  }
}
exports.RoleMappingsAPIClient = RoleMappingsAPIClient;
