// @kbn/security-solution-plugin/public/common/lib/kibana/services.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KibanaServices = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * This class is a singleton that holds references to core Kibana services.
 * It is initialized during the plugin start lifecycle.
 * Use with caution since it is not updated if services are changed after initialization.
 * useKibana hook should be used in React components to access these services.
 */
class KibanaServices {
  static init({
    http,
    application,
    data,
    unifiedSearch,
    kibanaBranch,
    kibanaVersion,
    buildFlavor,
    prebuiltRulesPackageVersion,
    uiSettings,
    notifications,
    expressions,
    savedSearch
  }) {
    this.services = {
      application,
      data,
      http,
      uiSettings,
      unifiedSearch,
      notifications,
      expressions,
      savedSearch
    };
    this.kibanaBranch = kibanaBranch;
    this.kibanaVersion = kibanaVersion;
    this.buildFlavor = buildFlavor;
    this.prebuiltRulesPackageVersion = prebuiltRulesPackageVersion;
  }
  static get() {
    if (!this.services) {
      this.throwUninitializedError();
    }
    return this.services;
  }
  static getKibanaBranch() {
    if (!this.kibanaBranch) {
      this.throwUninitializedError();
    }
    return this.kibanaBranch;
  }
  static getKibanaVersion() {
    if (!this.kibanaVersion) {
      this.throwUninitializedError();
    }
    return this.kibanaVersion;
  }
  static getPrebuiltRulesPackageVersion() {
    return this.prebuiltRulesPackageVersion;
  }
  static getBuildFlavor() {
    if (!this.buildFlavor) {
      this.throwUninitializedError();
    }
    return this.buildFlavor;
  }
  static throwUninitializedError() {
    throw new Error('Kibana services not initialized - are you trying to import this module from outside of the SIEM app?');
  }
}
exports.KibanaServices = KibanaServices;
/**
 * Whether the environment is 'serverless' or 'traditional'
 */
(0, _defineProperty2.default)(KibanaServices, "buildFlavor", void 0);
/**
 * The current Kibana branch. e.g. 'main'
 */
(0, _defineProperty2.default)(KibanaServices, "kibanaBranch", void 0);
/**
 * The current Kibana version. e.g. '8.0.0' or '8.0.0-SNAPSHOT'
 */
(0, _defineProperty2.default)(KibanaServices, "kibanaVersion", void 0);
(0, _defineProperty2.default)(KibanaServices, "prebuiltRulesPackageVersion", void 0);
(0, _defineProperty2.default)(KibanaServices, "services", void 0);
