// @kbn/security-solution-plugin/public/common/components/and_or_badge/rounded_badge_antenna.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RoundedBadgeAntenna = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireWildcard(require("styled-components"));
var _rounded_badge = require("./rounded_badge");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const antennaStyles = (0, _styledComponents.css)`
  background: ${({
  theme
}) => theme.eui.euiColorLightShade};
  position: relative;
  width: 2px;
  &:after {
    background: ${({
  theme
}) => theme.eui.euiColorLightShade};
    content: '';
    height: 8px;
    right: -4px;
    position: absolute;
    width: 10px;
    clip-path: circle();
  }
`;
const TopAntenna = (0, _styledComponents.default)(_eui.EuiFlexItem)`
  ${antennaStyles}
  &:after {
    top: 0;
  }
`;
const BottomAntenna = (0, _styledComponents.default)(_eui.EuiFlexItem)`
  ${antennaStyles}
  &:after {
    bottom: 0;
  }
`;
const RoundedBadgeAntenna = ({
  type
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  className: "andBadgeContainer",
  gutterSize: "none",
  direction: "column",
  alignItems: "center"
}, /*#__PURE__*/_react.default.createElement(TopAntenna, {
  "data-test-subj": "andOrBadgeBarTop",
  grow: 1
}), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_rounded_badge.RoundedBadge, {
  type: type
})), /*#__PURE__*/_react.default.createElement(BottomAntenna, {
  "data-test-subj": "andOrBadgeBarBottom",
  grow: 1
}));
exports.RoundedBadgeAntenna = RoundedBadgeAntenna;
RoundedBadgeAntenna.displayName = 'RoundedBadgeAntenna';
