// @kbn/observability-plugin/public/pages/overview/components/chart_container/chart_container.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ChartContainer = ChartContainer;
var _charts = require("@elastic/charts");
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CHART_HEIGHT = 120;
function ChartContainer({
  isInitialLoad,
  children,
  iconSize = 'l',
  height = CHART_HEIGHT
}) {
  if (isInitialLoad) {
    return /*#__PURE__*/_react.default.createElement("div", {
      "data-test-subj": "loading",
      style: {
        height,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'center'
      }
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingChart, {
      size: iconSize
    }));
  }
  return /*#__PURE__*/_react.default.createElement(_charts.Chart, {
    size: {
      height
    }
  }, children);
}
