// @kbn/console-plugin/public/application/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "sendRequest", {
  enumerable: true,
  get: function () {
    return _use_send_current_request.sendRequest;
  }
});
Object.defineProperty(exports, "useDataInit", {
  enumerable: true,
  get: function () {
    return _use_data_init.useDataInit;
  }
});
Object.defineProperty(exports, "useRestoreRequestFromHistory", {
  enumerable: true,
  get: function () {
    return _use_restore_request_from_history.useRestoreRequestFromHistory;
  }
});
Object.defineProperty(exports, "useSaveCurrentTextObject", {
  enumerable: true,
  get: function () {
    return _use_save_current_text_object.useSaveCurrentTextObject;
  }
});
Object.defineProperty(exports, "useSendCurrentRequest", {
  enumerable: true,
  get: function () {
    return _use_send_current_request.useSendCurrentRequest;
  }
});
Object.defineProperty(exports, "useSetInputEditor", {
  enumerable: true,
  get: function () {
    return _use_set_input_editor.useSetInputEditor;
  }
});
var _use_set_input_editor = require("./use_set_input_editor");
var _use_restore_request_from_history = require("./use_restore_request_from_history");
var _use_send_current_request = require("./use_send_current_request");
var _use_save_current_text_object = require("./use_save_current_text_object");
var _use_data_init = require("./use_data_init");
