// @kbn/security-solution-plugin/public/detections/containers/detection_engine/alerts/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ALERTS_QUERY_NAMES = void 0;
var _constants = require("../../../../../common/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ALERTS_QUERY_NAMES = exports.ALERTS_QUERY_NAMES = {
  ADD_EXCEPTION_FLYOUT: `${_constants.APP_UI_ID} fetchAlerts addExceptionFlyout`,
  ALERTS_COUNT_BY_STATUS: `${_constants.APP_UI_ID} fetchAlerts byRulebyCount`,
  ALERTS_GROUPING: `${_constants.APP_UI_ID} fetchAlerts grouping`,
  BY_ID: `${_constants.APP_UI_ID} fetchAlerts byId`,
  BY_RULE_BY_STATUS: `${_constants.APP_UI_ID} fetchAlerts byRulebyStatus`,
  BY_RULE_ID: `${_constants.APP_UI_ID} fetchAlerts byRuleId`,
  BY_SEVERITY: `${_constants.APP_UI_ID} fetchAlerts bySeverity`,
  BY_STATUS: `${_constants.APP_UI_ID} fetchAlerts byStatus`,
  CASES: `${_constants.APP_UI_ID} fetchAlerts cases`,
  COUNT: `${_constants.APP_UI_ID} fetchAlerts count`,
  HISTOGRAM: `${_constants.APP_UI_ID} fetchAlerts histogram`,
  PREVALENCE: `${_constants.APP_UI_ID} fetchAlerts prevalence`,
  SOC_TRENDS: `${_constants.APP_UI_ID} fetchAlerts socTrends`,
  TREE_MAP: `${_constants.APP_UI_ID} fetchAlerts treeMap`,
  VULNERABLE_HOSTS: `${_constants.APP_UI_ID} fetchAlerts vulnerableHosts`,
  VULNERABLE_USERS: `${_constants.APP_UI_ID} fetchAlerts vulnerableUsers`
};
