// @kbn/reporting-plugin/server/lib/tasks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ExecuteReportTask", {
  enumerable: true,
  get: function () {
    return _execute_report.ExecuteReportTask;
  }
});
exports.TIME_BETWEEN_ATTEMPTS = exports.ReportingTaskStatus = exports.REPORTING_EXECUTE_TYPE = void 0;
var _execute_report = require("./execute_report");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const REPORTING_EXECUTE_TYPE = exports.REPORTING_EXECUTE_TYPE = 'report:execute';
const TIME_BETWEEN_ATTEMPTS = exports.TIME_BETWEEN_ATTEMPTS = 10 * 1000; // 10 seconds
let ReportingTaskStatus = exports.ReportingTaskStatus = /*#__PURE__*/function (ReportingTaskStatus) {
  ReportingTaskStatus["UNINITIALIZED"] = "uninitialized";
  ReportingTaskStatus["INITIALIZED"] = "initialized";
  return ReportingTaskStatus;
}({});
