// @kbn/content-management-plugin/public/content_client/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ContentClient", {
  enumerable: true,
  get: function () {
    return _content_client.ContentClient;
  }
});
Object.defineProperty(exports, "ContentClientProvider", {
  enumerable: true,
  get: function () {
    return _content_client_context.ContentClientProvider;
  }
});
Object.defineProperty(exports, "useContentClient", {
  enumerable: true,
  get: function () {
    return _content_client_context.useContentClient;
  }
});
Object.defineProperty(exports, "useCreateContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client_mutation_hooks.useCreateContentMutation;
  }
});
Object.defineProperty(exports, "useDeleteContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client_mutation_hooks.useDeleteContentMutation;
  }
});
Object.defineProperty(exports, "useGetContentQuery", {
  enumerable: true,
  get: function () {
    return _content_client_query_hooks.useGetContentQuery;
  }
});
Object.defineProperty(exports, "useSearchContentQuery", {
  enumerable: true,
  get: function () {
    return _content_client_query_hooks.useSearchContentQuery;
  }
});
Object.defineProperty(exports, "useUpdateContentMutation", {
  enumerable: true,
  get: function () {
    return _content_client_mutation_hooks.useUpdateContentMutation;
  }
});
var _content_client = require("./content_client");
var _content_client_context = require("./content_client_context");
var _content_client_query_hooks = require("./content_client_query_hooks");
var _content_client_mutation_hooks = require("./content_client_mutation_hooks");
