// @langtrase/trace-attributes/dist/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QdrantDBMethods = exports.PineconeMethods = exports.ChromaDBMethods = exports.OpenAIMethods = exports.Event = void 0;
var Event;
(function (Event) {
    Event["STREAM_START"] = "stream.start";
    Event["STREAM_OUTPUT"] = "stream.output";
    Event["STREAM_END"] = "stream.end";
})(Event || (exports.Event = Event = {}));
var OpenAIMethods;
(function (OpenAIMethods) {
    OpenAIMethods["CHAT_COMPLETION"] = "openai.chat.completions.create";
    OpenAIMethods["IMAGES_GENERATION"] = "openai.images.generate";
    OpenAIMethods["EMBEDDINGS_CREATE"] = "openai.embeddings.create";
})(OpenAIMethods || (exports.OpenAIMethods = OpenAIMethods = {}));
var ChromaDBMethods;
(function (ChromaDBMethods) {
    ChromaDBMethods["ADD"] = "chromadb.collection.add";
    ChromaDBMethods["QUERY"] = "chromadb.collection.query";
    ChromaDBMethods["DELETE"] = "chromadb.collection.delete";
    ChromaDBMethods["PEEK"] = "chromadb.collection.peek";
    ChromaDBMethods["UPDATE"] = "chromadb.collection.update";
    ChromaDBMethods["MODIFY"] = "chromadb.collection.modify";
    ChromaDBMethods["COUNT"] = "chromadb.collection.count";
})(ChromaDBMethods || (exports.ChromaDBMethods = ChromaDBMethods = {}));
var PineconeMethods;
(function (PineconeMethods) {
    PineconeMethods["UPSERT"] = "pinecone.index.upsert";
    PineconeMethods["QUERY"] = "pinecone.index.query";
    PineconeMethods["DELETE_ONE"] = "pinecone.index.deleteOne";
    PineconeMethods["DELETE_MANY"] = "pinecone.index.deleteMany";
    PineconeMethods["DELETE_ALL"] = "pinecone.index.deleteAll";
})(PineconeMethods || (exports.PineconeMethods = PineconeMethods = {}));
var QdrantDBMethods;
(function (QdrantDBMethods) {
    QdrantDBMethods["ADD"] = "qdrantdb.add";
    QdrantDBMethods["GET_COLLECTION"] = "qdrantdb.get_collection";
    QdrantDBMethods["GET_COLLECTIONS"] = "qdrantdb.get_collections";
    QdrantDBMethods["QUERY"] = "qdrantdb.query";
    QdrantDBMethods["QUERY_BATCH"] = "qdrantdb.query_batch";
    QdrantDBMethods["DELETE"] = "qdrantdb.delete";
    QdrantDBMethods["DISCOVER"] = "qdrantdb.discover";
    QdrantDBMethods["DISCOVER_BATCH"] = "qdrantdb.discover_batch";
    QdrantDBMethods["RECOMMEND"] = "qdrantdb.recommend";
    QdrantDBMethods["RECOMMEND_BATCH"] = "qdrantdb.recommend_batch";
    QdrantDBMethods["RETRIEVE"] = "qdrantdb.retrieve";
    QdrantDBMethods["SEARCH"] = "qdrantdb.search";
    QdrantDBMethods["SEARCH_BATCH"] = "qdrantdb.search_batch";
    QdrantDBMethods["UPSERT"] = "qdrantdb.upsert";
    QdrantDBMethods["COUNT"] = "qdrantdb.count";
    QdrantDBMethods["UPDATE_COLLECTION"] = "qdrantdb.update_collection";
    QdrantDBMethods["UPDATE_VECTORS"] = "qdrantdb.update_vectors";
})(QdrantDBMethods || (exports.QdrantDBMethods = QdrantDBMethods = {}));
