// @kbn/maps-plugin/public/classes/layers/wizards/choropleth_layer_wizard/choropleth_layer_wizard.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.choroplethLayerWizardConfig = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _constants = require("../../../../../common/constants");
var _layer_template = require("./layer_template");
var _cloropleth_layer_icon = require("../icons/cloropleth_layer_icon");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const choroplethLayerWizardConfig = exports.choroplethLayerWizardConfig = {
  id: _constants.WIZARD_ID.CHOROPLETH,
  order: 10,
  categories: [_constants.LAYER_WIZARD_CATEGORY.ELASTICSEARCH],
  description: _i18n.i18n.translate('xpack.maps.choropleth.desc', {
    defaultMessage: 'Shade areas to compare statistics across boundaries'
  }),
  icon: _cloropleth_layer_icon.ChoroplethLayerIcon,
  renderWizard: renderWizardArguments => {
    return /*#__PURE__*/_react.default.createElement(_layer_template.LayerTemplate, renderWizardArguments);
  },
  title: _i18n.i18n.translate('xpack.maps.choropleth.title', {
    defaultMessage: 'Choropleth'
  })
};
