// @kbn/security-solution-navigation/src/landing_links/lazy.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LandingLinksImages = exports.LandingLinksImageCards = exports.LandingLinksIconsGroups = exports.LandingLinksIconsCategoriesGroups = exports.LandingLinksIconsCategories = exports.LandingLinksIcons = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const centerSpinnerStyle = {
  display: 'flex',
  margin: 'auto'
};
const withSuspense = Component => function LazyPageWithSuspense(props) {
  return /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
      size: "xl",
      style: centerSpinnerStyle
    })
  }, /*#__PURE__*/_react.default.createElement(Component, props));
};
const LandingLinksIconsCategoriesLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_icons_categories'))));
const LandingLinksIconsCategories = exports.LandingLinksIconsCategories = withSuspense(LandingLinksIconsCategoriesLazy);
const LandingLinksIconsLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_icons'))));
const LandingLinksIcons = exports.LandingLinksIcons = withSuspense(LandingLinksIconsLazy);
const LandingLinksIconsGroupsLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_icons_groups'))));
const LandingLinksIconsGroups = exports.LandingLinksIconsGroups = withSuspense(LandingLinksIconsGroupsLazy);
const LandingLinksImagesLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_images'))));
const LandingLinksImages = exports.LandingLinksImages = withSuspense(LandingLinksImagesLazy);
const LandingLinksImageCardsLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_images_cards'))));
const LandingLinksImageCards = exports.LandingLinksImageCards = withSuspense(LandingLinksImageCardsLazy);
const LandingLinksIconsCategoriesGroupsLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('./landing_links_icons_categories_groups'))));
const LandingLinksIconsCategoriesGroups = exports.LandingLinksIconsCategoriesGroups = withSuspense(LandingLinksIconsCategoriesGroupsLazy);
