// @kbn/security-solution-plugin/public/timelines/components/row_renderers_browser/examples/system_endgame_process.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SystemEndgameProcessExample = void 0;
var _react = _interopRequireDefault(require("react"));
var _generic_row_renderer = require("../../timeline/body/renderers/system/generic_row_renderer");
var _creation = require("../../../../common/demo_data/endgame_ecs/creation");
var _translations = require("../../timeline/body/renderers/system/translations");
var _constants = require("../constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SystemEndgameProcessExampleComponent = () => {
  const systemEndgameProcessRowRenderer = (0, _generic_row_renderer.createEndgameProcessRowRenderer)({
    actionName: 'creation_event',
    text: _translations.PROCESS_STARTED
  });
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, systemEndgameProcessRowRenderer.renderRow({
    data: _creation.demoEndgameCreationEvent,
    isDraggable: false,
    scopeId: _constants.ROW_RENDERER_BROWSER_EXAMPLE_TIMELINE_ID
  }));
};
const SystemEndgameProcessExample = exports.SystemEndgameProcessExample = /*#__PURE__*/_react.default.memo(SystemEndgameProcessExampleComponent);
