// @kbn/security-solution-plugin/public/sourcerer/store/actions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setSourcererScopeLoading = exports.setSourcererDataViews = exports.setSignalIndexName = exports.setSelectedDataView = exports.setDataViewLoading = exports.setDataView = void 0;
var _typescriptFsa = _interopRequireDefault(require("typescript-fsa"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const actionCreator = (0, _typescriptFsa.default)('x-pack/security_solution/local/sourcerer');
const setDataView = exports.setDataView = actionCreator('SET_DATA_VIEW');
const setDataViewLoading = exports.setDataViewLoading = actionCreator('SET_DATA_VIEW_LOADING');
const setSignalIndexName = exports.setSignalIndexName = actionCreator('SET_SIGNAL_INDEX_NAME');
const setSourcererDataViews = exports.setSourcererDataViews = actionCreator('SET_SOURCERER_DATA_VIEWS');
const setSourcererScopeLoading = exports.setSourcererScopeLoading = actionCreator('SET_SOURCERER_SCOPE_LOADING');
const setSelectedDataView = exports.setSelectedDataView = actionCreator('SET_SELECTED_DATA_VIEW');
