// @kbn/inspector-plugin/public/views/requests/components/details/clusters_view/clusters_table/shards_view/shards_view.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ShardsView = ShardsView;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
var _open_shard_failure_flyout_button = require("./open_shard_failure_flyout_button");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function ShardsView({
  failures,
  shardStats
}) {
  return !shardStats && failures.length === 0 ? null : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "xxxs"
  }, /*#__PURE__*/_react.default.createElement("h4", null, _i18n.i18n.translate('inspector.requests.clusters.shards.shardsTitle', {
    defaultMessage: 'Shards'
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_open_shard_failure_flyout_button.OpenShardFailureFlyoutButton, {
    failures: failures
  }))), shardStats ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, _i18n.i18n.translate('inspector.requests.clusters.shards.totalShardsLabel', {
    defaultMessage: '{total} total shards',
    values: {
      total: shardStats.total
    }
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, _i18n.i18n.translate('inspector.requests.clusters.shards.successfulShardsLabel', {
    defaultMessage: '{successful} of {total} successful',
    values: {
      successful: shardStats.successful,
      total: shardStats.total
    }
  }))) : null);
}
