// @kbn/security-solution-plugin/public/detection_engine/rule_exceptions/utils/highlighted_fields_config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.highlightedFieldsPrefixToExclude = exports.getKibanaAlertIdField = exports.KIBANA_ALERT_RULE_UUID = exports.KIBANA_ALERT_RULE_TYPE = exports.EVENT_CODE = exports.EVENT_CATEGORY = exports.ENDPOINT_ALERT = exports.AGENT_TYPE = exports.AGENT_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * The highlightedFieldsPrefixToExclude is an array of prefixes
 * that should be disregarded in the Rule Exception.These prefixes
 * are irrelevant to the exception and should be ignored,even if
 * they were retrieved as Highlighted Fields from the "getEventFieldsToDisplay".
 */
const highlightedFieldsPrefixToExclude = exports.highlightedFieldsPrefixToExclude = ['kibana.alert.rule', 'signal.rule', 'rule'];
const getKibanaAlertIdField = id => `kibana.alert.${id}`;
exports.getKibanaAlertIdField = getKibanaAlertIdField;
const EVENT_CATEGORY = exports.EVENT_CATEGORY = 'event.category';
const EVENT_CODE = exports.EVENT_CODE = 'event.code';
const KIBANA_ALERT_RULE_TYPE = exports.KIBANA_ALERT_RULE_TYPE = 'kibana.alert.rule.type';
const AGENT_ID = exports.AGENT_ID = 'agent.id';
const AGENT_TYPE = exports.AGENT_TYPE = 'agent.type';
const KIBANA_ALERT_RULE_UUID = exports.KIBANA_ALERT_RULE_UUID = 'kibana.alert.rule.uuid';
const ENDPOINT_ALERT = exports.ENDPOINT_ALERT = 'endpoint';
