// @kbn/lists-plugin/server/routes/list_item/create_list_item_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createListItemRoute = void 0;
var _securitysolutionEsUtils = require("@kbn/securitysolution-es-utils");
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
var _zodHelpers = require("@kbn/zod-helpers");
var _api = require("@kbn/securitysolution-lists-common/api");
var _utils = require("../utils");
var _ = require("..");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createListItemRoute = router => {
  router.versioned.post({
    access: 'public',
    options: {
      tags: ['access:lists-all']
    },
    path: _securitysolutionListConstants.LIST_ITEM_URL
  }).addVersion({
    validate: {
      request: {
        body: (0, _zodHelpers.buildRouteValidationWithZod)(_api.CreateListItemRequestBody)
      }
    },
    version: '2023-10-31'
  }, async (context, request, response) => {
    const siemResponse = (0, _utils.buildSiemResponse)(response);
    try {
      const {
        id,
        list_id: listId,
        value,
        meta,
        refresh
      } = request.body;
      const lists = await (0, _.getListClient)(context);
      const list = await lists.getList({
        id: listId
      });
      if (list == null) {
        return siemResponse.error({
          body: `list id: "${listId}" does not exist`,
          statusCode: 404
        });
      } else {
        if (id != null) {
          const listItem = await lists.getListItem({
            id
          });
          if (listItem != null) {
            return siemResponse.error({
              body: `list item id: "${id}" already exists`,
              statusCode: 409
            });
          }
        }
        const createdListItem = await lists.createListItem({
          deserializer: list.deserializer,
          id,
          listId,
          meta,
          refresh,
          serializer: list.serializer,
          type: list.type,
          value
        });
        if (createdListItem != null) {
          return response.ok({
            body: _api.CreateListItemResponse.parse(createdListItem)
          });
        } else {
          return siemResponse.error({
            body: 'list item invalid',
            statusCode: 400
          });
        }
      }
    } catch (err) {
      const error = (0, _securitysolutionEsUtils.transformError)(err);
      return siemResponse.error({
        body: error.message,
        statusCode: error.statusCode
      });
    }
  });
};
exports.createListItemRoute = createListItemRoute;
