// @kbn/lists-plugin/server/routes/list_item/read_list_item_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.readListItemRoute = void 0;
var _securitysolutionEsUtils = require("@kbn/securitysolution-es-utils");
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
var _zodHelpers = require("@kbn/zod-helpers");
var _api = require("@kbn/securitysolution-lists-common/api");
var _utils = require("../utils");
var _ = require("..");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const readListItemRoute = router => {
  router.versioned.get({
    access: 'public',
    options: {
      tags: ['access:lists-read']
    },
    path: _securitysolutionListConstants.LIST_ITEM_URL
  }).addVersion({
    validate: {
      request: {
        query: (0, _zodHelpers.buildRouteValidationWithZod)(_api.GetListItemRequestQuery)
      }
    },
    version: '2023-10-31'
  }, async (context, request, response) => {
    const siemResponse = (0, _utils.buildSiemResponse)(response);
    try {
      const {
        id,
        list_id: listId,
        value
      } = request.query;
      const lists = await (0, _.getListClient)(context);
      if (id != null) {
        const listItem = await lists.getListItem({
          id
        });
        if (listItem == null) {
          return siemResponse.error({
            body: `list item id: "${id}" does not exist`,
            statusCode: 404
          });
        } else {
          return response.ok({
            body: _api.GetListItemResponse.parse(listItem)
          });
        }
      } else if (listId != null && value != null) {
        const list = await lists.getList({
          id: listId
        });
        if (list == null) {
          return siemResponse.error({
            body: `list id: "${listId}" does not exist`,
            statusCode: 404
          });
        } else {
          const listItem = await lists.getListItemByValue({
            listId,
            type: list.type,
            value
          });
          if (listItem.length === 0) {
            return siemResponse.error({
              body: `list_id: "${listId}" item of ${value} does not exist`,
              statusCode: 404
            });
          } else {
            return response.ok({
              body: _api.GetListItemResponse.parse(listItem)
            });
          }
        }
      } else {
        return siemResponse.error({
          body: 'Either "list_id" or "id" needs to be defined in the request',
          statusCode: 400
        });
      }
    } catch (err) {
      const error = (0, _securitysolutionEsUtils.transformError)(err);
      return siemResponse.error({
        body: error.message,
        statusCode: error.statusCode
      });
    }
  });
};
exports.readListItemRoute = readListItemRoute;
