// @kbn/canvas-plugin/public/components/home/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useCloneWorkpad", {
  enumerable: true,
  get: function () {
    return _use_clone_workpad.useCloneWorkpad;
  }
});
Object.defineProperty(exports, "useCreateFromTemplate", {
  enumerable: true,
  get: function () {
    return _use_create_from_template.useCreateFromTemplate;
  }
});
Object.defineProperty(exports, "useCreateWorkpad", {
  enumerable: true,
  get: function () {
    return _use_create_workpad.useCreateWorkpad;
  }
});
Object.defineProperty(exports, "useDeleteWorkpads", {
  enumerable: true,
  get: function () {
    return _use_delete_workpad.useDeleteWorkpads;
  }
});
Object.defineProperty(exports, "useFindTemplates", {
  enumerable: true,
  get: function () {
    return _use_find_templates.useFindTemplates;
  }
});
Object.defineProperty(exports, "useFindWorkpads", {
  enumerable: true,
  get: function () {
    return _use_find_workpad.useFindWorkpads;
  }
});
Object.defineProperty(exports, "useImportWorkpad", {
  enumerable: true,
  get: function () {
    return _use_upload_workpad.useImportWorkpad;
  }
});
var _use_clone_workpad = require("./use_clone_workpad");
var _use_create_workpad = require("./use_create_workpad");
var _use_delete_workpad = require("./use_delete_workpad");
var _use_find_templates = require("./use_find_templates");
var _use_find_workpad = require("./use_find_workpad");
var _use_upload_workpad = require("./use_upload_workpad");
var _use_create_from_template = require("./use_create_from_template");
