// @kbn/monitoring-plugin/public/components/elasticsearch/shard_activity/total_time.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TotalTime = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TotalTime = ({
  startTime,
  totalTime
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.monitoring.elasticsearch.shardActivity.totalTimeTooltip",
      defaultMessage: "Started: {startTime}",
      values: {
        startTime
      }
    }),
    position: "bottom"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, null, totalTime));
};
exports.TotalTime = TotalTime;
