// @kbn/cases-plugin/public/client/ui/get_all_cases_selector_modal.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getAllCasesSelectorModalNoProviderLazy = exports.getAllCasesSelectorModalLazy = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _cases_context = require("../../components/cases_context");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; } /*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
const AllCasesSelectorModalLazy = /*#__PURE__*/(0, _react.lazy)(() => Promise.resolve().then(() => _interopRequireWildcard(require('../../components/all_cases/selector_modal'))));
const getAllCasesSelectorModalLazy = ({
  externalReferenceAttachmentTypeRegistry,
  persistableStateAttachmentTypeRegistry,
  getFilesClient,
  owner,
  permissions,
  hiddenStatuses,
  onRowClick,
  onClose
}) => /*#__PURE__*/_react.default.createElement(_cases_context.CasesProvider, {
  value: {
    externalReferenceAttachmentTypeRegistry,
    persistableStateAttachmentTypeRegistry,
    getFilesClient,
    owner,
    permissions
  }
}, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
  fallback: /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, null)
}, /*#__PURE__*/_react.default.createElement(AllCasesSelectorModalLazy, {
  hiddenStatuses: hiddenStatuses,
  onRowClick: onRowClick,
  onClose: onClose
})));

/**
 * Same as getAllCasesSelectorModalLazy but without injecting the
 * cases provider. to be further refactored https://github.com/elastic/kibana/issues/123183
 */
exports.getAllCasesSelectorModalLazy = getAllCasesSelectorModalLazy;
const getAllCasesSelectorModalNoProviderLazy = ({
  hiddenStatuses,
  onRowClick,
  onClose
}) => /*#__PURE__*/_react.default.createElement(_react.Suspense, {
  fallback: /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, null)
}, /*#__PURE__*/_react.default.createElement(AllCasesSelectorModalLazy, {
  hiddenStatuses: hiddenStatuses,
  onRowClick: onRowClick,
  onClose: onClose
}));
exports.getAllCasesSelectorModalNoProviderLazy = getAllCasesSelectorModalNoProviderLazy;
