// @kbn/stack-connectors-plugin/server/plugin.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StackConnectorsPlugin = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _connector_types = require("./connector_types");
var _routes = require("./routes");
var _experimental_features = require("../common/experimental_features");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class StackConnectorsPlugin {
  constructor(context) {
    (0, _defineProperty2.default)(this, "logger", void 0);
    (0, _defineProperty2.default)(this, "config", void 0);
    (0, _defineProperty2.default)(this, "experimentalFeatures", void 0);
    this.logger = context.logger.get();
    this.config = context.config.get();
    this.experimentalFeatures = (0, _experimental_features.parseExperimentalConfigValue)(this.config.enableExperimental || []);
  }
  setup(core, plugins) {
    const router = core.http.createRouter();
    const {
      actions
    } = plugins;
    (0, _routes.getWellKnownEmailServiceRoute)(router);
    (0, _routes.validSlackApiChannelsRoute)(router, actions.getActionsConfigurationUtilities(), this.logger);
    (0, _connector_types.registerConnectorTypes)({
      actions,
      publicBaseUrl: core.http.basePath.publicBaseUrl,
      experimentalFeatures: this.experimentalFeatures
    });
  }
  start() {}
  stop() {}
}
exports.StackConnectorsPlugin = StackConnectorsPlugin;
