// @kbn/custom-integrations-plugin/public/language_components.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ElasticsearchDotnetClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_dotnet_readme.ElasticsearchDotnetClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchGoClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_go_readme.ElasticsearchGoClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchJavaClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_java_readme.ElasticsearchJavaClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchJsClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_js_readme.ElasticsearchJsClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchPhpClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_php_readme.ElasticsearchPhpClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchPyClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_py_readme.ElasticsearchPyClientReadme;
  }
});
Object.defineProperty(exports, "ElasticsearchRubyClientReadme", {
  enumerable: true,
  get: function () {
    return _elasticsearch_ruby_readme.ElasticsearchRubyClientReadme;
  }
});
Object.defineProperty(exports, "SampleClientReadme", {
  enumerable: true,
  get: function () {
    return _sample_client_readme.SampleClientReadme;
  }
});
var _sample_client_readme = require("./components/fleet_integration/sample/sample_client_readme");
var _elasticsearch_js_readme = require("./components/fleet_integration/elasticsearch_js/elasticsearch_js_readme");
var _elasticsearch_py_readme = require("./components/fleet_integration/elasticsearch_py/elasticsearch_py_readme");
var _elasticsearch_go_readme = require("./components/fleet_integration/elasticsearch_go/elasticsearch_go_readme");
var _elasticsearch_ruby_readme = require("./components/fleet_integration/elasticsearch_ruby/elasticsearch_ruby_readme");
var _elasticsearch_java_readme = require("./components/fleet_integration/elasticsearch_java/elasticsearch_java_readme");
var _elasticsearch_php_readme = require("./components/fleet_integration/elasticsearch_php/elasticsearch_php_readme");
var _elasticsearch_dotnet_readme = require("./components/fleet_integration/elasticsearch_dotnet/elasticsearch_dotnet_readme");
