// @kbn/vis-type-timeseries-plugin/public/services.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.setUsageCollectionStart = exports.setUnifiedSearchStart = exports.setUISettings = exports.setI18n = exports.setFieldFormats = exports.setDataViewsStart = exports.setDataStart = exports.setCoreStart = exports.setCharts = exports.getUsageCollectionStart = exports.getUnifiedSearchStart = exports.getUISettings = exports.getI18n = exports.getFieldFormats = exports.getDataViewsStart = exports.getDataStart = exports.getCoreStart = exports.getCharts = void 0;
var _public = require("@kbn/kibana-utils-plugin/public");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const [getUISettings, setUISettings] = (0, _public.createGetterSetter)('UISettings');
exports.setUISettings = setUISettings;
exports.getUISettings = getUISettings;
const [getFieldFormats, setFieldFormats] = (0, _public.createGetterSetter)('FieldFormats');
exports.setFieldFormats = setFieldFormats;
exports.getFieldFormats = getFieldFormats;
const [getCoreStart, setCoreStart] = (0, _public.createGetterSetter)('CoreStart');
exports.setCoreStart = setCoreStart;
exports.getCoreStart = getCoreStart;
const [getDataStart, setDataStart] = (0, _public.createGetterSetter)('DataStart');
exports.setDataStart = setDataStart;
exports.getDataStart = getDataStart;
const [getUnifiedSearchStart, setUnifiedSearchStart] = (0, _public.createGetterSetter)('unifiedSearchStart');
exports.setUnifiedSearchStart = setUnifiedSearchStart;
exports.getUnifiedSearchStart = getUnifiedSearchStart;
const [getDataViewsStart, setDataViewsStart] = (0, _public.createGetterSetter)('dataViews');
exports.setDataViewsStart = setDataViewsStart;
exports.getDataViewsStart = getDataViewsStart;
const [getI18n, setI18n] = (0, _public.createGetterSetter)('I18n');
exports.setI18n = setI18n;
exports.getI18n = getI18n;
const [getCharts, setCharts] = (0, _public.createGetterSetter)('ChartsPluginStart');
exports.setCharts = setCharts;
exports.getCharts = getCharts;
const [getUsageCollectionStart, setUsageCollectionStart] = (0, _public.createGetterSetter)('UsageCollection');
exports.setUsageCollectionStart = setUsageCollectionStart;
exports.getUsageCollectionStart = getUsageCollectionStart;
