// @kbn/security-solution-plugin/public/timelines/components/side_panel/network_details/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NetworkDetailsPanel = void 0;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _expandable_network = require("./expandable_network");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StyledEuiFlyoutBody = (0, _styledComponents.default)(_eui.EuiFlyoutBody)`
  .euiFlyoutBody__overflow {
    display: flex;
    flex: 1;
    overflow-x: hidden;
    overflow-y: scroll;

    .euiFlyoutBody__overflowContent {
      flex: 1;
      overflow-x: hidden;
      overflow-y: scroll;
      padding: ${({
  theme
}) => `${theme.eui.euiSizeXS} ${theme.eui.euiSizeM} 64px`};
    }
  }
`;
const StyledEuiFlexGroup = (0, _styledComponents.default)(_eui.EuiFlexGroup)`
  flex: 1 0 auto;
`;
const StyledEuiFlexButtonWrapper = (0, _styledComponents.default)(_eui.EuiFlexItem)`
  align-self: flex-start;
  flex: 1 0 auto;
`;
const StyledPanelContent = _styledComponents.default.div`
  display: block;
  height: 100%;
  overflow-y: scroll;
  overflow-x: hidden;
`;
// eslint-disable-next-line react/display-name
const NetworkDetailsPanel = exports.NetworkDetailsPanel = /*#__PURE__*/_react.default.memo(({
  contextID,
  expandedNetwork,
  handleOnNetworkClosed,
  isFlyoutView,
  isDraggable
}) => {
  const {
    ip
  } = expandedNetwork;
  return isFlyoutView ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: true
  }, /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetailsTitle, {
    ip: ip
  })), /*#__PURE__*/_react.default.createElement(StyledEuiFlyoutBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetailsPageLink, {
    expandedNetwork: expandedNetwork
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetails, {
    contextID: contextID,
    expandedNetwork: expandedNetwork
  }))) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(StyledEuiFlexGroup, {
    justifyContent: "spaceBetween",
    wrap: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetailsTitle, {
    ip: ip
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButtonIcon, {
    iconType: "cross",
    "aria-label": _i18n.i18n.translate('xpack.securitySolution.timeline.sidePanel.networkDetails.close', {
      defaultMessage: 'close'
    }),
    onClick: handleOnNetworkClosed
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(StyledEuiFlexButtonWrapper, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetailsPageLink, {
    expandedNetwork: expandedNetwork
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(StyledPanelContent, null, /*#__PURE__*/_react.default.createElement(_expandable_network.ExpandableNetworkDetails, {
    contextID: contextID,
    expandedNetwork: expandedNetwork,
    isDraggable: isDraggable
  })));
});
