// @kbn/visualizations-plugin/public/embeddable/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.COMMON_VISUALIZATION_GROUPING = void 0;
Object.defineProperty(exports, "VISUALIZE_EMBEDDABLE_TYPE", {
  enumerable: true,
  get: function () {
    return _constants.VISUALIZE_EMBEDDABLE_TYPE;
  }
});
var _i18n = require("@kbn/i18n");
var _constants = require("../../common/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const COMMON_VISUALIZATION_GROUPING = exports.COMMON_VISUALIZATION_GROUPING = [{
  id: 'visualizations',
  getDisplayName: () => _i18n.i18n.translate('visualizations.common.constants.grouping.legacy', {
    defaultMessage: 'Visualizations'
  }),
  getIconType: () => {
    return 'visGauge';
  },
  order: 1000
}];
