// @kbn/discover-plugin/public/embeddable/saved_search_grid.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DiscoverGridEmbeddable = DiscoverGridEmbeddable;
exports.DiscoverGridMemoized = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _discoverUtils = require("@kbn/discover-utils");
var _unifiedDataTable = require("@kbn/unified-data-table");
var _discover_grid = require("../components/discover_grid");
require("./saved_search_grid.scss");
var _discover_grid_flyout = require("../components/discover_grid_flyout");
var _saved_search_embeddable_base = require("./saved_search_embeddable_base");
var _total_documents = require("../application/main/components/total_documents/total_documents");
var _context_awareness = require("../context_awareness");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const DiscoverGridMemoized = exports.DiscoverGridMemoized = /*#__PURE__*/_react.default.memo(_discover_grid.DiscoverGrid);
function DiscoverGridEmbeddable(props) {
  const {
    interceptedWarnings,
    ...gridProps
  } = props;
  const [expandedDoc, setExpandedDoc] = (0, _react.useState)(undefined);
  const renderDocumentView = (0, _react.useCallback)((hit, displayedRows, displayedColumns, customColumnsMeta) => /*#__PURE__*/_react.default.createElement(_discover_grid_flyout.DiscoverGridFlyout, {
    dataView: props.dataView,
    hit: hit,
    hits: displayedRows
    // if default columns are used, dont make them part of the URL - the context state handling will take care to restore them
    ,
    columns: displayedColumns,
    columnsMeta: customColumnsMeta,
    savedSearchId: props.savedSearchId,
    onFilter: props.onFilter,
    onRemoveColumn: props.onRemoveColumn,
    onAddColumn: props.onAddColumn,
    onClose: () => setExpandedDoc(undefined),
    setExpandedDoc: setExpandedDoc,
    query: props.query
  }), [props.dataView, props.onAddColumn, props.onFilter, props.onRemoveColumn, props.query, props.savedSearchId]);
  const renderCustomToolbarWithElements = (0, _react.useMemo)(() => (0, _unifiedDataTable.getRenderCustomToolbarWithElements)({
    leftSide: typeof props.totalHitCount === 'number' ? /*#__PURE__*/_react.default.createElement(_total_documents.TotalDocuments, {
      totalHitCount: props.totalHitCount
    }) : undefined
  }), [props.totalHitCount]);
  const getCellRenderersAccessor = (0, _context_awareness.useProfileAccessor)('getCellRenderers');
  const cellRenderers = (0, _react.useMemo)(() => {
    const getCellRenderers = getCellRenderersAccessor(() => ({}));
    return getCellRenderers();
  }, [getCellRenderersAccessor]);
  return /*#__PURE__*/_react.default.createElement(_saved_search_embeddable_base.SavedSearchEmbeddableBase, {
    totalHitCount: undefined // it will be rendered inside the custom grid toolbar instead
    ,
    isLoading: props.loadingState === _unifiedDataTable.DataLoadingState.loading,
    dataTestSubj: "embeddedSavedSearchDocTable",
    interceptedWarnings: props.interceptedWarnings
  }, /*#__PURE__*/_react.default.createElement(DiscoverGridMemoized, (0, _extends2.default)({}, gridProps, {
    totalHits: props.totalHitCount,
    setExpandedDoc: setExpandedDoc,
    expandedDoc: expandedDoc,
    configRowHeight: props.services.uiSettings.get(_discoverUtils.ROW_HEIGHT_OPTION),
    showMultiFields: props.services.uiSettings.get(_discoverUtils.SHOW_MULTIFIELDS),
    maxDocFieldsDisplayed: props.services.uiSettings.get(_discoverUtils.MAX_DOC_FIELDS_DISPLAYED),
    renderDocumentView: renderDocumentView,
    renderCustomToolbar: renderCustomToolbarWithElements,
    externalCustomRenderers: cellRenderers,
    enableComparisonMode: true,
    showColumnTokens: true,
    configHeaderRowHeight: 3,
    showFullScreenButton: false,
    className: "unifiedDataTable"
  })));
}
