// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/configs.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProductTier = exports.ProductLine = exports.ALL_PRODUCT_LINES = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
/**
 * We do not want serverless code in this plugin. Please Do Not Reuse.
 * This is temporary, will be deleted when issue-174766 is resolved.
 */
let ProductLine = exports.ProductLine = /*#__PURE__*/function (ProductLine) {
  ProductLine["security"] = "security";
  ProductLine["endpoint"] = "endpoint";
  ProductLine["cloud"] = "cloud";
  return ProductLine;
}({});
/**
 * We do not want serverless code in this plugin. Please Do Not Reuse.
 * This is temporary, will be deleted when issue-174766 is resolved.
 */
let ProductTier = exports.ProductTier = /*#__PURE__*/function (ProductTier) {
  ProductTier["essentials"] = "essentials";
  ProductTier["complete"] = "complete";
  return ProductTier;
}({});
/**
 * We do not want serverless code in this plugin. Please Do Not Reuse.
 * This is temporary, will be deleted when issue-174766 is resolved.
 */
/**
 * We do not want serverless code in this plugin. Please Do Not Reuse.
 * This is temporary, will be deleted when issue-174766 is resolved.
 */
const ALL_PRODUCT_LINES = exports.ALL_PRODUCT_LINES = Object.values(ProductLine);
