// @kbn/security-solution-plugin/public/common/components/landing_page/onboarding/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ViewDashboardSteps = exports.ViewAlertsSteps = exports.SectionId = exports.QuickStartSectionCardsId = exports.OverviewSteps = exports.OnboardingActions = exports.GetStartedWithAlertsCardsId = exports.EnablePrebuiltRulesSteps = exports.CreateProjectSteps = exports.BadgeId = exports.AddIntegrationsSteps = exports.AddAndValidateYourDataCardsId = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let SectionId = exports.SectionId = /*#__PURE__*/function (SectionId) {
  SectionId["quickStart"] = "quick_start";
  SectionId["addAndValidateYourData"] = "add_and_validate_your_data";
  SectionId["getStartedWithAlerts"] = "get_started_with_alerts";
  return SectionId;
}({});
let QuickStartSectionCardsId = exports.QuickStartSectionCardsId = /*#__PURE__*/function (QuickStartSectionCardsId) {
  QuickStartSectionCardsId["createFirstProject"] = "create_first_project";
  QuickStartSectionCardsId["watchTheOverviewVideo"] = "watch_the_overview_video";
  return QuickStartSectionCardsId;
}({});
let AddAndValidateYourDataCardsId = exports.AddAndValidateYourDataCardsId = /*#__PURE__*/function (AddAndValidateYourDataCardsId) {
  AddAndValidateYourDataCardsId["addIntegrations"] = "add_integrations";
  AddAndValidateYourDataCardsId["viewDashboards"] = "view_dashboards";
  return AddAndValidateYourDataCardsId;
}({});
let GetStartedWithAlertsCardsId = exports.GetStartedWithAlertsCardsId = /*#__PURE__*/function (GetStartedWithAlertsCardsId) {
  GetStartedWithAlertsCardsId["enablePrebuiltRules"] = "enable_prebuilt_rules";
  GetStartedWithAlertsCardsId["viewAlerts"] = "view_alerts";
  return GetStartedWithAlertsCardsId;
}({});
let BadgeId = exports.BadgeId = /*#__PURE__*/function (BadgeId) {
  BadgeId["analytics"] = "analytics";
  BadgeId["cloud"] = "cloud";
  BadgeId["edr"] = "edr";
  return BadgeId;
}({});
let CreateProjectSteps = exports.CreateProjectSteps = /*#__PURE__*/function (CreateProjectSteps) {
  CreateProjectSteps["createFirstProject"] = "create_your_first_project";
  return CreateProjectSteps;
}({});
let OverviewSteps = exports.OverviewSteps = /*#__PURE__*/function (OverviewSteps) {
  OverviewSteps["getToKnowElasticSecurity"] = "watch_the_overview_video";
  return OverviewSteps;
}({});
let AddIntegrationsSteps = exports.AddIntegrationsSteps = /*#__PURE__*/function (AddIntegrationsSteps) {
  AddIntegrationsSteps["connectToDataSources"] = "add_integrations";
  return AddIntegrationsSteps;
}({});
let ViewDashboardSteps = exports.ViewDashboardSteps = /*#__PURE__*/function (ViewDashboardSteps) {
  ViewDashboardSteps["analyzeData"] = "view_and_analyze_your_data";
  return ViewDashboardSteps;
}({});
let EnablePrebuiltRulesSteps = exports.EnablePrebuiltRulesSteps = /*#__PURE__*/function (EnablePrebuiltRulesSteps) {
  EnablePrebuiltRulesSteps["enablePrebuiltRules"] = "enable_prebuilt_rules";
  return EnablePrebuiltRulesSteps;
}({});
let ViewAlertsSteps = exports.ViewAlertsSteps = /*#__PURE__*/function (ViewAlertsSteps) {
  ViewAlertsSteps["viewAlerts"] = "view_alerts";
  return ViewAlertsSteps;
}({});
let OnboardingActions = exports.OnboardingActions = /*#__PURE__*/function (OnboardingActions) {
  OnboardingActions["AddFinishedStep"] = "addFinishedStep";
  OnboardingActions["RemoveFinishedStep"] = "removeFinishedStep";
  OnboardingActions["ToggleProduct"] = "toggleProduct";
  OnboardingActions["ToggleExpandedStep"] = "toggleExpandedStep";
  return OnboardingActions;
}({});
