// @kbn/unified-histogram-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "UnifiedBreakdownFieldSelector", {
  enumerable: true,
  get: function () {
    return _lazy.UnifiedBreakdownFieldSelector;
  }
});
Object.defineProperty(exports, "UnifiedHistogramContainer", {
  enumerable: true,
  get: function () {
    return _container.UnifiedHistogramContainer;
  }
});
Object.defineProperty(exports, "UnifiedHistogramExternalVisContextStatus", {
  enumerable: true,
  get: function () {
    return _types.UnifiedHistogramExternalVisContextStatus;
  }
});
Object.defineProperty(exports, "UnifiedHistogramFetchStatus", {
  enumerable: true,
  get: function () {
    return _types.UnifiedHistogramFetchStatus;
  }
});
Object.defineProperty(exports, "canImportVisContext", {
  enumerable: true,
  get: function () {
    return _external_vis_context.canImportVisContext;
  }
});
Object.defineProperty(exports, "fieldSupportsBreakdown", {
  enumerable: true,
  get: function () {
    return _field_supports_breakdown.fieldSupportsBreakdown;
  }
});
Object.defineProperty(exports, "getBreakdownField", {
  enumerable: true,
  get: function () {
    return _container.getBreakdownField;
  }
});
Object.defineProperty(exports, "getChartHidden", {
  enumerable: true,
  get: function () {
    return _container.getChartHidden;
  }
});
Object.defineProperty(exports, "getTopPanelHeight", {
  enumerable: true,
  get: function () {
    return _container.getTopPanelHeight;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "setBreakdownField", {
  enumerable: true,
  get: function () {
    return _container.setBreakdownField;
  }
});
Object.defineProperty(exports, "setChartHidden", {
  enumerable: true,
  get: function () {
    return _container.setChartHidden;
  }
});
Object.defineProperty(exports, "setTopPanelHeight", {
  enumerable: true,
  get: function () {
    return _container.setTopPanelHeight;
  }
});
var _plugin = require("./plugin");
var _lazy = require("./chart/lazy");
var _container = require("./container");
var _types = require("./types");
var _external_vis_context = require("./utils/external_vis_context");
var _field_supports_breakdown = require("./utils/field_supports_breakdown");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const plugin = () => new _plugin.UnifiedHistogramPublicPlugin();
exports.plugin = plugin;
