// @kbn/security-solution-plugin/public/timelines/components/side_panel/event_details/flyout/header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutHeaderContent = exports.FlyoutHeader = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _expandable_event = require("../expandable_event");
var _back_to_alert_details_link = require("./back_to_alert_details_link");
var _constants = require("../../../../../../common/endpoint/service/response_actions/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FlyoutHeaderContentComponent = ({
  eventId,
  eventIndex,
  isAlert,
  isHostIsolationPanelOpen,
  isolateAction,
  loading,
  promptContextId,
  ruleName,
  showAlertDetails,
  timestamp,
  scopeId,
  refetchFlyoutData,
  getFieldsData
}) => {
  const isSentinelOneAlert = (0, _react.useMemo)(() => {
    var _getFieldsData;
    return !!(isAlert && (_getFieldsData = getFieldsData(_constants.RESPONSE_ACTIONS_ALERT_AGENT_ID_FIELD.sentinel_one)) !== null && _getFieldsData !== void 0 && _getFieldsData.length);
  }, [getFieldsData, isAlert]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, isHostIsolationPanelOpen ? /*#__PURE__*/_react.default.createElement(_back_to_alert_details_link.BackToAlertDetailsLink, {
    isolateAction: isolateAction,
    showAlertDetails: showAlertDetails,
    showExperimentalBadge: isSentinelOneAlert
  }) : /*#__PURE__*/_react.default.createElement(_expandable_event.ExpandableEventTitle, {
    eventId: eventId,
    eventIndex: eventIndex,
    isAlert: isAlert,
    loading: loading,
    promptContextId: promptContextId,
    ruleName: ruleName,
    timestamp: timestamp,
    scopeId: scopeId,
    refetchFlyoutData: refetchFlyoutData,
    getFieldsData: getFieldsData
  }));
};
const FlyoutHeaderContent = exports.FlyoutHeaderContent = /*#__PURE__*/_react.default.memo(FlyoutHeaderContentComponent);
const FlyoutHeaderComponent = ({
  eventId,
  eventIndex,
  isAlert,
  isHostIsolationPanelOpen,
  isolateAction,
  loading,
  promptContextId,
  ruleName,
  showAlertDetails,
  timestamp,
  scopeId,
  refetchFlyoutData,
  getFieldsData
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: isHostIsolationPanelOpen
  }, /*#__PURE__*/_react.default.createElement(FlyoutHeaderContentComponent, {
    eventId: eventId,
    eventIndex: eventIndex,
    isAlert: isAlert,
    isHostIsolationPanelOpen: isHostIsolationPanelOpen,
    isolateAction: isolateAction,
    loading: loading,
    promptContextId: promptContextId,
    ruleName: ruleName,
    showAlertDetails: showAlertDetails,
    timestamp: timestamp,
    scopeId: scopeId,
    refetchFlyoutData: refetchFlyoutData,
    getFieldsData: getFieldsData
  }));
};
const FlyoutHeader = exports.FlyoutHeader = /*#__PURE__*/_react.default.memo(FlyoutHeaderComponent);
