// @kbn/ml-plugin/public/application/components/ml_inference/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EDIT_MESSAGE = exports.CREATE_FIELD_MAPPING_MESSAGE = exports.CONTINUE_BUTTON_LABEL = exports.CLOSE_BUTTON_LABEL = exports.CLEAR_BUTTON_LABEL = exports.CANCEL_EDIT_MESSAGE = exports.CANCEL_BUTTON_LABEL = exports.BACK_BUTTON_LABEL = exports.ADD_INFERENCE_PIPELINE_STEPS = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ADD_INFERENCE_PIPELINE_STEPS = exports.ADD_INFERENCE_PIPELINE_STEPS = {
  DETAILS: 'Details',
  CONFIGURE_PROCESSOR: 'Configure processor',
  ON_FAILURE: 'Failure handling',
  TEST: 'Test',
  CREATE: 'create'
};
const CANCEL_BUTTON_LABEL = exports.CANCEL_BUTTON_LABEL = _i18n.i18n.translate('xpack.ml.trainedModels.actions.cancelButtonLabel', {
  defaultMessage: 'Cancel'
});
const CLEAR_BUTTON_LABEL = exports.CLEAR_BUTTON_LABEL = _i18n.i18n.translate('xpack.ml.trainedModels.actions.clearButtonLabel', {
  defaultMessage: 'Clear'
});
const CLOSE_BUTTON_LABEL = exports.CLOSE_BUTTON_LABEL = _i18n.i18n.translate('xpack.ml.trainedModels.actions.closeButtonLabel', {
  defaultMessage: 'Close'
});
const BACK_BUTTON_LABEL = exports.BACK_BUTTON_LABEL = _i18n.i18n.translate('xpack.ml.trainedModels.actions.backButtonLabel', {
  defaultMessage: 'Back'
});
const CONTINUE_BUTTON_LABEL = exports.CONTINUE_BUTTON_LABEL = _i18n.i18n.translate('xpack.ml.trainedModels.actions.continueButtonLabel', {
  defaultMessage: 'Continue'
});
const EDIT_MESSAGE = exports.EDIT_MESSAGE = _i18n.i18n.translate('xpack.ml.trainedModels.actions.create.advancedDetails.editButtonText', {
  defaultMessage: 'Edit'
});
const CREATE_FIELD_MAPPING_MESSAGE = exports.CREATE_FIELD_MAPPING_MESSAGE = _i18n.i18n.translate('xpack.ml.trainedModels.actions.create.advancedDetails.createFieldMapText', {
  defaultMessage: 'Create field map'
});
const CANCEL_EDIT_MESSAGE = exports.CANCEL_EDIT_MESSAGE = _i18n.i18n.translate('xpack.ml.trainedModels.actions.create.advancedDetails.cancelEditButtonText', {
  defaultMessage: 'Cancel'
});
