// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/processor_form/processors/common_fields/properties_field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PropertiesField = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _shared_imports = require("../../../../../../../shared_imports");
var _shared = require("../shared");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const fieldsConfig = {
  properties: {
    type: _shared_imports.FIELD_TYPES.COMBO_BOX,
    deserializer: _shared.to.arrayOfStrings,
    serializer: v => v.length ? v : undefined,
    label: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.commonFields.propertiesFieldLabel', {
      defaultMessage: 'Properties (optional)'
    })
  }
};
const PropertiesField = ({
  helpText,
  euiFieldProps
}) => {
  return /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    config: {
      ...fieldsConfig.properties,
      helpText
    },
    component: _shared_imports.ComboBoxField,
    path: "fields.properties",
    componentProps: {
      euiFieldProps
    }
  });
};
exports.PropertiesField = PropertiesField;
