// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/processor_form/processors/common_fields/target_field.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TargetField = exports.TARGET_FIELD_PATH = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _shared_imports = require("../../../../../../../shared_imports");
var _shared = require("../shared");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const fieldsConfig = {
  target_field: {
    type: _shared_imports.FIELD_TYPES.TEXT,
    deserializer: String,
    serializer: _shared.from.emptyStringToUndefined,
    label: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.commonFields.targetFieldLabel', {
      defaultMessage: 'Target field (optional)'
    }),
    helpText: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.commonFields.targetFieldHelpText', {
      defaultMessage: 'Output field. If empty, the input field is updated in place.'
    })
  }
};
const TARGET_FIELD_PATH = exports.TARGET_FIELD_PATH = 'fields.target_field';
const TargetField = props => {
  return /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    config: {
      ...fieldsConfig.target_field,
      ...props
    },
    component: _shared_imports.Field,
    path: TARGET_FIELD_PATH,
    "data-test-subj": "targetField"
  });
};
exports.TargetField = TargetField;
