// @kbn/aiops-plugin/public/components/log_categorization/technical_preview_badge/technical_preview_badge.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TechnicalPreviewBadge = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TechnicalPreviewBadge = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiBetaBadge, {
    label: _i18n.i18n.translate('xpack.aiops.techPreviewBadge.label', {
      defaultMessage: 'Technical preview'
    }),
    size: "m",
    color: "hollow",
    tooltipContent: _i18n.i18n.translate('xpack.aiops.techPreviewBadge.tooltip', {
      defaultMessage: 'This functionality is in technical preview and may be changed or removed completely in a future release. Elastic will work to fix any issues, but features in technical preview are not subject to the support SLA of official GA features.'
    }),
    tooltipPosition: 'right'
  });
};
exports.TechnicalPreviewBadge = TechnicalPreviewBadge;
