// @kbn/alerting-plugin/common/routes/rule/apis/create/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "actionAlertsFilterSchema", {
  enumerable: true,
  get: function () {
    return _latest.actionAlertsFilterSchema;
  }
});
Object.defineProperty(exports, "actionAlertsFilterSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.actionAlertsFilterSchema;
  }
});
Object.defineProperty(exports, "actionFrequencySchema", {
  enumerable: true,
  get: function () {
    return _latest.actionFrequencySchema;
  }
});
Object.defineProperty(exports, "actionFrequencySchemaV1", {
  enumerable: true,
  get: function () {
    return _v.actionFrequencySchema;
  }
});
Object.defineProperty(exports, "actionSchema", {
  enumerable: true,
  get: function () {
    return _latest.actionSchema;
  }
});
Object.defineProperty(exports, "actionSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.actionSchema;
  }
});
Object.defineProperty(exports, "createBodySchema", {
  enumerable: true,
  get: function () {
    return _latest.createBodySchema;
  }
});
Object.defineProperty(exports, "createBodySchemaV1", {
  enumerable: true,
  get: function () {
    return _v.createBodySchema;
  }
});
Object.defineProperty(exports, "createParamsSchema", {
  enumerable: true,
  get: function () {
    return _latest.createParamsSchema;
  }
});
Object.defineProperty(exports, "createParamsSchemaV1", {
  enumerable: true,
  get: function () {
    return _v.createParamsSchema;
  }
});
var _latest = require("./schemas/latest");
var _v = require("./schemas/v1");
