// @kbn/observability-ai-assistant-plugin/public/utils/create_function_response_error.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createPublicFunctionResponseError = createPublicFunctionResponseError;
var _create_function_response_message = require("../../common/utils/create_function_response_message");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// this one does not check for ES errors, they don't get to the browser
// and we prevent importing Node.js-only code
function createPublicFunctionResponseError({
  name,
  error,
  message
}) {
  const sanitizedError = JSON.parse('toJSON' in error && typeof error.toJSON === 'function' ? error.toJSON() : JSON.stringify(error));
  return (0, _create_function_response_message.createFunctionResponseMessage)({
    name,
    content: {
      error: {
        ...sanitizedError,
        name: error.name,
        message: error.message,
        cause: error.cause
      },
      message: message || error.message
    },
    data: {
      stack: error.stack
    }
  });
}
