// @kbn/synthetics-plugin/server/routes/settings/params/delete_param.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.deleteSyntheticsParamsRoute = void 0;
var _configSchema = require("@kbn/config-schema");
var _saved_objects = require("../../../../common/types/saved_objects");
var _constants = require("../../../../common/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const deleteSyntheticsParamsRoute = () => ({
  method: 'DELETE',
  path: _constants.SYNTHETICS_API_URLS.PARAMS,
  validate: {},
  validation: {
    request: {
      body: _configSchema.schema.object({
        ids: _configSchema.schema.arrayOf(_configSchema.schema.string())
      })
    }
  },
  handler: async ({
    savedObjectsClient,
    request
  }) => {
    const {
      ids
    } = request.body;
    const result = await savedObjectsClient.bulkDelete(ids.map(id => ({
      type: _saved_objects.syntheticsParamType,
      id
    })), {
      force: true
    });
    return result.statuses.map(({
      id,
      success
    }) => ({
      id,
      deleted: success
    }));
  }
});
exports.deleteSyntheticsParamsRoute = deleteSyntheticsParamsRoute;
