// @kbn/console-plugin/public/lib/autocomplete/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  AutocompleteComponent: true,
  SharedComponent: true,
  ConstantComponent: true,
  ListComponent: true,
  SimpleParamComponent: true,
  ConditionalProxy: true,
  GlobalOnlyComponent: true,
  ObjectComponent: true,
  AcceptEndpointComponent: true,
  URL_PATH_END_MARKER: true,
  UrlPatternMatcher: true,
  IndexAutocompleteComponent: true,
  FieldAutocompleteComponent: true,
  IndexTemplateAutocompleteComponent: true,
  ComponentTemplateAutocompleteComponent: true,
  DataStreamAutocompleteComponent: true
};
Object.defineProperty(exports, "AcceptEndpointComponent", {
  enumerable: true,
  get: function () {
    return _accept_endpoint_component.AcceptEndpointComponent;
  }
});
Object.defineProperty(exports, "AutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _autocomplete_component.AutocompleteComponent;
  }
});
Object.defineProperty(exports, "ComponentTemplateAutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _component_template_autocomplete_component.ComponentTemplateAutocompleteComponent;
  }
});
Object.defineProperty(exports, "ConditionalProxy", {
  enumerable: true,
  get: function () {
    return _conditional_proxy.ConditionalProxy;
  }
});
Object.defineProperty(exports, "ConstantComponent", {
  enumerable: true,
  get: function () {
    return _constant_component.ConstantComponent;
  }
});
Object.defineProperty(exports, "DataStreamAutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _data_stream_autocomplete_component.DataStreamAutocompleteComponent;
  }
});
Object.defineProperty(exports, "FieldAutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _field_autocomplete_component.FieldAutocompleteComponent;
  }
});
Object.defineProperty(exports, "GlobalOnlyComponent", {
  enumerable: true,
  get: function () {
    return _global_only_component.GlobalOnlyComponent;
  }
});
Object.defineProperty(exports, "IndexAutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _index_autocomplete_component.IndexAutocompleteComponent;
  }
});
Object.defineProperty(exports, "IndexTemplateAutocompleteComponent", {
  enumerable: true,
  get: function () {
    return _index_template_autocomplete_component.IndexTemplateAutocompleteComponent;
  }
});
Object.defineProperty(exports, "ListComponent", {
  enumerable: true,
  get: function () {
    return _list_component.ListComponent;
  }
});
Object.defineProperty(exports, "ObjectComponent", {
  enumerable: true,
  get: function () {
    return _object_component.ObjectComponent;
  }
});
Object.defineProperty(exports, "SharedComponent", {
  enumerable: true,
  get: function () {
    return _shared_component.SharedComponent;
  }
});
Object.defineProperty(exports, "SimpleParamComponent", {
  enumerable: true,
  get: function () {
    return _simple_param_component.SimpleParamComponent;
  }
});
Object.defineProperty(exports, "URL_PATH_END_MARKER", {
  enumerable: true,
  get: function () {
    return _accept_endpoint_component.URL_PATH_END_MARKER;
  }
});
Object.defineProperty(exports, "UrlPatternMatcher", {
  enumerable: true,
  get: function () {
    return _url_pattern_matcher.UrlPatternMatcher;
  }
});
var _autocomplete_component = require("./autocomplete_component");
var _shared_component = require("./shared_component");
var _constant_component = require("./constant_component");
var _list_component = require("./list_component");
var _simple_param_component = require("./simple_param_component");
var _conditional_proxy = require("./conditional_proxy");
var _global_only_component = require("./global_only_component");
var _object_component = require("./object_component");
var _accept_endpoint_component = require("./accept_endpoint_component");
var _url_pattern_matcher = require("./url_pattern_matcher");
var _index_autocomplete_component = require("./index_autocomplete_component");
var _field_autocomplete_component = require("./field_autocomplete_component");
var _index_template_autocomplete_component = require("./index_template_autocomplete_component");
var _component_template_autocomplete_component = require("./component_template_autocomplete_component");
var _data_stream_autocomplete_component = require("./data_stream_autocomplete_component");
var _legacy = require("./legacy");
Object.keys(_legacy).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _legacy[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _legacy[key];
    }
  });
});
