// @kbn/synthetics-plugin/public/apps/synthetics/components/common/screenshot/screenshot_image.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ScreenshotImage = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _empty_thumbnail = require("./empty_thumbnail");
var _screenshot_size = require("./screenshot_size");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DEFAULT_SIZE = [512, 512];
const ScreenshotImage = ({
  label,
  imgSrc,
  isLoading,
  animateLoading = true,
  unavailableMessage,
  borderColor,
  borderRadius,
  hasBorder = true,
  size = [100, 64],
  onMouseEnter,
  onMouseLeave,
  onClick
}) => {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const [naturalSize, setNaturalSize] = (0, _react.useState)(DEFAULT_SIZE);
  const {
    width,
    height
  } = (0, _screenshot_size.getConfinedScreenshotSize)(size, naturalSize);
  return imgSrc ?
  /*#__PURE__*/
  // eslint-disable-next-line jsx-a11y/no-noninteractive-element-interactions
  _react.default.createElement("img", {
    css: {
      outline: 0,
      objectFit: 'contain',
      width,
      height,
      cursor: 'pointer',
      border: euiTheme.border.thin,
      ...(borderColor ? {
        borderColor
      } : {}),
      ...(borderRadius ? {
        borderRadius
      } : {}),
      ...(!hasBorder ? {
        borderRadius
      } : {}),
      ...(borderColor ? {
        borderColor
      } : {})
    },
    src: imgSrc,
    alt: label,
    "data-test-subj": "stepScreenshotThumbnail",
    onLoad: evt => {
      var _naturalWidth, _evt$target, _naturalHeight, _evt$target2;
      const updatedSize = [(_naturalWidth = evt === null || evt === void 0 ? void 0 : (_evt$target = evt.target) === null || _evt$target === void 0 ? void 0 : _evt$target.naturalWidth) !== null && _naturalWidth !== void 0 ? _naturalWidth : naturalSize[0], (_naturalHeight = evt === null || evt === void 0 ? void 0 : (_evt$target2 = evt.target) === null || _evt$target2 === void 0 ? void 0 : _evt$target2.naturalHeight) !== null && _naturalHeight !== void 0 ? _naturalHeight : naturalSize[1]];
      setNaturalSize(updatedSize);
    },
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    onKeyDown: undefined
  }) : /*#__PURE__*/_react.default.createElement(_empty_thumbnail.EmptyThumbnail, {
    isLoading: isLoading,
    size: size === 'full' ? naturalSize : size,
    unavailableMessage: unavailableMessage,
    borderRadius: borderRadius,
    animateLoading: animateLoading
  });
};
exports.ScreenshotImage = ScreenshotImage;
