// @kbn/ui-actions-enhanced-plugin/public/drilldowns/url_drilldown/components/variable_popover/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.VariablePopover = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("./i18n");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const VariablePopover = ({
  variables,
  onSelect,
  variablesHelpLink
}) => {
  const [isVariablesPopoverOpen, setIsVariablesPopoverOpen] = (0, _react.useState)(false);
  const closePopover = () => setIsVariablesPopoverOpen(false);
  const options = variables.map(({
    label
  }) => ({
    key: label,
    label
  }));
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPopover, {
    ownFocus: true,
    button: /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "xs"
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
      onClick: () => setIsVariablesPopoverOpen(true)
    }, _i18n.txtAddVariableButtonTitle, " ", /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
      type: "indexOpen"
    }))),
    isOpen: isVariablesPopoverOpen,
    closePopover: closePopover,
    panelPaddingSize: "none",
    anchorPosition: "downLeft"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSelectable, {
    singleSelection: true,
    searchable: true,
    searchProps: {
      placeholder: _i18n.txtUrlTemplateVariablesFilterPlaceholderText,
      compressed: true
    },
    options: options,
    onChange: newOptions => {
      const selected = newOptions.find(o => o.checked === 'on');
      if (!selected) return;
      onSelect(selected.key);
      closePopover();
    },
    listProps: {
      showIcons: false
    }
  }, (list, search) => /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: 320
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiPopoverTitle, null, search), list, variablesHelpLink && /*#__PURE__*/_react.default.createElement(_eui.EuiPopoverFooter, {
    className: 'eui-textRight'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    external: true,
    href: variablesHelpLink,
    target: "_blank"
  }, _i18n.txtUrlTemplateVariablesHelpLinkText)))));
};
exports.VariablePopover = VariablePopover;
