// @kbn/elastic-assistant-common/impl/schemas/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  API_VERSIONS: true,
  PUBLIC_API_ACCESS: true,
  INTERNAL_API_ACCESS: true,
  PromptResponse: true,
  PromptTypeEnum: true
};
exports.PUBLIC_API_ACCESS = exports.INTERNAL_API_ACCESS = exports.API_VERSIONS = void 0;
Object.defineProperty(exports, "PromptResponse", {
  enumerable: true,
  get: function () {
    return _bulk_crud_prompts_route.PromptResponse;
  }
});
Object.defineProperty(exports, "PromptTypeEnum", {
  enumerable: true,
  get: function () {
    return _bulk_crud_prompts_route.PromptTypeEnum;
  }
});
var _common_attributes = require("./common_attributes.gen");
Object.keys(_common_attributes).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common_attributes[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common_attributes[key];
    }
  });
});
var _common_attributes2 = require("./attack_discovery/common_attributes.gen");
Object.keys(_common_attributes2).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common_attributes2[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common_attributes2[key];
    }
  });
});
var _get_attack_discovery_route = require("./attack_discovery/get_attack_discovery_route.gen");
Object.keys(_get_attack_discovery_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _get_attack_discovery_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _get_attack_discovery_route[key];
    }
  });
});
var _post_attack_discovery_route = require("./attack_discovery/post_attack_discovery_route.gen");
Object.keys(_post_attack_discovery_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _post_attack_discovery_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _post_attack_discovery_route[key];
    }
  });
});
var _cancel_attack_discovery_route = require("./attack_discovery/cancel_attack_discovery_route.gen");
Object.keys(_cancel_attack_discovery_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _cancel_attack_discovery_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _cancel_attack_discovery_route[key];
    }
  });
});
var _post_chat_complete_route = require("./chat/post_chat_complete_route.gen");
Object.keys(_post_chat_complete_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _post_chat_complete_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _post_chat_complete_route[key];
    }
  });
});
var _post_evaluate_route = require("./evaluation/post_evaluate_route.gen");
Object.keys(_post_evaluate_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _post_evaluate_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _post_evaluate_route[key];
    }
  });
});
var _get_evaluate_route = require("./evaluation/get_evaluate_route.gen");
Object.keys(_get_evaluate_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _get_evaluate_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _get_evaluate_route[key];
    }
  });
});
var _get_capabilities_route = require("./capabilities/get_capabilities_route.gen");
Object.keys(_get_capabilities_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _get_capabilities_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _get_capabilities_route[key];
    }
  });
});
var _bulk_crud_conversations_route = require("./conversations/bulk_crud_conversations_route.gen");
Object.keys(_bulk_crud_conversations_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bulk_crud_conversations_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bulk_crud_conversations_route[key];
    }
  });
});
var _common_attributes3 = require("./conversations/common_attributes.gen");
Object.keys(_common_attributes3).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common_attributes3[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common_attributes3[key];
    }
  });
});
var _crud_conversation_route = require("./conversations/crud_conversation_route.gen");
Object.keys(_crud_conversation_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _crud_conversation_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _crud_conversation_route[key];
    }
  });
});
var _find_conversations_route = require("./conversations/find_conversations_route.gen");
Object.keys(_find_conversations_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _find_conversations_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _find_conversations_route[key];
    }
  });
});
var _post_actions_connector_execute_route = require("./actions_connector/post_actions_connector_execute_route.gen");
Object.keys(_post_actions_connector_execute_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _post_actions_connector_execute_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _post_actions_connector_execute_route[key];
    }
  });
});
var _crud_kb_route = require("./knowledge_base/crud_kb_route.gen");
Object.keys(_crud_kb_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _crud_kb_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _crud_kb_route[key];
    }
  });
});
var _bulk_crud_knowledge_base_route = require("./knowledge_base/bulk_crud_knowledge_base_route.gen");
Object.keys(_bulk_crud_knowledge_base_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _bulk_crud_knowledge_base_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _bulk_crud_knowledge_base_route[key];
    }
  });
});
var _common_attributes4 = require("./knowledge_base/common_attributes.gen");
Object.keys(_common_attributes4).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _common_attributes4[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _common_attributes4[key];
    }
  });
});
var _crud_knowledge_base_route = require("./knowledge_base/crud_knowledge_base_route.gen");
Object.keys(_crud_knowledge_base_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _crud_knowledge_base_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _crud_knowledge_base_route[key];
    }
  });
});
var _find_knowledge_base_entries_route = require("./knowledge_base/find_knowledge_base_entries_route.gen");
Object.keys(_find_knowledge_base_entries_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _find_knowledge_base_entries_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _find_knowledge_base_entries_route[key];
    }
  });
});
var _find_prompts_route = require("./prompts/find_prompts_route.gen");
Object.keys(_find_prompts_route).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _find_prompts_route[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _find_prompts_route[key];
    }
  });
});
var _bulk_crud_prompts_route = require("./prompts/bulk_crud_prompts_route.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// API versioning constants
const API_VERSIONS = exports.API_VERSIONS = {
  public: {
    v1: '2023-10-31'
  },
  internal: {
    v1: '1'
  }
};
const PUBLIC_API_ACCESS = exports.PUBLIC_API_ACCESS = 'public';
const INTERNAL_API_ACCESS = exports.INTERNAL_API_ACCESS = 'internal';

// Common Schemas

// Attack discovery Schemas

// Chat Schemas

// Evaluation Schemas

// Capabilities Schemas

// Conversations Schemas

// Actions Connector Schemas

// Knowledge Base Schemas
