// @kbn/es-ui-shared-plugin/public/authorization/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AuthorizationContext", {
  enumerable: true,
  get: function () {
    return _authorization.AuthorizationContext;
  }
});
Object.defineProperty(exports, "AuthorizationProvider", {
  enumerable: true,
  get: function () {
    return _authorization.AuthorizationProvider;
  }
});
Object.defineProperty(exports, "NotAuthorizedSection", {
  enumerable: true,
  get: function () {
    return _authorization.NotAuthorizedSection;
  }
});
Object.defineProperty(exports, "PageError", {
  enumerable: true,
  get: function () {
    return _authorization.PageError;
  }
});
Object.defineProperty(exports, "SectionError", {
  enumerable: true,
  get: function () {
    return _authorization.SectionError;
  }
});
Object.defineProperty(exports, "WithPrivileges", {
  enumerable: true,
  get: function () {
    return _authorization.WithPrivileges;
  }
});
Object.defineProperty(exports, "useAuthorizationContext", {
  enumerable: true,
  get: function () {
    return _authorization.useAuthorizationContext;
  }
});
var _authorization = require("../../__packages_do_not_import__/authorization");
