// @kbn/fleet-plugin/public/applications/fleet/sections/settings/components/download_source_flyout/use_delete_download_source.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.useDeleteDownloadSource = useDeleteDownloadSource;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _i18n = require("@kbn/i18n");
var _hooks = require("../../../../hooks");
var _use_confirm_modal = require("../../hooks/use_confirm_modal");
var _get_count = require("./services/get_count");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ConfirmTitle = () => /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
  id: "xpack.fleet.settings.deleteDowloadSource.confirmModalTitle",
  defaultMessage: "Delete and deploy changes?"
});
const ConfirmDeleteDescription = ({
  downloadSource,
  agentCount,
  agentPolicyCount
}) => /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
  id: "xpack.fleet.settings.deleteDowloadSource.confirmModalText",
  defaultMessage: "This action will delete {downloadSourceName} agent binary source. It will update {policies} and {agents}. This action can not be undone. Are you sure you wish to continue?",
  values: {
    downloadSourceName: /*#__PURE__*/_react.default.createElement("strong", null, downloadSource.name),
    agents: /*#__PURE__*/_react.default.createElement("strong", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.settings.deleteDowloadSource.agentsCount",
      defaultMessage: "{agentCount, plural, one {# agent} other {# agents}}",
      values: {
        agentCount
      }
    })),
    policies: /*#__PURE__*/_react.default.createElement("strong", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.settings.deleteDowloadSource.agentPolicyCount",
      defaultMessage: "{agentPolicyCount, plural, one {# agent policy} other {# agent policies}}",
      values: {
        agentPolicyCount
      }
    }))
  }
});
function useDeleteDownloadSource(onSuccess) {
  const {
    confirm
  } = (0, _use_confirm_modal.useConfirmModal)();
  const {
    notifications
  } = (0, _hooks.useStartServices)();
  const deleteDownloadSource = (0, _react.useCallback)(async downloadSource => {
    try {
      const {
        agentCount,
        agentPolicyCount
      } = await (0, _get_count.getCountsForDownloadSource)(downloadSource);
      const isConfirmed = await confirm( /*#__PURE__*/_react.default.createElement(ConfirmTitle, null), /*#__PURE__*/_react.default.createElement(ConfirmDeleteDescription, {
        "data-test-subj": "editDownloadSourcesDeleteModal.confirmModalText",
        downloadSource: downloadSource,
        agentCount: agentCount,
        agentPolicyCount: agentPolicyCount
      }), {
        buttonColor: 'danger',
        confirmButtonText: _i18n.i18n.translate('xpack.fleet.settings.deleteDownloadSource.confirmButtonLabel', {
          defaultMessage: 'Delete and deploy'
        })
      });
      if (!isConfirmed) {
        return;
      }
      const res = await (0, _hooks.sendDeleteDownloadSource)(downloadSource.id);
      if (res.error) {
        throw res.error;
      }
      onSuccess();
    } catch (err) {
      notifications.toasts.addError(err, {
        title: _i18n.i18n.translate('xpack.fleet.settings.deleteDownloadSource.errorToastTitle', {
          defaultMessage: 'Error deleting agent binary source.'
        })
      });
    }
  }, [confirm, notifications.toasts, onSuccess]);
  return {
    deleteDownloadSource
  };
}
