// @kbn/data-views-plugin/public/types.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ResolveIndexResponseItemIndexAttrs = exports.IndicesResponseItemIndexAttrs = exports.INDEX_PATTERN_TYPE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */
let INDEX_PATTERN_TYPE = exports.INDEX_PATTERN_TYPE = /*#__PURE__*/function (INDEX_PATTERN_TYPE) {
  INDEX_PATTERN_TYPE["ROLLUP"] = "rollup";
  INDEX_PATTERN_TYPE["DEFAULT"] = "default";
  return INDEX_PATTERN_TYPE;
}({});
let IndicesResponseItemIndexAttrs = exports.IndicesResponseItemIndexAttrs = /*#__PURE__*/function (IndicesResponseItemIndexAttrs) {
  IndicesResponseItemIndexAttrs["OPEN"] = "open";
  IndicesResponseItemIndexAttrs["CLOSED"] = "closed";
  IndicesResponseItemIndexAttrs["HIDDEN"] = "hidden";
  IndicesResponseItemIndexAttrs["FROZEN"] = "frozen";
  return IndicesResponseItemIndexAttrs;
}({});
/**
 * Data views public setup dependencies
 */
/**
 * Data views public start dependencies
 */
/**
 * Data plugin public Setup contract
 */
/**
 * Data views plugin public Start contract
 */
// for showing index matches
let ResolveIndexResponseItemIndexAttrs = exports.ResolveIndexResponseItemIndexAttrs = /*#__PURE__*/function (ResolveIndexResponseItemIndexAttrs) {
  ResolveIndexResponseItemIndexAttrs["OPEN"] = "open";
  ResolveIndexResponseItemIndexAttrs["CLOSED"] = "closed";
  ResolveIndexResponseItemIndexAttrs["HIDDEN"] = "hidden";
  ResolveIndexResponseItemIndexAttrs["FROZEN"] = "frozen";
  return ResolveIndexResponseItemIndexAttrs;
}({});
