// @kbn/vis-default-editor-plugin/public/components/controls/top_size.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TopSizeParamEditor = TopSizeParamEditor;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _size = require("./size");
var _top_aggregate = require("./top_aggregate");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function TopSizeParamEditor(props) {
  const iconTip = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, ' ', /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    position: "right",
    content: _i18n.i18n.translate('visDefaultEditor.controls.sizeTooltip', {
      defaultMessage: "Request top-K hits. Multiple hits will be combined via 'aggregate with'."
    }),
    type: "questionInCircle"
  }));
  const fieldType = props.agg.params.field && props.agg.params.field.type;
  const disabled = fieldType && !(0, _top_aggregate.getCompatibleAggs)(props.agg).length;
  return /*#__PURE__*/_react.default.createElement(_size.SizeParamEditor, (0, _extends2.default)({}, props, {
    iconTip: iconTip,
    disabled: disabled
  }));
}
