// @kbn/osquery-plugin/common/api/live_query/create_live_query_route.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createLiveQueryRequestBodySchema = void 0;
var t = _interopRequireWildcard(require("io-ts"));
var _osqueryIoTsTypes = require("@kbn/osquery-io-ts-types");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const createLiveQueryRequestBodySchema = exports.createLiveQueryRequestBodySchema = t.partial({
  agent_ids: t.array(t.string),
  agent_all: t.union([t.boolean, t.undefined]),
  agent_platforms: t.array(t.string),
  agent_policy_ids: t.array(t.string),
  query: _osqueryIoTsTypes.queryOrUndefined,
  queries: _osqueryIoTsTypes.arrayQueries,
  saved_query_id: _osqueryIoTsTypes.savedQueryIdOrUndefined,
  timeout: _osqueryIoTsTypes.timeoutOrUndefined,
  ecs_mapping: _osqueryIoTsTypes.ecsMappingOrUndefined,
  pack_id: _osqueryIoTsTypes.packIdOrUndefined,
  alert_ids: t.array(t.string),
  case_ids: t.array(t.string),
  event_ids: t.array(t.string),
  metadata: t.union([t.object, t.undefined])
});
