// @kbn/osquery-plugin/common/api/live_query/find_live_query.gen.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SuccessResponse = exports.FindLiveQueryRequestQuery = void 0;
var _zod = require("zod");
var _common_attributes = require("../model/schema/common_attributes.gen");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/*
 * NOTICE: Do not edit this file manually.
 * This file is automatically generated by the OpenAPI Generator, @kbn/openapi-generator.
 *
 * info:
 *   title: Find Live Queries Schema
 *   version: 2023-10-31
 */

const FindLiveQueryRequestQuery = exports.FindLiveQueryRequestQuery = _zod.z.object({
  kuery: _common_attributes.KueryOrUndefined.optional(),
  page: _common_attributes.PageOrUndefined.optional(),
  pageSize: _common_attributes.PageSizeOrUndefined.optional(),
  sort: _common_attributes.SortOrUndefined.optional(),
  sortOrder: _common_attributes.SortOrderOrUndefined.optional()
});
const SuccessResponse = exports.SuccessResponse = _zod.z.object({});
