// @kbn/integration-assistant-plugin/public/components/create_integration/create_integration_assistant/header/steps.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Steps = void 0;
var _eui = require("@elastic/eui");
var _react = require("@emotion/react");
var _i18n = require("@kbn/i18n");
var _react2 = _interopRequireWildcard(require("react"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

// Custom styles for the EuiStepsHorizontal component suggested by design team
const stepsCss = (0, _react.css)`
  margin-left: -66px;
  margin-right: -59px;
`;
const STEP_CONNECTOR = _i18n.i18n.translate('xpack.integrationAssistant.step.connector', {
  defaultMessage: 'Connector'
});
const STEP_INTEGRATION = _i18n.i18n.translate('xpack.integrationAssistant.step.integration', {
  defaultMessage: 'Integration'
});
const STEP_DATA_STREAM = _i18n.i18n.translate('xpack.integrationAssistant.step.dataStream', {
  defaultMessage: 'Data stream'
});
const STEP_REVIEW = _i18n.i18n.translate('xpack.integrationAssistant.step.review', {
  defaultMessage: 'Review'
});
const getStepStatus = (step, currentStep, loading) => {
  if (step === currentStep) {
    return loading ? 'loading' : 'current';
  } else if (step < currentStep) {
    return 'complete';
  } else {
    return 'disabled';
  }
};
const getStepOnClick = (step, currentStep, setStep) => {
  if (step < currentStep) {
    return () => setStep(step);
  }
  return () => {};
};
const Steps = exports.Steps = /*#__PURE__*/_react2.default.memo(({
  currentStep,
  setStep,
  isGenerating
}) => {
  const steps = (0, _react2.useMemo)(() => {
    return [{
      title: STEP_CONNECTOR,
      status: getStepStatus(1, currentStep),
      onClick: getStepOnClick(1, currentStep, setStep)
    }, {
      title: STEP_INTEGRATION,
      status: getStepStatus(2, currentStep),
      onClick: getStepOnClick(2, currentStep, setStep)
    }, {
      title: STEP_DATA_STREAM,
      status: getStepStatus(3, currentStep, isGenerating),
      onClick: getStepOnClick(3, currentStep, setStep)
    }, {
      title: STEP_REVIEW,
      status: getStepStatus(4, currentStep, isGenerating),
      onClick: getStepOnClick(4, currentStep, setStep)
    }];
  }, [currentStep, setStep, isGenerating]);
  return /*#__PURE__*/_react2.default.createElement(_eui.EuiStepsHorizontal, {
    steps: steps,
    size: "s",
    css: stepsCss
  });
});
Steps.displayName = 'Steps';
