// @kbn/security-solution-plugin/public/entity_analytics/components/top_risk_score_contributors_alerts/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TopRiskScoreContributorsAlerts = void 0;
var _react = _interopRequireWildcard(require("react"));
var _securitysolutionDataTable = require("@kbn/securitysolution-data-table");
var _eui = require("@elastic/eui");
var _header_section = require("../../../common/components/header_section");
var i18n = _interopRequireWildcard(require("./translations"));
var _risk_engine = require("../../../../common/entity_analytics/risk_engine");
var _constants = require("../../../../common/constants");
var _alerts_table = require("../../../detections/components/alerts_table");
var _alerts_grouping = require("../../../detections/components/alerts_table/alerts_grouping");
var _use_global_time = require("../../../common/containers/use_global_time");
var _use_selector = require("../../../common/hooks/use_selector");
var _inputs = require("../../../common/store/inputs");
var _user_info = require("../../../detections/components/user_info");
var _containers = require("../../../sourcerer/containers");
var _model = require("../../../sourcerer/store/model");
var _risk_information = require("../risk_information");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TopRiskScoreContributorsAlerts = ({
  toggleStatus,
  toggleQuery,
  riskScore,
  riskEntity,
  loading
}) => {
  const {
    to,
    from
  } = (0, _use_global_time.useGlobalTime)();
  const [{
    loading: userInfoLoading,
    signalIndexName,
    hasIndexWrite,
    hasIndexMaintenance
  }] = (0, _user_info.useUserData)();
  const {
    runtimeMappings
  } = (0, _containers.useSourcererDataView)(_model.SourcererScopeName.detections);
  const getGlobalFiltersQuerySelector = (0, _react.useMemo)(() => _inputs.inputsSelectors.globalFiltersQuerySelector(), []);
  const getGlobalQuerySelector = (0, _react.useMemo)(() => _inputs.inputsSelectors.globalQuerySelector(), []);
  const query = (0, _use_selector.useDeepEqualSelector)(getGlobalQuerySelector);
  const filters = (0, _use_selector.useDeepEqualSelector)(getGlobalFiltersQuerySelector);
  const inputFilters = (0, _react.useMemo)(() => {
    var _riskScoreEntity$risk, _riskScoreEntity$risk2;
    const riskScoreEntity = riskEntity === _risk_engine.RiskScoreEntity.host ? riskScore.host : riskScore.user;
    const riskInputs = (_riskScoreEntity$risk = riskScoreEntity === null || riskScoreEntity === void 0 ? void 0 : (_riskScoreEntity$risk2 = riskScoreEntity.risk) === null || _riskScoreEntity$risk2 === void 0 ? void 0 : _riskScoreEntity$risk2.inputs) !== null && _riskScoreEntity$risk !== void 0 ? _riskScoreEntity$risk : [];
    return [{
      meta: {
        alias: null,
        negate: false,
        disabled: false
      },
      query: {
        terms: {
          _id: riskInputs.map(item => item.id)
        }
      }
    }];
  }, [riskScore, riskEntity]);
  const renderGroupedAlertTable = (0, _react.useCallback)(groupingFilters => {
    return /*#__PURE__*/_react.default.createElement(_alerts_table.AlertsTableComponent, {
      configId: _constants.ALERTS_TABLE_REGISTRY_CONFIG_IDS.RISK_INPUTS,
      inputFilters: [...inputFilters, ...filters, ...groupingFilters],
      tableId: _securitysolutionDataTable.TableId.alertsRiskInputs
    });
  }, [inputFilters, filters]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
    hasBorder: true,
    "data-test-subj": "topRiskScoreContributorsAlerts"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: 'none'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: 1
  }, /*#__PURE__*/_react.default.createElement(_header_section.HeaderSection, {
    title: i18n.TOP_RISK_SCORE_CONTRIBUTORS,
    hideSubtitle: true,
    toggleQuery: toggleQuery,
    toggleStatus: toggleStatus,
    headerFilters: /*#__PURE__*/_react.default.createElement(_risk_information.RiskInformationButtonEmpty, {
      riskEntity: riskEntity
    })
  }))), toggleStatus && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    "data-test-subj": "topRiskScoreContributorsAlerts-table",
    gutterSize: "none",
    direction: "column"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: 1
  }, /*#__PURE__*/_react.default.createElement(_alerts_grouping.GroupedAlertsTable, {
    defaultFilters: [...inputFilters, ...filters],
    from: from,
    globalFilters: filters,
    globalQuery: query,
    hasIndexMaintenance: hasIndexMaintenance !== null && hasIndexMaintenance !== void 0 ? hasIndexMaintenance : false,
    hasIndexWrite: hasIndexWrite !== null && hasIndexWrite !== void 0 ? hasIndexWrite : false,
    loading: userInfoLoading || loading,
    renderChildComponent: renderGroupedAlertTable,
    runtimeMappings: runtimeMappings,
    signalIndexName: signalIndexName,
    tableId: _securitysolutionDataTable.TableId.alertsRiskInputs,
    to: to
  }))));
};
exports.TopRiskScoreContributorsAlerts = TopRiskScoreContributorsAlerts;
