// @kbn/enterprise-search-plugin/public/applications/app_search/components/synonyms/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UPDATE_SUCCESS = exports.SYNONYM_UPDATE_TITLE = exports.SYNONYM_IMPACT_MESSAGE = exports.SYNONYM_CREATE_TITLE = exports.SYNONYMS_TITLE = exports.SYNONYMS_PAGE_META = exports.DELETE_SUCCESS = exports.DELETE_CONFIRMATION = exports.CREATE_SUCCESS = void 0;
var _i18n = require("@kbn/i18n");
var _constants = require("../../../shared/constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SYNONYMS_PAGE_META = exports.SYNONYMS_PAGE_META = {
  page: {
    ..._constants.DEFAULT_META.page,
    size: 12 // Use a multiple of 3, since synonym cards are in rows of 3
  }
};
const SYNONYMS_TITLE = exports.SYNONYMS_TITLE = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.title', {
  defaultMessage: 'Synonyms'
});
const SYNONYM_CREATE_TITLE = exports.SYNONYM_CREATE_TITLE = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.createSynonymSetTitle', {
  defaultMessage: 'Add a synonym set'
});
const SYNONYM_UPDATE_TITLE = exports.SYNONYM_UPDATE_TITLE = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.updateSynonymSetTitle', {
  defaultMessage: 'Manage synonym set'
});
const CREATE_SUCCESS = exports.CREATE_SUCCESS = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.createSuccessMessage', {
  defaultMessage: 'Synonym set created'
});
const UPDATE_SUCCESS = exports.UPDATE_SUCCESS = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.updateSuccessMessage', {
  defaultMessage: 'Synonym set updated'
});
const DELETE_CONFIRMATION = exports.DELETE_CONFIRMATION = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.deleteConfirmationMessage', {
  defaultMessage: 'Are you sure you want to delete this synonym set?'
});
const DELETE_SUCCESS = exports.DELETE_SUCCESS = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.deleteSuccessMessage', {
  defaultMessage: 'Synonym set deleted'
});
const SYNONYM_IMPACT_MESSAGE = exports.SYNONYM_IMPACT_MESSAGE = _i18n.i18n.translate('xpack.enterpriseSearch.appSearch.engine.synonyms.impactDescription', {
  defaultMessage: 'The set will impact your results shortly.'
});
