// @kbn/infra-plugin/public/components/asset_details/tabs/profiling/description_callout.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DescriptionCallout = DescriptionCallout;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
var _useLocalStorage = _interopRequireDefault(require("react-use/lib/useLocalStorage"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LOCAL_STORAGE_KEY = 'infra-profiling-description-callout-dismissed';
function DescriptionCallout() {
  const [isDismissed, setIsDismissed] = (0, _useLocalStorage.default)(LOCAL_STORAGE_KEY, false);
  const onDismissClick = (0, _react.useCallback)(() => setIsDismissed(true), [setIsDismissed]);
  if (isDismissed) {
    return null;
  }
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    title: _i18n.i18n.translate('xpack.infra.profiling.descriptionCallout.title', {
      defaultMessage: 'Displaying Resource Consumption for this Host'
    }),
    iconType: "iInCircle"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, _i18n.i18n.translate('xpack.infra.profiling.descriptionCallout.body', {
    defaultMessage: 'Universal Profiling helps you optimize resource usage and find performance bottlenecks by showing which lines of code are consuming resources on your host, down to the application code, kernel, and third-party libraries.'
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    "data-test-subj": "infraProfilingDescriptionCalloutLearnMoreLink",
    href: "https://www.elastic.co/guide/en/observability/current/profiling-get-started.html",
    external: true,
    target: "_blank"
  }, _i18n.i18n.translate('xpack.infra.profiling.descriptionCallout.learnMore', {
    defaultMessage: 'Learn more'
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "infraDescriptionCalloutDismissButton",
    onClick: onDismissClick
  }, _i18n.i18n.translate('xpack.infra.profiling.descriptionCallout.dismiss', {
    defaultMessage: 'Dismiss'
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null));
}
