// @kbn/fleet-plugin/public/components/agent_enrollment_flyout/steps/run_k8s_apply_command_step.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.KubernetesManifestApplyStep = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _styledComponents = _interopRequireDefault(require("styled-components"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const KubernetesManifestApplyStep = ({
  isComplete,
  fullCopyButton,
  onCopy
}) => {
  const [copyButtonClicked, setCopyButtonClicked] = (0, _react.useState)(false);
  // Otherwise the copy button is over the text
  const CommandCode = _styledComponents.default.pre({
    overflow: 'auto'
  });
  const onTextAreaClick = () => {
    if (onCopy) onCopy();
  };
  const onCopyButtonClick = copy => {
    copy();
    setCopyButtonClicked(true);
    if (onCopy) onCopy();
  };
  const status = isComplete ? 'complete' : undefined;
  return {
    status,
    title: _i18n.i18n.translate('xpack.fleet.agentEnrollment.stepKubernetesApplyManifest', {
      defaultMessage: 'Run the apply command'
    }),
    children: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    }), /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.agentEnrollment.kubernetesCommandInstructions",
      defaultMessage: "From the directory where the manifest is downloaded, run the apply command."
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "m"
    })), /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
      onClick: onTextAreaClick,
      fontSize: "m",
      isCopyable: !fullCopyButton,
      paddingSize: "m",
      css: `
            max-width: 1100px;
          `
    }, /*#__PURE__*/_react.default.createElement(CommandCode, null, "kubectl apply -f elastic-agent-managed-kubernetes.yml")), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "s"
    }), fullCopyButton && /*#__PURE__*/_react.default.createElement(_eui.EuiCopy, {
      textToCopy: "kubectl apply -f elastic-agent-managed-kubernetes.yml"
    }, copy => /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
      color: "primary",
      iconType: "copyClipboard",
      size: "m",
      onClick: () => onCopyButtonClick(copy)
    }, copyButtonClicked ? /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.enrollmentInstructions.copyButtonClicked",
      defaultMessage: "Copied"
    }) : /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.fleet.enrollmentInstructions.copyButton",
      defaultMessage: "Copy to clipboard"
    }))))
  };
};
exports.KubernetesManifestApplyStep = KubernetesManifestApplyStep;
