// @kbn/data-plugin/server/saved_objects/search_telemetry.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.searchTelemetry = void 0;
var _coreSavedObjectsServer = require("@kbn/core-saved-objects-server");
var _to_v7_12_ = require("./migrations/to_v7_12_0");
var _search_telemetry = require("./schemas/search_telemetry");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const searchTelemetry = exports.searchTelemetry = {
  name: 'search-telemetry',
  indexPattern: _coreSavedObjectsServer.ANALYTICS_SAVED_OBJECT_INDEX,
  namespaceType: 'agnostic',
  hidden: false,
  mappings: {
    dynamic: false,
    properties: {}
  },
  migrations: {
    '7.12.0': _to_v7_12_.migrate712
  },
  schemas: {
    '8.8.0': _search_telemetry.SCHEMA_SEARCH_TELEMETRY_V8_8_0
  }
};
