// @kbn/triggers-actions-ui-plugin/public/application/sections/alert_summary_widget/components/alert_summary_widget_loader.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertSummaryWidgetLoader = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AlertSummaryWidgetLoader = ({
  fullSize,
  isLoadingWithoutChart
}) => {
  return /*#__PURE__*/_react.default.createElement("div", {
    style: {
      minHeight: isLoadingWithoutChart ? 44 : fullSize ? 238 : 224,
      display: 'flex',
      alignItems: 'center',
      justifyContent: isLoadingWithoutChart ? 'flex-start' : 'center'
    }
  }, isLoadingWithoutChart ? /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
    size: "m"
  }) : /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingChart, {
    size: "l",
    "data-test-subj": "alertSummaryWidgetLoading"
  }));
};
exports.AlertSummaryWidgetLoader = AlertSummaryWidgetLoader;
