// @kbn/triggers-actions-ui-plugin/public/application/sections/alert_summary_widget/components/alert_summary_widget_error.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertSummaryWidgetError = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AlertSummaryWidgetError = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    "data-test-subj": "alertSummaryWidgetError",
    iconType: "warning",
    color: "danger",
    title: /*#__PURE__*/_react.default.createElement("h5", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.triggersActionsUI.sections.alertsSummaryWidget.errorTitle",
      defaultMessage: "Unable to load the alerts summary"
    })),
    body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.triggersActionsUI.sections.alertsSummaryWidget.errorBody",
      defaultMessage: "There was an error loading the alerts summary. Contact your administrator for help."
    }))
  });
};
exports.AlertSummaryWidgetError = AlertSummaryWidgetError;
