// @kbn/rollup-plugin/common/ui_metric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIM_SHOW_DETAILS_CLICK = exports.UIM_JOB_STOP_MANY = exports.UIM_JOB_STOP = exports.UIM_JOB_START_MANY = exports.UIM_JOB_START = exports.UIM_JOB_DELETE_MANY = exports.UIM_JOB_DELETE = exports.UIM_JOB_CREATE = exports.UIM_DETAIL_PANEL_TERMS_TAB_CLICK = exports.UIM_DETAIL_PANEL_SUMMARY_TAB_CLICK = exports.UIM_DETAIL_PANEL_METRICS_TAB_CLICK = exports.UIM_DETAIL_PANEL_JSON_TAB_CLICK = exports.UIM_DETAIL_PANEL_HISTOGRAM_TAB_CLICK = exports.UIM_APP_NAME = exports.UIM_APP_LOAD = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UIM_APP_NAME = exports.UIM_APP_NAME = 'rollup_jobs';
const UIM_APP_LOAD = exports.UIM_APP_LOAD = 'app_load';
const UIM_JOB_CREATE = exports.UIM_JOB_CREATE = 'job_create';
const UIM_JOB_DELETE = exports.UIM_JOB_DELETE = 'job_delete';
const UIM_JOB_DELETE_MANY = exports.UIM_JOB_DELETE_MANY = 'job_delete_many';
const UIM_JOB_START = exports.UIM_JOB_START = 'job_start';
const UIM_JOB_START_MANY = exports.UIM_JOB_START_MANY = 'job_start_many';
const UIM_JOB_STOP = exports.UIM_JOB_STOP = 'job_stop';
const UIM_JOB_STOP_MANY = exports.UIM_JOB_STOP_MANY = 'job_stop_many';
const UIM_SHOW_DETAILS_CLICK = exports.UIM_SHOW_DETAILS_CLICK = 'show_details_click';
const UIM_DETAIL_PANEL_SUMMARY_TAB_CLICK = exports.UIM_DETAIL_PANEL_SUMMARY_TAB_CLICK = 'detail_panel_summary_tab_click';
const UIM_DETAIL_PANEL_TERMS_TAB_CLICK = exports.UIM_DETAIL_PANEL_TERMS_TAB_CLICK = 'detail_panel_terms_tab_click';
const UIM_DETAIL_PANEL_HISTOGRAM_TAB_CLICK = exports.UIM_DETAIL_PANEL_HISTOGRAM_TAB_CLICK = 'detail_panel_histogram_tab_click';
const UIM_DETAIL_PANEL_METRICS_TAB_CLICK = exports.UIM_DETAIL_PANEL_METRICS_TAB_CLICK = 'detail_panel_metrics_tab_click';
const UIM_DETAIL_PANEL_JSON_TAB_CLICK = exports.UIM_DETAIL_PANEL_JSON_TAB_CLICK = 'detail_panel_json_tab_click';
