// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/pick_fields_step/components/bucket_span_estimator/bucket_span_estimator.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.BucketSpanEstimator = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _mlAnomalyUtils = require("@kbn/ml-anomaly-utils");
var _job_creator = require("../../../../../common/job_creator");
var _job_creator_context = require("../../../job_creator_context");
var _estimate_bucket_span = require("./estimate_bucket_span");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const BucketSpanEstimator = ({
  setEstimating
}) => {
  const {
    jobCreator,
    jobCreatorUpdated
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const {
    status,
    estimateBucketSpan
  } = (0, _estimate_bucket_span.useEstimateBucketSpan)();
  const [noDetectors, setNoDetectors] = (0, _react.useState)(jobCreator.detectors.length === 0);
  const [isUsingMlCategory, setIsUsingMlCategory] = (0, _react.useState)(checkIsUsingMlCategory());
  (0, _react.useEffect)(() => {
    setEstimating(status === _estimate_bucket_span.ESTIMATE_STATUS.RUNNING);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [status]);
  (0, _react.useEffect)(() => {
    setNoDetectors(jobCreator.detectors.length === 0);
    setIsUsingMlCategory(checkIsUsingMlCategory());
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreatorUpdated]);
  function checkIsUsingMlCategory() {
    return (0, _job_creator.isAdvancedJobCreator)(jobCreator) && jobCreator.detectors.some(d => {
      if (d.partition_field_name === _mlAnomalyUtils.MLCATEGORY || d.over_field_name === _mlAnomalyUtils.MLCATEGORY || d.by_field_name === _mlAnomalyUtils.MLCATEGORY) {
        return true;
      }
    });
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    disabled: status === _estimate_bucket_span.ESTIMATE_STATUS.RUNNING || noDetectors === true || isUsingMlCategory === true,
    onClick: estimateBucketSpan
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.ml.newJob.wizard.pickFieldsStep.bucketSpanEstimatorButton",
    defaultMessage: "Estimate bucket span"
  }));
};
exports.BucketSpanEstimator = BucketSpanEstimator;
