// @kbn/stack-connectors-plugin/public/connector_types/resilient/resilient_connectors.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _public = require("@kbn/triggers-actions-ui-plugin/public");
var i18n = _interopRequireWildcard(require("./translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const configFormSchema = [{
  id: 'apiUrl',
  label: i18n.API_URL_LABEL,
  isUrlField: true
}, {
  id: 'orgId',
  label: i18n.ORG_ID_LABEL
}];
const secretsFormSchema = [{
  id: 'apiKeyId',
  label: i18n.API_KEY_ID_LABEL
}, {
  id: 'apiKeySecret',
  label: i18n.API_KEY_SECRET_LABEL,
  isPasswordField: true
}];
const ResilientConnectorFields = ({
  readOnly,
  isEdit
}) => {
  return /*#__PURE__*/_react.default.createElement(_public.SimpleConnectorForm, {
    isEdit: isEdit,
    readOnly: readOnly,
    configFormSchema: configFormSchema,
    secretsFormSchema: secretsFormSchema
  });
};

// eslint-disable-next-line import/no-default-export
exports.default = ResilientConnectorFields;
module.exports = exports.default;
