// @kbn/synthetics-plugin/public/apps/synthetics/components/monitor_details/monitor_errors/failed_tests_by_step.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FailedTestsByStep = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _panel_with_title = require("../../common/components/panel_with_title");
var _use_selected_monitor = require("../hooks/use_selected_monitor");
var _use_failed_tests_by_step = require("../hooks/use_failed_tests_by_step");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FailedTestsByStep = ({
  time
}) => {
  const {
    failedSteps,
    loading
  } = (0, _use_failed_tests_by_step.useFailedTestByStep)(time);
  const {
    monitor
  } = (0, _use_selected_monitor.useSelectedMonitor)();
  if ((monitor === null || monitor === void 0 ? void 0 : monitor.type) !== 'browser') {
    return null;
  }
  if (loading && !failedSteps) {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiSkeletonText, {
      lines: 3
    });
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: 1
  }, /*#__PURE__*/_react.default.createElement(_panel_with_title.PanelWithTitle, {
    title: FAILED_TESTS_BY_STEPS_LABEL
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement("div", null, failedSteps === null || failedSteps === void 0 ? void 0 : failedSteps.map(item => /*#__PURE__*/_react.default.createElement(_react.Fragment, {
    key: item.name
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiProgress, {
    valueText: /*#__PURE__*/_react.default.createElement("span", null, _i18n.i18n.translate('xpack.synthetics.monitorDetails.summary.failedTests.count', {
      defaultMessage: 'Failed {count}',
      values: {
        count: item.count
      }
    })),
    max: 100,
    color: "danger",
    size: "l",
    value: item.percent,
    label: `${item.index}. ${item.name}`
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }))))));
};
exports.FailedTestsByStep = FailedTestsByStep;
const FAILED_TESTS_BY_STEPS_LABEL = _i18n.i18n.translate('xpack.synthetics.errors.failedTests.byStep', {
  defaultMessage: 'Failed tests by step'
});
