// @kbn/index-lifecycle-management-plugin/public/application/sections/edit_policy/components/phases/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "ColdPhase", {
  enumerable: true,
  get: function () {
    return _cold_phase.ColdPhase;
  }
});
Object.defineProperty(exports, "DeletePhase", {
  enumerable: true,
  get: function () {
    return _delete_phase.DeletePhase;
  }
});
Object.defineProperty(exports, "FrozenPhase", {
  enumerable: true,
  get: function () {
    return _frozen_phase.FrozenPhase;
  }
});
Object.defineProperty(exports, "HotPhase", {
  enumerable: true,
  get: function () {
    return _hot_phase.HotPhase;
  }
});
Object.defineProperty(exports, "Phase", {
  enumerable: true,
  get: function () {
    return _phase.Phase;
  }
});
Object.defineProperty(exports, "WarmPhase", {
  enumerable: true,
  get: function () {
    return _warm_phase.WarmPhase;
  }
});
var _hot_phase = require("./hot_phase");
var _warm_phase = require("./warm_phase");
var _cold_phase = require("./cold_phase");
var _frozen_phase = require("./frozen_phase");
var _delete_phase = require("./delete_phase");
var _phase = require("./phase");
