// @kbn/expression-gauge-plugin/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GaugeTicksPositions = exports.GaugeShapes = exports.GaugeLabelMajorModes = exports.GaugeColorModes = exports.GaugeCentralMajorModes = exports.GAUGE_FUNCTION_RENDERER_NAME = exports.EXPRESSION_GAUGE_NAME = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const EXPRESSION_GAUGE_NAME = exports.EXPRESSION_GAUGE_NAME = 'gauge';
const GAUGE_FUNCTION_RENDERER_NAME = exports.GAUGE_FUNCTION_RENDERER_NAME = 'gauge_renderer';
const GaugeShapes = exports.GaugeShapes = {
  HORIZONTAL_BULLET: 'horizontalBullet',
  VERTICAL_BULLET: 'verticalBullet',
  SEMI_CIRCLE: 'semiCircle',
  ARC: 'arc',
  CIRCLE: 'circle'
};
const GaugeTicksPositions = exports.GaugeTicksPositions = {
  HIDDEN: 'hidden',
  AUTO: 'auto',
  BANDS: 'bands'
};
const GaugeLabelMajorModes = exports.GaugeLabelMajorModes = {
  AUTO: 'auto',
  CUSTOM: 'custom',
  NONE: 'none'
};
const GaugeCentralMajorModes = exports.GaugeCentralMajorModes = {
  AUTO: 'auto',
  CUSTOM: 'custom',
  NONE: 'none'
};
const GaugeColorModes = exports.GaugeColorModes = {
  PALETTE: 'palette',
  NONE: 'none'
};
