// @kbn/console-plugin/public/lib/autocomplete_entities/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Alias", {
  enumerable: true,
  get: function () {
    return _alias.Alias;
  }
});
Object.defineProperty(exports, "ComponentTemplate", {
  enumerable: true,
  get: function () {
    return _component_template.ComponentTemplate;
  }
});
Object.defineProperty(exports, "DataStream", {
  enumerable: true,
  get: function () {
    return _data_stream.DataStream;
  }
});
Object.defineProperty(exports, "IndexTemplate", {
  enumerable: true,
  get: function () {
    return _index_template.IndexTemplate;
  }
});
Object.defineProperty(exports, "LegacyTemplate", {
  enumerable: true,
  get: function () {
    return _legacy.LegacyTemplate;
  }
});
Object.defineProperty(exports, "Mapping", {
  enumerable: true,
  get: function () {
    return _mapping.Mapping;
  }
});
var _alias = require("./alias");
var _mapping = require("./mapping");
var _data_stream = require("./data_stream");
var _legacy = require("./legacy");
var _index_template = require("./index_template");
var _component_template = require("./component_template");
