// @kbn/cases-plugin/public/components/case_view/metrics/status.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CaseStatusMetrics = void 0;
var _react = _interopRequireWildcard(require("react"));
var _prettyMs = _interopRequireDefault(require("pretty-ms"));
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
var _api = require("../../../../common/types/api");
var _translations = require("./translations");
var _maybe_date = require("../../formatted_date/maybe_date");
var _formatted_date = require("../../formatted_date");
var _empty_value = require("../../empty_value");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CaseStatusMetrics = exports.CaseStatusMetrics = /*#__PURE__*/_react.default.memo(({
  metrics,
  features
}) => {
  const lifespanMetrics = useGetLifespanMetrics(metrics, features);
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  if (!lifespanMetrics) {
    return null;
  }
  const items = [{
    key: 'case-created',
    component: /*#__PURE__*/_react.default.createElement(CaseStatusMetricsItem, {
      title: _translations.CASE_CREATED,
      value: /*#__PURE__*/_react.default.createElement(CreationDate, {
        date: lifespanMetrics.creationDate
      }),
      euiTheme: euiTheme
    }),
    dataTestSubject: 'case-metrics-lifespan-item-creation-date'
  }, {
    key: 'in-progress-duration',
    component: /*#__PURE__*/_react.default.createElement(CaseStatusMetricsItem, {
      title: _translations.CASE_IN_PROGRESS_DURATION,
      value: getInProgressDuration(lifespanMetrics.statusInfo.inProgressDuration),
      euiTheme: euiTheme
    }),
    dataTestSubject: 'case-metrics-lifespan-item-inProgress-duration'
  }, {
    key: 'open-duration',
    component: /*#__PURE__*/_react.default.createElement(CaseStatusMetricsItem, {
      title: _translations.CASE_OPEN_DURATION,
      value: formatDuration(lifespanMetrics.statusInfo.openDuration),
      euiTheme: euiTheme
    }),
    dataTestSubject: 'case-metrics-lifespan-item-open-duration'
  }, {
    key: 'duration-from-creation-to-close',
    component: /*#__PURE__*/_react.default.createElement(CaseStatusMetricsOpenCloseDuration, {
      title: _translations.CASE_OPEN_TO_CLOSE_DURATION,
      value: getOpenCloseDuration(lifespanMetrics.creationDate, lifespanMetrics.closeDate),
      reopens: lifespanMetrics.statusInfo.reopenDates,
      euiTheme: euiTheme
    }),
    dataTestSubject: 'case-metrics-lifespan-item-open-to-close-duration'
  }];
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: 3
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGrid, {
    columns: 2,
    gutterSize: "s",
    responsive: false
  }, items.map(({
    component,
    dataTestSubject,
    key
  }) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": dataTestSubject,
    key: key
  }, component))));
});
CaseStatusMetrics.displayName = 'CaseStatusMetrics';
const useGetLifespanMetrics = (metrics, features) => {
  return (0, _react.useMemo)(() => {
    var _metrics$lifespan;
    const lifespan = (_metrics$lifespan = metrics === null || metrics === void 0 ? void 0 : metrics.lifespan) !== null && _metrics$lifespan !== void 0 ? _metrics$lifespan : {
      closeDate: '',
      creationDate: '',
      statusInfo: {
        inProgressDuration: 0,
        reopenDates: [],
        openDuration: 0
      }
    };
    if (!features.includes(_api.CaseMetricsFeature.LIFESPAN)) {
      return;
    }
    return lifespan;
  }, [features, metrics]);
};
const CreationDate = /*#__PURE__*/_react.default.memo(({
  date
}) => {
  const creationDate = (0, _maybe_date.getMaybeDate)(date);
  if (!creationDate.isValid()) {
    return (0, _empty_value.getEmptyCellValue)();
  }
  return /*#__PURE__*/_react.default.createElement(_formatted_date.FormattedRelativePreferenceDate, {
    "data-test-subj": 'case-metrics-lifespan-creation-date',
    value: date,
    stripMs: true
  });
});
CreationDate.displayName = 'CreationDate';
const getInProgressDuration = duration => {
  if (duration <= 0) {
    return (0, _empty_value.getEmptyCellValue)();
  }
  return formatDuration(duration);
};
const formatDuration = milliseconds => {
  return (0, _prettyMs.default)(milliseconds, {
    compact: true,
    verbose: true
  });
};
const getOpenCloseDuration = (openDate, closeDate) => {
  if (closeDate == null) {
    return;
  }
  const openDateObject = (0, _maybe_date.getMaybeDate)(openDate);
  const closeDateObject = (0, _maybe_date.getMaybeDate)(closeDate);
  if (!openDateObject.isValid() || !closeDateObject.isValid()) {
    return;
  }
  return formatDuration(closeDateObject.diff(openDateObject));
};
const CaseStatusMetricsItem = /*#__PURE__*/_react.default.memo(({
  title,
  value,
  euiTheme
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  direction: "column",
  gutterSize: "s",
  responsive: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  css: (0, _react2.css)`
        font-size: ${euiTheme.size.m};
        font-weight: bold;
      `
}, title), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, value)));
CaseStatusMetricsItem.displayName = 'CaseStatusMetricsItem';
const CaseStatusMetricsOpenCloseDuration = /*#__PURE__*/_react.default.memo(({
  title,
  value,
  reopens,
  euiTheme
}) => {
  const valueText = getOpenCloseDurationText(value, reopens);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "s",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    css: (0, _react2.css)`
          font-size: ${euiTheme.size.m};
          font-weight: bold;
        `
  }, title), value != null && caseWasReopened(reopens) ? /*#__PURE__*/_react.default.createElement(ValueWithExplanationIcon, {
    value: valueText,
    explanationValues: reopens
  }) : /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, valueText));
});
CaseStatusMetricsOpenCloseDuration.displayName = 'OpenCloseDuration';
const getOpenCloseDurationText = (value, reopens) => {
  if (value == null) {
    return (0, _empty_value.getEmptyCellValue)();
  } else if (reopens.length > 0) {
    return `${value} ${_translations.CASE_REOPENED}`;
  }
  return value;
};
const caseWasReopened = reopens => {
  return reopens.length > 0;
};
const ValueWithExplanationIcon = /*#__PURE__*/_react.default.memo(({
  value,
  explanationValues
}) => {
  const content = /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, _translations.CASE_REOPENED_ON, explanationValues.map((explanationValue, index) => {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, {
      key: `explanation-value-${index}`
    }, /*#__PURE__*/_react.default.createElement(_formatted_date.FormattedRelativePreferenceDate, {
      "data-test-subj": `case-metrics-lifespan-reopen-${index}`,
      value: explanationValue,
      stripMs: true
    }), isNotLastItem(index, explanationValues.length) ? /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "xs"
    }) : null);
  }));
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "case-metrics-lifespan-reopen-icon"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    responsive: false,
    gutterSize: "s",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, value), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIconTip, {
    content: content,
    position: "right"
  }))));
});
ValueWithExplanationIcon.displayName = 'ValueWithExplanationIcon';
const isNotLastItem = (index, arrayLength) => index + 1 < arrayLength;
