// @kbn/security-solution-plugin/public/flyout/document_details/shared/constants/panel_keys.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentDetailsRuleOverviewPanelKey = exports.DocumentDetailsRightPanelKey = exports.DocumentDetailsPreviewPanelKey = exports.DocumentDetailsLeftPanelKey = exports.DocumentDetailsIsolateHostPanelKey = exports.DocumentDetailsAlertReasonPanelKey = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DocumentDetailsRightPanelKey = exports.DocumentDetailsRightPanelKey = 'document-details-right';
const DocumentDetailsLeftPanelKey = exports.DocumentDetailsLeftPanelKey = 'document-details-left';
const DocumentDetailsPreviewPanelKey = exports.DocumentDetailsPreviewPanelKey = 'document-details-preview';
const DocumentDetailsIsolateHostPanelKey = exports.DocumentDetailsIsolateHostPanelKey = 'document-details-isolate-host';
const DocumentDetailsAlertReasonPanelKey = exports.DocumentDetailsAlertReasonPanelKey = 'document-details-alert-reason';
const DocumentDetailsRuleOverviewPanelKey = exports.DocumentDetailsRuleOverviewPanelKey = 'document-details-rule-overview';
