// @kbn/slo-plugin/public/pages/slo_settings/slo_settings.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SloSettingsPage = SloSettingsPage;
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _public = require("@kbn/observability-shared-plugin/public");
var _settings_form = require("./settings_form");
var _kibana_react = require("../../utils/kibana_react");
var _use_plugin_context = require("../../hooks/use_plugin_context");
var _paths = require("../../../common/locators/paths");
var _header_menu = require("../../components/header_menu/header_menu");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function SloSettingsPage() {
  const {
    http: {
      basePath
    }
  } = (0, _kibana_react.useKibana)().services;
  const {
    ObservabilityPageTemplate
  } = (0, _use_plugin_context.usePluginContext)();
  (0, _public.useBreadcrumbs)([{
    href: basePath.prepend(_paths.paths.slosSettings),
    text: _i18n.i18n.translate('xpack.slo.breadcrumbs.slosSettingsText', {
      defaultMessage: 'SLOs Settings'
    })
  }]);
  return /*#__PURE__*/_react.default.createElement(ObservabilityPageTemplate, {
    "data-test-subj": "slosSettingsPage",
    pageHeader: {
      pageTitle: _i18n.i18n.translate('xpack.slo.pageHeader.title.', {
        defaultMessage: 'SLOs Settings'
      }),
      rightSideItems: []
    }
  }, /*#__PURE__*/_react.default.createElement(_header_menu.HeaderMenu, null), /*#__PURE__*/_react.default.createElement(_settings_form.SettingsForm, null));
}
