// @kbn/upgrade-assistant-plugin/public/shared_imports.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "APP_WRAPPER_CLASS", {
  enumerable: true,
  get: function () {
    return _public4.APP_WRAPPER_CLASS;
  }
});
Object.defineProperty(exports, "AuthorizationContext", {
  enumerable: true,
  get: function () {
    return _public.AuthorizationContext;
  }
});
Object.defineProperty(exports, "AuthorizationProvider", {
  enumerable: true,
  get: function () {
    return _public.AuthorizationProvider;
  }
});
Object.defineProperty(exports, "GlobalFlyout", {
  enumerable: true,
  get: function () {
    return _public.GlobalFlyout;
  }
});
Object.defineProperty(exports, "KibanaContextProvider", {
  enumerable: true,
  get: function () {
    return _public3.KibanaContextProvider;
  }
});
Object.defineProperty(exports, "KibanaRenderContextProvider", {
  enumerable: true,
  get: function () {
    return _reactKibanaContextRender.KibanaRenderContextProvider;
  }
});
Object.defineProperty(exports, "NotAuthorizedSection", {
  enumerable: true,
  get: function () {
    return _public.NotAuthorizedSection;
  }
});
Object.defineProperty(exports, "RedirectAppLinks", {
  enumerable: true,
  get: function () {
    return _sharedUxLinkRedirectApp.RedirectAppLinks;
  }
});
Object.defineProperty(exports, "SectionLoading", {
  enumerable: true,
  get: function () {
    return _public.SectionLoading;
  }
});
Object.defineProperty(exports, "Storage", {
  enumerable: true,
  get: function () {
    return _public2.Storage;
  }
});
Object.defineProperty(exports, "WithPrivileges", {
  enumerable: true,
  get: function () {
    return _public.WithPrivileges;
  }
});
Object.defineProperty(exports, "reactRouterNavigate", {
  enumerable: true,
  get: function () {
    return _public3.reactRouterNavigate;
  }
});
Object.defineProperty(exports, "sendRequest", {
  enumerable: true,
  get: function () {
    return _public.sendRequest;
  }
});
Object.defineProperty(exports, "useRequest", {
  enumerable: true,
  get: function () {
    return _public.useRequest;
  }
});
var _public = require("@kbn/es-ui-shared-plugin/public");
var _public2 = require("@kbn/kibana-utils-plugin/public");
var _public3 = require("@kbn/kibana-react-plugin/public");
var _reactKibanaContextRender = require("@kbn/react-kibana-context-render");
var _sharedUxLinkRedirectApp = require("@kbn/shared-ux-link-redirect-app");
var _public4 = require("@kbn/core/public");
