// @kbn/cases-components/src/tooltip/skeleton.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Skeleton = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const SkeletonComponent = () => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    css: {
      width: 240
    },
    "data-test-subj": "tooltip-loading-content"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSkeletonTitle, {
    size: "xxxs",
    css: {
      width: 70,
      marginBottom: '12px'
    }
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSkeletonText, {
    lines: 3
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, {
    margin: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }));
};
SkeletonComponent.displayName = 'Skeleton';
const Skeleton = exports.Skeleton = SkeletonComponent;
