// @kbn/triggers-actions-ui-plugin/public/application/sections/action_connector_form/edit_connector_flyout/header.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FlyoutHeader = void 0;
var _react = _interopRequireWildcard(require("react"));
var _react2 = require("@emotion/react");
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _i18n = require("@kbn/i18n");
var _translations = require("../../translations");
var _types = require("../../../../types");
var _kibana = require("../../../../common/lib/kibana");
var _capabilities = require("../../../lib/capabilities");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FlyoutHeaderComponent = ({
  icon,
  isExperimental = false,
  isPreconfigured,
  connectorName,
  connectorTypeId,
  connectorTypeDesc,
  selectedTab,
  setTab
}) => {
  const {
    application: {
      capabilities
    }
  } = (0, _kibana.useKibana)().services;
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  const canExecute = (0, _capabilities.hasExecuteActionsCapability)(capabilities, connectorTypeId);
  const setConfigurationTab = (0, _react.useCallback)(() => {
    setTab(_types.EditConnectorTabs.Configuration);
  }, [setTab]);
  const setTestTab = (0, _react.useCallback)(() => {
    setTab(_types.EditConnectorTabs.Test);
  }, [setTab]);
  const setRulesTab = (0, _react.useCallback)(() => {
    setTab(_types.EditConnectorTabs.Rules);
  }, [setTab]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlyoutHeader, {
    hasBorder: true,
    "data-test-subj": "edit-connector-flyout-header"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    alignItems: "center"
  }, icon ? /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: icon,
    size: "m",
    "data-test-subj": "edit-connector-flyout-header-icon"
  })) : null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, isPreconfigured ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    justifyContent: "center",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("h3", {
    id: "flyoutTitle"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    defaultMessage: "{connectorName}",
    id: "xpack.triggersActionsUI.sections.preconfiguredConnectorForm.flyoutTitle",
    values: {
      connectorName
    }
  })))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiBetaBadge, {
    label: "Preconfigured",
    "data-test-subj": "preconfiguredBadge",
    tooltipContent: _i18n.i18n.translate('xpack.triggersActionsUI.sections.preconfiguredConnectorForm.tooltipContent', {
      defaultMessage: 'This connector is preconfigured and cannot be edited'
    })
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, isExperimental && /*#__PURE__*/_react.default.createElement(_eui.EuiBetaBadge, {
    label: _translations.TECH_PREVIEW_LABEL,
    tooltipContent: _translations.TECH_PREVIEW_DESCRIPTION
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    defaultMessage: "{connectorTypeDesc}",
    id: "xpack.triggersActionsUI.sections.editConnectorForm.actionTypeDescription",
    values: {
      connectorTypeDesc
    }
  }))) : /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    justifyContent: "center",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("h3", {
    id: "flyoutTitle"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    defaultMessage: "Edit connector",
    id: "xpack.triggersActionsUI.sections.editConnectorForm.flyoutPreconfiguredTitle"
  })))), isExperimental && /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiBetaBadge, {
    label: _translations.TECH_PREVIEW_LABEL,
    tooltipContent: _translations.TECH_PREVIEW_DESCRIPTION
  }))))), /*#__PURE__*/_react.default.createElement(_eui.EuiTabs, {
    className: "connectorEditFlyoutTabs",
    bottomBorder: false,
    css: (0, _react2.css)`
          margin-bottom: -${euiTheme.size.l};
        `
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTab, {
    onClick: setConfigurationTab,
    "data-test-subj": "configureConnectorTab",
    isSelected: _types.EditConnectorTabs.Configuration === selectedTab
  }, _i18n.i18n.translate('xpack.triggersActionsUI.sections.editConnectorForm.tabText', {
    defaultMessage: 'Configuration'
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiTab, {
    onClick: setRulesTab,
    "data-test-subj": "rulesConnectorTab",
    isSelected: _types.EditConnectorTabs.Rules === selectedTab
  }, _i18n.i18n.translate('xpack.triggersActionsUI.sections.rulesConnectorList.tabText', {
    defaultMessage: 'Rules'
  })), canExecute && /*#__PURE__*/_react.default.createElement(_eui.EuiTab, {
    onClick: setTestTab,
    "data-test-subj": "testConnectorTab",
    isSelected: _types.EditConnectorTabs.Test === selectedTab
  }, _i18n.i18n.translate('xpack.triggersActionsUI.sections.testConnectorForm.tabText', {
    defaultMessage: 'Test'
  }))));
};
const FlyoutHeader = exports.FlyoutHeader = /*#__PURE__*/(0, _react.memo)(FlyoutHeaderComponent);
