// @kbn/enterprise-search-plugin/public/applications/enterprise_search_content/api/pipelines/simulate_ml_inference_pipeline_processors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.simulateMlInferencePipeline = exports.SimulateMlInterfacePipelineApiLogic = void 0;
var _create_api_logic = require("../../../shared/api_logic/create_api_logic");
var _http = require("../../../shared/http");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const simulateMlInferencePipeline = async ({
  docs,
  indexName,
  pipeline
}) => {
  const route = `/internal/enterprise_search/indices/${indexName}/ml_inference/pipeline_processors/simulate`;
  return await _http.HttpLogic.values.http.post(route, {
    body: JSON.stringify({
      docs,
      pipeline: {
        description: pipeline.description,
        processors: pipeline.processors
      }
    })
  });
};
exports.simulateMlInferencePipeline = simulateMlInferencePipeline;
const SimulateMlInterfacePipelineApiLogic = exports.SimulateMlInterfacePipelineApiLogic = (0, _create_api_logic.createApiLogic)(['simulate_ml_inference_pipeline_api_logic'], simulateMlInferencePipeline);
