// @kbn/security-solution-plugin/common/detection_engine/rule_management/rule_fields.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TAGS_FIELD = exports.RULE_PARAMS_FIELDS = exports.RULE_NAME_FIELD = exports.PARAMS_TYPE_FIELD = exports.PARAMS_IMMUTABLE_FIELD = exports.LAST_RUN_OUTCOME_FIELD = exports.ENABLED_FIELD = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const RULE_NAME_FIELD = exports.RULE_NAME_FIELD = 'alert.attributes.name';
const RULE_PARAMS_FIELDS = exports.RULE_PARAMS_FIELDS = {
  INDEX: 'alert.attributes.params.index',
  TACTIC_ID: 'alert.attributes.params.threat.tactic.id',
  TACTIC_NAME: 'alert.attributes.params.threat.tactic.name',
  TECHNIQUE_ID: 'alert.attributes.params.threat.technique.id',
  TECHNIQUE_NAME: 'alert.attributes.params.threat.technique.name',
  SUBTECHNIQUE_ID: 'alert.attributes.params.threat.technique.subtechnique.id',
  SUBTECHNIQUE_NAME: 'alert.attributes.params.threat.technique.subtechnique.name'
};
const ENABLED_FIELD = exports.ENABLED_FIELD = 'alert.attributes.enabled';
const TAGS_FIELD = exports.TAGS_FIELD = 'alert.attributes.tags';
const PARAMS_TYPE_FIELD = exports.PARAMS_TYPE_FIELD = 'alert.attributes.params.type';
const PARAMS_IMMUTABLE_FIELD = exports.PARAMS_IMMUTABLE_FIELD = 'alert.attributes.params.immutable';
const LAST_RUN_OUTCOME_FIELD = exports.LAST_RUN_OUTCOME_FIELD = 'alert.attributes.lastRun.outcome';
