// @kbn/index-management-plugin/public/application/components/mappings_editor/components/load_mappings/load_from_json_button.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadMappingsFromJsonButton = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _load_mappings_provider = require("./load_mappings_provider");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LoadMappingsFromJsonButton = ({
  onJson,
  esNodesPlugins
}) => /*#__PURE__*/_react.default.createElement(_load_mappings_provider.LoadMappingsProvider, {
  onJson: onJson,
  esNodesPlugins: esNodesPlugins
}, openModal => /*#__PURE__*/_react.default.createElement(_eui.EuiButtonEmpty, {
  onClick: openModal,
  size: "s"
}, _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.loadFromJsonButtonLabel', {
  defaultMessage: 'Load JSON'
})));
exports.LoadMappingsFromJsonButton = LoadMappingsFromJsonButton;
