// @kbn/apm-plugin/public/components/app/profiling_overview/host_names_filter_warning.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HostnamesFilterWarning = HostnamesFilterWarning;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function HostnamesFilterWarning({
  hostNames = []
}) {
  function renderTooltipOptions() {
    return /*#__PURE__*/_react.default.createElement("ul", null, hostNames.map(hostName => /*#__PURE__*/_react.default.createElement("li", {
      key: hostName
    }, `- ${hostName}`)));
  }
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs",
    color: "subdued"
  }, _i18n.i18n.translate('xpack.apm.profiling.flamegraph.filteredLabel', {
    defaultMessage: "Displaying profiling insights from the service's host(s)"
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    content: renderTooltipOptions()
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "questionInCircle"
  }))));
}
