// @kbn/apm-plugin/public/components/app/profiling_overview/profiling_hosts_top_functions.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProfilingHostsTopNFunctions = ProfilingHostsTopNFunctions;
var _eui = require("@elastic/eui");
var _public = require("@kbn/observability-shared-plugin/public");
var _react = _interopRequireDefault(require("react"));
var _apm = require("../../../../common/es_fields/apm");
var _kuery_utils = require("../../../../common/utils/kuery_utils");
var _use_fetcher = require("../../../hooks/use_fetcher");
var _top_functions_link = require("../../shared/profiling/top_functions/top_functions_link");
var _host_names_filter_warning = require("./host_names_filter_warning");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function ProfilingHostsTopNFunctions({
  serviceName,
  start,
  end,
  environment,
  startIndex,
  endIndex,
  dataSource,
  kuery,
  rangeFrom,
  rangeTo
}) {
  const {
    data,
    status
  } = (0, _use_fetcher.useFetcher)(callApmApi => {
    if (dataSource) {
      return callApmApi('GET /internal/apm/services/{serviceName}/profiling/hosts/functions', {
        params: {
          path: {
            serviceName
          },
          query: {
            start,
            end,
            environment,
            startIndex,
            endIndex,
            documentType: dataSource.documentType,
            rollupInterval: dataSource.rollupInterval,
            kuery
          }
        }
      });
    }
  }, [dataSource, serviceName, start, end, environment, startIndex, endIndex, kuery]);
  const hostNamesKueryFormat = (0, _kuery_utils.toKueryFilterFormat)(_apm.HOST_NAME, (data === null || data === void 0 ? void 0 : data.hostNames) || []);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_host_names_filter_warning.HostnamesFilterWarning, {
    hostNames: data === null || data === void 0 ? void 0 : data.hostNames
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_top_functions_link.ProfilingTopNFunctionsLink, {
    kuery: (0, _kuery_utils.mergeKueries)([`(${hostNamesKueryFormat})`, kuery]),
    rangeFrom: rangeFrom,
    rangeTo: rangeTo,
    justifyContent: "flexEnd"
  }))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null), /*#__PURE__*/_react.default.createElement(_public.EmbeddableFunctions, {
    data: data === null || data === void 0 ? void 0 : data.functions,
    isLoading: (0, _use_fetcher.isPending)(status),
    rangeFrom: new Date(start).valueOf(),
    rangeTo: new Date(end).valueOf()
  }));
}
