// @kbn/transform-plugin/server/routes/api/stop_transforms/stop_transforms.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.stopTransforms = stopTransforms;
var _transform = require("../../../../common/types/transform");
var _constants = require("../../../../common/constants");
var _error_utils = require("../../utils/error_utils");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

async function stopTransforms(transformsInfo, esClient) {
  const results = {};
  for (const transformInfo of transformsInfo) {
    const transformId = transformInfo.id;
    try {
      await esClient.transform.stopTransform({
        transform_id: transformId,
        force: transformInfo.state !== undefined ? transformInfo.state === _constants.TRANSFORM_STATE.FAILED : false,
        wait_for_completion: true
      });
      results[transformId] = {
        success: true
      };
    } catch (e) {
      if ((0, _error_utils.isRequestTimeout)(e)) {
        return (0, _error_utils.fillResultsWithTimeouts)({
          results,
          id: transformId,
          items: transformsInfo,
          action: _transform.TRANSFORM_ACTIONS.STOP
        });
      }
      results[transformId] = {
        success: false,
        error: e.meta.body.error
      };
    }
  }
  return results;
}
