// @kbn/synthetics-plugin/public/apps/synthetics/state/monitor_details/selectors.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.selectorMonitorDetailsState = exports.selectorError = exports.selectStatusFilter = exports.selectSelectedLocationId = exports.selectPingsLoading = exports.selectPingsError = exports.selectMonitorPingsMetadata = exports.selectLastRunMetadata = void 0;
var _reselect = require("reselect");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const getState = appState => appState.monitorDetails;
const selectorMonitorDetailsState = exports.selectorMonitorDetailsState = (0, _reselect.createSelector)(getState, state => state);
const selectorError = exports.selectorError = (0, _reselect.createSelector)(getState, state => state.error);
const selectSelectedLocationId = exports.selectSelectedLocationId = (0, _reselect.createSelector)(getState, state => state.selectedLocationId);
const selectLastRunMetadata = exports.selectLastRunMetadata = (0, _reselect.createSelector)(getState, state => state.lastRun);
const selectPingsLoading = exports.selectPingsLoading = (0, _reselect.createSelector)(getState, state => state.pings.loading);
const selectMonitorPingsMetadata = exports.selectMonitorPingsMetadata = (0, _reselect.createSelector)(getState, state => state.pings);
const selectPingsError = exports.selectPingsError = (0, _reselect.createSelector)(getState, state => state.error);
const selectStatusFilter = exports.selectStatusFilter = (0, _reselect.createSelector)(getState, state => state.statusFilter);
