// @kbn/canvas-plugin/public/expression_types/arg_types/container_style/simple_template.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SimpleTemplate = void 0;
var _react = _interopRequireDefault(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _color_picker_popover = require("../../../components/color_picker_popover");
var _i18n = require("../../../../i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const {
  ContainerStyle: strings
} = _i18n.ArgTypesStrings;
const SimpleTemplate = ({
  getArgValue,
  setArgValue,
  workpad
}) => /*#__PURE__*/_react.default.createElement("div", {
  style: {
    fontSize: 0
  }
}, /*#__PURE__*/_react.default.createElement(_color_picker_popover.ColorPickerPopover, {
  value: getArgValue('backgroundColor'),
  onChange: color => setArgValue('backgroundColor', color),
  colors: workpad.colors,
  anchorPosition: "leftCenter",
  ariaLabel: strings.getDisplayName()
}));
exports.SimpleTemplate = SimpleTemplate;
SimpleTemplate.displayName = 'ContainerStyleArgSimpleInput';
SimpleTemplate.propTypes = {
  getArgValue: _propTypes.default.func.isRequired,
  setArgValue: _propTypes.default.func.isRequired,
  workpad: _propTypes.default.shape({
    colors: _propTypes.default.array.isRequired
  })
};
