// @kbn/logs-explorer-plugin/public/components/virtual_columns/column_tooltips/actions_column_tooltip.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionsColumnTooltip = void 0;
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
var _eui = require("@elastic/eui");
var _uiTheme = require("@kbn/ui-theme");
var _translations = require("../../common/translations");
var _tooltip_button = require("./tooltip_button");
var constants = _interopRequireWildcard(require("../../../../common/constants"));
var _field_with_token = require("./field_with_token");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const spacingCSS = (0, _react2.css)`
  margin-bottom: ${_uiTheme.euiThemeVars.euiSizeS};
`;
const ActionsColumnTooltip = () => {
  return /*#__PURE__*/_react.default.createElement(_tooltip_button.TooltipButton, {
    displayText: _translations.actionsLabelLowerCase,
    popoverTitle: _translations.actionsLabel
  }, /*#__PURE__*/_react.default.createElement("div", {
    style: {
      width: '230px'
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s",
    css: spacingCSS
  }, /*#__PURE__*/_react.default.createElement("p", null, _translations.actionsHeaderTooltipParagraph)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    responsive: false,
    alignItems: "baseline",
    justifyContent: "flexStart",
    gutterSize: "s",
    css: spacingCSS
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "expand",
    size: "s"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("p", null, _translations.actionsHeaderTooltipExpandAction)))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    responsive: false,
    alignItems: "baseline",
    justifyContent: "flexStart",
    gutterSize: "s",
    css: spacingCSS
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "indexClose",
    size: "s",
    color: "danger"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("p", null, _translations.actionsHeaderTooltipDegradedAction)))), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    responsive: false,
    alignItems: "baseline",
    justifyContent: "flexStart",
    gutterSize: "s",
    css: spacingCSS
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "apmTrace",
    size: "s"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s"
  }, /*#__PURE__*/_react.default.createElement("p", null, _translations.actionsHeaderTooltipStacktraceAction)))), /*#__PURE__*/_react.default.createElement("div", {
    style: {
      marginLeft: '15px'
    }
  }, [constants.ERROR_STACK_TRACE, constants.ERROR_EXCEPTION_STACKTRACE, constants.ERROR_LOG_STACKTRACE].map(field => /*#__PURE__*/_react.default.createElement(_field_with_token.FieldWithToken, {
    field: field,
    key: field
  })))));
};
exports.ActionsColumnTooltip = ActionsColumnTooltip;
