// @kbn/saved-objects-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "SavedObjectSaveModal", {
  enumerable: true,
  get: function () {
    return _save_modal.SavedObjectSaveModal;
  }
});
Object.defineProperty(exports, "SavedObjectSaveModalOrigin", {
  enumerable: true,
  get: function () {
    return _save_modal.SavedObjectSaveModalOrigin;
  }
});
Object.defineProperty(exports, "checkForDuplicateTitle", {
  enumerable: true,
  get: function () {
    return _saved_object.checkForDuplicateTitle;
  }
});
Object.defineProperty(exports, "isErrorNonFatal", {
  enumerable: true,
  get: function () {
    return _saved_object.isErrorNonFatal;
  }
});
exports.plugin = void 0;
Object.defineProperty(exports, "saveWithConfirmation", {
  enumerable: true,
  get: function () {
    return _saved_object.saveWithConfirmation;
  }
});
Object.defineProperty(exports, "showSaveModal", {
  enumerable: true,
  get: function () {
    return _save_modal.showSaveModal;
  }
});
var _plugin = require("./plugin");
var _save_modal = require("./save_modal");
var _saved_object = require("./saved_object");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const plugin = () => new _plugin.SavedObjectsPublicPlugin();
exports.plugin = plugin;
