// @kbn/integration-assistant-plugin/public/common/components/integration_image_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IntegrationImageHeader = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _sharedUxPageKibanaTemplate = require("@kbn/shared-ux-page-kibana-template");
var _react2 = require("@emotion/react");
var _integrations_light = _interopRequireDefault(require("../images/integrations_light.svg"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const headerCss = (0, _react2.css)`
  > div {
    padding-block: 0;
  }
`;
const imageCss = (0, _react2.css)`
  width: 318px;
  height: 183px;
  object-fit: cover;
  object-position: center top;
`;
const IntegrationImageHeader = exports.IntegrationImageHeader = /*#__PURE__*/_react.default.memo(() => {
  return /*#__PURE__*/_react.default.createElement(_sharedUxPageKibanaTemplate.KibanaPageTemplate.Header, {
    css: headerCss
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    alignItems: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xl"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiImage, {
    alt: "create integration background",
    src: _integrations_light.default,
    css: imageCss
  }))));
});
IntegrationImageHeader.displayName = 'IntegrationImageHeader';
