// @kbn/enterprise-search-plugin/public/applications/shared/constants/units.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WEEKS_UNIT_LABEL = exports.MONTHS_UNIT_LABEL = exports.MINUTES_UNIT_LABEL = exports.HOURS_UNIT_LABEL = exports.DAYS_UNIT_LABEL = exports.DAYS_OF_WEEK_LABELS = exports.ALL_DAYS_LABEL = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ALL_DAYS_LABEL = exports.ALL_DAYS_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.allDaysLabel', {
  defaultMessage: 'All days'
});
const MINUTES_UNIT_LABEL = exports.MINUTES_UNIT_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.minutesLabel', {
  defaultMessage: 'Minutes'
});
const HOURS_UNIT_LABEL = exports.HOURS_UNIT_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.hoursLabel', {
  defaultMessage: 'Hours'
});
const DAYS_UNIT_LABEL = exports.DAYS_UNIT_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.daysLabel', {
  defaultMessage: 'Days'
});
const WEEKS_UNIT_LABEL = exports.WEEKS_UNIT_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.weeksLabel', {
  defaultMessage: 'Weeks'
});
const MONTHS_UNIT_LABEL = exports.MONTHS_UNIT_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.units.monthsLabel', {
  defaultMessage: 'Months'
});
const DAYS_OF_WEEK_LABELS = exports.DAYS_OF_WEEK_LABELS = {
  SUNDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.sunday', {
    defaultMessage: 'Sunday'
  }),
  MONDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.monday', {
    defaultMessage: 'Monday'
  }),
  TUESDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.tuesday', {
    defaultMessage: 'Tuesday'
  }),
  WEDNESDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.wednesday', {
    defaultMessage: 'Wednesday'
  }),
  THURSDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.thursday', {
    defaultMessage: 'Thursday'
  }),
  FRIDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.friday', {
    defaultMessage: 'Friday'
  }),
  SATURDAY: _i18n.i18n.translate('xpack.enterpriseSearch.units.daysOfWeekLabel.saturday', {
    defaultMessage: 'Saturday'
  })
};
