// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/pick_fields_step/components/geo_view/metric_selection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GeoDetector = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
var _job_creator_context = require("../../../job_creator_context");
var _geo_field = require("../geo_field");
var _geo_map_examples = require("./geo_map_examples");
var _kibana = require("../../../../../../../contexts/kibana");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const GeoDetector = ({
  setIsValid
}) => {
  const {
    jobCreator: jc,
    jobCreatorUpdated,
    chartLoader
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const jobCreator = jc;
  const [fieldValues, setFieldValues] = (0, _react.useState)([]);
  const [layerList, setLayerList] = (0, _react.useState)([]);
  const {
    services: {
      notifications: toasts
    }
  } = (0, _kibana.useMlKibana)();
  const {
    mapLoader
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  (0, _react.useEffect)(() => {
    let valid = false;
    if (jobCreator.geoField !== null) {
      valid = true;
    }
    setIsValid(valid);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreatorUpdated]);

  // Load example field values when split field changes
  // changes to fieldValues here will trigger the card effect
  (0, _react.useEffect)(() => {
    if (jobCreator.splitField !== null) {
      chartLoader.loadFieldExampleValues(jobCreator.splitField, jobCreator.runtimeMappings, jobCreator.datafeedConfig.indices_options).then(setFieldValues).catch(error => {
        // @ts-ignore
        toasts.addDanger({
          title: _i18n.i18n.translate('xpack.ml.newJob.geoWizard.fieldValuesFetchErrorTitle', {
            defaultMessage: 'Error fetching field example values: {error}',
            values: {
              error
            }
          })
        });
      });
    } else {
      setFieldValues([]);
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreator.splitField]);

  // Update the layer list  with updated geo points upon refresh
  (0, _react.useEffect)(() => {
    async function getMapLayersForGeoJob() {
      if (jobCreator.geoField) {
        var _jobCreator$savedSear;
        const {
          query
        } = (_jobCreator$savedSear = jobCreator.savedSearchQuery) !== null && _jobCreator$savedSear !== void 0 ? _jobCreator$savedSear : {};
        const layers = await mapLoader.getMapLayersForGeoJob(jobCreator.geoField, jobCreator.splitField, fieldValues, query);
        setLayerList(layers);
      }
    }
    getMapLayersForGeoJob();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [jobCreator.geoField, jobCreator.splitField, fieldValues]);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, jobCreator.geoField !== null && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_geo_map_examples.GeoMapExamples, {
    geoField: jobCreator.geoField,
    splitField: jobCreator.splitField,
    fieldValues: fieldValues,
    geoAgg: jobCreator.geoAgg,
    layerList: layerList
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, null)), /*#__PURE__*/_react.default.createElement(_geo_field.GeoField, null));
};
exports.GeoDetector = GeoDetector;
