// @kbn/search-homepage/public/components/search_homepage_body.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchHomepageBody = void 0;
var _react = _interopRequireDefault(require("react"));
var _sharedUxPageKibanaTemplate = require("@kbn/shared-ux-page-kibana-template");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const SearchHomepageBody = () => /*#__PURE__*/_react.default.createElement(_sharedUxPageKibanaTemplate.KibanaPageTemplate.Section, {
  alignment: "top",
  restrictWidth: false,
  grow: true,
  css: {
    position: 'relative'
  },
  contentProps: {
    css: {
      display: 'flex',
      flexGrow: 1,
      position: 'absolute',
      inset: 0
    }
  },
  paddingSize: "none",
  className: "eui-fullHeight"
}, /*#__PURE__*/_react.default.createElement("div", null));
exports.SearchHomepageBody = SearchHomepageBody;
