// @kbn/monitoring-plugin/public/application/pages/elasticsearch/index_advanced_page.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ElasticsearchIndexAdvancedPage = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _lodash = require("lodash");
var _reactRouterDom = require("react-router-dom");
var _public = require("@kbn/kibana-react-plugin/public");
var _global_state_context = require("../../contexts/global_state_context");
var _setup_mode = require("../../../components/renderers/setup_mode");
var _setup_mode_context = require("../../../components/setup_mode/setup_mode_context");
var _use_charts = require("../../hooks/use_charts");
var _item_template = require("./item_template");
var _advanced = require("../../../components/elasticsearch/index/advanced");
var _fetch_alerts = require("../../../lib/fetch_alerts");
var _constants = require("../../../../common/constants");
var _use_breadcrumbs = require("../../hooks/use_breadcrumbs");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ElasticsearchIndexAdvancedPage = ({
  clusters
}) => {
  var _services$data2;
  const globalState = (0, _react.useContext)(_global_state_context.GlobalStateContext);
  const {
    generate: generateBreadcrumbs
  } = (0, _use_breadcrumbs.useBreadcrumbContainerContext)();
  const {
    services
  } = (0, _public.useKibana)();
  const {
    index
  } = (0, _reactRouterDom.useParams)();
  const {
    zoomInfo,
    onBrush
  } = (0, _use_charts.useCharts)();
  const clusterUuid = globalState.cluster_uuid;
  const ccs = globalState.ccs;
  const [data, setData] = (0, _react.useState)({});
  const [alerts, setAlerts] = (0, _react.useState)({});
  const cluster = (0, _lodash.find)(clusters, {
    cluster_uuid: clusterUuid
  });
  (0, _react.useEffect)(() => {
    if (cluster) {
      generateBreadcrumbs(cluster.cluster_name, {
        inElasticsearch: true,
        name: 'indices',
        instance: index
      });
    }
  }, [cluster, generateBreadcrumbs, index]);
  const title = _i18n.i18n.translate('xpack.monitoring.elasticsearch.index.advanced.title', {
    defaultMessage: 'Elasticsearch - Indices - {indexName} - Advanced',
    values: {
      indexName: index
    }
  });
  const getPageData = (0, _react.useCallback)(async () => {
    var _services$data, _services$http;
    const bounds = (_services$data = services.data) === null || _services$data === void 0 ? void 0 : _services$data.query.timefilter.timefilter.getBounds();
    const url = `../api/monitoring/v1/clusters/${clusterUuid}/elasticsearch/indices/${index}`;
    if ((_services$http = services.http) !== null && _services$http !== void 0 && _services$http.fetch && clusterUuid) {
      var _services$http2;
      const response = await ((_services$http2 = services.http) === null || _services$http2 === void 0 ? void 0 : _services$http2.fetch(url, {
        method: 'POST',
        body: JSON.stringify({
          ccs,
          timeRange: {
            min: bounds.min.toISOString(),
            max: bounds.max.toISOString()
          },
          is_advanced: true
        })
      }));
      setData(response);
      const alertsResponse = await (0, _fetch_alerts.fetchAlerts)({
        fetch: services.http.fetch,
        alertTypeIds: [_constants.RULE_LARGE_SHARD_SIZE],
        filters: [{
          shardIndex: index
        }],
        clusterUuid,
        timeRange: {
          min: bounds.min.valueOf(),
          max: bounds.max.valueOf()
        }
      });
      setAlerts(alertsResponse);
    }
  }, [(_services$data2 = services.data) === null || _services$data2 === void 0 ? void 0 : _services$data2.query.timefilter.timefilter, services.http, clusterUuid, index, ccs]);
  return /*#__PURE__*/_react.default.createElement(_item_template.ItemTemplate, {
    title: title,
    getPageData: getPageData,
    id: index,
    pageType: "indices",
    pageTitle: index
  }, /*#__PURE__*/_react.default.createElement(_setup_mode.SetupModeRenderer, {
    productName: _constants.ELASTICSEARCH_SYSTEM_ID,
    render: ({
      setupMode,
      flyoutComponent,
      bottomBarComponent
    }) => /*#__PURE__*/_react.default.createElement(_setup_mode_context.SetupModeContext.Provider, {
      value: {
        setupModeSupported: true
      }
    }, flyoutComponent, /*#__PURE__*/_react.default.createElement(_advanced.AdvancedIndex, {
      setupMode: setupMode,
      alerts: alerts,
      indexSummary: data.indexSummary,
      metrics: data.metrics,
      onBrush: onBrush,
      zoomInfo: zoomInfo
    }), bottomBarComponent)
  }));
};
exports.ElasticsearchIndexAdvancedPage = ElasticsearchIndexAdvancedPage;
