// @kbn/synthetics-plugin/common/field_names.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.URL_PORT = exports.URL_FULL = exports.STATE_ID = exports.OBSERVER_GEO_NAME = exports.MONITOR_TYPE = exports.MONITOR_NAME = exports.MONITOR_ID = exports.ERROR_MESSAGE = exports.CERT_VALID_NOT_BEFORE = exports.CERT_VALID_NOT_AFTER = exports.CERT_ISSUER_NAME = exports.CERT_HASH_SHA256 = exports.CERT_COMMON_NAME = exports.AGENT_NAME = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AGENT_NAME = exports.AGENT_NAME = 'agent.name';
const MONITOR_ID = exports.MONITOR_ID = 'monitor.id';
const MONITOR_NAME = exports.MONITOR_NAME = 'monitor.name';
const MONITOR_TYPE = exports.MONITOR_TYPE = 'monitor.type';
const URL_FULL = exports.URL_FULL = 'url.full';
const URL_PORT = exports.URL_PORT = 'url.port';
const OBSERVER_GEO_NAME = exports.OBSERVER_GEO_NAME = 'observer.geo.name';
const ERROR_MESSAGE = exports.ERROR_MESSAGE = 'error.message';
const STATE_ID = exports.STATE_ID = 'monitor.state.id';
const CERT_COMMON_NAME = exports.CERT_COMMON_NAME = 'tls.server.x509.subject.common_name';
const CERT_ISSUER_NAME = exports.CERT_ISSUER_NAME = 'tls.server.x509.issuer.common_name';
const CERT_VALID_NOT_AFTER = exports.CERT_VALID_NOT_AFTER = 'tls.server.x509.not_after';
const CERT_VALID_NOT_BEFORE = exports.CERT_VALID_NOT_BEFORE = 'tls.server.x509.not_before';
const CERT_HASH_SHA256 = exports.CERT_HASH_SHA256 = 'tls.server.hash.sha256';
