// @kbn/expression-partition-vis-plugin/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WAFFLE_VIS_EXPRESSION_NAME = exports.TREEMAP_VIS_EXPRESSION_NAME = exports.PLUGIN_NAME = exports.PLUGIN_ID = exports.PIE_VIS_EXPRESSION_NAME = exports.PARTITION_VIS_RENDERER_NAME = exports.PARTITION_LABELS_VALUE = exports.PARTITION_LABELS_FUNCTION = exports.MOSAIC_VIS_EXPRESSION_NAME = exports.DEFAULT_PERCENT_DECIMALS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const PLUGIN_ID = exports.PLUGIN_ID = 'expressionPartitionVis';
const PLUGIN_NAME = exports.PLUGIN_NAME = 'expressionPartitionVis';
const PIE_VIS_EXPRESSION_NAME = exports.PIE_VIS_EXPRESSION_NAME = 'pieVis';
const TREEMAP_VIS_EXPRESSION_NAME = exports.TREEMAP_VIS_EXPRESSION_NAME = 'treemapVis';
const MOSAIC_VIS_EXPRESSION_NAME = exports.MOSAIC_VIS_EXPRESSION_NAME = 'mosaicVis';
const WAFFLE_VIS_EXPRESSION_NAME = exports.WAFFLE_VIS_EXPRESSION_NAME = 'waffleVis';
const PARTITION_VIS_RENDERER_NAME = exports.PARTITION_VIS_RENDERER_NAME = 'partitionVis';
const PARTITION_LABELS_VALUE = exports.PARTITION_LABELS_VALUE = 'partitionLabelsValue';
const PARTITION_LABELS_FUNCTION = exports.PARTITION_LABELS_FUNCTION = 'partitionLabels';
const DEFAULT_PERCENT_DECIMALS = exports.DEFAULT_PERCENT_DECIMALS = 2;
