// @kbn/security-solution-plugin/common/api/detection_engine/prebuilt_rules/urls.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.REVIEW_RULE_UPGRADE_URL = exports.REVIEW_RULE_INSTALLATION_URL = exports.PREBUILT_RULES_URL = exports.PREBUILT_RULES_STATUS_URL = exports.PERFORM_RULE_UPGRADE_URL = exports.PERFORM_RULE_INSTALLATION_URL = exports.GET_PREBUILT_RULES_STATUS_URL = void 0;
var _constants = require("../../../constants");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const OLD_BASE_URL = `${_constants.DETECTION_ENGINE_RULES_URL}/prepackaged`;
const NEW_BASE_URL = `${_constants.INTERNAL_DETECTION_ENGINE_URL}/prebuilt_rules`;
const PREBUILT_RULES_URL = exports.PREBUILT_RULES_URL = OLD_BASE_URL;
const PREBUILT_RULES_STATUS_URL = exports.PREBUILT_RULES_STATUS_URL = `${OLD_BASE_URL}/_status`;
const GET_PREBUILT_RULES_STATUS_URL = exports.GET_PREBUILT_RULES_STATUS_URL = `${NEW_BASE_URL}/status`;
const REVIEW_RULE_UPGRADE_URL = exports.REVIEW_RULE_UPGRADE_URL = `${NEW_BASE_URL}/upgrade/_review`;
const PERFORM_RULE_UPGRADE_URL = exports.PERFORM_RULE_UPGRADE_URL = `${NEW_BASE_URL}/upgrade/_perform`;
const REVIEW_RULE_INSTALLATION_URL = exports.REVIEW_RULE_INSTALLATION_URL = `${NEW_BASE_URL}/installation/_review`;
const PERFORM_RULE_INSTALLATION_URL = exports.PERFORM_RULE_INSTALLATION_URL = `${NEW_BASE_URL}/installation/_perform`;
