// @kbn/aiops-plugin/public/components/log_categorization/log_categorization_for_embeddable/log_categorization_wrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.LogCategorizationWrapper = void 0;
var _react = _interopRequireWildcard(require("react"));
var _mlDatePicker = require("@kbn/ml-date-picker");
var _mlLocalStorage = require("@kbn/ml-local-storage");
var _lodash = require("lodash");
var _public = require("@kbn/data-plugin/public");
var _public2 = require("@kbn/kibana-utils-plugin/public");
var _storage = require("../../../types/storage");
var _use_aiops_app_context = require("../../../hooks/use_aiops_app_context");
var _log_categorization_for_embeddable = require("./log_categorization_for_embeddable");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const localStorage = new _public2.Storage(window.localStorage);
const LogCategorizationWrapper = ({
  deps,
  props,
  embeddingOrigin
}) => {
  const I18nContext = deps.i18n.Context;
  const aiopsAppContextValue = {
    embeddingOrigin,
    ...deps
  };
  const datePickerDeps = {
    ...(0, _lodash.pick)(deps, ['data', 'http', 'notifications', 'theme', 'uiSettings', 'i18n']),
    uiSettingsKeys: _public.UI_SETTINGS
  };
  return /*#__PURE__*/_react.default.createElement(I18nContext, null, /*#__PURE__*/_react.default.createElement(_use_aiops_app_context.AiopsAppContext.Provider, {
    value: aiopsAppContextValue
  }, /*#__PURE__*/_react.default.createElement(_mlDatePicker.DatePickerContextProvider, datePickerDeps, /*#__PURE__*/_react.default.createElement(_mlLocalStorage.StorageContextProvider, {
    storage: localStorage,
    storageKeys: _storage.AIOPS_STORAGE_KEYS
  }, /*#__PURE__*/_react.default.createElement(_react.Suspense, {
    fallback: null
  }, /*#__PURE__*/_react.default.createElement(_log_categorization_for_embeddable.LogCategorizationEmbeddable, {
    input: props.input,
    renderViewModeToggle: props.renderViewModeToggle
  }))))));
};

// eslint-disable-next-line import/no-default-export
exports.LogCategorizationWrapper = LogCategorizationWrapper;
var _default = exports.default = LogCategorizationWrapper;
