// @kbn/alerting-plugin/public/services/maintenance_windows_api/create.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.createMaintenanceWindow = createMaintenanceWindow;
var _common = require("../../../common");
var _transform_maintenance_window_response = require("./transform_maintenance_window_response");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const transformCreateBodySchema = createParams => {
  return {
    title: createParams.title,
    duration: createParams.duration,
    r_rule: createParams.rRule,
    ...(createParams.categoryIds !== undefined ? {
      category_ids: createParams.categoryIds
    } : {}),
    ...(createParams.scopedQuery !== undefined ? {
      scoped_query: createParams.scopedQuery
    } : {})
  };
};
async function createMaintenanceWindow({
  http,
  createParams
}) {
  const res = await http.post(`${_common.INTERNAL_BASE_ALERTING_API_PATH}/rules/maintenance_window`, {
    body: JSON.stringify(transformCreateBodySchema(createParams))
  });
  return (0, _transform_maintenance_window_response.transformMaintenanceWindowResponse)(res);
}
