// @kbn/uptime-plugin/public/legacy_uptime/components/overview/alerts/monitor_expressions/status_expression_select.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StatusExpressionSelect = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _down_number_select = require("./down_number_select");
var _time_expression_select = require("./time_expression_select");
var _translations = require("./translations");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StatusExpressionSelect = ({
  ruleParams,
  hasFilters,
  setRuleParams
}) => {
  var _ruleParams$shouldChe;
  const [isEnabled, setIsEnabled] = (0, _react.useState)((_ruleParams$shouldChe = ruleParams.shouldCheckStatus) !== null && _ruleParams$shouldChe !== void 0 ? _ruleParams$shouldChe : true);
  (0, _react.useEffect)(() => {
    setRuleParams('shouldCheckStatus', isEnabled);
  }, [isEnabled, setRuleParams]);
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "s"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiCheckbox, {
    id: "statusEnabled",
    label: _translations.statusExpLabels.ENABLED_CHECKBOX,
    checked: isEnabled,
    onChange: () => setIsEnabled(!isEnabled)
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_down_number_select.DownNoExpressionSelect, {
    defaultNumTimes: ruleParams.numTimes,
    hasFilters: hasFilters,
    isEnabled: isEnabled,
    setRuleParams: setRuleParams
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_time_expression_select.TimeExpressionSelect, {
    defaultTimerangeUnit: ruleParams.timerangeUnit,
    defaultTimerangeCount: ruleParams.timerangeCount,
    isEnabled: isEnabled,
    setRuleParams: setRuleParams
  })));
};
exports.StatusExpressionSelect = StatusExpressionSelect;
