// @kbn/deeplinks-observability/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SYNTHETICS_APP_ID = exports.SLO_APP_ID = exports.OBSERVABILITY_OVERVIEW_APP_ID = exports.OBSERVABILITY_ONBOARDING_APP_ID = exports.OBSERVABILITY_LOGS_EXPLORER_APP_ID = exports.OBLT_UX_APP_ID = exports.OBLT_PROFILING_APP_ID = exports.METRICS_APP_ID = exports.LOGS_APP_ID = exports.APM_APP_ID = exports.AI_ASSISTANT_APP_ID = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const LOGS_APP_ID = exports.LOGS_APP_ID = 'logs';
const OBSERVABILITY_LOGS_EXPLORER_APP_ID = exports.OBSERVABILITY_LOGS_EXPLORER_APP_ID = 'observability-logs-explorer';
const OBSERVABILITY_OVERVIEW_APP_ID = exports.OBSERVABILITY_OVERVIEW_APP_ID = 'observability-overview';
const METRICS_APP_ID = exports.METRICS_APP_ID = 'metrics';
const APM_APP_ID = exports.APM_APP_ID = 'apm';
const SYNTHETICS_APP_ID = exports.SYNTHETICS_APP_ID = 'synthetics';
const OBSERVABILITY_ONBOARDING_APP_ID = exports.OBSERVABILITY_ONBOARDING_APP_ID = 'observabilityOnboarding';
const SLO_APP_ID = exports.SLO_APP_ID = 'slo';
const AI_ASSISTANT_APP_ID = exports.AI_ASSISTANT_APP_ID = 'observabilityAIAssistant';
const OBLT_UX_APP_ID = exports.OBLT_UX_APP_ID = 'ux';
const OBLT_PROFILING_APP_ID = exports.OBLT_PROFILING_APP_ID = 'profiling';
