// @kbn/deeplinks-observability/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LOGS_APP_ID: true,
  OBSERVABILITY_LOGS_EXPLORER_APP_ID: true,
  OBSERVABILITY_ONBOARDING_APP_ID: true,
  OBSERVABILITY_OVERVIEW_APP_ID: true,
  AI_ASSISTANT_APP_ID: true,
  OBLT_UX_APP_ID: true,
  OBLT_PROFILING_APP_ID: true
};
Object.defineProperty(exports, "AI_ASSISTANT_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.AI_ASSISTANT_APP_ID;
  }
});
Object.defineProperty(exports, "LOGS_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.LOGS_APP_ID;
  }
});
Object.defineProperty(exports, "OBLT_PROFILING_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.OBLT_PROFILING_APP_ID;
  }
});
Object.defineProperty(exports, "OBLT_UX_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.OBLT_UX_APP_ID;
  }
});
Object.defineProperty(exports, "OBSERVABILITY_LOGS_EXPLORER_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.OBSERVABILITY_LOGS_EXPLORER_APP_ID;
  }
});
Object.defineProperty(exports, "OBSERVABILITY_ONBOARDING_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.OBSERVABILITY_ONBOARDING_APP_ID;
  }
});
Object.defineProperty(exports, "OBSERVABILITY_OVERVIEW_APP_ID", {
  enumerable: true,
  get: function () {
    return _constants.OBSERVABILITY_OVERVIEW_APP_ID;
  }
});
var _constants = require("./constants");
var _locators = require("./locators");
Object.keys(_locators).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _locators[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function () {
      return _locators[key];
    }
  });
});
