// @kbn/links-plugin/common/content_management/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "CONTENT_ID", {
  enumerable: true,
  get: function () {
    return _constants.CONTENT_ID;
  }
});
Object.defineProperty(exports, "DASHBOARD_LINK_TYPE", {
  enumerable: true,
  get: function () {
    return _latest.DASHBOARD_LINK_TYPE;
  }
});
Object.defineProperty(exports, "EXTERNAL_LINK_TYPE", {
  enumerable: true,
  get: function () {
    return _latest.EXTERNAL_LINK_TYPE;
  }
});
Object.defineProperty(exports, "LATEST_VERSION", {
  enumerable: true,
  get: function () {
    return _constants.LATEST_VERSION;
  }
});
Object.defineProperty(exports, "LINKS_HORIZONTAL_LAYOUT", {
  enumerable: true,
  get: function () {
    return _latest.LINKS_HORIZONTAL_LAYOUT;
  }
});
Object.defineProperty(exports, "LINKS_VERTICAL_LAYOUT", {
  enumerable: true,
  get: function () {
    return _latest.LINKS_VERTICAL_LAYOUT;
  }
});
exports.LinksV1 = void 0;
var _constants = require("../constants");
var _latest = require("./latest");
var _LinksV = _interopRequireWildcard(require("./v1"));
exports.LinksV1 = _LinksV;
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
