// @elastic/eui/lib/components/page/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "EuiPage", {
  enumerable: true,
  get: function get() {
    return _page.EuiPage;
  }
});
Object.defineProperty(exports, "EuiPageBody", {
  enumerable: true,
  get: function get() {
    return _page_body.EuiPageBody;
  }
});
Object.defineProperty(exports, "EuiPageHeader", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeader;
  }
});
Object.defineProperty(exports, "EuiPageHeaderContent", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeaderContent;
  }
});
Object.defineProperty(exports, "EuiPageHeaderSection", {
  enumerable: true,
  get: function get() {
    return _page_header.EuiPageHeaderSection;
  }
});
Object.defineProperty(exports, "EuiPageSection", {
  enumerable: true,
  get: function get() {
    return _page_section.EuiPageSection;
  }
});
Object.defineProperty(exports, "EuiPageSidebar", {
  enumerable: true,
  get: function get() {
    return _page_sidebar.EuiPageSidebar;
  }
});
var _page = require("./page");
var _page_body = require("./page_body");
var _page_header = require("./page_header");
var _page_section = require("./page_section");
var _page_sidebar = require("./page_sidebar");
