// @kbn/security-solution-plugin/public/common/components/import_data_modal/action_connectors_warning/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ActionConnectorWarnings = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _kibana_react = require("../../../lib/kibana/kibana_react");
var i18n = _interopRequireWildcard(require("../translations"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ActionConnectorWarningsComponent = ({
  actionConnectorsWarnings,
  importedActionConnectorsCount
}) => {
  const {
    http
  } = (0, _kibana_react.useKibana)().services;
  if (!importedActionConnectorsCount || !actionConnectorsWarnings.length) return null;
  const {
    actionPath,
    message,
    buttonLabel
  } = actionConnectorsWarnings[0];
  return /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    "data-test-subj": "actionConnectorsWarningsCallOut",
    size: "m",
    heading: "h2",
    iconType: "warning",
    title: /*#__PURE__*/_react.default.createElement("span", {
      "data-test-subj": "actionConnectorsWarningsCallOutTitle"
    }, i18n.ACTION_CONNECTORS_WARNING_TITLE(importedActionConnectorsCount)),
    color: "warning"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "none"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    "data-test-subj": "actionConnectorsWarningsCallOutMessage"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "xs"
  }, message)), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none",
    direction: "columnReverse",
    alignItems: "flexEnd"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    "data-test-subj": "actionConnectorsWarningsCallOutButton",
    color: "warning",
    href: http.basePath.prepend(actionPath)
  }, buttonLabel || i18n.ACTION_CONNECTORS_WARNING_BUTTON)))));
};
ActionConnectorWarningsComponent.displayName = 'ActionConnectorWarningsComponent';
const ActionConnectorWarnings = exports.ActionConnectorWarnings = /*#__PURE__*/_react.default.memo(ActionConnectorWarningsComponent);
ActionConnectorWarnings.displayName = 'ActionConnectorWarnings';
