// @kbn/osquery-plugin/public/routes/live_queries/list/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LiveQueriesPage = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _react = _interopRequireWildcard(require("react"));
var _kibana = require("../../../common/lib/kibana");
var _actions_table = require("../../../actions/actions_table");
var _layouts = require("../../../components/layouts");
var _use_breadcrumbs = require("../../../common/hooks/use_breadcrumbs");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const LiveQueriesPageComponent = () => {
  (0, _use_breadcrumbs.useBreadcrumbs)('live_queries');
  const LeftColumn = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    direction: "column",
    gutterSize: "m"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("h1", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.liveQueriesHistory.pageTitle",
    defaultMessage: "Live queries history"
  }))))), []);
  return /*#__PURE__*/_react.default.createElement(_layouts.WithHeaderLayout, {
    leftColumn: LeftColumn,
    rightColumn: /*#__PURE__*/_react.default.createElement(NewLiveQueryButton, null),
    rightColumnGrow: false
  }, /*#__PURE__*/_react.default.createElement(_actions_table.ActionsTable, null));
};
const LiveQueriesPage = exports.LiveQueriesPage = /*#__PURE__*/_react.default.memo(LiveQueriesPageComponent);
const NewLiveQueryButton = /*#__PURE__*/_react.default.memo(() => {
  const permissions = (0, _kibana.useKibana)().services.application.capabilities.osquery;
  const newQueryLinkProps = (0, _kibana.useRouterNavigate)('live_queries/new');
  return /*#__PURE__*/_react.default.createElement(_eui.EuiButton, (0, _extends2.default)({
    fill: true
  }, newQueryLinkProps, {
    iconType: "plusInCircle",
    isDisabled: !(permissions.writeLiveQueries || permissions.runSavedQueries && (permissions.readSavedQueries || permissions.readPacks))
  }), /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.osquery.liveQueriesHistory.newLiveQueryButtonLabel",
    defaultMessage: "New live query"
  }));
});
NewLiveQueryButton.displayName = 'NewLiveQueryButton';
