// @kbn/ml-plugin/common/types/storage.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ML_STORAGE_KEYS = exports.ML_OVERVIEW_PANELS = exports.ML_NOTIFICATIONS_LAST_CHECKED_AT = exports.ML_GETTING_STARTED_CALLOUT_DISMISSED = exports.ML_FROZEN_TIER_PREFERENCE = exports.ML_ENTITY_FIELDS_CONFIG = exports.ML_ELSER_CALLOUT_DISMISSED = exports.ML_APPLY_TIME_RANGE_CONFIG = exports.ML_ANOMALY_EXPLORER_PANELS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ML_ENTITY_FIELDS_CONFIG = exports.ML_ENTITY_FIELDS_CONFIG = 'ml.singleMetricViewer.partitionFields';
const ML_APPLY_TIME_RANGE_CONFIG = exports.ML_APPLY_TIME_RANGE_CONFIG = 'ml.jobSelectorFlyout.applyTimeRange';
const ML_GETTING_STARTED_CALLOUT_DISMISSED = exports.ML_GETTING_STARTED_CALLOUT_DISMISSED = 'ml.gettingStarted.isDismissed';
const ML_FROZEN_TIER_PREFERENCE = exports.ML_FROZEN_TIER_PREFERENCE = 'ml.frozenDataTierPreference';
const ML_ANOMALY_EXPLORER_PANELS = exports.ML_ANOMALY_EXPLORER_PANELS = 'ml.anomalyExplorerPanels';
const ML_NOTIFICATIONS_LAST_CHECKED_AT = exports.ML_NOTIFICATIONS_LAST_CHECKED_AT = 'ml.notificationsLastCheckedAt';
const ML_OVERVIEW_PANELS = exports.ML_OVERVIEW_PANELS = 'ml.overviewPanels';
const ML_ELSER_CALLOUT_DISMISSED = exports.ML_ELSER_CALLOUT_DISMISSED = 'ml.elserUpdateCalloutDismissed';
const ML_STORAGE_KEYS = exports.ML_STORAGE_KEYS = [ML_ENTITY_FIELDS_CONFIG, ML_APPLY_TIME_RANGE_CONFIG, ML_GETTING_STARTED_CALLOUT_DISMISSED, ML_FROZEN_TIER_PREFERENCE, ML_ANOMALY_EXPLORER_PANELS, ML_NOTIFICATIONS_LAST_CHECKED_AT, ML_OVERVIEW_PANELS, ML_ELSER_CALLOUT_DISMISSED];
