// @kbn/securitysolution-io-ts-list-types/src/request/internal/create_exception_list_schema/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.internalCreateExceptionListSchema = void 0;
var _securitysolutionListConstants = require("@kbn/securitysolution-list-constants");
var t = _interopRequireWildcard(require("io-ts"));
var _create_exception_list_schema = require("../../create_exception_list_schema");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const internalCreateExceptionListSchema = exports.internalCreateExceptionListSchema = t.intersection([t.exact(t.type({
  type: t.keyof({
    endpoint: null,
    endpoint_events: null,
    endpoint_host_isolation_exceptions: null,
    endpoint_blocklists: null
  })
})), t.exact(t.partial({
  list_id: t.keyof(_securitysolutionListConstants.ENDPOINT_ARTIFACT_LIST_IDS.reduce((mapOfListIds, listId) => {
    mapOfListIds[listId] = null;
    return mapOfListIds;
  }, {}))
})), _create_exception_list_schema.createExceptionListSchema]);

// This type is used after a decode since some things are defaults after a decode.
