// @kbn/painless-lab-plugin/public/application/components/output_pane/output_pane.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.OutputPane = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _output_tab = require("./output_tab");
var _parameters_tab = require("./parameters_tab");
var _context_tab = require("./context_tab");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const OutputPane = ({
  isLoading,
  response
}) => {
  const outputTabLabel = /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "s",
    alignItems: "center",
    responsive: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, isLoading ? /*#__PURE__*/_react.default.createElement(_eui.EuiLoadingSpinner, {
    size: "m"
  }) : response && response.error ? /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "warning",
    color: "danger"
  }) : /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "check",
    color: "success"
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, _i18n.i18n.translate('xpack.painlessLab.outputTabLabel', {
    defaultMessage: 'Output'
  })));
  return /*#__PURE__*/_react.default.createElement("div", {
    className: "painlessLabRightPane"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiTabbedContent, {
    className: "painlessLabRightPane__tabs",
    "data-test-subj": "painlessTabs",
    size: "s",
    tabs: [{
      id: 'output',
      // TODO: Currently this causes an Eui prop error because it is expecting string, but we give it React.ReactNode - should fix.
      name: outputTabLabel,
      content: /*#__PURE__*/_react.default.createElement(_output_tab.OutputTab, {
        response: response
      })
    }, {
      id: 'parameters',
      name: _i18n.i18n.translate('xpack.painlessLab.parametersTabLabel', {
        defaultMessage: 'Parameters'
      }),
      content: /*#__PURE__*/_react.default.createElement(_parameters_tab.ParametersTab, null)
    }, {
      id: 'context',
      name: _i18n.i18n.translate('xpack.painlessLab.contextTabLabel', {
        defaultMessage: 'Context'
      }),
      content: /*#__PURE__*/_react.default.createElement(_context_tab.ContextTab, null)
    }]
  }));
};
exports.OutputPane = OutputPane;
