// @kbn/data-visualizer-plugin/public/application/common/components/stats_table/components/field_data_expanded_row/text_content.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TextContent = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18nReact = require("@kbn/i18n-react");
var _i18n = require("@kbn/i18n");
var _examples_list = require("../../../examples_list");
var _expanded_row_content = require("./expanded_row_content");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const TextContent = ({
  config
}) => {
  const {
    stats
  } = config;
  if (stats === undefined) return null;
  const {
    examples
  } = stats;
  if (examples === undefined) return null;
  const numExamples = examples.length;
  return /*#__PURE__*/_react.default.createElement(_expanded_row_content.ExpandedRowContent, {
    dataTestSubj: 'dataVisualizerTextContent'
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    className: "dvText__wrapper"
  }, numExamples > 0 && /*#__PURE__*/_react.default.createElement(_examples_list.ExamplesList, {
    examples: examples
  }), numExamples === 0 && /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "xs"
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiCallOut, {
    size: "s",
    title: _i18n.i18n.translate('xpack.dataVisualizer.dataGrid.fieldText.noExamplesForFieldsTitle', {
      defaultMessage: 'No examples were obtained for this field'
    }),
    iconType: "warning"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.dataGrid.fieldText.fieldNotPresentDescription",
    defaultMessage: "This field was not present in the {sourceParam} field of documents queried.",
    values: {
      sourceParam: /*#__PURE__*/_react.default.createElement("span", {
        className: "dvExpandedRow__codeContent"
      }, "_source")
    }
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.dataVisualizer.dataGrid.fieldText.fieldMayBePopulatedDescription",
    defaultMessage: "It may be populated, for example, using a {copyToParam} parameter in the document mapping, or be pruned from the {sourceParam} field after indexing through the use of {includesParam} and {excludesParam} parameters.",
    values: {
      copyToParam: /*#__PURE__*/_react.default.createElement("span", {
        className: "dvExpandedRow__codeContent"
      }, "copy_to"),
      sourceParam: /*#__PURE__*/_react.default.createElement("span", {
        className: "dvExpandedRow__codeContent"
      }, "_source"),
      includesParam: /*#__PURE__*/_react.default.createElement("span", {
        className: "dvExpandedRow__codeContent"
      }, "includes"),
      excludesParam: /*#__PURE__*/_react.default.createElement("span", {
        className: "dvExpandedRow__codeContent"
      }, "excludes")
    }
  })))));
};
exports.TextContent = TextContent;
