// @kbn/enterprise-search-plugin/public/applications/workplace_search/views/oauth_authorize/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WRITE_SCOPE_DESCRIPTION = exports.SEARCH_SCOPE_DESCRIPTION = exports.SCOPES_LEAD_IN_MESSAGE = exports.HTTP_REDIRECT_WARNING_MESSAGE = exports.DENY_BUTTON_LABEL = exports.AUTHORIZE_BUTTON_LABEL = exports.AUTHORIZATION_REQUIRED_TITLE = void 0;
var _i18n = require("@kbn/i18n");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const HTTP_REDIRECT_WARNING_MESSAGE = exports.HTTP_REDIRECT_WARNING_MESSAGE = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.httpRedirectWarningMessage', {
  defaultMessage: 'This application is using an insecure redirect URI (http)'
});
const SCOPES_LEAD_IN_MESSAGE = exports.SCOPES_LEAD_IN_MESSAGE = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.scopesLeadInMessage', {
  defaultMessage: 'This application will be able to'
});
const AUTHORIZATION_REQUIRED_TITLE = exports.AUTHORIZATION_REQUIRED_TITLE = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.authorizationTitle', {
  defaultMessage: 'Authorization required'
});
const AUTHORIZE_BUTTON_LABEL = exports.AUTHORIZE_BUTTON_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.authorizeButtonLabel', {
  defaultMessage: 'Authorize'
});
const DENY_BUTTON_LABEL = exports.DENY_BUTTON_LABEL = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.denyButtonLabel', {
  defaultMessage: 'Deny'
});
const SEARCH_SCOPE_DESCRIPTION = exports.SEARCH_SCOPE_DESCRIPTION = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.searchScopeDescription', {
  defaultMessage: 'Search your data'
});
const WRITE_SCOPE_DESCRIPTION = exports.WRITE_SCOPE_DESCRIPTION = _i18n.i18n.translate('xpack.enterpriseSearch.workplaceSearch.oauthAuthorize.writeScopeDescription', {
  defaultMessage: 'Modify your data'
});
