// @kbn/data-plugin/public/search/search_interceptor/search_session_incomplete_warning.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SearchSessionIncompleteWarning = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireDefault(require("react"));
var _i18nReact = require("@kbn/i18n-react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const SearchSessionIncompleteWarning = docLinks => /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "s"
}), "It needs more time to fully render. You can wait here or come back to it later.", /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
  size: "m"
}), /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
  textAlign: "right"
}, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
  href: docLinks.links.search.sessionLimits,
  color: "warning",
  target: "_blank",
  "data-test-subj": "searchSessionIncompleteWarning",
  external: true
}, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
  id: "data.searchSession.warning.readDocs",
  defaultMessage: "Read More"
}))));
exports.SearchSessionIncompleteWarning = SearchSessionIncompleteWarning;
