// @kbn/security-plugin/public/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AuthenticationProvider", {
  enumerable: true,
  get: function () {
    return _use_current_user.AuthenticationProvider;
  }
});
Object.defineProperty(exports, "SecurityApiClientsProvider", {
  enumerable: true,
  get: function () {
    return _security_api_clients_provider.SecurityApiClientsProvider;
  }
});
Object.defineProperty(exports, "useAuthentication", {
  enumerable: true,
  get: function () {
    return _use_current_user.useAuthentication;
  }
});
Object.defineProperty(exports, "useCurrentUser", {
  enumerable: true,
  get: function () {
    return _use_current_user.useCurrentUser;
  }
});
Object.defineProperty(exports, "useSecurityApiClients", {
  enumerable: true,
  get: function () {
    return _security_api_clients_provider.useSecurityApiClients;
  }
});
Object.defineProperty(exports, "useUserProfile", {
  enumerable: true,
  get: function () {
    return _use_current_user.useUserProfile;
  }
});
var _security_api_clients_provider = require("./security_api_clients_provider");
var _use_current_user = require("./use_current_user");
