// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/processor_form/processors/fail.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Fail = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _shared_imports = require("../../../../../../shared_imports");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const {
  emptyField
} = _shared_imports.fieldValidators;
const fieldsConfig = {
  message: {
    type: _shared_imports.FIELD_TYPES.TEXT,
    deserializer: String,
    label: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.failForm.messageFieldLabel', {
      defaultMessage: 'Message'
    }),
    helpText: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.failForm.messageHelpText', {
      defaultMessage: 'Error message returned by the processor.'
    }),
    validations: [{
      validator: emptyField(_i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.failForm.valueRequiredError', {
        defaultMessage: 'A message is required.'
      }))
    }]
  }
};
const Fail = () => {
  return /*#__PURE__*/_react.default.createElement(_shared_imports.UseField, {
    "data-test-subj": "messageField",
    component: _shared_imports.Field,
    config: fieldsConfig.message,
    path: "fields.message"
  });
};
exports.Fail = Fail;
