// @kbn/lens-plugin/public/datasources/form_based/operations/definitions/formula/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "formulaOperation", {
  enumerable: true,
  get: function () {
    return _formula.formulaOperation;
  }
});
Object.defineProperty(exports, "insertOrReplaceFormulaColumn", {
  enumerable: true,
  get: function () {
    return _parse.insertOrReplaceFormulaColumn;
  }
});
Object.defineProperty(exports, "intervalOperation", {
  enumerable: true,
  get: function () {
    return _context_variables.intervalOperation;
  }
});
Object.defineProperty(exports, "mathOperation", {
  enumerable: true,
  get: function () {
    return _math.mathOperation;
  }
});
Object.defineProperty(exports, "nowOperation", {
  enumerable: true,
  get: function () {
    return _context_variables.nowOperation;
  }
});
Object.defineProperty(exports, "timeRangeOperation", {
  enumerable: true,
  get: function () {
    return _context_variables.timeRangeOperation;
  }
});
var _formula = require("./formula");
var _parse = require("./parse");
var _math = require("./math");
var _context_variables = require("./context_variables");
