// @kbn/security-solution-plugin/server/lib/detection_engine/rule_management/logic/bulk_actions/validations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.validateBulkScheduleBackfill = exports.validateBulkEnableRule = exports.validateBulkEditRule = exports.validateBulkDuplicateRule = exports.validateBulkDisableRule = exports.dryRunValidateBulkEditRule = void 0;
var _invariant = require("../../../../../../common/utils/invariant");
var _helpers = require("../../../../../../common/machine_learning/helpers");
var _utils = require("../../../../../../common/detection_engine/utils");
var _constants = require("../../../../../../common/constants");
var _rule_management = require("../../../../../../common/api/detection_engine/rule_management");
var _utils2 = require("./utils");
var _dry_run = require("./dry_run");
var _validation = require("../../../../machine_learning/validation");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * throws ML authorization error wrapped with MACHINE_LEARNING_AUTH error code
 * @param mlAuthz - {@link MlAuthz}
 * @param ruleType - {@link RuleType}
 */
const throwMlAuthError = (mlAuthz, ruleType) => (0, _dry_run.throwDryRunError)(async () => (0, _validation.throwAuthzError)(await mlAuthz.validateRuleType(ruleType)), _constants.BulkActionsDryRunErrCode.MACHINE_LEARNING_AUTH);

/**
 * runs validation for bulk enable for a single rule
 * @param params - {@link BulkActionsValidationArgs}
 */
const validateBulkEnableRule = async ({
  rule,
  mlAuthz
}) => {
  await throwMlAuthError(mlAuthz, rule.params.type);
};

/**
 * runs validation for bulk disable for a single rule
 * @param params - {@link BulkActionsValidationArgs}
 */
exports.validateBulkEnableRule = validateBulkEnableRule;
const validateBulkDisableRule = async ({
  rule,
  mlAuthz
}) => {
  await throwMlAuthError(mlAuthz, rule.params.type);
};

/**
 * runs validation for bulk duplicate for a single rule
 * @param params - {@link BulkActionsValidationArgs}
 */
exports.validateBulkDisableRule = validateBulkDisableRule;
const validateBulkDuplicateRule = async ({
  rule,
  mlAuthz
}) => {
  await throwMlAuthError(mlAuthz, rule.params.type);
};

/**
 * runs validation for bulk schedule backfill for a single rule
 * @param params - {@link DryRunManualRuleRunBulkActionsValidationArgs}
 */
exports.validateBulkDuplicateRule = validateBulkDuplicateRule;
const validateBulkScheduleBackfill = async ({
  rule,
  experimentalFeatures
}) => {
  // check whether "manual rule run" feature is enabled
  await (0, _dry_run.throwDryRunError)(() => (0, _invariant.invariant)(experimentalFeatures === null || experimentalFeatures === void 0 ? void 0 : experimentalFeatures.manualRuleRunEnabled, 'Manual rule run feature is disabled.'), _constants.BulkActionsDryRunErrCode.MANUAL_RULE_RUN_FEATURE);
  await (0, _dry_run.throwDryRunError)(() => (0, _invariant.invariant)(rule.enabled, 'Cannot schedule manual rule run for a disabled rule'), _constants.BulkActionsDryRunErrCode.MANUAL_RULE_RUN_DISABLED_RULE);
};

/**
 * runs validation for bulk edit for a single rule
 * @param params - {@link BulkActionsValidationArgs}
 */
exports.validateBulkScheduleBackfill = validateBulkScheduleBackfill;
const validateBulkEditRule = async ({
  ruleType,
  mlAuthz,
  edit,
  immutable
}) => {
  await throwMlAuthError(mlAuthz, ruleType);

  // if rule can't be edited error will be thrown
  const canRuleBeEdited = !immutable || istEditApplicableToImmutableRule(edit);
  await (0, _dry_run.throwDryRunError)(() => (0, _invariant.invariant)(canRuleBeEdited, "Elastic rule can't be edited"), _constants.BulkActionsDryRunErrCode.IMMUTABLE);
};

/**
 * add_rule_actions, set_rule_actions can be applied to prebuilt/immutable rules
 */
exports.validateBulkEditRule = validateBulkEditRule;
const istEditApplicableToImmutableRule = edit => {
  const applicableActions = [_rule_management.BulkActionEditTypeEnum.set_rule_actions, _rule_management.BulkActionEditTypeEnum.add_rule_actions];
  return edit.every(({
    type
  }) => applicableActions.includes(type));
};

/**
 * executes dry run validations for bulk edit of a single rule
 * @param params - {@link DryRunBulkEditBulkActionsValidationArgs}
 */
const dryRunValidateBulkEditRule = async ({
  rule,
  edit,
  mlAuthz
}) => {
  await validateBulkEditRule({
    ruleType: rule.params.type,
    mlAuthz,
    edit,
    immutable: rule.params.immutable
  });

  // if rule is machine_learning, index pattern action can't be applied to it
  await (0, _dry_run.throwDryRunError)(() => (0, _invariant.invariant)(!(0, _helpers.isMlRule)(rule.params.type) || !edit.some(action => (0, _utils2.isIndexPatternsBulkEditAction)(action.type)), "Machine learning rule doesn't have index patterns"), _constants.BulkActionsDryRunErrCode.MACHINE_LEARNING_INDEX_PATTERN);

  // if rule is es|ql, index pattern action can't be applied to it
  await (0, _dry_run.throwDryRunError)(() => (0, _invariant.invariant)(!(0, _utils.isEsqlRule)(rule.params.type) || !edit.some(action => (0, _utils2.isIndexPatternsBulkEditAction)(action.type)), "ES|QL rule doesn't have index patterns"), _constants.BulkActionsDryRunErrCode.ESQL_INDEX_PATTERN);
};
exports.dryRunValidateBulkEditRule = dryRunValidateBulkEditRule;
