// @kbn/uptime-plugin/public/legacy_uptime/components/monitor/synthetics/waterfall/components/waterfall_flyout_table.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Table = void 0;
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _eui = require("@elastic/eui");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StyledText = (0, _styledComponents.default)(_eui.EuiText)`
  width: 100%;
`;
class TableWithoutHeader extends _eui.EuiBasicTable {
  renderTableHead() {
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null);
  }
}
const Table = props => {
  const {
    rows,
    title
  } = props;
  const columns = (0, _react.useMemo)(() => [{
    field: 'name',
    name: '',
    sortable: false,
    render: (_name, item) => /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      size: "xs"
    }, /*#__PURE__*/_react.default.createElement("strong", null, item.name))
  }, {
    field: 'value',
    name: '',
    sortable: false,
    render: (_name, item) => {
      var _item$value;
      return /*#__PURE__*/_react.default.createElement(StyledText, {
        size: "xs",
        textAlign: "right"
      }, (_item$value = item.value) !== null && _item$value !== void 0 ? _item$value : '--');
    }
  }], []);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, /*#__PURE__*/_react.default.createElement("h4", null, title)), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), /*#__PURE__*/_react.default.createElement(TableWithoutHeader, {
    tableLayout: "fixed",
    compressed: true,
    responsiveBreakpoint: false,
    columns: columns,
    items: rows
  }));
};
exports.Table = Table;
