// @kbn/uptime-plugin/public/legacy_uptime/components/monitor/synthetics/waterfall/components/waterfall_chart.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.WaterfallChart = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _useWindowSize = _interopRequireDefault(require("react-use/lib/useWindowSize"));
var _waterfall_chart = require("../context/waterfall_chart");
var _styles = require("./styles");
var _constants = require("./constants");
var _sidebar = require("./sidebar");
var _legend = require("./legend");
var _use_bar_charts = require("./use_bar_charts");
var _waterfall_bar_chart = require("./waterfall_bar_chart");
var _waterfall_chart_fixed_axis = require("./waterfall_chart_fixed_axis");
var _network_requests_total = require("./network_requests_total");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const WaterfallChart = ({
  tickFormat,
  domain,
  barStyleAccessor,
  renderSidebarItem,
  renderLegendItem,
  renderFilter,
  renderFlyout,
  maxHeight = '800px',
  fullHeight = false
}) => {
  const {
    data,
    showOnlyHighlightedNetworkRequests,
    sidebarItems,
    legendItems,
    totalNetworkRequests,
    highlightedNetworkRequests,
    fetchedNetworkRequests
  } = (0, _waterfall_chart.useWaterfallContext)();
  const {
    width
  } = (0, _useWindowSize.default)();
  const chartWrapperDivRef = (0, _react.useRef)(null);
  const legendDivRef = (0, _react.useRef)(null);
  const [height, setHeight] = (0, _react.useState)(maxHeight);
  const shouldRenderSidebar = !!(sidebarItems && renderSidebarItem);
  const shouldRenderLegend = !!(legendItems && legendItems.length > 0 && renderLegendItem);
  (0, _react.useEffect)(() => {
    if (fullHeight && chartWrapperDivRef.current && legendDivRef.current) {
      const chartOffset = chartWrapperDivRef.current.getBoundingClientRect().top;
      const legendOffset = legendDivRef.current.getBoundingClientRect().height;
      setHeight(`calc(100vh - ${chartOffset + _constants.CHART_LEGEND_PADDING + legendOffset}px)`);
    }
  }, [chartWrapperDivRef, fullHeight, legendDivRef, width]);
  const chartsToDisplay = (0, _use_bar_charts.useBarCharts)({
    data
  });
  return /*#__PURE__*/_react.default.createElement(_styles.RelativeContainer, null, /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartFixedTopContainer, null, /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartTopContainer, {
    gutterSize: "none",
    responsive: false
  }, shouldRenderSidebar && /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartSidebarWrapper, {
    grow: _constants.SIDEBAR_GROW_SIZE
  }, /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartFixedTopContainerSidebarCover, {
    paddingSize: "none",
    hasShadow: false
  }), /*#__PURE__*/_react.default.createElement(_network_requests_total.NetworkRequestsTotal, {
    totalNetworkRequests: totalNetworkRequests,
    highlightedNetworkRequests: highlightedNetworkRequests,
    fetchedNetworkRequests: fetchedNetworkRequests,
    showHighlightedNetworkRequests: showOnlyHighlightedNetworkRequests
  }), renderFilter && /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartFilterContainer, null, renderFilter())), /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartAxisOnlyContainer, {
    grow: shouldRenderSidebar ? _constants.MAIN_GROW_SIZE : true,
    "data-test-subj": "axisOnlyWrapper"
  }, /*#__PURE__*/_react.default.createElement(_waterfall_chart_fixed_axis.WaterfallChartFixedAxis, {
    domain: domain,
    barStyleAccessor: barStyleAccessor,
    tickFormat: tickFormat
  })))), /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartOuterContainer, {
    height: height,
    "data-test-subj": "waterfallOuterContainer",
    ref: chartWrapperDivRef
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "none",
    responsive: false
  }, shouldRenderSidebar && /*#__PURE__*/_react.default.createElement(_sidebar.Sidebar, {
    items: sidebarItems,
    render: renderSidebarItem
  }), /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartAxisOnlyContainer, {
    grow: shouldRenderSidebar ? _constants.MAIN_GROW_SIZE : true,
    "data-test-subj": "dataOnlyWrapper"
  }, chartsToDisplay.map((chartData, ind) => /*#__PURE__*/_react.default.createElement(_waterfall_bar_chart.WaterfallBarChart, {
    index: ind,
    key: ind,
    chartData: chartData,
    domain: domain,
    barStyleAccessor: barStyleAccessor,
    tickFormat: tickFormat
  }))))), shouldRenderLegend && /*#__PURE__*/_react.default.createElement(_styles.WaterfallChartLegendContainer, {
    ref: legendDivRef
  }, /*#__PURE__*/_react.default.createElement(_legend.Legend, {
    items: legendItems,
    render: renderLegendItem
  })), renderFlyout && renderFlyout());
};
exports.WaterfallChart = WaterfallChart;
