// @kbn/cloud-security-posture-plugin/public/pages/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Benchmarks", {
  enumerable: true,
  get: function () {
    return _benchmarks.Benchmarks;
  }
});
Object.defineProperty(exports, "ComplianceDashboard", {
  enumerable: true,
  get: function () {
    return _compliance_dashboard.ComplianceDashboard;
  }
});
Object.defineProperty(exports, "Configurations", {
  enumerable: true,
  get: function () {
    return _configurations.Configurations;
  }
});
Object.defineProperty(exports, "Findings", {
  enumerable: true,
  get: function () {
    return _findings.Findings;
  }
});
Object.defineProperty(exports, "Rules", {
  enumerable: true,
  get: function () {
    return _rules.Rules;
  }
});
Object.defineProperty(exports, "VulnerabilityDashboard", {
  enumerable: true,
  get: function () {
    return _vulnerability_dashboard.VulnerabilityDashboard;
  }
});
var _findings = require("./findings");
var _configurations = require("./configurations");
var _compliance_dashboard = require("./compliance_dashboard");
var _benchmarks = require("./benchmarks");
var _rules = require("./rules");
var _vulnerability_dashboard = require("./vulnerability_dashboard/vulnerability_dashboard");
