// @kbn/enterprise-search-plugin/public/applications/analytics/components/add_analytics_collections/add_analytics_collection.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AddAnalyticsCollection = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _add_analytics_collection_modal = require("./add_analytics_collection_modal");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const AddAnalyticsCollection = ({
  render,
  disabled
}) => {
  const [isModalVisible, setIsModalVisible] = (0, _react.useState)(false);
  const closeModal = () => setIsModalVisible(false);
  const showModal = () => setIsModalVisible(true);
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, render ? render(showModal) : /*#__PURE__*/_react.default.createElement(_eui.EuiButton, {
    fill: true,
    iconType: "plusInCircle",
    onClick: showModal,
    disabled: disabled,
    "data-test-subj": "create-analytics-collection-btn"
  }, _i18n.i18n.translate('xpack.enterpriseSearch.analytics.collections.create.buttonTitle', {
    defaultMessage: 'Create collection'
  })), isModalVisible && /*#__PURE__*/_react.default.createElement(_add_analytics_collection_modal.AddAnalyticsCollectionModal, {
    onClose: closeModal
  }));
};
exports.AddAnalyticsCollection = AddAnalyticsCollection;
