// @kbn/rule-data-utils/src/legacy_alerts_as_data.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TAGS = exports.EVENT_KIND = exports.EVENT_ACTION = exports.ECS_VERSION = exports.ALERT_WORKFLOW_USER = exports.ALERT_WORKFLOW_STATUS_UPDATED_AT = exports.ALERT_WORKFLOW_REASON = exports.ALERT_SYSTEM_STATUS = exports.ALERT_SUPPRESSION_VALUE = exports.ALERT_SUPPRESSION_TERMS = exports.ALERT_SUPPRESSION_START = exports.ALERT_SUPPRESSION_FIELD = exports.ALERT_SUPPRESSION_END = exports.ALERT_SUPPRESSION_DOCS_COUNT = exports.ALERT_SEVERITY = exports.ALERT_RULE_VERSION = exports.ALERT_RULE_UPDATED_BY = exports.ALERT_RULE_UPDATED_AT = exports.ALERT_RULE_TYPE = exports.ALERT_RULE_TO = exports.ALERT_RULE_RULE_NAME_OVERRIDE = exports.ALERT_RULE_RULE_ID = exports.ALERT_RULE_REFERENCES = exports.ALERT_RULE_NOTE = exports.ALERT_RULE_LICENSE = exports.ALERT_RULE_INTERVAL = exports.ALERT_RULE_FROM = exports.ALERT_RULE_ENABLED = exports.ALERT_RULE_DESCRIPTION = exports.ALERT_RULE_CREATED_BY = exports.ALERT_RULE_CREATED_AT = exports.ALERT_RULE_AUTHOR = exports.ALERT_RISK_SCORE = void 0;
var _default_alerts_as_data = require("./default_alerts_as_data");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const ECS_VERSION = exports.ECS_VERSION = 'ecs.version';
const EVENT_ACTION = exports.EVENT_ACTION = 'event.action';
const EVENT_KIND = exports.EVENT_KIND = 'event.kind';
const TAGS = exports.TAGS = 'tags';

// These are the fields that are in the rule registry technical component template
// that are NOT in the framework alerts as data common component template

// We will maintain a legacy component template that can be used by legacy
// rule registry rules with these fields.
const ALERT_RISK_SCORE = exports.ALERT_RISK_SCORE = `${_default_alerts_as_data.ALERT_NAMESPACE}.risk_score`;
const ALERT_RULE_AUTHOR = exports.ALERT_RULE_AUTHOR = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.author`;
const ALERT_RULE_CREATED_AT = exports.ALERT_RULE_CREATED_AT = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.created_at`;
const ALERT_RULE_CREATED_BY = exports.ALERT_RULE_CREATED_BY = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.created_by`;
const ALERT_RULE_DESCRIPTION = exports.ALERT_RULE_DESCRIPTION = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.description`;
const ALERT_RULE_ENABLED = exports.ALERT_RULE_ENABLED = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.enabled`;
const ALERT_RULE_FROM = exports.ALERT_RULE_FROM = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.from`;
const ALERT_RULE_INTERVAL = exports.ALERT_RULE_INTERVAL = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.interval`;
const ALERT_RULE_LICENSE = exports.ALERT_RULE_LICENSE = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.license`;
const ALERT_RULE_NOTE = exports.ALERT_RULE_NOTE = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.note`;
const ALERT_RULE_REFERENCES = exports.ALERT_RULE_REFERENCES = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.references`;
const ALERT_RULE_RULE_ID = exports.ALERT_RULE_RULE_ID = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.rule_id`;
const ALERT_RULE_RULE_NAME_OVERRIDE = exports.ALERT_RULE_RULE_NAME_OVERRIDE = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.rule_name_override`;
const ALERT_RULE_TO = exports.ALERT_RULE_TO = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.to`;
const ALERT_RULE_TYPE = exports.ALERT_RULE_TYPE = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.type`;
const ALERT_RULE_UPDATED_AT = exports.ALERT_RULE_UPDATED_AT = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.updated_at`;
const ALERT_RULE_UPDATED_BY = exports.ALERT_RULE_UPDATED_BY = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.updated_by`;
const ALERT_RULE_VERSION = exports.ALERT_RULE_VERSION = `${_default_alerts_as_data.ALERT_RULE_NAMESPACE}.version`;
const ALERT_SEVERITY = exports.ALERT_SEVERITY = `${_default_alerts_as_data.ALERT_NAMESPACE}.severity`;
const ALERT_SUPPRESSION_META = `${_default_alerts_as_data.ALERT_NAMESPACE}.suppression`;
const ALERT_SUPPRESSION_TERMS = exports.ALERT_SUPPRESSION_TERMS = `${ALERT_SUPPRESSION_META}.terms`;
const ALERT_SUPPRESSION_FIELD = exports.ALERT_SUPPRESSION_FIELD = `${ALERT_SUPPRESSION_TERMS}.field`;
const ALERT_SUPPRESSION_VALUE = exports.ALERT_SUPPRESSION_VALUE = `${ALERT_SUPPRESSION_TERMS}.value`;
const ALERT_SUPPRESSION_START = exports.ALERT_SUPPRESSION_START = `${ALERT_SUPPRESSION_META}.start`;
const ALERT_SUPPRESSION_END = exports.ALERT_SUPPRESSION_END = `${ALERT_SUPPRESSION_META}.end`;
const ALERT_SUPPRESSION_DOCS_COUNT = exports.ALERT_SUPPRESSION_DOCS_COUNT = `${ALERT_SUPPRESSION_META}.docs_count`;
const ALERT_SYSTEM_STATUS = exports.ALERT_SYSTEM_STATUS = `${_default_alerts_as_data.ALERT_NAMESPACE}.system_status`;
const ALERT_WORKFLOW_REASON = exports.ALERT_WORKFLOW_REASON = `${_default_alerts_as_data.ALERT_NAMESPACE}.workflow_reason`;
const ALERT_WORKFLOW_USER = exports.ALERT_WORKFLOW_USER = `${_default_alerts_as_data.ALERT_NAMESPACE}.workflow_user`;
const ALERT_WORKFLOW_STATUS_UPDATED_AT = exports.ALERT_WORKFLOW_STATUS_UPDATED_AT = `${_default_alerts_as_data.ALERT_NAMESPACE}.workflow_status_updated_at`;
