// @kbn/security-solution-plugin/public/flyout/document_details/preview/footer.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.PreviewPanelFooter = void 0;
var _eui = require("@elastic/eui");
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _expandableFlyout = require("@kbn/expandable-flyout");
var _flyout_footer = require("../../shared/components/flyout_footer");
var _panel_keys = require("../shared/constants/panel_keys");
var _context = require("../shared/context");
var _test_ids = require("./test_ids");
var _kibana = require("../../../common/lib/kibana");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

/**
 * Footer at the bottom of preview panel with a link to open document details flyout
 */
const PreviewPanelFooter = () => {
  const {
    eventId,
    indexName,
    scopeId
  } = (0, _context.useDocumentDetailsContext)();
  const {
    openFlyout
  } = (0, _expandableFlyout.useExpandableFlyoutApi)();
  const {
    telemetry
  } = (0, _kibana.useKibana)().services;
  const openDocumentFlyout = (0, _react.useCallback)(() => {
    openFlyout({
      right: {
        id: _panel_keys.DocumentDetailsRightPanelKey,
        params: {
          id: eventId,
          indexName,
          scopeId
        }
      }
    });
    telemetry.reportDetailsFlyoutOpened({
      location: scopeId,
      panel: 'right'
    });
  }, [openFlyout, eventId, indexName, scopeId, telemetry]);
  return /*#__PURE__*/_react.default.createElement(_flyout_footer.FlyoutFooter, {
    "data-test-subj": _test_ids.PREVIEW_FOOTER_TEST_ID
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "center"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    onClick: openDocumentFlyout,
    target: "_blank",
    "data-test-subj": _test_ids.PREVIEW_FOOTER_LINK_TEST_ID
  }, _i18n.i18n.translate('xpack.securitySolution.flyout.preview.openFlyoutLabel', {
    defaultMessage: 'Show full alert details'
  })))));
};
exports.PreviewPanelFooter = PreviewPanelFooter;
