// @kbn/saved-search-plugin/common/service/saved_searches_utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.fromSavedSearchAttributes = void 0;
Object.defineProperty(exports, "getSavedSearchFullPathUrl", {
  enumerable: true,
  get: function () {
    return _.getSavedSearchFullPathUrl;
  }
});
Object.defineProperty(exports, "getSavedSearchUrl", {
  enumerable: true,
  get: function () {
    return _.getSavedSearchUrl;
  }
});
exports.toSavedSearchAttributes = void 0;
var _lodash = require("lodash");
var _ = require("..");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const fromSavedSearchAttributes = (id, attributes, tags, references, searchSource, sharingSavedObjectProps, managed) => ({
  ...(0, _.fromSavedSearchAttributes)(id, attributes, tags, searchSource, managed),
  sharingSavedObjectProps,
  references
});
exports.fromSavedSearchAttributes = fromSavedSearchAttributes;
const toSavedSearchAttributes = (savedSearch, searchSourceJSON) => {
  var _savedSearch$title, _savedSearch$sort, _savedSearch$columns, _savedSearch$descript, _savedSearch$grid, _savedSearch$hideChar, _savedSearch$isTextBa, _savedSearch$timeRest;
  return {
    kibanaSavedObjectMeta: {
      searchSourceJSON
    },
    title: (_savedSearch$title = savedSearch.title) !== null && _savedSearch$title !== void 0 ? _savedSearch$title : '',
    sort: (_savedSearch$sort = savedSearch.sort) !== null && _savedSearch$sort !== void 0 ? _savedSearch$sort : [],
    columns: (_savedSearch$columns = savedSearch.columns) !== null && _savedSearch$columns !== void 0 ? _savedSearch$columns : [],
    description: (_savedSearch$descript = savedSearch.description) !== null && _savedSearch$descript !== void 0 ? _savedSearch$descript : '',
    grid: (_savedSearch$grid = savedSearch.grid) !== null && _savedSearch$grid !== void 0 ? _savedSearch$grid : {},
    hideChart: (_savedSearch$hideChar = savedSearch.hideChart) !== null && _savedSearch$hideChar !== void 0 ? _savedSearch$hideChar : false,
    viewMode: savedSearch.viewMode,
    hideAggregatedPreview: savedSearch.hideAggregatedPreview,
    rowHeight: savedSearch.rowHeight,
    headerRowHeight: savedSearch.headerRowHeight,
    isTextBasedQuery: (_savedSearch$isTextBa = savedSearch.isTextBasedQuery) !== null && _savedSearch$isTextBa !== void 0 ? _savedSearch$isTextBa : false,
    usesAdHocDataView: savedSearch.usesAdHocDataView,
    timeRestore: (_savedSearch$timeRest = savedSearch.timeRestore) !== null && _savedSearch$timeRest !== void 0 ? _savedSearch$timeRest : false,
    timeRange: savedSearch.timeRange ? (0, _lodash.pick)(savedSearch.timeRange, ['from', 'to']) : undefined,
    refreshInterval: savedSearch.refreshInterval,
    rowsPerPage: savedSearch.rowsPerPage,
    sampleSize: savedSearch.sampleSize,
    breakdownField: savedSearch.breakdownField,
    visContext: savedSearch.visContext
  };
};
exports.toSavedSearchAttributes = toSavedSearchAttributes;
