// @kbn/uptime-plugin/public/legacy_uptime/components/synthetics/check_steps/step_image.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StepImage = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _ping_timestamp = require("../../monitor/ping_list/columns/ping_timestamp");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StepImage = ({
  step,
  compactView,
  allStepsLoaded
}) => {
  var _step$synthetics, _step$synthetics$step, _step$synthetics$payl, _step$synthetics2, _step$synthetics2$ste;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    alignItems: "center",
    gutterSize: "s",
    wrap: true
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false
  }, /*#__PURE__*/_react.default.createElement(_ping_timestamp.PingTimestamp, {
    checkGroup: step.monitor.check_group,
    initialStepNo: (_step$synthetics = step.synthetics) === null || _step$synthetics === void 0 ? void 0 : (_step$synthetics$step = _step$synthetics.step) === null || _step$synthetics$step === void 0 ? void 0 : _step$synthetics$step.index,
    stepStatus: (_step$synthetics$payl = step.synthetics.payload) === null || _step$synthetics$payl === void 0 ? void 0 : _step$synthetics$payl.status,
    allStepsLoaded: allStepsLoaded
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
    grow: false,
    style: {
      minWidth: 80
    }
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: compactView ? 's' : 'm'
  }, (_step$synthetics2 = step.synthetics) === null || _step$synthetics2 === void 0 ? void 0 : (_step$synthetics2$ste = _step$synthetics2.step) === null || _step$synthetics2$ste === void 0 ? void 0 : _step$synthetics2$ste.name)));
};
exports.StepImage = StepImage;
