// @kbn/cross-cluster-replication-plugin/public/app/components/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "AutoFollowPatternActionMenu", {
  enumerable: true,
  get: function () {
    return _auto_follow_pattern_action_menu.AutoFollowPatternActionMenu;
  }
});
Object.defineProperty(exports, "AutoFollowPatternDeleteProvider", {
  enumerable: true,
  get: function () {
    return _auto_follow_pattern_delete_provider.AutoFollowPatternDeleteProvider;
  }
});
Object.defineProperty(exports, "AutoFollowPatternForm", {
  enumerable: true,
  get: function () {
    return _auto_follow_pattern_form.AutoFollowPatternForm;
  }
});
Object.defineProperty(exports, "AutoFollowPatternIndicesPreview", {
  enumerable: true,
  get: function () {
    return _auto_follow_pattern_indices_preview.AutoFollowPatternIndicesPreview;
  }
});
Object.defineProperty(exports, "AutoFollowPatternPageTitle", {
  enumerable: true,
  get: function () {
    return _auto_follow_pattern_page_title.AutoFollowPatternPageTitle;
  }
});
Object.defineProperty(exports, "FollowerIndexActionsProvider", {
  enumerable: true,
  get: function () {
    return _follower_index_actions_providers.FollowerIndexActionsProvider;
  }
});
Object.defineProperty(exports, "FollowerIndexForm", {
  enumerable: true,
  get: function () {
    return _follower_index_form.FollowerIndexForm;
  }
});
Object.defineProperty(exports, "FollowerIndexPageTitle", {
  enumerable: true,
  get: function () {
    return _follower_index_page_title.FollowerIndexPageTitle;
  }
});
Object.defineProperty(exports, "FollowerIndexPauseProvider", {
  enumerable: true,
  get: function () {
    return _follower_index_actions_providers.FollowerIndexPauseProvider;
  }
});
Object.defineProperty(exports, "FollowerIndexResumeProvider", {
  enumerable: true,
  get: function () {
    return _follower_index_actions_providers.FollowerIndexResumeProvider;
  }
});
Object.defineProperty(exports, "FollowerIndexUnfollowProvider", {
  enumerable: true,
  get: function () {
    return _follower_index_actions_providers.FollowerIndexUnfollowProvider;
  }
});
Object.defineProperty(exports, "FormEntryRow", {
  enumerable: true,
  get: function () {
    return _form_entry_row.FormEntryRow;
  }
});
Object.defineProperty(exports, "RemoteClustersFormField", {
  enumerable: true,
  get: function () {
    return _remote_clusters_form_field.RemoteClustersFormField;
  }
});
Object.defineProperty(exports, "RemoteClustersProvider", {
  enumerable: true,
  get: function () {
    return _remote_clusters_provider.RemoteClustersProvider;
  }
});
Object.defineProperty(exports, "SectionError", {
  enumerable: true,
  get: function () {
    return _section_error.SectionError;
  }
});
Object.defineProperty(exports, "SectionLoading", {
  enumerable: true,
  get: function () {
    return _section_loading.SectionLoading;
  }
});
Object.defineProperty(exports, "SectionUnauthorized", {
  enumerable: true,
  get: function () {
    return _section_unauthorized.SectionUnauthorized;
  }
});
var _section_loading = require("./section_loading");
var _section_error = require("./section_error");
var _section_unauthorized = require("./section_unauthorized");
var _remote_clusters_provider = require("./remote_clusters_provider");
var _auto_follow_pattern_form = require("./auto_follow_pattern_form");
var _auto_follow_pattern_delete_provider = require("./auto_follow_pattern_delete_provider");
var _auto_follow_pattern_page_title = require("./auto_follow_pattern_page_title");
var _auto_follow_pattern_indices_preview = require("./auto_follow_pattern_indices_preview");
var _follower_index_actions_providers = require("./follower_index_actions_providers");
var _follower_index_form = require("./follower_index_form");
var _follower_index_page_title = require("./follower_index_page_title");
var _form_entry_row = require("./form_entry_row");
var _remote_clusters_form_field = require("./remote_clusters_form_field");
var _auto_follow_pattern_action_menu = require("./auto_follow_pattern_action_menu");
