// @kbn/ingest-pipelines-plugin/public/application/components/pipeline_editor/components/processors_empty_prompt.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ProcessorsEmptyPrompt = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _shared_imports = require("../../../../shared_imports");
var _context = require("../context");
var _add_processor_button = require("./add_processor_button");
var _load_from_json = require("./load_from_json");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const i18nTexts = {
  emptyPromptTitle: _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.emptyPrompt.title', {
    defaultMessage: 'Add your first processor'
  })
};
const ProcessorsEmptyPrompt = ({
  onLoadJson
}) => {
  const {
    onTreeAction
  } = (0, _context.usePipelineProcessorsContext)();
  const {
    services
  } = (0, _shared_imports.useKibana)();
  return /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
    title: /*#__PURE__*/_react.default.createElement("h2", null, i18nTexts.emptyPromptTitle),
    "data-test-subj": "processorsEmptyPrompt",
    body: /*#__PURE__*/_react.default.createElement("p", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.ingestPipelines.pipelineEditor.emptyPrompt.description",
      defaultMessage: "Use processors to transform data before indexing. {learnMoreLink}",
      values: {
        learnMoreLink: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
          href: services.documentation.getProcessorsUrl(),
          target: "_blank",
          external: true
        }, _i18n.i18n.translate('xpack.ingestPipelines.pipelineEditor.processorsDocumentationLink', {
          defaultMessage: 'Learn more.'
        }))
      }
    })),
    actions: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_add_processor_button.AddProcessorButton, {
      onClick: () => {
        onTreeAction({
          type: 'addProcessor',
          payload: {
            target: ['processors']
          }
        });
      }
    }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
      size: "m"
    }), /*#__PURE__*/_react.default.createElement(_load_from_json.LoadFromJsonButton, {
      onDone: onLoadJson
    }))
  });
};
exports.ProcessorsEmptyPrompt = ProcessorsEmptyPrompt;
