// @kbn/unified-doc-viewer-plugin/public/components/doc_viewer_logs_overview/sub_components/log_level.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LogLevel = LogLevel;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const LEVEL_DICT = {
  error: 'danger',
  warn: 'warning',
  info: 'primary',
  debug: 'accent'
};
function LogLevel({
  level
}) {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  if (!level) return null;
  const colorName = LEVEL_DICT[level];
  const computedColor = colorName ? euiTheme.colors[colorName] : null;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiBadge, {
    color: "hollow",
    css: (0, _react2.css)`
        max-width: 100px;
        ${computedColor ? `border: 2px solid ${computedColor};` : ''}
      `,
    "data-test-subj": "unifiedDocViewLogsOverviewLogLevel"
  }, level);
}
