// @kbn/ml-plugin/public/application/jobs/new_job/pages/components/job_details_step/components/advanced_section/components/ignore_unavailable/ignore_unavailable_switch.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.IgnoreUnavailableSwitch = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
var _job_creator_context = require("../../../../../job_creator_context");
var _description = require("./description");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const IgnoreUnavailableSwitch = () => {
  const {
    jobCreator,
    jobCreatorUpdate
  } = (0, _react.useContext)(_job_creator_context.JobCreatorContext);
  const [ignoreUnavailable, setIgnoreUnavailable] = (0, _react.useState)(jobCreator.ignoreUnavailable);
  (0, _react.useEffect)(() => {
    jobCreator.ignoreUnavailable = ignoreUnavailable;
    jobCreatorUpdate();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [ignoreUnavailable]);
  function toggleIgnoreUnavailable() {
    setIgnoreUnavailable(!ignoreUnavailable);
  }
  return /*#__PURE__*/_react.default.createElement(_description.Description, null, /*#__PURE__*/_react.default.createElement(_eui.EuiSwitch, {
    name: "ignoreUnavailableSwitch",
    checked: ignoreUnavailable,
    onChange: toggleIgnoreUnavailable,
    "data-test-subj": "mlJobWizardSwitchIgnoreUnavailableIndex",
    label: _i18n.i18n.translate('xpack.ml.newJob.wizard.jobDetailsStep.advancedSection.ignoreUnavailable.title', {
      defaultMessage: 'Ignore unavailable indices'
    })
  }));
};
exports.IgnoreUnavailableSwitch = IgnoreUnavailableSwitch;
