// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/field_parameters/coerce_shape_parameter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CoerceShapeParameter = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _edit_field = require("../fields/edit_field");
var _documentation = require("../../../../../services/documentation");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const CoerceShapeParameter = () => /*#__PURE__*/_react.default.createElement(_edit_field.EditFieldFormRow, {
  title: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.coerceShapeFieldTitle', {
    defaultMessage: 'Coerce to shape'
  }),
  description: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.coerceShapeDescription', {
    defaultMessage: 'If disabled, then documents that contain polygons with unclosed linear rings are rejected.'
  }),
  docLink: {
    text: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.coerceShapeDocLinkText', {
      defaultMessage: 'Coerce documentation'
    }),
    href: _documentation.documentationService.getCoerceLink()
  },
  formFieldPath: "coerce",
  configPath: "coerce_shape"
});
exports.CoerceShapeParameter = CoerceShapeParameter;
