// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/field_parameters/enabled_parameter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.EnabledParameter = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _eui = require("@elastic/eui");
var _documentation = require("../../../../../services/documentation");
var _edit_field = require("../fields/edit_field");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const EnabledParameter = () => {
  return /*#__PURE__*/_react.default.createElement(_edit_field.EditFieldFormRow, {
    title: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.searchableProperties.fieldTitle', {
      defaultMessage: 'Searchable properties'
    }),
    description: /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
      id: "xpack.idxMgmt.mappingsEditor.searchableProperties.fieldDescription",
      defaultMessage: "Allow object properties to be searched. The JSON can still be retrieved from the {source} field even after disabling this setting.",
      values: {
        source: /*#__PURE__*/_react.default.createElement(_eui.EuiCode, null, "_source")
      }
    }),
    docLink: {
      text: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.enabledDocLinkText', {
        defaultMessage: 'Enabled documentation'
      }),
      href: _documentation.documentationService.getEnabledLink()
    },
    formFieldPath: "enabled"
  });
};
exports.EnabledParameter = EnabledParameter;
