// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/field_parameters/norms_parameter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.NormsParameter = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _documentation = require("../../../../../services/documentation");
var _edit_field = require("../fields/edit_field");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const NormsParameter = ({
  configPath = 'norms'
}) => /*#__PURE__*/_react.default.createElement(_edit_field.EditFieldFormRow, {
  title: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.useNormsFieldTitle', {
    defaultMessage: 'Use norms'
  }),
  description: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.useNormsFieldDescription', {
    defaultMessage: 'Account for field length when scoring queries. Norms require significant memory and are not necessary for fields that are used solely for filtering or aggregations.'
  }),
  docLink: {
    text: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.normsDocLinkText', {
      defaultMessage: 'Norms documentation'
    }),
    href: _documentation.documentationService.getNormsLink()
  },
  formFieldPath: "norms",
  configPath: configPath
});
exports.NormsParameter = NormsParameter;
