// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/field_parameters/store_parameter.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.StoreParameter = void 0;
var _react = _interopRequireDefault(require("react"));
var _i18n = require("@kbn/i18n");
var _documentation = require("../../../../../services/documentation");
var _edit_field = require("../fields/edit_field");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const StoreParameter = () => /*#__PURE__*/_react.default.createElement(_edit_field.EditFieldFormRow, {
  title: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.storeFieldValueFieldTitle', {
    defaultMessage: 'Store field value outside of _source'
  }),
  description: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.storeFieldValueFieldDescription', {
    defaultMessage: 'This can be useful when the _source field is very large and you want to retrieve a few select fields without extracting them from _source.'
  }),
  docLink: {
    text: _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.storeDocLinkText', {
      defaultMessage: 'Store documentation'
    }),
    href: _documentation.documentationService.getStoreLink()
  },
  formFieldPath: "store"
});
exports.StoreParameter = StoreParameter;
