// @kbn/ui-actions-enhanced-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "DEFAULT_URL_DRILLDOWN_OPTIONS", {
  enumerable: true,
  get: function () {
    return _url_drilldown.DEFAULT_URL_DRILLDOWN_OPTIONS;
  }
});
Object.defineProperty(exports, "Plugin", {
  enumerable: true,
  get: function () {
    return _plugin.AdvancedUiActionsPublicPlugin;
  }
});
Object.defineProperty(exports, "UiActionsEnhancedAbstractActionStorage", {
  enumerable: true,
  get: function () {
    return _dynamic_actions.AbstractActionStorage;
  }
});
Object.defineProperty(exports, "UiActionsEnhancedActionFactory", {
  enumerable: true,
  get: function () {
    return _dynamic_actions.ActionFactory;
  }
});
Object.defineProperty(exports, "UiActionsEnhancedDynamicActionManager", {
  enumerable: true,
  get: function () {
    return _dynamic_actions.DynamicActionManager;
  }
});
Object.defineProperty(exports, "UiActionsEnhancedMemoryActionStorage", {
  enumerable: true,
  get: function () {
    return _dynamic_actions.MemoryActionStorage;
  }
});
Object.defineProperty(exports, "UrlDrilldownCollectConfig", {
  enumerable: true,
  get: function () {
    return _url_drilldown.UrlDrilldownCollectConfig;
  }
});
Object.defineProperty(exports, "UrlDrilldownOptionsComponent", {
  enumerable: true,
  get: function () {
    return _url_drilldown.UrlDrilldownOptionsComponent;
  }
});
exports.plugin = plugin;
Object.defineProperty(exports, "urlDrilldownCompileUrl", {
  enumerable: true,
  get: function () {
    return _url_drilldown.urlDrilldownCompileUrl;
  }
});
Object.defineProperty(exports, "urlDrilldownGlobalScopeProvider", {
  enumerable: true,
  get: function () {
    return _url_drilldown.urlDrilldownGlobalScopeProvider;
  }
});
Object.defineProperty(exports, "urlDrilldownValidateUrl", {
  enumerable: true,
  get: function () {
    return _url_drilldown.urlDrilldownValidateUrl;
  }
});
Object.defineProperty(exports, "urlDrilldownValidateUrlTemplate", {
  enumerable: true,
  get: function () {
    return _url_drilldown.urlDrilldownValidateUrlTemplate;
  }
});
var _plugin = require("./plugin");
var _dynamic_actions = require("./dynamic_actions");
var _url_drilldown = require("./drilldowns/url_drilldown");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function plugin(initializerContext) {
  return new _plugin.AdvancedUiActionsPublicPlugin(initializerContext);
}
