// @kbn/monitoring-plugin/public/components/elasticsearch/ccr_shard/ccr_shard.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.CcrShard = CcrShard;
var _react = _interopRequireWildcard(require("react"));
var _public = require("@kbn/kibana-react-plugin/public");
var _eui = require("@elastic/eui");
var _chart = require("../../chart");
var _status = require("./status");
var _formatting = require("../../../../common/formatting");
var _i18nReact = require("@kbn/i18n-react");
var _i18n = require("@kbn/i18n");
var _callout = require("../../../alerts/callout");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function CcrShard(props) {
  var _services$uiSettings;
  const {
    services
  } = (0, _public.useKibana)();
  const timezone = (_services$uiSettings = services.uiSettings) === null || _services$uiSettings === void 0 ? void 0 : _services$uiSettings.get('dateFormat:tz');
  const {
    metrics,
    stat,
    timestamp,
    oldestStat,
    formattedLeader,
    alerts
  } = props;
  const renderCharts = () => {
    const seriesToShow = [metrics.ccr_sync_lag_ops, metrics.ccr_sync_lag_time];
    const charts = seriesToShow.map((data, index) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      style: {
        minWidth: '45%'
      },
      key: index
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, null, /*#__PURE__*/_react.default.createElement(_chart.MonitoringTimeseriesContainer, {
      series: data
    }))));
    return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, charts);
  };
  const renderErrors = () => {
    if (stat.read_exceptions && stat.read_exceptions.length > 0) {
      return /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
        size: "s",
        color: "danger"
      }, /*#__PURE__*/_react.default.createElement("h3", null, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
        color: "danger"
      }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.monitoring.elasticsearch.ccrShard.errorsTableTitle",
        defaultMessage: "Errors"
      })))), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
        size: "s"
      }), /*#__PURE__*/_react.default.createElement(_eui.EuiBasicTable, {
        items: stat.read_exceptions,
        columns: [{
          name: _i18n.i18n.translate('xpack.monitoring.elasticsearch.ccrShard.errorsTable.typeColumnTitle', {
            defaultMessage: 'Type'
          }),
          field: 'exception.type'
        }, {
          name: _i18n.i18n.translate('xpack.monitoring.elasticsearch.ccrShard.errorsTable.reasonColumnTitle', {
            defaultMessage: 'Reason'
          }),
          field: 'exception.reason',
          width: '75%'
        }]
      })), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, null));
    }
    return null;
  };
  const renderLatestStat = () => {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
      id: "ccrLatestStat",
      buttonContent: /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, null, /*#__PURE__*/_react.default.createElement("h2", null, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
        id: "xpack.monitoring.elasticsearch.ccrShard.latestStateAdvancedButtonLabel",
        defaultMessage: "Advanced"
      }))),
      paddingSize: "l"
    }, /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "s"
    }, /*#__PURE__*/_react.default.createElement("h2", null, (0, _formatting.formatDateTimeLocal)(timestamp, timezone))), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, null), /*#__PURE__*/_react.default.createElement(_eui.EuiCodeBlock, {
      language: "json"
    }, JSON.stringify(stat, null, 2))));
  };
  return /*#__PURE__*/_react.default.createElement(_eui.EuiPage, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPageBody, null, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, null, /*#__PURE__*/_react.default.createElement(_status.Status, {
    stat: stat,
    formattedLeader: formattedLeader,
    oldestStat: oldestStat,
    alerts: alerts
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), /*#__PURE__*/_react.default.createElement(_callout.AlertsCallout, {
    alerts: alerts
  }), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "m"
  }), renderErrors(), /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    wrap: true
  }, renderCharts()), /*#__PURE__*/_react.default.createElement(_eui.EuiHorizontalRule, null), renderLatestStat()));
}
