// @kbn/ml-plugin/common/constants/jobs_list.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RESETTING_JOBS_REFRESH_INTERVAL_MS = exports.PROGRESS_JOBS_REFRESH_INTERVAL_MS = exports.MINIMUM_REFRESH_INTERVAL_MS = exports.DEFAULT_REFRESH_INTERVAL_MS = exports.BLOCKED_JOBS_REFRESH_THRESHOLD_MS = exports.BLOCKED_JOBS_REFRESH_INTERVAL_SLOW_MS = exports.BLOCKED_JOBS_REFRESH_INTERVAL_MS = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DEFAULT_REFRESH_INTERVAL_MS = exports.DEFAULT_REFRESH_INTERVAL_MS = 30000;
const MINIMUM_REFRESH_INTERVAL_MS = exports.MINIMUM_REFRESH_INTERVAL_MS = 1000;
const BLOCKED_JOBS_REFRESH_INTERVAL_MS = exports.BLOCKED_JOBS_REFRESH_INTERVAL_MS = 2000;
const BLOCKED_JOBS_REFRESH_INTERVAL_SLOW_MS = exports.BLOCKED_JOBS_REFRESH_INTERVAL_SLOW_MS = 120000; // 2mins
const BLOCKED_JOBS_REFRESH_THRESHOLD_MS = exports.BLOCKED_JOBS_REFRESH_THRESHOLD_MS = 60000; // 1min
const RESETTING_JOBS_REFRESH_INTERVAL_MS = exports.RESETTING_JOBS_REFRESH_INTERVAL_MS = 1000;
const PROGRESS_JOBS_REFRESH_INTERVAL_MS = exports.PROGRESS_JOBS_REFRESH_INTERVAL_MS = 2000;
