// @kbn/ml-plugin/common/constants/new_job.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SHARED_RESULTS_INDEX_NAME = exports.NUMBER_OF_CATEGORY_EXAMPLES = exports.JOB_TYPE = exports.DEFAULT_RARE_BUCKET_SPAN = exports.DEFAULT_QUERY_DELAY = exports.DEFAULT_MODEL_MEMORY_LIMIT = exports.DEFAULT_BUCKET_SPAN = exports.CREATED_BY_LABEL = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */
let JOB_TYPE = exports.JOB_TYPE = /*#__PURE__*/function (JOB_TYPE) {
  JOB_TYPE["SINGLE_METRIC"] = "single_metric";
  JOB_TYPE["MULTI_METRIC"] = "multi_metric";
  JOB_TYPE["POPULATION"] = "population";
  JOB_TYPE["ADVANCED"] = "advanced";
  JOB_TYPE["CATEGORIZATION"] = "categorization";
  JOB_TYPE["RARE"] = "rare";
  JOB_TYPE["GEO"] = "geo";
  return JOB_TYPE;
}({});
let CREATED_BY_LABEL = exports.CREATED_BY_LABEL = /*#__PURE__*/function (CREATED_BY_LABEL) {
  CREATED_BY_LABEL["SINGLE_METRIC"] = "single-metric-wizard";
  CREATED_BY_LABEL["MULTI_METRIC"] = "multi-metric-wizard";
  CREATED_BY_LABEL["POPULATION"] = "population-wizard";
  CREATED_BY_LABEL["ADVANCED"] = "advanced-wizard";
  CREATED_BY_LABEL["CATEGORIZATION"] = "categorization-wizard";
  CREATED_BY_LABEL["RARE"] = "rare-wizard";
  CREATED_BY_LABEL["GEO"] = "geo-wizard";
  CREATED_BY_LABEL["GEO_FROM_LENS"] = "geo-wizard-from-lens";
  CREATED_BY_LABEL["APM_TRANSACTION"] = "ml-module-apm-transaction";
  CREATED_BY_LABEL["SINGLE_METRIC_FROM_LENS"] = "single-metric-wizard-from-lens";
  CREATED_BY_LABEL["MULTI_METRIC_FROM_LENS"] = "multi-metric-wizard-from-lens";
  CREATED_BY_LABEL["CATEGORIZATION_FROM_PATTERN_ANALYSIS"] = "categorization-wizard-from-pattern-analysis";
  return CREATED_BY_LABEL;
}({});
const DEFAULT_MODEL_MEMORY_LIMIT = exports.DEFAULT_MODEL_MEMORY_LIMIT = '10MB';
const DEFAULT_BUCKET_SPAN = exports.DEFAULT_BUCKET_SPAN = '15m';
const DEFAULT_RARE_BUCKET_SPAN = exports.DEFAULT_RARE_BUCKET_SPAN = '1h';
const DEFAULT_QUERY_DELAY = exports.DEFAULT_QUERY_DELAY = '60s';
const SHARED_RESULTS_INDEX_NAME = exports.SHARED_RESULTS_INDEX_NAME = 'shared';
const NUMBER_OF_CATEGORY_EXAMPLES = exports.NUMBER_OF_CATEGORY_EXAMPLES = 5;
