// @kbn/expressions-plugin/public/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Plugin", {
  enumerable: true,
  get: function () {
    return _plugin.ExpressionsPublicPlugin;
  }
});
Object.defineProperty(exports, "buildExpression", {
  enumerable: true,
  get: function () {
    return _common.buildExpression;
  }
});
Object.defineProperty(exports, "buildExpressionFunction", {
  enumerable: true,
  get: function () {
    return _common.buildExpressionFunction;
  }
});
Object.defineProperty(exports, "createDefaultInspectorAdapters", {
  enumerable: true,
  get: function () {
    return _common.createDefaultInspectorAdapters;
  }
});
Object.defineProperty(exports, "formatExpression", {
  enumerable: true,
  get: function () {
    return _common.formatExpression;
  }
});
Object.defineProperty(exports, "isExpressionAstBuilder", {
  enumerable: true,
  get: function () {
    return _common.isExpressionAstBuilder;
  }
});
Object.defineProperty(exports, "parseExpression", {
  enumerable: true,
  get: function () {
    return _common.parseExpression;
  }
});
exports.plugin = plugin;
Object.defineProperty(exports, "useExpressionRenderer", {
  enumerable: true,
  get: function () {
    return _react_expression_renderer.useExpressionRenderer;
  }
});
require("./index.scss");
var _plugin = require("./plugin");
var _react_expression_renderer = require("./react_expression_renderer");
var _common = require("../common");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

// Kibana Platform.

function plugin(initializerContext) {
  return new _plugin.ExpressionsPublicPlugin(initializerContext);
}

// Static exports.
