// @kbn/search-playground/common/doc_links.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.docLinks = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

class PlaygroundDocLinks {
  constructor() {
    (0, _defineProperty2.default)(this, "chatPlayground", '');
    (0, _defineProperty2.default)(this, "retrievalOptimize", '');
    (0, _defineProperty2.default)(this, "retrieval", '');
    (0, _defineProperty2.default)(this, "context", '');
    (0, _defineProperty2.default)(this, "hiddenFields", '');
  }
  setDocLinks(newDocLinks) {
    this.chatPlayground = newDocLinks.playground.chatPlayground;
    this.retrievalOptimize = newDocLinks.playground.retrievalOptimize;
    this.retrieval = newDocLinks.playground.retrieval;
    this.context = newDocLinks.playground.context;
    this.hiddenFields = newDocLinks.playground.hiddenFields;
  }
}
const docLinks = exports.docLinks = new PlaygroundDocLinks();
