// @kbn/security-solution-plugin/server/lib/detection_engine/prebuilt_rules/logic/rule_assets/prebuilt_rule_assets_type.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.prebuiltRuleAssetType = exports.PREBUILT_RULE_ASSETS_SO_TYPE = void 0;
var _coreSavedObjectsServer = require("@kbn/core-saved-objects-server");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const PREBUILT_RULE_ASSETS_SO_TYPE = exports.PREBUILT_RULE_ASSETS_SO_TYPE = 'security-rule';
const prebuiltRuleAssetMappings = {
  dynamic: false,
  properties: {
    rule_id: {
      type: 'keyword'
    },
    version: {
      type: 'long'
    }
  }
};
const prebuiltRuleAssetType = exports.prebuiltRuleAssetType = {
  name: PREBUILT_RULE_ASSETS_SO_TYPE,
  indexPattern: _coreSavedObjectsServer.SECURITY_SOLUTION_SAVED_OBJECT_INDEX,
  hidden: false,
  management: {
    importableAndExportable: true,
    visibleInManagement: false
  },
  namespaceType: 'agnostic',
  mappings: prebuiltRuleAssetMappings
};
