// @kbn/screenshotting-plugin/server/screenshots/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DEFAULT_PAGELOAD_SELECTOR = exports.CONTEXT_WAITFORRENDER = exports.CONTEXT_WAITFORELEMENTSTOBEINDOM = exports.CONTEXT_SKIPTELEMETRY = exports.CONTEXT_READMETADATA = exports.CONTEXT_INJECTCSS = exports.CONTEXT_GETTIMERANGE = exports.CONTEXT_GETRENDERERRORS = exports.CONTEXT_GETNUMBEROFITEMS = exports.CONTEXT_ELEMENTATTRIBUTES = exports.CONTEXT_DEBUG = void 0;
var _server = require("@kbn/core/server");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DEFAULT_PAGELOAD_SELECTOR = exports.DEFAULT_PAGELOAD_SELECTOR = `.${_server.APP_WRAPPER_CLASS}`;

// FIXME: cleanup: remove this file and use the EventLogger's Actions enum instead
const CONTEXT_GETNUMBEROFITEMS = exports.CONTEXT_GETNUMBEROFITEMS = 'GetNumberOfItems';
const CONTEXT_INJECTCSS = exports.CONTEXT_INJECTCSS = 'InjectCss';
const CONTEXT_WAITFORRENDER = exports.CONTEXT_WAITFORRENDER = 'WaitForRender';
const CONTEXT_GETRENDERERRORS = exports.CONTEXT_GETRENDERERRORS = 'GetVisualisationsRenderErrors';
const CONTEXT_GETTIMERANGE = exports.CONTEXT_GETTIMERANGE = 'GetTimeRange';
const CONTEXT_ELEMENTATTRIBUTES = exports.CONTEXT_ELEMENTATTRIBUTES = 'ElementPositionAndAttributes';
const CONTEXT_WAITFORELEMENTSTOBEINDOM = exports.CONTEXT_WAITFORELEMENTSTOBEINDOM = 'WaitForElementsToBeInDOM';
const CONTEXT_SKIPTELEMETRY = exports.CONTEXT_SKIPTELEMETRY = 'SkipTelemetry';
const CONTEXT_READMETADATA = exports.CONTEXT_READMETADATA = 'ReadVisualizationsMetadata';
const CONTEXT_DEBUG = exports.CONTEXT_DEBUG = 'Debug';
