// @kbn/index-management-plugin/public/application/components/mappings_editor/components/document_fields/document_fields_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentFieldsHeader = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _i18nReact = require("@kbn/i18n-react");
var _documentation = require("../../../../services/documentation");
var _document_fields_search = require("./document_fields_search");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const DocumentFieldsHeader = exports.DocumentFieldsHeader = /*#__PURE__*/_react.default.memo(({
  searchValue,
  onSearchChange
}) => {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    justifyContent: "spaceBetween"
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, null, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
    size: "s",
    color: "subdued"
  }, /*#__PURE__*/_react.default.createElement(_i18nReact.FormattedMessage, {
    id: "xpack.idxMgmt.mappingsEditor.documentFieldsDescription",
    defaultMessage: "Define the fields for your indexed documents. {docsLink}",
    values: {
      docsLink: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
        href: _documentation.documentationService.getMappingTypesLink(),
        target: "_blank",
        external: true
      }, _i18n.i18n.translate('xpack.idxMgmt.mappingsEditor.documentFieldsDocumentationLink', {
        defaultMessage: 'Learn more.'
      }))
    }
  }))), /*#__PURE__*/_react.default.createElement(_document_fields_search.DocumentFieldsSearch, {
    searchValue: searchValue,
    onSearchChange: onSearchChange
  }));
});
