// @kbn/lens-plugin/public/datasources/form_based/dimension_panel/advanced_options.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AdvancedOptions = AdvancedOptions;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
var _react2 = require("@emotion/react");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function AdvancedOptions(props) {
  const {
    euiTheme
  } = (0, _eui.useEuiTheme)();
  return /*#__PURE__*/_react.default.createElement(_eui.EuiAccordion, {
    id: "advancedOptionsAccordion",
    arrowProps: {
      color: 'primary'
    },
    "data-test-subj": "indexPattern-advanced-accordion",
    className: "lnsIndexPatternDimensionEditor-advancedOptions",
    buttonContent: /*#__PURE__*/_react.default.createElement(_eui.EuiTitle, {
      size: "xxs"
    }, /*#__PURE__*/_react.default.createElement("h5", null, /*#__PURE__*/_react.default.createElement(_eui.EuiTextColor, {
      color: euiTheme.colors.primary
    }, _i18n.i18n.translate('xpack.lens.indexPattern.advancedSettings', {
      defaultMessage: 'Advanced'
    })))),
    css: (0, _react2.css)`
        padding: 0 ${euiTheme.size.base} ${euiTheme.size.base};
      `
  }, props.options.map(({
    dataTestSubj,
    inlineElement
  }) => /*#__PURE__*/_react.default.createElement("div", {
    key: dataTestSubj,
    "data-test-subj": dataTestSubj
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "s"
  }), inlineElement)));
}
