// @kbn/lens-plugin/public/datasources/form_based/dimension_panel/dimension_panel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormBasedDimensionEditorComponent = exports.FormBasedDimensionEditor = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _react = _interopRequireWildcard(require("react"));
var _public = require("@kbn/expression-xy-plugin/public");
var _dimension_editor = require("./dimension_editor");
var _operation_support = require("./operation_support");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const FormBasedDimensionEditorComponent = exports.FormBasedDimensionEditorComponent = function FormBasedDimensionPanel(props) {
  var _props$state$layers$l;
  const layerId = props.layerId;
  const currentIndexPattern = props.indexPatterns[(_props$state$layers$l = props.state.layers[layerId]) === null || _props$state$layers$l === void 0 ? void 0 : _props$state$layers$l.indexPatternId];
  if (!currentIndexPattern) {
    return null;
  }
  const operationSupportMatrix = (0, _operation_support.getOperationSupportMatrix)(props);
  const selectedColumn = props.state.layers[layerId].columns[props.columnId] || null;
  return /*#__PURE__*/_react.default.createElement(_dimension_editor.DimensionEditor, (0, _extends2.default)({}, props, {
    layerType: props.layerType || _public.LayerTypes.DATA,
    currentIndexPattern: currentIndexPattern,
    selectedColumn: selectedColumn,
    operationSupportMatrix: operationSupportMatrix
  }));
};
const FormBasedDimensionEditor = exports.FormBasedDimensionEditor = /*#__PURE__*/(0, _react.memo)(FormBasedDimensionEditorComponent);
