// @kbn/infra-plugin/public/pages/metrics/hosts/components/table/column_header.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ColumnHeader = void 0;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
var _react2 = require("@emotion/react");
var _tooltip_content = require("../../../../../components/lens/metric_explanation/tooltip_content");
var _popover = require("../common/popover");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const ColumnHeader = exports.ColumnHeader = /*#__PURE__*/_react.default.memo(({
  label,
  toolTip,
  formula,
  showDocumentationLink = true
}) => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  gutterSize: "xs"
}, /*#__PURE__*/_react.default.createElement("div", {
  css: (0, _react2.css)`
          overflow-wrap: break-word !important;
          word-break: break-word;
          min-width: 0;
          text-overflow: ellipsis;
          overflow: hidden;
        `
}, label), toolTip && /*#__PURE__*/_react.default.createElement(_popover.Popover, null, /*#__PURE__*/_react.default.createElement(_tooltip_content.TooltipContent, {
  formula: formula,
  description: toolTip,
  showDocumentationLink: showDocumentationLink
}))));
