// @kbn/apm-plugin/public/components/shared/technical_preview_badge.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TechnicalPreviewBadge = TechnicalPreviewBadge;
var _eui = require("@elastic/eui");
var _i18n = require("@kbn/i18n");
var _react = _interopRequireDefault(require("react"));
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function TechnicalPreviewBadge({
  icon,
  size,
  style
}) {
  return /*#__PURE__*/_react.default.createElement(_eui.EuiBetaBadge, {
    label: _i18n.i18n.translate('xpack.apm.technicalPreviewBadgeLabel', {
      defaultMessage: 'Technical preview'
    }),
    tooltipContent: _i18n.i18n.translate('xpack.apm.technicalPreviewBadgeDescription', {
      defaultMessage: 'This functionality is in technical preview and may be changed or removed completely in a future release. Elastic will work to fix any issues, but features in technical preview are not subject to the support SLA of official GA features.'
    }),
    iconType: icon,
    size: size,
    style: style
  });
}
