// @kbn/observability-plugin/public/pages/alert_details/components/alert_summary.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.AlertSummary = AlertSummary;
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _eui = require("@elastic/eui");
var _observabilityGetPaddedAlertTimeRangeUtil = require("@kbn/observability-get-padded-alert-time-range-util");
var _ruleDataUtils = require("@kbn/rule-data-utils");
var _i18n = require("@kbn/i18n");
var _groups = require("../../../components/alert_sources/groups");
var _tags = require("../../../components/tags");
var _get_sources = require("../../../components/alert_sources/get_sources");
var _kibana_react = require("../../../utils/kibana_react");
var _paths = require("../../../../common/locators/paths");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

function AlertSummary({
  alert,
  alertSummaryFields
}) {
  const {
    http
  } = (0, _kibana_react.useKibana)().services;
  const [timeRange, setTimeRange] = (0, _react.useState)({
    from: 'now-15m',
    to: 'now'
  });
  const alertStart = alert.fields[_ruleDataUtils.ALERT_START];
  const alertEnd = alert.fields[_ruleDataUtils.ALERT_END];
  const ruleName = alert.fields[_ruleDataUtils.ALERT_RULE_NAME];
  const ruleId = alert.fields[_ruleDataUtils.ALERT_RULE_UUID];
  const tags = alert.fields[_ruleDataUtils.TAGS];
  const ruleLink = http.basePath.prepend(_paths.paths.observability.ruleDetails(ruleId));
  const commonFieldsAtStart = [];
  const commonFieldsAtEnd = [];
  const groups = (0, _get_sources.getSources)(alert);
  (0, _react.useEffect)(() => {
    setTimeRange((0, _observabilityGetPaddedAlertTimeRangeUtil.getPaddedAlertTimeRange)(alertStart, alertEnd));
  }, [alertStart, alertEnd]);
  if (groups && groups.length > 0) {
    commonFieldsAtStart.push({
      label: _i18n.i18n.translate('xpack.observability.alertDetails.alertSummaryField.source', {
        defaultMessage: 'Source'
      }),
      value: /*#__PURE__*/_react.default.createElement(_groups.Groups, {
        groups: groups,
        timeRange: alertEnd ? timeRange : {
          ...timeRange,
          to: 'now'
        }
      })
    });
  }
  if (tags && tags.length > 0) {
    commonFieldsAtEnd.push({
      label: _i18n.i18n.translate('xpack.observability.alertDetails.alertSummaryField.tags', {
        defaultMessage: 'Tags'
      }),
      value: /*#__PURE__*/_react.default.createElement(_tags.Tags, {
        tags: tags
      })
    });
  }
  commonFieldsAtEnd.push({
    label: _i18n.i18n.translate('xpack.observability.alertDetails.alertSummaryField.rule', {
      defaultMessage: 'Rule'
    }),
    value: /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
      "data-test-subj": "o11yAlertRuleLink",
      href: ruleLink
    }, ruleName)
  });
  const alertSummary = [...commonFieldsAtStart, ...(alertSummaryFields !== null && alertSummaryFields !== void 0 ? alertSummaryFields : []), ...commonFieldsAtEnd];
  return /*#__PURE__*/_react.default.createElement("div", {
    "data-test-subj": "alert-summary-container"
  }, alertSummary && alertSummary.length > 0 && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
    gutterSize: "xl"
  }, alertSummary.map((field, idx) => {
    return /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
      key: idx,
      grow: false
    }, /*#__PURE__*/_react.default.createElement(_eui.EuiText, {
      color: "subdued"
    }, field.label), /*#__PURE__*/_react.default.createElement(_eui.EuiText, null, field.value));
  })), /*#__PURE__*/_react.default.createElement(_eui.EuiSpacer, {
    size: "l"
  })));
}

// eslint-disable-next-line import/no-default-export
var _default = exports.default = AlertSummary;
