// @kbn/expression-shape-plugin/common/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.SVG = exports.PLUGIN_NAME = exports.PLUGIN_ID = exports.FONT_WEIGHT = exports.FONT_FAMILY = exports.CSS = exports.BOOLEAN_TRUE = exports.BOOLEAN_FALSE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

const PLUGIN_ID = exports.PLUGIN_ID = 'expressionShape';
const PLUGIN_NAME = exports.PLUGIN_NAME = 'expressionShape';
const SVG = exports.SVG = 'SVG';
const CSS = exports.CSS = 'CSS';
const FONT_FAMILY = exports.FONT_FAMILY = '`font-family`';
const FONT_WEIGHT = exports.FONT_WEIGHT = '`font-weight`';
const BOOLEAN_TRUE = exports.BOOLEAN_TRUE = '`true`';
const BOOLEAN_FALSE = exports.BOOLEAN_FALSE = '`false`';
