// @kbn/canvas-plugin/public/components/home/my_workpads/empty_prompt.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.HomeEmptyPrompt = void 0;
var _react = _interopRequireWildcard(require("react"));
var _i18n = require("@kbn/i18n");
var _eui = require("@elastic/eui");
var _constants = require("../../../../i18n/constants");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function (e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const HomeEmptyPrompt = () => /*#__PURE__*/_react.default.createElement(_eui.EuiFlexGroup, {
  justifyContent: "spaceAround",
  alignItems: "center",
  style: {
    minHeight: 600
  }
}, /*#__PURE__*/_react.default.createElement(_eui.EuiFlexItem, {
  grow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiPanel, {
  color: "subdued",
  borderRadius: "none",
  hasShadow: false
}, /*#__PURE__*/_react.default.createElement(_eui.EuiEmptyPrompt, {
  color: "subdued",
  iconType: "importAction",
  title: /*#__PURE__*/_react.default.createElement("h2", null, strings.getEmptyPromptTitle()),
  titleSize: "m",
  body: /*#__PURE__*/_react.default.createElement(_react.Fragment, null, /*#__PURE__*/_react.default.createElement("p", null, strings.getEmptyPromptGettingStartedDescription()), /*#__PURE__*/_react.default.createElement("p", null, strings.getEmptyPromptNewUserDescription(), ' ', /*#__PURE__*/_react.default.createElement(_eui.EuiLink, {
    href: "home#/tutorial_directory/sampleData"
  }, strings.getSampleDataLinkLabel()), "."))
}))));
exports.HomeEmptyPrompt = HomeEmptyPrompt;
const strings = {
  getEmptyPromptGettingStartedDescription: () => _i18n.i18n.translate('xpack.canvas.homeEmptyPrompt.emptyPromptGettingStartedDescription', {
    defaultMessage: 'Create a new workpad, start from a template, or import a workpad {JSON} file by dropping it here.',
    values: {
      JSON: _constants.JSON
    }
  }),
  getEmptyPromptNewUserDescription: () => _i18n.i18n.translate('xpack.canvas.homeEmptyPrompt.emptyPromptNewUserDescription', {
    defaultMessage: 'New to {CANVAS}?',
    values: {
      CANVAS: _constants.CANVAS
    }
  }),
  getEmptyPromptTitle: () => _i18n.i18n.translate('xpack.canvas.homeEmptyPrompt.emptyPromptTitle', {
    defaultMessage: 'Add your first workpad'
  }),
  getSampleDataLinkLabel: () => _i18n.i18n.translate('xpack.canvas.homeEmptyPrompt.sampleDataLinkLabel', {
    defaultMessage: 'Add your first workpad'
  })
};
