// @kbn/index-management-plugin/common/constants/index_statuses.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.INDEX_REFRESHING = exports.INDEX_OPENING = exports.INDEX_OPEN = exports.INDEX_MERGING = exports.INDEX_FORCEMERGING = exports.INDEX_FLUSHING = exports.INDEX_CLOSING = exports.INDEX_CLOSED = exports.INDEX_CLEARING_CACHE = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const INDEX_CLOSED = exports.INDEX_CLOSED = 'close';
const INDEX_OPEN = exports.INDEX_OPEN = 'open';
const INDEX_CLOSING = exports.INDEX_CLOSING = 'closing';
const INDEX_OPENING = exports.INDEX_OPENING = 'opening';
const INDEX_REFRESHING = exports.INDEX_REFRESHING = 'refreshing';
const INDEX_FLUSHING = exports.INDEX_FLUSHING = 'flushing';
const INDEX_FORCEMERGING = exports.INDEX_FORCEMERGING = 'forcemerging';
const INDEX_MERGING = exports.INDEX_MERGING = 'merging';
const INDEX_CLEARING_CACHE = exports.INDEX_CLEARING_CACHE = 'clearing cache';
