// @kbn/index-management-plugin/common/constants/ui_metric.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.UIM_UPDATE_SETTINGS = exports.UIM_TEMPLATE_UPDATE = exports.UIM_TEMPLATE_SIMULATE = exports.UIM_TEMPLATE_SHOW_DETAILS_CLICK = exports.UIM_TEMPLATE_LIST_LOAD = exports.UIM_TEMPLATE_DETAIL_PANEL_SUMMARY_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_SETTINGS_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_PREVIEW_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_MAPPINGS_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_ALIASES_TAB = exports.UIM_TEMPLATE_DELETE_MANY = exports.UIM_TEMPLATE_DELETE = exports.UIM_TEMPLATE_CREATE = exports.UIM_TEMPLATE_CLONE = exports.UIM_INDEX_UNFREEZE_MANY = exports.UIM_INDEX_UNFREEZE = exports.UIM_INDEX_SETTINGS_EDIT = exports.UIM_INDEX_REFRESH_MANY = exports.UIM_INDEX_REFRESH = exports.UIM_INDEX_OPEN_MANY = exports.UIM_INDEX_OPEN = exports.UIM_INDEX_FORCE_MERGE_MANY = exports.UIM_INDEX_FORCE_MERGE = exports.UIM_INDEX_FLUSH_MANY = exports.UIM_INDEX_FLUSH = exports.UIM_INDEX_DELETE_MANY = exports.UIM_INDEX_DELETE = exports.UIM_INDEX_CLOSE_MANY = exports.UIM_INDEX_CLOSE = exports.UIM_INDEX_CLEAR_CACHE_MANY = exports.UIM_INDEX_CLEAR_CACHE = exports.UIM_APP_NAME = exports.UIM_APP_LOAD = void 0;
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0; you may not use this file except in compliance with the Elastic License
 * 2.0.
 */

const UIM_APP_NAME = exports.UIM_APP_NAME = 'index_management';
const UIM_APP_LOAD = exports.UIM_APP_LOAD = 'app_load';
const UIM_UPDATE_SETTINGS = exports.UIM_UPDATE_SETTINGS = 'update_settings';
const UIM_INDEX_CLEAR_CACHE = exports.UIM_INDEX_CLEAR_CACHE = 'index_clear_cache';
const UIM_INDEX_CLEAR_CACHE_MANY = exports.UIM_INDEX_CLEAR_CACHE_MANY = 'index_clear_cache_many';
const UIM_INDEX_CLOSE = exports.UIM_INDEX_CLOSE = 'index_close';
const UIM_INDEX_CLOSE_MANY = exports.UIM_INDEX_CLOSE_MANY = 'index_close_many';
const UIM_INDEX_DELETE = exports.UIM_INDEX_DELETE = 'index_delete';
const UIM_INDEX_DELETE_MANY = exports.UIM_INDEX_DELETE_MANY = 'index_delete_many';
const UIM_INDEX_FLUSH = exports.UIM_INDEX_FLUSH = 'index_flush';
const UIM_INDEX_FLUSH_MANY = exports.UIM_INDEX_FLUSH_MANY = 'index_flush_many';
const UIM_INDEX_FORCE_MERGE = exports.UIM_INDEX_FORCE_MERGE = 'index_force_merge';
const UIM_INDEX_FORCE_MERGE_MANY = exports.UIM_INDEX_FORCE_MERGE_MANY = 'index_force_merge_many';
const UIM_INDEX_OPEN = exports.UIM_INDEX_OPEN = 'index_open';
const UIM_INDEX_OPEN_MANY = exports.UIM_INDEX_OPEN_MANY = 'index_open_many';
const UIM_INDEX_REFRESH = exports.UIM_INDEX_REFRESH = 'index_refresh';
const UIM_INDEX_REFRESH_MANY = exports.UIM_INDEX_REFRESH_MANY = 'index_refresh_many';
const UIM_INDEX_SETTINGS_EDIT = exports.UIM_INDEX_SETTINGS_EDIT = 'index_settings_edit';
const UIM_INDEX_UNFREEZE = exports.UIM_INDEX_UNFREEZE = 'index_unfreeze';
const UIM_INDEX_UNFREEZE_MANY = exports.UIM_INDEX_UNFREEZE_MANY = 'index_unfreeze_many';
const UIM_TEMPLATE_LIST_LOAD = exports.UIM_TEMPLATE_LIST_LOAD = 'template_list_load';
const UIM_TEMPLATE_DELETE = exports.UIM_TEMPLATE_DELETE = 'template_delete';
const UIM_TEMPLATE_DELETE_MANY = exports.UIM_TEMPLATE_DELETE_MANY = 'template_delete_many';
const UIM_TEMPLATE_SHOW_DETAILS_CLICK = exports.UIM_TEMPLATE_SHOW_DETAILS_CLICK = 'template_show_details_click';
const UIM_TEMPLATE_DETAIL_PANEL_SUMMARY_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_SUMMARY_TAB = 'template_details_summary_tab';
const UIM_TEMPLATE_DETAIL_PANEL_SETTINGS_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_SETTINGS_TAB = 'template_details_settings_tab';
const UIM_TEMPLATE_DETAIL_PANEL_MAPPINGS_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_MAPPINGS_TAB = 'template_details_mappings_tab';
const UIM_TEMPLATE_DETAIL_PANEL_ALIASES_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_ALIASES_TAB = 'template_details_aliases_tab';
const UIM_TEMPLATE_DETAIL_PANEL_PREVIEW_TAB = exports.UIM_TEMPLATE_DETAIL_PANEL_PREVIEW_TAB = 'template_details_preview_tab';
const UIM_TEMPLATE_CREATE = exports.UIM_TEMPLATE_CREATE = 'template_create';
const UIM_TEMPLATE_UPDATE = exports.UIM_TEMPLATE_UPDATE = 'template_update';
const UIM_TEMPLATE_CLONE = exports.UIM_TEMPLATE_CLONE = 'template_clone';
const UIM_TEMPLATE_SIMULATE = exports.UIM_TEMPLATE_SIMULATE = 'template_simulate';
