// @kbn/ml-plugin/public/application/model_management/test_models/models/text_classification/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FillMaskInference", {
  enumerable: true,
  get: function () {
    return _fill_mask_inference.FillMaskInference;
  }
});
Object.defineProperty(exports, "LangIdentInference", {
  enumerable: true,
  get: function () {
    return _lang_ident_inference.LangIdentInference;
  }
});
Object.defineProperty(exports, "TextClassificationInference", {
  enumerable: true,
  get: function () {
    return _text_classification_inference.TextClassificationInference;
  }
});
Object.defineProperty(exports, "ZeroShotClassificationInference", {
  enumerable: true,
  get: function () {
    return _zero_shot_classification_inference.ZeroShotClassificationInference;
  }
});
Object.defineProperty(exports, "getFillMaskOutputComponent", {
  enumerable: true,
  get: function () {
    return _fill_mask_output.getFillMaskOutputComponent;
  }
});
Object.defineProperty(exports, "getLangIdentOutputComponent", {
  enumerable: true,
  get: function () {
    return _lang_ident_output.getLangIdentOutputComponent;
  }
});
Object.defineProperty(exports, "getTextClassificationOutputComponent", {
  enumerable: true,
  get: function () {
    return _text_classification_output.getTextClassificationOutputComponent;
  }
});
Object.defineProperty(exports, "getZeroShotClassificationInput", {
  enumerable: true,
  get: function () {
    return _zero_shot_classification_input.getZeroShotClassificationInput;
  }
});
var _text_classification_inference = require("./text_classification_inference");
var _text_classification_output = require("./text_classification_output");
var _zero_shot_classification_inference = require("./zero_shot_classification_inference");
var _zero_shot_classification_input = require("./zero_shot_classification_input");
var _fill_mask_inference = require("./fill_mask_inference");
var _fill_mask_output = require("./fill_mask_output");
var _lang_ident_inference = require("./lang_ident_inference");
var _lang_ident_output = require("./lang_ident_output");
