// @kbn/input-control-vis-plugin/public/components/vis/form_row.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormRow = FormRow;
var _react = _interopRequireDefault(require("react"));
var _eui = require("@elastic/eui");
/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License
 * 2.0 and the Server Side Public License, v 1; you may not use this file except
 * in compliance with, at your election, the Elastic License 2.0 or the Server
 * Side Public License, v 1.
 */

function FormRow(props) {
  let control = props.children;
  if (props.disableMsg) {
    control = /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
      position: "top",
      content: props.disableMsg,
      anchorClassName: "eui-displayBlock"
    }, control);
  }
  const label = props.warningMsg ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_eui.EuiToolTip, {
    position: "top",
    content: props.warningMsg
  }, /*#__PURE__*/_react.default.createElement(_eui.EuiIcon, {
    type: "warning"
  })), props.label) : props.label;
  return /*#__PURE__*/_react.default.createElement(_eui.EuiFormRow, {
    label: label,
    id: props.id,
    "data-test-subj": 'inputControl' + props.controlIndex
  }, control);
}
